// tdesign-react/cjs/select/base/Select.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var toConsumableArray = require('../../_chunks/dep-1d0dbfc5.js');
var objectWithoutProperties = require('../../_chunks/dep-1c586f5c.js');
var defineProperty = require('../../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var _util_forwardRefWithStatics = require('../../_util/forwardRefWithStatics.js');
var _util_helper = require('../../_util/helper.js');
var _util_noop = require('../../_util/noop.js');
var _util_parseTNode = require('../../_util/parseTNode.js');
var _util_ref = require('../../_util/ref.js');
var common_FakeArrow = require('../../common/FakeArrow.js');
var hooks_useConfig = require('../../hooks/useConfig.js');
var hooks_useControlled = require('../../hooks/useControlled.js');
var hooks_useDefaultProps = require('../../hooks/useDefaultProps.js');
var loading_index = require('../../loading/index.js');
var locale_LocalReceiver = require('../../locale/LocalReceiver.js');
var selectInput_index = require('../../select-input/index.js');
var tag_index = require('../../tag/index.js');
var select_defaultProps = require('../defaultProps.js');
var select_hooks_useKeyboardControl = require('../hooks/useKeyboardControl.js');
var select_base_Option = require('../../_chunks/dep-bdc97672.js');
var select_base_OptionGroup = require('./OptionGroup.js');
var select_base_PopupContent = require('./PopupContent.js');
var get = require('../../_chunks/dep-05df1ea3.js');
var isFunction = require('../../_chunks/dep-cba78412.js');
var debounce = require('../../_chunks/dep-d3af8eb7.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_chunks/dep-3c2d5e8c.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-32593c54.js');
require('hoist-non-react-statics');
require('../../_chunks/dep-e44b03ec.js');
require('../../_chunks/dep-b4442734.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-876d090e.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-eb1b216c.js');
require('../../_chunks/dep-b529a0d7.js');
require('../../_chunks/dep-bcd16957.js');
require('../../_chunks/dep-b289b0be.js');
require('react-is');
require('../../_util/isFragment.js');
require('../../_chunks/dep-b011b5d3.js');
require('../../_chunks/dep-f0e884b9.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-14013769.js');
require('../../_chunks/dep-06bf4625.js');
require('dayjs');
require('../../_chunks/dep-043eee20.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../_chunks/dep-2d5cf7d6.js');
require('../../loading/Loading.js');
require('../../_chunks/dep-68f5fe7b.js');
require('../../common/Portal.js');
require('react-dom');
require('../../hooks/useLayoutEffect.js');
require('../../loading/defaultProps.js');
require('../../loading/gradient.js');
require('../../_chunks/dep-064e4dba.js');
require('../../hooks/useDomRefCallback.js');
require('../../loading/plugin.js');
require('../../_util/react-render.js');
require('../../_chunks/dep-e37cb830.js');
require('../../common/PluginContainer.js');
require('../../config-provider/ConfigProvider.js');
require('../../_chunks/dep-4a562e04.js');
require('../../_chunks/dep-a5b4810f.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-ea361d31.js');
require('../../select-input/SelectInput.js');
require('../../popup/index.js');
require('../../popup/Popup.js');
require('react-transition-group');
require('../../hooks/useAnimation.js');
require('../../hooks/useAttach.js');
require('../../hooks/useMutationObserver.js');
require('../../hooks/useLatest.js');
require('../../_chunks/dep-4e876c4b.js');
require('../../_chunks/dep-df8d2fc8.js');
require('../../_chunks/dep-66f91e5d.js');
require('../../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../../hooks/useWindowSize.js');
require('../../popup/defaultProps.js');
require('../../popup/hooks/useTrigger.js');
require('../../_util/listener.js');
require('../../popup/utils/transition.js');
require('../../_chunks/dep-eab994e8.js');
require('../../popup/PopupPlugin.js');
require('../../select-input/defaultProps.js');
require('../../select-input/useMultiple.js');
require('../../tag-input/index.js');
require('../../tag-input/TagInput.js');
require('tdesign-icons-react');
require('../../hooks/useDragSorter.js');
require('../../hooks/useEventCallback.js');
require('../../hooks/useGlobalIcon.js');
require('../../input/index.js');
require('../../input/Input.js');
require('../../input/defaultProps.js');
require('../../input/InputGroup.js');
require('../../input/useLengthLimit.js');
require('../../tag-input/defaultProps.js');
require('../../tag-input/useHover.js');
require('../../tag-input/useTagList.js');
require('../../tag-input/useTagScroll.js');
require('../../tag/Tag.js');
require('../../_chunks/dep-d14bfcfa.js');
require('../../tag/defaultProps.js');
require('../../tag/CheckTag.js');
require('../../_chunks/dep-dd0f5f6b.js');
require('../../tag/CheckTagGroup.js');
require('../../select-input/useOverlayInnerStyle.js');
require('../../hooks/useInnerPopupVisible.js');
require('../../select-input/useSingle.js');
require('../../_chunks/dep-f7bf3d17.js');
require('../../_chunks/dep-5e17a073.js');
require('../../_chunks/dep-ca918872.js');
require('../../_chunks/dep-13de2fef.js');
require('../../_chunks/dep-63308b84.js');
require('../../_chunks/dep-048e4314.js');
require('../../hooks/useRipple.js');
require('../hooks/usePanelVirtualScroll.js');
require('../../hooks/useVirtualScroll.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["overlayClassName", "onScroll", "onScrollToBottom"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Select = _util_forwardRefWithStatics["default"](function (originalProps, ref) {
  var _props$tagInputProps;
  var props = hooks_useDefaultProps["default"](originalProps, select_defaultProps.selectDefaultProps);
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("select"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var emptyText = t(local.loadingText);
  var borderless = props.borderless,
    autoWidth = props.autoWidth,
    creatable = props.creatable,
    _props$loadingText = props.loadingText,
    loadingText = _props$loadingText === void 0 ? emptyText : _props$loadingText,
    max = props.max,
    popupProps = props.popupProps,
    reserveKeyword = props.reserveKeyword,
    className = props.className,
    style = props.style,
    disabled = props.disabled,
    size = props.size,
    multiple = props.multiple,
    placeholder = props.placeholder,
    clearable = props.clearable,
    prefixIcon = props.prefixIcon,
    options = props.options,
    filterable = props.filterable,
    loading = props.loading,
    empty = props.empty,
    valueType = props.valueType,
    keys = props.keys,
    children = props.children,
    collapsedItems = props.collapsedItems,
    minCollapsedNum = props.minCollapsedNum,
    valueDisplay = props.valueDisplay,
    showArrow = props.showArrow,
    inputProps = props.inputProps,
    panelBottomContent = props.panelBottomContent,
    panelTopContent = props.panelTopContent,
    selectInputProps = props.selectInputProps,
    tagInputProps = props.tagInputProps,
    scroll = props.scroll,
    suffixIcon = props.suffixIcon,
    label = props.label,
    filter = props.filter,
    onFocus = props.onFocus,
    _onBlur = props.onBlur,
    _props$onClear = props.onClear,
    onClear = _props$onClear === void 0 ? _util_noop["default"] : _props$onClear,
    onCreate = props.onCreate,
    onRemove = props.onRemove,
    onSearch = props.onSearch,
    onEnter = props.onEnter,
    onPopupVisibleChange = props.onPopupVisibleChange;
  var readOnly = props.readOnly || props.readonly;
  var tagProps = _objectSpread(_objectSpread({}, props.tagProps), (_props$tagInputProps = props.tagInputProps) === null || _props$tagInputProps === void 0 ? void 0 : _props$tagInputProps.tagProps);
  var _useControlled = hooks_useControlled["default"](props, "value", props.onChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    value = _useControlled2[0],
    onChange = _useControlled2[1];
  var _useMemo = React.useMemo(function () {
      return select_base_Option.getKeyMapping(keys);
    }, [keys]),
    valueKey = _useMemo.valueKey,
    labelKey = _useMemo.labelKey,
    disabledKey = _useMemo.disabledKey;
  var selectInputRef = React.useRef(null);
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _ref = popupProps || {},
    overlayClassName = _ref.overlayClassName,
    onScroll = _ref.onScroll,
    onScrollToBottom = _ref.onScrollToBottom,
    restPopupProps = objectWithoutProperties._objectWithoutProperties(_ref, _excluded);
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    isScrolling = _useState2[0],
    toggleIsScrolling = _useState2[1];
  var name = "".concat(classPrefix, "-select");
  var _useControlled3 = hooks_useControlled["default"](props, "inputValue", props.onInputChange),
    _useControlled4 = slicedToArray._slicedToArray(_useControlled3, 2),
    inputValue = _useControlled4[0],
    onInputChange = _useControlled4[1];
  var _useControlled5 = hooks_useControlled["default"](props, "popupVisible", onPopupVisibleChange),
    _useControlled6 = slicedToArray._slicedToArray(_useControlled5, 2),
    innerPopupVisible = _useControlled6[0],
    setInnerPopupVisible = _useControlled6[1];
  var handlePopupVisibleChange = function handlePopupVisibleChange(visible, ctx) {
    if (disabled) return;
    visible ? toggleIsScrolling(false) : onInputChange("", {
      trigger: "blur"
    });
    setInnerPopupVisible(visible, ctx);
  };
  var _useOptions = select_base_Option.useOptions(keys, options, children, valueType, value, reserveKeyword),
    currentOptions = _useOptions.currentOptions,
    setCurrentOptions = _useOptions.setCurrentOptions,
    tmpPropOptions = _useOptions.tmpPropOptions,
    valueToOption = _useOptions.valueToOption,
    selectedOptions = _useOptions.selectedOptions,
    flattenedOptions = _useOptions.flattenedOptions;
  var onCheckAllChange = React.useCallback(function (checkAll, e) {
    var isDisabledCheckAll = function isDisabledCheckAll(opt) {
      return opt.checkAll && opt.disabled;
    };
    if (!multiple || currentOptions.some(function (opt) {
      return !select_base_Option.isSelectOptionGroup(opt) && isDisabledCheckAll(opt);
    })) return;
    var isObjectType = valueType === "object";
    var enabledOptions = [];
    currentOptions.forEach(function (option) {
      if (select_base_Option.isSelectOptionGroup(option)) {
        var _option$children;
        (_option$children = option.children) === null || _option$children === void 0 || _option$children.forEach(function (item) {
          if (!item.checkAll && !item.disabled) {
            enabledOptions.push(item);
          }
        });
      } else {
        !option.checkAll && !option.disabled && enabledOptions.push(option);
      }
    });
    var currentValues = Array.isArray(value) ? value : [];
    var disabledSelectedOptions = [];
    var isDisabledAndSelected = function isDisabledAndSelected(opt) {
      if (opt.checkAll || !opt.disabled) return false;
      if (isObjectType) return currentValues.some(function (v) {
        return get.get(v, valueKey) === opt[valueKey];
      });
      return currentValues.includes(opt[valueKey]);
    };
    currentOptions.forEach(function (opt) {
      if (select_base_Option.isSelectOptionGroup(opt)) {
        var _opt$children;
        (_opt$children = opt.children) === null || _opt$children === void 0 || _opt$children.forEach(function (item) {
          if (isDisabledAndSelected(item)) {
            disabledSelectedOptions.push(item);
          }
        });
      } else if (isDisabledAndSelected(opt)) {
        disabledSelectedOptions.push(opt);
      }
    });
    var checkAllValue;
    if (checkAll) {
      var enabledValues = enabledOptions.map(function (opt) {
        return isObjectType ? opt : opt[valueKey];
      });
      var disabledValues = disabledSelectedOptions.map(function (opt) {
        return isObjectType ? opt : opt[valueKey];
      });
      checkAllValue = [].concat(toConsumableArray._toConsumableArray(disabledValues), toConsumableArray._toConsumableArray(enabledValues));
    } else {
      checkAllValue = disabledSelectedOptions.map(function (opt) {
        return isObjectType ? opt : opt[valueKey];
      });
    }
    var _getSelectedOptions = select_base_Option.getSelectedOptions(checkAllValue, multiple, valueType, keys, valueToOption),
      currentSelectedOptions = _getSelectedOptions.currentSelectedOptions;
    onChange === null || onChange === void 0 || onChange(checkAllValue, {
      e: e,
      trigger: checkAll ? "check" : "uncheck",
      selectedOptions: currentSelectedOptions
    });
  }, [currentOptions, keys, multiple, value, valueKey, valueToOption, valueType, onChange]);
  var selectedLabel = React.useMemo(function () {
    if (multiple) {
      return selectedOptions.map(function (selectedOption) {
        return get.get(selectedOption || {}, labelKey) || "";
      });
    }
    return get.get(selectedOptions[0] || {}, labelKey) || void 0;
  }, [multiple, selectedOptions, labelKey]);
  var onTagChange = function onTagChange(_currentTags, context) {
    var handleRemove = function handleRemove(removeIndex, trigger2, e2, label2) {
      var values = select_base_Option.getSelectValueArr(value, value[removeIndex], true, valueType, keys);
      var _getSelectedOptions2 = select_base_Option.getSelectedOptions(values, multiple, valueType, keys, valueToOption),
        currentSelectedOptions = _getSelectedOptions2.currentSelectedOptions;
      onChange(values, {
        e: e2,
        trigger: trigger2,
        selectedOptions: currentSelectedOptions
      });
      onRemove === null || onRemove === void 0 || onRemove({
        value: value[removeIndex],
        data: {
          label: label2,
          value: value[removeIndex]
        },
        e: e2
      });
    };
    var trigger = context.trigger,
      index = context.index,
      item = context.item,
      e = context.e;
    e === null || e === void 0 || e.stopPropagation();
    if (trigger === "backspace") {
      var closest = -1;
      var len = index;
      while (len >= 0) {
        var option = selectedOptions[len];
        if (!select_base_Option.isSelectOptionGroup(option) && !option.disabled) {
          closest = len;
          break;
        }
        len -= 1;
      }
      if (closest >= 0) {
        var label2 = get.get(selectedOptions[closest], select_base_Option.getKeyMapping(keys).labelKey);
        handleRemove(closest, trigger, e, label2);
      }
      return;
    }
    if (trigger === "tag-remove") {
      var _e$stopPropagation;
      e === null || e === void 0 || (_e$stopPropagation = e.stopPropagation) === null || _e$stopPropagation === void 0 || _e$stopPropagation.call(e);
      handleRemove(index, trigger, e, item);
    }
  };
  var handleChange = function handleChange(value2, context) {
    var selectedValue = multiple ? context.value : value2;
    if (multiple) {
      !reserveKeyword && inputValue && onInputChange("", {
        e: context.e,
        trigger: "change"
      });
    }
    if (creatable && isFunction.isFunction(onCreate)) {
      if (options.filter(function (option) {
        return option.value === selectedValue;
      }).length === 0) {
        onCreate(selectedValue);
      }
    }
    var _getSelectedOptions3 = select_base_Option.getSelectedOptions(value2, multiple, valueType, keys, valueToOption, selectedValue),
      currentSelectedOptions = _getSelectedOptions3.currentSelectedOptions,
      currentOption = _getSelectedOptions3.currentOption;
    onChange === null || onChange === void 0 || onChange(value2, {
      e: context.e,
      trigger: context.trigger,
      selectedOptions: currentSelectedOptions,
      option: currentOption
    });
    if (multiple && (context === null || context === void 0 ? void 0 : context.trigger) === "uncheck" && isFunction.isFunction(onRemove)) {
      var value3 = context === null || context === void 0 ? void 0 : context.value;
      var option = options === null || options === void 0 ? void 0 : options.find(function (option2) {
        return option2.value === value3;
      });
      onRemove({
        value: value3,
        data: option,
        e: context.e
      });
    }
  };
  var _useKeyboardControl = select_hooks_useKeyboardControl["default"]({
      displayOptions: flattenedOptions,
      keys: keys,
      innerPopupVisible: innerPopupVisible,
      max: max,
      multiple: multiple,
      selectInputRef: selectInputRef,
      value: value,
      valueType: valueType,
      handlePopupVisibleChange: handlePopupVisibleChange,
      handleChange: handleChange,
      onCheckAllChange: onCheckAllChange,
      toggleIsScrolling: toggleIsScrolling
    }),
    hoverIndex = _useKeyboardControl.hoverIndex,
    handleKeyDown = _useKeyboardControl.handleKeyDown;
  var handleFilter = function handleFilter(value2) {
    var filteredOptions = [];
    if (filterable && isFunction.isFunction(onSearch)) {
      return;
    }
    if (!value2) {
      setCurrentOptions(tmpPropOptions);
      return;
    }
    var filterLabels = [];
    var filterMethods = function filterMethods(option) {
      if (filter && isFunction.isFunction(filter)) {
        return filter(value2, option);
      }
      var upperValue = value2.toUpperCase();
      return ((option === null || option === void 0 ? void 0 : option.label) || "").toUpperCase().includes(upperValue);
    };
    tmpPropOptions === null || tmpPropOptions === void 0 || tmpPropOptions.forEach(function (option) {
      if (select_base_Option.isSelectOptionGroup(option)) {
        var _option$children2;
        filteredOptions.push(_objectSpread(_objectSpread({}, option), {}, {
          children: (_option$children2 = option.children) === null || _option$children2 === void 0 ? void 0 : _option$children2.filter(function (child) {
            if (filterMethods(child)) {
              filterLabels.push(child.label);
              return true;
            }
            return false;
          })
        }));
      } else if (filterMethods(option)) {
        filterLabels.push(option.label);
        filteredOptions.push(option);
      }
    });
    var isSameLabelOptionExist = filterLabels.includes(value2);
    if (creatable && !isSameLabelOptionExist) {
      filteredOptions = filteredOptions.concat([{
        label: value2,
        value: value2
      }]);
    }
    setCurrentOptions(filteredOptions);
  };
  var handleInputChange = function handleInputChange(value2, context) {
    if (context.trigger !== "clear") {
      onInputChange(value2, {
        e: context.e,
        trigger: "input"
      });
    }
    if (value2 === void 0) {
      return;
    }
    if (isFunction.isFunction(onSearch)) {
      onSearch(value2, {
        e: context.e
      });
      return;
    }
  };
  var handleClear = function handleClear(context) {
    context.e.stopPropagation();
    if (Array.isArray(value)) {
      onChange([], _objectSpread(_objectSpread({}, context), {}, {
        trigger: "clear",
        selectedOptions: []
      }));
    } else {
      onChange(null, _objectSpread(_objectSpread({}, context), {}, {
        trigger: "clear",
        selectedOptions: []
      }));
    }
    onClear(context);
  };
  React.useEffect(function () {
    if (typeof inputValue !== "undefined") {
      handleFilter(String(inputValue));
    }
  }, [inputValue, tmpPropOptions]);
  var renderSuffixIcon = function renderSuffixIcon() {
    if (suffixIcon) {
      return suffixIcon;
    }
    if (loading) {
      return /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, {
        className: classNames__default["default"]("".concat(name, "__right-icon"), "".concat(name, "__active-icon")),
        loading: true,
        size: "small"
      });
    }
    return showArrow && /* @__PURE__ */React__default["default"].createElement(common_FakeArrow["default"], {
      className: "".concat(name, "__right-icon"),
      isActive: innerPopupVisible,
      disabled: disabled
    });
  };
  var getPopupInstance = React.useCallback(function () {
    var _selectInputRef$curre;
    return (_selectInputRef$curre = selectInputRef.current) === null || _selectInputRef$curre === void 0 ? void 0 : _selectInputRef$curre.getPopupContentElement();
  }, []);
  var childrenWithProps = React.Children.map(children, function (child) {
    if (/*#__PURE__*/React.isValidElement(child)) {
      var addedProps = {
        multiple: multiple
      };
      return /*#__PURE__*/React.cloneElement(child, _objectSpread({}, addedProps));
    }
    return child;
  });
  var renderContent = function renderContent() {
    var popupContentProps = {
      onChange: handleChange,
      value: value,
      className: className,
      size: size,
      multiple: multiple,
      showPopup: innerPopupVisible,
      setShowPopup: function setShowPopup(show) {
        return handlePopupVisibleChange(show, {});
      },
      options: currentOptions,
      empty: empty,
      max: max,
      loadingText: loadingText,
      loading: loading,
      valueType: valueType,
      keys: keys,
      panelBottomContent: panelBottomContent,
      panelTopContent: panelTopContent,
      onCheckAllChange: onCheckAllChange,
      getPopupInstance: getPopupInstance,
      scroll: scroll,
      hoverIndex: hoverIndex
    };
    return /* @__PURE__ */React__default["default"].createElement(select_base_PopupContent["default"], _objectSpread({}, popupContentProps), childrenWithProps);
  };
  var renderValueDisplay = React.useMemo(function () {
    if (!valueDisplay) {
      if (!multiple) {
        if (typeof selectedLabel !== "string") {
          return selectedLabel;
        }
        return "";
      }
      return function (_ref2) {
        var val = _ref2.value;
        return val.slice(0, minCollapsedNum ? minCollapsedNum : val.length).map(function (_, index) {
          var targetVal = get.get(selectedOptions[index], valueKey);
          var targetLabel = get.get(selectedOptions[index], labelKey);
          var targetOption = valueToOption[targetVal];
          return /* @__PURE__ */React__default["default"].createElement(tag_index.Tag, _objectSpread(_objectSpread({
            key: index,
            closable: !get.get(targetOption, disabledKey) && !disabled && !readOnly,
            size: size
          }, tagProps), {}, {
            onClose: function onClose(_ref3) {
              var _e$nativeEvent, _e$nativeEvent$stopIm, _tagProps$onClose;
              var e = _ref3.e;
              e.stopPropagation();
              e === null || e === void 0 || (_e$nativeEvent = e.nativeEvent) === null || _e$nativeEvent === void 0 || (_e$nativeEvent$stopIm = _e$nativeEvent.stopImmediatePropagation) === null || _e$nativeEvent$stopIm === void 0 || _e$nativeEvent$stopIm.call(_e$nativeEvent);
              var values = select_base_Option.getSelectValueArr(value, value[index], true, valueType, keys);
              var _getSelectedOptions4 = select_base_Option.getSelectedOptions(values, multiple, valueType, keys, valueToOption, value),
                currentSelectedOptions = _getSelectedOptions4.currentSelectedOptions;
              onChange(values, {
                e: e,
                selectedOptions: currentSelectedOptions,
                trigger: "tag-remove"
              });
              tagProps === null || tagProps === void 0 || (_tagProps$onClose = tagProps.onClose) === null || _tagProps$onClose === void 0 || _tagProps$onClose.call(tagProps, {
                e: e
              });
              onRemove === null || onRemove === void 0 || onRemove({
                value: targetVal,
                data: {
                  label: targetLabel,
                  value: targetVal
                },
                e: e
              });
            }
          }), targetLabel);
        });
      };
    }
    if (typeof valueDisplay === "string") {
      return valueDisplay;
    }
    if (multiple) {
      return function (_ref4) {
        var onClose = _ref4.onClose;
        return _util_parseTNode.parseContentTNode(valueDisplay, {
          value: selectedOptions,
          onClose: onClose
        });
      };
    }
    return _util_parseTNode.parseContentTNode(valueDisplay, {
      value: selectedLabel,
      onClose: _util_noop["default"]
    });
  }, [valueDisplay, multiple, selectedLabel, minCollapsedNum, options, disabled, readOnly, size, tagProps, value, valueType, keys, valueToOption, onRemove, selectedOptions]);
  var updateScrollTop = function updateScrollTop(content) {
    if (!content || isScrolling) {
      return;
    }
    var firstSelectedNode = content.querySelector(".".concat(classPrefix, "-is-selected"));
    if (!multiple && firstSelectedNode) {
      var _getComputedStyle = getComputedStyle(firstSelectedNode),
        paddingBottom = _getComputedStyle.paddingBottom;
      var _getComputedStyle2 = getComputedStyle(content),
        marginBottom = _getComputedStyle2.marginBottom;
      var elementBottomHeight = parseInt(paddingBottom, 10) + parseInt(marginBottom, 10);
      var updateValue = _util_helper.getOffsetTopToContainer(firstSelectedNode, content) - content.offsetTop - (content.clientHeight - firstSelectedNode.clientHeight) + elementBottomHeight;
      setTimeout(function () {
        content.scrollTop = updateValue;
      });
    }
  };
  var onMouseEnter = props.onMouseEnter,
    onMouseLeave = props.onMouseLeave;
  var handleEnter = function handleEnter(_, context) {
    onEnter === null || onEnter === void 0 || onEnter(_objectSpread(_objectSpread({}, context), {}, {
      value: value
    }));
  };
  var handleScroll = function handleScroll(_ref5) {
    var e = _ref5.e;
    toggleIsScrolling(true);
    onScroll === null || onScroll === void 0 || onScroll({
      e: e
    });
    if (onScrollToBottom) {
      var debounceOnScrollBottom = debounce.debounce(function (e2) {
        return onScrollToBottom({
          e: e2
        });
      }, 100);
      var _e$target = e.target,
        scrollTop = _e$target.scrollTop,
        clientHeight = _e$target.clientHeight,
        scrollHeight = _e$target.scrollHeight;
      if (Math.abs(scrollHeight - clientHeight - scrollTop) <= 1) {
        debounceOnScrollBottom(e);
      }
    }
  };
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(name, "__wrap"), className),
    style: style,
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave
  }, /* @__PURE__ */React__default["default"].createElement(selectInput_index.SelectInput, _objectSpread({
    autoWidth: !(style !== null && style !== void 0 && style.width) && autoWidth,
    ref: _util_ref.composeRefs(ref, selectInputRef),
    className: name,
    readonly: readOnly,
    autofocus: props.autofocus,
    allowInput: (filterable !== null && filterable !== void 0 ? filterable : local.filterable) || isFunction.isFunction(filter),
    multiple: multiple,
    value: selectedLabel,
    options: selectedOptions,
    valueDisplay: renderValueDisplay,
    clearable: clearable,
    disabled: disabled,
    status: props.status,
    tips: props.tips,
    borderless: borderless,
    label: label,
    suffix: props.suffix,
    prefixIcon: prefixIcon,
    suffixIcon: renderSuffixIcon(),
    panel: renderContent(),
    placeholder: !multiple && innerPopupVisible && selectedLabel ? selectedLabel : placeholder || t(local.placeholder),
    inputValue: inputValue,
    tagInputProps: _objectSpread({
      size: size
    }, tagInputProps),
    tagProps: _objectSpread({
      size: size
    }, tagProps),
    inputProps: _objectSpread(_objectSpread({
      size: size
    }, inputProps), {}, {
      onKeydown: handleKeyDown
    }),
    minCollapsedNum: minCollapsedNum,
    collapsedItems: collapsedItems,
    updateScrollTop: updateScrollTop,
    popupProps: _objectSpread({
      overlayClassName: ["".concat(name, "__dropdown"), overlayClassName],
      onScroll: handleScroll
    }, restPopupProps),
    popupVisible: innerPopupVisible,
    onPopupVisibleChange: handlePopupVisibleChange,
    onTagChange: onTagChange,
    onInputChange: handleInputChange,
    onFocus: onFocus,
    onEnter: handleEnter,
    onBlur: function onBlur(_, context) {
      _onBlur === null || _onBlur === void 0 || _onBlur({
        value: value,
        e: context.e
      });
    },
    onClear: handleClear
  }, selectInputProps)));
}, {
  Option: select_base_Option.Option,
  OptionGroup: select_base_OptionGroup["default"]
});
Select.displayName = "Select";

exports["default"] = Select;
