// tdesign-react/lib/color-picker/ColorPicker.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-5e72c8b9.js';
import { _ as _slicedToArray } from '../_chunks/dep-94a7af64.js';
import { _ as _objectWithoutProperties } from '../_chunks/dep-4461a684.js';
import React, { useRef } from 'react';
import { Popup } from '../popup/index.js';
import useClassNames from './hooks/useClassNames.js';
import useControlled from '../hooks/useControlled.js';
import ColorTrigger from './components/trigger.js';
import ColorPanel from './components/panel/index.js';
import { colorPickerDefaultProps } from './defaultProps.js';
import useDefaultProps from '../hooks/useDefaultProps.js';
import '../_chunks/dep-43f7da7b.js';
import '../_chunks/dep-0054b574.js';
import '../_chunks/dep-8e8ac0d4.js';
import '../popup/Popup.js';
import 'react-transition-group';
import 'classnames';
import 'lodash-es';
import '../_chunks/dep-a347b1b5.js';
import '../common/Portal.js';
import 'react-dom';
import '../hooks/useConfig.js';
import '../config-provider/ConfigContext.js';
import '../_chunks/dep-923c86e8.js';
import '../_chunks/dep-03cf63f2.js';
import 'dayjs';
import '../hooks/useLayoutEffect.js';
import '../hooks/useAnimation.js';
import '../hooks/useAttach.js';
import '../hooks/useMutationObserver.js';
import '../hooks/useLatest.js';
import '../hooks/usePopper.js';
import '../_chunks/dep-1e9f924f.js';
import '../_chunks/dep-57e67945.js';
import 'react-fast-compare';
import '@popperjs/core';
import '../hooks/useWindowSize.js';
import '../popup/defaultProps.js';
import '../popup/hooks/useTrigger.js';
import '../_util/listener.js';
import '../_util/ref.js';
import 'react-is';
import '../_util/isFragment.js';
import '../popup/utils/transition.js';
import '../_util/noop.js';
import '../popup/PopupPlugin.js';
import '../_chunks/dep-b9d1804c.js';
import '../_util/react-render.js';
import '../common/PluginContainer.js';
import '../config-provider/ConfigProvider.js';
import '../_chunks/dep-ee6c694f.js';
import '../_chunks/dep-c02aebd8.js';
import '../_chunks/dep-dfdbe6bd.js';
import '../_chunks/dep-8df8dad3.js';
import '../input/index.js';
import '../input/Input.js';
import 'tdesign-icons-react';
import '../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../_util/parseTNode.js';
import '../_chunks/dep-65d82dda.js';
import '../hooks/useGlobalIcon.js';
import '../locale/LocalReceiver.js';
import '../input/defaultProps.js';
import '../input/InputGroup.js';
import '../input/useLengthLimit.js';
import '../_chunks/dep-ee4dfb46.js';
import '../_chunks/dep-7fd3d1d9.js';
import '../input-number/index.js';
import '../input-number/InputNumber.js';
import '../button/index.js';
import '../button/Button.js';
import '../hooks/useDomRefCallback.js';
import '../hooks/useRipple.js';
import '../_chunks/dep-2481dfad.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../loading/defaultProps.js';
import '../loading/gradient.js';
import '../loading/plugin.js';
import '../button/defaultProps.js';
import '../input-number/useInputNumber.js';
import '../_chunks/dep-35a55ece.js';
import '../hooks/useCommonClassName.js';
import '../input-number/defaultProps.js';
import './components/panel/alpha.js';
import './components/panel/slider.js';
import '../hooks/useMouseEvent.js';
import './hooks/useStyles.js';
import './components/panel/format/index.js';
import '../select/index.js';
import '../select/base/Select.js';
import '../_util/helper.js';
import '../common/FakeArrow.js';
import '../select-input/index.js';
import '../select-input/SelectInput.js';
import '../select-input/defaultProps.js';
import '../select-input/useMultiple.js';
import '../tag-input/index.js';
import '../tag-input/TagInput.js';
import '../hooks/useDragSorter.js';
import '../hooks/useEventCallback.js';
import '../tag-input/defaultProps.js';
import '../tag-input/useHover.js';
import '../tag-input/useTagList.js';
import '../tag/index.js';
import '../tag/Tag.js';
import '../tag/defaultProps.js';
import '../tag/CheckTag.js';
import '../_chunks/dep-41651104.js';
import '../tag/CheckTagGroup.js';
import '../tag-input/useTagScroll.js';
import '../select-input/useOverlayInnerStyle.js';
import '../hooks/useInnerPopupVisible.js';
import '../select-input/useSingle.js';
import '../select/defaultProps.js';
import '../select/hooks/useKeyboardControl.js';
import '../_chunks/dep-a58e8ab4.js';
import '../select/base/OptionGroup.js';
import '../select/base/PopupContent.js';
import '../select/hooks/usePanelVirtualScroll.js';
import '../hooks/useVirtualScroll.js';
import './components/panel/header.js';
import '../radio/index.js';
import '../_chunks/dep-d96ef314.js';
import '../common/Check.js';
import '../_chunks/dep-be76de2c.js';
import '../hooks/useDeepEffect.js';
import '../radio/defaultProps.js';
import '../radio/useKeyboard.js';
import './components/panel/hue.js';
import './components/panel/linear-gradient.js';
import './components/panel/saturation.js';
import './components/panel/swatches.js';

var _excluded = ["popupProps", "clearable", "disabled", "borderless", "inputProps", "onChange", "onClear", "colorModes"],
  _excluded2 = ["overlayClassName", "overlayInnerStyle"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var ColorPicker = function ColorPicker(props) {
  var baseClassName = useClassNames();
  var _useDefaultProps = useDefaultProps(props, colorPickerDefaultProps),
    popupProps = _useDefaultProps.popupProps,
    clearable = _useDefaultProps.clearable,
    disabled = _useDefaultProps.disabled,
    borderless = _useDefaultProps.borderless,
    inputProps = _useDefaultProps.inputProps,
    onChange = _useDefaultProps.onChange,
    onClear = _useDefaultProps.onClear,
    colorModes = _useDefaultProps.colorModes,
    rest = _objectWithoutProperties(_useDefaultProps, _excluded);
  var _ref = popupProps || {},
    overlayClassName = _ref.overlayClassName,
    _ref$overlayInnerStyl = _ref.overlayInnerStyle,
    overlayInnerStyle = _ref$overlayInnerStyl === void 0 ? {} : _ref$overlayInnerStyl,
    restPopupProps = _objectWithoutProperties(_ref, _excluded2);
  var _useControlled = useControlled(props, "value", onChange),
    _useControlled2 = _slicedToArray(_useControlled, 2),
    innerValue = _useControlled2[0],
    setInnerValue = _useControlled2[1];
  var triggerRef = useRef(null);
  var colorPanelRef = useRef(null);
  var popProps = _objectSpread(_objectSpread({
    placement: "bottom-left",
    expandAnimation: true,
    trigger: "click"
  }, restPopupProps), {}, {
    overlayClassName: [baseClassName, overlayClassName],
    overlayInnerStyle: _objectSpread({
      padding: 0
    }, overlayInnerStyle)
  });
  return /* @__PURE__ */React.createElement(Popup, _objectSpread(_objectSpread({}, popProps), {}, {
    content: !disabled && /* @__PURE__ */React.createElement(ColorPanel, _objectSpread(_objectSpread({}, rest), {}, {
      disabled: disabled,
      value: innerValue,
      colorModes: colorModes,
      onChange: function onChange(value, context) {
        return setInnerValue(value, context);
      },
      ref: colorPanelRef
    }))
  }), /* @__PURE__ */React.createElement("div", {
    className: "".concat(baseClassName, "__trigger"),
    ref: triggerRef
  }, /* @__PURE__ */React.createElement(ColorTrigger, {
    clearable: clearable,
    disabled: disabled,
    borderless: borderless,
    inputProps: inputProps,
    value: innerValue,
    onChange: setInnerValue,
    onClear: onClear
  })));
};
ColorPicker.displayName = "ColorPicker";
var _ColorPicker = /*#__PURE__*/React.memo(ColorPicker);

export { _ColorPicker as default };
