// tdesign-react/es/date-picker/panel/RangePanel.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-5e72c8b9.js';
import { _ as _slicedToArray } from '../../_chunks/dep-94a7af64.js';
import React, { forwardRef, useState, useMemo, useEffect } from 'react';
import classNames from 'classnames';
import { isArray, isFunction } from 'lodash-es';
import { g as getDefaultFormat, p as parseToDayjs } from '../../_chunks/dep-07ae3139.js';
import '../../_chunks/dep-3561cd86.js';
import useConfig from '../../hooks/useConfig.js';
import PanelContent from './PanelContent.js';
import ExtraContent from './ExtraContent.js';
import useTableData from '../hooks/useTableData.js';
import useDisableDate from '../hooks/useDisableDate.js';
import useDefaultProps from '../../hooks/useDefaultProps.js';
import { parseToDateTime } from '../utils.js';
import { TimePickerPanel } from '../../time-picker/index.js';
import { l as log } from '../../_chunks/dep-65d82dda.js';
import '../../_chunks/dep-43f7da7b.js';
import '../../_chunks/dep-0054b574.js';
import '../../_chunks/dep-8e8ac0d4.js';
import 'dayjs';
import '../../_chunks/dep-7d48ddac.js';
import '../../_chunks/dep-9b727dd4.js';
import '../../config-provider/ConfigContext.js';
import '../../locale/zh_CN.js';
import '../../_chunks/dep-766c4603.js';
import '../base/Header.js';
import '../../_chunks/dep-1e9f924f.js';
import '../../_chunks/dep-57e67945.js';
import '../../locale/LocalReceiver.js';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';
import '../../config-provider/type.js';
import '../../pagination/index.js';
import '../../pagination/Pagination.js';
import '../../_chunks/dep-4461a684.js';
import '../../_util/noop.js';
import '../../hooks/useControlled.js';
import '../../select/index.js';
import '../../select/base/Select.js';
import '../../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../../_util/helper.js';
import '../../_util/parseTNode.js';
import '../../_util/ref.js';
import 'react-is';
import '../../_util/isFragment.js';
import '../../common/FakeArrow.js';
import '../../_chunks/dep-9721fff3.js';
import '../../loading/index.js';
import '../../loading/Loading.js';
import '../../_chunks/dep-a347b1b5.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../hooks/useLayoutEffect.js';
import '../../loading/defaultProps.js';
import '../../loading/gradient.js';
import '../../_chunks/dep-2481dfad.js';
import '../../hooks/useDomRefCallback.js';
import '../../loading/plugin.js';
import '../../_util/react-render.js';
import '../../_chunks/dep-fe79dc5d.js';
import '../../common/PluginContainer.js';
import '../../loading/style/css.js';
import '../../loading/type.js';
import '../../select-input/index.js';
import '../../select-input/SelectInput.js';
import '../../popup/index.js';
import '../../popup/Popup.js';
import 'react-transition-group';
import '../../hooks/useAnimation.js';
import '../../hooks/useAttach.js';
import '../../hooks/useMutationObserver.js';
import '../../hooks/useLatest.js';
import '../../hooks/usePopper.js';
import 'react-fast-compare';
import '@popperjs/core';
import '../../hooks/useWindowSize.js';
import '../../popup/defaultProps.js';
import '../../popup/hooks/useTrigger.js';
import '../../_util/listener.js';
import '../../popup/utils/transition.js';
import '../../popup/PopupPlugin.js';
import '../../popup/style/css.js';
import '../../popup/type.js';
import '../../select-input/defaultProps.js';
import '../../select-input/useMultiple.js';
import '../../tag-input/index.js';
import '../../tag-input/TagInput.js';
import 'tdesign-icons-react';
import '../../hooks/useDragSorter.js';
import '../../hooks/useEventCallback.js';
import '../../hooks/useGlobalIcon.js';
import '../../input/index.js';
import '../../input/Input.js';
import '../../input/defaultProps.js';
import '../../input/InputGroup.js';
import '../../input/useLengthLimit.js';
import '../../input/style/css.js';
import '../../input/type.js';
import '../../tag-input/defaultProps.js';
import '../../tag-input/useHover.js';
import '../../tag-input/useTagList.js';
import '../../tag/index.js';
import '../../tag/Tag.js';
import '../../_chunks/dep-dfdbe6bd.js';
import '../../tag/defaultProps.js';
import '../../tag/CheckTag.js';
import '../../_chunks/dep-29a70a9d.js';
import '../../tag/CheckTagGroup.js';
import '../../tag/style/css.js';
import '../../tag/type.js';
import '../../tag-input/useTagScroll.js';
import '../../tag-input/style/css.js';
import '../../tag-input/type.js';
import '../../select-input/useOverlayInnerStyle.js';
import '../../hooks/useInnerPopupVisible.js';
import '../../select-input/useSingle.js';
import '../../select-input/style/css.js';
import '../../select-input/type.js';
import '../../select/defaultProps.js';
import '../../select/hooks/useKeyboardControl.js';
import '../../_chunks/dep-29be6164.js';
import '../../hooks/useRipple.js';
import '../../select/base/OptionGroup.js';
import '../../select/base/PopupContent.js';
import '../../select/hooks/usePanelVirtualScroll.js';
import '../../hooks/useVirtualScroll.js';
import '../../select/style/css.js';
import '../../select/type.js';
import '../../input-number/index.js';
import '../../input-number/InputNumber.js';
import '../../button/index.js';
import '../../button/Button.js';
import '../../button/defaultProps.js';
import '../../button/style/css.js';
import '../../button/type.js';
import '../../input-number/useInputNumber.js';
import '../../_chunks/dep-35a55ece.js';
import '../../hooks/useCommonClassName.js';
import '../../input-number/defaultProps.js';
import '../../input-number/style/css.js';
import '../../input-number/type.js';
import '../../input-adornment/index.js';
import '../../input-adornment/InputAdornment.js';
import '../../input-adornment/style/css.js';
import '../../input-adornment/type.js';
import '../../pagination/hooks/useBoundaryJumper.js';
import '../../pagination/hooks/usePrevNextJumper.js';
import '../../pagination/hooks/usePageNumber.js';
import '../../pagination/hooks/useTotal.js';
import '../../pagination/validators.js';
import '../../pagination/defaultProps.js';
import '../../pagination/PaginationMini.js';
import '../../pagination/style/css.js';
import '../../pagination/type.js';
import '../hooks/useSelectRange.js';
import '../../_chunks/dep-7ff1070c.js';
import '../../_chunks/dep-35769bc6.js';
import '../base/Table.js';
import '../base/Cell.js';
import '../base/Footer.js';
import '../../time-picker/TimePicker.js';
import '../../_chunks/dep-35931e80.js';
import '../../time-picker/hooks/useTimePickerTextConfig.js';
import '../../hooks/useDebounce.js';
import '../../hooks/usePersistFn.js';
import '../../hooks/usePropsRef.js';
import '../../time-picker/TimeRangePicker.js';
import '../../range-input/index.js';
import '../../range-input/RangeInput.js';
import '../../range-input/defaultProps.js';
import '../../range-input/RangeInputPopup.js';
import '../../range-input/style/css.js';
import '../../range-input/type.js';
import '../../time-picker/panel/TimePickerPanel.js';
import '../../time-picker/defaultProps.js';
import '../../time-picker/style/css.js';
import '../../time-picker/type.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var RangePanel = /*#__PURE__*/forwardRef(function (originalProps, ref) {
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix,
    globalDatePickerConfig = _useConfig.datePicker;
  var panelName = "".concat(classPrefix, "-date-range-picker__panel");
  var _useState = useState(true),
    _useState2 = _slicedToArray(_useState, 2),
    isDateRangeContent = _useState2[0],
    toggleDateRangeContent = _useState2[1];
  var props = useDefaultProps(originalProps, {
    mode: "date",
    panelPreselection: true,
    enableTimePicker: false,
    presetsPlacement: "bottom",
    needConfirm: true
  });
  var _props$value = props.value,
    value = _props$value === void 0 ? [] : _props$value,
    _props$hoverValue = props.hoverValue,
    hoverValue = _props$hoverValue === void 0 ? [] : _props$hoverValue,
    mode = props.mode,
    presets = props.presets,
    enableTimePicker = props.enableTimePicker,
    presetsPlacement = props.presetsPlacement,
    disableDateFromProps = props.disableDate,
    _props$firstDayOfWeek = props.firstDayOfWeek,
    firstDayOfWeek = _props$firstDayOfWeek === void 0 ? globalDatePickerConfig.firstDayOfWeek : _props$firstDayOfWeek,
    isFirstValueSelected = props.isFirstValueSelected,
    needConfirm = props.needConfirm,
    style = props.style,
    className = props.className,
    activeIndex = props.activeIndex,
    year = props.year,
    month = props.month,
    range = props.range,
    cell = props.cell,
    _props$time = props.time,
    time = _props$time === void 0 ? [] : _props$time,
    panelPreselection = props.panelPreselection,
    onPresetClick = props.onPresetClick,
    cancelRangeSelectLimit = props.cancelRangeSelectLimit,
    _onClick = props.onClick,
    onConfirmClick = props.onConfirmClick,
    disableTime = props.disableTime,
    onTimePickerChange = props.onTimePickerChange,
    popupVisible = props.popupVisible,
    isSwitchTimeMode = props.isSwitchTimeMode;
  var _getDefaultFormat = getDefaultFormat({
      mode: props.mode,
      format: props.format,
      enableTimePicker: props.enableTimePicker
    }),
    format = _getDefaultFormat.format,
    timeFormat = _getDefaultFormat.timeFormat;
  var disableDateOptions = useDisableDate({
    disableDate: disableDateFromProps,
    mode: mode,
    format: format,
    start: isFirstValueSelected && activeIndex === 1 ? new Date(parseToDayjs(value[0], format, "start").toDate().setHours(0, 0, 0)) : void 0,
    end: isFirstValueSelected && activeIndex === 0 ? new Date(parseToDayjs(value[1], format).toDate().setHours(23, 59, 59)) : void 0
  });
  var rangeValue = useMemo(function () {
    var startRange = range;
    var endRange = range;
    if (isArray(range)) {
      if (range.length !== 2) {
        log.warn("DateRangePicker", "`range` length must be 2 when `range` is an array.");
      }
      var first = range[0];
      var second = range[1];
      if ((isArray(first) || isFunction(first)) && (isArray(second) || isFunction(second))) {
        startRange = first;
        endRange = second;
      }
    }
    return {
      start: startRange,
      end: endRange
    };
  }, [range]);
  var disableTimeOptions = function disableTimeOptions(h, m, s, ms) {
    if (!isFunction(disableTime)) {
      return {};
    }
    var _ref = value || [],
      _ref2 = _slicedToArray(_ref, 2),
      startTime = _ref2[0],
      endTime = _ref2[1];
    return disableTime([parseToDateTime(startTime, format), parseToDateTime(endTime, format, [h, m, s, ms])], {
      partial: activeIndex === 0 ? "start" : "end"
    });
  };
  var _year = _slicedToArray(year, 2),
    startYear = _year[0],
    endYear = _year[1];
  var _month = _slicedToArray(month, 2),
    startMonth = _month[0],
    endMonth = _month[1];
  var hidePreselection = !panelPreselection && value.length === 2;
  var startTableData = useTableData(_objectSpread({
    isRange: true,
    start: value[0] ? parseToDayjs(value[0], format).toDate() : void 0,
    end: value[1] ? parseToDayjs(value[1], format).toDate() : void 0,
    hoverStart: !hidePreselection && hoverValue[0] ? parseToDayjs(hoverValue[0], format).toDate() : void 0,
    hoverEnd: !hidePreselection && hoverValue[1] ? parseToDayjs(hoverValue[1], format).toDate() : void 0,
    year: startYear,
    month: startMonth,
    range: rangeValue.start,
    mode: mode,
    firstDayOfWeek: firstDayOfWeek,
    cancelRangeSelectLimit: cancelRangeSelectLimit
  }, disableDateOptions));
  var endTableData = useTableData(_objectSpread({
    isRange: true,
    start: value[0] ? parseToDayjs(value[0], format).toDate() : void 0,
    end: value[1] ? parseToDayjs(value[1], format).toDate() : void 0,
    hoverStart: !hidePreselection && hoverValue[0] ? parseToDayjs(hoverValue[0], format).toDate() : void 0,
    hoverEnd: !hidePreselection && hoverValue[1] ? parseToDayjs(hoverValue[1], format).toDate() : void 0,
    year: mode === "year" && endYear - startYear <= 9 ? endYear + 9 : endYear,
    month: endMonth,
    range: rangeValue.end,
    mode: mode,
    firstDayOfWeek: firstDayOfWeek,
    cancelRangeSelectLimit: cancelRangeSelectLimit
  }, disableDateOptions));
  var panelContentProps = {
    mode: mode,
    format: format,
    cell: cell,
    firstDayOfWeek: firstDayOfWeek,
    popupVisible: popupVisible,
    enableTimePicker: props.enableTimePicker,
    isSwitchTimeMode: isSwitchTimeMode,
    timePickerProps: _objectSpread({
      disableTime: disableTimeOptions
    }, props.timePickerProps),
    onMonthChange: props.onMonthChange,
    onYearChange: props.onYearChange,
    onJumperClick: props.onJumperClick,
    onCellClick: props.onCellClick,
    onCellMouseEnter: props.onCellMouseEnter,
    onCellMouseLeave: props.onCellMouseLeave,
    onTimePickerChange: props.onTimePickerChange
  };
  var onTimeModeChange = function onTimeModeChange() {
    toggleDateRangeContent(function (prev) {
      return !prev;
    });
  };
  var renderPanelContent = function renderPanelContent() {
    if (!enableTimePicker || isSwitchTimeMode) {
      if (isDateRangeContent) return /* @__PURE__ */React.createElement(React.Fragment, null, /* @__PURE__ */React.createElement(PanelContent, _objectSpread({
        key: "startPanel",
        partial: "start",
        year: startYear,
        month: startMonth,
        time: time[activeIndex],
        tableData: startTableData,
        value: value,
        range: rangeValue.start
      }, panelContentProps)), /* @__PURE__ */React.createElement(PanelContent, _objectSpread({
        key: "endPanel",
        partial: "end",
        year: endYear,
        month: endMonth,
        time: time[activeIndex],
        value: value,
        tableData: endTableData,
        internalYear: year,
        range: rangeValue.end
      }, panelContentProps)));
      return /* @__PURE__ */React.createElement(React.Fragment, null, /* @__PURE__ */React.createElement(TimePickerPanel, _objectSpread({
        key: "startPanel",
        position: "start",
        format: timeFormat,
        value: time[0],
        onChange: function onChange(val) {
          return onTimePickerChange(val, {
            activeIndex: 0
          });
        }
      }, props.timePickerProps)), /* @__PURE__ */React.createElement(TimePickerPanel, _objectSpread({
        key: "endPanel",
        position: "end",
        format: timeFormat,
        value: time[1],
        onChange: function onChange(val) {
          return onTimePickerChange(val, {
            activeIndex: 1
          });
        }
      }, props.timePickerProps)));
    }
    return /* @__PURE__ */React.createElement(PanelContent, _objectSpread({
      key: "start",
      partial: activeIndex ? "end" : "start",
      year: activeIndex ? endYear : startYear,
      month: activeIndex ? endMonth : startMonth,
      time: activeIndex ? time[1] : time[0],
      value: value,
      tableData: activeIndex ? endTableData : startTableData,
      range: activeIndex ? rangeValue.end : rangeValue.start
    }, panelContentProps));
  };
  useEffect(function () {
    if (popupVisible && isSwitchTimeMode) {
      toggleDateRangeContent(true);
    }
  }, [popupVisible, isSwitchTimeMode]);
  return /* @__PURE__ */React.createElement("div", {
    ref: ref,
    style: style,
    className: classNames(panelName, className, _defineProperty(_defineProperty({}, "".concat(panelName, "--direction-row"), ["left", "right"].includes(presetsPlacement)), "".concat(panelName, "--switch-mode"), isSwitchTimeMode)),
    onClick: function onClick(e) {
      return _onClick === null || _onClick === void 0 ? void 0 : _onClick({
        e: e
      });
    }
  }, ["top", "left"].includes(presetsPlacement) ? /* @__PURE__ */React.createElement(ExtraContent, {
    presets: presets,
    selectedValue: value[activeIndex],
    enableTimePicker: enableTimePicker,
    onPresetClick: onPresetClick,
    onConfirmClick: onConfirmClick,
    presetsPlacement: presetsPlacement,
    needConfirm: needConfirm,
    onTimeModeChange: onTimeModeChange,
    isDateRangeContent: isDateRangeContent,
    isSwitchTimeMode: isSwitchTimeMode
  }) : null, /* @__PURE__ */React.createElement("div", {
    className: "".concat(panelName, "-content-wrapper")
  }, renderPanelContent()), ["bottom", "right"].includes(presetsPlacement) ? /* @__PURE__ */React.createElement(ExtraContent, {
    presets: presets,
    selectedValue: value[activeIndex],
    enableTimePicker: enableTimePicker,
    onPresetClick: onPresetClick,
    onConfirmClick: onConfirmClick,
    onTimeModeChange: onTimeModeChange,
    isDateRangeContent: isDateRangeContent,
    isSwitchTimeMode: isSwitchTimeMode,
    presetsPlacement: presetsPlacement,
    needConfirm: needConfirm
  }) : null);
});
RangePanel.displayName = "RangePanel";

export { RangePanel as default };
