// tdesign-react/esm/tabs/useTabClass.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import useConfig from '../hooks/useConfig.js';
import 'react';
import '../config-provider/ConfigContext.js';
import '../_chunks/dep-5e72c8b9.js';
import '../_chunks/dep-43f7da7b.js';
import 'lodash-es';
import '../locale/zh_CN.js';
import '../_chunks/dep-7d48ddac.js';
import 'dayjs';
import '../_chunks/dep-766c4603.js';

var useTabClass = function useTabClass() {
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix;
  var tdTabsClassPrefix = "".concat(classPrefix, "-tabs");
  var tdTabPanelClassPrefix = "".concat(classPrefix, "-tab-panel");
  var tdClassGenerator = function tdClassGenerator(append) {
    return "".concat(classPrefix, "-").concat(append);
  };
  var tdTabsClassGenerator = function tdTabsClassGenerator(append) {
    return "".concat(tdTabsClassPrefix, "__").concat(append);
  };
  var tdTabPanelClassGenerator = function tdTabPanelClassGenerator(append) {
    return "".concat(tdTabPanelClassPrefix, "__").concat(append);
  };
  var tdSizeClassGenerator = function tdSizeClassGenerator(size) {
    return "".concat(classPrefix, "-size-").concat(size === "large" ? "l" : "m");
  };
  return {
    tdTabsClassPrefix: tdTabsClassPrefix,
    tdTabPanelClassPrefix: tdTabPanelClassPrefix,
    tdClassGenerator: tdClassGenerator,
    tdTabsClassGenerator: tdTabsClassGenerator,
    tdTabPanelClassGenerator: tdTabPanelClassGenerator,
    tdSizeClassGenerator: tdSizeClassGenerator
  };
};

export { useTabClass };
