// tdesign-react/es/range-input/RangeInputPopup.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-5e72c8b9.js';
import React from 'react';
import classNames from 'classnames';
import { Popup } from '../popup/index.js';
import useConfig from '../hooks/useConfig.js';
import RangeInput from './RangeInput.js';
import useOverlayInnerStyle from '../select-input/useOverlayInnerStyle.js';
import '../_chunks/dep-43f7da7b.js';
import '../popup/Popup.js';
import '../_chunks/dep-94a7af64.js';
import '../_chunks/dep-0054b574.js';
import '../_chunks/dep-8e8ac0d4.js';
import 'react-transition-group';
import 'lodash-es';
import '../_chunks/dep-a347b1b5.js';
import '../common/Portal.js';
import 'react-dom';
import '../hooks/useLayoutEffect.js';
import '../hooks/useAnimation.js';
import '../config-provider/ConfigContext.js';
import '../locale/zh_CN.js';
import '../_chunks/dep-7d48ddac.js';
import 'dayjs';
import '../_chunks/dep-766c4603.js';
import '../hooks/useAttach.js';
import '../hooks/useControlled.js';
import '../_util/noop.js';
import '../hooks/useDefaultProps.js';
import '../hooks/useMutationObserver.js';
import '../hooks/useLatest.js';
import '../hooks/usePopper.js';
import '../_chunks/dep-1e9f924f.js';
import '../_chunks/dep-57e67945.js';
import 'react-fast-compare';
import '@popperjs/core';
import '../hooks/useWindowSize.js';
import '../popup/defaultProps.js';
import '../popup/hooks/useTrigger.js';
import '../_util/listener.js';
import '../_util/ref.js';
import 'react-is';
import '../_util/isFragment.js';
import '../popup/utils/transition.js';
import '../popup/PopupPlugin.js';
import '../_chunks/dep-fe79dc5d.js';
import '../_util/react-render.js';
import '../common/PluginContainer.js';
import '../config-provider/index.js';
import '../config-provider/ConfigProvider.js';
import '../config-provider/type.js';
import '../popup/style/css.js';
import '../popup/type.js';
import '../_chunks/dep-4461a684.js';
import 'tdesign-icons-react';
import '../input/index.js';
import '../input/Input.js';
import '../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../_util/parseTNode.js';
import '../_chunks/dep-3561cd86.js';
import '../_chunks/dep-65d82dda.js';
import '../hooks/useGlobalIcon.js';
import '../locale/LocalReceiver.js';
import '../input/defaultProps.js';
import '../input/InputGroup.js';
import '../input/useLengthLimit.js';
import '../_chunks/dep-9721fff3.js';
import '../input/style/css.js';
import '../input/type.js';
import './defaultProps.js';
import '../hooks/useInnerPopupVisible.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var RangeInputPopup = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix;
  var name = "".concat(classPrefix, "-range-input-popup");
  var className = props.className,
    style = props.style,
    inputValue = props.inputValue,
    panel = props.panel,
    popupProps = props.popupProps,
    rangeInputProps = props.rangeInputProps,
    popupVisible = props.popupVisible,
    onInputChange = props.onInputChange,
    disabled = props.disabled,
    status = props.status,
    tips = props.tips;
  var _useOverlayInnerStyle = useOverlayInnerStyle(props),
    tOverlayInnerStyle = _useOverlayInnerStyle.tOverlayInnerStyle,
    innerPopupVisible = _useOverlayInnerStyle.innerPopupVisible,
    onInnerPopupVisibleChange = _useOverlayInnerStyle.onInnerPopupVisibleChange;
  var popupClasses = classNames([name, _defineProperty({}, "".concat(name, "--visible"), popupVisible !== null && popupVisible !== void 0 ? popupVisible : innerPopupVisible)]);
  return /* @__PURE__ */React.createElement("div", {
    ref: ref,
    style: style,
    className: classNames(name, className, popupClasses)
  }, /* @__PURE__ */React.createElement(Popup, _objectSpread(_objectSpread({
    hideEmptyPopup: true,
    content: panel,
    trigger: "click",
    placement: "bottom-left",
    visible: popupVisible !== null && popupVisible !== void 0 ? popupVisible : innerPopupVisible,
    onVisibleChange: onInnerPopupVisibleChange,
    disabled: disabled
  }, popupProps), {}, {
    overlayInnerStyle: tOverlayInnerStyle
  }), /* @__PURE__ */React.createElement(RangeInput, _objectSpread({
    disabled: disabled,
    status: status,
    tips: tips,
    value: inputValue,
    onChange: onInputChange
  }, rangeInputProps))));
});
RangeInputPopup.displayName = "RangeInputPopup";

export { RangeInputPopup as default };
