// tdesign-react/cjs/progress/Progress.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var classNames = require('classnames');
var helper = require('../_chunks/dep-b011b5d3.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var progress_defaultProps = require('./defaultProps.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-1c586f5c.js');
require('../_chunks/dep-f0e884b9.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-794b42ce.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var Progress = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      CheckCircleIcon: tdesignIconsReact.CheckIcon,
      CloseCircleIcon: tdesignIconsReact.CloseIcon,
      ErrorCircleIcon: tdesignIconsReact.ErrorIcon,
      CheckCircleFilledIcon: tdesignIconsReact.CheckCircleFilledIcon,
      CloseCircleFilledIcon: tdesignIconsReact.CloseCircleFilledIcon,
      ErrorCircleFilledIcon: tdesignIconsReact.ErrorCircleFilledIcon
    }),
    CheckCircleIcon = _useGlobalIcon.CheckCircleIcon,
    CloseCircleIcon = _useGlobalIcon.CloseCircleIcon,
    ErrorCircleIcon = _useGlobalIcon.ErrorCircleIcon,
    CheckCircleFilledIcon = _useGlobalIcon.CheckCircleFilledIcon,
    CloseCircleFilledIcon = _useGlobalIcon.CloseCircleFilledIcon,
    ErrorCircleFilledIcon = _useGlobalIcon.ErrorCircleFilledIcon;
  var _useDefaultProps = hooks_useDefaultProps["default"](props, progress_defaultProps.progressDefaultProps),
    theme = _useDefaultProps.theme,
    percentage = _useDefaultProps.percentage,
    label = _useDefaultProps.label,
    _useDefaultProps$colo = _useDefaultProps.color,
    color = _useDefaultProps$colo === void 0 ? "" : _useDefaultProps$colo,
    trackColor = _useDefaultProps.trackColor,
    strokeWidth = _useDefaultProps.strokeWidth,
    size = _useDefaultProps.size,
    className = _useDefaultProps.className,
    style = _useDefaultProps.style,
    customizeStatus = _useDefaultProps.status;
  var status = !customizeStatus && percentage >= 100 ? "success" : customizeStatus;
  var innerRef = React.useRef(null);
  var labelRef = React.useRef(null);
  var _useState = React.useState("inside"),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    labelPosition = _useState2[0],
    setLabelPosition = _useState2[1];
  var getLabelPosition = React.useCallback(function () {
    if (!innerRef.current || !labelRef.current) return "inside";
    var innerWidth = innerRef.current.offsetWidth;
    var labelWidth = labelRef.current.offsetWidth;
    var padding = 8;
    return innerWidth >= labelWidth + padding ? "inside" : "outside";
  }, []);
  React.useEffect(function () {
    if (theme !== "plump") return;
    setLabelPosition(getLabelPosition());
    if (typeof ResizeObserver === "undefined") return;
    var observer = new ResizeObserver(function () {
      setLabelPosition(getLabelPosition());
    });
    if (innerRef.current) {
      observer.observe(innerRef.current);
    }
    return function () {
      return observer.disconnect();
    };
  }, [theme, percentage, label, getLabelPosition]);
  var iconMap = {
    success: CheckCircleFilledIcon,
    error: CloseCircleFilledIcon,
    warning: ErrorCircleFilledIcon
  };
  var getInfoContent = function getInfoContent() {
    if (!label) {
      return "";
    }
    var info;
    if (typeof label === "boolean") {
      info = /* @__PURE__ */React__default["default"].createElement("div", {
        className: "".concat(classPrefix, "-progress__info")
      }, "".concat(percentage, "%"));
      if (["success", "error", "warning"].includes(status)) {
        var Icon = iconMap[status];
        info = /* @__PURE__ */React__default["default"].createElement("div", {
          className: "".concat(classPrefix, "-progress__info")
        }, /* @__PURE__ */React__default["default"].createElement(Icon, {
          className: "".concat(classPrefix, "-progress__icon")
        }));
      }
    } else {
      info = /* @__PURE__ */React__default["default"].createElement("div", {
        className: "".concat(classPrefix, "-progress__info")
      }, label);
    }
    return info;
  };
  var statusClassName = "".concat(classPrefix, "-progress--status--").concat(status);
  var progressDom;
  if (theme === "circle") {
    iconMap = {
      success: CheckCircleIcon,
      error: CloseCircleIcon,
      warning: ErrorCircleIcon
    };
    var getDiameter = function getDiameter() {
      var diameter2 = 112;
      if (!size) {
        return diameter2;
      }
      switch (size) {
        default:
          diameter2 = Number(size);
          break;
        case "small":
          diameter2 = 72;
          break;
        case "medium":
          diameter2 = 112;
          break;
        case "large":
          diameter2 = 160;
          break;
      }
      return diameter2;
    };
    var getCircleStokeWidth = function getCircleStokeWidth() {
      if (!strokeWidth) {
        if (size === "small") {
          return 4;
        }
      }
      if (typeof strokeWidth !== "number" || Number.isNaN(strokeWidth)) {
        return 6;
      }
      return strokeWidth;
    };
    var circleStokeWidth = getCircleStokeWidth();
    var diameter = getDiameter();
    var radius = diameter / 2;
    var innerRadius = radius - circleStokeWidth;
    var perimeter = Math.PI * 2 * radius;
    var percent = percentage / 100;
    var strokeDasharray = "".concat(perimeter * percent, "  ").concat(perimeter * (1 - percent));
    var fontSizeRatio = innerRadius * 0.27;
    var circleBoxStyle = {
      width: diameter,
      height: diameter,
      fontSize: 4 + fontSizeRatio
    };
    var circlePathStyle = {
      stroke: color,
      strokeLinecap: circleStokeWidth < 30 ? "round" : "buff"
    };
    var circleCenterInViewBox = radius + circleStokeWidth / 2;
    progressDom = /* @__PURE__ */React__default["default"].createElement("div", {
      ref: ref,
      className: classNames__default["default"]("".concat(classPrefix, "-progress--circle"), defineProperty._defineProperty({}, "".concat(statusClassName), status)),
      style: circleBoxStyle
    }, getInfoContent(), /* @__PURE__ */React__default["default"].createElement("svg", {
      width: diameter,
      height: diameter,
      viewBox: "0 0 ".concat(diameter + circleStokeWidth, " ").concat(diameter + circleStokeWidth)
    }, /* @__PURE__ */React__default["default"].createElement("circle", {
      cx: circleCenterInViewBox,
      cy: circleCenterInViewBox,
      r: radius,
      strokeWidth: circleStokeWidth,
      stroke: trackColor || "var(--td-bg-color-component)",
      fill: "none"
    }), percentage > 0 && /* @__PURE__ */React__default["default"].createElement("circle", {
      cx: circleCenterInViewBox,
      cy: circleCenterInViewBox,
      r: radius,
      strokeWidth: circleStokeWidth,
      fill: "none",
      transform: "matrix(0,-1,1,0,0,".concat(diameter + circleStokeWidth, ")"),
      strokeDasharray: strokeDasharray,
      className: "".concat(classPrefix, "-progress__circle-inner"),
      style: circlePathStyle
    })));
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: className
    }, progressDom);
  }
  var getHeight = function getHeight() {
    if (strokeWidth) {
      return typeof strokeWidth === "string" ? strokeWidth : "".concat(strokeWidth, "px");
    }
  };
  var trackStyle = {
    height: getHeight(),
    backgroundColor: trackColor,
    borderRadius: getHeight()
  };
  var barStyle = {
    width: "".concat(percentage, "%"),
    background: helper.getBackgroundColor(color),
    borderRadius: getHeight()
  };
  if (theme === "plump") {
    progressDom = /* @__PURE__ */React__default["default"].createElement("div", {
      ref: ref,
      className: classNames__default["default"]("".concat(classPrefix, "-progress__bar"), "".concat(classPrefix, "-progress--plump"), defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(statusClassName), status), "".concat(classPrefix, "-progress--over-ten"), labelPosition === "inside"), "".concat(classPrefix, "-progress--under-ten"), labelPosition === "outside")),
      style: trackStyle
    }, labelPosition === "inside" ? /* @__PURE__ */React__default["default"].createElement("div", {
      ref: innerRef,
      className: "".concat(classPrefix, "-progress__inner"),
      style: barStyle
    }, label && /* @__PURE__ */React__default["default"].createElement("div", {
      ref: labelRef,
      className: "".concat(classPrefix, "-progress__info"),
      style: {
        color: "#fff"
      }
    }, /*#__PURE__*/React.isValidElement(label) ? label : "".concat(percentage, "%"))) : /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement("div", {
      ref: innerRef,
      className: "".concat(classPrefix, "-progress__inner"),
      style: barStyle
    }), label && /* @__PURE__ */React__default["default"].createElement("div", {
      ref: labelRef,
      className: "".concat(classPrefix, "-progress__info")
    }, /*#__PURE__*/React.isValidElement(label) ? label : "".concat(percentage, "%"))));
  } else {
    progressDom = /* @__PURE__ */React__default["default"].createElement("div", {
      ref: ref,
      className: classNames__default["default"]("".concat(classPrefix, "-progress--thin"), defineProperty._defineProperty({}, "".concat(statusClassName), status))
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-progress__bar"),
      style: trackStyle
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-progress__inner"),
      style: barStyle
    })), getInfoContent());
  }
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: className,
    style: style
  }, progressDom);
});
Progress.displayName = "Progress";

exports["default"] = Progress;
