// tdesign-react/lib/cascader/hooks.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-5e72c8b9.js';
import { _ as _slicedToArray } from '../_chunks/dep-94a7af64.js';
import { useState, useMemo, useRef, useEffect } from 'react';
import { isEqual, isFunction, isArray } from 'lodash-es';
import { p as pathToKey } from '../_chunks/dep-5ecb9314.js';
import { T as TreeStore } from '../_chunks/dep-8c837c77.js';
import useControlled from '../hooks/useControlled.js';
import useDefaultProps from '../hooks/useDefaultProps.js';
import { treeNodesEffect, treeStoreExpendEffect } from './core/effect.js';
import { isValueInvalid, isEmptyValues, getTreeValue } from './core/helper.js';
import { cascaderDefaultProps } from './defaultProps.js';
import '../_chunks/dep-43f7da7b.js';
import '../_chunks/dep-0054b574.js';
import '../_chunks/dep-8e8ac0d4.js';
import '../_chunks/dep-65d82dda.js';
import '../_chunks/dep-c02aebd8.js';
import 'mitt';
import '../_chunks/dep-69c21c7e.js';
import '../_chunks/dep-b9d1804c.js';
import '../_chunks/dep-1e9f924f.js';
import '../_chunks/dep-57e67945.js';
import '../_util/noop.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var useCascaderContext = function useCascaderContext(originalProps) {
  var props = useDefaultProps(originalProps, cascaderDefaultProps);
  var disabled = props.disabled,
    options = props.options,
    _props$keys = props.keys,
    keys = _props$keys === void 0 ? {} : _props$keys,
    checkStrictly = props.checkStrictly,
    lazy = props.lazy,
    multiple = props.multiple,
    reserveKeyword = props.reserveKeyword,
    valueMode = props.valueMode,
    load = props.load;
  var _useControlled = useControlled(props, "value", props.onChange),
    _useControlled2 = _slicedToArray(_useControlled, 2),
    innerValue = _useControlled2[0],
    setInnerValue = _useControlled2[1];
  var _useControlled3 = useControlled(props, "popupVisible", props.onPopupVisibleChange),
    _useControlled4 = _slicedToArray(_useControlled3, 2),
    innerPopupVisible = _useControlled4[0],
    setPopupVisible = _useControlled4[1];
  var _useState = useState(""),
    _useState2 = _slicedToArray(_useState, 2),
    inputVal = _useState2[0],
    setInputVal = _useState2[1];
  var _useState3 = useState(null),
    _useState4 = _slicedToArray(_useState3, 2),
    treeStore = _useState4[0],
    setTreeStore = _useState4[1];
  var _useState5 = useState([]),
    _useState6 = _slicedToArray(_useState5, 2),
    treeNodes = _useState6[0],
    setTreeNodes = _useState6[1];
  var _useState7 = useState([]),
    _useState8 = _slicedToArray(_useState7, 2),
    expend = _useState8[0],
    setExpend = _useState8[1];
  var _useState9 = useState(void 0),
    _useState0 = _slicedToArray(_useState9, 2),
    scopeVal = _useState0[0],
    setScopeVal = _useState0[1];
  var isParentFilterable = useMemo(function () {
    return !!((["parentFirst", "all"].includes(props.valueMode) || props.checkStrictly) && inputVal);
  }, [props.valueMode, props.checkStrictly, inputVal]);
  var cascaderContext = useMemo(function () {
    var size = props.size,
      checkStrictly2 = props.checkStrictly,
      lazy2 = props.lazy,
      multiple2 = props.multiple,
      filterable = props.filterable,
      clearable = props.clearable,
      checkProps = props.checkProps,
      max = props.max,
      disabled2 = props.disabled,
      showAllLevels = props.showAllLevels,
      minCollapsedNum = props.minCollapsedNum,
      valueType = props.valueType;
    return {
      value: scopeVal,
      size: size,
      checkStrictly: checkStrictly2,
      lazy: lazy2,
      multiple: multiple2,
      filterable: filterable,
      clearable: clearable,
      checkProps: checkProps,
      max: max,
      disabled: disabled2,
      showAllLevels: showAllLevels,
      minCollapsedNum: minCollapsedNum,
      valueType: valueType,
      treeStore: treeStore,
      setValue: function setValue(val, source, node) {
        if (isEqual(val, scopeVal)) return;
        setInnerValue(val, {
          source: source,
          node: node
        });
      },
      visible: innerPopupVisible,
      setVisible: setPopupVisible,
      treeNodes: treeNodes,
      setTreeNodes: setTreeNodes,
      inputVal: inputVal,
      setInputVal: setInputVal,
      setExpend: setExpend,
      isParentFilterable: isParentFilterable
    };
  }, [props, scopeVal, innerPopupVisible, treeStore, treeNodes, inputVal, setInnerValue, setPopupVisible, isParentFilterable]);
  var isFilterable = useMemo(function () {
    return Boolean(props.filterable || isFunction(props.filter));
  }, [props.filterable, props.filter]);
  var optionCurrent = useRef([]);
  useEffect(function () {
    if (!isEqual(optionCurrent.current, options)) {
      optionCurrent.current = options;
      handleTreeStore();
    }
  }, [options]);
  var handleTreeStore = function handleTreeStore() {
    if (!treeStore) {
      if (!options.length) return;
      var store = new TreeStore({
        keys: _objectSpread(_objectSpread({}, keys), {}, {
          children: typeof keys.children === "string" ? keys.children : "children"
        }),
        allowDuplicateValue: props.valueType === "full",
        onLoad: function onLoad() {
          setTimeout(function () {
            store.refreshNodes();
            treeNodesEffect(inputVal, store, setTreeNodes, props.filter, isParentFilterable);
          });
        }
      });
      store.append(options);
      setTreeStore(store);
    } else {
      treeStore.reload(options);
      treeStore.refreshNodes();
      treeStoreExpendEffect(treeStore, scopeVal, [], props.valueType);
      treeNodesEffect(inputVal, treeStore, setTreeNodes, props.filter, isParentFilterable);
    }
  };
  useEffect(function () {
    if (!treeStore) return;
    var treeProps = {
      keys: _objectSpread(_objectSpread({}, keys), {}, {
        children: typeof keys.children === "string" ? keys.children : "children"
      }),
      checkable: true,
      expandMutex: true,
      expandParent: true,
      checkStrictly: checkStrictly,
      disabled: disabled,
      load: load,
      lazy: lazy,
      valueMode: valueMode
    };
    treeStore.setConfig(treeProps);
  }, [checkStrictly, disabled, keys, lazy, load, valueMode, treeStore]);
  useEffect(function () {
    var setValue = cascaderContext.setValue,
      multiple2 = cascaderContext.multiple;
    if (isValueInvalid(innerValue, cascaderContext)) {
      setValue(multiple2 ? [] : "", "invalid-value");
    }
    if (!isEmptyValues(innerValue)) {
      setScopeVal(innerValue);
    } else {
      setScopeVal(multiple2 ? [] : "");
    }
    if (multiple2 && !reserveKeyword) {
      setInputVal("");
    }
  }, [innerValue, multiple, reserveKeyword]);
  useEffect(function () {
    if (!treeStore) return;
    treeStoreExpendEffect(treeStore, scopeVal, expend, props.valueType);
  }, [treeStore, scopeVal, expend, props.valueType]);
  useEffect(function () {
    if (!treeStore) return;
    treeNodesEffect(inputVal, treeStore, setTreeNodes, props.filter, isParentFilterable);
  }, [inputVal, treeStore, props.filter, isParentFilterable]);
  useEffect(function () {
    if (!treeStore) return;
    var valueType = props.valueType;
    var getCheckedKeys = function getCheckedKeys() {
      if (valueType !== "full") {
        return getTreeValue(scopeVal);
      }
      var isValidPath = function isValidPath(path) {
        return Array.isArray(path) && path.length > 0;
      };
      if (multiple && Array.isArray(scopeVal)) {
        return scopeVal.filter(isValidPath).map(pathToKey);
      }
      return isValidPath(scopeVal) ? [pathToKey(scopeVal)] : [];
    };
    treeStore.replaceChecked(getCheckedKeys());
  }, [options, scopeVal, treeStore, multiple, props]);
  useEffect(function () {
    if (!innerPopupVisible && isFilterable) {
      setInputVal("");
    }
  }, [innerPopupVisible, isFilterable]);
  useEffect(function () {
    var inputVal2 = cascaderContext.inputVal,
      treeStore2 = cascaderContext.treeStore,
      setTreeNodes2 = cascaderContext.setTreeNodes;
    treeNodesEffect(inputVal2, treeStore2, setTreeNodes2, props.filter, isParentFilterable);
  }, [inputVal, scopeVal, isParentFilterable]);
  var getCascaderItems = function getCascaderItems(arrValue, valueType, multiple2) {
    var treeStore2 = cascaderContext.treeStore;
    if (!treeStore2) return [];
    var optionsData = [];
    var pushNodeData = function pushNodeData(node) {
      if (node !== null && node !== void 0 && node.data) optionsData.push(node.data);
    };
    if (valueType === "full") {
      var values = multiple2 ? arrValue : [arrValue];
      values.forEach(function (v) {
        if (isArray(v) && v.length) {
          pushNodeData(treeStore2.getNode(v));
        }
      });
      return optionsData;
    }
    arrValue.forEach(function (v) {
      var _treeStore2$getNodes;
      var node = (_treeStore2$getNodes = treeStore2.getNodes(v)) === null || _treeStore2$getNodes === void 0 ? void 0 : _treeStore2$getNodes[0];
      pushNodeData(node);
    });
    return optionsData;
  };
  return {
    cascaderContext: cascaderContext,
    isFilterable: isFilterable,
    innerValue: innerValue,
    getCascaderItems: getCascaderItems
  };
};

export { useCascaderContext };
