// tdesign-react/cjs/input/Input.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var classNames = require('classnames');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var _util_parseTNode = require('../_util/parseTNode.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var hooks_useLayoutEffect = require('../hooks/useLayoutEffect.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var input_defaultProps = require('./defaultProps.js');
var input_InputGroup = require('./InputGroup.js');
var input_useLengthLimit = require('./useLengthLimit.js');
var isFunction = require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('hoist-non-react-statics');
require('../_chunks/dep-b289b0be.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-f0e884b9.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["type", "autoWidth", "borderless", "placeholder", "disabled", "status", "size", "className", "inputClass", "style", "prefixIcon", "suffixIcon", "clearable", "tips", "align", "maxlength", "maxcharacter", "showClearIconOnEmpty", "autofocus", "autocomplete", "label", "suffix", "showInput", "keepWrapperWidth", "showLimitNumber", "allowInput", "allowInputOverMax", "name", "readOnly", "readonly", "format", "onClick", "onClear", "onEnter", "onKeydown", "onKeyup", "onKeypress", "onFocus", "onBlur", "onPaste", "onMouseenter", "onMouseleave", "onWheel", "onCompositionstart", "onCompositionend", "onValidate", "onChange"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var renderIcon = function renderIcon(classPrefix, type, icon) {
  var result = _util_parseTNode["default"](icon);
  var iconClassName = icon ? "".concat(classPrefix, "-input__").concat(type, "-icon") : "";
  return result ? /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-input__").concat(type, " ").concat(iconClassName)
  }, result) : null;
};
var Input = _util_forwardRefWithStatics["default"](function (originalProps, ref) {
  var _classNames3;
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("input"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      BrowseIcon: tdesignIconsReact.BrowseIcon,
      BrowseOffIcon: tdesignIconsReact.BrowseOffIcon,
      CloseCircleFilledIcon: tdesignIconsReact.CloseCircleFilledIcon
    }),
    BrowseIcon = _useGlobalIcon.BrowseIcon,
    BrowseOffIcon = _useGlobalIcon.BrowseOffIcon,
    CloseCircleFilledIcon = _useGlobalIcon.CloseCircleFilledIcon;
  var props = hooks_useDefaultProps["default"](originalProps, input_defaultProps.inputDefaultProps);
  var type = props.type,
    autoWidth = props.autoWidth,
    borderless = props.borderless,
    _props$placeholder = props.placeholder,
    placeholder = _props$placeholder === void 0 ? t(local.placeholder) : _props$placeholder,
    disabled = props.disabled,
    status = props.status,
    size = props.size,
    className = props.className,
    inputClass = props.inputClass,
    style = props.style,
    prefixIcon = props.prefixIcon,
    suffixIcon = props.suffixIcon,
    clearable = props.clearable,
    tips = props.tips,
    align = props.align,
    maxlength = props.maxlength,
    maxcharacter = props.maxcharacter,
    showClearIconOnEmpty = props.showClearIconOnEmpty,
    autofocus = props.autofocus,
    autocomplete = props.autocomplete,
    label = props.label,
    suffix = props.suffix,
    _props$showInput = props.showInput,
    showInput = _props$showInput === void 0 ? true : _props$showInput,
    keepWrapperWidth = props.keepWrapperWidth,
    showLimitNumber = props.showLimitNumber,
    allowInput = props.allowInput,
    allowInputOverMax = props.allowInputOverMax,
    name = props.name,
    readOnly = props.readOnly,
    readonly = props.readonly,
    format = props.format,
    _onClick = props.onClick,
    onClear = props.onClear,
    onEnter = props.onEnter,
    onKeydown = props.onKeydown,
    onKeyup = props.onKeyup,
    onKeypress = props.onKeypress,
    onFocus = props.onFocus,
    onBlur = props.onBlur,
    onPaste = props.onPaste,
    onMouseenter = props.onMouseenter,
    onMouseleave = props.onMouseleave,
    _onWheel = props.onWheel,
    onCompositionstart = props.onCompositionstart,
    onCompositionend = props.onCompositionend,
    onValidate = props.onValidate,
    onChangeFromProps = props.onChange,
    restProps = objectWithoutProperties._objectWithoutProperties(props, _excluded);
  var readOnlyProp = readOnly || readonly;
  var _useControlled = hooks_useControlled["default"](props, "value", onChangeFromProps),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    value = _useControlled2[0],
    onChange = _useControlled2[1];
  var _useLengthLimit = input_useLengthLimit["default"]({
      value: value === void 0 ? void 0 : String(value),
      status: status,
      maxlength: maxlength,
      maxcharacter: maxcharacter,
      allowInputOverMax: allowInputOverMax,
      onValidate: onValidate
    }),
    limitNumber = _useLengthLimit.limitNumber,
    getValueByLimitNumber = _useLengthLimit.getValueByLimitNumber,
    tStatus = _useLengthLimit.tStatus;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix,
    inputConfig = _useConfig.input;
  var composingRef = React.useRef(false);
  var inputRef = React.useRef(null);
  var inputPreRef = React.useRef(null);
  var wrapperRef = React.useRef(null);
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    isHover = _useState2[0],
    toggleIsHover = _useState2[1];
  var _useState3 = React.useState(false),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    isFocused = _useState4[0],
    toggleIsFocused = _useState4[1];
  var _useState5 = React.useState(type),
    _useState6 = slicedToArray._slicedToArray(_useState5, 2),
    renderType = _useState6[0],
    setRenderType = _useState6[1];
  var _useState7 = React.useState(""),
    _useState8 = slicedToArray._slicedToArray(_useState7, 2),
    composingValue = _useState8[0],
    setComposingValue = _useState8[1];
  var isInnerInputReadonly = readOnlyProp || !allowInput;
  var isValueEnabled = value && !disabled;
  var alwaysShowClearIcon = (inputConfig === null || inputConfig === void 0 ? void 0 : inputConfig.clearTrigger) === "always";
  var isShowClearIcon = (clearable && isValueEnabled || showClearIconOnEmpty) && isHover || isValueEnabled && alwaysShowClearIcon;
  var prefixIconContent = renderIcon(classPrefix, "prefix", _util_parseTNode["default"](prefixIcon));
  var suffixIconNew = suffixIcon;
  if (isShowClearIcon) suffixIconNew = /* @__PURE__ */React__default["default"].createElement(CloseCircleFilledIcon, {
    className: "".concat(classPrefix, "-input__suffix-clear"),
    onClick: handleClear
  });
  if (type === "password" && typeof suffixIcon === "undefined") {
    if (renderType === "password") {
      suffixIconNew = /* @__PURE__ */React__default["default"].createElement(BrowseOffIcon, {
        className: "".concat(classPrefix, "-input__suffix-clear"),
        onClick: togglePasswordVisible
      });
    } else if (renderType === "text") {
      suffixIconNew = /* @__PURE__ */React__default["default"].createElement(BrowseIcon, {
        className: "".concat(classPrefix, "-input__suffix-clear"),
        onClick: togglePasswordVisible
      });
    }
  }
  var suffixIconContent = renderIcon(classPrefix, "suffix", _util_parseTNode["default"](suffixIconNew));
  var labelContent = isFunction.isFunction(label) ? label() : label;
  var suffixContent = isFunction.isFunction(suffix) ? suffix() : suffix;
  var limitNumberNode = limitNumber && showLimitNumber ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-input__limit-number"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-disabled"), disabled))
  }, limitNumber) : null;
  var updateInputWidth = function updateInputWidth() {
    if (!autoWidth || !inputRef.current) return;
    var offsetWidth = inputPreRef.current.offsetWidth;
    var _inputPreRef$current$ = inputPreRef.current.getBoundingClientRect(),
      width = _inputPreRef$current$.width;
    var calcWidth = width < offsetWidth ? offsetWidth + 1 : width;
    inputRef.current.style.width = "".concat(calcWidth, "px");
  };
  hooks_useLayoutEffect["default"](function () {
    requestAnimationFrame(function () {
      updateInputWidth();
    });
  }, [autoWidth, value, placeholder, inputRef, composingValue]);
  React.useEffect(function () {
    var resizeObserver = null;
    if (typeof window.ResizeObserver === "undefined" || !inputRef.current) return;
    resizeObserver = new window.ResizeObserver(function () {
      updateInputWidth();
    });
    resizeObserver.observe(inputRef.current);
    return function () {
      var _resizeObserver$disco, _resizeObserver;
      (_resizeObserver$disco = (_resizeObserver = resizeObserver).disconnect) === null || _resizeObserver$disco === void 0 || _resizeObserver$disco.call(_resizeObserver);
      resizeObserver = null;
    };
  }, [inputRef]);
  React.useEffect(function () {
    setRenderType(type);
  }, [type]);
  React.useEffect(function () {
    if (value) {
      var limitedValue = getValueByLimitNumber(value);
      if (limitedValue.length !== value.length && !allowInputOverMax) {
        onChange === null || onChange === void 0 || onChange(limitedValue, {
          trigger: "initial"
        });
      }
    }
  }, []);
  var innerValue = composingRef.current ? composingValue : value !== null && value !== void 0 ? value : "";
  var formatDisplayValue = format && !isFocused ? format(innerValue) : innerValue;
  var renderInput = /* @__PURE__ */React__default["default"].createElement("input", {
    ref: inputRef,
    placeholder: placeholder,
    type: renderType,
    className: classNames__default["default"]("".concat(classPrefix, "-input__inner"), defineProperty._defineProperty({}, "".concat(classPrefix, "-input--soft-hidden"), !showInput)),
    value: formatDisplayValue,
    readOnly: isInnerInputReadonly,
    disabled: disabled,
    autoComplete: autocomplete !== null && autocomplete !== void 0 ? autocomplete : local.autocomplete || void 0,
    autoFocus: autofocus,
    onChange: handleChange,
    onKeyDown: handleKeyDown,
    onKeyUp: handleKeyUp,
    onKeyPress: handleKeyPress,
    onCompositionStart: handleCompositionStart,
    onCompositionEnd: handleCompositionEnd,
    onFocus: handleFocus,
    onBlur: handleBlur,
    onPaste: handlePaste,
    name: name
  });
  var renderInputNode = /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](inputClass, "".concat(classPrefix, "-input"), (_classNames3 = {}, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(_classNames3, "".concat(classPrefix, "-is-readonly"), readOnlyProp), "".concat(classPrefix, "-is-disabled"), disabled), "".concat(classPrefix, "-is-focused"), isFocused), "".concat(classPrefix, "-size-s"), size === "small"), "".concat(classPrefix, "-size-l"), size === "large"), "".concat(classPrefix, "-align-").concat(align), align), "".concat(classPrefix, "-is-").concat(tStatus), tStatus && tStatus !== "default"), "".concat(classPrefix, "-input--prefix"), prefixIcon || labelContent), "".concat(classPrefix, "-input--suffix"), suffixIconContent || suffixContent), "".concat(classPrefix, "-input--borderless"), borderless), defineProperty._defineProperty(_classNames3, "".concat(classPrefix, "-input--focused"), isFocused))),
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    onWheel: function onWheel(e) {
      return _onWheel === null || _onWheel === void 0 ? void 0 : _onWheel({
        e: e
      });
    },
    onClick: function onClick(e) {
      var _inputRef$current;
      (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 || _inputRef$current.focus();
      _onClick === null || _onClick === void 0 || _onClick({
        e: e
      });
    }
  }, prefixIconContent, labelContent ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-input__prefix")
  }, labelContent) : null, renderInput, autoWidth && /* @__PURE__ */React__default["default"].createElement("span", {
    ref: inputPreRef,
    className: "".concat(classPrefix, "-input__input-pre")
  }, innerValue || placeholder), suffixContent || limitNumberNode ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-input__suffix")
  }, suffixContent, limitNumberNode) : null, suffixIconContent);
  function togglePasswordVisible() {
    var _inputRef$current2;
    if (disabled) return;
    var inputEl = inputRef.current;
    var cursorPosition = ((_inputRef$current2 = inputRef.current) === null || _inputRef$current2 === void 0 ? void 0 : _inputRef$current2.selectionStart) || 0;
    var toggleType = renderType === "password" ? "text" : "password";
    setRenderType(toggleType);
    requestAnimationFrame(function () {
      inputEl === null || inputEl === void 0 || inputEl.setSelectionRange(cursorPosition, cursorPosition);
    });
  }
  function handleChange(e) {
    var trigger = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "input";
    var newStr = e.currentTarget.value;
    if (composingRef.current) {
      setComposingValue(newStr);
    } else {
      if (props.type !== "number") {
        newStr = getValueByLimitNumber(newStr);
      }
      setComposingValue(newStr);
      onChange(newStr, {
        e: e,
        trigger: trigger
      });
    }
  }
  function handleClear(e) {
    onChange === null || onChange === void 0 || onChange("", {
      e: e,
      trigger: "clear"
    });
    onClear === null || onClear === void 0 || onClear({
      e: e
    });
  }
  function handleKeyDown(e) {
    var key = e.key,
      value2 = e.currentTarget.value;
    key === "Enter" && (onEnter === null || onEnter === void 0 ? void 0 : onEnter(value2, {
      e: e
    }));
    onKeydown === null || onKeydown === void 0 || onKeydown(value2, {
      e: e
    });
  }
  function handleKeyUp(e) {
    var value2 = e.currentTarget.value;
    onKeyup === null || onKeyup === void 0 || onKeyup(value2, {
      e: e
    });
  }
  function handleKeyPress(e) {
    var value2 = e.currentTarget.value;
    onKeypress === null || onKeypress === void 0 || onKeypress(value2, {
      e: e
    });
  }
  function handleCompositionStart(e) {
    composingRef.current = true;
    var value2 = e.currentTarget.value;
    onCompositionstart === null || onCompositionstart === void 0 || onCompositionstart(value2, {
      e: e
    });
  }
  function handleCompositionEnd(e) {
    var value2 = e.currentTarget.value;
    if (composingRef.current) {
      composingRef.current = false;
      handleChange(e);
    }
    onCompositionend === null || onCompositionend === void 0 || onCompositionend(value2, {
      e: e
    });
  }
  function handleFocus(e) {
    var value2 = e.currentTarget.value;
    onFocus === null || onFocus === void 0 || onFocus(value2, {
      e: e
    });
    if (isInnerInputReadonly) return;
    toggleIsFocused(true);
  }
  function handleBlur(e) {
    var value2 = e.currentTarget.value;
    onBlur === null || onBlur === void 0 || onBlur(value2, {
      e: e
    });
    if (isInnerInputReadonly) return;
    toggleIsFocused(false);
  }
  function handlePaste(e) {
    var clipData = e.clipboardData;
    var pasteValue = clipData === null || clipData === void 0 ? void 0 : clipData.getData("text/plain");
    onPaste === null || onPaste === void 0 || onPaste({
      e: e,
      pasteValue: pasteValue
    });
  }
  function handleMouseEnter(e) {
    !readOnly && toggleIsHover(true);
    onMouseenter === null || onMouseenter === void 0 || onMouseenter({
      e: e
    });
  }
  function handleMouseLeave(e) {
    !readOnly && toggleIsHover(false);
    onMouseleave === null || onMouseleave === void 0 || onMouseleave({
      e: e
    });
  }
  React.useImperativeHandle(ref, function () {
    return {
      currentElement: wrapperRef.current,
      inputElement: inputRef.current,
      focus: function focus() {
        var _inputRef$current3;
        return (_inputRef$current3 = inputRef.current) === null || _inputRef$current3 === void 0 ? void 0 : _inputRef$current3.focus();
      },
      blur: function blur() {
        var _inputRef$current4;
        return (_inputRef$current4 = inputRef.current) === null || _inputRef$current4 === void 0 ? void 0 : _inputRef$current4.blur();
      },
      select: function select() {
        var _inputRef$current5;
        return (_inputRef$current5 = inputRef.current) === null || _inputRef$current5 === void 0 ? void 0 : _inputRef$current5.select();
      }
    };
  });
  return /* @__PURE__ */React__default["default"].createElement("div", _objectSpread({
    ref: wrapperRef,
    style: style,
    className: classNames__default["default"]("".concat(classPrefix, "-input__wrap"), className, defineProperty._defineProperty({}, "".concat(classPrefix, "-input--auto-width"), autoWidth && !keepWrapperWidth))
  }, restProps), renderInputNode, tips && /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-input__tips"), "".concat(classPrefix, "-input__tips--").concat(tStatus || "default"))
  }, tips));
}, {
  Group: input_InputGroup["default"]
});
Input.displayName = "Input";

exports["default"] = Input;
