// tdesign-react/lib/collapse/Collapse.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-5e72c8b9.js';
import { _ as _toConsumableArray } from '../_chunks/dep-1e9f924f.js';
import { _ as _slicedToArray } from '../_chunks/dep-94a7af64.js';
import { _ as _objectWithoutProperties } from '../_chunks/dep-4461a684.js';
import React, { useRef, useEffect } from 'react';
import classNames from 'classnames';
import useConfig from '../hooks/useConfig.js';
import forwardRefWithStatics from '../_util/forwardRefWithStatics.js';
import useControlled from '../hooks/useControlled.js';
import CollapsePanel from './CollapsePanel.js';
import CollapseContext from './CollapseContext.js';
import { collapseDefaultProps } from './defaultProps.js';
import useDefaultProps from '../hooks/useDefaultProps.js';
import '../_chunks/dep-43f7da7b.js';
import '../_chunks/dep-8e8ac0d4.js';
import '../_chunks/dep-57e67945.js';
import '../_chunks/dep-0054b574.js';
import '../config-provider/ConfigContext.js';
import 'lodash-es';
import '../_chunks/dep-923c86e8.js';
import '../_chunks/dep-03cf63f2.js';
import 'dayjs';
import 'hoist-non-react-statics';
import '../_util/noop.js';
import 'react-transition-group';
import '../common/FakeArrow.js';
import '../_chunks/dep-ee4dfb46.js';

var _excluded = ["defaultExpandAll", "disabled", "expandIconPlacement", "expandOnRowClick", "expandIcon"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Collapse = forwardRefWithStatics(function (originalProps, ref) {
  var props = useDefaultProps(originalProps, collapseDefaultProps);
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix;
  var componentName = "".concat(classPrefix, "-collapse");
  var borderlessClass = "".concat(classPrefix, "--border-less");
  var defaultExpandAll = props.defaultExpandAll,
    disabled = props.disabled,
    expandIconPlacement = props.expandIconPlacement,
    expandOnRowClick = props.expandOnRowClick,
    expandIcon = props.expandIcon,
    rest = _objectWithoutProperties(props, _excluded);
  var children = rest.children,
    className = rest.className,
    style = rest.style,
    expandMutex = rest.expandMutex,
    borderless = rest.borderless,
    onChange = rest.onChange;
  var _useControlled = useControlled(props, "value", onChange),
    _useControlled2 = _slicedToArray(_useControlled, 2),
    collapseValue = _useControlled2[0],
    setCollapseValue = _useControlled2[1];
  var collapseValues = useRef(collapseValue);
  useEffect(function () {
    collapseValues.current = collapseValue;
  }, [collapseValue]);
  var updateCollapseValue = function updateCollapseValue(value, context) {
    var newValue = [].concat(collapseValues.current || []);
    var index = newValue.indexOf(value);
    if (index >= 0) {
      newValue.splice(index, 1);
    } else if (expandMutex) {
      newValue = [value];
    } else {
      newValue.push(value);
    }
    collapseValues.current = _toConsumableArray(newValue);
    setCollapseValue(newValue, context);
  };
  var classes = [componentName, _defineProperty({}, borderlessClass, !!borderless), className];
  var childrenList = React.Children.toArray(children).filter(function (child) {
    return child.type.displayName === CollapsePanel.displayName;
  });
  var collapsePanelList = function collapsePanelList() {
    return childrenList.map(function (child, index) {
      var key = child.key || String(index);
      var childProps = _objectSpread({
        key: key,
        index: index + 1
      }, child.props);
      return /*#__PURE__*/React.cloneElement(child, childProps);
    });
  };
  return /* @__PURE__ */React.createElement(CollapseContext.Provider, {
    value: {
      defaultExpandAll: defaultExpandAll,
      disabled: disabled,
      expandIconPlacement: expandIconPlacement,
      expandOnRowClick: expandOnRowClick,
      expandIcon: expandIcon,
      updateCollapseValue: updateCollapseValue,
      collapseValue: collapseValue
    }
  }, /* @__PURE__ */React.createElement("div", {
    className: classNames(classes),
    style: style,
    ref: ref
  }, collapsePanelList()));
}, {
  Panel: CollapsePanel
});
Collapse.displayName = "Collapse";

export { Collapse as default };
