// tdesign-react/lib/image-viewer/defaultTrigger.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-94a7af64.js';
import React from 'react';
import { BrowseIcon } from 'tdesign-icons-react';
import { Image } from '../image/index.js';
import useConfig from '../hooks/useConfig.js';
import { useLocaleReceiver } from '../locale/LocalReceiver.js';
import '../_chunks/dep-0054b574.js';
import '../_chunks/dep-8e8ac0d4.js';
import '../image/Image.js';
import '../_chunks/dep-5e72c8b9.js';
import '../_chunks/dep-43f7da7b.js';
import '../_chunks/dep-4461a684.js';
import 'classnames';
import 'lodash-es';
import '../_chunks/dep-be76de2c.js';
import '../image/defaultProps.js';
import '../space/index.js';
import '../space/Space.js';
import 'react-is';
import '../_chunks/dep-ee4dfb46.js';
import '../_chunks/dep-1e9f924f.js';
import '../_chunks/dep-57e67945.js';
import '../space/defaultProps.js';
import '../hooks/useDefaultProps.js';
import '../config-provider/ConfigContext.js';
import '../_chunks/dep-923c86e8.js';
import '../_chunks/dep-03cf63f2.js';
import 'dayjs';
import '../hooks/useGlobalIcon.js';
import '../hooks/useImagePreviewUrl.js';
import '../_chunks/dep-65d82dda.js';
import '../_chunks/dep-c37bf4b4.js';
import '../config-provider/ConfigProvider.js';

var DefaultTrigger = function DefaultTrigger(_ref) {
  var showImage = _ref.showImage,
    onClick = _ref.onClick;
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix;
  var _useLocaleReceiver = useLocaleReceiver("imageViewer"),
    _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 1),
    local = _useLocaleReceiver2[0];
  return /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-image-viewer__trigger"),
    onClick: onClick
  }, /* @__PURE__ */React.createElement(Image, {
    src: showImage,
    alt: "preview",
    fit: "contain",
    className: "".concat(classPrefix, "-image-viewer__trigger-img")
  }), /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-image-viewer__trigger--hover")
  }, /* @__PURE__ */React.createElement("span", null, /* @__PURE__ */React.createElement(BrowseIcon, {
    size: "1.4em",
    className: "".concat(classPrefix, "-image-viewer__trigger-icon")
  }), local.previewText)));
};

export { DefaultTrigger };
