// tdesign-react/lib/image-viewer/ImageViewer.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-94a7af64.js';
import React, { useState, useMemo, useCallback } from 'react';
import { createPortal } from 'react-dom';
import { isNumber, isFunction } from 'lodash-es';
import { f as formatImages, I as ImageModal } from '../_chunks/dep-9068fa4c.js';
import { c as canUseDocument } from '../_chunks/dep-a347b1b5.js';
import useAttach from '../hooks/useAttach.js';
import useControlled from '../hooks/useControlled.js';
import useDefaultProps from '../hooks/useDefaultProps.js';
import { imageViewerDefaultProps } from './defaultProps.js';
import useImageScale from './hooks/useImageScale.js';
import useIndex from './hooks/useIndex.js';
import useViewerScale from './hooks/useViewerScale.js';
import { DefaultTrigger } from './defaultTrigger.js';
import '../_chunks/dep-0054b574.js';
import '../_chunks/dep-8e8ac0d4.js';
import '../_chunks/dep-5e72c8b9.js';
import '../_chunks/dep-43f7da7b.js';
import '../_chunks/dep-b9d1804c.js';
import 'classnames';
import 'tdesign-icons-react';
import '../_chunks/dep-35a55ece.js';
import '../_chunks/dep-65d82dda.js';
import '../hooks/useConfig.js';
import '../config-provider/ConfigContext.js';
import '../_chunks/dep-923c86e8.js';
import '../_chunks/dep-03cf63f2.js';
import 'dayjs';
import '../hooks/useGlobalIcon.js';
import '../hooks/useImagePreviewUrl.js';
import '../_chunks/dep-c37bf4b4.js';
import '../image/index.js';
import '../image/Image.js';
import '../_chunks/dep-4461a684.js';
import '../_chunks/dep-be76de2c.js';
import '../locale/LocalReceiver.js';
import '../config-provider/ConfigProvider.js';
import '../image/defaultProps.js';
import '../space/index.js';
import '../space/Space.js';
import 'react-is';
import '../_chunks/dep-ee4dfb46.js';
import '../_chunks/dep-1e9f924f.js';
import '../_chunks/dep-57e67945.js';
import '../space/defaultProps.js';
import '../tooltip/index.js';
import '../tooltip/Tooltip.js';
import '../popup/index.js';
import '../popup/Popup.js';
import 'react-transition-group';
import '../common/Portal.js';
import '../hooks/useLayoutEffect.js';
import '../hooks/useAnimation.js';
import '../hooks/useMutationObserver.js';
import '../hooks/useLatest.js';
import '../hooks/usePopper.js';
import 'react-fast-compare';
import '@popperjs/core';
import '../hooks/useWindowSize.js';
import '../popup/defaultProps.js';
import '../popup/hooks/useTrigger.js';
import '../_util/listener.js';
import '../_util/ref.js';
import '../_util/isFragment.js';
import '../popup/utils/transition.js';
import '../_util/noop.js';
import '../popup/PopupPlugin.js';
import '../_util/react-render.js';
import '../common/PluginContainer.js';
import '../tooltip/defaultProps.js';
import '../tooltip/TooltipLite.js';
import '../hooks/useSwitch.js';
import '../hooks/usePersistFn.js';
import '../dialog/index.js';
import '../dialog/Dialog.js';
import '../hooks/useDeepEffect.js';
import '../hooks/useSetState.js';
import '../dialog/defaultProps.js';
import '../dialog/DialogCard.js';
import '../_util/parseTNode.js';
import '../button/index.js';
import '../button/Button.js';
import '../hooks/useDomRefCallback.js';
import '../hooks/useRipple.js';
import '../_chunks/dep-2481dfad.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../loading/defaultProps.js';
import '../loading/gradient.js';
import '../loading/plugin.js';
import '../button/defaultProps.js';
import '../dialog/hooks/useDialogDrag.js';
import '../hooks/useMouseEvent.js';
import '../dialog/hooks/useDialogEsc.js';
import '../dialog/hooks/useDialogPosition.js';
import '../dialog/hooks/useLockStyle.js';
import '../_chunks/dep-7aa18730.js';
import '../dialog/plugin.js';
import './hooks/useIconMap.js';
import './hooks/useMirror.js';
import './hooks/usePosition.js';
import './hooks/useRotate.js';
import './hooks/useScale.js';

var ImageViewer = function ImageViewer(originalProps) {
  var props = useDefaultProps(originalProps, imageViewerDefaultProps);
  var attach = props.attach,
    mode = props.mode,
    trigger = props.trigger,
    images = props.images,
    title = props.title,
    imageScaleD = props.imageScale,
    viewerScaleD = props.viewerScale,
    className = props.className,
    style = props.style,
    innerClassName = props.innerClassName;
  var imageViewerAttach = useAttach("imageViewer", attach);
  var _useControlled = useControlled(props, "visible", function (visible2, context) {
      var _props$onClose;
      !visible2 && (props === null || props === void 0 || (_props$onClose = props.onClose) === null || _props$onClose === void 0 ? void 0 : _props$onClose.call(props, context));
    }),
    _useControlled2 = _slicedToArray(_useControlled, 2),
    visible = _useControlled2[0],
    setVisible = _useControlled2[1];
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    visibled = _useState2[0],
    setVisibled = _useState2[1];
  var imageList = useMemo(function () {
    return formatImages(images);
  }, [images]);
  var _useIndex = useIndex(props, images),
    index = _useIndex.index,
    setIndex = _useIndex.setIndex;
  var imageScale = useImageScale(imageScaleD);
  var viewerScale = useViewerScale(viewerScaleD);
  var isMini = mode === "modeless";
  var close = useCallback(function (context) {
    setVisible(false, context);
    setTimeout(function () {
      return setVisibled(false);
    }, 196);
  }, [setVisible]);
  var open = useCallback(function (index2) {
    if (!images) return;
    if (isNumber(index2)) {
      setIndex(index2);
    }
    setVisible(true, null);
    setVisibled(true);
  }, [images, setVisible, setIndex]);
  var uiImage = useMemo(function () {
    if (!trigger && imageList.length) {
      var _ref = imageList[props.defaultIndex || 0] || {},
        mainImage = _ref.mainImage,
        thumbnail = _ref.thumbnail;
      var showImage = mainImage || thumbnail;
      return /* @__PURE__ */React.createElement(DefaultTrigger, {
        showImage: showImage,
        onClick: open
      });
    }
    return isFunction(trigger) ? trigger({
      open: open,
      close: close,
      onOpen: open,
      onClose: close
    }) : trigger;
  }, [close, imageList, open, props.defaultIndex, trigger]);
  var attachElement = useMemo(function () {
    if (!canUseDocument || !imageViewerAttach) return null;
    if (typeof imageViewerAttach === "string") {
      return document.querySelector(imageViewerAttach);
    }
    if (isFunction(imageViewerAttach)) {
      return imageViewerAttach();
    }
  }, [imageViewerAttach]);
  return /* @__PURE__ */React.createElement(React.Fragment, null, uiImage, (visibled || visible) && /*#__PURE__*/createPortal(/* @__PURE__ */React.createElement(ImageModal, {
    innerClassName: innerClassName,
    title: title,
    visible: visible,
    images: imageList,
    isMini: isMini,
    imageScale: imageScale,
    viewerScale: viewerScale,
    zIndex: props.zIndex,
    defaultIndex: props.defaultIndex,
    index: index,
    onIndexChange: setIndex,
    onDownload: props.onDownload,
    draggable: props.draggable,
    closeOnOverlay: props.closeOnOverlay,
    closeBtn: props.closeBtn,
    showOverlay: props.showOverlay,
    closeOnEscKeydown: props.closeOnEscKeydown,
    onClose: close,
    onOpen: open,
    imageReferrerpolicy: props.imageReferrerpolicy,
    className: className,
    style: style
  }), attachElement));
};
ImageViewer.displayName = "ImageViewer";

export { ImageViewer as default };
