// tdesign-react/cjs/back-top/BackTop.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var classNames = require('classnames');
var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var hooks_useConfig = require('../hooks/useConfig.js');
var backTop_defaultProps = require('./defaultProps.js');
var backTop_useScroll = require('./useScroll.js');
var _util_scroll = require('../_util/scroll.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('raf');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../_util/easing.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var getContainer = function getContainer(container) {
  if (typeof container === "string") {
    if (typeof document !== "undefined") {
      return document.querySelector(container);
    }
  }
  if (typeof container === "function") {
    return container();
  }
  return null;
};
var InternalBackTop = function InternalBackTop(props, ref) {
  var _useDefaultProps = hooks_useDefaultProps["default"](props, backTop_defaultProps.backTopDefaultProps),
    theme = _useDefaultProps.theme,
    size = _useDefaultProps.size,
    shape = _useDefaultProps.shape,
    target = _useDefaultProps.target,
    visibleHeight = _useDefaultProps.visibleHeight,
    container = _useDefaultProps.container,
    duration = _useDefaultProps.duration,
    content = _useDefaultProps.content,
    offset = _useDefaultProps.offset,
    children = _useDefaultProps.children,
    cusContent = _useDefaultProps["default"],
    className = _useDefaultProps.className,
    style = _useDefaultProps.style,
    onClick = _useDefaultProps.onClick;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var scrollContainer = React.useMemo(function () {
    return getContainer(container);
  }, [container]);
  var _useScroll = backTop_useScroll["default"]({
      target: scrollContainer
    }),
    scrollTop = _useScroll.scrollTop;
  var defaultContent = /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.BacktopIcon, {
    className: "".concat(classPrefix, "-back-top__icon"),
    size: 24
  }), /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-back-top__text")
  }, "TOP"));
  var renderChildren = children || content || cusContent || defaultContent;
  var backTopRef = React__default["default"].useRef(null);
  React__default["default"].useImperativeHandle(ref, function () {
    return backTopRef.current;
  });
  var visible = React.useMemo(function () {
    if (typeof visibleHeight === "string") {
      return scrollTop >= Number(visibleHeight.replace("px", ""));
    }
    return scrollTop >= visibleHeight;
  }, [scrollTop, visibleHeight]);
  var backTopStyle = React.useMemo(function () {
    return _objectSpread({
      insetInlineEnd: offset[0],
      insetBlockEnd: offset[1]
    }, style);
  }, [offset, style]);
  var cls = classNames__default["default"]("".concat(classPrefix, "-back-top"), "".concat(classPrefix, "-back-top--theme-").concat(theme), "".concat(classPrefix, "-back-top--").concat(shape), defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-back-top--show"), visible), "".concat(classPrefix, "-size-s"), size === "small"), "".concat(classPrefix, "-size-m"), size === "medium"), className);
  var getBackTo = React.useCallback(function () {
    if (!target) return 0;
    var targetNode = getContainer(target);
    if (!targetNode) return 0;
    var rect = targetNode.getBoundingClientRect();
    var y = rect.y;
    return y;
  }, [target]);
  var handleClick = React.useCallback(function (e) {
    e.stopPropagation();
    onClick === null || onClick === void 0 || onClick({
      e: e
    });
    var backTo = getBackTo();
    _util_scroll.scrollTo(backTo, {
      container: scrollContainer,
      duration: duration
    });
  }, [duration, getBackTo, onClick, scrollContainer]);
  return /* @__PURE__ */React__default["default"].createElement("button", {
    type: "button",
    ref: backTopRef,
    className: cls,
    style: backTopStyle,
    onClick: handleClick
  }, renderChildren);
};
var BackTop = /*#__PURE__*/React__default["default"].forwardRef(InternalBackTop);
BackTop.displayName = "BackTop";

exports["default"] = BackTop;
