// tdesign-react/cjs/pagination/PaginationMini.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var button_index = require('../button/index.js');
var pagination_defaultProps = require('./defaultProps.js');
var _util_noop = require('../_util/noop.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var isObject = require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../button/Button.js');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-1c586f5c.js');
require('../hooks/useDomRefCallback.js');
require('../hooks/useRipple.js');
require('../_chunks/dep-064e4dba.js');
require('../hooks/useAnimation.js');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-f0e884b9.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../button/defaultProps.js');
require('../_util/parseTNode.js');
require('../_chunks/dep-b289b0be.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var PaginationMini = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      RoundIcon: tdesignIconsReact.RoundIcon,
      ChevronUpIcon: tdesignIconsReact.ChevronUpIcon,
      ChevronDownIcon: tdesignIconsReact.ChevronDownIcon,
      ChevronLeftIcon: tdesignIconsReact.ChevronLeftIcon,
      ChevronRightIcon: tdesignIconsReact.ChevronRightIcon
    }),
    RoundIcon = _useGlobalIcon.RoundIcon,
    ChevronUpIcon = _useGlobalIcon.ChevronUpIcon,
    ChevronDownIcon = _useGlobalIcon.ChevronDownIcon,
    ChevronLeftIcon = _useGlobalIcon.ChevronLeftIcon,
    ChevronRightIcon = _useGlobalIcon.ChevronRightIcon;
  var _useDefaultProps = hooks_useDefaultProps["default"](props, pagination_defaultProps.paginationMiniDefaultProps),
    variant = _useDefaultProps.variant,
    tips = _useDefaultProps.tips,
    showCurrent = _useDefaultProps.showCurrent,
    disabled = _useDefaultProps.disabled,
    layout = _useDefaultProps.layout,
    size = _useDefaultProps.size,
    _useDefaultProps$onCh = _useDefaultProps.onChange,
    onChange = _useDefaultProps$onCh === void 0 ? _util_noop["default"] : _useDefaultProps$onCh,
    className = _useDefaultProps.className,
    style = _useDefaultProps.style;
  var titleConfig = React.useMemo(function () {
    if (isObject.isObject(tips)) {
      return tips;
    }
    if (tips === true) {
      return {
        prev: "\u4E0A\u4E00\u9875",
        current: "\u5F53\u524D",
        next: "\u4E0B\u4E00\u9875"
      };
    }
    return {};
  }, [tips]);
  var disabledConfig = React.useMemo(function () {
    if (isObject.isObject(disabled)) {
      return disabled;
    }
    if (disabled === true) {
      return {
        prev: true,
        current: true,
        next: true
      };
    }
    return {
      prev: false,
      current: false,
      next: false
    };
  }, [disabled]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-pagination-mini"), className, defineProperty._defineProperty({}, "".concat(classPrefix, "-pagination-mini--outline"), variant === "outline")),
    ref: ref,
    style: style
  }, /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    title: titleConfig.prev,
    variant: variant,
    size: size,
    shape: "square",
    onClick: function onClick(e) {
      return onChange({
        e: e,
        trigger: "prev"
      });
    },
    icon: layout === "horizontal" ? /* @__PURE__ */React__default["default"].createElement(ChevronLeftIcon, null) : /* @__PURE__ */React__default["default"].createElement(ChevronUpIcon, null),
    className: "".concat(classPrefix, "-pagination-mini__prev"),
    disabled: disabledConfig.prev
  }), showCurrent && /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    title: titleConfig.current,
    variant: variant,
    size: size,
    shape: "square",
    onClick: function onClick(e) {
      return onChange({
        e: e,
        trigger: "current"
      });
    },
    icon: /* @__PURE__ */React__default["default"].createElement(RoundIcon, null),
    className: "".concat(classPrefix, "-pagination-mini__current"),
    disabled: disabledConfig.current
  }), /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    title: titleConfig.next,
    variant: variant,
    size: size,
    shape: "square",
    onClick: function onClick(e) {
      return onChange({
        e: e,
        trigger: "next"
      });
    },
    icon: layout === "horizontal" ? /* @__PURE__ */React__default["default"].createElement(ChevronRightIcon, null) : /* @__PURE__ */React__default["default"].createElement(ChevronDownIcon, null),
    className: "".concat(classPrefix, "-pagination-mini__next"),
    disabled: disabledConfig.next
  }));
});
PaginationMini.displayName = "PaginationMini";

exports["default"] = PaginationMini;
