// tdesign-react/cjs/pagination/Pagination.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var _util_noop = require('../_util/noop.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var select_index = require('../select/index.js');
var inputNumber_index = require('../input-number/index.js');
var inputAdornment_InputAdornment = require('../input-adornment/InputAdornment.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var pagination_hooks_useBoundaryJumper = require('./hooks/useBoundaryJumper.js');
var pagination_hooks_usePrevNextJumper = require('./hooks/usePrevNextJumper.js');
var pagination_hooks_usePageNumber = require('./hooks/usePageNumber.js');
var pagination_hooks_useTotal = require('./hooks/useTotal.js');
var pagination_validators = require('./validators.js');
var pagination_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var omit = require('../_chunks/dep-5c3139cb.js');
var isNumber = require('../_chunks/dep-f0e884b9.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../select/base/Select.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../_util/helper.js');
require('../_chunks/dep-e44b03ec.js');
require('../_util/parseTNode.js');
require('../_chunks/dep-b289b0be.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');
require('../common/FakeArrow.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-17e2e72f.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../_chunks/dep-68f5fe7b.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../hooks/useDomRefCallback.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../select-input/index.js');
require('../select-input/SelectInput.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../hooks/useAnimation.js');
require('../hooks/useAttach.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/listener.js');
require('../popup/utils/transition.js');
require('../popup/PopupPlugin.js');
require('../select-input/defaultProps.js');
require('../select-input/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/TagInput.js');
require('tdesign-icons-react');
require('../hooks/useDragSorter.js');
require('../hooks/useEventCallback.js');
require('../hooks/useGlobalIcon.js');
require('../input/index.js');
require('../input/Input.js');
require('../input/defaultProps.js');
require('../input/InputGroup.js');
require('../input/useLengthLimit.js');
require('../tag-input/defaultProps.js');
require('../tag-input/useHover.js');
require('../tag-input/useTagList.js');
require('../tag/index.js');
require('../tag/Tag.js');
require('../_chunks/dep-d14bfcfa.js');
require('../tag/defaultProps.js');
require('../tag/CheckTag.js');
require('../_chunks/dep-dd0f5f6b.js');
require('../tag/CheckTagGroup.js');
require('../tag-input/useTagScroll.js');
require('../select-input/useOverlayInnerStyle.js');
require('../hooks/useInnerPopupVisible.js');
require('../select-input/useSingle.js');
require('../_chunks/dep-f7bf3d17.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-ca918872.js');
require('../_chunks/dep-13de2fef.js');
require('../_chunks/dep-63308b84.js');
require('../_chunks/dep-048e4314.js');
require('../select/defaultProps.js');
require('../select/hooks/useKeyboardControl.js');
require('../_chunks/dep-bdc97672.js');
require('../hooks/useRipple.js');
require('../select/base/OptionGroup.js');
require('../_chunks/dep-05df1ea3.js');
require('../select/base/PopupContent.js');
require('../select/hooks/usePanelVirtualScroll.js');
require('../hooks/useVirtualScroll.js');
require('../input-number/InputNumber.js');
require('../button/index.js');
require('../button/Button.js');
require('../button/defaultProps.js');
require('../input-number/useInputNumber.js');
require('../_chunks/dep-04deac0d.js');
require('../_chunks/dep-8f62fb96.js');
require('../hooks/useCommonClassName.js');
require('../input-number/defaultProps.js');
require('../_chunks/dep-2e455d91.js');
require('../_chunks/dep-894a5ece.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

/**
 * Checks if `value` is `NaN`.
 *
 * **Note:** This method is based on
 * [`Number.isNaN`](https://mdn.io/Number/isNaN) and is not the same as
 * global [`isNaN`](https://mdn.io/isNaN) which returns `true` for
 * `undefined` and other non-number values.
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Lang
 * @param {*} value The value to check.
 * @returns {boolean} Returns `true` if `value` is `NaN`, else `false`.
 * @example
 *
 * _.isNaN(NaN);
 * // => true
 *
 * _.isNaN(new Number(NaN));
 * // => true
 *
 * isNaN(undefined);
 * // => true
 *
 * _.isNaN(undefined);
 * // => false
 */
function isNaN(value) {
  // An `NaN` primitive is the only value that is not equal to itself.
  // Perform the `toStringTag` check first to avoid errors with some
  // ActiveX objects in IE.
  return isNumber.isNumber(value) && value != +value;
}

var _excluded = ["theme", "size", "total", "showPageSize", "showPageNumber", "showPreviousAndNextBtn", "showFirstAndLastPageBtn", "showJumper", "pageEllipsisMode", "disabled", "foldedMaxPageBtn", "maxPageBtn", "totalContent", "pageSizeOptions", "onChange", "onCurrentChange", "onPageSizeChange", "style", "className", "selectProps"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Option = select_index.Select.Option;
var Pagination = /*#__PURE__*/React.forwardRef(function (originalProps, ref) {
  var props = hooks_useDefaultProps["default"](originalProps, pagination_defaultProps.paginationDefaultProps);
  var theme = props.theme,
    size = props.size,
    total = props.total,
    showPageSize = props.showPageSize,
    showPageNumber = props.showPageNumber,
    showPreviousAndNextBtn = props.showPreviousAndNextBtn,
    showFirstAndLastPageBtn = props.showFirstAndLastPageBtn,
    showJumper = props.showJumper,
    pageEllipsisMode = props.pageEllipsisMode,
    disabled = props.disabled,
    foldedMaxPageBtn = props.foldedMaxPageBtn,
    maxPageBtn = props.maxPageBtn,
    totalContent = props.totalContent,
    pageSizeOptions = props.pageSizeOptions,
    _props$onChange = props.onChange,
    onChange = _props$onChange === void 0 ? _util_noop["default"] : _props$onChange,
    onCurrentChange = props.onCurrentChange,
    onPageSizeChange = props.onPageSizeChange,
    style = props.style,
    className = props.className,
    selectProps = props.selectProps,
    otherProps = objectWithoutProperties._objectWithoutProperties(props, _excluded);
  var restProps = omit.omit(otherProps, ["current", "pageSize", "defaultPageSize", "defaultCurrent"]);
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("pagination"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    locale = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var _useControlled = hooks_useControlled["default"](props, "pageSize", onPageSizeChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    pageSize = _useControlled2[0],
    setPageSize = _useControlled2[1];
  var _useControlled3 = hooks_useControlled["default"](props, "current", onCurrentChange),
    _useControlled4 = slicedToArray._slicedToArray(_useControlled3, 2),
    current = _useControlled4[0],
    setCurrent = _useControlled4[1];
  var _useState = React.useState(current),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    jumpValue = _useState2[0],
    setJumpValue = _useState2[1];
  var min = 1;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix,
    paginationConfig = _useConfig.pagination;
  var name = "".concat(classPrefix, "-pagination");
  var pageCount = React.useMemo(function () {
    var calCount = Math.ceil(total / pageSize);
    return calCount > 0 ? calCount : 1;
  }, [pageSize, total]);
  var changeCurrent = function changeCurrent(_nextCurrent, _nextPageSize) {
    if (disabled || current === _nextCurrent) {
      return;
    }
    var nextCurrent = Math.trunc(_nextCurrent);
    if (isNaN(nextCurrent)) return;
    var nextPageSize = _nextPageSize;
    if (!nextPageSize && !pagination_validators.pageSizeValidator(nextPageSize)) {
      var _pageSizeOptions$;
      nextPageSize = pageSize !== null && pageSize !== void 0 ? pageSize : typeof pageSizeOptions[0] === "number" ? pageSizeOptions[0] : (_pageSizeOptions$ = pageSizeOptions[0]) === null || _pageSizeOptions$ === void 0 ? void 0 : _pageSizeOptions$.value;
    }
    if (nextCurrent < min) nextCurrent = min;
    if (nextCurrent > pageCount) nextCurrent = pageCount;
    setCurrent(nextCurrent, {
      current: nextCurrent,
      previous: current,
      pageSize: nextPageSize
    });
    onChange({
      current: nextCurrent,
      previous: current,
      pageSize: nextPageSize
    });
  };
  var changePageSize = function changePageSize(nextPageSize) {
    var nextCurrent = Math.min(current, Math.max(1, Math.ceil(total / nextPageSize)));
    var pageInfo = {
      current: nextCurrent,
      previous: current,
      pageSize: nextPageSize
    };
    setPageSize(nextPageSize, pageInfo);
    if (current !== nextCurrent) {
      changeCurrent(nextCurrent, nextPageSize);
    } else {
      onChange(pageInfo);
    }
  };
  var _useTotal = pagination_hooks_useTotal["default"]({
      totalContent: totalContent,
      pageSize: pageSize,
      current: current,
      total: total
    }),
    totalControl = _useTotal.totalControl;
  var _useBoundaryJumper = pagination_hooks_useBoundaryJumper["default"]({
      disabled: disabled,
      current: current,
      pageCount: pageCount,
      showFirstAndLastPageBtn: showFirstAndLastPageBtn,
      changeCurrent: changeCurrent
    }),
    firstPageJumper = _useBoundaryJumper.firstPageJumper,
    lastPageJumper = _useBoundaryJumper.lastPageJumper;
  var _usePrevNextJumper = pagination_hooks_usePrevNextJumper["default"]({
      disabled: disabled,
      current: current,
      pageCount: pageCount,
      showPreviousAndNextBtn: showPreviousAndNextBtn,
      changeCurrent: changeCurrent
    }),
    prevJumper = _usePrevNextJumper.prevJumper,
    nextJumper = _usePrevNextJumper.nextJumper;
  var _usePageNumber = pagination_hooks_usePageNumber["default"]({
      showPageNumber: showPageNumber,
      maxPageBtn: maxPageBtn,
      disabled: disabled,
      current: current,
      pageCount: pageCount,
      foldedMaxPageBtn: foldedMaxPageBtn,
      changeCurrent: changeCurrent,
      pageEllipsisMode: pageEllipsisMode
    }),
    pageNumberControl = _usePageNumber.pageNumberControl;
  React.useEffect(function () {
    setJumpValue(current);
  }, [current]);
  var pageSizeControl = showPageSize && pageSizeOptions.length ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(name, "__select")
  }, /* @__PURE__ */React__default["default"].createElement(select_index.Select, _objectSpread({
    autoWidth: true,
    size: size,
    value: pageSize,
    disabled: disabled,
    onChange: changePageSize
  }, selectProps), pageSizeOptions.map(function (item) {
    return typeof item === "number" ? /* @__PURE__ */React__default["default"].createElement(Option, {
      key: item,
      label: t(locale.itemsPerPage, {
        size: item
      }),
      value: item
    }) : /* @__PURE__ */React__default["default"].createElement(Option, {
      key: item.value,
      label: item.label,
      value: item.value
    });
  }))) : null;
  var DefaultJumper = /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(name, "__jump")
  }, t(locale.jumpTo), /* @__PURE__ */React__default["default"].createElement(inputAdornment_InputAdornment["default"], {
    append: "/ ".concat(pageCount, " ").concat(t(locale.page))
  }, /* @__PURE__ */React__default["default"].createElement(inputNumber_index.InputNumber, {
    className: "".concat(classPrefix, "-pagination__input"),
    min: min,
    size: size,
    theme: "normal",
    max: pageCount,
    disabled: disabled,
    value: jumpValue,
    onChange: function onChange(val) {
      return setJumpValue(val);
    },
    onBlur: function onBlur(val) {
      return changeCurrent(val);
    },
    onEnter: function onEnter(val) {
      return changeCurrent(val);
    },
    placeholder: ""
  })));
  var globalJumper = paginationConfig.jumper;
  var Jumper = globalJumper ? globalJumper({
    current: current,
    pageCount: pageCount,
    onChange: changeCurrent
  }) : DefaultJumper;
  return /* @__PURE__ */React__default["default"].createElement("div", _objectSpread({
    className: classNames__default["default"](name, className, defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-size-s"), size === "small"), "".concat(classPrefix, "-is-disabled"), disabled)),
    style: style,
    ref: ref
  }, restProps), totalControl, pageSizeControl, firstPageJumper, prevJumper, theme === "default" && pageNumberControl, theme === "simple" && Jumper, nextJumper, lastPageJumper, theme === "default" && showJumper && Jumper);
});
Pagination.displayName = "Pagination";

exports["default"] = Pagination;
