// tdesign-react/cjs/transfer/TransferList.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var transfer_utils = require('./utils.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var checkbox_index = require('../checkbox/index.js');
var input_index = require('../input/index.js');
var pagination_index = require('../pagination/index.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var _util_parseTNode = require('../_util/parseTNode.js');
var isEmpty = require('../_chunks/dep-545dd164.js');
var isFunction = require('../_chunks/dep-cba78412.js');
var isString = require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-83710d84.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../common/Check.js');
require('../_chunks/dep-1c586f5c.js');
require('../_util/helper.js');
require('../_chunks/dep-e44b03ec.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-089e8106.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../hooks/useDefaultProps.js');
require('../checkbox/defaultProps.js');
require('../_chunks/dep-f0e884b9.js');
require('../input/Input.js');
require('../hooks/useLayoutEffect.js');
require('../_chunks/dep-68f5fe7b.js');
require('../input/defaultProps.js');
require('../input/InputGroup.js');
require('../input/useLengthLimit.js');
require('../_chunks/dep-b289b0be.js');
require('../_chunks/dep-b011b5d3.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../pagination/Pagination.js');
require('../select/index.js');
require('../select/base/Select.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');
require('../common/FakeArrow.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../common/Portal.js');
require('react-dom');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../hooks/useDomRefCallback.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../select-input/index.js');
require('../select-input/SelectInput.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../hooks/useAnimation.js');
require('../hooks/useAttach.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/listener.js');
require('../popup/utils/transition.js');
require('../popup/PopupPlugin.js');
require('../select-input/defaultProps.js');
require('../select-input/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/TagInput.js');
require('../hooks/useDragSorter.js');
require('../hooks/useEventCallback.js');
require('../tag-input/defaultProps.js');
require('../tag-input/useHover.js');
require('../tag-input/useTagList.js');
require('../tag/index.js');
require('../tag/Tag.js');
require('../_chunks/dep-d14bfcfa.js');
require('../tag/defaultProps.js');
require('../tag/CheckTag.js');
require('../_chunks/dep-dd0f5f6b.js');
require('../tag/CheckTagGroup.js');
require('../tag-input/useTagScroll.js');
require('../select-input/useOverlayInnerStyle.js');
require('../hooks/useInnerPopupVisible.js');
require('../select-input/useSingle.js');
require('../_chunks/dep-f7bf3d17.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-ca918872.js');
require('../_chunks/dep-13de2fef.js');
require('../_chunks/dep-63308b84.js');
require('../_chunks/dep-048e4314.js');
require('../select/defaultProps.js');
require('../select/hooks/useKeyboardControl.js');
require('../_chunks/dep-bdc97672.js');
require('../hooks/useRipple.js');
require('../select/base/OptionGroup.js');
require('../_chunks/dep-05df1ea3.js');
require('../select/base/PopupContent.js');
require('../select/hooks/usePanelVirtualScroll.js');
require('../hooks/useVirtualScroll.js');
require('../input-number/index.js');
require('../input-number/InputNumber.js');
require('../button/index.js');
require('../button/Button.js');
require('../button/defaultProps.js');
require('../input-number/useInputNumber.js');
require('../_chunks/dep-04deac0d.js');
require('../_chunks/dep-8f62fb96.js');
require('../hooks/useCommonClassName.js');
require('../input-number/defaultProps.js');
require('../input-adornment/InputAdornment.js');
require('../pagination/hooks/useBoundaryJumper.js');
require('../pagination/hooks/usePrevNextJumper.js');
require('../pagination/hooks/usePageNumber.js');
require('../pagination/hooks/useTotal.js');
require('../pagination/validators.js');
require('../pagination/defaultProps.js');
require('../_chunks/dep-5c3139cb.js');
require('../_chunks/dep-2e455d91.js');
require('../_chunks/dep-894a5ece.js');
require('../pagination/PaginationMini.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var TransferList = function TransferList(props) {
  var className = props.className,
    style = props.style,
    data = props.data,
    _props$search = props.search,
    search = _props$search === void 0 ? false : _props$search,
    _props$checked = props.checked,
    checked = _props$checked === void 0 ? [] : _props$checked,
    empty = props.empty,
    title = props.title,
    footer = props.footer,
    content = props.content,
    onCheckbox = props.onCheckbox,
    onSearch = props.onSearch,
    _props$disabled = props.disabled,
    disabled = _props$disabled === void 0 ? false : _props$disabled,
    pagination = props.pagination,
    transferItem = props.transferItem,
    treeNode = props.tree,
    showCheckAll = props.showCheckAll,
    listType = props.listType;
  var notDisabledData = !treeNode ? data.filter(function (item) {
    return !item.disabled;
  }) : transfer_utils.getLeafNodes(data).filter(function (item) {
    return !item.disabled;
  });
  var _useState = React.useState(1),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    currentPage = _useState2[0],
    setCurrentPage = _useState2[1];
  var _useMemo = React.useMemo(function () {
      var activeLen = checked.length;
      var notLen = notDisabledData.length;
      return [activeLen && activeLen < notLen, activeLen && activeLen === notLen];
    }, [checked, notDisabledData]),
    _useMemo2 = slicedToArray._slicedToArray(_useMemo, 2),
    indeterminate = _useMemo2[0],
    allChecked = _useMemo2[1];
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      SearchIcon: tdesignIconsReact.SearchIcon
    }),
    SearchIcon = _useGlobalIcon.SearchIcon;
  var CLASSPREFIX = "".concat(classPrefix, "-transfer__list");
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("transfer"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var handleCheckbox = function handleCheckbox(vals) {
    if (isFunction.isFunction(onCheckbox)) onCheckbox(vals);
  };
  var handleAllCheckbox = function handleAllCheckbox(checked2) {
    if (isFunction.isFunction(onCheckbox)) onCheckbox(checked2 ? notDisabledData.map(function (item) {
      return item.value;
    }) : []);
  };
  var HeaderCmp = function HeaderCmp() {
    var total = treeNode ? transfer_utils.getLeafNodes(data).length : data.length;
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(CLASSPREFIX, "-header")
    }, /* @__PURE__ */React__default["default"].createElement("div", null, showCheckAll ? /* @__PURE__ */React__default["default"].createElement(checkbox_index.Checkbox, {
      indeterminate: indeterminate,
      checked: allChecked,
      disabled: disabled,
      onChange: handleAllCheckbox
    }) : null, /* @__PURE__ */React__default["default"].createElement("span", null, t(local.title, {
      checked: checked.length,
      total: total
    }))), /* @__PURE__ */React__default["default"].createElement("span", null, title));
  };
  var SearchCmp = function SearchCmp() {
    return search ? /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-transfer__search-wrapper")
    }, /* @__PURE__ */React__default["default"].createElement(input_index.Input, {
      placeholder: local.placeholder,
      suffixIcon: /* @__PURE__ */React__default["default"].createElement(SearchIcon, null),
      onChange: onSearch
    })) : null;
  };
  var viewData = React.useMemo(function () {
    if (!isEmpty.isEmpty(pagination)) {
      var pageSize = pagination.pageSize || 10;
      var start = (currentPage - 1) * pageSize;
      return data.slice(start, start + pageSize);
    }
    return data;
  }, [currentPage, data, pagination]);
  var EmptyCmp = function EmptyCmp() {
    return isString.isString(empty) ? /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-transfer__empty")
    }, /* @__PURE__ */React__default["default"].createElement("span", null, empty || "\u6682\u65E0\u6570\u636E")) : empty;
  };
  var contentCmp = function contentCmp() {
    if (typeof treeNode === "function") {
      return treeNode({
        data: viewData,
        value: checked,
        onChange: handleCheckbox
      });
    }
    if (typeof content === "function") {
      return content({
        data: viewData
      });
    }
    return /* @__PURE__ */React__default["default"].createElement(checkbox_index.Checkbox.Group, {
      value: checked,
      onChange: handleCheckbox,
      disabled: disabled
    }, viewData.map(function (item, index) {
      return /* @__PURE__ */React__default["default"].createElement(checkbox_index.Checkbox, {
        key: item.value,
        value: item.value,
        disabled: item.disabled,
        className: "".concat(CLASSPREFIX, "-item")
      }, /* @__PURE__ */React__default["default"].createElement("span", null, transferItem ? _util_parseTNode.parseContentTNode(transferItem, {
        data: item,
        index: index,
        type: listType
      }) : item.label));
    }));
  };
  var BodyCmp = function BodyCmp() {
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: classNames__default["default"]("".concat(CLASSPREFIX, "-body"), defineProperty._defineProperty({}, "".concat(CLASSPREFIX, "--with-search"), search))
    }, SearchCmp(), /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(CLASSPREFIX, "-content narrow-scrollbar")
    }, viewData.length ? contentCmp() : EmptyCmp()));
  };
  var PaginationCmp = function PaginationCmp() {
    var handleCurrentPagination = function handleCurrentPagination(current) {
      setCurrentPage(current);
      if (isFunction.isFunction(pagination.onPageChange)) {
        pagination.onPageChange(current);
      }
    };
    return pagination ? /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(CLASSPREFIX, "-pagination")
    }, /* @__PURE__ */React__default["default"].createElement(pagination_index.Pagination, {
      size: "small",
      theme: "simple",
      pageSizeOptions: [],
      total: data.length,
      totalContent: false,
      onCurrentChange: handleCurrentPagination,
      pageSize: pagination.pageSize
    })) : null;
  };
  var footerCmp = function footerCmp() {
    return !isEmpty.isEmpty(footer) ? /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-transfer__footer")
    }, footer) : null;
  };
  return /* @__PURE__ */React__default["default"].createElement("div", {
    style: style,
    className: classNames__default["default"](CLASSPREFIX, className)
  }, HeaderCmp(), BodyCmp(), PaginationCmp(), footerCmp());
};

exports["default"] = TransferList;
