// tdesign-react/cjs/_chunks/dep-83710d84.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

var defineProperty = require('./dep-2b5ba78d.js');
var React = require('react');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var common_Check = require('../common/Check.js');
var toConsumableArray = require('./dep-1d0dbfc5.js');
var slicedToArray = require('./dep-1a3dddaf.js');
var objectWithoutProperties = require('./dep-1c586f5c.js');
var _typeof = require('./dep-c4d1ce0c.js');
var classNames = require('classnames');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var checkbox_defaultProps = require('../checkbox/defaultProps.js');
var isNumber = require('./dep-f0e884b9.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["onChange", "disabled", "className", "style", "children", "max", "options"];
function ownKeys$1(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread$1(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys$1(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys$1(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var getCheckboxValue = function getCheckboxValue(v) {
  switch (_typeof._typeof(v)) {
    case "number":
      return v;
    case "string":
      return v;
    case "object":
      {
        var vs = v;
        return vs.value;
      }
    default:
      return void 0;
  }
};
var CheckboxGroup = function CheckboxGroup(props) {
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _useDefaultProps = hooks_useDefaultProps["default"](props, checkbox_defaultProps.checkboxGroupDefaultProps),
    onChange = _useDefaultProps.onChange,
    disabled = _useDefaultProps.disabled,
    className = _useDefaultProps.className,
    style = _useDefaultProps.style,
    children = _useDefaultProps.children,
    max = _useDefaultProps.max,
    _useDefaultProps$opti = _useDefaultProps.options,
    options = _useDefaultProps$opti === void 0 ? [] : _useDefaultProps$opti,
    resetProps = objectWithoutProperties._objectWithoutProperties(_useDefaultProps, _excluded);
  var readOnly = resetProps.readOnly || resetProps.readonly;
  var intervalOptions = Array.isArray(options) && options.length > 0 ? options : React__default["default"].Children.map(children, function (child) {
    var _child$type;
    return (child === null || child === void 0 || (_child$type = child.type) === null || _child$type === void 0 ? void 0 : _child$type.displayName) === Checkbox.displayName && child.props;
  }) || [];
  var optionsWithoutCheckAll = intervalOptions.filter(function (t) {
    return _typeof._typeof(t) !== "object" || !t.checkAll;
  });
  optionsWithoutCheckAll.forEach(function (v) {
    getCheckboxValue(v);
  });
  var _useMemo = React.useMemo(function () {
      var enabledValues2 = [];
      var disabledValues2 = [];
      optionsWithoutCheckAll.forEach(function (option) {
        var isOptionDisabled = _typeof._typeof(option) === "object" && (option.disabled || option.readOnly || option.readonly);
        var value = getCheckboxValue(option);
        if (isOptionDisabled || disabled || readOnly) {
          disabledValues2.push(value);
        } else {
          enabledValues2.push(value);
        }
      });
      return {
        enabledValues: enabledValues2,
        disabledValues: disabledValues2
      };
    }, [optionsWithoutCheckAll, disabled, readOnly]),
    enabledValues = _useMemo.enabledValues,
    disabledValues = _useMemo.disabledValues;
  var _useControlled = hooks_useControlled["default"](props, "value", onChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    internalValue = _useControlled2[0],
    setInternalValue = _useControlled2[1];
  var _useState = React.useState(max),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    localMax = _useState2[0],
    setLocalMax = _useState2[1];
  var getCheckedSet = React.useCallback(function () {
    if (!Array.isArray(internalValue)) {
      return /* @__PURE__ */new Set([]);
    }
    return new Set([].concat(internalValue));
  }, [internalValue]);
  var checkedSet = React.useMemo(function () {
    return getCheckedSet();
  }, [getCheckedSet]);
  var indeterminate = React.useMemo(function () {
    var allValues = [].concat(toConsumableArray._toConsumableArray(enabledValues), toConsumableArray._toConsumableArray(disabledValues));
    var checkedCount = allValues.filter(function (value) {
      return checkedSet.has(value);
    }).length;
    return checkedCount > 0 && checkedCount < allValues.length;
  }, [checkedSet, enabledValues, disabledValues]);
  var checkAllChecked = React.useMemo(function () {
    var checkableValues = enabledValues.filter(function (value) {
      return checkedSet.has(value);
    });
    return enabledValues.length > 0 && checkableValues.length === enabledValues.length;
  }, [checkedSet, enabledValues]);
  React.useEffect(function () {
    if (!isNumber.isNumber(max)) {
      return;
    }
    if (max < checkedSet.size) {
      console.warn("[TDesign] max should be less than the length of value, change is invalid");
    } else {
      setLocalMax(max);
    }
  }, [max, checkedSet]);
  var context = {
    inject: function inject(checkProps) {
      if (typeof checkProps.checked !== "undefined") {
        return checkProps;
      }
      var checkValue = checkProps.value;
      return _objectSpread$1(_objectSpread$1({}, checkProps), {}, {
        checked: checkProps.checkAll ? checkAllChecked : checkedSet.has(checkValue),
        indeterminate: checkProps.checkAll ? indeterminate : checkProps.indeterminate,
        disabled: checkProps.disabled || disabled || checkedSet.size >= localMax && !checkedSet.has(checkValue),
        readOnly: checkProps.readonly || readOnly,
        onChange: function onChange(checked, _ref) {
          var e = _ref.e;
          if (typeof checkProps.onChange === "function") {
            checkProps.onChange(checked, {
              e: e
            });
          }
          var checkedSet2 = getCheckedSet();
          if (checkProps.checkAll) {
            var checkedEnabledValues = enabledValues.filter(function (value) {
              return checkedSet2.has(value);
            });
            var allEnabledChecked = enabledValues.length > 0 && checkedEnabledValues.length === enabledValues.length;
            if (!allEnabledChecked) {
              enabledValues.forEach(function (value) {
                if (!checkedSet2.has(value)) {
                  checkedSet2.add(value);
                }
              });
            } else {
              enabledValues.forEach(function (value) {
                if (checkedSet2.has(value)) {
                  checkedSet2["delete"](value);
                }
              });
            }
          } else if (checked) {
            if (checkedSet2.size >= localMax && isNumber.isNumber(max)) return;
            checkedSet2.add(checkValue);
          } else {
            checkedSet2["delete"](checkValue);
          }
          var currentOptionChecked = optionsWithoutCheckAll.find(function (item) {
            return item.value === checkValue;
          });
          setInternalValue(Array.from(checkedSet2), {
            e: e,
            current: checkProps.checkAll ? void 0 : checkValue,
            type: checked ? "check" : "uncheck",
            option: checkProps.checkAll ? void 0 : currentOptionChecked
          });
        }
      });
    }
  };
  var useOptions = Array.isArray(options) && options.length !== 0;
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-checkbox-group"), className),
    style: style
  }, /* @__PURE__ */React__default["default"].createElement(common_Check.CheckContext.Provider, {
    value: context
  }, useOptions ? options.map(function (v, index) {
    switch (_typeof._typeof(v)) {
      case "string":
        return /* @__PURE__ */React__default["default"].createElement(Checkbox, {
          key: index,
          label: v,
          value: v
        }, v);
      case "number":
        {
          return /* @__PURE__ */React__default["default"].createElement(Checkbox, {
            key: index,
            label: v,
            value: v
          }, String(v));
        }
      case "object":
        {
          var vs = v;
          return vs.checkAll ? /* @__PURE__ */React__default["default"].createElement(Checkbox, _objectSpread$1(_objectSpread$1({}, vs), {}, {
            key: "checkAll_".concat(index),
            indeterminate: indeterminate
          })) : /* @__PURE__ */React__default["default"].createElement(Checkbox, _objectSpread$1(_objectSpread$1({}, vs), {}, {
            key: index,
            disabled: vs.disabled || disabled,
            readOnly: vs.readOnly || vs.readonly || readOnly
          }));
        }
      default:
        return null;
    }
  }) : children));
};
CheckboxGroup.displayName = "CheckboxGroup";

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Checkbox = _util_forwardRefWithStatics["default"](function (props, ref) {
  return /* @__PURE__ */React__default["default"].createElement(common_Check["default"], _objectSpread({
    ref: ref,
    type: "checkbox"
  }, hooks_useDefaultProps["default"](props, checkbox_defaultProps.checkboxDefaultProps)));
}, {
  Group: CheckboxGroup
});
Checkbox.displayName = "Checkbox";

exports.Checkbox = Checkbox;
exports.CheckboxGroup = CheckboxGroup;
