// tdesign-react/cjs/_chunks/dep-a7cf54af.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

var slicedToArray = require('./dep-1a3dddaf.js');
var objectWithoutProperties = require('./dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var _util_noop = require('../_util/noop.js');
var hooks_usePersistFn = require('../hooks/usePersistFn.js');
var message_MessageClose = require('../message/MessageClose.js');
var message_MessageIcon = require('../message/MessageIcon.js');
var message_useMessageClass = require('../message/useMessageClass.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var DISTANCE = "32px";
var PLACEMENT_OFFSET = {
  center: {
    left: "50%",
    top: "50%",
    transform: "translateX(-50%) translateY(-50%)"
  },
  left: {
    left: DISTANCE,
    top: "50%",
    transform: "translateY(-50%)"
  },
  bottom: {
    bottom: DISTANCE,
    left: "50%",
    transform: "translateX(-50%)"
  },
  right: {
    right: DISTANCE,
    top: "50%",
    transform: "translateY(-50%)"
  },
  top: {
    top: DISTANCE,
    left: "50%",
    transform: "translateX(-50%)"
  },
  "top-left": {
    left: DISTANCE,
    top: DISTANCE
  },
  "top-right": {
    right: DISTANCE,
    top: DISTANCE
  },
  "bottom-left": {
    left: DISTANCE,
    bottom: DISTANCE
  },
  "bottom-right": {
    right: DISTANCE,
    bottom: DISTANCE
  }
};
var PLACEMENT_LIST = Object.keys(PLACEMENT_OFFSET);

var ANIMATION_OPTION = {
  duration: 200,
  easing: "linear"
};
function setDomStyleAfterAnimation(dom, styleAfterAnimation) {
  var keys = Object.keys(styleAfterAnimation);
  for (var i = 0; i < keys.length; i += 1) {
    var key = keys[i];
    dom.style[key] = styleAfterAnimation[key];
  }
}
function getFadeInKeyframes(placement, offsetWidth, offsetHeight) {
  if (!PLACEMENT_LIST.includes(placement)) return null;
  if (["top-left", "left", "bottom-left"].includes(placement)) {
    return [{
      opacity: 0,
      marginLeft: "-".concat(offsetWidth, "px")
    }, {
      opacity: 1,
      marginLeft: "0"
    }];
  }
  if (["top-right", "right", "bottom-right"].includes(placement)) {
    return [{
      opacity: 0,
      marginRight: "-".concat(offsetWidth, "px")
    }, {
      opacity: 1,
      marginRight: "0"
    }];
  }
  if (["top", "center"].includes(placement)) {
    return [{
      opacity: 0,
      marginTop: "-".concat(offsetHeight, "px")
    }, {
      opacity: 1,
      marginTop: "0"
    }];
  }
  if (["bottom"].includes(placement)) {
    return [{
      opacity: 0,
      transform: "translate3d(0, ".concat(offsetHeight, "px, 0)")
    }, {
      opacity: 1,
      transform: "translate3d(0, 0, 0)"
    }];
  }
  return null;
}
function getFadeOutKeyframes(placement, offsetHeight) {
  if (!PLACEMENT_LIST.includes(placement)) return null;
  if (["bottom-left", "bottom", "bottom-right"].includes(placement)) {
    var marginOffset2 = "".concat(offsetHeight, "px");
    return [{
      opacity: 1,
      marginTop: "0px"
    }, {
      opacity: 0,
      marginTop: marginOffset2
    }];
  }
  var marginOffset = "-".concat(offsetHeight, "px");
  return [{
    opacity: 1,
    marginTop: "0px"
  }, {
    opacity: 0,
    marginTop: marginOffset
  }];
}
function fadeIn(dom, placement) {
  var _dom$animate;
  if (!dom) return;
  var offsetWidth = (dom === null || dom === void 0 ? void 0 : dom.offsetWidth) || 0;
  var offsetHeight = (dom === null || dom === void 0 ? void 0 : dom.offsetHeight) || 0;
  var fadeInKeyframes = getFadeInKeyframes(placement, offsetWidth, offsetHeight);
  if (!fadeInKeyframes) return;
  var styleAfterFadeIn = fadeInKeyframes[fadeInKeyframes.length - 1];
  setDomStyleAfterAnimation(dom, styleAfterFadeIn);
  (_dom$animate = dom.animate) === null || _dom$animate === void 0 || _dom$animate.call(dom, fadeInKeyframes, ANIMATION_OPTION);
}
function fadeOut(dom, placement, onFinish) {
  if (!dom) return;
  var offsetHeight = dom.offsetHeight || 0;
  var fadeOutKeyframes = getFadeOutKeyframes(placement, offsetHeight);
  if (!fadeOutKeyframes) {
    onFinish();
    return;
  }
  var styleAfterFadeOut = fadeOutKeyframes[fadeOutKeyframes.length - 1];
  setDomStyleAfterAnimation(dom, styleAfterFadeOut);
  var animation = dom.animate && dom.animate(fadeOutKeyframes, ANIMATION_OPTION);
  if (animation) {
    animation.onfinish = function () {
      dom.style.display = "none";
      onFinish();
    };
  } else {
    dom.style.display = "none";
    onFinish();
  }
}

var _excluded = ["placement"];
var MessageComponent = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var placement = props.placement,
    otherProps = objectWithoutProperties._objectWithoutProperties(props, _excluded);
  var _useMessageClass = message_useMessageClass.useMessageClass(),
    tdMessagePrefix = _useMessageClass.tdMessagePrefix,
    tdClassIsGenerator = _useMessageClass.tdClassIsGenerator;
  var _otherProps$theme = otherProps.theme,
    theme = _otherProps$theme === void 0 ? "info" : _otherProps$theme,
    className = otherProps.className,
    children = otherProps.children,
    style = otherProps.style,
    _otherProps$icon = otherProps.icon,
    icon = _otherProps$icon === void 0 ? true : _otherProps$icon,
    content = otherProps.content,
    closeBtn = otherProps.closeBtn,
    _otherProps$onCloseBt = otherProps.onCloseBtnClick,
    onCloseBtnClick = _otherProps$onCloseBt === void 0 ? _util_noop["default"] : _otherProps$onCloseBt,
    _otherProps$onDuratio = otherProps.onDurationEnd,
    onDurationEnd = _otherProps$onDuratio === void 0 ? _util_noop["default"] : _otherProps$onDuratio,
    _otherProps$onClose = otherProps.onClose,
    onClose = _otherProps$onClose === void 0 ? _util_noop["default"] : _otherProps$onClose,
    duration = otherProps.duration;
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    isHovering = _useState2[0],
    setIsHovering = _useState2[1];
  var onCloseFn = hooks_usePersistFn.usePersistFn(onClose);
  var onDurationEndFn = hooks_usePersistFn.usePersistFn(onDurationEnd);
  var onCloseBtnClickFn = hooks_usePersistFn.usePersistFn(onCloseBtnClick);
  function handleCloseBtnClick(e) {
    onCloseBtnClickFn(e);
    onCloseFn({
      trigger: "close-click"
    });
  }
  React.useEffect(function () {
    if (ref && "current" in ref && ref.current && placement) {
      fadeIn(ref.current, placement);
    }
  }, []);
  React.useEffect(function () {
    if (!isHovering && duration > 0) {
      var timer = setTimeout(function () {
        onDurationEndFn();
        onCloseFn({
          trigger: "duration-end"
        });
      }, duration);
      return function () {
        clearTimeout(timer);
      };
    }
  }, [duration, isHovering, onCloseFn, onDurationEndFn]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: ref,
    key: "message",
    style: style,
    className: classNames__default["default"](className, "".concat(tdMessagePrefix), tdClassIsGenerator(theme), closeBtn ? tdClassIsGenerator("closable") : ""),
    onMouseEnter: function onMouseEnter() {
      return setIsHovering(true);
    },
    onMouseLeave: function onMouseLeave() {
      return setIsHovering(false);
    }
  }, icon === true ? /* @__PURE__ */React__default["default"].createElement(message_MessageIcon["default"], {
    theme: theme
  }) : icon, content ? content : children, /* @__PURE__ */React__default["default"].createElement(message_MessageClose["default"], {
    closeBtn: closeBtn,
    onCloseBtnClick: handleCloseBtnClick
  }));
});
MessageComponent.displayName = "Message";

exports.MessageComponent = MessageComponent;
exports.PLACEMENT_OFFSET = PLACEMENT_OFFSET;
exports.fadeOut = fadeOut;
