// tdesign-react/cjs/rate/Rate.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var toConsumableArray = require('../_chunks/dep-1d0dbfc5.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var tooltip_index = require('../tooltip/index.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var rate_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../tooltip/Tooltip.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-ecfe656e.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../hooks/useAnimation.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../hooks/useAttach.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../_chunks/dep-876d090e.js');
require('../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/listener.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');
require('../popup/utils/transition.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../popup/PopupPlugin.js');
require('../_chunks/dep-e37cb830.js');
require('../_util/react-render.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../tooltip/defaultProps.js');
require('../tooltip/TooltipLite.js');
require('../hooks/useSwitch.js');
require('../hooks/usePersistFn.js');
require('../_chunks/dep-2a85392e.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["icon"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var RateIcon = function RateIcon(_ref) {
  var icon = _ref.icon,
    props = objectWithoutProperties._objectWithoutProperties(_ref, _excluded);
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      StarFilledIcon: tdesignIconsReact.StarFilledIcon
    }),
    StarFilledIcon = _useGlobalIcon.StarFilledIcon;
  if (/*#__PURE__*/React__default["default"].isValidElement(icon)) {
    return /*#__PURE__*/React__default["default"].cloneElement(icon, props);
  }
  return /* @__PURE__ */React__default["default"].createElement(StarFilledIcon, _objectSpread({}, props));
};
var Rate = /*#__PURE__*/React__default["default"].forwardRef(function (originalProps, ref) {
  var props = hooks_useDefaultProps["default"](originalProps, rate_defaultProps.rateDefaultProps);
  var allowHalf = props.allowHalf,
    color = props.color,
    count = props.count,
    disabled = props.disabled,
    gap = props.gap,
    showText = props.showText,
    size = props.size,
    icon = props.icon,
    className = props.className,
    style = props.style,
    onChange = props.onChange,
    texts = props.texts,
    clearable = props.clearable;
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("rate"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    locale = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var displayTexts = texts.length ? texts : t(locale.rateText);
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _useControlled = hooks_useControlled["default"](props, "value", onChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    starValue = _useControlled2[0],
    setStarValue = _useControlled2[1];
  var _useState = React.useState(void 0),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    hoverValue = _useState2[0],
    setHoverValue = _useState2[1];
  var displayValue = hoverValue || starValue;
  var rootRef = React__default["default"].useRef(null);
  var activeColor = Array.isArray(color) ? color[0] : color;
  var defaultColor = Array.isArray(color) ? color[1] : "var(--td-bg-color-component)";
  var getStarValue = function getStarValue(event, index) {
    if (allowHalf) {
      var rootNode = rootRef.current;
      var _rootNode$getBounding = rootNode.getBoundingClientRect(),
        left = _rootNode$getBounding.left;
      var firstStar = rootNode.firstChild;
      var _firstStar$getBoundin = firstStar.getBoundingClientRect(),
        width = _firstStar$getBoundin.width;
      var clientX = event.clientX;
      var starMiddle = width * (index - 0.5) + gap * (index - 1);
      if (clientX - left >= starMiddle) {
        return index;
      }
      if (clientX - left < starMiddle) {
        return index - 0.5;
      }
    }
    return index;
  };
  var mouseEnterHandler = function mouseEnterHandler(event, index) {
    if (disabled) {
      return;
    }
    setHoverValue(getStarValue(event, index));
  };
  var mouseLeaveHandler = function mouseLeaveHandler() {
    if (disabled) {
      return;
    }
    setHoverValue(void 0);
  };
  var clickHandler = function clickHandler(event, index) {
    if (disabled) {
      return;
    }
    var value = getStarValue(event, index);
    if (clearable && value === starValue) {
      value = 0;
      setHoverValue(void 0);
    }
    setStarValue(value);
  };
  var getStarCls = function getStarCls(index) {
    if (allowHalf && index + 0.5 === displayValue) {
      return "".concat(classPrefix, "-rate__item--half");
    }
    if (index >= displayValue) {
      return "";
    }
    if (index < displayValue) {
      return "".concat(classPrefix, "-rate__item--full");
    }
  };
  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: ref,
    style: style,
    className: classNames__default["default"]("".concat(classPrefix, "-rate"), className),
    onMouseLeave: mouseLeaveHandler
  }, /* @__PURE__ */React__default["default"].createElement("ul", {
    className: "".concat(classPrefix, "-rate__list"),
    style: {
      gap: gap
    },
    ref: rootRef
  }, toConsumableArray._toConsumableArray(Array(count)).map(function (_, index) {
    return /* @__PURE__ */React__default["default"].createElement("li", {
      key: index,
      className: classNames__default["default"]("".concat(classPrefix, "-rate__item"), getStarCls(index)),
      onClick: function onClick(event) {
        return clickHandler(event, index + 1);
      },
      onMouseMove: function onMouseMove(event) {
        return mouseEnterHandler(event, index + 1);
      }
    }, showText ? /* @__PURE__ */React__default["default"].createElement(tooltip_index.TooltipLite, {
      key: index,
      content: displayTexts[displayValue - 1]
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-rate__star-top")
    }, /* @__PURE__ */React__default["default"].createElement(RateIcon, {
      size: size,
      color: activeColor,
      icon: icon
    })), /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-rate__star-bottom")
    }, /* @__PURE__ */React__default["default"].createElement(RateIcon, {
      size: size,
      color: defaultColor,
      icon: icon
    }))) : /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-rate__star-top")
    }, /* @__PURE__ */React__default["default"].createElement(RateIcon, {
      size: size,
      color: activeColor,
      icon: icon
    })), /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-rate__star-bottom")
    }, /* @__PURE__ */React__default["default"].createElement(RateIcon, {
      size: size,
      color: defaultColor,
      icon: icon
    }))));
  })), showText && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-rate__text")
  }, displayTexts[displayValue - 1]));
});
Rate.displayName = "Rate";

exports["default"] = Rate;
