// tdesign-react/cjs/comment/Comment.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var classNames = require('classnames');
var button_index = require('../button/index.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var comment_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../button/Button.js');
require('../_chunks/dep-2b5ba78d.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-1c586f5c.js');
require('../hooks/useDomRefCallback.js');
require('../hooks/useRipple.js');
require('../_chunks/dep-064e4dba.js');
require('../hooks/useAnimation.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-f0e884b9.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../button/defaultProps.js');
require('../_util/parseTNode.js');
require('../_chunks/dep-b289b0be.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var Comment = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _useDefaultProps = hooks_useDefaultProps["default"](props, comment_defaultProps.commentDefaultProps),
    actions = _useDefaultProps.actions,
    author = _useDefaultProps.author,
    avatar = _useDefaultProps.avatar,
    content = _useDefaultProps.content,
    datetime = _useDefaultProps.datetime,
    reply = _useDefaultProps.reply,
    quote = _useDefaultProps.quote,
    className = _useDefaultProps.className,
    style = _useDefaultProps.style;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var avatarElement = avatar ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-comment__avatar")
  }, typeof avatar === "string" ? /* @__PURE__ */React__default["default"].createElement("img", {
    src: avatar,
    alt: "",
    className: "".concat(classPrefix, "-comment__avatar-image")
  }) : avatar) : null;
  var authorDatetimeContent = (author || datetime) && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-comment__author")
  }, author && /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-comment__name")
  }, author), datetime && /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-comment__time")
  }, datetime));
  var quoteElement = quote ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-comment__quote")
  }, quote) : null;
  var actionsElement = actions && actions.length ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-comment__actions")
  }, actions.map(function (action, index) {
    return /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
      key: "action-".concat(index),
      size: "small",
      variant: "text"
    }, action);
  })) : null;
  var contentElement = /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-comment__content")
  }, authorDatetimeContent, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-comment__detail")
  }, content), quoteElement, actionsElement);
  var replyElement = reply ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-comment__reply"))
  }, reply) : null;
  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: ref,
    style: style,
    className: classNames__default["default"]("".concat(classPrefix, "-comment"), className)
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-comment__inner")
  }, avatarElement, contentElement), replyElement);
});
Comment.displayName = "Comment";

exports["default"] = Comment;
