// tdesign-react/cjs/date-picker/panel/RangePanel.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var format = require('../../_chunks/dep-06ac19a1.js');
var log = require('../../_chunks/dep-b289b0be.js');
var hooks_useConfig = require('../../hooks/useConfig.js');
var datePicker_panel_PanelContent = require('./PanelContent.js');
var datePicker_panel_ExtraContent = require('./ExtraContent.js');
var datePicker_hooks_useTableData = require('../hooks/useTableData.js');
var datePicker_hooks_useDisableDate = require('../hooks/useDisableDate.js');
var hooks_useDefaultProps = require('../../hooks/useDefaultProps.js');
var datePicker_utils = require('../utils.js');
var timePicker_index = require('../../time-picker/index.js');
var isArray = require('../../_chunks/dep-794b42ce.js');
var isFunction = require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('dayjs');
require('../../_chunks/dep-06bf4625.js');
require('../../_chunks/dep-bf18bb11.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-14013769.js');
require('../../_chunks/dep-043eee20.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../_chunks/dep-2d5cf7d6.js');
require('../base/Header.js');
require('../../_chunks/dep-1d0dbfc5.js');
require('../../_chunks/dep-3c2d5e8c.js');
require('../../locale/LocalReceiver.js');
require('../../config-provider/ConfigProvider.js');
require('../../_chunks/dep-4a562e04.js');
require('../../_chunks/dep-a5b4810f.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-ea361d31.js');
require('../../pagination/index.js');
require('../../pagination/Pagination.js');
require('../../_chunks/dep-1c586f5c.js');
require('../../_util/noop.js');
require('../../hooks/useControlled.js');
require('../../_chunks/dep-eb1b216c.js');
require('../../_chunks/dep-b529a0d7.js');
require('../../_chunks/dep-bcd16957.js');
require('../../_chunks/dep-b4442734.js');
require('../../_chunks/dep-876d090e.js');
require('../../select/index.js');
require('../../select/base/Select.js');
require('../../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../../_util/helper.js');
require('../../_chunks/dep-e44b03ec.js');
require('../../_util/parseTNode.js');
require('../../_util/ref.js');
require('react-is');
require('../../_util/isFragment.js');
require('../../common/FakeArrow.js');
require('../../_chunks/dep-b011b5d3.js');
require('../../_chunks/dep-f0e884b9.js');
require('../../loading/index.js');
require('../../loading/Loading.js');
require('../../_chunks/dep-68f5fe7b.js');
require('../../common/Portal.js');
require('react-dom');
require('../../hooks/useLayoutEffect.js');
require('../../loading/defaultProps.js');
require('../../loading/gradient.js');
require('../../_chunks/dep-064e4dba.js');
require('../../hooks/useDomRefCallback.js');
require('../../loading/plugin.js');
require('../../_util/react-render.js');
require('../../_chunks/dep-e37cb830.js');
require('../../common/PluginContainer.js');
require('../../select-input/index.js');
require('../../select-input/SelectInput.js');
require('../../popup/index.js');
require('../../popup/Popup.js');
require('react-transition-group');
require('../../hooks/useAnimation.js');
require('../../hooks/useAttach.js');
require('../../hooks/useMutationObserver.js');
require('../../hooks/useLatest.js');
require('../../_chunks/dep-4e876c4b.js');
require('../../_chunks/dep-df8d2fc8.js');
require('../../_chunks/dep-66f91e5d.js');
require('../../_chunks/dep-d3af8eb7.js');
require('../../_chunks/dep-eab994e8.js');
require('../../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../../hooks/useWindowSize.js');
require('../../popup/defaultProps.js');
require('../../popup/hooks/useTrigger.js');
require('../../_util/listener.js');
require('../../popup/utils/transition.js');
require('../../popup/PopupPlugin.js');
require('../../select-input/defaultProps.js');
require('../../select-input/useMultiple.js');
require('../../tag-input/index.js');
require('../../tag-input/TagInput.js');
require('tdesign-icons-react');
require('../../hooks/useDragSorter.js');
require('../../hooks/useEventCallback.js');
require('../../hooks/useGlobalIcon.js');
require('../../input/index.js');
require('../../input/Input.js');
require('../../input/defaultProps.js');
require('../../input/InputGroup.js');
require('../../input/useLengthLimit.js');
require('../../tag-input/defaultProps.js');
require('../../tag-input/useHover.js');
require('../../tag-input/useTagList.js');
require('../../tag/index.js');
require('../../tag/Tag.js');
require('../../_chunks/dep-d14bfcfa.js');
require('../../tag/defaultProps.js');
require('../../tag/CheckTag.js');
require('../../_chunks/dep-dd0f5f6b.js');
require('../../tag/CheckTagGroup.js');
require('../../tag-input/useTagScroll.js');
require('../../select-input/useOverlayInnerStyle.js');
require('../../hooks/useInnerPopupVisible.js');
require('../../select-input/useSingle.js');
require('../../_chunks/dep-f7bf3d17.js');
require('../../_chunks/dep-5e17a073.js');
require('../../_chunks/dep-ca918872.js');
require('../../_chunks/dep-13de2fef.js');
require('../../_chunks/dep-63308b84.js');
require('../../_chunks/dep-048e4314.js');
require('../../select/defaultProps.js');
require('../../select/hooks/useKeyboardControl.js');
require('../../_chunks/dep-bdc97672.js');
require('../../hooks/useRipple.js');
require('../../select/base/OptionGroup.js');
require('../../_chunks/dep-05df1ea3.js');
require('../../select/base/PopupContent.js');
require('../../select/hooks/usePanelVirtualScroll.js');
require('../../hooks/useVirtualScroll.js');
require('../../input-number/index.js');
require('../../input-number/InputNumber.js');
require('../../button/index.js');
require('../../button/Button.js');
require('../../button/defaultProps.js');
require('../../input-number/useInputNumber.js');
require('../../_chunks/dep-04deac0d.js');
require('../../_chunks/dep-8f62fb96.js');
require('../../hooks/useCommonClassName.js');
require('../../input-number/defaultProps.js');
require('../../input-adornment/InputAdornment.js');
require('../../pagination/hooks/useBoundaryJumper.js');
require('../../pagination/hooks/usePrevNextJumper.js');
require('../../pagination/hooks/usePageNumber.js');
require('../../pagination/hooks/useTotal.js');
require('../../pagination/validators.js');
require('../../pagination/defaultProps.js');
require('../../_chunks/dep-5c3139cb.js');
require('../../_chunks/dep-2e455d91.js');
require('../../_chunks/dep-894a5ece.js');
require('../../pagination/PaginationMini.js');
require('../hooks/useSelectRange.js');
require('../../_chunks/dep-8227db70.js');
require('../../_chunks/dep-c990dbb4.js');
require('../base/Table.js');
require('../base/Cell.js');
require('../base/Footer.js');
require('../../_chunks/dep-4d78a412.js');
require('../../time-picker/TimePicker.js');
require('../../_chunks/dep-9220a4d6.js');
require('../../time-picker/hooks/useTimePickerTextConfig.js');
require('../../hooks/useDebounce.js');
require('../../hooks/usePersistFn.js');
require('../../hooks/usePropsRef.js');
require('../../time-picker/TimeRangePicker.js');
require('../../range-input/index.js');
require('../../range-input/RangeInput.js');
require('../../range-input/defaultProps.js');
require('../../range-input/RangeInputPopup.js');
require('../../time-picker/panel/TimePickerPanel.js');
require('../../time-picker/defaultProps.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var RangePanel = /*#__PURE__*/React.forwardRef(function (originalProps, ref) {
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix,
    globalDatePickerConfig = _useConfig.datePicker;
  var panelName = "".concat(classPrefix, "-date-range-picker__panel");
  var _useState = React.useState(true),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    isDateRangeContent = _useState2[0],
    toggleDateRangeContent = _useState2[1];
  var props = hooks_useDefaultProps["default"](originalProps, {
    mode: "date",
    panelPreselection: true,
    enableTimePicker: false,
    presetsPlacement: "bottom",
    needConfirm: true
  });
  var _props$value = props.value,
    value = _props$value === void 0 ? [] : _props$value,
    _props$hoverValue = props.hoverValue,
    hoverValue = _props$hoverValue === void 0 ? [] : _props$hoverValue,
    mode = props.mode,
    presets = props.presets,
    enableTimePicker = props.enableTimePicker,
    presetsPlacement = props.presetsPlacement,
    disableDateFromProps = props.disableDate,
    _props$firstDayOfWeek = props.firstDayOfWeek,
    firstDayOfWeek = _props$firstDayOfWeek === void 0 ? globalDatePickerConfig.firstDayOfWeek : _props$firstDayOfWeek,
    isFirstValueSelected = props.isFirstValueSelected,
    needConfirm = props.needConfirm,
    style = props.style,
    className = props.className,
    activeIndex = props.activeIndex,
    year = props.year,
    month = props.month,
    range = props.range,
    cell = props.cell,
    _props$time = props.time,
    time = _props$time === void 0 ? [] : _props$time,
    panelPreselection = props.panelPreselection,
    onPresetClick = props.onPresetClick,
    cancelRangeSelectLimit = props.cancelRangeSelectLimit,
    _onClick = props.onClick,
    onConfirmClick = props.onConfirmClick,
    disableTime = props.disableTime,
    onTimePickerChange = props.onTimePickerChange,
    popupVisible = props.popupVisible,
    isSwitchTimeMode = props.isSwitchTimeMode;
  var _getDefaultFormat = format.getDefaultFormat({
      mode: props.mode,
      format: props.format,
      enableTimePicker: props.enableTimePicker
    }),
    format$1 = _getDefaultFormat.format,
    timeFormat = _getDefaultFormat.timeFormat;
  var disableDateOptions = datePicker_hooks_useDisableDate["default"]({
    disableDate: disableDateFromProps,
    mode: mode,
    format: format$1,
    start: isFirstValueSelected && activeIndex === 1 ? new Date(format.parseToDayjs(value[0], format$1, "start").toDate().setHours(0, 0, 0)) : void 0,
    end: isFirstValueSelected && activeIndex === 0 ? new Date(format.parseToDayjs(value[1], format$1).toDate().setHours(23, 59, 59)) : void 0
  });
  var rangeValue = React.useMemo(function () {
    var startRange = range;
    var endRange = range;
    if (isArray.isArray(range)) {
      if (range.length !== 2) {
        log.log.warn("DateRangePicker", "`range` length must be 2 when `range` is an array.");
      }
      var first = range[0];
      var second = range[1];
      if ((isArray.isArray(first) || isFunction.isFunction(first)) && (isArray.isArray(second) || isFunction.isFunction(second))) {
        startRange = first;
        endRange = second;
      }
    }
    return {
      start: startRange,
      end: endRange
    };
  }, [range]);
  var disableTimeOptions = function disableTimeOptions(h, m, s, ms) {
    if (!isFunction.isFunction(disableTime)) {
      return {};
    }
    var _ref = value || [],
      _ref2 = slicedToArray._slicedToArray(_ref, 2),
      startTime = _ref2[0],
      endTime = _ref2[1];
    return disableTime([datePicker_utils.parseToDateTime(startTime, format$1), datePicker_utils.parseToDateTime(endTime, format$1, [h, m, s, ms])], {
      partial: activeIndex === 0 ? "start" : "end"
    });
  };
  var _year = slicedToArray._slicedToArray(year, 2),
    startYear = _year[0],
    endYear = _year[1];
  var _month = slicedToArray._slicedToArray(month, 2),
    startMonth = _month[0],
    endMonth = _month[1];
  var hidePreselection = !panelPreselection && value.length === 2;
  var startTableData = datePicker_hooks_useTableData["default"](_objectSpread({
    isRange: true,
    start: value[0] ? format.parseToDayjs(value[0], format$1).toDate() : void 0,
    end: value[1] ? format.parseToDayjs(value[1], format$1).toDate() : void 0,
    hoverStart: !hidePreselection && hoverValue[0] ? format.parseToDayjs(hoverValue[0], format$1).toDate() : void 0,
    hoverEnd: !hidePreselection && hoverValue[1] ? format.parseToDayjs(hoverValue[1], format$1).toDate() : void 0,
    year: startYear,
    month: startMonth,
    range: rangeValue.start,
    mode: mode,
    firstDayOfWeek: firstDayOfWeek,
    cancelRangeSelectLimit: cancelRangeSelectLimit
  }, disableDateOptions));
  var endTableData = datePicker_hooks_useTableData["default"](_objectSpread({
    isRange: true,
    start: value[0] ? format.parseToDayjs(value[0], format$1).toDate() : void 0,
    end: value[1] ? format.parseToDayjs(value[1], format$1).toDate() : void 0,
    hoverStart: !hidePreselection && hoverValue[0] ? format.parseToDayjs(hoverValue[0], format$1).toDate() : void 0,
    hoverEnd: !hidePreselection && hoverValue[1] ? format.parseToDayjs(hoverValue[1], format$1).toDate() : void 0,
    year: mode === "year" && endYear - startYear <= 9 ? endYear + 9 : endYear,
    month: endMonth,
    range: rangeValue.end,
    mode: mode,
    firstDayOfWeek: firstDayOfWeek,
    cancelRangeSelectLimit: cancelRangeSelectLimit
  }, disableDateOptions));
  var panelContentProps = {
    mode: mode,
    format: format$1,
    cell: cell,
    firstDayOfWeek: firstDayOfWeek,
    popupVisible: popupVisible,
    enableTimePicker: props.enableTimePicker,
    isSwitchTimeMode: isSwitchTimeMode,
    timePickerProps: _objectSpread({
      disableTime: disableTimeOptions
    }, props.timePickerProps),
    onMonthChange: props.onMonthChange,
    onYearChange: props.onYearChange,
    onJumperClick: props.onJumperClick,
    onCellClick: props.onCellClick,
    onCellMouseEnter: props.onCellMouseEnter,
    onCellMouseLeave: props.onCellMouseLeave,
    onTimePickerChange: props.onTimePickerChange
  };
  var onTimeModeChange = function onTimeModeChange() {
    toggleDateRangeContent(function (prev) {
      return !prev;
    });
  };
  var renderPanelContent = function renderPanelContent() {
    if (!enableTimePicker || isSwitchTimeMode) {
      if (isDateRangeContent) return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement(datePicker_panel_PanelContent["default"], _objectSpread({
        key: "startPanel",
        partial: "start",
        year: startYear,
        month: startMonth,
        time: time[activeIndex],
        tableData: startTableData,
        value: value,
        range: rangeValue.start
      }, panelContentProps)), /* @__PURE__ */React__default["default"].createElement(datePicker_panel_PanelContent["default"], _objectSpread({
        key: "endPanel",
        partial: "end",
        year: endYear,
        month: endMonth,
        time: time[activeIndex],
        value: value,
        tableData: endTableData,
        internalYear: year,
        range: rangeValue.end
      }, panelContentProps)));
      return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement(timePicker_index.TimePickerPanel, _objectSpread({
        key: "startPanel",
        position: "start",
        format: timeFormat,
        value: time[0],
        onChange: function onChange(val) {
          return onTimePickerChange(val, {
            activeIndex: 0
          });
        }
      }, props.timePickerProps)), /* @__PURE__ */React__default["default"].createElement(timePicker_index.TimePickerPanel, _objectSpread({
        key: "endPanel",
        position: "end",
        format: timeFormat,
        value: time[1],
        onChange: function onChange(val) {
          return onTimePickerChange(val, {
            activeIndex: 1
          });
        }
      }, props.timePickerProps)));
    }
    return /* @__PURE__ */React__default["default"].createElement(datePicker_panel_PanelContent["default"], _objectSpread({
      key: "start",
      partial: activeIndex ? "end" : "start",
      year: activeIndex ? endYear : startYear,
      month: activeIndex ? endMonth : startMonth,
      time: activeIndex ? time[1] : time[0],
      value: value,
      tableData: activeIndex ? endTableData : startTableData,
      range: activeIndex ? rangeValue.end : rangeValue.start
    }, panelContentProps));
  };
  React.useEffect(function () {
    if (popupVisible && isSwitchTimeMode) {
      toggleDateRangeContent(true);
    }
  }, [popupVisible, isSwitchTimeMode]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: ref,
    style: style,
    className: classNames__default["default"](panelName, className, defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(panelName, "--direction-row"), ["left", "right"].includes(presetsPlacement)), "".concat(panelName, "--switch-mode"), isSwitchTimeMode)),
    onClick: function onClick(e) {
      return _onClick === null || _onClick === void 0 ? void 0 : _onClick({
        e: e
      });
    }
  }, ["top", "left"].includes(presetsPlacement) ? /* @__PURE__ */React__default["default"].createElement(datePicker_panel_ExtraContent["default"], {
    presets: presets,
    selectedValue: value[activeIndex],
    enableTimePicker: enableTimePicker,
    onPresetClick: onPresetClick,
    onConfirmClick: onConfirmClick,
    presetsPlacement: presetsPlacement,
    needConfirm: needConfirm,
    onTimeModeChange: onTimeModeChange,
    isDateRangeContent: isDateRangeContent,
    isSwitchTimeMode: isSwitchTimeMode
  }) : null, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(panelName, "-content-wrapper")
  }, renderPanelContent()), ["bottom", "right"].includes(presetsPlacement) ? /* @__PURE__ */React__default["default"].createElement(datePicker_panel_ExtraContent["default"], {
    presets: presets,
    selectedValue: value[activeIndex],
    enableTimePicker: enableTimePicker,
    onPresetClick: onPresetClick,
    onConfirmClick: onConfirmClick,
    onTimeModeChange: onTimeModeChange,
    isDateRangeContent: isDateRangeContent,
    isSwitchTimeMode: isSwitchTimeMode,
    presetsPlacement: presetsPlacement,
    needConfirm: needConfirm
  }) : null);
});
RangePanel.displayName = "RangePanel";

exports["default"] = RangePanel;
