// tdesign-react/cjs/input-adornment/InputAdornment.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var hooks_useConfig = require('../hooks/useConfig.js');
var isString = require('../_chunks/dep-17e2e72f.js');
var isNumber = require('../_chunks/dep-f0e884b9.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["prepend", "append", "children", "className", "style"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var InputAdornment = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var prepend = props.prepend,
    append = props.append,
    children = props.children,
    className = props.className,
    style = props.style,
    wrapperProps = objectWithoutProperties._objectWithoutProperties(props, _excluded);
  var renderAddon = function renderAddon(type, classPrefix2, Content) {
    if (typeof Content === "undefined") return null;
    var addonNode;
    var isContentNode = isString.isString(Content) || isNumber.isNumber(Content);
    if (typeof Content === "function") {
      addonNode = /* @__PURE__ */React__default["default"].createElement(Content, null);
    } else {
      addonNode = isContentNode ? /* @__PURE__ */React__default["default"].createElement("span", {
        className: "".concat(classPrefix2, "-input-adornment__text")
      }, Content) : Content;
    }
    return addonNode ? /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(classPrefix2, "-input-adornment__").concat(type)
    }, addonNode) : addonNode;
  };
  var renderChildren = function renderChildren() {
    return React__default["default"].Children.map(children, function (child) {
      if (!child) return null;
      if (/*#__PURE__*/React__default["default"].isValidElement(child)) {
        var childProps = child.props;
        return /*#__PURE__*/React__default["default"].cloneElement(child, _objectSpread(_objectSpread(_objectSpread({}, wrapperProps), childProps), {}, {
          onChange: function onChange() {
            var _wrapperProps$onChang, _wrapperProps$onChang2, _childProps$onChange, _childProps$onChange$;
            for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
              args[_key] = arguments[_key];
            }
            wrapperProps === null || wrapperProps === void 0 || (_wrapperProps$onChang = wrapperProps.onChange) === null || _wrapperProps$onChang === void 0 || (_wrapperProps$onChang2 = _wrapperProps$onChang.call) === null || _wrapperProps$onChang2 === void 0 || _wrapperProps$onChang2.call.apply(_wrapperProps$onChang2, [_wrapperProps$onChang, null].concat(args));
            childProps === null || childProps === void 0 || (_childProps$onChange = childProps.onChange) === null || _childProps$onChange === void 0 || (_childProps$onChange$ = _childProps$onChange.call) === null || _childProps$onChange$ === void 0 || _childProps$onChange$.call.apply(_childProps$onChange$, [_childProps$onChange, null].concat(args));
          }
        }));
      }
      return child;
    });
  };
  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: ref,
    style: style,
    className: classNames__default["default"]("".concat(classPrefix, "-input-adornment"), className)
  }, renderAddon("prepend", classPrefix, prepend), renderChildren(), renderAddon("append", classPrefix, append));
});
InputAdornment.displayName = "InputAdornment";

exports["default"] = InputAdornment;
