// tdesign-react/esm/anchor/AnchorTarget.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import React from 'react';
import classNames from 'classnames';
import { FileCopyIcon } from 'tdesign-icons-react';
import { Popup } from '../popup/index.js';
import { MessagePlugin } from '../message/index.js';
import useConfig from '../hooks/useConfig.js';
import useGlobalIcon from '../hooks/useGlobalIcon.js';
import copyText from '../_util/copyText.js';
import { anchorTargetDefaultProps } from './defaultProps.js';
import useDefaultProps from '../hooks/useDefaultProps.js';
import '../popup/Popup.js';
import '../_chunks/dep-5e72c8b9.js';
import '../_chunks/dep-43f7da7b.js';
import '../_chunks/dep-94a7af64.js';
import '../_chunks/dep-0054b574.js';
import '../_chunks/dep-8e8ac0d4.js';
import 'react-transition-group';
import 'lodash-es';
import '../_chunks/dep-a347b1b5.js';
import '../common/Portal.js';
import 'react-dom';
import '../hooks/useLayoutEffect.js';
import '../config-provider/ConfigContext.js';
import '../locale/zh_CN.js';
import '../_chunks/dep-7d48ddac.js';
import 'dayjs';
import '../_chunks/dep-766c4603.js';
import '../hooks/useAnimation.js';
import '../hooks/useAttach.js';
import '../hooks/useControlled.js';
import '../_util/noop.js';
import '../hooks/useMutationObserver.js';
import '../hooks/useLatest.js';
import '../hooks/usePopper.js';
import '../_chunks/dep-1e9f924f.js';
import '../_chunks/dep-57e67945.js';
import 'react-fast-compare';
import '@popperjs/core';
import '../hooks/useWindowSize.js';
import '../popup/defaultProps.js';
import '../popup/hooks/useTrigger.js';
import '../_util/listener.js';
import '../_util/ref.js';
import 'react-is';
import '../_util/isFragment.js';
import '../popup/utils/transition.js';
import '../popup/PopupPlugin.js';
import '../_chunks/dep-fe79dc5d.js';
import '../_util/react-render.js';
import '../common/PluginContainer.js';
import '../config-provider/index.js';
import '../config-provider/ConfigProvider.js';
import '../config-provider/type.js';
import '../popup/style/index.js';
import '../popup/type.js';
import '../message/Message.js';
import '../_chunks/dep-4a2e3d21.js';
import '../_chunks/dep-4461a684.js';
import '../hooks/usePersistFn.js';
import '../message/MessageClose.js';
import '../message/useMessageClass.js';
import '../message/MessageIcon.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../loading/defaultProps.js';
import '../loading/gradient.js';
import '../_chunks/dep-2481dfad.js';
import '../_chunks/dep-9721fff3.js';
import '../hooks/useDomRefCallback.js';
import '../loading/plugin.js';
import '../loading/style/index.js';
import '../loading/type.js';
import '../message/config.js';
import '../message/defaultProps.js';
import '../message/style/index.js';
import '../message/type.js';

var AnchorTarget = function AnchorTarget(props) {
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix;
  var _useGlobalIcon = useGlobalIcon({
      FileCopyIcon: FileCopyIcon
    }),
    FileCopyIcon$1 = _useGlobalIcon.FileCopyIcon;
  var _useDefaultProps = useDefaultProps(props, anchorTargetDefaultProps),
    id = _useDefaultProps.id,
    tag = _useDefaultProps.tag,
    children = _useDefaultProps.children,
    className = _useDefaultProps.className,
    style = _useDefaultProps.style;
  var tagClassName = classNames("".concat(classPrefix, "-anchor__target"), className);
  var iconClassName = "".concat(classPrefix, "-anchor__copy");
  var handleCopyText = function handleCopyText() {
    var a = document.createElement("a");
    a.href = "#".concat(id);
    copyText(a.href);
    MessagePlugin.success("\u94FE\u63A5\u590D\u5236\u6210\u529F", 1e3);
  };
  var Content = function Content() {
    return /* @__PURE__ */React.createElement(React.Fragment, null, children, /* @__PURE__ */React.createElement(Popup, {
      content: "\u590D\u5236\u94FE\u63A5",
      placement: "top",
      showArrow: true
    }, /* @__PURE__ */React.createElement(FileCopyIcon$1, {
      className: iconClassName,
      onClick: handleCopyText
    })));
  };
  return /*#__PURE__*/React.createElement(tag, {
    id: id,
    className: tagClassName,
    style: style
  }, /*#__PURE__*/React.createElement(Content));
};
AnchorTarget.displayName = "AnchorTarget";

export { AnchorTarget as default };
