// tdesign-react/lib/date-picker/panel/ExtraContent.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import React from 'react';
import DatePickerFooter from '../base/Footer.js';
import '../../_chunks/dep-5e72c8b9.js';
import '../../_chunks/dep-43f7da7b.js';
import '../../_chunks/dep-94a7af64.js';
import '../../_chunks/dep-0054b574.js';
import '../../_chunks/dep-8e8ac0d4.js';
import 'classnames';
import 'lodash-es';
import '../../locale/LocalReceiver.js';
import '../../config-provider/ConfigProvider.js';
import '../../config-provider/ConfigContext.js';
import '../../_chunks/dep-923c86e8.js';
import '../../_chunks/dep-03cf63f2.js';
import 'dayjs';
import '../../button/index.js';
import '../../button/Button.js';
import '../../_chunks/dep-4461a684.js';
import '../../hooks/useConfig.js';
import '../../hooks/useDomRefCallback.js';
import '../../hooks/useRipple.js';
import '../../_chunks/dep-2481dfad.js';
import '../../hooks/useAnimation.js';
import '../../_chunks/dep-a347b1b5.js';
import '../../loading/index.js';
import '../../loading/Loading.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../hooks/useLayoutEffect.js';
import '../../hooks/useDefaultProps.js';
import '../../loading/defaultProps.js';
import '../../loading/gradient.js';
import '../../_chunks/dep-ee4dfb46.js';
import '../../_chunks/dep-1e9f924f.js';
import '../../_chunks/dep-57e67945.js';
import '../../loading/plugin.js';
import '../../_util/react-render.js';
import '../../_chunks/dep-b9d1804c.js';
import '../../common/PluginContainer.js';
import '../../button/defaultProps.js';
import '../../_util/parseTNode.js';
import '../../_chunks/dep-65d82dda.js';

function ExtraContent(props) {
  var presets = props.presets,
    enableTimePicker = props.enableTimePicker,
    presetsPlacement = props.presetsPlacement,
    onPresetClick = props.onPresetClick,
    onConfirmClick = props.onConfirmClick,
    selectedValue = props.selectedValue,
    needConfirm = props.needConfirm,
    isSwitchTimeMode = props.isSwitchTimeMode;
  var showPanelFooter = enableTimePicker && needConfirm || presets;
  return showPanelFooter ? /* @__PURE__ */React.createElement(DatePickerFooter, {
    presets: presets,
    onPresetClick: onPresetClick,
    enableTimePicker: enableTimePicker,
    onConfirmClick: onConfirmClick,
    presetsPlacement: presetsPlacement,
    selectedValue: selectedValue,
    needConfirm: needConfirm,
    onTimePanelChange: props.onTimeModeChange,
    isDateRangeContent: props.isDateRangeContent,
    isSwitchTimeMode: isSwitchTimeMode
  }) : null;
}

export { ExtraContent as default };
