// tdesign-react/cjs/grid/Col.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var hooks_useConfig = require('../hooks/useConfig.js');
var grid_defaultProps = require('./defaultProps.js');
var grid_Row = require('./Row.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var isObject = require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["flex", "offset", "order", "pull", "push", "span", "tag", "className", "children", "style"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var calcColPadding = function calcColPadding(gutter, currentSize) {
  var paddingObj = {};
  if (typeof gutter === "number") {
    Object.assign(paddingObj, {
      paddingLeft: "".concat(gutter / 2, "px"),
      paddingRight: "".concat(gutter / 2, "px")
    });
  } else if (Array.isArray(gutter) && gutter.length) {
    if (typeof gutter[0] === "number") {
      Object.assign(paddingObj, {
        paddingLeft: "".concat(gutter[0] / 2, "px"),
        paddingRight: "".concat(gutter[0] / 2, "px")
      });
    }
    if (isObject.isObject(gutter[0]) && gutter[0][currentSize]) {
      Object.assign(paddingObj, {
        paddingLeft: "".concat(gutter[0][currentSize] / 2, "px"),
        paddingRight: "".concat(gutter[0][currentSize] / 2, "px")
      });
    }
  } else if (isObject.isObject(gutter) && gutter[currentSize]) {
    Object.assign(paddingObj, {
      paddingLeft: "".concat(gutter[currentSize] / 2, "px"),
      paddingRight: "".concat(gutter[currentSize] / 2, "px")
    });
  }
  return paddingObj;
};
var parseFlex = function parseFlex(flex) {
  if (typeof flex === "number") {
    return "".concat(flex, " ").concat(flex, " auto");
  }
  if (/^\d+(\.\d+)?(px|r?em|%)$/.test(flex)) {
    return "0 0 ".concat(flex);
  }
  return flex;
};
var Col = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _useDefaultProps = hooks_useDefaultProps["default"](props, grid_defaultProps.colDefaultProps),
    flex = _useDefaultProps.flex,
    offset = _useDefaultProps.offset,
    order = _useDefaultProps.order,
    pull = _useDefaultProps.pull,
    push = _useDefaultProps.push,
    span = _useDefaultProps.span,
    tag = _useDefaultProps.tag,
    className = _useDefaultProps.className,
    children = _useDefaultProps.children,
    propStyle = _useDefaultProps.style,
    otherColProps = objectWithoutProperties._objectWithoutProperties(_useDefaultProps, _excluded);
  var _useContext = React.useContext(grid_Row.RowContext),
    rowGutter = _useContext.gutter,
    rowSize = _useContext.size;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var allSizes = ["xs", "sm", "md", "lg", "xl", "xxl"];
  var sizeClasses = allSizes.reduce(function (acc, currSize) {
    var sizeProp = props[currSize];
    var sizeObj = {};
    if (typeof sizeProp === "number") {
      sizeObj.span = sizeProp;
    } else if (isObject.isObject(sizeProp)) {
      sizeObj = sizeProp || {};
    }
    return _objectSpread(_objectSpread({}, acc), {}, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-col-").concat(currSize, "-").concat(sizeObj.span), sizeObj.span !== void 0), "".concat(classPrefix, "-col-").concat(currSize, "-order-").concat(sizeObj.order), parseInt(sizeObj.order, 10) >= 0), "".concat(classPrefix, "-col-").concat(currSize, "-offset-").concat(sizeObj.offset), parseInt(sizeObj.offset, 10) >= 0), "".concat(classPrefix, "-col-").concat(currSize, "-push-").concat(sizeObj.push), parseInt(sizeObj.push, 10) >= 0), "".concat(classPrefix, "-col-").concat(currSize, "-pull-").concat(sizeObj.pull), parseInt(sizeObj.pull, 10) >= 0));
  }, {});
  var colClassNames = classNames__default["default"]("".concat(classPrefix, "-col"), className, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-col-").concat(span), span !== void 0), "".concat(classPrefix, "-col-offset-").concat(offset), parseInt(offset, 10) >= 0), "".concat(classPrefix, "-col-pull-").concat(pull), parseInt(pull, 10) >= 0), "".concat(classPrefix, "-col-push-").concat(push), parseInt(push, 10) >= 0), "".concat(classPrefix, "-col-order-").concat(order), parseInt(order, 10) >= 0), sizeClasses);
  var colStyle = _objectSpread(_objectSpread({}, calcColPadding(rowGutter, rowSize)), propStyle);
  flex && (colStyle.flex = parseFlex(flex));
  return /*#__PURE__*/React__default["default"].createElement(tag, _objectSpread({
    ref: ref,
    className: colClassNames,
    style: colStyle
  }, otherColProps), children);
});
Col.displayName = "Col";

exports["default"] = Col;
