// tdesign-react/cjs/tree/Tree.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var toConsumableArray = require('../_chunks/dep-1d0dbfc5.js');
var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var reactTransitionGroup = require('react-transition-group');
var classNames = require('classnames');
var log = require('../_chunks/dep-b289b0be.js');
var _util_parseTNode = require('../_util/parseTNode.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var hooks_usePersistFn = require('../hooks/usePersistFn.js');
var tree_defaultProps = require('./defaultProps.js');
var tree_hooks_TreeDraggableContext = require('./hooks/TreeDraggableContext.js');
var tree_hooks_useControllable = require('./hooks/useControllable.js');
var tree_hooks_useStore = require('./hooks/useStore.js');
var tree_hooks_useTreeConfig = require('./hooks/useTreeConfig.js');
var tree_hooks_useTreeVirtualScroll = require('./hooks/useTreeVirtualScroll.js');
var tree_TreeItem = require('./TreeItem.js');
var get = require('../_chunks/dep-05df1ea3.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-aff91ab0.js');
require('../_util/createHookContext.js');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-5814cc47.js');
require('../_chunks/dep-427fb3a1.js');
require('mitt');
require('../_chunks/dep-78c47269.js');
require('../_chunks/dep-e37cb830.js');
require('../_chunks/dep-91c4f617.js');
require('../_chunks/dep-f7bf3d17.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-ca918872.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-13de2fef.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-63308b84.js');
require('../_chunks/dep-048e4314.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-089e8106.js');
require('../_chunks/dep-5c3139cb.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-2e455d91.js');
require('../_chunks/dep-894a5ece.js');
require('../_chunks/dep-8f62fb96.js');
require('../_chunks/dep-f0e884b9.js');
require('../_chunks/dep-526abd62.js');
require('../_chunks/dep-4d78a412.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-966765d8.js');
require('../_chunks/dep-4ef3269f.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-2a089253.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-e44b03ec.js');
require('../hooks/useUpdateLayoutEffect.js');
require('../hooks/useIsFirstRender.js');
require('../hooks/useLayoutEffect.js');
require('../_chunks/dep-68f5fe7b.js');
require('../hooks/usePrevious.js');
require('../_chunks/dep-4a562e04.js');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../locale/LocalReceiver.js');
require('../config-provider/ConfigProvider.js');
require('../hooks/useEventCallback.js');
require('../hooks/useVirtualScroll.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-1c586f5c.js');
require('tdesign-icons-react');
require('../loading/index.js');
require('../loading/Loading.js');
require('../common/Portal.js');
require('react-dom');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../_chunks/dep-b011b5d3.js');
require('../hooks/useDomRefCallback.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../common/PluginContainer.js');
require('../hooks/useRipple.js');
require('../hooks/useAnimation.js');
require('../hooks/useGlobalIcon.js');
require('../checkbox/index.js');
require('../_chunks/dep-83710d84.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../common/Check.js');
require('../_util/helper.js');
require('../checkbox/defaultProps.js');
require('./hooks/useDraggable.js');
require('../_chunks/dep-2a85392e.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Tree = /*#__PURE__*/React.forwardRef(function (originalProps, ref) {
  var _useTreeConfig = tree_hooks_useTreeConfig.useTreeConfig(),
    treeClassNames = _useTreeConfig.treeClassNames,
    transitionNames = _useTreeConfig.transitionNames,
    transitionClassNames = _useTreeConfig.transitionClassNames,
    transitionDuration = _useTreeConfig.transitionDuration,
    locale = _useTreeConfig.locale;
  var _useControllable = tree_hooks_useControllable["default"](originalProps),
    value = _useControllable.value,
    onChange = _useControllable.onChange,
    expanded = _useControllable.expanded,
    onExpand = _useControllable.onExpand,
    onActive = _useControllable.onActive,
    actived = _useControllable.actived,
    setTreeIndeterminate = _useControllable.setTreeIndeterminate,
    indeterminate = _useControllable.indeterminate;
  var props = hooks_useDefaultProps["default"](originalProps, tree_defaultProps.treeDefaultProps);
  var empty = props.empty,
    activable = props.activable,
    disabled = props.disabled,
    checkable = props.checkable,
    checkProps = props.checkProps,
    disableCheck = props.disableCheck,
    hover = props.hover,
    icon = props.icon,
    line = props.line,
    label = props.label,
    operations = props.operations,
    transition = props.transition,
    expandOnClickNode = props.expandOnClickNode,
    onClick = props.onClick,
    scroll = props.scroll,
    className = props.className,
    style = props.style,
    allowDrop = props.allowDrop,
    onScroll = props.onScroll;
  var _useState = React.useState([]),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    visibleNodes = _useState2[0],
    setVisibleNodes = _useState2[1];
  var emptyText = empty || locale("empty");
  var store = tree_hooks_useStore.useStore(_objectSpread(_objectSpread({}, props), {}, {
    value: value,
    onChange: onChange,
    expanded: expanded,
    onExpand: onExpand,
    onActive: onActive,
    actived: actived,
    indeterminate: indeterminate,
    setTreeIndeterminate: setTreeIndeterminate
  }), initial);
  function initial() {
    var nodes = store === null || store === void 0 ? void 0 : store.getNodes();
    var newVisibleNodes = nodes === null || nodes === void 0 ? void 0 : nodes.filter(function (node) {
      return node.visible;
    });
    setVisibleNodes(newVisibleNodes);
  }
  var setExpanded = hooks_usePersistFn.usePersistFn(function (node, isExpanded, ctx) {
    var e = ctx.e,
      trigger = ctx.trigger;
    var expanded2 = node.setExpanded(isExpanded);
    var treeNodeModel = node === null || node === void 0 ? void 0 : node.getModel();
    (e || trigger) && (onExpand === null || onExpand === void 0 ? void 0 : onExpand(expanded2, {
      node: treeNodeModel,
      e: e,
      trigger: trigger
    }));
    return expanded2;
  });
  var setIndeterminate = hooks_usePersistFn.usePersistFn(function (node, isIndeterminate, ctx) {
    ctx.e;
      ctx.trigger;
    var indeterminate2 = node.setIndeterminate(isIndeterminate);
    return indeterminate2;
  });
  var treeRef = React.useRef(null);
  var _useTreeVirtualScroll = tree_hooks_useTreeVirtualScroll["default"]({
      treeRef: treeRef,
      scroll: scroll,
      data: visibleNodes,
      onScroll: onScroll
    }),
    visibleData = _useTreeVirtualScroll.visibleData,
    isVirtual = _useTreeVirtualScroll.isVirtual,
    virtualTreeNodeStyle = _useTreeVirtualScroll.treeNodeStyle,
    cursorStyle = _useTreeVirtualScroll.cursorStyle,
    handleRowMounted = _useTreeVirtualScroll.handleRowMounted,
    scrollToElement = _useTreeVirtualScroll.scrollToElement;
  var setActived = hooks_usePersistFn.usePersistFn(function (node, isActived, ctx) {
    var actived2 = node.setActived(isActived);
    var treeNodeModel = node === null || node === void 0 ? void 0 : node.getModel();
    onActive === null || onActive === void 0 || onActive(actived2, _objectSpread({
      node: treeNodeModel
    }, ctx));
    return actived2;
  });
  var setChecked = hooks_usePersistFn.usePersistFn(function (node, isChecked, ctx) {
    var checked = node.setChecked(isChecked);
    var treeNodeModel = node === null || node === void 0 ? void 0 : node.getModel();
    onChange === null || onChange === void 0 || onChange(checked, _objectSpread({
      node: treeNodeModel
    }, ctx));
    return checked;
  });
  var handleItemClick = function handleItemClick(node, options) {
    if (!node) {
      return;
    }
    var isDisabled = disabled || node.disabled;
    var expand = options.expand,
      active = options.active,
      e = options.e,
      trigger = options.trigger;
    if (expand) setExpanded(node, !node.isExpanded(), {
      e: e,
      trigger: trigger
    });
    if (!isDisabled) {
      var treeNodeModel = node === null || node === void 0 ? void 0 : node.getModel();
      if (active) {
        setActived(node, !node.isActived(), {
          e: e,
          trigger: "node-click"
        });
        onClick === null || onClick === void 0 || onClick({
          node: treeNodeModel,
          e: e
        });
      } else {
        onClick === null || onClick === void 0 || onClick({
          node: treeNodeModel,
          e: e
        });
      }
    }
  };
  var handleChange = function handleChange(node, ctx) {
    if (!node || disabled || node.disabled) return;
    var checked = node.toggleChecked();
    var treeNodeModel = node === null || node === void 0 ? void 0 : node.getModel();
    onChange === null || onChange === void 0 || onChange(checked, _objectSpread(_objectSpread({
      node: treeNodeModel
    }, ctx), {}, {
      trigger: "node-click"
    }));
  };
  var handleScrollToElement = React.useCallback(function (params) {
    var index = params.index;
    if (!index && index !== 0) {
      if (!params.key) {
        log.log.error("Tree", "scrollToElement: one of `index` or `key` must exist.");
        return;
      }
      var data = visibleNodes;
      index = data === null || data === void 0 ? void 0 : data.findIndex(function (item) {
        return [get.get(item.data, "key"), get.get(item.data, "value")].includes(params.key);
      });
      if (index < 0) {
        log.log.error("Tree", "".concat(params.key, " does not exist in data, check `key` or `data` please."));
      }
    }
    scrollToElement(_objectSpread(_objectSpread({}, params), {}, {
      index: index
    }));
  }, [scrollToElement, visibleNodes]);
  React.useImperativeHandle(ref, function () {
    return {
      store: store,
      scrollTo: function scrollTo(p) {
        return handleScrollToElement(p);
      },
      appendTo: function appendTo(value2, newData) {
        var list = [];
        if (Array.isArray(newData)) {
          list = newData;
        } else {
          list = [newData];
        }
        list.forEach(function (item) {
          store.appendNodes(value2, item);
        });
      },
      getIndex: function getIndex(value2) {
        return store.getNodeIndex(value2);
      },
      getItem: function getItem(value2) {
        var node = store.getNode(value2);
        return node === null || node === void 0 ? void 0 : node.getModel();
      },
      getItems: function getItems(value2) {
        var nodes = store.getNodes(value2);
        return nodes.map(function (node) {
          return node.getModel();
        });
      },
      getParent: function getParent(value2) {
        var node = store.getParent(value2);
        return node === null || node === void 0 ? void 0 : node.getModel();
      },
      getParents: function getParents(value2) {
        var nodes = store.getParents(value2);
        return nodes.map(function (node) {
          return node.getModel();
        });
      },
      getPath: function getPath(value2) {
        var node = store.getNode(value2);
        var pathNodes = [];
        if (node) {
          pathNodes = node.getPath().map(function (node2) {
            return node2.getModel();
          });
        }
        return pathNodes;
      },
      insertAfter: function insertAfter(value2, newData) {
        return store.insertAfter(value2, newData);
      },
      insertBefore: function insertBefore(value2, newData) {
        return store.insertBefore(value2, newData);
      },
      remove: function remove(value2) {
        return store.remove(value2);
      },
      setItem: function setItem(value2, options) {
        var node = this.store.getNode(value2);
        var spec = options;
        if (node && spec) {
          if ("expanded" in options) {
            setExpanded(node, spec.expanded, {
              trigger: "setItem"
            });
            delete spec.expanded;
          }
          if ("actived" in options) {
            setActived(node, spec.actived, {
              trigger: "setItem"
            });
            delete spec.actived;
          }
          if ("checked" in options) {
            setChecked(node, spec.checked, {
              trigger: "setItem"
            });
            delete spec.checked;
          }
          if ("indeterminate" in options) {
            setTreeIndeterminate(function (prevIndeterminate) {
              return [].concat(toConsumableArray._toConsumableArray(prevIndeterminate), [value2]);
            });
            setIndeterminate(node, spec.indeterminate, {
              trigger: "setItem"
            });
            delete spec.indeterminate;
          }
          node.set(spec);
        }
      }
    };
  }, [store, setExpanded, setActived, setTreeIndeterminate, setChecked, setIndeterminate, handleScrollToElement]);
  var nodeList = React.useMemo(function () {
    return visibleNodes.map(function () {
      return /*#__PURE__*/React__default["default"].createRef();
    });
  }, [visibleNodes]);
  var renderEmpty = function renderEmpty() {
    return _util_parseTNode["default"](emptyText);
  };
  var renderItems = function renderItems(renderNode) {
    if (renderNode.length <= 0) {
      return renderEmpty();
    }
    if (isVirtual) return /* @__PURE__ */React__default["default"].createElement("div", {
      className: treeClassNames.treeList,
      style: virtualTreeNodeStyle
    }, renderNode.map(function (node, index) {
      return /* @__PURE__ */React__default["default"].createElement(tree_TreeItem["default"], {
        ref: nodeList[index],
        key: node.value,
        node: node,
        keys: props.keys,
        empty: empty,
        icon: icon,
        label: label,
        line: line,
        allowDrop: allowDrop,
        transition: transition,
        expandOnClickNode: expandOnClickNode,
        activable: activable,
        operations: operations,
        checkProps: checkProps,
        disableCheck: disableCheck,
        onClick: handleItemClick,
        onChange: handleChange,
        onTreeItemMounted: handleRowMounted,
        isVirtual: isVirtual
      });
    }));
    return /* @__PURE__ */React__default["default"].createElement(reactTransitionGroup.TransitionGroup, {
      name: transitionNames.treeNode,
      className: treeClassNames.treeList
    }, renderNode.map(function (node, index) {
      return /* @__PURE__ */React__default["default"].createElement(reactTransitionGroup.CSSTransition, {
        nodeRef: nodeList[index],
        key: node.value,
        timeout: transitionDuration,
        classNames: transitionClassNames
      }, /* @__PURE__ */React__default["default"].createElement(tree_TreeItem["default"], {
        ref: nodeList[index],
        node: node,
        keys: props.keys,
        empty: empty,
        icon: icon,
        label: label,
        line: line,
        allowDrop: allowDrop,
        transition: transition,
        expandOnClickNode: expandOnClickNode,
        activable: activable,
        operations: operations,
        checkProps: checkProps,
        disableCheck: disableCheck,
        onClick: handleItemClick,
        onChange: handleChange
      }));
    }));
  };
  var draggable = React.useMemo(function () {
    return {
      props: props,
      store: store
    };
  }, [props, store]);
  return /* @__PURE__ */React__default["default"].createElement(tree_hooks_TreeDraggableContext.TreeDraggableContext.Provider, {
    value: draggable
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](treeClassNames.tree, className, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, treeClassNames.disabled, disabled), treeClassNames.treeHoverable, hover), treeClassNames.treeCheckable, checkable), treeClassNames.treeFx, transition), treeClassNames.treeBlockNode, expandOnClickNode), treeClassNames.treeVscroll, props.scroll)),
    style: style,
    ref: treeRef
  }, isVirtual ? /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement("div", {
    className: treeClassNames.treeVscrollCursor,
    style: cursorStyle
  }), renderItems(visibleData)) : renderItems(visibleNodes)));
});
Tree.displayName = "Tree";

exports["default"] = Tree;
