// tdesign-react/cjs/tree/TreeItem.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var loading_index = require('../loading/index.js');
var hooks_useRipple = require('../hooks/useRipple.js');
var hooks_useDomRefCallback = require('../hooks/useDomRefCallback.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var checkbox_index = require('../checkbox/index.js');
var tree_hooks_useTreeConfig = require('./hooks/useTreeConfig.js');
var tree_hooks_useDraggable = require('./hooks/useDraggable.js');
var _util_ref = require('../_util/ref.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var isFunction = require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../loading/Loading.js');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-ecfe656e.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../hooks/useDefaultProps.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-f0e884b9.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../hooks/useAnimation.js');
require('../_chunks/dep-83710d84.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../common/Check.js');
require('../_util/helper.js');
require('../_chunks/dep-e44b03ec.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-089e8106.js');
require('../checkbox/defaultProps.js');
require('../locale/LocalReceiver.js');
require('./hooks/TreeDraggableContext.js');
require('../_util/createHookContext.js');
require('../hooks/usePersistFn.js');
require('../_chunks/dep-2a85392e.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('react-is');
require('../_util/isFragment.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["setData"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var TreeItem = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var node = props.node,
    icon = props.icon,
    label = props.label,
    line = props.line,
    expandOnClickNode = props.expandOnClickNode,
    activable = props.activable,
    checkProps = props.checkProps,
    operations = props.operations,
    onClick = props.onClick,
    _onChange = props.onChange,
    isVirtual = props.isVirtual,
    onTreeItemMounted = props.onTreeItemMounted,
    allowDrop = props.allowDrop;
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      CaretRightSmallIcon: tdesignIconsReact.CaretRightSmallIcon
    }),
    CaretRightSmallIcon = _useGlobalIcon.CaretRightSmallIcon;
  var level = node.level;
  var nodeRef = React.useRef(null);
  var _useTreeConfig = tree_hooks_useTreeConfig.useTreeConfig(),
    treeClassNames = _useTreeConfig.treeClassNames,
    locale = _useTreeConfig.locale;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  React.useEffect(function () {
    onTreeItemMounted === null || onTreeItemMounted === void 0 || onTreeItemMounted({
      ref: nodeRef.current,
      data: node
    });
  }, [isVirtual, nodeRef, node, onTreeItemMounted]);
  var handleClick = function handleClick(evt) {
    var _srcTarget$className, _srcTarget$className$;
    var srcTarget = evt.target;
    var isBranchTrigger = node.children && expandOnClickNode && (srcTarget.className === "".concat(classPrefix, "-checkbox__input") || srcTarget.tagName.toLowerCase() === "input");
    if (isBranchTrigger) return;
    if (expandOnClickNode && node.children && ((_srcTarget$className = srcTarget.className) === null || _srcTarget$className === void 0 || (_srcTarget$className$ = _srcTarget$className.indexOf) === null || _srcTarget$className$ === void 0 ? void 0 : _srcTarget$className$.call(_srcTarget$className, "".concat(classPrefix, "-tree__label"))) !== -1) evt.preventDefault();
    onClick === null || onClick === void 0 || onClick(node, {
      e: evt,
      expand: expandOnClickNode,
      active: activable,
      trigger: "node-click"
    });
  };
  var handleItemClick = function handleItemClick(evt) {
    if (node.loading) {
      return;
    }
    onClick === null || onClick === void 0 || onClick(node, {
      e: evt,
      expand: true,
      active: false,
      trigger: "icon-click"
    });
  };
  var handleIconClick = function handleIconClick(evt) {
    if (!icon) return;
    evt.stopPropagation();
    handleItemClick(evt);
  };
  var stopPropagation = function stopPropagation(e) {
    e.stopPropagation();
  };
  var renderIcon = function renderIcon() {
    var isDefaultIcon = false;
    var renderIconNode = function renderIconNode() {
      if (icon === false) {
        return null;
      }
      if (icon instanceof Function) {
        return icon(node.getModel());
      }
      if (/*#__PURE__*/React__default["default"].isValidElement(icon)) {
        return icon;
      }
      if (icon && icon !== true) {
        throw new Error("invalid type of icon");
      }
      if (!node.isLeaf()) {
        isDefaultIcon = true;
        if (node.loading && node.expanded) {
          return /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, {
            loading: true
          });
        }
        return /* @__PURE__ */React__default["default"].createElement(CaretRightSmallIcon, {
          className: treeClassNames.treeIconRight
        });
      }
      return null;
    };
    var iconNode = renderIconNode();
    return /* @__PURE__ */React__default["default"].createElement("span", {
      className: classNames__default["default"](treeClassNames.treeIcon, treeClassNames.folderIcon, defineProperty._defineProperty({}, treeClassNames.treeIconDefault, isDefaultIcon)),
      onClick: handleIconClick
    }, iconNode);
  };
  var renderLine = function renderLine() {
    var iconVisible = icon !== false;
    if (line === false) {
      return null;
    }
    if (isFunction.isFunction(line)) {
      return line(node.getModel());
    }
    if (/*#__PURE__*/React__default["default"].isValidElement(line)) {
      return line;
    }
    if (node.parent && node.tree) {
      var shadowStyles = [];
      var parents = node.getParents();
      parents.pop();
      parents.forEach(function (pnode, index) {
        if (!pnode.vmIsLast) {
          shadowStyles.push("calc(-".concat(index + 1, " * var(--space)) 0 var(--color)"));
        }
      });
      var styles = {
        "--level": level,
        boxShadow: shadowStyles.join(",")
      };
      return /* @__PURE__ */React__default["default"].createElement("span", {
        className: classNames__default["default"](treeClassNames.line, defineProperty._defineProperty(defineProperty._defineProperty({}, treeClassNames.lineIsLeaf, node.vmIsLeaf || !iconVisible), treeClassNames.lineIsFirst, node.vmIsFirst && iconVisible)),
        style: styles,
        onClick: stopPropagation
      });
    }
    return null;
  };
  var _useDomRefCallback = hooks_useDomRefCallback["default"](),
    _useDomRefCallback2 = slicedToArray._slicedToArray(_useDomRefCallback, 2),
    labelDom = _useDomRefCallback2[0],
    setRefCurrent = _useDomRefCallback2[1];
  hooks_useRipple["default"](labelDom);
  var _useState = React.useState({}),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    updateRender = _useState2[1];
  var renderLabel = function renderLabel() {
    var _node$data, _node$data2;
    var emptyView = locale("empty");
    var labelText = "";
    if (label instanceof Function) {
      var _node$getModel = node.getModel(),
        nodeSetData = _node$getModel.setData,
        rest = objectWithoutProperties._objectWithoutProperties(_node$getModel, _excluded);
      labelText = label(_objectSpread(_objectSpread({}, rest), {}, {
        setData: function setData(value) {
          nodeSetData(value);
          updateRender({});
        }
      })) || emptyView;
    } else {
      labelText = node.label || emptyView;
    }
    var labelClasses = classNames__default["default"](treeClassNames.treeLabel, treeClassNames.treeLabelStrictly, defineProperty._defineProperty({}, treeClassNames.actived, node.isActivable() ? node.actived : false));
    if (node.isCheckable()) {
      var checkboxProps;
      if (typeof checkProps === "function") {
        checkboxProps = checkProps(node.getModel());
      } else {
        checkboxProps = checkProps;
      }
      return /* @__PURE__ */React__default["default"].createElement(checkbox_index.Checkbox, _objectSpread({
        ref: setRefCurrent,
        checked: node.checked,
        indeterminate: node.indeterminate,
        disabled: node.disabled,
        name: String(node.value),
        onChange: function onChange(_, ctx) {
          return _onChange(node, ctx);
        },
        className: labelClasses,
        stopLabelTrigger: expandOnClickNode && !!node.children
      }, checkboxProps), /* @__PURE__ */React__default["default"].createElement("span", {
        "data-target": "label"
      }, labelText));
    }
    return /* @__PURE__ */React__default["default"].createElement("span", {
      ref: setRefCurrent,
      "data-target": "label",
      className: labelClasses,
      title: /*#__PURE__*/React.isValidElement(node.label) && !((_node$data = node.data) !== null && _node$data !== void 0 && _node$data.text) ? "" : String(((_node$data2 = node.data) === null || _node$data2 === void 0 ? void 0 : _node$data2.text) || node.label)
    }, /* @__PURE__ */React__default["default"].createElement("span", {
      style: {
        position: "relative"
      }
    }, labelText));
  };
  var renderOperations = function renderOperations() {
    var operationsView = null;
    if (operations) {
      if (/*#__PURE__*/React__default["default"].isValidElement(operations)) {
        operationsView = operations;
      } else if (operations instanceof Function) {
        var treeNodeModel = node === null || node === void 0 ? void 0 : node.getModel();
        operationsView = operations(treeNodeModel);
      } else {
        throw new Error("invalid type of operations");
      }
    }
    if (operationsView) {
      return /* @__PURE__ */React__default["default"].createElement("span", {
        className: treeClassNames.treeOperations,
        "data-target": "operations",
        onClick: stopPropagation
      }, operationsView);
    }
    return null;
  };
  var _useDraggable = tree_hooks_useDraggable["default"]({
      node: node,
      nodeRef: nodeRef,
      allowDrop: allowDrop
    }),
    setDragStatus = _useDraggable.setDragStatus,
    isDragging = _useDraggable.isDragging,
    dropPosition = _useDraggable.dropPosition,
    isDragOver = _useDraggable.isDragOver;
  var handleDragStart = function handleDragStart(evt) {
    var node2 = props.node;
    if (!node2.isDraggable()) return;
    evt.stopPropagation();
    setDragStatus("dragStart", evt);
    try {
      var _evt$dataTransfer;
      (_evt$dataTransfer = evt.dataTransfer) === null || _evt$dataTransfer === void 0 || _evt$dataTransfer.setData("text/plain", "");
    } catch (e) {}
  };
  var handleDragEnd = function handleDragEnd(evt) {
    var node2 = props.node;
    if (!node2.isDraggable()) return;
    evt.stopPropagation();
    setDragStatus("dragEnd", evt);
  };
  var handleDragOver = function handleDragOver(evt) {
    var node2 = props.node;
    if (!node2.isDraggable()) return;
    evt.stopPropagation();
    evt.preventDefault();
    setDragStatus("dragOver", evt);
  };
  var handleDragLeave = function handleDragLeave(evt) {
    var node2 = props.node;
    if (!node2.isDraggable()) return;
    evt.stopPropagation();
    setDragStatus("dragLeave", evt);
  };
  var handleDrop = function handleDrop(evt) {
    var node2 = props.node;
    if (!node2.isDraggable()) return;
    evt.stopPropagation();
    evt.preventDefault();
    setDragStatus("drop", evt);
  };
  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: _util_ref.composeRefs(ref, nodeRef),
    "data-value": node.value,
    "data-level": level,
    className: classNames__default["default"](treeClassNames.treeNode, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, treeClassNames.treeNodeOpen, node.expanded), treeClassNames.actived, node.isActivable() ? node.actived : false), treeClassNames.disabled, node.isDisabled()), treeClassNames.treeNodeDraggable, !node.isDisabled() && node.isDraggable()), treeClassNames.treeNodeDragging, isDragging), treeClassNames.treeNodeDragTipTop, isDragOver && dropPosition < 0), treeClassNames.treeNodeDragTipBottom, isDragOver && dropPosition > 0), treeClassNames.treeNodeDragTipHighlight, !isDragging && isDragOver && dropPosition === 0)),
    style: {
      "--level": level,
      boxShadow: ""
    },
    onClick: handleClick,
    draggable: !node.isDisabled() && node.isDraggable(),
    onDragStart: handleDragStart,
    onDragEnd: handleDragEnd,
    onDragOver: handleDragOver,
    onDragLeave: handleDragLeave,
    onDrop: handleDrop
  }, renderLine(), renderIcon(), renderLabel(), renderOperations());
});
TreeItem.displayName = "TreeItem";

exports["default"] = TreeItem;
