// tdesign-react/cjs/color-picker/components/panel/alpha.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../../_chunks/dep-2b5ba78d.js');
var objectWithoutProperties = require('../../../_chunks/dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var colorPicker_components_panel_slider = require('./slider.js');
require('../../../_chunks/dep-c4d1ce0c.js');
require('../../../_chunks/dep-da15e3e2.js');
require('../../../hooks/useMouseEvent.js');
require('../../hooks/useStyles.js');
require('../../../_chunks/dep-1a3dddaf.js');
require('../../../_chunks/dep-32593c54.js');
require('../../../_chunks/dep-51fc57b3.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["color", "baseClassName", "onChange"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Alpha = function Alpha(props) {
  var color = props.color,
    baseClassName = props.baseClassName,
    onChange = props.onChange,
    rest = objectWithoutProperties._objectWithoutProperties(props, _excluded);
  var handleChange = function handleChange(v) {
    onChange(v / 100);
  };
  var railStyle = {
    background: "linear-gradient(to right, rgba(0, 0, 0, 0), ".concat(props.color.rgb, ")")
  };
  return /* @__PURE__ */React__default["default"].createElement(colorPicker_components_panel_slider["default"], _objectSpread({
    baseClassName: baseClassName,
    className: classNames__default["default"](["".concat(baseClassName, "__alpha"), "".concat(baseClassName, "--bg-alpha")]),
    color: color,
    value: color.alpha * 100,
    onChange: handleChange,
    railStyle: railStyle,
    type: "alpha",
    maxValue: 100
  }, rest));
};
var AlphaSlider = /*#__PURE__*/React__default["default"].memo(Alpha);

exports["default"] = AlphaSlider;
