// tdesign-react/cjs/color-picker/components/panel/index.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../../_chunks/dep-2b5ba78d.js');
var toConsumableArray = require('../../../_chunks/dep-1d0dbfc5.js');
var slicedToArray = require('../../../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var color = require('../../../_chunks/dep-f40ba4c4.js');
var constants = require('../../../_chunks/dep-da15e3e2.js');
var colorPicker_components_panel_format_inputs = require('../../../_chunks/dep-a38cacd7.js');
var hooks_useCommonClassName = require('../../../hooks/useCommonClassName.js');
var hooks_useControlled = require('../../../hooks/useControlled.js');
var hooks_useDefaultProps = require('../../../hooks/useDefaultProps.js');
var locale_LocalReceiver = require('../../../locale/LocalReceiver.js');
var colorPicker_defaultProps = require('../../defaultProps.js');
var colorPicker_hooks_useClassNames = require('../../hooks/useClassNames.js');
var colorPicker_components_panel_alpha = require('./alpha.js');
var colorPicker_components_panel_format_index = require('./format/index.js');
var colorPicker_components_panel_header = require('./header.js');
var colorPicker_components_panel_hue = require('./hue.js');
var colorPicker_components_panel_linearGradient = require('./linear-gradient.js');
var colorPicker_components_panel_saturation = require('./saturation.js');
var colorPicker_components_panel_swatches = require('./swatches.js');
require('../../../_chunks/dep-c4d1ce0c.js');
require('../../../_chunks/dep-51fc57b3.js');
require('../../../_chunks/dep-3c2d5e8c.js');
require('../../../_chunks/dep-32593c54.js');
require('../../../_chunks/dep-427fb3a1.js');
require('../../../_chunks/dep-d14bfcfa.js');
require('../../../_chunks/dep-17e2e72f.js');
require('../../../_chunks/dep-5f51e3c1.js');
require('../../../_chunks/dep-794b42ce.js');
require('../../../_chunks/dep-ecfe656e.js');
require('../../../_chunks/dep-526abd62.js');
require('../../../input/index.js');
require('../../../input/Input.js');
require('../../../_chunks/dep-1c586f5c.js');
require('tdesign-icons-react');
require('../../../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../../../_util/parseTNode.js');
require('../../../_chunks/dep-b289b0be.js');
require('../../../_chunks/dep-cba78412.js');
require('../../../_chunks/dep-aff91ab0.js');
require('../../../hooks/useConfig.js');
require('../../../config-provider/ConfigContext.js');
require('../../../_chunks/dep-14013769.js');
require('../../../_chunks/dep-06bf4625.js');
require('dayjs');
require('../../../_chunks/dep-043eee20.js');
require('../../../_chunks/dep-e51f8066.js');
require('../../../_chunks/dep-67a947f0.js');
require('../../../_chunks/dep-cd29b2b2.js');
require('../../../_chunks/dep-ed46627b.js');
require('../../../_chunks/dep-746faf11.js');
require('../../../_chunks/dep-6abdd014.js');
require('../../../_chunks/dep-6d96dce8.js');
require('../../../_chunks/dep-977a0080.js');
require('../../../_chunks/dep-41c55905.js');
require('../../../_chunks/dep-b21b548e.js');
require('../../../_chunks/dep-69c18aee.js');
require('../../../_chunks/dep-83fd21a7.js');
require('../../../_chunks/dep-0bed591e.js');
require('../../../_chunks/dep-5f500f41.js');
require('../../../_chunks/dep-7577d7c1.js');
require('../../../_chunks/dep-2d5cf7d6.js');
require('../../../hooks/useGlobalIcon.js');
require('../../../hooks/useLayoutEffect.js');
require('../../../_chunks/dep-68f5fe7b.js');
require('../../../input/defaultProps.js');
require('../../../input/InputGroup.js');
require('../../../input/useLengthLimit.js');
require('../../../_chunks/dep-b011b5d3.js');
require('../../../_chunks/dep-f0e884b9.js');
require('../../../_util/noop.js');
require('../../../_chunks/dep-eb1b216c.js');
require('../../../_chunks/dep-b529a0d7.js');
require('../../../_chunks/dep-bcd16957.js');
require('../../../_chunks/dep-b4442734.js');
require('../../../_chunks/dep-876d090e.js');
require('../../../config-provider/ConfigProvider.js');
require('../../../_chunks/dep-4a562e04.js');
require('../../../_chunks/dep-a5b4810f.js');
require('../../../_chunks/dep-de8b5dd9.js');
require('../../../_chunks/dep-fc111b3d.js');
require('../../../_chunks/dep-ea361d31.js');
require('../../../input-number/index.js');
require('../../../input-number/InputNumber.js');
require('../../../button/index.js');
require('../../../button/Button.js');
require('../../../hooks/useDomRefCallback.js');
require('../../../hooks/useRipple.js');
require('../../../_chunks/dep-064e4dba.js');
require('../../../hooks/useAnimation.js');
require('../../../loading/index.js');
require('../../../loading/Loading.js');
require('../../../common/Portal.js');
require('react-dom');
require('../../../loading/defaultProps.js');
require('../../../loading/gradient.js');
require('../../../loading/plugin.js');
require('../../../_util/react-render.js');
require('../../../_chunks/dep-e37cb830.js');
require('../../../common/PluginContainer.js');
require('../../../button/defaultProps.js');
require('../../../input-number/useInputNumber.js');
require('../../../_chunks/dep-04deac0d.js');
require('../../../_chunks/dep-8f62fb96.js');
require('../../../input-number/defaultProps.js');
require('../../../_chunks/dep-2a85392e.js');
require('../../../_chunks/dep-d3af8eb7.js');
require('../../../_chunks/dep-eab994e8.js');
require('./slider.js');
require('../../../hooks/useMouseEvent.js');
require('../../hooks/useStyles.js');
require('../../../select/index.js');
require('../../../select/base/Select.js');
require('../../../_util/helper.js');
require('../../../_chunks/dep-e44b03ec.js');
require('../../../_util/ref.js');
require('react-is');
require('../../../_util/isFragment.js');
require('../../../common/FakeArrow.js');
require('../../../select-input/index.js');
require('../../../select-input/SelectInput.js');
require('../../../popup/index.js');
require('../../../popup/Popup.js');
require('react-transition-group');
require('../../../hooks/useAttach.js');
require('../../../hooks/useMutationObserver.js');
require('../../../hooks/useLatest.js');
require('../../../_chunks/dep-4e876c4b.js');
require('../../../_chunks/dep-df8d2fc8.js');
require('../../../_chunks/dep-66f91e5d.js');
require('../../../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../../../hooks/useWindowSize.js');
require('../../../popup/defaultProps.js');
require('../../../popup/hooks/useTrigger.js');
require('../../../_util/listener.js');
require('../../../popup/utils/transition.js');
require('../../../popup/PopupPlugin.js');
require('../../../select-input/defaultProps.js');
require('../../../select-input/useMultiple.js');
require('../../../tag-input/index.js');
require('../../../tag-input/TagInput.js');
require('../../../hooks/useDragSorter.js');
require('../../../hooks/useEventCallback.js');
require('../../../tag-input/defaultProps.js');
require('../../../tag-input/useHover.js');
require('../../../tag-input/useTagList.js');
require('../../../tag/index.js');
require('../../../tag/Tag.js');
require('../../../tag/defaultProps.js');
require('../../../tag/CheckTag.js');
require('../../../_chunks/dep-dd0f5f6b.js');
require('../../../tag/CheckTagGroup.js');
require('../../../tag-input/useTagScroll.js');
require('../../../select-input/useOverlayInnerStyle.js');
require('../../../hooks/useInnerPopupVisible.js');
require('../../../select-input/useSingle.js');
require('../../../_chunks/dep-f7bf3d17.js');
require('../../../_chunks/dep-5e17a073.js');
require('../../../_chunks/dep-ca918872.js');
require('../../../_chunks/dep-13de2fef.js');
require('../../../_chunks/dep-63308b84.js');
require('../../../_chunks/dep-048e4314.js');
require('../../../select/defaultProps.js');
require('../../../select/hooks/useKeyboardControl.js');
require('../../../_chunks/dep-bdc97672.js');
require('../../../select/base/OptionGroup.js');
require('../../../_chunks/dep-05df1ea3.js');
require('../../../select/base/PopupContent.js');
require('../../../select/hooks/usePanelVirtualScroll.js');
require('../../../hooks/useVirtualScroll.js');
require('../../../radio/index.js');
require('../../../_chunks/dep-27d626a1.js');
require('../../../common/Check.js');
require('../../../_chunks/dep-089e8106.js');
require('../../../_chunks/dep-e31ce0e8.js');
require('../../../hooks/useDeepEffect.js');
require('../../../radio/defaultProps.js');
require('../../../radio/useKeyboard.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Panel = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var baseClassName = colorPicker_hooks_useClassNames["default"]();
  var _useCommonClassName = hooks_useCommonClassName["default"](),
    STATUS = _useCommonClassName.STATUS;
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("colorPicker"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var _useDefaultProps = hooks_useDefaultProps["default"](props, colorPicker_defaultProps.colorPickerDefaultProps),
    className = _useDefaultProps.className,
    colorModes = _useDefaultProps.colorModes,
    defaultRecentColors = _useDefaultProps.defaultRecentColors,
    disabled = _useDefaultProps.disabled,
    enableAlpha = _useDefaultProps.enableAlpha,
    enableMultipleGradient = _useDefaultProps.enableMultipleGradient,
    format = _useDefaultProps.format,
    style = _useDefaultProps.style,
    swatchColors = _useDefaultProps.swatchColors,
    showPrimaryColorPreview = _useDefaultProps.showPrimaryColorPreview,
    onChange = _useDefaultProps.onChange,
    onPaletteBarChange = _useDefaultProps.onPaletteBarChange,
    onRecentColorsChange = _useDefaultProps.onRecentColorsChange;
  var _useControlled = hooks_useControlled["default"](props, "value", onChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    innerValue = _useControlled2[0],
    setInnerValue = _useControlled2[1];
  var _useState = React.useState(0),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    setUpdateId = _useState2[1];
  var getModeByColor = function getModeByColor(input) {
    if (colorModes.length === 1) return colorModes[0];
    return colorModes.includes("linear-gradient") && color.Color.isGradientColor(input) ? "linear-gradient" : "monochrome";
  };
  var _useState3 = React.useState(function () {
      return getModeByColor(innerValue);
    }),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    mode = _useState4[0],
    setMode = _useState4[1];
  var isGradient = mode === "linear-gradient";
  var defaultEmptyColor = isGradient ? constants.DEFAULT_LINEAR_GRADIENT : constants.DEFAULT_COLOR;
  var _useControlled3 = hooks_useControlled["default"](props, "recentColors", onRecentColorsChange, {
      defaultRecentColors: defaultRecentColors
    }),
    _useControlled4 = slicedToArray._slicedToArray(_useControlled3, 2),
    recentlyUsedColors = _useControlled4[0],
    setRecentlyUsedColors = _useControlled4[1];
  var colorInstanceRef = React.useRef(new color.Color(innerValue || defaultEmptyColor));
  var formatRef = React.useRef(colorPicker_components_panel_format_inputs.initColorFormat(format, enableAlpha));
  var baseProps = {
    color: colorInstanceRef.current,
    disabled: disabled,
    baseClassName: baseClassName
  };
  var updateColor = function updateColor(value) {
    colorInstanceRef.current.update(value);
    setUpdateId(performance.now());
  };
  var emitColorChange = React.useCallback(function (trigger) {
    var value = colorInstanceRef.current.getFormattedColor(formatRef.current, enableAlpha);
    setInnerValue(value, {
      color: color.getColorObject(colorInstanceRef.current),
      trigger: trigger || "palette-saturation-brightness"
    });
    setUpdateId(performance.now());
  }, [enableAlpha, setInnerValue]);
  React.useEffect(function () {
    var currentColor = colorInstanceRef.current.getFormattedColor(formatRef.current, enableAlpha);
    if (innerValue === currentColor) return;
    var newMode = getModeByColor(innerValue);
    setMode(newMode);
    colorInstanceRef.current.isGradient = newMode === "linear-gradient";
    updateColor(innerValue);
  }, [innerValue]);
  var handleModeChange = function handleModeChange(newMode) {
    setMode(newMode);
    var isGradientMode = newMode === "linear-gradient";
    colorInstanceRef.current.isGradient = isGradientMode;
    var _colorInstanceRef$cur = colorInstanceRef.current,
      rgba = _colorInstanceRef$cur.rgba,
      gradientColors = _colorInstanceRef$cur.gradientColors,
      linearGradient = _colorInstanceRef$cur.linearGradient;
    if (isGradientMode) {
      updateColor(gradientColors.length > 0 ? linearGradient : constants.DEFAULT_LINEAR_GRADIENT);
    } else {
      updateColor(rgba);
    }
    emitColorChange();
  };
  var handleRecentlyUsedColorsChange = function handleRecentlyUsedColorsChange(colors) {
    setRecentlyUsedColors(colors);
  };
  var addRecentlyUsedColor = function addRecentlyUsedColor() {
    var colors = toConsumableArray._toConsumableArray(recentlyUsedColors || []);
    var _colorInstanceRef$cur2 = colorInstanceRef.current,
      isGradient2 = _colorInstanceRef$cur2.isGradient,
      linearGradient = _colorInstanceRef$cur2.linearGradient,
      rgba = _colorInstanceRef$cur2.rgba;
    var currentColor = isGradient2 ? linearGradient : rgba;
    var index = colors.indexOf(currentColor);
    if (index > -1) {
      colors.splice(index, 1);
    }
    colors.unshift(currentColor);
    if (colors.length > constants.TD_COLOR_USED_COLORS_MAX_SIZE) {
      colors.length = constants.TD_COLOR_USED_COLORS_MAX_SIZE;
    }
    handleRecentlyUsedColorsChange(colors);
  };
  var handleSatAndValueChange = function handleSatAndValueChange(_ref) {
    var saturation = _ref.saturation,
      value = _ref.value;
    var _colorInstanceRef$cur3 = colorInstanceRef.current,
      sat = _colorInstanceRef$cur3.saturation,
      val = _colorInstanceRef$cur3.value;
    var changeTrigger = "palette-saturation-brightness";
    if (value !== val && saturation !== sat) {
      changeTrigger = "palette-saturation-brightness";
      colorInstanceRef.current.saturation = saturation;
      colorInstanceRef.current.value = value;
    } else if (saturation !== sat) {
      changeTrigger = "palette-saturation";
      colorInstanceRef.current.saturation = saturation;
    } else if (value !== val) {
      changeTrigger = "palette-brightness";
      colorInstanceRef.current.value = value;
    } else {
      return;
    }
    emitColorChange(changeTrigger);
  };
  var handleHueChange = function handleHueChange(hue) {
    colorInstanceRef.current.hue = hue;
    emitColorChange("palette-hue-bar");
    onPaletteBarChange === null || onPaletteBarChange === void 0 || onPaletteBarChange({
      color: color.getColorObject(colorInstanceRef.current)
    });
  };
  var handleAlphaChange = function handleAlphaChange(alpha) {
    colorInstanceRef.current.alpha = alpha;
    emitColorChange("palette-alpha-bar");
    onPaletteBarChange === null || onPaletteBarChange === void 0 || onPaletteBarChange({
      color: color.getColorObject(colorInstanceRef.current)
    });
  };
  var handleGradientChange = function handleGradientChange(_ref2) {
    var key = _ref2.key,
      payload = _ref2.payload;
    var trigger = "palette-saturation-brightness";
    switch (key) {
      case "degree":
        colorInstanceRef.current.gradientDegree = payload;
        trigger = "input";
        break;
      case "selectedId":
        colorInstanceRef.current.gradientSelectedId = payload;
        break;
      case "colors":
        colorInstanceRef.current.gradientColors = payload;
        break;
    }
    emitColorChange(trigger);
  };
  var handleInputChange = function handleInputChange() {
    emitColorChange("input");
  };
  var SwatchesArea = /*#__PURE__*/React__default["default"].memo(function () {
    var onlySupportGradient = colorModes.length === 1 && colorModes.includes("linear-gradient");
    var recentColors = recentlyUsedColors;
    if (onlySupportGradient && Array.isArray(recentColors)) {
      recentColors = recentColors.filter(function (color$1) {
        return color.Color.isGradientColor(color$1);
      });
    }
    var showUsedColors = Array.isArray(recentColors) || recentColors === true;
    var systemColors = swatchColors;
    if (systemColors === void 0) {
      systemColors = toConsumableArray._toConsumableArray(constants.DEFAULT_SYSTEM_SWATCH_COLORS);
    }
    if (onlySupportGradient) {
      var _systemColors;
      systemColors = (_systemColors = systemColors) === null || _systemColors === void 0 ? void 0 : _systemColors.filter(function (color$1) {
        return color.Color.isGradientColor(color$1);
      });
    }
    var showSystemColors = Array.isArray(systemColors);
    var handleSetColor = function handleSetColor(value, trigger) {
      var newMode = getModeByColor(value);
      setMode(newMode);
      colorInstanceRef.current.isGradient = newMode === "linear-gradient";
      updateColor(value);
      emitColorChange(trigger);
    };
    if (!showSystemColors && !showUsedColors) return null;
    return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(baseClassName, "__swatches-wrap")
    }, showUsedColors && /* @__PURE__ */React__default["default"].createElement(colorPicker_components_panel_swatches["default"], _objectSpread(_objectSpread({}, baseProps), {}, {
      title: t(local.recentColorTitle),
      editable: true,
      handleAddColor: addRecentlyUsedColor,
      colors: recentColors,
      onSetColor: function onSetColor(color) {
        return handleSetColor(color, "recent");
      },
      onChange: handleRecentlyUsedColorsChange
    })), showSystemColors && /* @__PURE__ */React__default["default"].createElement(colorPicker_components_panel_swatches["default"], _objectSpread(_objectSpread({}, baseProps), {}, {
      title: t(local.swatchColorTitle),
      colors: systemColors,
      onSetColor: function onSetColor(color) {
        return handleSetColor(color, "preset");
      }
    }))));
  });
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(baseClassName, "__panel"), disabled ? STATUS.disabled : false, className),
    onClick: function onClick(e) {
      return e.stopPropagation();
    },
    style: _objectSpread({}, style),
    ref: ref
  }, /* @__PURE__ */React__default["default"].createElement(colorPicker_components_panel_header["default"], {
    baseClassName: baseClassName,
    mode: mode,
    colorModes: colorModes,
    onModeChange: handleModeChange
  }), /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(baseClassName, "__body")
  }, isGradient && /* @__PURE__ */React__default["default"].createElement(colorPicker_components_panel_linearGradient["default"], _objectSpread(_objectSpread({}, baseProps), {}, {
    enableMultipleGradient: enableMultipleGradient,
    onChange: handleGradientChange
  })), /* @__PURE__ */React__default["default"].createElement(colorPicker_components_panel_saturation["default"], _objectSpread(_objectSpread({}, baseProps), {}, {
    onChange: handleSatAndValueChange
  })), /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(baseClassName, "__sliders-wrapper")
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(baseClassName, "__sliders")
  }, /* @__PURE__ */React__default["default"].createElement(colorPicker_components_panel_hue["default"], _objectSpread(_objectSpread({}, baseProps), {}, {
    onChange: handleHueChange
  })), enableAlpha && /* @__PURE__ */React__default["default"].createElement(colorPicker_components_panel_alpha["default"], _objectSpread(_objectSpread({}, baseProps), {}, {
    onChange: handleAlphaChange
  }))), showPrimaryColorPreview ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](["".concat(baseClassName, "__sliders-preview"), "".concat(baseClassName, "--bg-alpha")])
  }, /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(baseClassName, "__sliders-preview-inner"),
    style: {
      background: isGradient ? colorInstanceRef.current.linearGradient : colorInstanceRef.current.rgba
    }
  })) : null), /* @__PURE__ */React__default["default"].createElement(colorPicker_components_panel_format_index["default"], _objectSpread(_objectSpread(_objectSpread({}, props), baseProps), {}, {
    format: formatRef.current,
    onInputChange: handleInputChange
  })), /* @__PURE__ */React__default["default"].createElement(SwatchesArea, null)));
});
var ColorPanel = /*#__PURE__*/React__default["default"].memo(Panel);

exports["default"] = ColorPanel;
