// tdesign-react/cjs/color-picker/components/panel/swatches.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var color = require('../../../_chunks/dep-f40ba4c4.js');
var hooks_useGlobalIcon = require('../../../hooks/useGlobalIcon.js');
var hooks_useCommonClassName = require('../../../hooks/useCommonClassName.js');
require('../../../_chunks/dep-1a3dddaf.js');
require('../../../_chunks/dep-32593c54.js');
require('../../../_chunks/dep-51fc57b3.js');
require('../../../_chunks/dep-427fb3a1.js');
require('../../../_chunks/dep-2b5ba78d.js');
require('../../../_chunks/dep-c4d1ce0c.js');
require('../../../_chunks/dep-d14bfcfa.js');
require('../../../_chunks/dep-da15e3e2.js');
require('../../../_chunks/dep-17e2e72f.js');
require('../../../_chunks/dep-5f51e3c1.js');
require('../../../_chunks/dep-794b42ce.js');
require('../../../_chunks/dep-ecfe656e.js');
require('../../../_chunks/dep-526abd62.js');
require('../../../hooks/useConfig.js');
require('../../../config-provider/ConfigContext.js');
require('../../../_chunks/dep-14013769.js');
require('../../../_chunks/dep-06bf4625.js');
require('dayjs');
require('../../../_chunks/dep-043eee20.js');
require('../../../_chunks/dep-e51f8066.js');
require('../../../_chunks/dep-67a947f0.js');
require('../../../_chunks/dep-cd29b2b2.js');
require('../../../_chunks/dep-ed46627b.js');
require('../../../_chunks/dep-746faf11.js');
require('../../../_chunks/dep-6abdd014.js');
require('../../../_chunks/dep-6d96dce8.js');
require('../../../_chunks/dep-977a0080.js');
require('../../../_chunks/dep-cba78412.js');
require('../../../_chunks/dep-aff91ab0.js');
require('../../../_chunks/dep-41c55905.js');
require('../../../_chunks/dep-b21b548e.js');
require('../../../_chunks/dep-69c18aee.js');
require('../../../_chunks/dep-83fd21a7.js');
require('../../../_chunks/dep-0bed591e.js');
require('../../../_chunks/dep-5f500f41.js');
require('../../../_chunks/dep-7577d7c1.js');
require('../../../_chunks/dep-2d5cf7d6.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var Swatches = function Swatches(props) {
  var baseClassName = props.baseClassName,
    _props$colors = props.colors,
    colors = _props$colors === void 0 ? [] : _props$colors,
    _props$editable = props.editable,
    editable = _props$editable === void 0 ? false : _props$editable,
    title = props.title,
    onChange = props.onChange,
    disabled = props.disabled,
    onSetColor = props.onSetColor,
    handleAddColor = props.handleAddColor;
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      DeleteIcon: tdesignIconsReact.DeleteIcon,
      AddIcon: tdesignIconsReact.AddIcon
    }),
    DeleteIcon = _useGlobalIcon.DeleteIcon,
    AddIcon = _useGlobalIcon.AddIcon;
  var swatchesClass = "".concat(baseClassName, "__swatches");
  var _useCommonClassName = hooks_useCommonClassName["default"](),
    statusClassNames = _useCommonClassName.STATUS;
  var isEqualCurrentColor = function isEqualCurrentColor(color$1) {
    return color.Color.compare(color$1, props.color.css);
  };
  var selectedColorIndex = function selectedColorIndex() {
    return colors.findIndex(function (color) {
      return isEqualCurrentColor(color);
    });
  };
  var handleRemoveColor = function handleRemoveColor() {
    var selectedIndex = selectedColorIndex();
    if (selectedIndex > -1) {
      var newColors = colors.filter(function (item, index) {
        return index !== selectedIndex;
      });
      onChange(newColors);
    }
  };
  var handleClick = function handleClick(color) {
    return onSetColor(color);
  };
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: swatchesClass
  }, title ? /* @__PURE__ */React__default["default"].createElement("h3", {
    className: "".concat(swatchesClass, "--title")
  }, /* @__PURE__ */React__default["default"].createElement("span", null, title), editable && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(swatchesClass, "--actions")
  }, /* @__PURE__ */React__default["default"].createElement("span", {
    role: "button",
    className: "".concat(baseClassName, "__icon"),
    onClick: function onClick() {
      return handleAddColor();
    }
  }, /* @__PURE__ */React__default["default"].createElement(AddIcon, null)), colors.length > 0 ? /* @__PURE__ */React__default["default"].createElement("span", {
    role: "button",
    className: "".concat(baseClassName, "__icon"),
    onClick: function onClick() {
      return handleRemoveColor();
    }
  }, /* @__PURE__ */React__default["default"].createElement(DeleteIcon, null)) : null)) : /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null), /* @__PURE__ */React__default["default"].createElement("ul", {
    className: classNames__default["default"]("".concat(swatchesClass, "--items"), "narrow-scrollbar")
  }, colors.map(function (color) {
    return /* @__PURE__ */React__default["default"].createElement("li", {
      className: classNames__default["default"]("".concat(swatchesClass, "--item"), isEqualCurrentColor(color) && editable ? statusClassNames.active : ""),
      key: color,
      onClick: function onClick() {
        if (disabled) {
          return;
        }
        handleClick(color);
      }
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: classNames__default["default"]("".concat(swatchesClass, "--item__color"), "".concat(baseClassName, "--bg-alpha"))
    }, /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(swatchesClass, "--item__inner"),
      style: {
        background: color
      }
    })));
  })));
};
var SwatchesPanel = /*#__PURE__*/React__default["default"].memo(Swatches);

exports["default"] = SwatchesPanel;
