// tdesign-react/cjs/calendar/CalendarCellComp.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var hooks_useConfig = require('../hooks/useConfig.js');
var calendar_hooks_usePrefixClass = require('./hooks/usePrefixClass.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var calendar__util = require('./_util.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-2b5ba78d.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-c17bb178.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-17e2e72f.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var CalendarCellComp = function CalendarCellComp(props) {
  var mode = props.mode,
    cell = props.cell,
    cellAppend = props.cellAppend,
    theme = props.theme,
    _props$isDisabled = props.isDisabled,
    isDisabled = _props$isDisabled === void 0 ? false : _props$isDisabled,
    cellData = props.cellData,
    isCurrent = props.isCurrent,
    isNow = props.isNow,
    fillWithZero = props.fillWithZero,
    createCalendarCell = props.createCalendarCell,
    onCellClick = props.onCellClick,
    onCellDoubleClick = props.onCellDoubleClick,
    onCellRightClick = props.onCellRightClick;
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("calendar"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var monthLabelList = t(local.cellMonth).split(",");
  var getMonthCN = function getMonthCN(month) {
    return monthLabelList[month];
  };
  var _useConfig = hooks_useConfig["default"](),
    calendarConfig = _useConfig.calendar;
  var fix0 = function fix0(num) {
    var _ref;
    var fillZero = num < 10 && ((_ref = fillWithZero !== null && fillWithZero !== void 0 ? fillWithZero : calendarConfig.fillWithZero) !== null && _ref !== void 0 ? _ref : true);
    return fillZero ? "0".concat(num) : num;
  };
  var prefixCls = calendar_hooks_usePrefixClass["default"]();
  return /* @__PURE__ */React__default["default"].createElement("td", {
    className: prefixCls([calendar__util.blockName, "table-body-cell"], isDisabled ? "is-disabled" : null, isCurrent ? "is-checked" : null, isNow ? [calendar__util.blockName, "table-body-cell--now"] : null),
    onClick: onCellClick,
    onDoubleClick: onCellDoubleClick,
    onContextMenu: onCellRightClick
  }, function () {
    if (cell && typeof cell === "function") return cell(mode === "month" ? createCalendarCell(cellData) : cellData);
    if (cell && typeof cell !== "function") return cell;
    var cellCtx;
    if (mode === "year") {
      var mIndex = cellData.date.getMonth();
      cellCtx = theme === "full" ? getMonthCN(mIndex) : t(local.monthSelection, {
        month: "".concat(mIndex + 1)
      });
    } else {
      cellCtx = fix0(cellData.date.getDate());
    }
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: prefixCls([calendar__util.blockName, "table-body-cell-display"])
    }, cellCtx);
  }(), function () {
    var cellCtx;
    if (cellAppend && typeof cellAppend === "function") {
      cellCtx = cellAppend(mode === "month" ? createCalendarCell(cellData) : cellData);
    }
    if (cellAppend && typeof cellAppend !== "function") {
      cellCtx = cellAppend;
    }
    return cellAppend && /* @__PURE__ */React__default["default"].createElement("div", {
      className: prefixCls([calendar__util.blockName, "table-body-cell-content"])
    }, cellCtx);
  }());
};

exports["default"] = CalendarCellComp;
