// tdesign-react/cjs/table/Cell.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _typeof = require('../_chunks/dep-c4d1ce0c.js');
var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var table_Ellipsis = require('./Ellipsis.js');
var table_hooks_useFixed = require('./hooks/useFixed.js');
var table_utils = require('./utils.js');
var isFunction = require('../_chunks/dep-cba78412.js');
var get = require('../_chunks/dep-05df1ea3.js');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_util/scroll.js');
require('raf');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-ecfe656e.js');
require('../_util/easing.js');
require('../tooltip/index.js');
require('../tooltip/Tooltip.js');
require('../_chunks/dep-1c586f5c.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../hooks/useLayoutEffect.js');
require('../hooks/useAnimation.js');
require('../hooks/useAttach.js');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../hooks/useDefaultProps.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../hooks/usePopper.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/listener.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');
require('../popup/utils/transition.js');
require('../popup/PopupPlugin.js');
require('../_chunks/dep-e37cb830.js');
require('../_util/react-render.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../tooltip/defaultProps.js');
require('../tooltip/TooltipLite.js');
require('../hooks/useSwitch.js');
require('../hooks/usePersistFn.js');
require('../_chunks/dep-2a85392e.js');
require('../hooks/useDebounce.js');
require('../_chunks/dep-b289b0be.js');
require('../_chunks/dep-fb673f7a.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-f0e884b9.js');
require('../hooks/useDeepEffect.js');
require('../hooks/usePrevious.js');
require('../_chunks/dep-e43e6728.js');
require('../_chunks/dep-4ef3269f.js');
require('../_chunks/dep-2a089253.js');
require('../_chunks/dep-048e4314.js');
require('../_chunks/dep-19400123.js');
require('../_chunks/dep-894a5ece.js');
require('../_chunks/dep-f7bf3d17.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-ca918872.js');
require('../_chunks/dep-13de2fef.js');
require('../_chunks/dep-63308b84.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function renderCell(params, extra) {
  var col = params.col,
    row = params.row,
    rowIndex = params.rowIndex;
  if (col.colKey === "serial-number") {
    var _ref = (extra === null || extra === void 0 ? void 0 : extra.pagination) || {},
      current = _ref.current,
      pageSize = _ref.pageSize,
      defaultCurrent = _ref.defaultCurrent,
      defaultPageSize = _ref.defaultPageSize;
    var tCurrent = current || defaultCurrent;
    var tPageSize = pageSize || defaultPageSize;
    if (tPageSize && tCurrent) {
      return tPageSize * (tCurrent - 1) + rowIndex + 1;
    }
    return rowIndex + 1;
  }
  if (isFunction.isFunction(col.cell)) {
    return col.cell(params);
  }
  if (isFunction.isFunction(col.render)) {
    return col.render(_objectSpread(_objectSpread({}, params), {}, {
      type: "cell"
    }));
  }
  var r = col.cell || col.render || get.get(row, col.colKey);
  if (![void 0, "", null].includes(r)) return r;
  if (extra !== null && extra !== void 0 && extra.cellEmptyContent) return extra.cellEmptyContent;
  return r;
}
function renderEllipsisCell(cellParams, params) {
  var cellNode = params.cellNode,
    tableElm = params.tableElm,
    columnLength = params.columnLength,
    classPrefix = params.classPrefix,
    overlayClassName = params.overlayClassName;
  var col = cellParams.col,
    colIndex = cellParams.colIndex;
  var content = isFunction.isFunction(col.ellipsis) ? col.ellipsis(cellParams) : void 0;
  if (_typeof._typeof(col.ellipsis) === "object" && "content" in col.ellipsis && isFunction.isFunction(col.ellipsis.content)) {
    content = col.ellipsis.content(cellParams);
  }
  var tooltipProps = {};
  if (_typeof._typeof(col.ellipsis) === "object") {
    tooltipProps = "props" in col.ellipsis ? col.ellipsis.props : col.ellipsis || void 0;
  }
  var tableElement = tableElm;
  var placement = colIndex === 0 ? "top-left" : "top";
  placement = colIndex === columnLength - 1 ? "top-right" : placement;
  return /* @__PURE__ */React__default["default"].createElement(table_Ellipsis["default"], {
    placement: placement,
    attach: tableElement ? function () {
      return tableElement;
    } : void 0,
    popupContent: content,
    tooltipProps: tooltipProps,
    overlayClassName: overlayClassName,
    classPrefix: classPrefix
  }, cellNode);
}
var Cell = function Cell(props) {
  var cellParams = props.cellParams,
    tableClassNames = props.tableClassNames,
    tableRef = props.tableRef,
    columnLength = props.columnLength,
    classPrefix = props.classPrefix,
    overlayClassName = props.overlayClassName,
    pagination = props.pagination;
  var col = cellParams.col,
    colIndex = cellParams.colIndex,
    rowIndex = cellParams.rowIndex;
  var cellSpans = props.cellSpans,
    dataLength = props.dataLength,
    rowAndColFixedPosition = props.rowAndColFixedPosition,
    cellEmptyContent = props.cellEmptyContent,
    rowspanAndColspan = props.rowspanAndColspan,
    onClick = props.onClick;
  var tableColFixedClasses = tableClassNames.tableColFixedClasses,
    tdEllipsisClass = tableClassNames.tdEllipsisClass,
    tableBaseClass = tableClassNames.tableBaseClass,
    tdAlignClasses = tableClassNames.tdAlignClasses,
    tableDraggableClasses = tableClassNames.tableDraggableClasses;
  var cellNode = renderCell(cellParams, {
    cellEmptyContent: cellEmptyContent,
    pagination: pagination
  });
  var tdStyles = table_hooks_useFixed.getColumnFixedStyles(col, colIndex, rowAndColFixedPosition, tableColFixedClasses);
  var customClasses = table_utils.formatClassNames(col.className, _objectSpread(_objectSpread({}, cellParams), {}, {
    type: "td"
  }));
  var classes = [tdStyles.classes, customClasses, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, tdEllipsisClass, col.ellipsis), tableBaseClass.tdLastRow, rowIndex + cellSpans.rowspan === dataLength), tableBaseClass.tdFirstCol, colIndex === 0 && rowspanAndColspan), tdAlignClasses[col.align], col.align && col.align !== "left"), tableDraggableClasses.handle, col.colKey === "drag")];
  var normalAttrs = isFunction.isFunction(col.attrs) ? col.attrs(_objectSpread(_objectSpread({}, cellParams), {}, {
    type: "td"
  })) : col.attrs;
  var attrs = _objectSpread(_objectSpread({}, normalAttrs), {}, {
    rowSpan: cellSpans.rowspan,
    colSpan: cellSpans.colspan
  });
  return /* @__PURE__ */React__default["default"].createElement("td", _objectSpread(_objectSpread({
    key: col.colKey || colIndex,
    className: classNames__default["default"](classes) || void 0,
    style: tdStyles.style
  }, attrs), {}, {
    onClick: onClick
  }), col.ellipsis ? renderEllipsisCell(cellParams, {
    cellNode: cellNode,
    tableElm: tableRef.current,
    columnLength: columnLength,
    classPrefix: classPrefix,
    overlayClassName: overlayClassName
  }) : cellNode);
};
Cell.displayName = "Cell";

exports["default"] = Cell;
exports.renderCell = renderCell;
