// tdesign-react/cjs/table/PrimaryTable.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var table_BaseTable = require('./BaseTable.js');
var table_hooks_useColumnController = require('./hooks/useColumnController.js');
var table_hooks_useRowExpand = require('./hooks/useRowExpand.js');
var table_hooks_useTableHeader = require('./hooks/useTableHeader.js');
var table_hooks_useRowSelect = require('./hooks/useRowSelect.js');
var table_hooks_useSorter = require('./hooks/useSorter.js');
var table_hooks_useFilter = require('./hooks/useFilter.js');
var table_hooks_useDragSort = require('./hooks/useDragSort.js');
var table_hooks_useAsyncLoading = require('./hooks/useAsyncLoading.js');
var table_hooks_useClassName = require('./hooks/useClassName.js');
var table_hooks_useStyle = require('./hooks/useStyle.js');
var table_EditableCell = require('./EditableCell.js');
var table_hooks_useEditableRow = require('./hooks/useEditableRow.js');
var table_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var get = require('../_chunks/dep-05df1ea3.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-b289b0be.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-1c586f5c.js');
require('../_chunks/dep-f0e884b9.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-794b42ce.js');
require('../affix/index.js');
require('../affix/Affix.js');
require('../_chunks/dep-68f5fe7b.js');
require('../_util/scroll.js');
require('raf');
require('../_util/easing.js');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../affix/defaultProps.js');
require('../hooks/useElementLazyRender.js');
require('../_chunks/dep-e31ce0e8.js');
require('../hooks/useVirtualScroll.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../hooks/useDomRefCallback.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('./TBody.js');
require('../locale/LocalReceiver.js');
require('./hooks/useRowspanAndColspan.js');
require('./TR.js');
require('./Cell.js');
require('./Ellipsis.js');
require('../tooltip/index.js');
require('../tooltip/Tooltip.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../hooks/useAnimation.js');
require('../hooks/useAttach.js');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/listener.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');
require('../popup/utils/transition.js');
require('../popup/PopupPlugin.js');
require('../tooltip/defaultProps.js');
require('../tooltip/TooltipLite.js');
require('../hooks/useSwitch.js');
require('../hooks/usePersistFn.js');
require('../_chunks/dep-2a85392e.js');
require('../hooks/useDebounce.js');
require('./hooks/useFixed.js');
require('../_chunks/dep-fb673f7a.js');
require('../hooks/useDeepEffect.js');
require('../hooks/usePrevious.js');
require('./utils.js');
require('../_chunks/dep-e43e6728.js');
require('../_chunks/dep-4ef3269f.js');
require('../_chunks/dep-2a089253.js');
require('../_chunks/dep-048e4314.js');
require('../_chunks/dep-19400123.js');
require('../_chunks/dep-894a5ece.js');
require('../_chunks/dep-f7bf3d17.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-ca918872.js');
require('../_chunks/dep-13de2fef.js');
require('../_chunks/dep-63308b84.js');
require('./hooks/useLazyLoad.js');
require('../_chunks/dep-e44b03ec.js');
require('./TFoot.js');
require('./THead.js');
require('./hooks/useAffix.js');
require('./hooks/useColumnResize.js');
require('./hooks/usePagination.js');
require('../pagination/index.js');
require('../pagination/Pagination.js');
require('../select/index.js');
require('../select/base/Select.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../_util/helper.js');
require('../_util/parseTNode.js');
require('../common/FakeArrow.js');
require('../select-input/index.js');
require('../select-input/SelectInput.js');
require('../select-input/defaultProps.js');
require('../select-input/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/TagInput.js');
require('tdesign-icons-react');
require('../hooks/useDragSorter.js');
require('../hooks/useEventCallback.js');
require('../hooks/useGlobalIcon.js');
require('../input/index.js');
require('../input/Input.js');
require('../input/defaultProps.js');
require('../input/InputGroup.js');
require('../input/useLengthLimit.js');
require('../tag-input/defaultProps.js');
require('../tag-input/useHover.js');
require('../tag-input/useTagList.js');
require('../tag/index.js');
require('../tag/Tag.js');
require('../_chunks/dep-d14bfcfa.js');
require('../tag/defaultProps.js');
require('../tag/CheckTag.js');
require('../_chunks/dep-dd0f5f6b.js');
require('../tag/CheckTagGroup.js');
require('../tag-input/useTagScroll.js');
require('../select-input/useOverlayInnerStyle.js');
require('../hooks/useInnerPopupVisible.js');
require('../select-input/useSingle.js');
require('../select/defaultProps.js');
require('../select/hooks/useKeyboardControl.js');
require('../_chunks/dep-bdc97672.js');
require('../hooks/useRipple.js');
require('../select/base/OptionGroup.js');
require('../select/base/PopupContent.js');
require('../select/hooks/usePanelVirtualScroll.js');
require('../input-number/index.js');
require('../input-number/InputNumber.js');
require('../button/index.js');
require('../button/Button.js');
require('../button/defaultProps.js');
require('../input-number/useInputNumber.js');
require('../_chunks/dep-04deac0d.js');
require('../_chunks/dep-8f62fb96.js');
require('../hooks/useCommonClassName.js');
require('../input-number/defaultProps.js');
require('../input-adornment/InputAdornment.js');
require('../pagination/hooks/useBoundaryJumper.js');
require('../pagination/hooks/usePrevNextJumper.js');
require('../pagination/hooks/usePageNumber.js');
require('../pagination/hooks/useTotal.js');
require('../pagination/validators.js');
require('../pagination/defaultProps.js');
require('../_chunks/dep-5c3139cb.js');
require('../_chunks/dep-2e455d91.js');
require('../pagination/PaginationMini.js');
require('./hooks/useMultiHeader.js');
require('../checkbox/index.js');
require('../_chunks/dep-83710d84.js');
require('../common/Check.js');
require('../_chunks/dep-089e8106.js');
require('../checkbox/defaultProps.js');
require('../dialog/plugin.js');
require('../dialog/Dialog.js');
require('../hooks/useSetState.js');
require('../dialog/defaultProps.js');
require('../dialog/DialogCard.js');
require('../dialog/hooks/useDialogDrag.js');
require('../hooks/useMouseEvent.js');
require('../dialog/hooks/useDialogEsc.js');
require('../dialog/hooks/useDialogPosition.js');
require('../dialog/hooks/useLockStyle.js');
require('../_chunks/dep-f94fba39.js');
require('../_chunks/dep-f92ec514.js');
require('../radio/index.js');
require('../_chunks/dep-27d626a1.js');
require('../radio/defaultProps.js');
require('../radio/useKeyboard.js');
require('./SorterButton.js');
require('./FilterController.js');
require('../_chunks/dep-545dd164.js');
require('sortablejs');
require('../_chunks/dep-c17bb178.js');
require('../form/formModel.js');
require('../_chunks/dep-57a4aad8.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var PrimaryTable = /*#__PURE__*/React.forwardRef(function (originalProps, ref) {
  var props = hooks_useDefaultProps["default"](originalProps, table_defaultProps.primaryTableDefaultProps);
  var columns = props.columns,
    columnController = props.columnController,
    editableRowKeys = props.editableRowKeys,
    style = props.style,
    className = props.className;
  var primaryTableRef = React.useRef(null);
  var innerPagination = React.useRef(props.pagination);
  var _useClassName = table_hooks_useClassName["default"](),
    classPrefix = _useClassName.classPrefix,
    tableDraggableClasses = _useClassName.tableDraggableClasses,
    tableBaseClass = _useClassName.tableBaseClass,
    tableSelectedClasses = _useClassName.tableSelectedClasses,
    tableSortClasses = _useClassName.tableSortClasses;
  var _useStyle = table_hooks_useStyle["default"](props),
    sizeClassNames = _useStyle.sizeClassNames;
  var _useColumnController = table_hooks_useColumnController["default"](props, {
      onColumnReduce: onColumnReduce
    }),
    tDisplayColumns = _useColumnController.tDisplayColumns,
    renderColumnController = _useColumnController.renderColumnController;
  var _useRowExpand = table_hooks_useRowExpand["default"](props),
    showExpandedRow = _useRowExpand.showExpandedRow,
    showExpandIconColumn = _useRowExpand.showExpandIconColumn,
    getExpandColumn = _useRowExpand.getExpandColumn,
    renderExpandedRow = _useRowExpand.renderExpandedRow,
    onInnerExpandRowClick = _useRowExpand.onInnerExpandRowClick,
    getExpandedRowClass = _useRowExpand.getExpandedRowClass;
  var _useSorter = table_hooks_useSorter["default"](props),
    renderSortIcon = _useSorter.renderSortIcon;
  var _useRowSelect = table_hooks_useRowSelect["default"](props, tableSelectedClasses),
    selectedRowClassNames = _useRowSelect.selectedRowClassNames,
    setCurrentPaginateData = _useRowSelect.setCurrentPaginateData,
    formatToRowSelectColumn = _useRowSelect.formatToRowSelectColumn,
    setTSelectedRowKeys = _useRowSelect.setTSelectedRowKeys,
    onInnerSelectRowClick = _useRowSelect.onInnerSelectRowClick;
  var _useFilter = table_hooks_useFilter["default"](props, primaryTableRef),
    hasEmptyCondition = _useFilter.hasEmptyCondition,
    isTableOverflowHidden = _useFilter.isTableOverflowHidden,
    renderFilterIcon = _useFilter.renderFilterIcon,
    renderFirstFilterRow = _useFilter.renderFirstFilterRow;
  var _useDragSort = table_hooks_useDragSort["default"](props, {
      primaryTableRef: primaryTableRef,
      innerPagination: innerPagination
    }),
    isRowHandlerDraggable = _useDragSort.isRowHandlerDraggable,
    isRowDraggable = _useDragSort.isRowDraggable,
    isColDraggable = _useDragSort.isColDraggable,
    setDragSortColumns = _useDragSort.setDragSortColumns;
  var _useTableHeader = table_hooks_useTableHeader["default"]({
      columns: props.columns
    }),
    renderTitleWidthIcon = _useTableHeader.renderTitleWidthIcon;
  var _useAsyncLoading = table_hooks_useAsyncLoading["default"](props),
    renderAsyncLoading = _useAsyncLoading.renderAsyncLoading;
  var primaryTableClasses = defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, tableDraggableClasses.colDraggable, isColDraggable), tableDraggableClasses.rowHandlerDraggable, isRowHandlerDraggable), tableDraggableClasses.rowDraggable, isRowDraggable), tableBaseClass.overflowVisible, isTableOverflowHidden === false), tableBaseClass.tableRowEdit, editableRowKeys);
  var _useEditableRow = table_hooks_useEditableRow.useEditableRow(props),
    errorListMap = _useEditableRow.errorListMap,
    editableKeysMap = _useEditableRow.editableKeysMap,
    validateRowData = _useEditableRow.validateRowData,
    validateTableData = _useEditableRow.validateTableData,
    onRuleChange = _useEditableRow.onRuleChange,
    clearValidateData = _useEditableRow.clearValidateData;
  var tRowClassNames = function () {
    var tClassNames = [props.rowClassName, selectedRowClassNames, getExpandedRowClass];
    return tClassNames.filter(function (v) {
      return v;
    });
  }();
  var tRowAttributes = function () {
    var tAttributes = [props.rowAttributes];
    if (isRowHandlerDraggable || isRowDraggable) {
      tAttributes.push(function (_ref) {
        var row = _ref.row;
        return {
          "data-id": get.get(row, props.rowKey || "id")
        };
      });
    }
    return tAttributes.filter(function (v) {
      return v;
    });
  }();
  React.useImperativeHandle(ref, function () {
    return _objectSpread({
      validateRowData: validateRowData,
      validateTableData: validateTableData,
      clearValidateData: clearValidateData
    }, primaryTableRef.current);
  });
  function onColumnReduce(reduceKeys) {
    if (props.resizable) {
      primaryTableRef.current.updateTableWidthOnColumnChange(reduceKeys);
    }
  }
  var _getColumns = function getColumns(columns2) {
    var arr = [];
    var _loop = function _loop() {
      var _item$children, _item$edit, _item$children2, _item$children3;
      var item = _objectSpread({}, columns2[i]);
      var isDisplayColumn = ((_item$children = item.children) === null || _item$children === void 0 ? void 0 : _item$children.length) || (tDisplayColumns === null || tDisplayColumns === void 0 ? void 0 : tDisplayColumns.includes(item.colKey));
      if (!isDisplayColumn && props.columnController && tDisplayColumns) return 1; // continue
      item = formatToRowSelectColumn(item);
      var sort = props.sort;
      if (item.sorter && props.showSortColumnBgColor) {
        var sorts = sort instanceof Array ? sort : [sort];
        var sortedColumn = sorts.find(function (sort2) {
          return sort2 && sort2.sortBy === item.colKey && sort2.descending !== void 0;
        });
        if (sortedColumn) {
          item.className = item.className instanceof Array ? item.className.concat(tableSortClasses.sortColumn) : [item.className, tableSortClasses.sortColumn];
        }
      }
      if (item.sorter || item.filter) {
        var titleContent = table_hooks_useTableHeader.renderTitle(item, i);
        var _item = item,
          ellipsisTitle = _item.ellipsisTitle;
        item.title = function (p) {
          var _primaryTableRef$curr;
          var sortIcon = item.sorter ? renderSortIcon(p) : null;
          var filterIcon = item.filter ? renderFilterIcon(p) : null;
          var attach = (_primaryTableRef$curr = primaryTableRef.current) === null || _primaryTableRef$curr === void 0 ? void 0 : _primaryTableRef$curr.tableContentRef;
          return renderTitleWidthIcon([titleContent, sortIcon, filterIcon], p.col, p.colIndex, ellipsisTitle, attach, {
            classPrefix: classPrefix,
            ellipsisOverlayClassName: props.size !== "medium" ? sizeClassNames[props.size] : ""
          });
        };
        item.ellipsisTitle = false;
      }
      if ((_item$edit = item.edit) !== null && _item$edit !== void 0 && _item$edit.component) {
        var oldCell = item.cell;
        item.cell = function (p) {
          var cellProps = _objectSpread(_objectSpread({}, p), {}, {
            oldCell: oldCell,
            tableBaseClass: tableBaseClass,
            cellEmptyContent: props.cellEmptyContent,
            onChange: props.onRowEdit,
            onValidate: props.onRowValidate,
            onRuleChange: onRuleChange
          });
          if (props.editableRowKeys) {
            var rowValue = get.get(p.row, props.rowKey || "id");
            cellProps.editable = editableKeysMap[rowValue] || false;
            var key = [rowValue, p.col.colKey].join("__");
            var errorList = errorListMap[key];
            errorList && (cellProps.errors = errorList);
          }
          if (props.editableCellState) {
            cellProps.readonly = !props.editableCellState(p);
          }
          return /* @__PURE__ */React__default["default"].createElement(table_EditableCell["default"], _objectSpread({}, cellProps));
        };
      }
      if ((_item$children2 = item.children) !== null && _item$children2 !== void 0 && _item$children2.length) {
        item.children = _getColumns(item.children);
      }
      if (!item.children || (_item$children3 = item.children) !== null && _item$children3 !== void 0 && _item$children3.length) {
        arr.push(item);
      }
    };
    for (var i = 0, len = columns2.length; i < len; i++) {
      if (_loop()) continue;
    }
    return arr;
  };
  var tColumns = function () {
    var cols = _getColumns(columns);
    if (showExpandIconColumn) {
      cols.unshift(getExpandColumn());
    }
    return cols;
  }();
  var onInnerPageChange = function onInnerPageChange(pageInfo, newData) {
    var _props$onPageChange, _props$onChange;
    innerPagination.current = _objectSpread(_objectSpread({}, innerPagination), pageInfo);
    setCurrentPaginateData(newData);
    (_props$onPageChange = props.onPageChange) === null || _props$onPageChange === void 0 || _props$onPageChange.call(props, pageInfo, newData);
    var changeParams = [{
      pagination: pageInfo
    }, {
      trigger: "pagination",
      currentData: newData
    }];
    (_props$onChange = props.onChange) === null || _props$onChange === void 0 || _props$onChange.call.apply(_props$onChange, [props].concat(changeParams));
    if (!props.reserveSelectedRowOnPaginate) {
      setTSelectedRowKeys([], {
        selectedRowData: [],
        type: "uncheck",
        currentRowKey: "CLEAR_ON_PAGINATE"
      });
    }
  };
  var onInnerRowClick = function onInnerRowClick(params) {
    if (props.expandOnRowClick) {
      onInnerExpandRowClick(params);
    }
    if (props.selectOnRowClick) {
      onInnerSelectRowClick(params);
    }
  };
  function formatNode(api, renderInnerNode, condition, extra) {
    if (!condition) return props[api];
    var innerNode = renderInnerNode();
    var propsNode = props[api];
    if (innerNode && !propsNode) return innerNode;
    if (propsNode && !innerNode) return propsNode;
    if (innerNode && propsNode) {
      return extra !== null && extra !== void 0 && extra.reverse ? /* @__PURE__ */React__default["default"].createElement("div", null, innerNode, propsNode) : /* @__PURE__ */React__default["default"].createElement("div", null, propsNode, innerNode);
    }
    return null;
  }
  var isColumnController = !!(columnController && Object.keys(columnController).length);
  var placement = isColumnController ? columnController.placement || "top-right" : "";
  var isBottomController = isColumnController && (placement === null || placement === void 0 ? void 0 : placement.indexOf("bottom")) !== -1;
  var topContent = formatNode("topContent", renderColumnController, isColumnController && !isBottomController);
  var bottomContent = formatNode("bottomContent", renderColumnController, isBottomController, {
    reverse: true
  });
  var firstFullRow = formatNode("firstFullRow", renderFirstFilterRow, !hasEmptyCondition);
  var lastFullRow = formatNode("lastFullRow", renderAsyncLoading, !!props.asyncLoading);
  var baseTableProps = _objectSpread(_objectSpread({}, props), {}, {
    rowClassName: tRowClassNames,
    rowAttributes: tRowAttributes,
    columns: tColumns,
    topContent: topContent,
    bottomContent: bottomContent,
    firstFullRow: firstFullRow,
    lastFullRow: lastFullRow,
    thDraggable: ["col", "row-handler-col"].includes(props.dragSort),
    onPageChange: onInnerPageChange,
    renderExpandedRow: showExpandedRow ? renderExpandedRow : void 0
  });
  if (props.expandOnRowClick || props.selectOnRowClick) {
    baseTableProps.onRowClick = onInnerRowClick;
  }
  return /* @__PURE__ */React__default["default"].createElement(table_BaseTable["default"], _objectSpread(_objectSpread({
    ref: primaryTableRef
  }, baseTableProps), {}, {
    className: classNames__default["default"](primaryTableClasses, className),
    style: style,
    onLeafColumnsChange: setDragSortColumns
  }));
});
PrimaryTable.displayName = "PrimaryTable";

exports.BASE_TABLE_ALL_EVENTS = table_BaseTable.BASE_TABLE_ALL_EVENTS;
exports["default"] = PrimaryTable;
