// tdesign-react/cjs/table/TBody.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var table_hooks_useRowspanAndColspan = require('./hooks/useRowspanAndColspan.js');
var table_TR = require('./TR.js');
var pick = require('../_chunks/dep-f7bf3d17.js');
var get = require('../_chunks/dep-05df1ea3.js');
var camelCase = require('../_chunks/dep-e44b03ec.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-ecfe656e.js');
require('../config-provider/ConfigProvider.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-b289b0be.js');
require('../hooks/useLayoutEffect.js');
require('../_chunks/dep-68f5fe7b.js');
require('./Cell.js');
require('./Ellipsis.js');
require('../_util/scroll.js');
require('raf');
require('../_util/easing.js');
require('../tooltip/index.js');
require('../tooltip/Tooltip.js');
require('../_chunks/dep-1c586f5c.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useConfig.js');
require('../hooks/useAnimation.js');
require('../hooks/useAttach.js');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../hooks/useDefaultProps.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../hooks/usePopper.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/listener.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');
require('../popup/utils/transition.js');
require('../popup/PopupPlugin.js');
require('../_chunks/dep-e37cb830.js');
require('../_util/react-render.js');
require('../common/PluginContainer.js');
require('../tooltip/defaultProps.js');
require('../tooltip/TooltipLite.js');
require('../hooks/useSwitch.js');
require('../hooks/usePersistFn.js');
require('../_chunks/dep-2a85392e.js');
require('../hooks/useDebounce.js');
require('./hooks/useFixed.js');
require('../_chunks/dep-fb673f7a.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-f0e884b9.js');
require('../hooks/useDeepEffect.js');
require('../hooks/usePrevious.js');
require('./utils.js');
require('../_chunks/dep-e43e6728.js');
require('../_chunks/dep-4ef3269f.js');
require('../_chunks/dep-2a089253.js');
require('../_chunks/dep-048e4314.js');
require('../_chunks/dep-19400123.js');
require('../_chunks/dep-894a5ece.js');
require('./hooks/useClassName.js');
require('./hooks/useLazyLoad.js');
require('../_chunks/dep-e31ce0e8.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-ca918872.js');
require('../_chunks/dep-13de2fef.js');
require('../_chunks/dep-63308b84.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var ROW_AND_TD_LISTENERS = table_TR.ROW_LISTENERS.concat("cell-click");
var extendTableProps = ["rowKey", "rowClassName", "rowAttributes", "loading", "empty", "fixedRows", "firstFullRow", "lastFullRow", "rowspanAndColspan", "scroll", "cellEmptyContent", "onCellClick", "onPageChange", "onRowClick", "onRowDblclick", "onRowMouseover", "onRowMousedown", "onRowMouseenter", "onRowMouseleave", "onRowMouseup", "onScroll", "onScrollX", "onScrollY"];
var trProperties = ["classPrefix", "ellipsisOverlayClassName", "rowAndColFixedPosition", "scroll", "tableRef", "tableContentRef", "trs", "bufferSize", "isVirtual", "rowHeight", "scrollType"];
function TBody(props) {
  var data = props.data,
    columns = props.columns,
    rowKey = props.rowKey,
    firstFullRow = props.firstFullRow,
    lastFullRow = props.lastFullRow,
    virtualConfig = props.virtualConfig,
    allTableClasses = props.allTableClasses;
  var isVirtualScroll = virtualConfig.isVirtualScroll;
  var renderData = isVirtualScroll ? virtualConfig.visibleData : data;
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("table"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    global = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var _useRowspanAndColspan = table_hooks_useRowspanAndColspan["default"](renderData, columns, rowKey, props.rowspanAndColspan),
    skipSpansMap = _useRowspanAndColspan.skipSpansMap;
  var isSkipSnapsMapNotFinish = Boolean(props.rowspanAndColspan && !skipSpansMap.size);
  var tableFullRowClasses = allTableClasses.tableFullRowClasses,
    tableBaseClass = allTableClasses.tableBaseClass;
  var tbodyClasses = React.useMemo(function () {
    return [tableBaseClass.body];
  }, [tableBaseClass.body]);
  var hasFullRowConfig = React.useMemo(function () {
    return firstFullRow || lastFullRow;
  }, [firstFullRow, lastFullRow]);
  var isEmpty = !(data !== null && data !== void 0 && data.length) && !props.loading && !hasFullRowConfig;
  var renderEmpty = function renderEmpty(columns2) {
    return /* @__PURE__ */React__default["default"].createElement("tr", {
      className: classNames__default["default"]([tableBaseClass.emptyRow, defineProperty._defineProperty({}, tableFullRowClasses.base, props.isWidthOverflow)])
    }, /* @__PURE__ */React__default["default"].createElement("td", {
      colSpan: columns2.length
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: classNames__default["default"]([tableBaseClass.empty, defineProperty._defineProperty({}, tableFullRowClasses.innerFullRow, props.isWidthOverflow)]),
      style: props.isWidthOverflow ? {
        width: "".concat(props.tableWidth, "px")
      } : {}
    }, props.empty || t(global.empty))));
  };
  var renderFullRow = function renderFullRow(type) {
    var fullRowNode = {
      "first-full-row": firstFullRow,
      "last-full-row": lastFullRow
    }[type];
    if (!fullRowNode) return null;
    var rowProps = {};
    if (isVirtualScroll) {
      var fullRowIndex = {
        "first-full-row": 0,
        "last-full-row": data.length - 1
      }[type];
      var rowVisible = renderData.some(function (item) {
        return item.__VIRTUAL_SCROLL_INDEX === fullRowIndex;
      });
      if (!rowVisible) return null;
      rowProps.ref = function (ref) {
        if (ref) {
          props === null || props === void 0 || props.handleRowMounted({
            ref: ref,
            data: data[fullRowIndex]
          });
        }
      };
    }
    var isFixedToLeft = props.isWidthOverflow && columns.find(function (col) {
      return col.fixed === "left";
    });
    var tType = camelCase.camelCase(type);
    var classes = [tableFullRowClasses.base, tableFullRowClasses[tType]];
    return /* @__PURE__ */React__default["default"].createElement("tr", _objectSpread({
      key: type,
      className: classNames__default["default"](classes)
    }, rowProps), /* @__PURE__ */React__default["default"].createElement("td", {
      colSpan: columns.length
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: classNames__default["default"](defineProperty._defineProperty({}, tableFullRowClasses.innerFullRow, isFixedToLeft)) || void 0,
      style: isFixedToLeft ? {
        width: "".concat(props.tableWidth, "px")
      } : {}
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: tableFullRowClasses.innerFullElement
    }, fullRowNode))));
  };
  var renderExpandedRows = function renderExpandedRows(row, rowIndex) {
    var _props$renderExpanded;
    var p = {
      row: row,
      index: rowIndex,
      columns: columns,
      tableWidth: props.tableWidth,
      isWidthOverflow: props.isWidthOverflow
    };
    return (_props$renderExpanded = props.renderExpandedRow) === null || _props$renderExpanded === void 0 ? void 0 : _props$renderExpanded.call(props, p);
  };
  var renderTRNodeList = function renderTRNodeList() {
    if (isSkipSnapsMapNotFinish) return null;
    var trNodeList = [];
    var firstFullRowNode = renderFullRow("first-full-row");
    firstFullRowNode && trNodeList.push(firstFullRowNode);
    renderData === null || renderData === void 0 || renderData.forEach(function (row, rowIndex) {
      if (row.__VIRTUAL_FAKE_DATA) return;
      var getRowIndex = function getRowIndex() {
        var virtualIndex = row.__VIRTUAL_SCROLL_INDEX;
        if (isVirtualScroll && firstFullRow && virtualIndex) {
          return virtualIndex - 1;
        }
        return virtualIndex !== null && virtualIndex !== void 0 ? virtualIndex : rowIndex;
      };
      var trProps = _objectSpread(_objectSpread(_objectSpread({}, pick.pick(props, table_TR.TABLE_PROPS)), {}, {
        rowKey: props.rowKey || "id",
        row: row,
        columns: columns,
        rowIndex: getRowIndex(),
        dataLength: data.length,
        skipSpansMap: skipSpansMap,
        virtualConfig: virtualConfig,
        classPrefix: props.classPrefix,
        ellipsisOverlayClassName: props.ellipsisOverlayClassName
      }, pick.pick(props, trProperties)), {}, {
        pagination: props.pagination
      });
      if (props.onCellClick) {
        trProps.onCellClick = props.onCellClick;
      }
      var trNode = /* @__PURE__ */React__default["default"].createElement(table_TR["default"], _objectSpread(_objectSpread({
        key: get.get(row, props.rowKey || "id") || rowIndex
      }, trProps), {}, {
        onRowMounted: props.handleRowMounted
      }));
      trNodeList.push(trNode);
      var expandedRowNode = renderExpandedRows(row, rowIndex);
      expandedRowNode && trNodeList.push(expandedRowNode);
    });
    var lastFullRowNode = renderFullRow("last-full-row");
    lastFullRowNode && trNodeList.push(lastFullRowNode);
    return trNodeList;
  };
  var translate = "translateY(".concat(virtualConfig.translateY, "px)");
  var posStyle = isVirtualScroll ? {
    transform: translate,
    msTransform: translate,
    MozTransform: translate,
    WebkitTransform: translate
  } : void 0;
  return /* @__PURE__ */React__default["default"].createElement("tbody", {
    className: classNames__default["default"](tbodyClasses) || void 0,
    style: _objectSpread({}, posStyle)
  }, isEmpty ? renderEmpty(columns) : renderTRNodeList());
}
TBody.displayName = "TBody";

exports.ROW_AND_TD_LISTENERS = ROW_AND_TD_LISTENERS;
exports["default"] = TBody;
exports.extendTableProps = extendTableProps;
