// tdesign-react/cjs/table/TFoot.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var table_utils = require('./utils.js');
var table_hooks_useFixed = require('./hooks/useFixed.js');
var table_hooks_useRowspanAndColspan = require('./hooks/useRowspanAndColspan.js');
var table_hooks_useClassName = require('./hooks/useClassName.js');
var isFunction = require('../_chunks/dep-cba78412.js');
var get = require('../_chunks/dep-05df1ea3.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-1c586f5c.js');
require('../_chunks/dep-f0e884b9.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-b289b0be.js');
require('../_chunks/dep-fb673f7a.js');
require('../_util/listener.js');
require('../_chunks/dep-68f5fe7b.js');
require('../hooks/useDeepEffect.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-0bed591e.js');
require('../hooks/usePrevious.js');
require('../_chunks/dep-e43e6728.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-4ef3269f.js');
require('../_chunks/dep-2a089253.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-048e4314.js');
require('../_chunks/dep-19400123.js');
require('../_chunks/dep-894a5ece.js');
require('../_chunks/dep-f7bf3d17.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-ca918872.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-13de2fef.js');
require('../_chunks/dep-63308b84.js');
require('../hooks/useLayoutEffect.js');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-2d5cf7d6.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function TFoot(props) {
  var footData = props.footData,
    columns = props.columns,
    rowKey = props.rowKey,
    footerSummary = props.footerSummary,
    virtualScroll = props.virtualScroll;
  var tfooterRef = React.useRef(null);
  var classnames = table_hooks_useClassName["default"]();
  var _useRowspanAndColspan = table_hooks_useRowspanAndColspan["default"](footData, columns, rowKey, props.rowspanAndColspanInFooter),
    skipSpansMap = _useRowspanAndColspan.skipSpansMap;
  var renderTFootCell = function renderTFootCell(p) {
    var col = p.col,
      row = p.row;
    if (isFunction.isFunction(col.foot)) {
      return col.foot(p);
    }
    return col.foot || get.get(row, col.colKey);
  };
  var theadClasses = [classnames.tableFooterClasses.footer, defineProperty._defineProperty({}, classnames.tableFooterClasses.fixed, props.isFixedHeader)];
  if (!columns) return null;
  var footerDomList = footData === null || footData === void 0 ? void 0 : footData.map(function (row, rowIndex) {
    var trAttributes = table_utils.formatRowAttributes(props.rowAttributes, {
      row: row,
      rowIndex: rowIndex,
      type: "foot"
    });
    var customClasses = table_utils.formatRowClassNames(props.rowClassName, {
      row: row,
      rowIndex: rowIndex,
      type: "foot"
    }, props.rowKey || "id");
    return /* @__PURE__ */React__default["default"].createElement("tr", _objectSpread(_objectSpread({
      key: rowIndex
    }, trAttributes), {}, {
      className: classNames__default["default"](customClasses) || void 0
    }), columns.map(function (col, colIndex) {
      var _props$thWidthList;
      var cellSpans = {};
      var spanState = null;
      if (skipSpansMap.size) {
        var _spanState, _spanState2;
        var cellKey = table_hooks_useRowspanAndColspan.getCellKey(row, rowKey, col.colKey, colIndex);
        spanState = skipSpansMap.get(cellKey) || {};
        ((_spanState = spanState) === null || _spanState === void 0 ? void 0 : _spanState.rowspan) > 1 && (cellSpans.rowspan = spanState.rowspan);
        ((_spanState2 = spanState) === null || _spanState2 === void 0 ? void 0 : _spanState2.colspan) > 1 && (cellSpans.colspan = spanState.colspan);
        if (spanState.skipped) return null;
      }
      var tdStyles = table_hooks_useFixed.getColumnFixedStyles(col, colIndex, props.rowAndColFixedPosition, classnames.tableColFixedClasses);
      var style = _objectSpread({}, tdStyles.style);
      if ((_props$thWidthList = props.thWidthList) !== null && _props$thWidthList !== void 0 && _props$thWidthList[col.colKey]) {
        style.width = "".concat(props.thWidthList[col.colKey] || 0, "px");
      }
      return /* @__PURE__ */React__default["default"].createElement("td", {
        key: col.colKey,
        rowSpan: cellSpans.rowspan,
        colSpan: cellSpans.colspan,
        className: classNames__default["default"](tdStyles.classes) || void 0,
        style: style
      }, renderTFootCell({
        row: row,
        rowIndex: rowIndex,
        col: col,
        colIndex: colIndex
      }));
    }));
  });
  if (!footerSummary && (!props.footData || !props.footData.length)) return null;
  return /* @__PURE__ */React__default["default"].createElement("tfoot", {
    ref: tfooterRef,
    className: classNames__default["default"](theadClasses) || void 0,
    style: {
      visibility: virtualScroll ? "hidden" : "visible"
    }
  }, footerSummary && /* @__PURE__ */React__default["default"].createElement("tr", {
    className: classnames.tableFullRowClasses.base
  }, /* @__PURE__ */React__default["default"].createElement("td", {
    colSpan: columns.length
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: classnames.tableFullRowClasses.innerFullElement
  }, footerSummary))), footerDomList);
}

exports["default"] = TFoot;
