// tdesign-react/cjs/table/THead.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _typeof = require('../_chunks/dep-c4d1ce0c.js');
var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var table_hooks_useFixed = require('./hooks/useFixed.js');
var table_hooks_useClassName = require('./hooks/useClassName.js');
var table_hooks_useTableHeader = require('./hooks/useTableHeader.js');
var table_Ellipsis = require('./Ellipsis.js');
var table_utils = require('./utils.js');
var isFunction = require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-b289b0be.js');
require('../_chunks/dep-fb673f7a.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-1c586f5c.js');
require('../_chunks/dep-f0e884b9.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-794b42ce.js');
require('../_util/listener.js');
require('../_chunks/dep-68f5fe7b.js');
require('../hooks/useDeepEffect.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-0bed591e.js');
require('../hooks/usePrevious.js');
require('../_chunks/dep-e43e6728.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-4ef3269f.js');
require('../_chunks/dep-2a089253.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-048e4314.js');
require('../_chunks/dep-19400123.js');
require('../_chunks/dep-894a5ece.js');
require('../_chunks/dep-f7bf3d17.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-ca918872.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-13de2fef.js');
require('../_chunks/dep-63308b84.js');
require('../_chunks/dep-05df1ea3.js');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-2d5cf7d6.js');
require('./hooks/useMultiHeader.js');
require('../_util/scroll.js');
require('raf');
require('../_util/easing.js');
require('../tooltip/index.js');
require('../tooltip/Tooltip.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../hooks/useAnimation.js');
require('../hooks/useAttach.js');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../hooks/useDefaultProps.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');
require('../popup/utils/transition.js');
require('../popup/PopupPlugin.js');
require('../_chunks/dep-e37cb830.js');
require('../_util/react-render.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../tooltip/defaultProps.js');
require('../tooltip/TooltipLite.js');
require('../hooks/useSwitch.js');
require('../hooks/usePersistFn.js');
require('../_chunks/dep-2a85392e.js');
require('../hooks/useDebounce.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function THead(props) {
  var columnResizeParams = props.columnResizeParams,
    classPrefix = props.classPrefix,
    showColumnShadow = props.showColumnShadow;
  var theadRef = React.useRef(null);
  var classnames = table_hooks_useClassName["default"]();
  var tableHeaderClasses = classnames.tableHeaderClasses,
    tableBaseClass = classnames.tableBaseClass;
  var theadClasses = [tableHeaderClasses.header, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, tableHeaderClasses.fixed, Boolean(props.maxHeight || props.height)), tableBaseClass.bordered, props.bordered && props.isMultipleHeader), tableHeaderClasses.multipleHeader, props.isMultipleHeader)];
  var colspanSkipMap = React.useMemo(function () {
    var map = {};
    var list = props.thList[0];
    for (var i = 0, len = list.length; i < len; i++) {
      var item = list[i];
      if (item.colspan > 1) {
        for (var j = i + 1; j < i + item.colspan; j++) {
          if (list[j]) {
            map[list[j].colKey] = true;
          }
        }
      }
    }
    return map;
  }, [props.thList]);
  var getTableNode = function getTableNode(thead) {
    var parent = thead;
    while (parent) {
      var _parent;
      parent = parent.parentNode;
      if ((_parent = parent) !== null && _parent !== void 0 && (_parent = _parent.classList) !== null && _parent !== void 0 && _parent.contains("".concat(props.classPrefix, "-table"))) {
        break;
      }
    }
    return parent;
  };
  var renderThNodeList = function renderThNodeList(rowAndColFixedPosition, thWidthList) {
    var thBorderMap = /* @__PURE__ */new Map();
    var thRowspanAndColspan = props.spansAndLeafNodes.rowspanAndColspanMap;
    return props.thList.map(function (row, rowIndex) {
      var thRow = row.map(function (col, index) {
        var _col$children;
        if (colspanSkipMap[col.colKey]) return null;
        var rowspanAndColspan = thRowspanAndColspan.get(col);
        if (index === 0 && rowspanAndColspan.rowspan > 1) {
          for (var j = rowIndex + 1; j < rowIndex + rowspanAndColspan.rowspan; j++) {
            thBorderMap.set(props.thList[j][0], true);
          }
        }
        var thStyles = table_hooks_useFixed.getColumnFixedStyles(col, index, rowAndColFixedPosition, classnames.tableColFixedClasses);
        var colParams = {
          col: col,
          colIndex: index,
          row: {},
          rowIndex: -1
        };
        var isLeftFixedActive = showColumnShadow.left && col.fixed === "left";
        var isRightFixedActive = showColumnShadow.right && col.fixed === "right";
        var canDragSort = props.thDraggable && !(isLeftFixedActive || isRightFixedActive);
        var customClasses = table_utils.formatClassNames(col.className, _objectSpread(_objectSpread({}, colParams), {}, {
          type: "th"
        }));
        var thCustomClasses = table_utils.formatClassNames(col.thClassName, colParams);
        var thClasses = [thStyles.classes, customClasses, thCustomClasses, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, tableHeaderClasses.thBordered, thBorderMap.get(col)), "".concat(classPrefix, "-table__th-").concat(col.colKey), col.colKey), classnames.tdAlignClasses[col.align], col.align && col.align !== "left"), classnames.tableDraggableClasses.dragSortTh, canDragSort)];
        var withoutChildren = !((_col$children = col.children) !== null && _col$children !== void 0 && _col$children.length);
        var width = withoutChildren && thWidthList !== null && thWidthList !== void 0 && thWidthList[col.colKey] ? "".concat(thWidthList === null || thWidthList === void 0 ? void 0 : thWidthList[col.colKey], "px") : void 0;
        var styles = _objectSpread(_objectSpread({}, thStyles.style || {}), {}, {
          width: width
        });
        var innerTh = table_hooks_useTableHeader.renderTitle(col, index);
        if (!col.colKey) return null;
        var resizeColumnListener = props.resizable || !canDragSort ? {
          onMouseDown: function onMouseDown(e) {
            if (props.resizable) {
              var _columnResizeParams$o;
              columnResizeParams === null || columnResizeParams === void 0 || (_columnResizeParams$o = columnResizeParams.onColumnMousedown) === null || _columnResizeParams$o === void 0 || _columnResizeParams$o.call(columnResizeParams, e, col, index);
            }
            if (!canDragSort) {
              var timer = setTimeout(function () {
                var _thList$index;
                var thList = theadRef.current.querySelectorAll("th");
                (_thList$index = thList[index]) === null || _thList$index === void 0 || _thList$index.removeAttribute("draggable");
                clearTimeout(timer);
              }, 10);
            }
          },
          onMouseMove: function onMouseMove(e) {
            var _columnResizeParams$o2;
            props.resizable && (columnResizeParams === null || columnResizeParams === void 0 || (_columnResizeParams$o2 = columnResizeParams.onColumnMouseover) === null || _columnResizeParams$o2 === void 0 ? void 0 : _columnResizeParams$o2.call(columnResizeParams, e, col));
          }
        } : {};
        var content = isFunction.isFunction(col.ellipsisTitle) ? col.ellipsisTitle({
          col: col,
          colIndex: index
        }) : void 0;
        var isEllipsis = col.ellipsisTitle !== void 0 ? Boolean(col.ellipsisTitle) : Boolean(col.ellipsis);
        var attrs = (isFunction.isFunction(col.attrs) ? col.attrs(_objectSpread(_objectSpread({}, colParams), {}, {
          type: "th"
        })) : col.attrs) || {};
        if (col.colspan > 1) {
          attrs.colSpan = col.colspan;
        }
        return /* @__PURE__ */React__default["default"].createElement("th", _objectSpread(_objectSpread(_objectSpread({
          key: col.colKey,
          "data-colkey": col.colKey,
          className: classNames__default["default"](thClasses),
          style: styles
        }, {
          rowSpan: rowspanAndColspan.rowspan,
          colSpan: rowspanAndColspan.colspan
        }), attrs), resizeColumnListener), /* @__PURE__ */React__default["default"].createElement("div", {
          className: tableBaseClass.thCellInner
        }, isEllipsis ? /* @__PURE__ */React__default["default"].createElement(table_Ellipsis["default"], {
          placement: "bottom",
          attach: theadRef.current ? function () {
            return getTableNode(theadRef.current);
          } : void 0,
          popupContent: content,
          tooltipProps: _typeof._typeof(col.ellipsisTitle) === "object" ? col.ellipsisTitle : void 0,
          overlayClassName: props.ellipsisOverlayClassName,
          classPrefix: props.classPrefix
        }, innerTh) : innerTh));
      });
      return /* @__PURE__ */React__default["default"].createElement("tr", {
        key: rowIndex
      }, thRow);
    });
  };
  return /* @__PURE__ */React__default["default"].createElement("thead", {
    ref: theadRef,
    className: classNames__default["default"](theadClasses)
  }, renderThNodeList(props.rowAndColFixedPosition, props.thWidthList));
}

exports["default"] = THead;
