// tdesign-react/cjs/tag/CheckTag.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var common = require('../_chunks/dep-dd0f5f6b.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var _util_noop = require('../_util/noop.js');
var tag_defaultProps = require('./defaultProps.js');
var tag_Tag = require('./Tag.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-ecfe656e.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('tdesign-icons-react');
require('../_chunks/dep-d14bfcfa.js');
require('../hooks/useGlobalIcon.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["value", "content", "onClick", "disabled", "children", "size", "checkedProps", "uncheckedProps", "onChange", "className"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var CheckTag = /*#__PURE__*/React.forwardRef(function (originalProps, ref) {
  var props = hooks_useDefaultProps["default"](originalProps, tag_defaultProps.checkTagDefaultProps);
  var value = props.value,
    content = props.content,
    _props$onClick = props.onClick,
    onClick = _props$onClick === void 0 ? _util_noop["default"] : _props$onClick,
    disabled = props.disabled,
    children = props.children,
    size = props.size,
    checkedProps = props.checkedProps,
    uncheckedProps = props.uncheckedProps,
    onChange = props.onChange,
    className = props.className,
    tagOtherProps = objectWithoutProperties._objectWithoutProperties(props, _excluded);
  var _useControlled = hooks_useControlled["default"](props, "checked", onChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    innerChecked = _useControlled2[0],
    setInnerChecked = _useControlled2[1];
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var tagClassPrefix = "".concat(classPrefix, "-tag");
  var tagClass = React.useMemo(function () {
    return ["".concat(tagClassPrefix), "".concat(tagClassPrefix, "--check"), defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(tagClassPrefix, "--checked"), innerChecked), "".concat(tagClassPrefix, "--disabled"), disabled), "".concat(classPrefix, "-size-s"), size === "small"), "".concat(classPrefix, "-size-l"), size === "large"), className];
  }, [innerChecked, disabled, classPrefix, tagClassPrefix, size, className]);
  var checkTagProps = React.useMemo(function () {
    var tmpCheckedProps = _objectSpread({
      theme: "primary"
    }, checkedProps);
    var tmpUncheckedProps = _objectSpread({}, uncheckedProps);
    return innerChecked ? tmpCheckedProps : tmpUncheckedProps;
  }, [innerChecked, checkedProps, uncheckedProps]);
  var handleClick = function handleClick(_ref2) {
    var e = _ref2.e;
    if (!disabled) {
      onClick === null || onClick === void 0 || onClick({
        e: e
      });
      setInnerChecked(!innerChecked, {
        e: e,
        value: value
      });
    }
  };
  var keyboardEventListener = function keyboardEventListener(e) {
    var _e$key;
    var code = e.code || ((_e$key = e.key) === null || _e$key === void 0 ? void 0 : _e$key.trim());
    var isCheckedCode = common.SPACE_REG.test(code) || common.ENTER_REG.test(code);
    if (isCheckedCode) {
      e.preventDefault();
      setInnerChecked(!innerChecked, {
        e: e,
        value: value
      });
    }
  };
  var onCheckboxFocus = function onCheckboxFocus(e) {
    e.currentTarget.addEventListener("keydown", keyboardEventListener);
  };
  var onCheckboxBlur = function onCheckboxBlur(e) {
    e.currentTarget.removeEventListener("keydown", keyboardEventListener);
  };
  return /* @__PURE__ */React__default["default"].createElement(tag_Tag.Tag, _objectSpread(_objectSpread({
    ref: ref,
    className: classNames__default["default"](tagClass),
    disabled: props.disabled,
    tabIndex: props.disabled ? void 0 : 0,
    onFocus: onCheckboxFocus,
    onBlur: onCheckboxBlur
  }, checkTagProps), {}, {
    onClick: handleClick
  }, tagOtherProps), content || children);
});
CheckTag.displayName = "CheckTag";

exports["default"] = CheckTag;
