// tdesign-react/cjs/tag/CheckTagGroup.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var hooks_useControlled = require('../hooks/useControlled.js');
var tag_defaultProps = require('./defaultProps.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var tag_CheckTag = require('./CheckTag.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-ecfe656e.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-2b5ba78d.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-1c586f5c.js');
require('classnames');
require('../_chunks/dep-dd0f5f6b.js');
require('./Tag.js');
require('tdesign-icons-react');
require('../_chunks/dep-d14bfcfa.js');
require('../hooks/useGlobalIcon.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var CheckTagGroup = function CheckTagGroup(originalProps) {
  var props = hooks_useDefaultProps["default"](originalProps, tag_defaultProps.checkTagGroupDefaultProps);
  var options = props.options,
    onChange = props.onChange;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var componentName = "".concat(classPrefix, "-check-tag-group");
  var _useControlled = hooks_useControlled["default"](props, "value", onChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    innerValue = _useControlled2[0],
    setInnerValue = _useControlled2[1];
  var onCheckTagChange = function onCheckTagChange(checked, ctx) {
    var value = ctx.value;
    if (checked) {
      if (props.multiple) {
        setInnerValue(innerValue.concat(value), {
          e: ctx.e,
          type: "check",
          value: value
        });
      } else {
        setInnerValue([value], {
          e: ctx.e,
          type: "check",
          value: value
        });
      }
    } else {
      var newValue = [];
      if (props.multiple) {
        newValue = innerValue.filter(function (t) {
          return t !== value;
        });
      }
      setInnerValue(newValue, {
        e: ctx.e,
        type: "uncheck",
        value: value
      });
    }
  };
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: componentName
  }, options === null || options === void 0 ? void 0 : options.map(function (option) {
    var _ref, _option$content;
    return /* @__PURE__ */React__default["default"].createElement(tag_CheckTag["default"], {
      key: option.value,
      value: option.value,
      "data-value": option.value,
      checkedProps: props.checkedProps,
      uncheckedProps: props.uncheckedProps,
      checked: innerValue.includes(option.value),
      onChange: onCheckTagChange,
      disabled: option.disabled,
      size: option.size
    }, (_ref = (_option$content = option.content) !== null && _option$content !== void 0 ? _option$content : option.children) !== null && _ref !== void 0 ? _ref : option.label);
  }));
};
CheckTagGroup.displayName = "CheckTagGroup";

exports["default"] = CheckTagGroup;
