// tdesign-react/cjs/tag-input/TagInput.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var classNames = require('classnames');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var hooks_useDragSorter = require('../hooks/useDragSorter.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var input_index = require('../input/index.js');
var tagInput_defaultProps = require('./defaultProps.js');
var tagInput_useHover = require('./useHover.js');
var tagInput_useTagList = require('./useTagList.js');
var tagInput_useTagScroll = require('./useTagScroll.js');
var isFunction = require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../hooks/useEventCallback.js');
require('../input/Input.js');
require('../_chunks/dep-1c586f5c.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../_util/parseTNode.js');
require('../_chunks/dep-b289b0be.js');
require('../hooks/useLayoutEffect.js');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../locale/LocalReceiver.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../input/defaultProps.js');
require('../input/InputGroup.js');
require('../input/useLengthLimit.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-f0e884b9.js');
require('../tag/index.js');
require('../tag/Tag.js');
require('../_chunks/dep-d14bfcfa.js');
require('../tag/defaultProps.js');
require('../tag/CheckTag.js');
require('../_chunks/dep-dd0f5f6b.js');
require('../tag/CheckTagGroup.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var TagInput = /*#__PURE__*/React.forwardRef(function (originalProps, ref) {
  var props = hooks_useDefaultProps["default"](originalProps, tagInput_defaultProps.tagInputDefaultProps);
  var _useConfig = hooks_useConfig["default"](),
    prefix = _useConfig.classPrefix;
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      CloseCircleFilledIcon: tdesignIconsReact.CloseCircleFilledIcon
    }),
    CloseCircleFilledIcon = _useGlobalIcon.CloseCircleFilledIcon;
  var excessTagsDisplayType = props.excessTagsDisplayType,
    autoWidth = props.autoWidth,
    borderless = props.borderless,
    disabled = props.disabled,
    clearable = props.clearable,
    placeholder = props.placeholder,
    valueDisplay = props.valueDisplay,
    label = props.label,
    inputProps = props.inputProps,
    size = props.size,
    tips = props.tips,
    status = props.status,
    suffixIcon = props.suffixIcon,
    suffix = props.suffix,
    prefixIcon = props.prefixIcon,
    maxRows = props.maxRows,
    onClick = props.onClick,
    onPaste = props.onPaste,
    _onFocus = props.onFocus,
    _onBlur = props.onBlur;
  var readOnly = props.readOnly || props.readonly;
  var isBreakLine = excessTagsDisplayType === "break-line";
  var _useControlled = hooks_useControlled["default"](props, "inputValue", props.onInputChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    tInputValue = _useControlled2[0],
    setTInputValue = _useControlled2[1];
  var _useHover = tagInput_useHover["default"](props),
    isHover = _useHover.isHover,
    addHover = _useHover.addHover,
    cancelHover = _useHover.cancelHover;
  var _useDragSorter = hooks_useDragSorter["default"](_objectSpread(_objectSpread({}, props), {}, {
      sortOnDraggable: props.dragSort,
      onDragOverCheck: {
        x: true,
        targetClassNameRegExp: new RegExp("^".concat(prefix, "-tag"))
      }
    })),
    getDragProps = _useDragSorter.getDragProps;
  var suffixWidthRef = React.useRef(0);
  var suffixIconWidthRef = React.useRef(0);
  var isCompositionRef = React.useRef(false);
  var _useTagScroll = tagInput_useTagScroll["default"](props),
    scrollToRight = _useTagScroll.scrollToRight,
    onWheel = _useTagScroll.onWheel,
    scrollToRightOnEnter = _useTagScroll.scrollToRightOnEnter,
    scrollToLeftOnLeave = _useTagScroll.scrollToLeftOnLeave,
    tagInputRef = _useTagScroll.tagInputRef;
  var _useTagList = tagInput_useTagList["default"](_objectSpread(_objectSpread({}, props), {}, {
      getDragProps: getDragProps
    })),
    tagValue = _useTagList.tagValue,
    _onClose = _useTagList.onClose,
    onInnerEnter = _useTagList.onInnerEnter,
    onInputBackspaceKeyUp = _useTagList.onInputBackspaceKeyUp,
    clearAll = _useTagList.clearAll,
    renderLabel = _useTagList.renderLabel,
    onInputBackspaceKeyDown = _useTagList.onInputBackspaceKeyDown;
  var NAME_CLASS = "".concat(prefix, "-tag-input");
  var WITH_SUFFIX_ICON_CLASS = "".concat(prefix, "-tag-input__with-suffix-icon");
  var CLEAR_CLASS = "".concat(prefix, "-tag-input__suffix-clear");
  var BREAK_LINE_CLASS = "".concat(prefix, "-tag-input--break-line");
  var tagInputPlaceholder = !(tagValue !== null && tagValue !== void 0 && tagValue.length) ? placeholder : "";
  var showClearIcon = Boolean(!readOnly && !disabled && clearable && isHover && (tagValue === null || tagValue === void 0 ? void 0 : tagValue.length));
  React.useImperativeHandle(ref, function () {
    return _objectSpread({}, tagInputRef.current || {});
  });
  var updateSuffixWidth = function updateSuffixWidth(selector, cssVar, widthRef) {
    var _tagInputRef$current;
    var wrapperEl = (_tagInputRef$current = tagInputRef.current) === null || _tagInputRef$current === void 0 ? void 0 : _tagInputRef$current.currentElement;
    if (!wrapperEl) return;
    var inputEl = wrapperEl.querySelector(".".concat(prefix, "-input"));
    if (!inputEl) return;
    var targetEl = wrapperEl.querySelector(selector);
    var width = targetEl ? targetEl.getBoundingClientRect().width : 0;
    if (width !== widthRef.current) {
      widthRef.current = width;
      if (width) {
        inputEl.style.setProperty(cssVar, "".concat(Math.ceil(width + 8), "px"));
      } else {
        inputEl.style.removeProperty(cssVar);
      }
    }
  };
  React.useEffect(function () {
    if (!isBreakLine || !suffix) return;
    updateSuffixWidth(".".concat(prefix, "-input__suffix:not(.").concat(prefix, "-input__suffix-icon)"), "--".concat(prefix, "-tag-input-suffix-width"), suffixWidthRef);
    updateSuffixWidth(".".concat(prefix, "-input__suffix-icon"), "--".concat(prefix, "-tag-input-suffix-icon-width"), suffixIconWidthRef);
  }, [excessTagsDisplayType, suffix, suffixIcon, showClearIcon, prefix, tagInputRef, isBreakLine]);
  var onInputCompositionstart = function onInputCompositionstart(value, context) {
    var _inputProps$onComposi;
    isCompositionRef.current = true;
    inputProps === null || inputProps === void 0 || (_inputProps$onComposi = inputProps.onCompositionstart) === null || _inputProps$onComposi === void 0 || _inputProps$onComposi.call(inputProps, value, context);
  };
  var onInputCompositionend = function onInputCompositionend(value, context) {
    var _inputProps$onComposi2;
    isCompositionRef.current = false;
    inputProps === null || inputProps === void 0 || (_inputProps$onComposi2 = inputProps.onCompositionend) === null || _inputProps$onComposi2 === void 0 || _inputProps$onComposi2.call(inputProps, value, context);
  };
  var onInputEnter = function onInputEnter(value, context) {
    setTInputValue("", {
      e: context.e,
      trigger: "enter"
    });
    !isCompositionRef.current && onInnerEnter(value, context);
    scrollToRight();
  };
  var onInnerClick = function onInnerClick(context) {
    if (!props.disabled && !readOnly) {
      var _tagInputRef$current2, _tagInputRef$current3;
      (_tagInputRef$current2 = tagInputRef.current) === null || _tagInputRef$current2 === void 0 || (_tagInputRef$current2 = _tagInputRef$current2.inputElement) === null || _tagInputRef$current2 === void 0 || (_tagInputRef$current3 = _tagInputRef$current2.focus) === null || _tagInputRef$current3 === void 0 || _tagInputRef$current3.call(_tagInputRef$current2);
    }
    onClick === null || onClick === void 0 || onClick(context);
  };
  var onClearClick = function onClearClick(e) {
    var _props$onClear;
    clearAll({
      e: e
    });
    setTInputValue("", {
      e: e,
      trigger: "clear"
    });
    (_props$onClear = props.onClear) === null || _props$onClear === void 0 || _props$onClear.call(props, {
      e: e
    });
  };
  var onKeydown = function onKeydown(value, context) {
    var _inputProps$onKeydown;
    onInputBackspaceKeyDown(value, context);
    inputProps === null || inputProps === void 0 || (_inputProps$onKeydown = inputProps.onKeydown) === null || _inputProps$onKeydown === void 0 || _inputProps$onKeydown.call(inputProps, value, context);
  };
  var onKeyup = function onKeyup(value, context) {
    var _inputProps$onKeyup;
    onInputBackspaceKeyUp(value);
    inputProps === null || inputProps === void 0 || (_inputProps$onKeyup = inputProps.onKeyup) === null || _inputProps$onKeyup === void 0 || _inputProps$onKeyup.call(inputProps, value, context);
  };
  var suffixIconNode = showClearIcon ? /* @__PURE__ */React__default["default"].createElement(CloseCircleFilledIcon, {
    className: CLEAR_CLASS,
    onClick: onClearClick
  }) : suffixIcon;
  var displayNode = React.useMemo(function () {
    return isFunction.isFunction(valueDisplay) ? valueDisplay({
      value: tagValue,
      onClose: function onClose(index) {
        return _onClose({
          index: index
        });
      }
    }) : valueDisplay;
  }, [valueDisplay]);
  var isEmpty = !(Array.isArray(tagValue) && tagValue.length);
  var classes = [NAME_CLASS, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, BREAK_LINE_CLASS, isBreakLine), WITH_SUFFIX_ICON_CLASS, !!suffixIconNode), "".concat(prefix, "-is-empty"), isEmpty), "".concat(prefix, "-tag-input--with-tag"), !isEmpty), "".concat(prefix, "-tag-input--max-rows"), isBreakLine && maxRows), "".concat(prefix, "-tag-input--drag-sort"), props.dragSort && !disabled && !readOnly), props.className];
  var maxRowsStyle = maxRows ? {
    "--max-rows": maxRows,
    "--tag-input-size": size
  } : {};
  return /* @__PURE__ */React__default["default"].createElement(input_index.Input, _objectSpread(_objectSpread({
    ref: tagInputRef,
    value: tInputValue,
    onChange: function onChange(val, context) {
      setTInputValue(val, _objectSpread(_objectSpread({}, context), {}, {
        trigger: "input"
      }));
    },
    autoWidth: true,
    onWheel: onWheel,
    size: size,
    borderless: borderless,
    readOnly: readOnly,
    disabled: disabled,
    label: renderLabel({
      displayNode: displayNode,
      label: label
    }),
    style: _objectSpread(_objectSpread({}, props.style), maxRowsStyle),
    tips: tips,
    status: status,
    placeholder: tagInputPlaceholder,
    suffix: suffix,
    prefixIcon: prefixIcon,
    suffixIcon: suffixIconNode,
    showInput: !(inputProps !== null && inputProps !== void 0 && inputProps.readOnly) || !(inputProps !== null && inputProps !== void 0 && inputProps.readonly) || !tagValue || !(tagValue !== null && tagValue !== void 0 && tagValue.length),
    keepWrapperWidth: !autoWidth,
    onPaste: onPaste,
    onClick: onInnerClick,
    onEnter: onInputEnter,
    onMouseenter: function onMouseenter(context) {
      addHover(context);
      scrollToRightOnEnter();
    },
    onMouseleave: function onMouseleave(context) {
      cancelHover(context);
      scrollToLeftOnLeave();
    },
    onFocus: function onFocus(inputValue, context) {
      _onFocus === null || _onFocus === void 0 || _onFocus(tagValue, {
        e: context.e,
        inputValue: inputValue
      });
    },
    onBlur: function onBlur(inputValue, context) {
      if (tInputValue) {
        setTInputValue("", {
          e: context.e,
          trigger: "blur"
        });
      }
      _onBlur === null || _onBlur === void 0 || _onBlur(tagValue, {
        e: context.e,
        inputValue: inputValue
      });
    }
  }, inputProps), {}, {
    onCompositionstart: onInputCompositionstart,
    onCompositionend: onInputCompositionend,
    onKeydown: onKeydown,
    onKeyup: onKeyup,
    className: classNames__default["default"](classes, inputProps === null || inputProps === void 0 ? void 0 : inputProps.className)
  }));
});
TagInput.displayName = "TagInput";

exports["default"] = TagInput;
