// tdesign-react/cjs/pagination/hooks/useBoundaryJumper.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var hooks_useConfig = require('../../hooks/useConfig.js');
var hooks_useGlobalIcon = require('../../hooks/useGlobalIcon.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-14013769.js');
require('../../_chunks/dep-06bf4625.js');
require('dayjs');
require('../../_chunks/dep-043eee20.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../_chunks/dep-2d5cf7d6.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function useBoundaryJumper(props) {
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      PageLastIcon: tdesignIconsReact.PageLastIcon,
      PageFirstIcon: tdesignIconsReact.PageFirstIcon
    }),
    PageFirstIcon = _useGlobalIcon.PageFirstIcon,
    PageLastIcon = _useGlobalIcon.PageLastIcon;
  var name = "".concat(classPrefix, "-pagination");
  var showFirstAndLastPageBtn = props.showFirstAndLastPageBtn,
    disabled = props.disabled,
    current = props.current,
    pageCount = props.pageCount,
    changeCurrent = props.changeCurrent;
  var firstPageJumper = showFirstAndLastPageBtn && /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(name, "__btn"), "".concat(name, "__btn-first-page"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-disabled"), disabled || current === 1)),
    onClick: function onClick() {
      return changeCurrent(1);
    }
  }, /* @__PURE__ */React__default["default"].createElement(PageFirstIcon, null));
  var lastPageJumper = showFirstAndLastPageBtn && /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(name, "__btn"), "".concat(name, "__btn-first-page"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-disabled"), disabled || current === pageCount)),
    onClick: function onClick() {
      return changeCurrent(pageCount);
    }
  }, /* @__PURE__ */React__default["default"].createElement(PageLastIcon, null));
  return {
    firstPageJumper: firstPageJumper,
    lastPageJumper: lastPageJumper
  };
}

exports["default"] = useBoundaryJumper;
