// tdesign-react/cjs/common/Check.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var _util_helper = require('../_util/helper.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var isBoolean = require('../_chunks/dep-089e8106.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-e44b03ec.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_util/noop.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["allowUncheck", "type", "disabled", "name", "value", "onChange", "indeterminate", "children", "label", "className", "style", "onClick"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var CheckContext = /*#__PURE__*/React__default["default"].createContext(null);
var Check = /*#__PURE__*/React.forwardRef(function (_props, ref) {
  var context = React.useContext(CheckContext);
  var props = context ? context.inject(_props) : _props;
  var _props$allowUncheck = props.allowUncheck,
    allowUncheck = _props$allowUncheck === void 0 ? false : _props$allowUncheck,
    type = props.type,
    disabled = props.disabled,
    name = props.name,
    value = props.value,
    onChange = props.onChange,
    indeterminate = props.indeterminate,
    children = props.children,
    label = props.label,
    className = props.className,
    style = props.style,
    onClick = props.onClick,
    htmlProps = objectWithoutProperties._objectWithoutProperties(props, _excluded);
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var TOnChange = onChange;
  var _useControlled = hooks_useControlled["default"](props, "checked", TOnChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    internalChecked = _useControlled2[0],
    setInternalChecked = _useControlled2[1];
  var labelClassName = classNames__default["default"]("".concat(classPrefix, "-").concat(type), className, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-is-checked"), internalChecked), "".concat(classPrefix, "-is-disabled"), disabled), "".concat(classPrefix, "-is-indeterminate"), indeterminate));
  var readOnly = props.readOnly || props.readonly;
  var isDisabled = disabled || readOnly;
  var input = /* @__PURE__ */React__default["default"].createElement("input", {
    readOnly: readOnly,
    type: type === "radio-button" ? "radio" : type,
    className: "".concat(classPrefix, "-").concat(type, "__former"),
    checked: internalChecked,
    disabled: disabled,
    name: name,
    tabIndex: -1,
    value: isBoolean.isBoolean(value) ? Number(value) : value,
    "data-value": typeof value === "string" ? "'".concat(value, "'") : value,
    "data-allow-uncheck": allowUncheck || void 0,
    onClick: function onClick(e) {
      e.stopPropagation();
      if ((type === "radio-button" || type === "radio") && allowUncheck && internalChecked) {
        setInternalChecked(!e.currentTarget.checked, {
          e: e
        });
      }
    },
    onChange: function onChange(e) {
      if (isDisabled) return;
      setInternalChecked(e.currentTarget.checked, {
        e: e
      });
    }
  });
  var showLabel = !!(children || label);
  var handleLabelClick = function handleLabelClick(event) {
    if (props.stopLabelTrigger) {
      event.preventDefault();
    }
  };
  var onInnerClick = function onInnerClick(e) {
    if (isDisabled) return;
    onClick === null || onClick === void 0 || onClick({
      e: e
    });
  };
  return /* @__PURE__ */React__default["default"].createElement("label", _objectSpread(_objectSpread({
    ref: ref,
    tabIndex: disabled ? void 0 : 0,
    className: labelClassName,
    title: props.title,
    style: style
  }, _util_helper.omit(htmlProps, ["checkAll", "stopLabelTrigger", "readonly"])), {}, {
    onClick: onInnerClick
  }), input, /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-").concat(type, "__input")
  }), showLabel && /* @__PURE__ */React__default["default"].createElement("span", {
    key: "label",
    className: "".concat(classPrefix, "-").concat(type, "__label"),
    onClick: handleLabelClick
  }, children || label));
});
Check.displayName = "Check";

exports.CheckContext = CheckContext;
exports["default"] = Check;
