// tdesign-react/es/date-picker/DateRangePicker.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-5e72c8b9.js';
import { _ as _toConsumableArray } from '../_chunks/dep-1e9f924f.js';
import { _ as _slicedToArray } from '../_chunks/dep-94a7af64.js';
import React, { forwardRef, useState, useEffect } from 'react';
import classNames from 'classnames';
import dayjs from 'dayjs';
import { g as getDefaultFormat, p as parseToDayjs, i as isValidDate, f as formatDate, a as formatTime, b as initYearMonthTime } from '../_chunks/dep-07ae3139.js';
import { g as getRangeBounds, s as subtractMonth, a as addMonth, e as extractTimeObj } from '../_chunks/dep-7ff1070c.js';
import '../_chunks/dep-3561cd86.js';
import { isArray, isFunction } from 'lodash-es';
import useConfig from '../hooks/useConfig.js';
import useDefaultProps from '../hooks/useDefaultProps.js';
import useLatest from '../hooks/useLatest.js';
import useUpdateEffect from '../hooks/useUpdateEffect.js';
import { RangeInputPopup } from '../range-input/index.js';
import { dateRangePickerDefaultProps } from './defaultProps.js';
import useRange from './hooks/useRange.js';
import RangePanel from './panel/RangePanel.js';
import { dateCorrection, triggerMap } from './utils.js';
import { l as log } from '../_chunks/dep-65d82dda.js';
import '../_chunks/dep-43f7da7b.js';
import '../_chunks/dep-8e8ac0d4.js';
import '../_chunks/dep-57e67945.js';
import '../_chunks/dep-0054b574.js';
import '../_chunks/dep-7d48ddac.js';
import '../_chunks/dep-9b727dd4.js';
import '../_chunks/dep-35769bc6.js';
import '../config-provider/ConfigContext.js';
import '../locale/zh_CN.js';
import '../_chunks/dep-766c4603.js';
import '../hooks/useIsFirstRender.js';
import '../range-input/RangeInput.js';
import '../_chunks/dep-4461a684.js';
import 'tdesign-icons-react';
import '../input/index.js';
import '../input/Input.js';
import '../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../_util/parseTNode.js';
import '../hooks/useControlled.js';
import '../_util/noop.js';
import '../hooks/useGlobalIcon.js';
import '../hooks/useLayoutEffect.js';
import '../_chunks/dep-a347b1b5.js';
import '../locale/LocalReceiver.js';
import '../config-provider/index.js';
import '../config-provider/ConfigProvider.js';
import '../config-provider/type.js';
import '../input/defaultProps.js';
import '../input/InputGroup.js';
import '../input/useLengthLimit.js';
import '../_chunks/dep-9721fff3.js';
import '../input/style/css.js';
import '../input/type.js';
import '../range-input/defaultProps.js';
import '../range-input/RangeInputPopup.js';
import '../popup/index.js';
import '../popup/Popup.js';
import 'react-transition-group';
import '../common/Portal.js';
import 'react-dom';
import '../hooks/useAnimation.js';
import '../hooks/useAttach.js';
import '../hooks/useMutationObserver.js';
import '../hooks/usePopper.js';
import 'react-fast-compare';
import '@popperjs/core';
import '../hooks/useWindowSize.js';
import '../popup/defaultProps.js';
import '../popup/hooks/useTrigger.js';
import '../_util/listener.js';
import '../_util/ref.js';
import 'react-is';
import '../_util/isFragment.js';
import '../popup/utils/transition.js';
import '../popup/PopupPlugin.js';
import '../_chunks/dep-fe79dc5d.js';
import '../_util/react-render.js';
import '../common/PluginContainer.js';
import '../popup/style/css.js';
import '../popup/type.js';
import '../select-input/useOverlayInnerStyle.js';
import '../hooks/useInnerPopupVisible.js';
import '../range-input/style/css.js';
import '../range-input/type.js';
import './hooks/useRangeValue.js';
import './panel/PanelContent.js';
import './base/Header.js';
import '../pagination/index.js';
import '../pagination/Pagination.js';
import '../select/index.js';
import '../select/base/Select.js';
import '../_util/helper.js';
import '../common/FakeArrow.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../loading/defaultProps.js';
import '../loading/gradient.js';
import '../_chunks/dep-2481dfad.js';
import '../hooks/useDomRefCallback.js';
import '../loading/plugin.js';
import '../loading/style/css.js';
import '../loading/type.js';
import '../select-input/index.js';
import '../select-input/SelectInput.js';
import '../select-input/defaultProps.js';
import '../select-input/useMultiple.js';
import '../tag-input/index.js';
import '../tag-input/TagInput.js';
import '../hooks/useDragSorter.js';
import '../hooks/useEventCallback.js';
import '../tag-input/defaultProps.js';
import '../tag-input/useHover.js';
import '../tag-input/useTagList.js';
import '../tag/index.js';
import '../tag/Tag.js';
import '../_chunks/dep-dfdbe6bd.js';
import '../tag/defaultProps.js';
import '../tag/CheckTag.js';
import '../_chunks/dep-29a70a9d.js';
import '../tag/CheckTagGroup.js';
import '../tag/style/css.js';
import '../tag/type.js';
import '../tag-input/useTagScroll.js';
import '../tag-input/style/css.js';
import '../tag-input/type.js';
import '../select-input/useSingle.js';
import '../select-input/style/css.js';
import '../select-input/type.js';
import '../select/defaultProps.js';
import '../select/hooks/useKeyboardControl.js';
import '../_chunks/dep-29be6164.js';
import '../hooks/useRipple.js';
import '../select/base/OptionGroup.js';
import '../select/base/PopupContent.js';
import '../select/hooks/usePanelVirtualScroll.js';
import '../hooks/useVirtualScroll.js';
import '../select/style/css.js';
import '../select/type.js';
import '../input-number/index.js';
import '../input-number/InputNumber.js';
import '../button/index.js';
import '../button/Button.js';
import '../button/defaultProps.js';
import '../button/style/css.js';
import '../button/type.js';
import '../input-number/useInputNumber.js';
import '../_chunks/dep-35a55ece.js';
import '../hooks/useCommonClassName.js';
import '../input-number/defaultProps.js';
import '../input-number/style/css.js';
import '../input-number/type.js';
import '../input-adornment/index.js';
import '../input-adornment/InputAdornment.js';
import '../input-adornment/style/css.js';
import '../input-adornment/type.js';
import '../pagination/hooks/useBoundaryJumper.js';
import '../pagination/hooks/usePrevNextJumper.js';
import '../pagination/hooks/usePageNumber.js';
import '../pagination/hooks/useTotal.js';
import '../pagination/validators.js';
import '../pagination/defaultProps.js';
import '../pagination/PaginationMini.js';
import '../pagination/style/css.js';
import '../pagination/type.js';
import './hooks/useSelectRange.js';
import './base/Table.js';
import './base/Cell.js';
import '../time-picker/index.js';
import '../time-picker/TimePicker.js';
import '../_chunks/dep-35931e80.js';
import '../time-picker/hooks/useTimePickerTextConfig.js';
import '../hooks/useDebounce.js';
import '../hooks/usePersistFn.js';
import '../hooks/usePropsRef.js';
import '../time-picker/TimeRangePicker.js';
import '../time-picker/panel/TimePickerPanel.js';
import '../time-picker/defaultProps.js';
import '../time-picker/style/css.js';
import '../time-picker/type.js';
import './panel/ExtraContent.js';
import './base/Footer.js';
import './hooks/useTableData.js';
import './hooks/useDisableDate.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var DateRangePicker = /*#__PURE__*/forwardRef(function (originalProps, ref) {
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix,
    globalDatePickerConfig = _useConfig.datePicker;
  var props = useDefaultProps(originalProps, dateRangePickerDefaultProps);
  var className = props.className,
    style = props.style,
    disabled = props.disabled,
    mode = props.mode,
    enableTimePicker = props.enableTimePicker,
    disableDate = props.disableDate,
    _props$firstDayOfWeek = props.firstDayOfWeek,
    firstDayOfWeek = _props$firstDayOfWeek === void 0 ? globalDatePickerConfig.firstDayOfWeek : _props$firstDayOfWeek,
    presets = props.presets,
    timePickerProps = props.timePickerProps,
    presetsPlacement = props.presetsPlacement,
    panelPreselection = props.panelPreselection,
    cancelRangeSelectLimit = props.cancelRangeSelectLimit,
    range = props.range,
    panelActiveDate = props.panelActiveDate,
    cell = props.cell,
    onPick = props.onPick,
    disableTime = props.disableTime,
    needConfirm = props.needConfirm;
  var _useRange = useRange(props),
    inputValue = _useRange.inputValue,
    popupVisible = _useRange.popupVisible,
    rangeInputProps = _useRange.rangeInputProps,
    popupProps = _useRange.popupProps,
    value = _useRange.value,
    year = _useRange.year,
    month = _useRange.month,
    time = _useRange.time,
    activeIndex = _useRange.activeIndex,
    isHoverCell = _useRange.isHoverCell,
    setActiveIndex = _useRange.setActiveIndex,
    onChange = _useRange.onChange,
    setIsHoverCell = _useRange.setIsHoverCell,
    setInputValue = _useRange.setInputValue,
    setPopupVisible = _useRange.setPopupVisible,
    setTime = _useRange.setTime,
    setYear = _useRange.setYear,
    setMonth = _useRange.setMonth,
    isFirstValueSelected = _useRange.isFirstValueSelected,
    setIsFirstValueSelected = _useRange.setIsFirstValueSelected,
    cacheValue = _useRange.cacheValue,
    setCacheValue = _useRange.setCacheValue,
    isSwitchTimeMode = _useRange.isSwitchTimeMode;
  var _getDefaultFormat = getDefaultFormat({
      mode: mode,
      enableTimePicker: enableTimePicker,
      format: props.format,
      valueType: props.valueType
    }),
    format = _getDefaultFormat.format,
    timeFormat = _getDefaultFormat.timeFormat,
    valueType = _getDefaultFormat.valueType;
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    isSelected = _useState2[0],
    setIsSelected = _useState2[1];
  var handleSyncPanelValue = function handleSyncPanelValue(value2) {
    var nextMonth = value2.map(function (v) {
      return parseToDayjs(v, format).month();
    });
    var nextYear = value2.map(function (v) {
      return parseToDayjs(v, format).year();
    });
    if (nextYear[0] === nextYear[1] && nextMonth[0] === nextMonth[1]) {
      nextMonth[0] === 11 ? nextMonth[0] -= 1 : nextMonth[1] += 1;
    }
    if (["month", "quarter"].includes(props.mode) && nextYear[0] === nextYear[1]) {
      nextYear = [nextYear[0], nextYear[0] + 1];
    }
    setMonth(nextMonth);
    setYear(nextYear);
  };
  var handlePopupInvisible = function handlePopupInvisible() {
    var _props$popupProps, _props$popupProps$onV;
    setPopupVisible(false);
    (_props$popupProps = props.popupProps) === null || _props$popupProps === void 0 || (_props$popupProps$onV = _props$popupProps.onVisibleChange) === null || _props$popupProps$onV === void 0 || _props$popupProps$onV.call(_props$popupProps, false, {});
  };
  var onTriggerNeedConfirm = useLatest(function () {
    if (needConfirm || !enableTimePicker || popupVisible) return;
    var nextValue = _toConsumableArray(inputValue);
    var notValidIndex = nextValue.findIndex(function (v) {
      return !v || !isValidDate(v, format);
    });
    if (notValidIndex === -1 && nextValue.length === 2) {
      var currentValue = formatDate(value || [], {
        format: format
      });
      if (currentValue[0] !== nextValue[0] || currentValue[1] !== nextValue[1]) {
        var formattedValue = formatDate(nextValue, {
          format: format,
          targetFormat: valueType,
          autoSwap: true
        });
        onChange(formattedValue, {
          dayjsValue: nextValue.map(function (v) {
            return parseToDayjs(v, format);
          }),
          trigger: "confirm"
        });
      }
    } else {
      setInputValue(formatDate(value || [], {
        format: format
      }));
    }
  });
  useEffect(function () {
    if (value === cacheValue) return;
    if (popupVisible) {
      setIsSelected(false);
      setCacheValue(formatDate(value || [], {
        format: format
      }));
      setTime(formatTime(value || [dayjs().format(timeFormat), dayjs().format(timeFormat)], format, timeFormat, props.defaultTime));
      if (!value.length) {
        if (range && isArray(range) || panelActiveDate) {
          var _startRangeBounds$min, _startRangeBounds$min2, _startRangeBounds$max, _startRangeBounds$min3, _startRangeBounds$min4, _startRangeBounds$max2, _endRangeBounds$min$g, _endRangeBounds$min, _endRangeBounds$max, _endRangeBounds$min$g2, _endRangeBounds$min2, _endRangeBounds$max2, _startPanelActiveDate, _startPanelActiveDate2, _startPanelActiveDate3, _startPanelActiveDate4, _endPanelActiveDate$y, _endPanelActiveDate, _endPanelActiveDate2, _endPanelActiveDate3;
          var startRange = range;
          var endRange = range;
          if (isArray(range)) {
            var _range = _slicedToArray(range, 2),
              first = _range[0],
              second = _range[1];
            if (isArray(first) || isFunction(first) || isArray(second) || isFunction(second)) {
              startRange = first;
              endRange = second;
            }
          }
          var startRangeBounds = getRangeBounds(startRange);
          var endRangeBounds = getRangeBounds(endRange);
          var startYearFromRange = (_startRangeBounds$min = (_startRangeBounds$min2 = startRangeBounds.min) === null || _startRangeBounds$min2 === void 0 ? void 0 : _startRangeBounds$min2.getFullYear()) !== null && _startRangeBounds$min !== void 0 ? _startRangeBounds$min : (_startRangeBounds$max = startRangeBounds.max) === null || _startRangeBounds$max === void 0 ? void 0 : _startRangeBounds$max.getFullYear();
          var startMonthFromRange = (_startRangeBounds$min3 = (_startRangeBounds$min4 = startRangeBounds.min) === null || _startRangeBounds$min4 === void 0 ? void 0 : _startRangeBounds$min4.getMonth()) !== null && _startRangeBounds$min3 !== void 0 ? _startRangeBounds$min3 : (_startRangeBounds$max2 = startRangeBounds.max) === null || _startRangeBounds$max2 === void 0 ? void 0 : _startRangeBounds$max2.getMonth();
          var endYearFromRange = (_endRangeBounds$min$g = (_endRangeBounds$min = endRangeBounds.min) === null || _endRangeBounds$min === void 0 ? void 0 : _endRangeBounds$min.getFullYear()) !== null && _endRangeBounds$min$g !== void 0 ? _endRangeBounds$min$g : (_endRangeBounds$max = endRangeBounds.max) === null || _endRangeBounds$max === void 0 ? void 0 : _endRangeBounds$max.getFullYear();
          var endMonthFromRange = (_endRangeBounds$min$g2 = (_endRangeBounds$min2 = endRangeBounds.min) === null || _endRangeBounds$min2 === void 0 ? void 0 : _endRangeBounds$min2.getMonth()) !== null && _endRangeBounds$min$g2 !== void 0 ? _endRangeBounds$min$g2 : (_endRangeBounds$max2 = endRangeBounds.max) === null || _endRangeBounds$max2 === void 0 ? void 0 : _endRangeBounds$max2.getMonth();
          var startPanelActiveDate = panelActiveDate;
          var endPanelActiveDate = panelActiveDate;
          if (isArray(panelActiveDate)) {
            var _panelActiveDate = _slicedToArray(panelActiveDate, 2);
            startPanelActiveDate = _panelActiveDate[0];
            endPanelActiveDate = _panelActiveDate[1];
          }
          var leftYear = (_startPanelActiveDate = (_startPanelActiveDate2 = startPanelActiveDate) === null || _startPanelActiveDate2 === void 0 ? void 0 : _startPanelActiveDate2.year) !== null && _startPanelActiveDate !== void 0 ? _startPanelActiveDate : startYearFromRange;
          var leftMonth = (_startPanelActiveDate3 = startPanelActiveDate) !== null && _startPanelActiveDate3 !== void 0 && _startPanelActiveDate3.month ? Number((_startPanelActiveDate4 = startPanelActiveDate) === null || _startPanelActiveDate4 === void 0 ? void 0 : _startPanelActiveDate4.month) - 1 : startMonthFromRange;
          var rightYear = (_endPanelActiveDate$y = (_endPanelActiveDate = endPanelActiveDate) === null || _endPanelActiveDate === void 0 ? void 0 : _endPanelActiveDate.year) !== null && _endPanelActiveDate$y !== void 0 ? _endPanelActiveDate$y : endYearFromRange;
          var rightMonth = (_endPanelActiveDate2 = endPanelActiveDate) !== null && _endPanelActiveDate2 !== void 0 && _endPanelActiveDate2.month ? Number((_endPanelActiveDate3 = endPanelActiveDate) === null || _endPanelActiveDate3 === void 0 ? void 0 : _endPanelActiveDate3.month) - 1 : endMonthFromRange;
          var _initYearMonthTime = initYearMonthTime({
              value: value,
              mode: mode,
              format: format,
              enableTimePicker: enableTimePicker
            }),
            defaultYear = _initYearMonthTime.year,
            defaultMonth = _initYearMonthTime.month;
          var nextYear = [leftYear !== null && leftYear !== void 0 ? leftYear : defaultYear[0], rightYear !== null && rightYear !== void 0 ? rightYear : defaultYear[1]];
          var nextMonth = [leftMonth !== null && leftMonth !== void 0 ? leftMonth : defaultMonth[0], rightMonth !== null && rightMonth !== void 0 ? rightMonth : defaultMonth[1]];
          if (nextYear[0] === nextYear[1] && nextMonth[0] === nextMonth[1] && !enableTimePicker) {
            if (startRange === endRange) {
              if (mode === "year") nextYear[1] += 10;else if (mode === "month" || mode === "quarter") nextYear[1] += 1;else nextMonth[1] += 1;
            }
          }
          setYear(nextYear);
          setMonth(nextMonth);
        } else {
          var _initYearMonthTime2 = initYearMonthTime({
              value: value,
              mode: mode,
              format: format,
              enableTimePicker: enableTimePicker
            }),
            _defaultYear = _initYearMonthTime2.year,
            _defaultMonth = _initYearMonthTime2.month;
          setYear(_defaultYear);
          setMonth(_defaultMonth);
        }
      } else if (value.length === 2 && (!enableTimePicker || isSwitchTimeMode)) {
        handleSyncPanelValue(value);
      } else {
        var _nextYear = value.map(function (v) {
          return parseToDayjs(v, format).year();
        });
        if (_nextYear.length === 1) _nextYear = [_nextYear[0], _nextYear[0]];
        var _nextMonth = value.map(function (v) {
          return parseToDayjs(v, format).month();
        });
        if (_nextMonth.length === 1) _nextMonth = [_nextMonth[0], Math.min(_nextMonth[0] + 1, 11)];
        setYear(_nextYear);
        setMonth(_nextMonth);
      }
    } else {
      setActiveIndex(0);
      setIsHoverCell(false);
      setIsFirstValueSelected(false);
      setInputValue(formatDate(value || [], {
        format: format
      }));
    }
  }, [popupVisible]);
  useUpdateEffect(function () {
    onTriggerNeedConfirm.current();
  }, [popupVisible]);
  function onCellMouseEnter(date) {
    setIsHoverCell(true);
    var nextValue = _toConsumableArray(inputValue);
    nextValue[activeIndex] = formatDate(date, {
      format: format
    });
    setInputValue(nextValue);
  }
  function onCellMouseLeave() {
    setIsHoverCell(false);
    setInputValue(cacheValue);
  }
  function onCellClick(date, _ref) {
    var e = _ref.e;
    onPick === null || onPick === void 0 || onPick(date, {
      e: e,
      partial: activeIndex ? "end" : "start"
    });
    setIsHoverCell(false);
    setIsSelected(true);
    var nextValue = _toConsumableArray(inputValue);
    nextValue[activeIndex] = formatDate(date, {
      format: format
    });
    setCacheValue(nextValue);
    setInputValue(nextValue);
    if (enableTimePicker && !isSwitchTimeMode) return;
    var notValidIndex = nextValue.findIndex(function (v) {
      return !v || !isValidDate(v, format);
    });
    if (notValidIndex === -1 && nextValue.length === 2) {
      if (!isFirstValueSelected && parseToDayjs(nextValue[0], format).isAfter(parseToDayjs(nextValue[1], format))) {
        nextValue[activeIndex ? 0 : 1] = "";
        setCacheValue(nextValue);
        setInputValue(nextValue);
      } else {
        onChange(formatDate(nextValue, {
          format: format,
          targetFormat: valueType,
          autoSwap: true
        }), {
          dayjsValue: nextValue.map(function (v) {
            return parseToDayjs(v, format);
          }),
          trigger: "pick"
        });
      }
    }
    if (!isFirstValueSelected || !activeIndex || isSwitchTimeMode) {
      var nextIndex = notValidIndex;
      if (isSwitchTimeMode && activeIndex === 1 && isFirstValueSelected) return;
      if (nextIndex === -1) nextIndex = activeIndex ? 0 : 1;
      setActiveIndex(nextIndex);
      setIsFirstValueSelected(!!nextValue[0]);
    } else {
      handlePopupInvisible();
    }
  }
  function onJumperClick(_ref2) {
    var trigger = _ref2.trigger,
      partial = _ref2.partial;
    var partialIndex = partial === "start" ? 0 : 1;
    var monthCountMap = {
      date: 1,
      week: 1,
      month: 12,
      quarter: 12,
      year: 120
    };
    var monthCount = monthCountMap[mode] || 0;
    var current = new Date(year[partialIndex], month[partialIndex]);
    var next = null;
    if (trigger === "prev") {
      next = subtractMonth(current, monthCount);
    } else if (trigger === "current") {
      next = new Date();
    } else if (trigger === "next") {
      next = addMonth(current, monthCount);
    }
    var nextYear = _toConsumableArray(year);
    nextYear[partialIndex] = next.getFullYear();
    var nextMonth = _toConsumableArray(month);
    nextMonth[partialIndex] = next.getMonth();
    var onlyYearSelect = ["year", "quarter", "month"].includes(mode);
    var correctedDate = dateCorrection(partialIndex, nextYear, nextMonth, onlyYearSelect);
    nextYear = correctedDate.nextYear;
    nextMonth = correctedDate.nextMonth;
    setYear(nextYear);
    setMonth(nextMonth);
    var yearChanged = year[partialIndex] !== nextYear[partialIndex];
    var monthChanged = month[partialIndex] !== nextMonth[partialIndex];
    if (yearChanged) {
      var _props$onYearChange;
      (_props$onYearChange = props.onYearChange) === null || _props$onYearChange === void 0 || _props$onYearChange.call(props, {
        partial: partial,
        year: nextYear[partialIndex],
        date: value.map(function (v) {
          return dayjs(v).toDate();
        }),
        trigger: trigger === "current" ? "today" : "year-".concat(triggerMap[trigger])
      });
    }
    if (monthChanged) {
      var _props$onMonthChange;
      (_props$onMonthChange = props.onMonthChange) === null || _props$onMonthChange === void 0 || _props$onMonthChange.call(props, {
        partial: partial,
        month: nextMonth[partialIndex],
        date: value.map(function (v) {
          return dayjs(v).toDate();
        }),
        trigger: trigger === "current" ? "today" : "month-".concat(triggerMap[trigger])
      });
    }
  }
  function onTimePickerChange(val, context) {
    var _context$activeIndex;
    var _extractTimeObj = extractTimeObj(val),
      hours = _extractTimeObj.hours,
      minutes = _extractTimeObj.minutes,
      seconds = _extractTimeObj.seconds,
      milliseconds = _extractTimeObj.milliseconds,
      meridiem = _extractTimeObj.meridiem;
    var currentIndex = (_context$activeIndex = context.activeIndex) !== null && _context$activeIndex !== void 0 ? _context$activeIndex : activeIndex;
    var nextInputValue = _toConsumableArray(inputValue);
    var changedInputValue = inputValue[currentIndex];
    var currentDate = !dayjs(changedInputValue, format).isValid() ? dayjs().year(year[currentIndex]).month(month[currentIndex]) : dayjs(changedInputValue, format);
    var nextHours = hours;
    if (/am/i.test(meridiem) && nextHours === 12) nextHours -= 12;
    if (/pm/i.test(meridiem) && nextHours < 12) nextHours += 12;
    var nextDate = currentDate.hour(nextHours).minute(minutes).second(seconds).millisecond(milliseconds).toDate();
    nextInputValue[currentIndex] = nextDate;
    var nextTime = _toConsumableArray(time);
    nextTime[currentIndex] = val;
    setTime(nextTime);
    setIsSelected(true);
    setInputValue(formatDate(nextInputValue, {
      format: format
    }));
    setCacheValue(formatDate(nextInputValue, {
      format: format
    }));
  }
  function onConfirmClick(_ref3) {
    var _props$onConfirm;
    var e = _ref3.e;
    var nextValue = _toConsumableArray(inputValue);
    props === null || props === void 0 || (_props$onConfirm = props.onConfirm) === null || _props$onConfirm === void 0 || _props$onConfirm.call(props, {
      e: e,
      date: nextValue,
      partial: activeIndex ? "end" : "start"
    });
    var notValidIndex = nextValue.findIndex(function (v) {
      return !v || !isValidDate(v, format);
    });
    if (notValidIndex === -1 && nextValue.length === 2) {
      if (!isFirstValueSelected && parseToDayjs(nextValue[0], format).isAfter(parseToDayjs(nextValue[1], format))) {
        nextValue[activeIndex ? 0 : 1] = "";
        setCacheValue(nextValue);
        setInputValue(nextValue);
      } else {
        onChange(formatDate(nextValue, {
          format: format,
          targetFormat: valueType,
          autoSwap: true
        }), {
          dayjsValue: nextValue.map(function (v) {
            return parseToDayjs(v, format);
          }),
          trigger: "confirm"
        });
      }
    }
    if (isSwitchTimeMode && nextValue.every(Boolean)) handlePopupInvisible();
    if (!isFirstValueSelected || !activeIndex) {
      var nextIndex = notValidIndex;
      if (nextIndex === -1) nextIndex = activeIndex ? 0 : 1;
      setActiveIndex(nextIndex);
      setIsFirstValueSelected(!!nextValue[0]);
    } else if (nextValue.length === 2) {
      handlePopupInvisible();
    }
  }
  function onPresetClick(preset, context) {
    var presetValue = preset;
    if (typeof preset === "function") {
      presetValue = preset();
    }
    if (!Array.isArray(presetValue)) {
      log.error("DateRangePicker", "preset: ".concat(preset, " must be Array!"));
    } else {
      var _props$onPresetClick;
      var formattedPreset = formatDate(presetValue, {
        format: format,
        targetFormat: valueType
      });
      setInputValue(formattedPreset);
      setCacheValue(formattedPreset);
      setTime(formatTime(formattedPreset, format, timeFormat, props.defaultTime));
      setIsSelected(true);
      setIsFirstValueSelected(true);
      handleSyncPanelValue(formattedPreset);
      handlePopupInvisible();
      onChange(formattedPreset, {
        dayjsValue: formattedPreset.map(function (p) {
          return parseToDayjs(p, format);
        }),
        trigger: "preset"
      });
      (_props$onPresetClick = props.onPresetClick) === null || _props$onPresetClick === void 0 || _props$onPresetClick.call(props, context);
    }
  }
  function onYearChange(nextVal, _ref4) {
    var _props$onYearChange2;
    var partial = _ref4.partial;
    var partialIndex = partial === "start" ? 0 : 1;
    if (enableTimePicker) partialIndex = activeIndex;
    var nextYear = _toConsumableArray(year);
    var nextMonth = _toConsumableArray(month);
    nextYear[partialIndex] = nextVal;
    var onlyYearSelect = ["year", "quarter", "month"].includes(mode);
    var correctedDate = dateCorrection(partialIndex, nextYear, nextMonth, onlyYearSelect);
    nextYear = correctedDate.nextYear;
    nextMonth = correctedDate.nextMonth;
    setYear(nextYear);
    !onlyYearSelect && setMonth(nextMonth);
    (_props$onYearChange2 = props.onYearChange) === null || _props$onYearChange2 === void 0 || _props$onYearChange2.call(props, {
      partial: partial,
      year: nextYear[partialIndex],
      date: value.map(function (v) {
        return dayjs(v).toDate();
      }),
      trigger: "year-select"
    });
  }
  function onMonthChange(nextVal, _ref5) {
    var _props$onMonthChange2;
    var partial = _ref5.partial;
    var partialIndex = partial === "start" ? 0 : 1;
    var nextMonth = _toConsumableArray(month);
    if (enableTimePicker) partialIndex = activeIndex;
    setMonth(function () {
      nextMonth[partialIndex] = nextVal;
      if (year[0] === year[1]) {
        if (partialIndex === 0) {
          if (nextMonth[1] <= nextMonth[0]) {
            nextMonth[1] = nextMonth[0] + 1;
            if (nextMonth[1] === 12) {
              nextMonth[1] = 0;
              setYear(function (currentYear) {
                return [currentYear[0], currentYear[1] + 1];
              });
            }
          }
        }
        if (partialIndex === 1) {
          nextMonth[0] = Math.min(nextMonth[0], nextMonth[1]);
          if (nextMonth[0] >= nextMonth[1]) {
            nextMonth[0] -= 1;
            if (nextMonth[0] === -1) {
              nextMonth[0] = 11;
              setYear(function (currentYear) {
                return [currentYear[0] - 1, currentYear[1]];
              });
            }
          }
        }
      }
      return nextMonth;
    });
    (_props$onMonthChange2 = props.onMonthChange) === null || _props$onMonthChange2 === void 0 || _props$onMonthChange2.call(props, {
      partial: partial,
      month: nextMonth[partialIndex],
      date: value.map(function (v) {
        return dayjs(v).toDate();
      }),
      trigger: "month-select"
    });
  }
  var panelProps = {
    hoverValue: isHoverCell ? inputValue : [],
    value: isSelected ? cacheValue : value,
    isFirstValueSelected: isFirstValueSelected,
    panelPreselection: panelPreselection,
    year: year,
    month: month,
    range: range,
    mode: mode,
    format: format,
    presets: presets,
    time: time,
    disableDate: disableDate,
    firstDayOfWeek: firstDayOfWeek,
    timePickerProps: timePickerProps,
    enableTimePicker: enableTimePicker,
    presetsPlacement: presetsPlacement,
    activeIndex: activeIndex,
    popupVisible: popupVisible,
    cancelRangeSelectLimit: cancelRangeSelectLimit,
    needConfirm: needConfirm,
    cell: cell,
    onCellClick: onCellClick,
    onCellMouseEnter: onCellMouseEnter,
    onCellMouseLeave: onCellMouseLeave,
    onJumperClick: onJumperClick,
    onConfirmClick: onConfirmClick,
    onPresetClick: onPresetClick,
    onYearChange: onYearChange,
    onMonthChange: onMonthChange,
    onTimePickerChange: onTimePickerChange,
    disableTime: disableTime,
    isSwitchTimeMode: isSwitchTimeMode
  };
  return /* @__PURE__ */React.createElement("div", {
    className: classNames("".concat(classPrefix, "-date-range-picker"), className),
    style: style,
    ref: ref
  }, /* @__PURE__ */React.createElement(RangeInputPopup, {
    disabled: disabled,
    status: props.status,
    tips: props.tips,
    inputValue: inputValue,
    popupProps: popupProps,
    rangeInputProps: rangeInputProps,
    popupVisible: popupVisible,
    panel: /* @__PURE__ */React.createElement(RangePanel, _objectSpread({}, panelProps))
  }));
});
DateRangePicker.displayName = "DateRangePicker";

export { DateRangePicker as default };
