// tdesign-react/lib/loading/plugin.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-5e72c8b9.js';
import React from 'react';
import { render, unmount } from '../_util/react-render.js';
import Loading from './Loading.js';
import PluginContainer from '../common/PluginContainer.js';
import ConfigProvider from '../config-provider/ConfigProvider.js';
import '../config-provider/ConfigContext.js';
import '../_chunks/dep-43f7da7b.js';
import '../_chunks/dep-b9d1804c.js';
import 'react-dom';
import '../_chunks/dep-94a7af64.js';
import '../_chunks/dep-0054b574.js';
import '../_chunks/dep-8e8ac0d4.js';
import 'classnames';
import '../_chunks/dep-a347b1b5.js';
import 'lodash-es';
import '../common/Portal.js';
import '../hooks/useConfig.js';
import '../hooks/useLayoutEffect.js';
import '../hooks/useDefaultProps.js';
import './defaultProps.js';
import './gradient.js';
import '../_chunks/dep-2481dfad.js';
import '../_chunks/dep-ee4dfb46.js';
import '../_chunks/dep-1e9f924f.js';
import '../_chunks/dep-57e67945.js';
import '../_chunks/dep-4461a684.js';
import '../hooks/useDomRefCallback.js';
import '../_chunks/dep-923c86e8.js';
import '../_chunks/dep-03cf63f2.js';
import 'dayjs';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function createContainer(attach) {
  if (typeof attach === "string") return document.querySelector(attach);
  if (typeof attach === "function") return attach();
  return document.body;
}
var LoadingPlugin = function LoadingPlugin(options) {
  if (options === false) return {
    hide: function hide() {
      return null;
    }
  };
  var props = typeof options === "boolean" ? {} : options;
  var attach = props.attach;
  var container = createContainer(attach);
  var div = document.createElement("div");
  div.setAttribute("style", "width: 100%; height: 100%; position: absolute; top: 0;");
  var defaultProps = {
    loading: true,
    attach: null,
    fullscreen: !attach,
    showOverlay: !!attach
  };
  var dGlobalConfig = ConfigProvider.getGlobalConfig();
  render(/* @__PURE__ */React.createElement(PluginContainer, {
    globalConfig: dGlobalConfig
  }, /* @__PURE__ */React.createElement(Loading, _objectSpread(_objectSpread(_objectSpread({}, defaultProps), props), {}, {
    attach: null
  }))), div);
  container.appendChild(div);
  return {
    hide: function hide() {
      unmount(div);
      div.remove();
    }
  };
};

export { LoadingPlugin };
