// tdesign-react/es/time-picker/hooks/useTimePickerTextConfig.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../../_chunks/dep-94a7af64.js';
import { useLocaleReceiver } from '../../locale/LocalReceiver.js';
import '../../_chunks/dep-0054b574.js';
import '../../_chunks/dep-8e8ac0d4.js';
import '../../_chunks/dep-5e72c8b9.js';
import '../../_chunks/dep-43f7da7b.js';
import 'react';
import 'lodash-es';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';
import '../../config-provider/ConfigContext.js';
import '../../locale/zh_CN.js';
import '../../_chunks/dep-7d48ddac.js';
import 'dayjs';
import '../../_chunks/dep-766c4603.js';
import '../../config-provider/type.js';

var useTimePickerTextConfig = function useTimePickerTextConfig() {
  var _useLocaleReceiver = useLocaleReceiver("timePicker"),
    _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  return {
    nowTime: t(local.now),
    confirm: t(local.confirm),
    am: t(local.anteMeridiem),
    pm: t(local.postMeridiem),
    placeholder: t(local.placeholder),
    rangePlaceholder: [t(local.placeholder), t(local.placeholder)]
  };
};

export { useTimePickerTextConfig };
