// tdesign-react/lib/cascader/core/helper.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _typeof } from '../../_chunks/dep-43f7da7b.js';
import { P as PATH_SEPARATOR } from '../../_chunks/dep-5ecb9314.js';
import { isEmpty } from 'lodash-es';
import '../../_chunks/dep-5e72c8b9.js';
import '../../_chunks/dep-65d82dda.js';

function getSingleContent(cascaderContext) {
  var value = cascaderContext.value,
    multiple = cascaderContext.multiple,
    treeStore = cascaderContext.treeStore,
    showAllLevels = cascaderContext.showAllLevels,
    valueType = cascaderContext.valueType;
  var isEmpty2 = !value && value !== 0 || Array.isArray(value) && value.length === 0;
  var isInvalidFullPathValue = valueType === "full" && !Array.isArray(value);
  if (multiple || isEmpty2 || isInvalidFullPathValue) return "";
  var formatContent = function formatContent(path2) {
    if (!(path2 !== null && path2 !== void 0 && path2.length)) return "";
    return showAllLevels ? path2.map(function (node) {
      return node.label;
    }).join(" ".concat(PATH_SEPARATOR, " ")) : path2[path2.length - 1].label;
  };
  var getDefaultDisplay = function getDefaultDisplay(val) {
    return Array.isArray(val) ? val.join(" ".concat(PATH_SEPARATOR, " ")) : String(val);
  };
  if (valueType === "full" && Array.isArray(value) && value.length > 0) {
    var node = treeStore === null || treeStore === void 0 ? void 0 : treeStore.getNode(value);
    var path2 = node === null || node === void 0 ? void 0 : node.getPath();
    return path2 !== null && path2 !== void 0 && path2.length ? formatContent(path2) : getDefaultDisplay(value);
  }
  var nodes = treeStore === null || treeStore === void 0 ? void 0 : treeStore.getNodes(value);
  if (!(nodes !== null && nodes !== void 0 && nodes.length)) {
    return getDefaultDisplay(value);
  }
  var path = nodes[0].getPath();
  return path !== null && path !== void 0 && path.length ? formatContent(path) : getDefaultDisplay(value);
}
function getMultipleContent(cascaderContext) {
  var value = cascaderContext.value,
    multiple = cascaderContext.multiple,
    treeStore = cascaderContext.treeStore,
    showAllLevels = cascaderContext.showAllLevels,
    valueType = cascaderContext.valueType;
  if (!multiple) return [];
  if (multiple && !Array.isArray(value)) return [];
  return value.map(function (item) {
    var _node;
    var node;
    if (valueType === "full" && Array.isArray(item)) {
      node = treeStore === null || treeStore === void 0 ? void 0 : treeStore.getNode(item);
    } else {
      var nodes = treeStore === null || treeStore === void 0 ? void 0 : treeStore.getNodes(item);
      node = nodes === null || nodes === void 0 ? void 0 : nodes[0];
    }
    if (!node) return void 0;
    return showAllLevels ? getFullPathLabel(node) : (_node = node) === null || _node === void 0 ? void 0 : _node.label;
  }).filter(function (item) {
    return !!item;
  });
}
function getPanels(treeNodes) {
  var panels = [];
  treeNodes.forEach(function (node) {
    if (panels[node.level]) {
      panels[node.level].push(node);
    } else {
      panels[node.level] = [node];
    }
  });
  return panels;
}
function getFullPathLabel(node) {
  var separator = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "/";
  return node === null || node === void 0 ? void 0 : node.getPath().map(function (node2) {
    return node2.label;
  }).join(separator);
}
var getTreeValue = function getTreeValue(value) {
  var treeValue = [];
  if (Array.isArray(value)) {
    if (value.length > 0 && _typeof(value[0]) === "object") {
      treeValue = value.map(function (val) {
        return val.value;
      });
    } else if (value.length) {
      treeValue = value;
    }
  } else if (value) {
    if (_typeof(value) === "object") {
      treeValue = [value.value];
    } else {
      treeValue = [value];
    }
  }
  return treeValue;
};
var getExpandValue = function getExpandValue(value, valueType) {
  if (valueType === "full" && Array.isArray(value) && value.length > 0) {
    return value[value.length - 1];
  }
  var treeValue = getTreeValue(value);
  return treeValue[0];
};
function isEmptyValues(value) {
  if (typeof value === "number" && !isNaN(value)) return false;
  return isEmpty(value);
}
function isValueInvalid(value, cascaderContext) {
  var multiple = cascaderContext.multiple,
    showAllLevels = cascaderContext.showAllLevels,
    valueType = cascaderContext.valueType;
  if (multiple) {
    return !Array.isArray(value);
  }
  if (valueType === "full") {
    return !Array.isArray(value);
  }
  if (Array.isArray(value)) {
    return !showAllLevels;
  }
  return false;
}

export { getExpandValue, getFullPathLabel, getMultipleContent, getPanels, getSingleContent, getTreeValue, isEmptyValues, isValueInvalid };
