// tdesign-react/cjs/time-picker/panel/TimePickerPanel.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var React = require('react');
var dayjs = require('dayjs');
var timePicker_panel_SinglePanel = require('../../_chunks/dep-9220a4d6.js');
var log = require('../../_chunks/dep-b289b0be.js');
var hooks_useConfig = require('../../hooks/useConfig.js');
var button_index = require('../../button/index.js');
var timePicker_hooks_useTimePickerTextConfig = require('../hooks/useTimePickerTextConfig.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('classnames');
require('../../_chunks/dep-bf18bb11.js');
require('../../_chunks/dep-06bf4625.js');
require('../../_chunks/dep-c990dbb4.js');
require('../../_chunks/dep-eab994e8.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-876d090e.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_util/noop.js');
require('../../hooks/useDebounce.js');
require('../../hooks/usePersistFn.js');
require('../../_chunks/dep-d3af8eb7.js');
require('../../hooks/usePropsRef.js');
require('../../_chunks/dep-2d5cf7d6.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-b4442734.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-b529a0d7.js');
require('../../_chunks/dep-bcd16957.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-14013769.js');
require('../../_chunks/dep-043eee20.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../button/Button.js');
require('../../_chunks/dep-1c586f5c.js');
require('../../hooks/useDomRefCallback.js');
require('../../hooks/useRipple.js');
require('../../_chunks/dep-064e4dba.js');
require('../../hooks/useAnimation.js');
require('../../_chunks/dep-68f5fe7b.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../loading/index.js');
require('../../loading/Loading.js');
require('../../common/Portal.js');
require('react-dom');
require('../../hooks/useLayoutEffect.js');
require('../../hooks/useDefaultProps.js');
require('../../loading/defaultProps.js');
require('../../loading/gradient.js');
require('../../_chunks/dep-b011b5d3.js');
require('../../_chunks/dep-1d0dbfc5.js');
require('../../_chunks/dep-3c2d5e8c.js');
require('../../_chunks/dep-f0e884b9.js');
require('../../loading/plugin.js');
require('../../_util/react-render.js');
require('../../_chunks/dep-e37cb830.js');
require('../../common/PluginContainer.js');
require('../../config-provider/ConfigProvider.js');
require('../../_chunks/dep-4a562e04.js');
require('../../_chunks/dep-a5b4810f.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-ea361d31.js');
require('../../button/defaultProps.js');
require('../../_util/parseTNode.js');
require('../../locale/LocalReceiver.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var TimePickerPanel = function TimePickerPanel(props) {
  var _props$format = props.format,
    format = _props$format === void 0 ? timePicker_panel_SinglePanel.DEFAULT_FORMAT : _props$format,
    _props$steps = props.steps,
    steps = _props$steps === void 0 ? timePicker_panel_SinglePanel.DEFAULT_STEPS : _props$steps,
    handleConfirmClick = props.handleConfirmClick,
    isFooterDisplay = props.isFooterDisplay,
    onChange = props.onChange,
    value = props.value,
    _props$isShowPanel = props.isShowPanel,
    isShowPanel = _props$isShowPanel === void 0 ? true : _props$isShowPanel,
    _props$presets = props.presets,
    presets = _props$presets === void 0 ? null : _props$presets;
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    triggerScroll = _useState2[0],
    toggleTriggerScroll = _useState2[1];
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var TEXT_CONFIG = timePicker_hooks_useTimePickerTextConfig.useTimePickerTextConfig();
  var panelClassName = "".concat(classPrefix, "-time-picker__panel");
  var showNowTimeBtn = !!steps.filter(function (v) {
    return Number(v) > 1;
  }).length;
  var defaultValue = React.useMemo(function () {
    var formattedValue = dayjs__default["default"](value, format);
    if (value && formattedValue.isValid()) {
      return formattedValue.format(format);
    }
    return dayjs__default["default"]().hour(0).minute(0).second(0).format(format);
  }, [value, format]);
  React.useEffect(function () {
    if (isShowPanel) toggleTriggerScroll(true);
  }, [isShowPanel]);
  var resetTriggerScroll = React.useCallback(function () {
    toggleTriggerScroll(false);
  }, [toggleTriggerScroll]);
  var handlePresetClick = function handlePresetClick(presetValue) {
    var presetVal = typeof presetValue === "function" ? presetValue() : presetValue;
    if (typeof props.activeIndex === "number") {
      if (Array.isArray(presetVal)) {
        var _props$onChange;
        (_props$onChange = props.onChange) === null || _props$onChange === void 0 || _props$onChange.call(props, presetVal[props.activeIndex]);
      } else {
        log.log.error("TimePicker", "preset: ".concat(presets, " \u9884\u8BBE\u503C\u5FC5\u987B\u662F\u6570\u7EC4!"));
      }
    } else {
      var _props$onChange2;
      (_props$onChange2 = props.onChange) === null || _props$onChange2 === void 0 || _props$onChange2.call(props, presetVal);
    }
  };
  var renderFooter = function renderFooter() {
    if (presets) {
      return Object.keys(presets).map(function (preset) {
        return /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
          key: preset,
          theme: "primary",
          size: "small",
          variant: "text",
          onClick: function onClick() {
            handlePresetClick(presets[preset]);
          }
        }, preset);
      });
    }
    return !showNowTimeBtn ? /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
      theme: "primary",
      variant: "text",
      size: "small",
      onClick: function onClick() {
        onChange === null || onChange === void 0 || onChange(dayjs__default["default"]().format(format));
      }
    }, TEXT_CONFIG.nowTime) : null;
  };
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: panelClassName
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(panelClassName, "-section-body")
  }, /* @__PURE__ */React__default["default"].createElement(timePicker_panel_SinglePanel.SinglePanel, _objectSpread(_objectSpread({}, props), {}, {
    onChange: onChange,
    format: format,
    steps: steps,
    value: dayjs__default["default"](value, format).isValid() ? value : defaultValue,
    triggerScroll: triggerScroll,
    isVisible: isShowPanel,
    resetTriggerScroll: resetTriggerScroll
  }))), isFooterDisplay ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(panelClassName, "-section-footer")
  }, /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    size: "small",
    theme: "primary",
    variant: "base",
    disabled: !props.value,
    onClick: function onClick() {
      handleConfirmClick === null || handleConfirmClick === void 0 || handleConfirmClick(defaultValue);
    }
  }, TEXT_CONFIG.confirm), renderFooter()) : null);
};

exports["default"] = TimePickerPanel;
