// tdesign-react/lib/dialog/plugin.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-5e72c8b9.js';
import React from 'react';
import { l as log } from '../_chunks/dep-65d82dda.js';
import { render } from '../_util/react-render.js';
import Dialog from './Dialog.js';
import { b as getAttach } from '../_chunks/dep-a347b1b5.js';
import PluginContainer from '../common/PluginContainer.js';
import ConfigProvider from '../config-provider/ConfigProvider.js';
import '../config-provider/ConfigContext.js';
import '../_chunks/dep-43f7da7b.js';
import '../_chunks/dep-b9d1804c.js';
import 'react-dom';
import '../_chunks/dep-94a7af64.js';
import '../_chunks/dep-0054b574.js';
import '../_chunks/dep-8e8ac0d4.js';
import '../_chunks/dep-4461a684.js';
import 'react-transition-group';
import 'classnames';
import 'lodash-es';
import '../_chunks/dep-ee4dfb46.js';
import '../_chunks/dep-1e9f924f.js';
import '../_chunks/dep-57e67945.js';
import '../common/Portal.js';
import '../hooks/useConfig.js';
import '../hooks/useLayoutEffect.js';
import '../hooks/useAttach.js';
import '../hooks/useDeepEffect.js';
import '../hooks/useDefaultProps.js';
import '../hooks/useSetState.js';
import '../locale/LocalReceiver.js';
import './defaultProps.js';
import './DialogCard.js';
import 'tdesign-icons-react';
import '../_util/parseTNode.js';
import '../button/index.js';
import '../button/Button.js';
import '../hooks/useDomRefCallback.js';
import '../hooks/useRipple.js';
import '../_chunks/dep-2481dfad.js';
import '../hooks/useAnimation.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../loading/defaultProps.js';
import '../loading/gradient.js';
import '../_chunks/dep-923c86e8.js';
import '../_chunks/dep-03cf63f2.js';
import 'dayjs';
import '../loading/plugin.js';
import '../button/defaultProps.js';
import '../hooks/useGlobalIcon.js';
import './hooks/useDialogDrag.js';
import '../hooks/useMouseEvent.js';
import './hooks/useDialogEsc.js';
import './hooks/useDialogPosition.js';
import './hooks/useLockStyle.js';
import '../_chunks/dep-7aa18730.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var createDialog = function createDialog(props) {
  var dialogRef = /*#__PURE__*/React.createRef();
  var options = _objectSpread({}, props);
  var _options$visible = options.visible,
    visible = _options$visible === void 0 ? true : _options$visible;
  var fragment = document.createDocumentFragment();
  var dGlobalConfig = ConfigProvider.getGlobalConfig();
  render(/* @__PURE__ */React.createElement(PluginContainer, {
    globalConfig: dGlobalConfig
  }, /* @__PURE__ */React.createElement(Dialog, _objectSpread(_objectSpread({}, options), {}, {
    visible: visible,
    ref: dialogRef,
    isPlugin: true
  }))), fragment);
  var container = getAttach(options.attach);
  if (container) {
    container.appendChild(fragment);
  } else {
    log.error("Dialog", "attach is not exist");
  }
  var dialogNode = {
    show: function show() {
      requestAnimationFrame(function () {
        var _dialogRef$current;
        container.appendChild(fragment);
        (_dialogRef$current = dialogRef.current) === null || _dialogRef$current === void 0 || _dialogRef$current.show();
      });
    },
    hide: function hide() {
      requestAnimationFrame(function () {
        var _dialogRef$current2;
        (_dialogRef$current2 = dialogRef.current) === null || _dialogRef$current2 === void 0 || _dialogRef$current2.destroy();
      });
    },
    setConfirmLoading: function setConfirmLoading(loading) {
      requestAnimationFrame(function () {
        var _dialogRef$current3;
        (_dialogRef$current3 = dialogRef.current) === null || _dialogRef$current3 === void 0 || _dialogRef$current3.setConfirmLoading(loading);
      });
    },
    update: function update(updateOptions) {
      requestAnimationFrame(function () {
        var _dialogRef$current4;
        (_dialogRef$current4 = dialogRef.current) === null || _dialogRef$current4 === void 0 || _dialogRef$current4.update(updateOptions);
      });
    },
    destroy: function destroy() {
      requestAnimationFrame(function () {
        var _dialogRef$current5;
        (_dialogRef$current5 = dialogRef.current) === null || _dialogRef$current5 === void 0 || _dialogRef$current5.destroy();
      });
    }
  };
  return dialogNode;
};
var confirm = function confirm(props) {
  return createDialog(props);
};
var alert = function alert(props) {
  var options = _objectSpread({}, props);
  options.cancelBtn = null;
  return createDialog(options);
};
createDialog.alert = alert;
createDialog.confirm = confirm;
var DialogPlugin = createDialog;

export { DialogPlugin };
