// tdesign-react/cjs/list/List.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var _util_noop = require('../_util/noop.js');
var _util_parseTNode = require('../_util/parseTNode.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var loading_index = require('../loading/index.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var list_ListItem = require('./ListItem.js');
var list_ListItemMeta = require('./ListItemMeta.js');
var list_defaultProps = require('./defaultProps.js');
var list_hooks_useListVirtualScroll = require('./hooks/useListVirtualScroll.js');
var isString = require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('hoist-non-react-statics');
require('../_chunks/dep-b289b0be.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-aff91ab0.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../loading/Loading.js');
require('../_chunks/dep-68f5fe7b.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-1c586f5c.js');
require('../_chunks/dep-f0e884b9.js');
require('../hooks/useDomRefCallback.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../hooks/useVirtualScroll.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

/**
 * Creates an array with all falsey values removed. The values `false`, `null`,
 * `0`, `-0`, `0n`, `""`, `undefined`, and `NaN` are falsy.
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Array
 * @param {Array} array The array to compact.
 * @returns {Array} Returns the new array of filtered values.
 * @example
 *
 * _.compact([0, 1, false, 2, '', 3]);
 * // => [1, 2, 3]
 */
function compact(array) {
  var index = -1,
    length = array == null ? 0 : array.length,
    resIndex = 0,
    result = [];
  while (++index < length) {
    var value = array[index];
    if (value) {
      result[resIndex++] = value;
    }
  }
  return result;
}

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var List = _util_forwardRefWithStatics["default"](function (props, ref) {
  var _useDefaultProps = hooks_useDefaultProps["default"](props, list_defaultProps.listDefaultProps),
    header = _useDefaultProps.header,
    footer = _useDefaultProps.footer,
    asyncLoading = _useDefaultProps.asyncLoading,
    size = _useDefaultProps.size,
    split = _useDefaultProps.split,
    stripe = _useDefaultProps.stripe,
    layout = _useDefaultProps.layout,
    children = _useDefaultProps.children,
    className = _useDefaultProps.className,
    _useDefaultProps$onLo = _useDefaultProps.onLoadMore,
    onLoadMore = _useDefaultProps$onLo === void 0 ? _util_noop["default"] : _useDefaultProps$onLo,
    _useDefaultProps$onSc = _useDefaultProps.onScroll,
    onScroll = _useDefaultProps$onSc === void 0 ? _util_noop["default"] : _useDefaultProps$onSc,
    style = _useDefaultProps.style,
    scroll = _useDefaultProps.scroll;
  var wrapperRef = React.useRef(null);
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("list"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var listItems = React.useMemo(function () {
    var _compact;
    return (_compact = compact(React__default["default"].Children.map(children, function (child) {
      return child === null || child === void 0 ? void 0 : child.props;
    }))) !== null && _compact !== void 0 ? _compact : [];
  }, [children]);
  var _useListVirtualScroll = list_hooks_useListVirtualScroll.useListVirtualScroll(scroll, wrapperRef, listItems),
    virtualConfig = _useListVirtualScroll.virtualConfig,
    cursorStyle = _useListVirtualScroll.cursorStyle,
    listStyle = _useListVirtualScroll.listStyle,
    isVirtualScroll = _useListVirtualScroll.isVirtualScroll,
    onInnerVirtualScroll = _useListVirtualScroll.onInnerVirtualScroll,
    scrollToElement = _useListVirtualScroll.scrollToElement;
  var COMPONENT_NAME = "".concat(classPrefix, "-list");
  var handleClickLoad = function handleClickLoad(e) {
    if (asyncLoading === "load-more") {
      onLoadMore({
        e: e
      });
    }
  };
  var handleScroll = function handleScroll(event) {
    var currentTarget = event.currentTarget;
    var scrollTop = currentTarget.scrollTop,
      offsetHeight = currentTarget.offsetHeight,
      scrollHeight = currentTarget.scrollHeight;
    var scrollBottom = scrollHeight - scrollTop - offsetHeight;
    if (isVirtualScroll) onInnerVirtualScroll(event);
    onScroll({
      e: event,
      scrollTop: scrollTop,
      scrollBottom: scrollBottom
    });
  };
  var loadElement = isString.isString(asyncLoading) ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-list__load"), defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-list__load--loading"), asyncLoading === "loading"), "".concat(classPrefix, "-list__load--load-more"), asyncLoading === "load-more")),
    onClick: handleClickLoad
  }, asyncLoading === "loading" && /* @__PURE__ */React__default["default"].createElement("div", null, /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, {
    loading: true
  }), /* @__PURE__ */React__default["default"].createElement("span", null, t(local.loadingText))), asyncLoading === "load-more" && /* @__PURE__ */React__default["default"].createElement("span", null, t(local.loadingMoreText))) : asyncLoading;
  React.useImperativeHandle(ref, function () {
    return {
      scrollTo: scrollToElement
    };
  });
  var renderContent = function renderContent() {
    return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, isVirtualScroll ? /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement("div", {
      style: cursorStyle
    }), /* @__PURE__ */React__default["default"].createElement("ul", {
      className: "".concat(COMPONENT_NAME, "__inner"),
      style: listStyle
    }, virtualConfig.visibleData.map(function (item, index) {
      return /* @__PURE__ */React__default["default"].createElement(list_ListItem["default"], _objectSpread({
        key: index
      }, item));
    }))) : /* @__PURE__ */React__default["default"].createElement("ul", {
      className: "".concat(COMPONENT_NAME, "__inner")
    }, children));
  };
  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: wrapperRef,
    style: _objectSpread(_objectSpread({}, style), {}, {
      position: isVirtualScroll ? "relative" : void 0
    }),
    onScroll: handleScroll,
    className: classNames__default["default"]("".concat(COMPONENT_NAME), className, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(COMPONENT_NAME, "--split"), split), "".concat(COMPONENT_NAME, "--stripe"), stripe), "".concat(COMPONENT_NAME, "--vertical-action"), layout === "vertical"), "".concat(classPrefix, "-size-s"), size === "small"), "".concat(classPrefix, "-size-l"), size === "large"))
  }, header && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(COMPONENT_NAME, "__header")
  }, _util_parseTNode["default"](header)), renderContent(), asyncLoading && loadElement, footer && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(COMPONENT_NAME, "__footer")
  }, _util_parseTNode["default"](footer)));
}, {
  ListItem: list_ListItem["default"],
  ListItemMeta: list_ListItemMeta["default"]
});
List.displayName = "List";

exports["default"] = List;
