// tdesign-react/cjs/cascader/components/Item.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var loading_index = require('../../loading/index.js');
var checkbox_index = require('../../checkbox/index.js');
var hooks_useConfig = require('../../hooks/useConfig.js');
var hooks_useGlobalIcon = require('../../hooks/useGlobalIcon.js');
var hooks_useDomRefCallback = require('../../hooks/useDomRefCallback.js');
var hooks_useCommonClassName = require('../../hooks/useCommonClassName.js');
var cascader_core_helper = require('../core/helper.js');
var cascader_core_className = require('../core/className.js');
var isFunction = require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../loading/Loading.js');
require('../../_chunks/dep-68f5fe7b.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../common/Portal.js');
require('react-dom');
require('../../hooks/useLayoutEffect.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-14013769.js');
require('../../_chunks/dep-06bf4625.js');
require('dayjs');
require('../../_chunks/dep-043eee20.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../_chunks/dep-2d5cf7d6.js');
require('../../hooks/useDefaultProps.js');
require('../../loading/defaultProps.js');
require('../../loading/gradient.js');
require('../../_chunks/dep-064e4dba.js');
require('../../_chunks/dep-b011b5d3.js');
require('../../_chunks/dep-1d0dbfc5.js');
require('../../_chunks/dep-3c2d5e8c.js');
require('../../_chunks/dep-1c586f5c.js');
require('../../_chunks/dep-f0e884b9.js');
require('../../loading/plugin.js');
require('../../_util/react-render.js');
require('../../_chunks/dep-e37cb830.js');
require('../../common/PluginContainer.js');
require('../../config-provider/ConfigProvider.js');
require('../../_chunks/dep-4a562e04.js');
require('../../_chunks/dep-a5b4810f.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-ea361d31.js');
require('../../_chunks/dep-83710d84.js');
require('../../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../../common/Check.js');
require('../../_util/helper.js');
require('../../_chunks/dep-e44b03ec.js');
require('../../_chunks/dep-b4442734.js');
require('../../_chunks/dep-876d090e.js');
require('../../_chunks/dep-eb1b216c.js');
require('../../_chunks/dep-b529a0d7.js');
require('../../_chunks/dep-bcd16957.js');
require('../../hooks/useControlled.js');
require('../../_util/noop.js');
require('../../_chunks/dep-089e8106.js');
require('../../checkbox/defaultProps.js');
require('../../_chunks/dep-91c4f617.js');
require('../../_chunks/dep-b289b0be.js');
require('../../_chunks/dep-f7bf3d17.js');
require('../../_chunks/dep-5e17a073.js');
require('../../_chunks/dep-ca918872.js');
require('../../_chunks/dep-13de2fef.js');
require('../../_chunks/dep-63308b84.js');
require('../../_chunks/dep-048e4314.js');
require('../../_chunks/dep-5c3139cb.js');
require('../../_chunks/dep-2e455d91.js');
require('../../_chunks/dep-894a5ece.js');
require('../../_chunks/dep-8f62fb96.js');
require('../../_chunks/dep-545dd164.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Item = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var node = props.node,
    optionChild = props.optionChild,
    multiple = props.cascaderContext.multiple,
    _onClick = props.onClick,
    _onChange = props.onChange,
    _onMouseEnter = props.onMouseEnter,
    cascaderContext = props.cascaderContext;
  var _useConfig = hooks_useConfig["default"](),
    prefix = _useConfig.classPrefix;
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      ChevronRightIcon: tdesignIconsReact.ChevronRightIcon
    }),
    ChevronRightIcon = _useGlobalIcon.ChevronRightIcon;
  var COMPONENT_NAME = "".concat(prefix, "-cascader__item");
  var _useDomRefCallback = hooks_useDomRefCallback["default"](),
    _useDomRefCallback2 = slicedToArray._slicedToArray(_useDomRefCallback, 2);
    _useDomRefCallback2[0];
    _useDomRefCallback2[1];
  var _useCommonClassName = hooks_useCommonClassName["default"](),
    STATUS = _useCommonClassName.STATUS,
    SIZE = _useCommonClassName.SIZE;
  var itemClass = React.useMemo(function () {
    return classNames__default["default"](cascader_core_className.getCascaderItemClass(prefix, node, SIZE, STATUS, cascaderContext));
  }, [prefix, node, SIZE, STATUS, cascaderContext]);
  var iconClass = React.useMemo(function () {
    return classNames__default["default"](cascader_core_className.getCascaderItemIconClass(prefix, node, STATUS, cascaderContext));
  }, [prefix, node, STATUS, cascaderContext]);
  var RenderLabelInner = function RenderLabelInner(node2, cascaderContext2) {
    var inputVal = cascaderContext2.inputVal;
    if (!inputVal && optionChild) {
      return optionChild;
    }
    var labelText = inputVal ? cascader_core_helper.getFullPathLabel(node2) : node2.label;
    if (inputVal) {
      var texts = labelText.split(inputVal);
      var doms = [];
      for (var index = 0; index < texts.length; index++) {
        doms.push(/* @__PURE__ */React__default["default"].createElement("span", {
          key: index
        }, texts[index]));
        if (index === texts.length - 1) break;
        doms.push(/* @__PURE__ */React__default["default"].createElement("span", {
          key: "".concat(index, "filter"),
          className: "".concat(COMPONENT_NAME, "-label--filter")
        }, inputVal));
      }
      return doms;
    }
    return labelText;
  };
  var RenderLabelContent = function RenderLabelContent(node2, cascaderContext2) {
    var label = RenderLabelInner(node2, cascaderContext2);
    var labelCont = /* @__PURE__ */React__default["default"].createElement("span", {
      title: cascaderContext2.inputVal ? cascader_core_helper.getFullPathLabel(node2) : node2.label,
      className: classNames__default["default"]("".concat(COMPONENT_NAME, "-label"), "".concat(COMPONENT_NAME, "-label--ellipsis")),
      role: "label"
    }, label);
    return labelCont;
  };
  var RenderCheckBox = function RenderCheckBox(node2, cascaderContext2) {
    var checkProps = cascaderContext2.checkProps,
      value = cascaderContext2.value,
      max = cascaderContext2.max,
      inputVal = cascaderContext2.inputVal,
      isParentFilterable = cascaderContext2.isParentFilterable;
    var label = RenderLabelInner(node2, cascaderContext2);
    return /* @__PURE__ */React__default["default"].createElement(checkbox_index.Checkbox, _objectSpread({
      checked: node2.checked,
      indeterminate: node2.indeterminate,
      disabled: node2.isDisabled() || value && value.length >= max && max !== 0,
      name: String(node2.value),
      stopLabelTrigger: !!node2.children && !isParentFilterable,
      title: inputVal ? cascader_core_helper.getFullPathLabel(node2) : node2.label,
      onChange: function onChange() {
        _onChange(node2);
      }
    }, checkProps), label);
  };
  var isFiltering = React.useMemo(function () {
    return Boolean(cascaderContext.filterable) || isFunction.isFunction(cascaderContext.filter);
  }, [cascaderContext.filterable, cascaderContext.filter]);
  return /* @__PURE__ */React__default["default"].createElement("li", {
    ref: ref,
    className: itemClass,
    onClick: function onClick(e) {
      var _e$nativeEvent, _e$nativeEvent$stopIm;
      e.stopPropagation();
      e === null || e === void 0 || (_e$nativeEvent = e.nativeEvent) === null || _e$nativeEvent === void 0 || (_e$nativeEvent$stopIm = _e$nativeEvent.stopImmediatePropagation) === null || _e$nativeEvent$stopIm === void 0 || _e$nativeEvent$stopIm.call(_e$nativeEvent);
      if (multiple && cascaderContext.inputVal && isFiltering) return;
      _onClick(node);
    },
    onMouseEnter: function onMouseEnter(e) {
      e.stopPropagation();
      if (cascaderContext.inputVal && isFiltering) return;
      _onMouseEnter(node);
    }
  }, multiple ? RenderCheckBox(node, cascaderContext) : RenderLabelContent(node, cascaderContext), node.children && !cascaderContext.isParentFilterable && (node.loading ? /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, {
    className: iconClass,
    loading: true,
    size: "small"
  }) : /* @__PURE__ */React__default["default"].createElement(ChevronRightIcon, {
    className: iconClass
  })));
});

exports["default"] = Item;
