// tdesign-react/cjs/menu/SubMenu.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var classNames = require('classnames');
var React = require('react');
var reactTransitionGroup = require('react-transition-group');
var _util_parseTNode = require('../_util/parseTNode.js');
var common_FakeArrow = require('../common/FakeArrow.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var hooks_useDomRefCallback = require('../hooks/useDomRefCallback.js');
var hooks_useRipple = require('../hooks/useRipple.js');
var popup_index = require('../popup/index.js');
var menu__util_calculatePaddingLeft = require('./_util/calculatePaddingLeft.js');
var menu__util_checkMenuType = require('./_util/checkMenuType.js');
var menu__util_checkSubMenuChildrenActive = require('./_util/checkSubMenuChildrenActive.js');
var menu__util_getSubMenuChildStyle = require('./_util/getSubMenuChildStyle.js');
var menu_MenuContext = require('./MenuContext.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-b289b0be.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-f0e884b9.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-794b42ce.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-064e4dba.js');
require('../hooks/useAnimation.js');
require('../_chunks/dep-68f5fe7b.js');
require('../popup/Popup.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../hooks/useAttach.js');
require('../hooks/useDefaultProps.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/listener.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');
require('../popup/utils/transition.js');
require('../popup/PopupPlugin.js');
require('../_chunks/dep-e37cb830.js');
require('../_util/react-render.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('./_util/type.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var _excluded = ["overlayClassName", "overlayInnerClassName"],
  _excluded2 = ["overlayClassName", "overlayInnerClassName"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var SubAccordion = function SubAccordion(props) {
  var content = props.content,
    _props$children = props.children,
    children = _props$children === void 0 ? content : _props$children,
    disabled = props.disabled,
    icon = props.icon,
    title = props.title,
    value = props.value,
    className = props.className,
    style = props.style,
    _props$level = props.level,
    level = _props$level === void 0 ? 1 : _props$level,
    popupProps = props.popupProps;
  var _ref = popupProps || {},
    overlayClassName = _ref.overlayClassName,
    overlayInnerClassName = _ref.overlayInnerClassName,
    restPopupProps = objectWithoutProperties._objectWithoutProperties(_ref, _excluded);
  var _useControlled = hooks_useControlled["default"](popupProps, "visible", restPopupProps.onVisibleChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    open = _useControlled2[0],
    setOpen = _useControlled2[1];
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _useContext = React.useContext(menu_MenuContext.MenuContext),
    _useContext$expanded = _useContext.expanded,
    expanded = _useContext$expanded === void 0 ? [] : _useContext$expanded,
    onExpand = _useContext.onExpand,
    active = _useContext.active,
    expandType = _useContext.expandType,
    _useContext$theme = _useContext.theme,
    theme = _useContext$theme === void 0 ? "light" : _useContext$theme;
  var isPopUp = expandType === "popup";
  var isExpand = expanded.includes(value) && !disabled && !isPopUp;
  var handleClick = function handleClick(e) {
    e.stopPropagation();
    onExpand(value, expanded);
    setOpen(false, {});
  };
  var handleVisibleChange = function handleVisibleChange(visible, ctx) {
    setOpen(visible, ctx);
  };
  var popupChildren = React__default["default"].Children.map(children, function (child) {
    var _child$props;
    return /*#__PURE__*/React__default["default"].cloneElement(child, {
      className: classNames__default["default"]("".concat(classPrefix, "-menu__item--plain"), "".concat(classPrefix, "-submenu__item"), "".concat(classPrefix, "-submenu__item--icon"), (_child$props = child.props) === null || _child$props === void 0 ? void 0 : _child$props.className)
    });
  });
  var childStyle = {
    maxHeight: isExpand || open && isPopUp ? menu__util_getSubMenuChildStyle.getSubMenuMaxHeight(children) : 0
  };
  var isOpen = React.useMemo(function () {
    if (disabled) return false;
    if (isPopUp) return open;
    return isExpand;
  }, [disabled, isPopUp, open, isExpand]);
  var menuPaddingLeft = menu__util_calculatePaddingLeft.calculatePaddingLeft(level - 1);
  var fakeArrowStyle = isPopUp && level > 1 ? {
    transform: "rotate(-90deg)"
  } : {};
  var contentRef = React.useRef(null);
  var transitionCallbacks = {
    onEnter: function onEnter() {
      contentRef.current.style.height = "".concat(contentRef === null || contentRef === void 0 ? void 0 : contentRef.current.scrollHeight, "px");
    },
    onEntered: function onEntered() {
      contentRef.current.style.height = "auto";
    },
    onExit: function onExit() {
      contentRef.current.style.height = "".concat(contentRef === null || contentRef === void 0 ? void 0 : contentRef.current.scrollHeight, "px");
    },
    onExiting: function onExiting() {
      contentRef.current.style.height = "0px";
    }
  };
  var pupContent = /* @__PURE__ */React__default["default"].createElement("ul", {
    className: classNames__default["default"]("".concat(classPrefix, "-menu__popup-wrapper"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-opened"), isOpen)),
    key: "popup",
    style: childStyle
  }, popupChildren);
  var submenu = /* @__PURE__ */React__default["default"].createElement("li", {
    className: classNames__default["default"]("".concat(classPrefix, "-submenu"), className, defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-is-disabled"), disabled), "".concat(classPrefix, "-is-opened"), isOpen)),
    style: style
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-menu__item"), defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-is-opened"), isOpen), "".concat(classPrefix, "-is-active"), menu__util_checkSubMenuChildrenActive["default"](children, active))),
    onClick: handleClick
  }, _util_parseTNode["default"](icon), /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-menu__content")
  }, title), /* @__PURE__ */React__default["default"].createElement(common_FakeArrow["default"], {
    style: fakeArrowStyle,
    isActive: isOpen,
    disabled: disabled
  })), !isPopUp && /* @__PURE__ */React__default["default"].createElement(reactTransitionGroup.CSSTransition, _objectSpread({
    "in": isOpen,
    appear: true,
    timeout: 300,
    nodeRef: contentRef
  }, transitionCallbacks), /* @__PURE__ */React__default["default"].createElement("ul", {
    key: "normal",
    style: _objectSpread(_objectSpread({}, childStyle), {}, {
      "--padding-left": "".concat(menuPaddingLeft, "px"),
      overflow: "hidden"
    }),
    className: classNames__default["default"]("".concat(classPrefix, "-menu__sub"), "".concat(classPrefix, "-slide-down-enter-active")),
    ref: contentRef
  }, popupChildren)));
  if (isPopUp) {
    return /* @__PURE__ */React__default["default"].createElement(popup_index.Popup, _objectSpread(_objectSpread({}, restPopupProps), {}, {
      overlayInnerClassName: ["".concat(classPrefix, "-menu__popup"), "".concat(classPrefix, "-is-vertical"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-opened"), isOpen), overlayInnerClassName],
      overlayClassName: ["".concat(classPrefix, "-menu--").concat(theme), defineProperty._defineProperty({}, "".concat(classPrefix, "-menu-is-nested"), level > 1), overlayClassName],
      placement: "right-top",
      content: pupContent,
      visible: open,
      onVisibleChange: handleVisibleChange
    }), submenu);
  }
  return submenu;
};
var SubTitleMenu = function SubTitleMenu(props) {
  var className = props.className,
    style = props.style,
    children = props.children,
    disabled = props.disabled,
    icon = props.icon,
    title = props.title,
    value = props.value,
    _props$level2 = props.level,
    level = _props$level2 === void 0 ? 1 : _props$level2,
    popupProps = props.popupProps;
  var _ref4 = popupProps || {},
    overlayClassName = _ref4.overlayClassName,
    overlayInnerClassName = _ref4.overlayInnerClassName,
    restPopupProps = objectWithoutProperties._objectWithoutProperties(_ref4, _excluded2);
  var _useControlled3 = hooks_useControlled["default"](popupProps, "visible", restPopupProps.onVisibleChange),
    _useControlled4 = slicedToArray._slicedToArray(_useControlled3, 2),
    open = _useControlled4[0],
    setOpen = _useControlled4[1];
  var _useContext2 = React.useContext(menu_MenuContext.MenuContext),
    active = _useContext2.active,
    onChange = _useContext2.onChange,
    expandType = _useContext2.expandType,
    _useContext2$theme = _useContext2.theme,
    theme = _useContext2$theme === void 0 ? "light" : _useContext2$theme;
  var _useConfig2 = hooks_useConfig["default"](),
    classPrefix = _useConfig2.classPrefix;
  var handleClick = function handleClick() {
    return onChange(value);
  };
  var handleVisibleChange = function handleVisibleChange(visible, ctx) {
    setOpen(visible, ctx);
  };
  var _useDomRefCallback = hooks_useDomRefCallback["default"](),
    _useDomRefCallback2 = slicedToArray._slicedToArray(_useDomRefCallback, 2),
    subMenuDom = _useDomRefCallback2[0],
    setRefCurrent = _useDomRefCallback2[1];
  hooks_useRipple["default"](subMenuDom);
  var isPopUp = expandType === "popup";
  var isActive = menu__util_checkSubMenuChildrenActive["default"](children, active) || active === value;
  var isOpen = React.useMemo(function () {
    if (disabled) return false;
    if (isPopUp) return open;
    return false;
  }, [disabled, isPopUp, open]);
  var fakeArrowStyle = level > 1 ? {
    transform: "rotate(-90deg)"
  } : {};
  var pupContent = /* @__PURE__ */React__default["default"].createElement("ul", {
    className: classNames__default["default"]("".concat(classPrefix, "-menu__popup-wrapper"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-opened"), isOpen))
  }, children);
  var placement = "right-top";
  if (level < 2) {
    placement = "bottom-left";
  }
  var submenu = /* @__PURE__ */React__default["default"].createElement("li", {
    className: classNames__default["default"]("".concat(classPrefix, "-submenu"), className, defineProperty._defineProperty({}, "".concat(classPrefix, "-is-opened"), open))
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    ref: setRefCurrent,
    className: classNames__default["default"]("".concat(classPrefix, "-menu__item"), defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-is-active"), isActive), "".concat(classPrefix, "-is-opened"), open)),
    onClick: handleClick,
    style: style
  }, _util_parseTNode["default"](icon), /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-menu__content")
  }, title), isPopUp && /* @__PURE__ */React__default["default"].createElement(common_FakeArrow["default"], {
    style: fakeArrowStyle,
    isActive: level === 1 && open
  })));
  if (isPopUp) {
    return /* @__PURE__ */React__default["default"].createElement(popup_index.Popup, _objectSpread(_objectSpread({}, restPopupProps), {}, {
      overlayInnerClassName: ["".concat(classPrefix, "-menu__popup"), "".concat(classPrefix, "-is-vertical"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-opened"), isOpen), overlayInnerClassName],
      overlayClassName: ["".concat(classPrefix, "-menu--").concat(theme), "".concat(classPrefix, "-is-head-menu"), defineProperty._defineProperty({}, "".concat(classPrefix, "-menu-is-nested"), level > 1), overlayClassName],
      placement: placement,
      content: pupContent,
      visible: open,
      onVisibleChange: handleVisibleChange
    }), submenu);
  }
  return submenu;
};
var SubMenu = function SubMenu(props) {
  var _useContext3 = React.useContext(menu_MenuContext.MenuContext),
    mode = _useContext3.mode;
  var children = props.children,
    _props$level3 = props.level,
    level = _props$level3 === void 0 ? 1 : _props$level3;
  var _changeItemLevel = function changeItemLevel(item) {
    if (menu__util_checkMenuType.checkIsSubMenu(item)) {
      return /*#__PURE__*/React__default["default"].cloneElement(item, {
        level: level + 1
      });
    }
    if (menu__util_checkMenuType.checkIsMenuGroup(item)) {
      var groupChildren = React__default["default"].Children.map(item.props.children, function (item2) {
        return _changeItemLevel(item2);
      });
      return /*#__PURE__*/React__default["default"].cloneElement(item, {
        level: level + 1
      }, groupChildren);
    }
    return item;
  };
  var childElement = React__default["default"].Children.map(children, function (item) {
    return _changeItemLevel(item);
  });
  if (mode === "accordion") return /* @__PURE__ */React__default["default"].createElement(SubAccordion, _objectSpread({}, props), childElement);
  if (mode === "title") return /* @__PURE__ */React__default["default"].createElement(SubTitleMenu, _objectSpread({}, props), childElement);
  return null;
};
SubMenu.displayName = "SubMenu";

exports["default"] = SubMenu;
