// tdesign-react/esm/select/util/helper.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import '../../_chunks/dep-1e9f924f.js';
import '../../_chunks/dep-5e72c8b9.js';
import 'lodash-es';
export { g as getKeyMapping, d as getLabel, e as getMultipleTags, b as getSelectValueArr, a as getSelectedOptions, c as getValueToOption, s as setValueToOptionFormOptionDom } from '../../_chunks/dep-29be6164.js';
import '../base/OptionGroup.js';
import '../../_chunks/dep-8e8ac0d4.js';
import '../../_chunks/dep-57e67945.js';
import '../../_chunks/dep-43f7da7b.js';
import '../../_chunks/dep-94a7af64.js';
import '../../_chunks/dep-0054b574.js';
import 'react';
import 'classnames';
import '../../hooks/useConfig.js';
import '../../config-provider/ConfigContext.js';
import '../../locale/zh_CN.js';
import '../../_chunks/dep-7d48ddac.js';
import 'dayjs';
import '../../_chunks/dep-766c4603.js';
import '../../hooks/useDomRefCallback.js';
import '../../hooks/useRipple.js';
import '../../_chunks/dep-2481dfad.js';
import '../../hooks/useAnimation.js';
import '../../_chunks/dep-a347b1b5.js';
import '../defaultProps.js';
import '../../hooks/useDefaultProps.js';
