// tdesign-react/cjs/table/hooks/useAffix.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _typeof = require('../../_chunks/dep-c4d1ce0c.js');
var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var React = require('react');
var _util_listener = require('../../_util/listener.js');
var _util_scroll = require('../../_util/scroll.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_chunks/dep-68f5fe7b.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-ecfe656e.js');
require('raf');
require('../../_util/easing.js');

function useAffix(props, _ref) {
  var showElement = _ref.showElement;
  var tableContentRef = React.useRef(null);
  var scrollContainersRef = React.useRef([]);
  var lastScrollLeftRef = React.useRef(0);
  var affixHeaderRef = React.useRef(null);
  var affixFooterRef = React.useRef(null);
  var horizontalScrollbarRef = React.useRef(null);
  var paginationRef = React.useRef(null);
  var _useState = React.useState(true),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    showAffixHeader = _useState2[0],
    setShowAffixHeader = _useState2[1];
  var _useState3 = React.useState(true),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    showAffixFooter = _useState4[0],
    setShowAffixFooter = _useState4[1];
  var _useState5 = React.useState(true),
    _useState6 = slicedToArray._slicedToArray(_useState5, 2),
    showAffixPagination = _useState6[0],
    setShowAffixPagination = _useState6[1];
  var isVirtualScroll = React.useMemo(function () {
    return props.scroll && props.scroll.type === "virtual" && (props.scroll.threshold || 100) < props.data.length;
  }, [props.data.length, props.scroll]);
  var isAffixed = React.useMemo(function () {
    return !!(props.headerAffixedTop || props.footerAffixedBottom || props.horizontalScrollAffixedBottom);
  }, [props.footerAffixedBottom, props.headerAffixedTop, props.horizontalScrollAffixedBottom]);
  var onHorizontalScroll = function onHorizontalScroll(scrollElement) {
    if (!isAffixed && !isVirtualScroll) return;
    var target = scrollElement;
    if (!target && tableContentRef.current) {
      lastScrollLeftRef.current = 0;
      target = tableContentRef.current;
    }
    if (!target) return;
    var left = target.scrollLeft;
    if (lastScrollLeftRef.current === left) return;
    lastScrollLeftRef.current = left;
    var toUpdateScrollElement = [tableContentRef.current, affixHeaderRef.current, affixFooterRef.current, horizontalScrollbarRef.current];
    for (var i = 0, len = toUpdateScrollElement.length; i < len; i++) {
      if (toUpdateScrollElement[i] && scrollElement !== toUpdateScrollElement[i]) {
        toUpdateScrollElement[i].scrollLeft = left;
      }
    }
  };
  var isAffixedBottomElementShow = React.useCallback(function (elementRect, tableRect, headerHeight, scrollContainer) {
    if (scrollContainer) {
      var containerRect = scrollContainer.getBoundingClientRect();
      var containerBottom = containerRect.bottom;
      var tableVisibleInContainer = tableRect.top + headerHeight < containerBottom;
      var tableBottomBelowContainer = tableRect.bottom > containerBottom;
      return tableVisibleInContainer && tableBottomBelowContainer;
    }
    return tableRect.top + headerHeight < elementRect.top && elementRect.top > elementRect.height;
  }, []);
  var getOffsetTop = function getOffsetTop(props2) {
    if (typeof props2 === "boolean") return 0;
    return props2.offsetTop || 0;
  };
  var updateAffixHeaderOrFooter = function updateAffixHeaderOrFooter() {
    var _tableContentRef$curr, _tableContentRef$curr2, _affixFooterRef$curre;
    if (!isAffixed && !isVirtualScroll) return;
    var pos = (_tableContentRef$curr = tableContentRef.current) === null || _tableContentRef$curr === void 0 ? void 0 : _tableContentRef$curr.getBoundingClientRect();
    if (!pos) return;
    var headerRect = (_tableContentRef$curr2 = tableContentRef.current) === null || _tableContentRef$curr2 === void 0 || (_tableContentRef$curr2 = _tableContentRef$curr2.querySelector("thead")) === null || _tableContentRef$curr2 === void 0 ? void 0 : _tableContentRef$curr2.getBoundingClientRect();
    var headerHeight = (headerRect === null || headerRect === void 0 ? void 0 : headerRect.height) || 0;
    var footerRect = (_affixFooterRef$curre = affixFooterRef.current) === null || _affixFooterRef$curre === void 0 ? void 0 : _affixFooterRef$curre.getBoundingClientRect();
    if ((props.headerAffixedTop || isVirtualScroll) && affixHeaderRef.current) {
      var offsetTop = getOffsetTop(props.headerAffixProps || props.headerAffixedTop);
      var footerHeight = (footerRect === null || footerRect === void 0 ? void 0 : footerRect.height) || 0;
      var r = Math.abs(pos.top) < pos.height - headerHeight - offsetTop - footerHeight;
      if (isVirtualScroll) {
        r = pos.top > -1 * headerRect.height;
      }
      setShowAffixHeader(r);
    }
    if (props.footerAffixedBottom && affixFooterRef.current) {
      var show = isAffixedBottomElementShow(footerRect, pos, headerHeight);
      setShowAffixFooter(show);
    } else if (props.horizontalScrollAffixedBottom && horizontalScrollbarRef.current) {
      var horizontalScrollbarRect = horizontalScrollbarRef.current.getBoundingClientRect();
      var _show = isAffixedBottomElementShow(horizontalScrollbarRect, pos, headerHeight);
      setShowAffixFooter(_show);
    }
    if (props.paginationAffixedBottom && paginationRef.current) {
      var pageRect = paginationRef.current.getBoundingClientRect();
      var _show2 = isAffixedBottomElementShow(pageRect, pos, headerHeight);
      setShowAffixPagination(_show2);
    }
  };
  var onDocumentScroll = function onDocumentScroll() {
    updateAffixHeaderOrFooter();
  };
  var onFootScroll = function onFootScroll() {
    onHorizontalScroll(affixFooterRef.current);
  };
  var onHeaderScroll = function onHeaderScroll() {
    onHorizontalScroll(affixHeaderRef.current);
  };
  var horizontalScrollbarScroll = function horizontalScrollbarScroll() {
    onHorizontalScroll(horizontalScrollbarRef.current);
  };
  var onTableContentScroll = function onTableContentScroll() {
    onHorizontalScroll(tableContentRef.current);
  };
  var onFootMouseEnter = function onFootMouseEnter() {
    _util_listener.on(affixFooterRef.current, "scroll", onFootScroll);
  };
  var onFootMouseLeave = function onFootMouseLeave() {
    _util_listener.off(affixFooterRef.current, "scroll", onFootScroll);
  };
  var onHeaderMouseEnter = function onHeaderMouseEnter() {
    _util_listener.on(affixHeaderRef.current, "scroll", onHeaderScroll);
  };
  var onHeaderMouseLeave = function onHeaderMouseLeave() {
    _util_listener.off(affixHeaderRef.current, "scroll", onHeaderScroll);
  };
  var onScrollbarMouseEnter = function onScrollbarMouseEnter() {
    _util_listener.on(horizontalScrollbarRef.current, "scroll", horizontalScrollbarScroll);
  };
  var onScrollbarMouseLeave = function onScrollbarMouseLeave() {
    _util_listener.off(horizontalScrollbarRef.current, "scroll", horizontalScrollbarScroll);
  };
  var onTableContentMouseEnter = function onTableContentMouseEnter() {
    _util_listener.on(tableContentRef.current, "scroll", onTableContentScroll);
  };
  var onTableContentMouseLeave = function onTableContentMouseLeave() {
    _util_listener.off(tableContentRef.current, "scroll", onTableContentScroll);
  };
  var addHorizontalScrollListeners = function addHorizontalScrollListeners() {
    if (affixHeaderRef.current) {
      _util_listener.on(affixHeaderRef.current, "mouseenter", onHeaderMouseEnter);
      _util_listener.on(affixHeaderRef.current, "mouseleave", onHeaderMouseLeave);
    }
    if (props.footerAffixedBottom && affixFooterRef.current) {
      _util_listener.on(affixFooterRef.current, "mouseenter", onFootMouseEnter);
      _util_listener.on(affixFooterRef.current, "mouseleave", onFootMouseLeave);
    }
    if (props.horizontalScrollAffixedBottom && horizontalScrollbarRef.current) {
      _util_listener.on(horizontalScrollbarRef.current, "mouseenter", onScrollbarMouseEnter);
      _util_listener.on(horizontalScrollbarRef.current, "mouseleave", onScrollbarMouseLeave);
    }
    if ((isAffixed || isVirtualScroll) && tableContentRef.current) {
      _util_listener.on(tableContentRef.current, "mouseenter", onTableContentMouseEnter);
      _util_listener.on(tableContentRef.current, "mouseleave", onTableContentMouseLeave);
    }
  };
  var removeHorizontalScrollListeners = function removeHorizontalScrollListeners() {
    if (affixHeaderRef.current) {
      _util_listener.off(affixHeaderRef.current, "mouseenter", onHeaderMouseEnter);
      _util_listener.off(affixHeaderRef.current, "mouseleave", onHeaderMouseLeave);
    }
    if (affixFooterRef.current) {
      _util_listener.off(affixFooterRef.current, "mouseenter", onFootMouseEnter);
      _util_listener.off(affixFooterRef.current, "mouseleave", onFootMouseLeave);
    }
    if (tableContentRef.current) {
      _util_listener.off(tableContentRef.current, "mouseenter", onTableContentMouseEnter);
      _util_listener.off(tableContentRef.current, "mouseleave", onTableContentMouseLeave);
    }
    if (horizontalScrollbarRef.current) {
      _util_listener.off(horizontalScrollbarRef.current, "mouseenter", onScrollbarMouseEnter);
      _util_listener.off(horizontalScrollbarRef.current, "mouseleave", onScrollbarMouseLeave);
    }
  };
  var getAffixContainers = React.useCallback(function () {
    var containers = [];
    var affixConfigs = [props.headerAffixedTop, props.footerAffixedBottom, props.horizontalScrollAffixedBottom];
    for (var i = 0; i < affixConfigs.length; i++) {
      var config = affixConfigs[i];
      if (_typeof._typeof(config) === "object" && config !== null && config !== void 0 && config.container) {
        var el = _util_scroll.getScrollContainer(config.container);
        if (el instanceof HTMLElement && !containers.includes(el)) {
          containers.push(el);
        }
      }
    }
    return containers;
  }, [props.headerAffixedTop, props.footerAffixedBottom, props.horizontalScrollAffixedBottom]);
  var addVerticalScrollListener = function addVerticalScrollListener() {
    if (typeof document === "undefined") return;
    if (!isAffixed && !props.paginationAffixedBottom) return;
    if (isAffixed || props.paginationAffixedBottom) {
      _util_listener.on(document, "scroll", onDocumentScroll);
      var containers = getAffixContainers();
      for (var i = 0; i < scrollContainersRef.current.length; i++) {
        _util_listener.off(scrollContainersRef.current[i], "scroll", onDocumentScroll);
      }
      scrollContainersRef.current = containers;
      for (var _i = 0; _i < containers.length; _i++) {
        _util_listener.on(containers[_i], "scroll", onDocumentScroll);
      }
      updateAffixHeaderOrFooter();
    } else {
      _util_listener.off(document, "scroll", onDocumentScroll);
      for (var _i2 = 0; _i2 < scrollContainersRef.current.length; _i2++) {
        _util_listener.off(scrollContainersRef.current[_i2], "scroll", onDocumentScroll);
      }
      scrollContainersRef.current = [];
    }
  };
  React.useEffect(function () {
    var timer = setTimeout(function () {
      addHorizontalScrollListeners();
      onHorizontalScroll();
      updateAffixHeaderOrFooter();
      clearTimeout(timer);
    });
    return removeHorizontalScrollListeners;
  }, [affixHeaderRef, affixFooterRef, horizontalScrollbarRef, tableContentRef, showElement]);
  React.useEffect(function () {
    addVerticalScrollListener();
    return function () {
      _util_listener.off(document, "scroll", onDocumentScroll);
      for (var i = 0; i < scrollContainersRef.current.length; i++) {
        _util_listener.off(scrollContainersRef.current[i], "scroll", onDocumentScroll);
      }
      scrollContainersRef.current = [];
    };
  }, [isAffixed]);
  React.useEffect(function () {
    addHorizontalScrollListeners();
    onHorizontalScroll();
  }, [props.data, props.columns, props.headerAffixedTop, props.footerAffixedBottom, props.horizontalScrollAffixedBottom, props.lazyLoad]);
  var setTableContentRef = function setTableContentRef(tableContent) {
    tableContentRef.current = tableContent;
    addVerticalScrollListener();
  };
  return {
    showAffixHeader: showAffixHeader,
    showAffixFooter: showAffixFooter,
    showAffixPagination: showAffixPagination,
    affixHeaderRef: affixHeaderRef,
    affixFooterRef: affixFooterRef,
    horizontalScrollbarRef: horizontalScrollbarRef,
    paginationRef: paginationRef,
    onHorizontalScroll: onHorizontalScroll,
    setTableContentRef: setTableContentRef,
    updateAffixHeaderOrFooter: updateAffixHeaderOrFooter
  };
}

exports["default"] = useAffix;
