// tdesign-react/cjs/table/hooks/useRowSelect.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var toConsumableArray = require('../../_chunks/dep-1d0dbfc5.js');
var defineProperty = require('../../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var React = require('react');
var utils = require('../../_chunks/dep-f92ec514.js');
var log = require('../../_chunks/dep-b289b0be.js');
var hooks_useControlled = require('../../hooks/useControlled.js');
var checkbox_index = require('../../checkbox/index.js');
var radio_index = require('../../radio/index.js');
var intersection = require('../../_chunks/dep-f94fba39.js');
var get = require('../../_chunks/dep-05df1ea3.js');
var isFunction = require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_chunks/dep-3c2d5e8c.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-32593c54.js');
require('../../_util/noop.js');
require('../../_chunks/dep-eb1b216c.js');
require('../../_chunks/dep-b529a0d7.js');
require('../../_chunks/dep-bcd16957.js');
require('../../_chunks/dep-b4442734.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-876d090e.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-83710d84.js');
require('../../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../../common/Check.js');
require('../../_chunks/dep-1c586f5c.js');
require('classnames');
require('../../_util/helper.js');
require('../../_chunks/dep-e44b03ec.js');
require('../../hooks/useConfig.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-14013769.js');
require('../../_chunks/dep-06bf4625.js');
require('dayjs');
require('../../_chunks/dep-043eee20.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../_chunks/dep-2d5cf7d6.js');
require('../../_chunks/dep-089e8106.js');
require('../../hooks/useDefaultProps.js');
require('../../checkbox/defaultProps.js');
require('../../_chunks/dep-f0e884b9.js');
require('../../_chunks/dep-27d626a1.js');
require('../../_chunks/dep-e31ce0e8.js');
require('../../hooks/useCommonClassName.js');
require('../../hooks/useDeepEffect.js');
require('../../_chunks/dep-df8d2fc8.js');
require('../../_chunks/dep-66f91e5d.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-ea361d31.js');
require('../../radio/defaultProps.js');
require('../../radio/useKeyboard.js');
require('../../_chunks/dep-dd0f5f6b.js');
require('../../_util/listener.js');
require('../../_chunks/dep-68f5fe7b.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../_chunks/dep-2a089253.js');
require('../../_chunks/dep-5e17a073.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var selectedRowDataMap = /* @__PURE__ */new Map();
function useRowSelect(props, tableSelectedClasses) {
  var selectedRowKeys = props.selectedRowKeys,
    columns = props.columns,
    data = props.data,
    rowKey = props.rowKey,
    indeterminateSelectedRowKeys = props.indeterminateSelectedRowKeys;
  var pagination = props.pagination,
    reserveSelectedRowOnPaginate = props.reserveSelectedRowOnPaginate;
  var _useState = React.useState(data),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    currentPaginateData = _useState2[0],
    setCurrentPaginateData = _useState2[1];
  var _useState3 = React.useState(),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    selectedRowClassNames = _useState4[0],
    setSelectedRowClassNames = _useState4[1];
  var _useControlled = hooks_useControlled["default"](props, "selectedRowKeys", props.onSelectChange, {
      defaultSelectedRowKeys: props.defaultSelectedRowKeys || []
    }),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    tSelectedRowKeys = _useControlled2[0],
    setTSelectedRowKeys = _useControlled2[1];
  var selectColumn = columns.find(function (_ref) {
    var type = _ref.type;
    return ["multiple", "single"].includes(type);
  });
  var canSelectedRows = React.useMemo(function () {
    var currentData = reserveSelectedRowOnPaginate ? data : currentPaginateData;
    return (currentData === null || currentData === void 0 ? void 0 : currentData.filter(function (row, rowIndex) {
      return !isDisabled(row, rowIndex);
    })) || [];
  }, [reserveSelectedRowOnPaginate, data, currentPaginateData]);
  var intersectionKeys = intersection.intersection(tSelectedRowKeys, canSelectedRows.map(function (t) {
    return get.get(t, rowKey || "id");
  }));
  React.useEffect(function () {
    if (reserveSelectedRowOnPaginate) return;
    var pageSize = pagination.pageSize,
      current = pagination.current,
      defaultPageSize = pagination.defaultPageSize,
      defaultCurrent = pagination.defaultCurrent;
    var tPageSize = pageSize || defaultPageSize;
    var tCurrent = current || defaultCurrent;
    var newData = data.slice(tPageSize * (tCurrent - 1), tPageSize * tCurrent);
    setCurrentPaginateData(newData);
  }, [data, reserveSelectedRowOnPaginate]);
  React.useEffect(function () {
    if (!selectColumn && (!tSelectedRowKeys || !tSelectedRowKeys.length)) return;
    var disabledRowFunc = function disabledRowFunc(p) {
      return selectColumn.disabled(p) ? tableSelectedClasses.disabled : "";
    };
    var disabledRowClass = selectColumn !== null && selectColumn !== void 0 && selectColumn.disabled ? disabledRowFunc : void 0;
    var selected = new Set(tSelectedRowKeys);
    var selectedRowClassFunc = function selectedRowClassFunc(_ref2) {
      var row = _ref2.row;
      var rowId = get.get(row, rowKey || "id");
      return selected.has(rowId) ? tableSelectedClasses.selected : "";
    };
    var selectedRowClass = selected.size ? selectedRowClassFunc : void 0;
    setSelectedRowClassNames([disabledRowClass, selectedRowClass]);
  }, [data, columns, tSelectedRowKeys, selectColumn, rowKey]);
  function isDisabled(row, rowIndex) {
    return utils.isRowSelectedDisabled(selectColumn, row, rowIndex);
  }
  function getSelectedHeader() {
    return function () {
      var isIndeterminate = intersectionKeys.length > 0 && intersectionKeys.length < canSelectedRows.length || intersectionKeys.length < tSelectedRowKeys.length;
      var isChecked = canSelectedRows.length !== 0 && intersectionKeys.length === canSelectedRows.length && intersectionKeys.length === tSelectedRowKeys.length;
      return /* @__PURE__ */React__default["default"].createElement(checkbox_index.Checkbox, {
        checked: isChecked,
        indeterminate: isIndeterminate,
        disabled: !canSelectedRows.length,
        onChange: handleSelectAll
      });
    };
  }
  function getRowSelectDisabledData(p) {
    var col = p.col,
      row = p.row,
      rowIndex = p.rowIndex;
    var disabled = typeof col.disabled === "function" ? col.disabled({
      row: row,
      rowIndex: rowIndex
    }) : col.disabled;
    var checkProps = isFunction.isFunction(col.checkProps) ? col.checkProps({
      row: row,
      rowIndex: rowIndex
    }) : col.checkProps;
    return {
      disabled: disabled || (checkProps === null || checkProps === void 0 ? void 0 : checkProps.disabled),
      checkProps: checkProps
    };
  }
  function renderSelectCell(p) {
    var column = p.col,
      _p$row = p.row,
      row = _p$row === void 0 ? {} : _p$row;
    var checked = tSelectedRowKeys.includes(get.get(row, rowKey || "id"));
    var _getRowSelectDisabled = getRowSelectDisabledData(p),
      disabled = _getRowSelectDisabled.disabled,
      checkProps = _getRowSelectDisabled.checkProps;
    var selectBoxProps = _objectSpread(_objectSpread({
      checked: checked,
      disabled: disabled
    }, checkProps), {}, {
      onChange: function onChange() {
        handleSelectChange(row);
      }
    });
    var onCheckClick = function onCheckClick(p2) {
      var e = "e" in p2 ? p2.e : p2;
      e === null || e === void 0 || e.stopPropagation();
    };
    if (column.type === "single") return /* @__PURE__ */React__default["default"].createElement(radio_index.Radio, _objectSpread(_objectSpread({}, selectBoxProps), {}, {
      onClick: onCheckClick
    }));
    if (column.type === "multiple") {
      var isIndeterminate = indeterminateSelectedRowKeys !== null && indeterminateSelectedRowKeys !== void 0 && indeterminateSelectedRowKeys.length ? indeterminateSelectedRowKeys.includes(get.get(row, rowKey)) : false;
      return /* @__PURE__ */React__default["default"].createElement(checkbox_index.Checkbox, _objectSpread(_objectSpread({
        indeterminate: isIndeterminate
      }, selectBoxProps), {}, {
        onClick: onCheckClick
      }));
    }
    return null;
  }
  var allowUncheck = React.useMemo(function () {
    var singleSelectCol = columns.find(function (col) {
      return col.type === "single";
    });
    if (!singleSelectCol || !singleSelectCol.checkProps || !("allowUncheck" in singleSelectCol.checkProps)) return false;
    return singleSelectCol.checkProps.allowUncheck;
  }, [columns]);
  function handleSelectChange() {
    var row = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var selectedRowKeys2 = toConsumableArray._toConsumableArray(tSelectedRowKeys);
    var reRowKey = rowKey || "id";
    var id = get.get(row, reRowKey);
    var selectedRowIndex = selectedRowKeys2.indexOf(id);
    var isExisted = selectedRowIndex !== -1;
    if (selectColumn.type === "multiple") {
      isExisted ? selectedRowKeys2.splice(selectedRowIndex, 1) : selectedRowKeys2.push(id);
    } else if (selectColumn.type === "single") {
      selectedRowKeys2 = isExisted && allowUncheck ? [] : [id];
    } else {
      log.log.warn("Table", "`column.type` must be one of `multiple` and `single`");
      return;
    }
    setTSelectedRowKeys(selectedRowKeys2, {
      selectedRowData: selectedRowKeys2.map(function (t) {
        return selectedRowDataMap.get(t);
      }),
      currentRowKey: id,
      currentRowData: row,
      type: isExisted ? "uncheck" : "check"
    });
  }
  function handleSelectAll(checked) {
    var reRowKey = rowKey || "id";
    var canSelectedRowKeys = canSelectedRows.map(function (record) {
      return get.get(record, reRowKey);
    });
    var disabledSelectedRowKeys = (selectedRowKeys === null || selectedRowKeys === void 0 ? void 0 : selectedRowKeys.filter(function (id) {
      return !canSelectedRowKeys.includes(id);
    })) || [];
    var allIds = checked ? [].concat(toConsumableArray._toConsumableArray(disabledSelectedRowKeys), toConsumableArray._toConsumableArray(canSelectedRowKeys)) : toConsumableArray._toConsumableArray(disabledSelectedRowKeys);
    setTSelectedRowKeys(allIds, {
      selectedRowData: checked ? allIds.map(function (t) {
        return selectedRowDataMap.get(t);
      }) : [],
      type: checked ? "check" : "uncheck",
      currentRowKey: "CHECK_ALL_BOX"
    });
  }
  function formatToRowSelectColumn(col) {
    var isSelection = ["multiple", "single"].includes(col.type);
    if (!isSelection) return col;
    return _objectSpread(_objectSpread({}, col), {}, {
      width: col.width || 64,
      className: tableSelectedClasses.checkCell,
      cell: function cell(p) {
        return renderSelectCell(p);
      },
      title: col.type === "multiple" ? getSelectedHeader() : col.title
    });
  }
  var onInnerSelectRowClick = function onInnerSelectRowClick(_ref3) {
    var row = _ref3.row,
      index = _ref3.index;
    var selectedColIndex = props.columns.findIndex(function (item) {
      return item.colKey === "row-select";
    });
    if (selectedColIndex === -1) return;
    var _getRowSelectDisabled2 = getRowSelectDisabledData({
        row: row,
        rowIndex: index,
        col: props.columns[selectedColIndex],
        colIndex: selectedColIndex
      }),
      disabled = _getRowSelectDisabled2.disabled;
    if (disabled) return;
    handleSelectChange(row);
  };
  React.useEffect(function () {
    for (var i = 0, len = data.length; i < len; i++) {
      selectedRowDataMap.set(get.get(data[i], rowKey || "id"), data[i]);
    }
  }, [data, rowKey]);
  return {
    selectedRowClassNames: selectedRowClassNames,
    currentPaginateData: currentPaginateData,
    setCurrentPaginateData: setCurrentPaginateData,
    setTSelectedRowKeys: setTSelectedRowKeys,
    formatToRowSelectColumn: formatToRowSelectColumn,
    onInnerSelectRowClick: onInnerSelectRowClick
  };
}

exports["default"] = useRowSelect;
