// tdesign-react/cjs/table/hooks/useRowspanAndColspan.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var React = require('react');
var log = require('../../_chunks/dep-b289b0be.js');
var hooks_useLayoutEffect = require('../../hooks/useLayoutEffect.js');
var get = require('../../_chunks/dep-05df1ea3.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_chunks/dep-68f5fe7b.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-5e17a073.js');
require('../../_chunks/dep-876d090e.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-b4442734.js');

function getCellKey(row, rowKey, colKey, colIndex) {
  var rowValue = get.get(row, rowKey);
  if (rowValue === void 0) {
    log.log.error("Table", "rowKey is wrong, can not get unique identifier of row.");
  }
  return [rowValue, colKey || colIndex].join("_");
}
function useRowspanAndColspan(data, columns, rowKey, rowspanAndColspan) {
  var _useState = React.useState(function () {
      return /* @__PURE__ */new Map();
    }),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    skipSpansMap = _useState2[0],
    setKipSnapsMap = _useState2[1];
  var onTrRowspanOrColspan = function onTrRowspanOrColspan(params, skipSpansValue, map) {
    var rowIndex = params.rowIndex,
      colIndex = params.colIndex;
    if (!skipSpansValue.rowspan && !skipSpansValue.colspan) return;
    var maxRowIndex = rowIndex + (skipSpansValue.rowspan || 1);
    var maxColIndex = colIndex + (skipSpansValue.colspan || 1);
    for (var i = rowIndex; i < maxRowIndex; i++) {
      for (var j = colIndex; j < maxColIndex; j++) {
        if (i !== rowIndex || j !== colIndex) {
          if (!data[i] || !columns[j]) return;
          var cellKey = getCellKey(data[i], rowKey, columns[j].colKey, j);
          var state = map.get(cellKey) || {};
          state.skipped = true;
          map.set(cellKey, state);
        }
      }
    }
  };
  var getSkipSpansMap = function getSkipSpansMap(data2, columns2, rowspanAndColspan2) {
    if (!data2 || !rowspanAndColspan2) return;
    var map = /* @__PURE__ */new Map();
    for (var i = 0, len = data2.length; i < len; i++) {
      var row = data2[i];
      for (var j = 0, colLen = columns2.length; j < colLen; j++) {
        var col = columns2[j];
        var params = {
          row: row,
          col: col,
          rowIndex: i,
          colIndex: j
        };
        var cellKey = getCellKey(row, rowKey, col.colKey, j);
        var state = map.get(cellKey) || {};
        var o = rowspanAndColspan2(params) || {};
        if (o.rowspan || o.colspan || state.rowspan || state.colspan) {
          o.rowspan && (state.rowspan = o.rowspan);
          o.colspan && (state.colspan = o.colspan);
          map.set(cellKey, state);
        }
        onTrRowspanOrColspan === null || onTrRowspanOrColspan === void 0 || onTrRowspanOrColspan(params, state, map);
      }
    }
    return map;
  };
  hooks_useLayoutEffect["default"](function () {
    if (!rowspanAndColspan) return;
    skipSpansMap.clear();
    var result = getSkipSpansMap(data, columns, rowspanAndColspan);
    setKipSnapsMap(result);
  }, [data, columns, rowspanAndColspan]);
  return {
    skipSpansMap: skipSpansMap,
    getSkipSpansMap: getSkipSpansMap
  };
}

exports["default"] = useRowspanAndColspan;
exports.getCellKey = getCellKey;
