// tdesign-react/cjs/table/hooks/useTreeData.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-2b5ba78d.js');
var toConsumableArray = require('../../_chunks/dep-1d0dbfc5.js');
var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var table_hooks_useTreeDataExpand = require('../../_chunks/dep-e87dbebb.js');
var tdesignIconsReact = require('tdesign-icons-react');
var _util_parseTNode = require('../../_util/parseTNode.js');
var hooks_useGlobalIcon = require('../../hooks/useGlobalIcon.js');
var locale_LocalReceiver = require('../../locale/LocalReceiver.js');
var table_Cell = require('../Cell.js');
var table_hooks_useClassName = require('./useClassName.js');
var get = require('../../_chunks/dep-05df1ea3.js');
var isEqual = require('../../_chunks/dep-4e876c4b.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_chunks/dep-3c2d5e8c.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-427fb3a1.js');
require('../../_chunks/dep-f92ec514.js');
require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-b289b0be.js');
require('../../_chunks/dep-8f62fb96.js');
require('../../_chunks/dep-57a4aad8.js');
require('../../_chunks/dep-ca918872.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-5e17a073.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-876d090e.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-b4442734.js');
require('../../_chunks/dep-6abdd014.js');
require('../../hooks/useControlled.js');
require('../../_util/noop.js');
require('../../_chunks/dep-eb1b216c.js');
require('../../_chunks/dep-b529a0d7.js');
require('../../_chunks/dep-bcd16957.js');
require('../../hooks/usePrevious.js');
require('../../hooks/useConfig.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-14013769.js');
require('../../_chunks/dep-06bf4625.js');
require('dayjs');
require('../../_chunks/dep-043eee20.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../_chunks/dep-2d5cf7d6.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../config-provider/ConfigProvider.js');
require('../../_chunks/dep-4a562e04.js');
require('../../_chunks/dep-a5b4810f.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-ea361d31.js');
require('../Ellipsis.js');
require('../../_util/scroll.js');
require('raf');
require('../../_chunks/dep-68f5fe7b.js');
require('../../_util/easing.js');
require('../../tooltip/index.js');
require('../../tooltip/Tooltip.js');
require('../../_chunks/dep-1c586f5c.js');
require('../../popup/index.js');
require('../../popup/Popup.js');
require('react-transition-group');
require('../../common/Portal.js');
require('react-dom');
require('../../hooks/useLayoutEffect.js');
require('../../hooks/useAnimation.js');
require('../../hooks/useAttach.js');
require('../../hooks/useDefaultProps.js');
require('../../hooks/useMutationObserver.js');
require('../../hooks/useLatest.js');
require('../../_chunks/dep-d3af8eb7.js');
require('../../_chunks/dep-eab994e8.js');
require('../../_chunks/dep-df8d2fc8.js');
require('../../_chunks/dep-66f91e5d.js');
require('../../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../../hooks/useWindowSize.js');
require('../../popup/defaultProps.js');
require('../../popup/hooks/useTrigger.js');
require('../../_util/listener.js');
require('../../_util/ref.js');
require('react-is');
require('../../_util/isFragment.js');
require('../../popup/utils/transition.js');
require('../../popup/PopupPlugin.js');
require('../../_chunks/dep-e37cb830.js');
require('../../_util/react-render.js');
require('../../common/PluginContainer.js');
require('../../tooltip/defaultProps.js');
require('../../tooltip/TooltipLite.js');
require('../../hooks/useSwitch.js');
require('../../hooks/usePersistFn.js');
require('../../_chunks/dep-2a85392e.js');
require('../../hooks/useDebounce.js');
require('./useFixed.js');
require('../../_chunks/dep-fb673f7a.js');
require('../../_chunks/dep-b011b5d3.js');
require('../../_chunks/dep-f0e884b9.js');
require('../../hooks/useDeepEffect.js');
require('../utils.js');
require('../../_chunks/dep-e43e6728.js');
require('../../_chunks/dep-4ef3269f.js');
require('../../_chunks/dep-2a089253.js');
require('../../_chunks/dep-048e4314.js');
require('../../_chunks/dep-19400123.js');
require('../../_chunks/dep-894a5ece.js');
require('../../_chunks/dep-f7bf3d17.js');
require('../../_chunks/dep-13de2fef.js');
require('../../_chunks/dep-63308b84.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function useTreeData(props) {
  var _props$tree;
  var data = props.data,
    columns = props.columns,
    tree = props.tree,
    rowKey = props.rowKey,
    treeExpandAndFoldIcon = props.treeExpandAndFoldIcon,
    expandedTreeNodes = props.expandedTreeNodes;
  var _useClassName = table_hooks_useClassName["default"](),
    tableTreeClasses = _useClassName.tableTreeClasses;
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("table"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    locale = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      AddRectangleIcon: tdesignIconsReact.AddRectangleIcon,
      MinusRectangleIcon: tdesignIconsReact.MinusRectangleIcon
    }),
    AddRectangleIcon = _useGlobalIcon.AddRectangleIcon,
    MinusRectangleIcon = _useGlobalIcon.MinusRectangleIcon;
  var dataRef = React.useRef(data);
  var isManuallyModified = React.useRef(false);
  var _useState = React.useState(function () {
      return new table_hooks_useTreeDataExpand.TableTreeStore();
    }),
    _useState2 = slicedToArray._slicedToArray(_useState, 1),
    store = _useState2[0];
  var _useState3 = React.useState(function () {
      return getTreeNodeColumnCol();
    }),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    treeNodeCol = _useState4[0],
    setTreeNodeCol = _useState4[1];
  var _useState5 = React.useState(data || []),
    _useState6 = slicedToArray._slicedToArray(_useState5, 2),
    dataSource = _useState6[0],
    setDataSource = _useState6[1];
  var updateDataSourceManually = function updateDataSourceManually(newDataSource) {
    setDataSource(toConsumableArray._toConsumableArray(newDataSource));
    isManuallyModified.current = true;
  };
  var rowDataKeys = React.useMemo(function () {
    return {
      rowKey: rowKey || "id",
      childrenKey: (tree === null || tree === void 0 ? void 0 : tree.childrenKey) || "children"
    };
  }, [rowKey, tree === null || tree === void 0 ? void 0 : tree.childrenKey]);
  var _useTreeDataExpand = table_hooks_useTreeDataExpand.useTreeDataExpand(props, {
      store: store,
      dataSource: dataSource,
      rowDataKeys: rowDataKeys,
      setDataSource: setDataSource
    }),
    expandAll = _useTreeDataExpand.expandAll,
    foldAll = _useTreeDataExpand.foldAll,
    updateExpandOnDataChange = _useTreeDataExpand.updateExpandOnDataChange,
    onExpandFoldIconClick = _useTreeDataExpand.onExpandFoldIconClick;
  var checkedColumn = React.useMemo(function () {
    return columns.find(function (col) {
      return col.colKey === "row-select";
    });
  }, [columns]);
  React.useEffect(function () {
    if (!store || !checkedColumn) return;
    var rowValue = get.get(dataSource[0], rowDataKeys.rowKey);
    if (!store.treeDataMap.get(rowValue)) return;
    store.updateDisabledState(dataSource, checkedColumn, rowDataKeys);
  }, [checkedColumn]);
  React.useEffect(function () {
    if (!data || !store) return;
    dataRef.current = data;
    if (!tree) {
      setDataSource(data);
      return;
    }
    var dataChanged = !isEqual.isEqual(dataRef.current, data);
    if (dataChanged || !isManuallyModified.current) {
      resetData(data);
      if (dataChanged) {
        isManuallyModified.current = false;
      }
    } else {
      updateExpandOnDataChange(toConsumableArray._toConsumableArray(dataSource));
      isManuallyModified.current = false;
    }
  }, [data, expandedTreeNodes]);
  React.useEffect(function () {
    var treeNodeColTmp = getTreeNodeColumnCol();
    setTreeNodeCol(treeNodeColTmp);
  }, [columns, (_props$tree = props.tree) === null || _props$tree === void 0 ? void 0 : _props$tree.treeNodeColumnIndex]);
  function resetData(data2) {
    var columns2 = props.columns;
    store.initialTreeStore(data2, columns2, rowDataKeys);
    updateExpandOnDataChange(toConsumableArray._toConsumableArray(data2));
  }
  function getTreeNodeStyle(level) {
    var _props$tree2, _props$tree3;
    if (level === void 0) return;
    var indent = ((_props$tree2 = props.tree) === null || _props$tree2 === void 0 ? void 0 : _props$tree2.indent) === void 0 ? 24 : (_props$tree3 = props.tree) === null || _props$tree3 === void 0 ? void 0 : _props$tree3.indent;
    return indent ? {
      paddingLeft: "".concat(level * indent || 1, "px")
    } : {};
  }
  function toggleExpandData(p, trigger) {
    var _props$onTreeExpandCh, _store$treeDataMap;
    var currentData = _objectSpread({}, p);
    if (p.row.__VIRTUAL_SCROLL_INDEX !== void 0) {
      currentData.rowIndex = p.row.__VIRTUAL_SCROLL_INDEX;
    }
    var newData = store.toggleExpandData(currentData, dataSource, rowDataKeys);
    setDataSource(toConsumableArray._toConsumableArray(newData));
    var rowValue = get.get(p.row, rowDataKeys.rowKey);
    (_props$onTreeExpandCh = props.onTreeExpandChange) === null || _props$onTreeExpandCh === void 0 || _props$onTreeExpandCh.call(props, {
      row: p.row,
      rowIndex: p.rowIndex,
      rowState: store === null || store === void 0 || (_store$treeDataMap = store.treeDataMap) === null || _store$treeDataMap === void 0 ? void 0 : _store$treeDataMap.get(rowValue),
      trigger: trigger
    });
  }
  function getTreeNodeColumnCol() {
    var _props$tree4;
    var columns2 = props.columns;
    var treeNodeColumnIndex = ((_props$tree4 = props.tree) === null || _props$tree4 === void 0 ? void 0 : _props$tree4.treeNodeColumnIndex) || 0;
    while ((_columns2$treeNodeCol = columns2[treeNodeColumnIndex]) !== null && _columns2$treeNodeCol !== void 0 && _columns2$treeNodeCol.type || ((_columns2$treeNodeCol2 = columns2[treeNodeColumnIndex]) === null || _columns2$treeNodeCol2 === void 0 ? void 0 : _columns2$treeNodeCol2.colKey) === "__EXPAND_ROW_ICON_COLUMN__") {
      var _columns2$treeNodeCol, _columns2$treeNodeCol2;
      treeNodeColumnIndex += 1;
    }
    return columns2[treeNodeColumnIndex];
  }
  function formatTreeColumn(col) {
    if (!col || !treeNodeCol || !store) return {};
    if (!props.tree || !Object.keys(props.tree).length || col.colKey !== treeNodeCol.colKey) return col;
    var newCol = _objectSpread({}, treeNodeCol);
    newCol.cell = function (p) {
      var cellInfo = table_Cell.renderCell(_objectSpread(_objectSpread({}, p), {}, {
        col: _objectSpread({}, treeNodeCol)
      }), {
        cellEmptyContent: props.cellEmptyContent
      });
      var currentState = store.treeDataMap.get(get.get(p.row, rowDataKeys.rowKey));
      var colStyle = getTreeNodeStyle(currentState === null || currentState === void 0 ? void 0 : currentState.level);
      var classes = defineProperty._defineProperty({}, tableTreeClasses.inlineCol, !!col.ellipsis);
      var childrenNodes = get.get(p.row, rowDataKeys.childrenKey);
      if (Array.isArray(childrenNodes) && childrenNodes.length > 0 || childrenNodes === true) {
        var _store$treeDataMap$ge;
        var expanded = (_store$treeDataMap$ge = store.treeDataMap.get(get.get(p.row, rowDataKeys.rowKey))) === null || _store$treeDataMap$ge === void 0 ? void 0 : _store$treeDataMap$ge.expanded;
        var type = expanded ? "fold" : "expand";
        var defaultIconNode = t(locale.treeExpandAndFoldIcon, {
          type: type
        }) || (expanded ? /* @__PURE__ */React__default["default"].createElement(MinusRectangleIcon, null) : /* @__PURE__ */React__default["default"].createElement(AddRectangleIcon, null));
        var iconNode = treeExpandAndFoldIcon ? _util_parseTNode.parseContentTNode(treeExpandAndFoldIcon, _objectSpread({
          type: type
        }, p)) : defaultIconNode;
        return /* @__PURE__ */React__default["default"].createElement("div", {
          className: classNames__default["default"]([tableTreeClasses.col, classes]),
          style: colStyle
        }, !!(childrenNodes.length || childrenNodes === true) && /* @__PURE__ */React__default["default"].createElement("span", {
          className: tableTreeClasses.icon,
          onClick: function onClick(e) {
            onExpandFoldIconClick(p, "expand-fold-icon");
            e.stopPropagation();
          }
        }, iconNode), cellInfo);
      }
      return /* @__PURE__ */React__default["default"].createElement("div", {
        style: colStyle,
        className: classNames__default["default"]([classes, tableTreeClasses.leafNode])
      }, /* @__PURE__ */React__default["default"].createElement("span", {
        className: tableTreeClasses.icon
      }), cellInfo);
    };
    if (newCol.ellipsis === true) {
      newCol.ellipsis = function (p) {
        return table_Cell.renderCell(_objectSpread(_objectSpread({}, p), {}, {
          col: _objectSpread({}, treeNodeCol)
        }));
      };
    }
    return newCol;
  }
  function setData(key, newRowData) {
    var rowIndex = store.updateData(key, newRowData, dataSource, rowDataKeys);
    var newData = toConsumableArray._toConsumableArray(dataSource);
    newData[rowIndex] = newRowData;
    updateDataSourceManually(newData);
  }
  function getData(key) {
    return store.getData(key);
  }
  function remove(key) {
    var newData = store.remove(key, dataSource, rowDataKeys);
    updateDataSourceManually(newData);
  }
  function removeChildren(key) {
    var newData = store.removeChildren(key, dataSource, rowDataKeys);
    updateDataSourceManually(newData);
  }
  function appendTo(key, newData) {
    if (!key) {
      updateDataSourceManually(toConsumableArray._toConsumableArray(store.appendToRoot(newData, dataSource, rowDataKeys)));
      return;
    }
    updateDataSourceManually(toConsumableArray._toConsumableArray(store.appendTo(key, newData, dataSource, rowDataKeys)));
  }
  function insertAfter(rowValue, newData) {
    updateDataSourceManually(toConsumableArray._toConsumableArray(store.insertAfter(rowValue, newData, dataSource, rowDataKeys)));
  }
  function insertBefore(rowValue, newData) {
    updateDataSourceManually(store.insertBefore(rowValue, newData, dataSource, rowDataKeys));
  }
  function swapData(params) {
    var r = store.swapData(params.data, params, rowDataKeys);
    if (r.result) {
      updateDataSourceManually(r.dataSource);
    } else {
      var _props$onAbnormalDrag;
      var params2 = {
        code: r.code,
        reason: r.reason
      };
      (_props$onAbnormalDrag = props.onAbnormalDragSort) === null || _props$onAbnormalDrag === void 0 || _props$onAbnormalDrag.call(props, params2);
    }
  }
  function getTreeNode() {
    return store.getTreeNode(dataSource, rowDataKeys);
  }
  function getTreeExpandedRow() {
    var type = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "data";
    return store.getTreeExpandedRow(dataSource, rowDataKeys, type);
  }
  return {
    store: store,
    rowDataKeys: rowDataKeys,
    dataSource: dataSource,
    swapData: swapData,
    setData: setData,
    getData: getData,
    remove: remove,
    removeChildren: removeChildren,
    appendTo: appendTo,
    insertAfter: insertAfter,
    insertBefore: insertBefore,
    formatTreeColumn: formatTreeColumn,
    toggleExpandData: toggleExpandData,
    expandAll: expandAll,
    foldAll: foldAll,
    getTreeNode: getTreeNode,
    resetData: resetData,
    getTreeExpandedRow: getTreeExpandedRow,
    onExpandFoldIconClick: onExpandFoldIconClick
  };
}

exports["default"] = useTreeData;
