// tdesign-react/cjs/switch/Switch.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var log = require('../_chunks/dep-b289b0be.js');
var loading_index = require('../loading/index.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useCommonClassName = require('../hooks/useCommonClassName.js');
var switch_defaultProps = require('./defaultProps.js');
var _util_parseTNode = require('../_util/parseTNode.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../loading/Loading.js');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-ecfe656e.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-f0e884b9.js');
require('../hooks/useDomRefCallback.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["className", "value", "defaultValue", "disabled", "loading", "size", "label", "customValue", "onChange", "beforeChange"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Switch = /*#__PURE__*/React__default["default"].forwardRef(function (originalProps, ref) {
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var props = hooks_useDefaultProps["default"](originalProps, switch_defaultProps.switchDefaultProps);
  var className = props.className,
    value = props.value,
    defaultValue = props.defaultValue,
    disabled = props.disabled,
    loading = props.loading,
    size = props.size,
    label = props.label,
    customValue = props.customValue,
    onChange = props.onChange,
    beforeChange = props.beforeChange,
    restProps = objectWithoutProperties._objectWithoutProperties(props, _excluded);
  var _ref = customValue || [],
    _ref2 = slicedToArray._slicedToArray(_ref, 2),
    _ref2$ = _ref2[0],
    activeValue = _ref2$ === void 0 ? true : _ref2$,
    _ref2$2 = _ref2[1],
    inactiveValue = _ref2$2 === void 0 ? false : _ref2$2;
  var isControlled = typeof value !== "undefined";
  var initChecked = defaultValue === activeValue || value === activeValue;
  var _useState = React.useState(initChecked),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    innerChecked = _useState2[0],
    setInnerChecked = _useState2[1];
  var contentNode = React__default["default"].useMemo(function () {
    if (Array.isArray(label)) {
      var _label = slicedToArray._slicedToArray(label, 2),
        _label$ = _label[0],
        activeContent = _label$ === void 0 ? "" : _label$,
        _label$2 = _label[1],
        inactiveContent = _label$2 === void 0 ? "" : _label$2;
      var content = innerChecked ? activeContent : inactiveContent;
      return _util_parseTNode["default"](content, {
        value: value
      });
    }
    return _util_parseTNode["default"](label, {
      value: value
    });
  }, [label, innerChecked, value]);
  var handleChange = function handleChange(e) {
    !isControlled && setInnerChecked(!innerChecked);
    var changedValue = !innerChecked ? activeValue : inactiveValue;
    onChange === null || onChange === void 0 || onChange(changedValue, {
      e: e
    });
  };
  var onInternalClick = function onInternalClick(e) {
    if (disabled) {
      return;
    }
    if (!beforeChange) {
      handleChange(e);
      return;
    }
    Promise.resolve(beforeChange()).then(function (v) {
      if (v) {
        handleChange(e);
      }
    })["catch"](function (e2) {
      log.log.error("Switch", "some error occurred: ".concat(e2));
    });
  };
  React.useEffect(function () {
    if (Array.isArray(customValue) && !customValue.includes(value)) {
      log.log.error("Switch", "value is not in customValue: ".concat(JSON.stringify(customValue)));
    }
    isControlled && setInnerChecked(value === activeValue);
  }, [value, customValue, activeValue, isControlled]);
  var _useCommonClassName = hooks_useCommonClassName["default"](),
    SIZE = _useCommonClassName.SIZE,
    STATUS = _useCommonClassName.STATUS;
  var switchClassName = classNames__default["default"]("".concat(classPrefix, "-switch"), className, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, STATUS.checked, innerChecked), STATUS.disabled, disabled), STATUS.loading, loading), SIZE[size]);
  return /* @__PURE__ */React__default["default"].createElement("button", _objectSpread(_objectSpread({}, restProps), {}, {
    type: "button",
    role: "switch",
    disabled: disabled || loading,
    className: switchClassName,
    ref: ref,
    onClick: onInternalClick
  }), /* @__PURE__ */React__default["default"].createElement("span", {
    className: "".concat(classPrefix, "-switch__handle")
  }, loading && /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, {
    loading: true,
    size: "small"
  })), /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-switch__content")
  }, contentNode));
});
Switch.displayName = "Switch";

exports["default"] = Switch;
