// tdesign-react/cjs/descriptions/Descriptions.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var descriptions_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useCommonClassName = require('../hooks/useCommonClassName.js');
var descriptions_DescriptionsContext = require('./DescriptionsContext.js');
var descriptions_DescriptionsItem = require('./DescriptionsItem.js');
var descriptions_Row = require('./Row.js');
var isArray = require('../_chunks/dep-794b42ce.js');
var _assignValue = require('../_chunks/dep-b21b548e.js');
var _initCloneObject = require('../_chunks/dep-83fd21a7.js');
var merge = require('../_chunks/dep-043eee20.js');
var isArrayLike = require('../_chunks/dep-0bed591e.js');
var _overArg = require('../_chunks/dep-cd29b2b2.js');
var _getAllKeys = require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../locale/LocalReceiver.js');
require('../_chunks/dep-17e2e72f.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

/** Used for built-in method references. */
var objectProto = Object.prototype;

/** Used to check objects for own properties. */
var hasOwnProperty = objectProto.hasOwnProperty;

/**
 * Assigns own enumerable string keyed properties of source objects to the
 * destination object. Source objects are applied from left to right.
 * Subsequent sources overwrite property assignments of previous sources.
 *
 * **Note:** This method mutates `object` and is loosely based on
 * [`Object.assign`](https://mdn.io/Object/assign).
 *
 * @static
 * @memberOf _
 * @since 0.10.0
 * @category Object
 * @param {Object} object The destination object.
 * @param {...Object} [sources] The source objects.
 * @returns {Object} Returns `object`.
 * @see _.assignIn
 * @example
 *
 * function Foo() {
 *   this.a = 1;
 * }
 *
 * function Bar() {
 *   this.c = 3;
 * }
 *
 * Foo.prototype.b = 2;
 * Bar.prototype.d = 4;
 *
 * _.assign({ 'a': 0 }, new Foo, new Bar);
 * // => { 'a': 1, 'c': 3 }
 */
var assign = merge.createAssigner(function (object, source) {
  if (_overArg.isPrototype(source) || isArrayLike.isArrayLike(source)) {
    _initCloneObject.copyObject(source, _getAllKeys.keys(source), object);
    return;
  }
  for (var key in source) {
    if (hasOwnProperty.call(source, key)) {
      _assignValue.assignValue(object, key, source[key]);
    }
  }
});
var assign$1 = assign;

var Descriptions = function Descriptions(DescriptionsProps) {
  var props = hooks_useDefaultProps["default"](DescriptionsProps, descriptions_defaultProps.descriptionsDefaultProps);
  var className = props.className,
    style = props.style,
    title = props.title,
    bordered = props.bordered,
    column = props.column,
    layout = props.layout,
    rowItems = props.items,
    children = props.children,
    tableLayout = props.tableLayout;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var COMPONENT_NAME = "".concat(classPrefix, "-descriptions");
  var _useCommonClassName = hooks_useCommonClassName["default"](),
    SIZE = _useCommonClassName.SIZE;
  var getRows = function getRows() {
    var items = [];
    if (isArray.isArray(rowItems)) {
      items = rowItems.map(function (item) {
        var _assign = assign$1({}, descriptions_defaultProps.descriptionsItemDefaultProps, item),
          span = _assign.span;
        return {
          label: item.label,
          content: item.content,
          span: span
        };
      });
    } else {
      var childrenList = React__default["default"].Children.toArray(children).filter(function (child) {
        var _child$type;
        return ((_child$type = child.type) === null || _child$type === void 0 ? void 0 : _child$type.displayName) === descriptions_DescriptionsItem["default"].displayName;
      });
      if (childrenList.length !== 0) {
        items = childrenList.map(function (_ref) {
          var _child$content;
          var child = _ref.props;
          var _assign2 = assign$1({}, descriptions_defaultProps.descriptionsItemDefaultProps, child),
            span = _assign2.span;
          return {
            label: child.label,
            content: (_child$content = child.content) !== null && _child$content !== void 0 ? _child$content : child.children,
            span: span
          };
        });
      }
    }
    if (layout === "vertical") {
      return [items];
    }
    var temp = [];
    var reset = column;
    var res = [];
    items.forEach(function (item, index) {
      var span = item.span;
      if (reset >= span) {
        temp.push(item);
        reset -= span;
      } else {
        res.push(temp);
        temp = [item];
        reset = column - span;
      }
      if (index === items.length - 1) {
        Reflect.set(item, "span", span + reset);
        res.push(temp);
      }
    });
    return res;
  };
  var renderHeader = function renderHeader() {
    return title ? /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(COMPONENT_NAME, "__header")
    }, title) : "";
  };
  var renderBody = function renderBody() {
    var tableClass = ["".concat(COMPONENT_NAME, "__body"), SIZE[props.size], defineProperty._defineProperty({}, "".concat(COMPONENT_NAME, "__body--fixed"), tableLayout === "fixed"), defineProperty._defineProperty({}, "".concat(COMPONENT_NAME, "__body--border"), bordered)];
    return /* @__PURE__ */React__default["default"].createElement("table", {
      className: classNames__default["default"](tableClass)
    }, /* @__PURE__ */React__default["default"].createElement("tbody", null, getRows().map(function (row, i) {
      return /* @__PURE__ */React__default["default"].createElement(descriptions_Row["default"], {
        row: row,
        key: i
      });
    })));
  };
  return /* @__PURE__ */React__default["default"].createElement(descriptions_DescriptionsContext.DescriptionsContext.Provider, {
    value: props
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](className, COMPONENT_NAME),
    style: style
  }, renderHeader(), renderBody()));
};
Descriptions.displayName = "Descriptions";
Descriptions.DescriptionsItem = descriptions_DescriptionsItem["default"];

exports["default"] = Descriptions;
