// tdesign-react/cjs/descriptions/Row.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var hooks_useConfig = require('../hooks/useConfig.js');
var descriptions_DescriptionsContext = require('./DescriptionsContext.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-2b5ba78d.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-17e2e72f.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var Row = function Row(props) {
  var row = props.row;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("descriptions"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    locale = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var descriptionsContext = React.useContext(descriptions_DescriptionsContext.DescriptionsContext);
  var COMPONENT_NAME = "".concat(classPrefix, "-descriptions");
  var label = function label(node) {
    var layout = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "horizontal";
    var rowKey = arguments.length > 2 ? arguments[2] : undefined;
    var span = node.span;
    var labelSpan = layout === "horizontal" ? 1 : span;
    return /* @__PURE__ */React__default["default"].createElement("td", {
      key: rowKey,
      colSpan: labelSpan,
      className: "".concat(COMPONENT_NAME, "__label"),
      style: descriptionsContext.labelStyle
    }, node.label, descriptionsContext.colon && t(locale.colonText));
  };
  var content = function content(node) {
    var layout = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "horizontal";
    var rowKey = arguments.length > 2 ? arguments[2] : undefined;
    var span = node.span;
    var contentSpan = span > 1 && layout === "horizontal" ? span * 2 - 1 : span;
    return /* @__PURE__ */React__default["default"].createElement("td", {
      key: rowKey,
      colSpan: contentSpan,
      className: "".concat(COMPONENT_NAME, "__content"),
      style: descriptionsContext.contentStyle
    }, node.content);
  };
  var hh = function hh() {
    return /* @__PURE__ */React__default["default"].createElement("tr", null, row.map(function (node, i) {
      return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, {
        key: i
      }, label(node), content(node));
    }));
  };
  var hv = function hv() {
    return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement("tr", null, row.map(function (node, i) {
      return label(node, "vertical", "top_".concat(i));
    })), /* @__PURE__ */React__default["default"].createElement("tr", null, row.map(function (node, i) {
      return content(node, "vertical", "bottom_".concat(i));
    })));
  };
  var vh = function vh() {
    return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, row.map(function (node, i) {
      return /* @__PURE__ */React__default["default"].createElement("tr", {
        key: i
      }, label(node), content(node));
    }));
  };
  var vv = function vv() {
    return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, row.map(function (node, i) {
      return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, {
        key: i
      }, /* @__PURE__ */React__default["default"].createElement("tr", null, label(node)), /* @__PURE__ */React__default["default"].createElement("tr", null, content(node)));
    }));
  };
  if (descriptionsContext.layout === "horizontal") {
    if (descriptionsContext.itemLayout === "horizontal") {
      return hh();
    }
    return hv();
  }
  if (descriptionsContext.itemLayout === "horizontal") {
    return vh();
  }
  return vv();
};
Row.displayName = "DescriptionsRow";

exports["default"] = Row;
