// tdesign-react/esm/form/const.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

var ValidateStatus = /* @__PURE__ */function (ValidateStatus2) {
  ValidateStatus2["SUCCESS"] = "success";
  ValidateStatus2["WARNING"] = "warning";
  ValidateStatus2["ERROR"] = "error";
  ValidateStatus2["VALIDATING"] = "validating";
  return ValidateStatus2;
}(ValidateStatus || {});
var TD_CTRL_PROP_MAP = /* @__PURE__ */new Map([["Checkbox", "checked"], ["CheckTag", "checked"], ["Upload", "files"]]);
var TD_BOOLEAN_COMP = ["Checkbox"];
var TD_ARRAY_COMP = ["Tree", "Upload", "Transfer", "TagInput", "RangeInput", "CheckboxGroup", "DateRangePicker", "TimeRangePicker"];
var TD_DEFAULT_VALUE_MAP = function () {
  var map = /* @__PURE__ */new Map();
  TD_ARRAY_COMP.forEach(function (componentName) {
    map.set(componentName, []);
  });
  TD_BOOLEAN_COMP.forEach(function (componentName) {
    map.set(componentName, false);
  });
  return map;
}();
var NATIVE_INPUT_COMP = ["input", "textarea", "select", "progress"];

export { NATIVE_INPUT_COMP, TD_CTRL_PROP_MAP, TD_DEFAULT_VALUE_MAP, ValidateStatus };
