// tdesign-react/es/date-picker/hooks/useTableData.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../../_chunks/dep-94a7af64.js';
import { isArray, isFunction } from 'lodash-es';
import { g as getRangeBounds, m as monthHasAnyAllowed, y as yearHasAnyAllowed, b as getWeeks, d as getQuarters, f as getMonths, h as getYears, j as flagActive } from '../../_chunks/dep-7ff1070c.js';
import dayjs from 'dayjs';
import { useLocaleReceiver } from '../../locale/LocalReceiver.js';
import '../../_chunks/dep-0054b574.js';
import '../../_chunks/dep-8e8ac0d4.js';
import '../../_chunks/dep-5e72c8b9.js';
import '../../_chunks/dep-43f7da7b.js';
import '../../_chunks/dep-7d48ddac.js';
import '../../_chunks/dep-35769bc6.js';
import '../../_chunks/dep-9b727dd4.js';
import '../../_chunks/dep-07ae3139.js';
import '../../_chunks/dep-3561cd86.js';
import '../../_chunks/dep-65d82dda.js';
import 'react';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';
import '../../config-provider/ConfigContext.js';
import '../../locale/zh_CN.js';
import '../../_chunks/dep-766c4603.js';
import '../../config-provider/type.js';

function useTableData(props) {
  var _useLocaleReceiver = useLocaleReceiver("datePicker"),
    _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var monthLocal = t(local.months);
  var quarterLocal = t(local.quarters);
  var start = props.start,
    end = props.end,
    hoverStart = props.hoverStart,
    hoverEnd = props.hoverEnd,
    year = props.year,
    month = props.month,
    mode = props.mode,
    firstDayOfWeek = props.firstDayOfWeek,
    disableDate = props.disableDate,
    minDate = props.minDate,
    maxDate = props.maxDate,
    isRange = props.isRange,
    value = props.value,
    multiple = props.multiple;
  var options = {
    minDate: minDate,
    maxDate: maxDate,
    disableDate: disableDate,
    firstDayOfWeek: firstDayOfWeek,
    monthLocal: monthLocal,
    quarterLocal: quarterLocal,
    showWeekOfYear: mode === "week",
    dayjsLocale: local.dayjsLocale,
    cancelRangeSelectLimit: props.cancelRangeSelectLimit
  };
  var baseDisableDate = options.disableDate;
  var finalDisableDate = function finalDisableDate(date) {
    return baseDisableDate ? !!baseDisableDate(date) : false;
  };
  if (isArray(props.range)) {
    var _getRangeBounds = getRangeBounds(props.range),
      rawMin = _getRangeBounds.min,
      rawMax = _getRangeBounds.max;
    var rangeMin = rawMin ? new Date(dayjs(rawMin).startOf("day").format()) : null;
    var rangeMax = rawMax ? new Date(dayjs(rawMax).endOf("day").format()) : null;
    var min = rangeMin;
    var max = rangeMax;
    if (["date", "week"].includes(props.mode)) {
      if (min) {
        options.minDate = options.minDate ? new Date(Math.max(options.minDate.getTime(), min.getTime())) : min;
      }
      if (max) {
        options.maxDate = options.maxDate ? new Date(Math.min(options.maxDate.getTime(), max.getTime())) : max;
      }
    }
    var intervalOverlap = function intervalOverlap(aStart, aEnd, bStart, bEnd) {
      var s1 = aStart.getTime();
      var e1 = aEnd.getTime();
      var s2 = bStart ? bStart.getTime() : -Infinity;
      var e2 = bEnd ? bEnd.getTime() : Infinity;
      return !(e1 < s2 || s1 > e2);
    };
    finalDisableDate = function finalDisableDate(date) {
      var base = baseDisableDate ? !!baseDisableDate(date) : false;
      if (props.mode === "date" || props.mode === "week") {
        var t2 = date.getTime();
        var aboveMin = min ? t2 >= min.getTime() : true;
        var belowMax = max ? t2 <= max.getTime() : true;
        var allowed = aboveMin && belowMax;
        return base || !allowed;
      }
      if (props.mode === "month") {
        var start2 = new Date(date.getFullYear(), date.getMonth(), 1);
        var end2 = new Date(date.getFullYear(), date.getMonth() + 1, 0);
        end2.setHours(23, 59, 59, 999);
        var _allowed = intervalOverlap(start2, end2, min, max);
        return base || !_allowed;
      }
      if (props.mode === "quarter") {
        var startMonth = Math.floor(date.getMonth() / 3) * 3;
        var _start = new Date(date.getFullYear(), startMonth, 1);
        var _end = new Date(date.getFullYear(), startMonth + 3, 0);
        _end.setHours(23, 59, 59, 999);
        var _allowed2 = intervalOverlap(_start, _end, min, max);
        return base || !_allowed2;
      }
      if (props.mode === "year") {
        var _start2 = new Date(date.getFullYear(), 0, 1);
        var _end2 = new Date(date.getFullYear(), 11, 31);
        _end2.setHours(23, 59, 59, 999);
        var _allowed3 = intervalOverlap(_start2, _end2, min, max);
        return base || !_allowed3;
      }
      return base;
    };
  } else if (isFunction(props.range)) {
    var allow = props.range;
    finalDisableDate = function finalDisableDate(date) {
      var base = baseDisableDate ? !!baseDisableDate(date) : false;
      if (props.mode === "date" || props.mode === "week") {
        var disallowedByRange = !allow(date);
        return base || disallowedByRange;
      }
      if (props.mode === "month") {
        var allowed = monthHasAnyAllowed(props.range, date.getFullYear(), date.getMonth());
        return base || !allowed;
      }
      if (props.mode === "quarter") {
        var startMonth = Math.floor(date.getMonth() / 3) * 3;
        var _allowed4 = false;
        for (var m = startMonth; m < startMonth + 3; m++) {
          if (monthHasAnyAllowed(props.range, date.getFullYear(), m)) {
            _allowed4 = true;
            break;
          }
        }
        return base || !_allowed4;
      }
      if (props.mode === "year") {
        var _allowed5 = yearHasAnyAllowed(props.range, date.getFullYear());
        return base || !_allowed5;
      }
      return base;
    };
  }
  options.disableDate = finalDisableDate;
  var data = [];
  if (mode === "date") {
    data = getWeeks({
      year: year,
      month: month
    }, options);
  } else if (mode === "week") {
    data = getWeeks({
      year: year,
      month: month
    }, options);
  } else if (mode === "quarter") {
    data = getQuarters(year, options);
  } else if (mode === "month") {
    data = getMonths(year, options);
  } else if (mode === "year") {
    data = getYears(year, options);
  }
  return flagActive(data, {
    start: start,
    end: end,
    hoverStart: hoverStart,
    hoverEnd: hoverEnd,
    type: mode,
    isRange: isRange,
    value: value,
    multiple: multiple
  });
}

export { useTableData as default };
