// tdesign-react/cjs/form/FormItem.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var toConsumableArray = require('../_chunks/dep-1d0dbfc5.js');
var _typeof = require('../_chunks/dep-c4d1ce0c.js');
var index = require('../_chunks/dep-e37cb830.js');
var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var form_const = require('./const.js');
var form_defaultProps = require('./defaultProps.js');
var form_FormContext = require('./FormContext.js');
var form_formModel = require('./formModel.js');
var form_hooks_useForm = require('./hooks/useForm.js');
var form_hooks_useFormItemInitialData = require('./hooks/useFormItemInitialData.js');
var form_hooks_useFormItemStyle = require('./hooks/useFormItemStyle.js');
var form_utils_index = require('./utils/index.js');
var isEqual = require('../_chunks/dep-4e876c4b.js');
var get = require('../_chunks/dep-05df1ea3.js');
var isFunction = require('../_chunks/dep-cba78412.js');
var set = require('../_chunks/dep-57a4aad8.js');
var cloneDeep = require('../_chunks/dep-4a562e04.js');
var isString = require('../_chunks/dep-17e2e72f.js');
var isObject = require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-32593c54.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1c586f5c.js');
require('../_chunks/dep-f0e884b9.js');
require('../_chunks/dep-545dd164.js');
require('../_chunks/dep-427fb3a1.js');
require('../_chunks/dep-b289b0be.js');
require('../_chunks/dep-c17bb178.js');
require('../_chunks/dep-9b3a719d.js');
require('../_chunks/dep-13de2fef.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-441ad6a5.js');
require('../_chunks/dep-2e455d91.js');
require('../_chunks/dep-894a5ece.js');
require('../_chunks/dep-bcd16957.js');
require('classnames');
require('../_util/parseTNode.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-ca918872.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var FormItem = /*#__PURE__*/React.forwardRef(function (originalProps, ref) {
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("form"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    locale = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix,
    globalFormConfig = _useConfig.form;
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      CheckCircleFilledIcon: tdesignIconsReact.CheckCircleFilledIcon,
      CloseCircleFilledIcon: tdesignIconsReact.CloseCircleFilledIcon,
      ErrorCircleFilledIcon: tdesignIconsReact.ErrorCircleFilledIcon
    }),
    CheckCircleFilledIcon = _useGlobalIcon.CheckCircleFilledIcon,
    CloseCircleFilledIcon = _useGlobalIcon.CloseCircleFilledIcon,
    ErrorCircleFilledIcon = _useGlobalIcon.ErrorCircleFilledIcon;
  var _useFormContext = form_FormContext.useFormContext(),
    form = _useFormContext.form,
    colon = _useFormContext.colon,
    layout = _useFormContext.layout,
    requiredMarkFromContext = _useFormContext.requiredMark,
    requiredMarkPosition = _useFormContext.requiredMarkPosition,
    labelAlignFromContext = _useFormContext.labelAlign,
    labelWidthFromContext = _useFormContext.labelWidth,
    showErrorMessageFromContext = _useFormContext.showErrorMessage,
    disabledFromContext = _useFormContext.disabled,
    readOnlyFromContext = _useFormContext.readOnly,
    resetTypeFromContext = _useFormContext.resetType,
    rulesFromContext = _useFormContext.rules,
    statusIconFromContext = _useFormContext.statusIcon,
    errorMessage = _useFormContext.errorMessage,
    formMapRef = _useFormContext.formMapRef,
    onFormItemValueChange = _useFormContext.onFormItemValueChange;
  var _useFormListContext = form_FormContext.useFormListContext(),
    formListName = _useFormListContext.name,
    parentFullPath = _useFormListContext.fullPath,
    formListRules = _useFormListContext.rules,
    formListMapRef = _useFormListContext.formListMapRef,
    formOfFormList = _useFormListContext.form;
  var props = hooks_useDefaultProps["default"](originalProps, form_defaultProps.formItemDefaultProps);
  var children = props.children,
    style = props.style,
    label = props.label,
    name = props.name,
    status = props.status,
    tips = props.tips,
    help = props.help,
    valueFormat = props.valueFormat,
    initialData = props.initialData,
    className = props.className,
    shouldUpdate = props.shouldUpdate,
    successBorder = props.successBorder,
    _props$statusIcon = props.statusIcon,
    statusIcon = _props$statusIcon === void 0 ? statusIconFromContext : _props$statusIcon,
    _props$rules = props.rules,
    innerRules = _props$rules === void 0 ? getInnerRules(name, rulesFromContext, formListName, formListRules) : _props$rules,
    _props$labelWidth = props.labelWidth,
    labelWidth = _props$labelWidth === void 0 ? labelWidthFromContext : _props$labelWidth,
    _props$labelAlign = props.labelAlign,
    labelAlign = _props$labelAlign === void 0 ? labelAlignFromContext : _props$labelAlign,
    _props$requiredMark = props.requiredMark,
    requiredMark = _props$requiredMark === void 0 ? requiredMarkFromContext : _props$requiredMark;
  var isSameForm = React.useMemo(function () {
    return isEqual.isEqual(form, formOfFormList);
  }, [form, formOfFormList]);
  var fullPath = React.useMemo(function () {
    var validParentFullPath = formListName && isSameForm ? parentFullPath : void 0;
    return form_utils_index.concatName(validParentFullPath, name);
  }, [formListName, parentFullPath, name, isSameForm]);
  var _useFormItemInitialDa = form_hooks_useFormItemInitialData["default"](name, fullPath, initialData, children),
    defaultInitialData = _useFormItemInitialDa.defaultInitialData;
  var _useState = React.useState({}),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    forceUpdate = _useState2[1];
  var _useState3 = React.useState(void 0),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    freeShowErrorMessage = _useState4[0],
    setFreeShowErrorMessage = _useState4[1];
  var _useState5 = React.useState([]),
    _useState6 = slicedToArray._slicedToArray(_useState5, 2),
    errorList = _useState6[0],
    setErrorList = _useState6[1];
  var _useState7 = React.useState([]),
    _useState8 = slicedToArray._slicedToArray(_useState7, 2),
    successList = _useState8[0],
    setSuccessList = _useState8[1];
  var _useState9 = React.useState("validating"),
    _useState0 = slicedToArray._slicedToArray(_useState9, 2),
    verifyStatus = _useState0[0],
    setVerifyStatus = _useState0[1];
  var _useState1 = React.useState(false),
    _useState10 = slicedToArray._slicedToArray(_useState1, 2),
    resetValidating = _useState10[0],
    setResetValidating = _useState10[1];
  var _useState11 = React.useState(false),
    _useState12 = slicedToArray._slicedToArray(_useState11, 2),
    needResetField = _useState12[0],
    setNeedResetField = _useState12[1];
  var _useState13 = React.useState(defaultInitialData),
    _useState14 = slicedToArray._slicedToArray(_useState13, 2),
    formValue = _useState14[0],
    setFormValue = _useState14[1];
  var formItemRef = React.useRef(null);
  var innerFormItemsRef = React.useRef([]);
  var shouldEmitChangeRef = React.useRef(false);
  var shouldValidate = React.useRef(false);
  var valueRef = React.useRef(formValue);
  var errorListMapRef = React.useRef(/* @__PURE__ */new Map());
  var snakeName = [].concat(isSameForm ? formListName : void 0, name).filter(function (item) {
    return item !== void 0;
  }).toString();
  var errorMessages = React.useMemo(function () {
    return errorMessage !== null && errorMessage !== void 0 ? errorMessage : globalFormConfig.errorMessage;
  }, [errorMessage, globalFormConfig]);
  var showErrorMessage = React.useMemo(function () {
    if (typeof freeShowErrorMessage === "boolean") return freeShowErrorMessage;
    if (typeof props.showErrorMessage === "boolean") return props.showErrorMessage;
    return showErrorMessageFromContext;
  }, [freeShowErrorMessage, props.showErrorMessage, showErrorMessageFromContext]);
  var _useFormItemStyle = form_hooks_useFormItemStyle["default"]({
      className: className,
      help: help,
      tips: tips,
      snakeName: snakeName,
      status: status,
      successBorder: successBorder,
      errorList: errorList,
      successList: successList,
      layout: layout,
      verifyStatus: verifyStatus,
      label: label,
      labelWidth: labelWidth,
      labelAlign: labelAlign,
      requiredMark: requiredMark,
      requiredMarkPosition: requiredMarkPosition,
      showErrorMessage: showErrorMessage,
      innerRules: innerRules
    }),
    formItemClass = _useFormItemStyle.formItemClass,
    formItemLabelClass = _useFormItemStyle.formItemLabelClass,
    contentClass = _useFormItemStyle.contentClass,
    labelStyle = _useFormItemStyle.labelStyle,
    contentStyle = _useFormItemStyle.contentStyle,
    helpNode = _useFormItemStyle.helpNode,
    extraNode = _useFormItemStyle.extraNode;
  var updateFormValue = function updateFormValue(newVal) {
    var _form$getInternalHook, _form$getFieldsValue;
    var validate2 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
    var shouldEmitChange = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
    var _ref = (form === null || form === void 0 || (_form$getInternalHook = form.getInternalHooks) === null || _form$getInternalHook === void 0 ? void 0 : _form$getInternalHook.call(form, form_hooks_useForm.HOOK_MARK)) || {},
      setPrevStore = _ref.setPrevStore;
    setPrevStore === null || setPrevStore === void 0 || setPrevStore(form === null || form === void 0 || (_form$getFieldsValue = form.getFieldsValue) === null || _form$getFieldsValue === void 0 ? void 0 : _form$getFieldsValue.call(form, true));
    shouldEmitChangeRef.current = shouldEmitChange;
    shouldValidate.current = validate2;
    valueRef.current = newVal;
    var fieldValue = get.get(form === null || form === void 0 ? void 0 : form.store, fullPath);
    if (isEqual.isEqual(fieldValue, newVal)) return;
    set.set(form === null || form === void 0 ? void 0 : form.store, fullPath, newVal);
    setFormValue(newVal);
  };
  function getInnerRules(name2, formRules, formListName2, formListRules2) {
    if (Array.isArray(name2)) {
      return get.get(formRules === null || formRules === void 0 ? void 0 : formRules[formListName2], name2) || get.get(formListRules2, name2) || get.get(formRules, name2.join(".")) || [];
    }
    return (formRules === null || formRules === void 0 ? void 0 : formRules[name2]) || formListRules2 || [];
  }
  var renderSuffixIcon = function renderSuffixIcon() {
    if (statusIcon === false) return null;
    var resultIcon = function resultIcon(iconSlot) {
      return /* @__PURE__ */React__default["default"].createElement("span", {
        className: "".concat(classPrefix, "-form__status")
      }, iconSlot);
    };
    var getDefaultIcon = function getDefaultIcon() {
      var iconMap = {
        success: /* @__PURE__ */React__default["default"].createElement(CheckCircleFilledIcon, {
          size: "25px"
        }),
        error: /* @__PURE__ */React__default["default"].createElement(CloseCircleFilledIcon, {
          size: "25px"
        }),
        warning: /* @__PURE__ */React__default["default"].createElement(ErrorCircleFilledIcon, {
          size: "25px"
        })
      };
      if (verifyStatus === form_const.ValidateStatus.SUCCESS) {
        return resultIcon(iconMap[verifyStatus]);
      }
      if (errorList && errorList[0]) {
        var type = errorList[0].type || "error";
        return resultIcon(iconMap[type]);
      }
      return null;
    };
    if (/*#__PURE__*/React__default["default"].isValidElement(statusIcon)) {
      return resultIcon(/*#__PURE__*/React__default["default"].cloneElement(statusIcon, _objectSpread({
        style: {
          color: "unset"
        }
      }, statusIcon.props)));
    }
    if (statusIcon === true) {
      return getDefaultIcon();
    }
    return null;
  };
  function analysisValidateResult(_x) {
    return _analysisValidateResult.apply(this, arguments);
  }
  function _analysisValidateResult() {
    _analysisValidateResult = index._asyncToGenerator(/*#__PURE__*/index.regenerator.mark(function _callee(trigger) {
      var _result$rules;
      var result;
      return index.regenerator.wrap(function (_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            result = {
              successList: [],
              errorList: [],
              rules: [],
              resultList: [],
              allowSetValue: false
            };
            result.rules = trigger === "all" ? innerRules : innerRules.filter(function (item) {
              return (item.trigger || "change") === trigger;
            });
            if ((_result$rules = result.rules) !== null && _result$rules !== void 0 && _result$rules.length) {
              _context.next = 1;
              break;
            }
            setResetValidating(false);
            return _context.abrupt("return", result);
          case 1:
            result.allowSetValue = true;
            _context.next = 2;
            return form_formModel.validate(formValue, result.rules);
          case 2:
            result.resultList = _context.sent;
            result.errorList = result.resultList.filter(function (item) {
              return item.result !== true;
            }).map(function (item) {
              Object.keys(item).forEach(function (key) {
                if (!item.message && errorMessages[key]) {
                  item.message = form_formModel.parseMessage(errorMessages[key], {
                    validate: item[key],
                    name: isString.isString(label) ? label : String(name)
                  });
                }
              });
              return item;
            });
            result.successList = result.resultList.filter(function (item) {
              return item.result === true && item.message && item.type === "success";
            });
            return _context.abrupt("return", result);
          case 3:
          case "end":
            return _context.stop();
        }
      }, _callee);
    }));
    return _analysisValidateResult.apply(this, arguments);
  }
  function validate() {
    return _validate.apply(this, arguments);
  }
  function _validate() {
    _validate = index._asyncToGenerator(/*#__PURE__*/index.regenerator.mark(function _callee2() {
      var trigger,
        showErrorMessage2,
        _yield$analysisValida,
        innerSuccessList,
        innerErrorList,
        validateRules,
        resultList,
        allowSetValue,
        cacheErrorList,
        status2,
        _innerErrorList$,
        _cacheErrorList$,
        _args2 = arguments;
      return index.regenerator.wrap(function (_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            trigger = _args2.length > 0 && _args2[0] !== undefined ? _args2[0] : "all";
            showErrorMessage2 = _args2.length > 1 ? _args2[1] : undefined;
            if (!innerFormItemsRef.current.length) {
              _context2.next = 1;
              break;
            }
            return _context2.abrupt("return", innerFormItemsRef.current.map(function (innerFormItem) {
              return innerFormItem === null || innerFormItem === void 0 ? void 0 : innerFormItem.validate(trigger, showErrorMessage2);
            }));
          case 1:
            setResetValidating(true);
            setFreeShowErrorMessage(showErrorMessage2);
            _context2.next = 2;
            return analysisValidateResult(trigger);
          case 2:
            _yield$analysisValida = _context2.sent;
            innerSuccessList = _yield$analysisValida.successList;
            innerErrorList = _yield$analysisValida.errorList;
            validateRules = _yield$analysisValida.rules;
            resultList = _yield$analysisValida.resultList;
            allowSetValue = _yield$analysisValida.allowSetValue;
            if (innerErrorList.length && trigger !== "all") {
              errorListMapRef.current.set(trigger, innerErrorList);
            } else {
              errorListMapRef.current["delete"](trigger);
            }
            if (!innerErrorList.length && trigger === "all") {
              errorListMapRef.current.clear();
            }
            cacheErrorList = toConsumableArray._toConsumableArray(errorListMapRef.current.values()).flat();
            if (allowSetValue) {
              setSuccessList(innerSuccessList);
              setErrorList(cacheErrorList.length ? cacheErrorList : innerErrorList);
            }
            if (validateRules.length) {
              status2 = form_const.ValidateStatus.SUCCESS;
              if (innerErrorList.length || cacheErrorList.length) {
                status2 = (innerErrorList === null || innerErrorList === void 0 || (_innerErrorList$ = innerErrorList[0]) === null || _innerErrorList$ === void 0 ? void 0 : _innerErrorList$.type) || (cacheErrorList === null || cacheErrorList === void 0 || (_cacheErrorList$ = cacheErrorList[0]) === null || _cacheErrorList$ === void 0 ? void 0 : _cacheErrorList$.type) || form_const.ValidateStatus.ERROR;
              }
              setVerifyStatus(status2);
            } else {
              setVerifyStatus(form_const.ValidateStatus.VALIDATING);
            }
            if (needResetField) {
              resetHandler();
            }
            setResetValidating(false);
            return _context2.abrupt("return", defineProperty._defineProperty({}, snakeName, innerErrorList.length === 0 ? true : resultList));
          case 3:
          case "end":
            return _context2.stop();
        }
      }, _callee2);
    }));
    return _validate.apply(this, arguments);
  }
  function validateOnly() {
    return _validateOnly.apply(this, arguments);
  }
  function _validateOnly() {
    _validateOnly = index._asyncToGenerator(/*#__PURE__*/index.regenerator.mark(function _callee3() {
      var trigger,
        _yield$analysisValida2,
        innerErrorList,
        resultList,
        _args3 = arguments;
      return index.regenerator.wrap(function (_context3) {
        while (1) switch (_context3.prev = _context3.next) {
          case 0:
            trigger = _args3.length > 0 && _args3[0] !== undefined ? _args3[0] : "all";
            _context3.next = 1;
            return analysisValidateResult(trigger);
          case 1:
            _yield$analysisValida2 = _context3.sent;
            innerErrorList = _yield$analysisValida2.errorList;
            resultList = _yield$analysisValida2.resultList;
            return _context3.abrupt("return", defineProperty._defineProperty({}, snakeName, innerErrorList.length === 0 ? true : resultList));
          case 2:
          case "end":
            return _context3.stop();
        }
      }, _callee3);
    }));
    return _validateOnly.apply(this, arguments);
  }
  function handleItemBlur() {
    var filterRules = innerRules.filter(function (item) {
      return item.trigger === "blur";
    });
    filterRules.length && validate("blur");
  }
  function getResetValue(resetType) {
    if (resetType === "initial") {
      return defaultInitialData;
    }
    var emptyValue;
    if (Array.isArray(formValue)) {
      emptyValue = [];
    } else if (isObject.isObject(formValue)) {
      emptyValue = {};
    } else if (isString.isString(formValue)) {
      emptyValue = "";
    }
    return emptyValue;
  }
  function resetField(type) {
    if (typeof name === "undefined") return;
    var resetType = type || resetTypeFromContext;
    var resetValue = getResetValue(resetType);
    updateFormValue(resetValue, false);
    if (resetValidating) {
      setNeedResetField(true);
    } else {
      resetHandler();
    }
  }
  function resetHandler() {
    setNeedResetField(false);
    setErrorList([]);
    setSuccessList([]);
    setVerifyStatus(form_const.ValidateStatus.VALIDATING);
  }
  function setField(field) {
    var value = field.value,
      status2 = field.status,
      validateMessage = field.validateMessage;
    if (typeof status2 !== "undefined") {
      setErrorList(validateMessage ? [validateMessage] : []);
      setSuccessList(validateMessage ? [validateMessage] : []);
      setNeedResetField(false);
      setVerifyStatus(status2);
    }
    if (typeof value !== "undefined") {
      updateFormValue(value, typeof status2 === "undefined" ? true : false, true);
    }
  }
  function setValidateMessage(validateMessage) {
    var _validateMessage$;
    if (!validateMessage || !Array.isArray(validateMessage)) return;
    if (validateMessage.length === 0) {
      setErrorList([]);
      setVerifyStatus(form_const.ValidateStatus.SUCCESS);
      return;
    }
    setErrorList(validateMessage);
    var status2 = (validateMessage === null || validateMessage === void 0 || (_validateMessage$ = validateMessage[0]) === null || _validateMessage$ === void 0 ? void 0 : _validateMessage$.type) || form_const.ValidateStatus.ERROR;
    setVerifyStatus(status2);
  }
  function getValidateMessage() {
    return errorList;
  }
  React.useEffect(function () {
    var _form$getInternalHook2;
    if (!shouldUpdate || !form) return;
    var _ref2 = (form === null || form === void 0 || (_form$getInternalHook2 = form.getInternalHooks) === null || _form$getInternalHook2 === void 0 ? void 0 : _form$getInternalHook2.call(form, form_hooks_useForm.HOOK_MARK)) || {},
      getPrevStore = _ref2.getPrevStore,
      registerWatch = _ref2.registerWatch;
    var cancelRegister = registerWatch === null || registerWatch === void 0 ? void 0 : registerWatch(function () {
      var _form$getFieldsValue2;
      var currStore = (form === null || form === void 0 || (_form$getFieldsValue2 = form.getFieldsValue) === null || _form$getFieldsValue2 === void 0 ? void 0 : _form$getFieldsValue2.call(form, true)) || {};
      var updateFlag = shouldUpdate;
      if (isFunction.isFunction(shouldUpdate)) updateFlag = shouldUpdate(getPrevStore === null || getPrevStore === void 0 ? void 0 : getPrevStore(), currStore);
      if (updateFlag) forceUpdate({});
    });
    return cancelRegister;
  }, [shouldUpdate, form]);
  React.useEffect(function () {
    if (typeof name === "undefined") return;
    var isFormList = formListName && isSameForm;
    var mapRef = isFormList ? formListMapRef : formMapRef;
    if (!(mapRef !== null && mapRef !== void 0 && mapRef.current)) return;
    mapRef.current.set(fullPath, formItemRef);
    set.set(form === null || form === void 0 ? void 0 : form.store, fullPath, defaultInitialData);
    setFormValue(defaultInitialData);
    return function () {
      mapRef.current["delete"](fullPath);
    };
  }, [snakeName, formListName]);
  React.useEffect(function () {
    var _form$getInternalHook3, _form$getInternalHook4;
    form === null || form === void 0 || (_form$getInternalHook3 = form.getInternalHooks) === null || _form$getInternalHook3 === void 0 || (_form$getInternalHook3 = _form$getInternalHook3.call(form, form_hooks_useForm.HOOK_MARK)) === null || _form$getInternalHook3 === void 0 || (_form$getInternalHook4 = _form$getInternalHook3.notifyWatch) === null || _form$getInternalHook4 === void 0 || _form$getInternalHook4.call(_form$getInternalHook3, name);
    if (!shouldValidate.current) return;
    if (typeof name !== "undefined" && shouldEmitChangeRef.current) {
      var fieldValue = form_utils_index.calcFieldValue(fullPath, formValue);
      onFormItemValueChange === null || onFormItemValueChange === void 0 || onFormItemValueChange(fieldValue);
    }
    var filterRules = innerRules.filter(function (item) {
      return (item.trigger || "change") === "change";
    });
    filterRules.length && validate("change");
  }, [formValue, snakeName]);
  var instance = {
    name: name,
    fullPath: fullPath,
    value: formValue,
    initialData: initialData,
    isFormList: false,
    getValue: function getValue() {
      return cloneDeep.cloneDeep(valueRef.current);
    },
    setValue: function setValue(newVal) {
      return updateFormValue(newVal, true, true);
    },
    setField: setField,
    validate: validate,
    validateOnly: validateOnly,
    resetField: resetField,
    setValidateMessage: setValidateMessage,
    getValidateMessage: getValidateMessage,
    resetValidate: resetHandler
  };
  React.useImperativeHandle(ref, function () {
    return instance;
  });
  React.useImperativeHandle(formItemRef, function () {
    return instance;
  });
  if (isFunction.isFunction(children)) return children(form);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: formItemClass,
    style: style
  }, label && /* @__PURE__ */React__default["default"].createElement("div", {
    className: formItemLabelClass,
    style: labelStyle
  }, /* @__PURE__ */React__default["default"].createElement("label", {
    htmlFor: props === null || props === void 0 ? void 0 : props["for"]
  }, label), colon && t(locale.colonText)), /* @__PURE__ */React__default["default"].createElement("div", {
    className: contentClass(),
    style: contentStyle
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-form__controls-content")
  }, React__default["default"].Children.map(children, function (child, index) {
    if (!child) return null;
    if (! /*#__PURE__*/React__default["default"].isValidElement(child) || child.type === React__default["default"].Fragment) return child;
    var childType = child.type;
    var isCustomComp = _typeof._typeof(childType) === "object" || typeof childType === "function";
    var componentName = isCustomComp ? childType.displayName : childType;
    if (componentName === "FormItem") {
      return /*#__PURE__*/React__default["default"].cloneElement(child, {
        ref: function ref(el) {
          if (!el) return;
          innerFormItemsRef.current[index] = el;
        }
      });
    }
    var childProps = child.props;
    var commonProps = _objectSpread({
      disabled: disabledFromContext,
      readOnly: readOnlyFromContext
    }, childProps);
    if (!isCustomComp && !form_const.NATIVE_INPUT_COMP.includes(componentName)) {
      return /*#__PURE__*/React__default["default"].cloneElement(child, commonProps);
    }
    var ctrlKey = "value";
    if (isCustomComp) {
      ctrlKey = form_const.TD_CTRL_PROP_MAP.get(componentName) || "value";
    }
    return /*#__PURE__*/React__default["default"].cloneElement(child, _objectSpread(_objectSpread({
      disabled: disabledFromContext,
      readOnly: readOnlyFromContext
    }, childProps), {}, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, ctrlKey, formValue), "onChange", function onChange(value) {
      var _childProps$onChange, _childProps$onChange$;
      var newValue = valueFormat ? valueFormat(value) : value;
      updateFormValue(newValue, true, true);
      for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        args[_key - 1] = arguments[_key];
      }
      childProps === null || childProps === void 0 || (_childProps$onChange = childProps.onChange) === null || _childProps$onChange === void 0 || (_childProps$onChange$ = _childProps$onChange.call) === null || _childProps$onChange$ === void 0 || _childProps$onChange$.call.apply(_childProps$onChange$, [_childProps$onChange, null, value].concat(args));
    }), "onBlur", function onBlur(value) {
      var _childProps$onBlur, _childProps$onBlur$ca;
      handleItemBlur();
      for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
        args[_key2 - 1] = arguments[_key2];
      }
      childProps === null || childProps === void 0 || (_childProps$onBlur = childProps.onBlur) === null || _childProps$onBlur === void 0 || (_childProps$onBlur$ca = _childProps$onBlur.call) === null || _childProps$onBlur$ca === void 0 || _childProps$onBlur$ca.call.apply(_childProps$onBlur$ca, [_childProps$onBlur, null, value].concat(args));
    })));
  }), renderSuffixIcon()), helpNode, extraNode));
});
FormItem.displayName = "FormItem";

exports["default"] = FormItem;
