// tdesign-react/cjs/form/Form.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var toConsumableArray = require('../_chunks/dep-1d0dbfc5.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var _util_noop = require('../_util/noop.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var form_FormContext = require('./FormContext.js');
var form_FormItem = require('./FormItem.js');
var form_FormList = require('./FormList.js');
var form_defaultProps = require('./defaultProps.js');
var form_hooks_useForm = require('./hooks/useForm.js');
var form_hooks_useInstance = require('./hooks/useInstance.js');
var form_hooks_useWatch = require('./hooks/useWatch.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-c4d1ce0c.js');
require('hoist-non-react-statics');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-e37cb830.js');
require('tdesign-icons-react');
require('../hooks/useGlobalIcon.js');
require('../locale/LocalReceiver.js');
require('../_chunks/dep-17e2e72f.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('./const.js');
require('./formModel.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1c586f5c.js');
require('../_chunks/dep-f0e884b9.js');
require('../_chunks/dep-545dd164.js');
require('./hooks/useFormItemInitialData.js');
require('../_chunks/dep-c17bb178.js');
require('../_chunks/dep-05df1ea3.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-9b3a719d.js');
require('../_chunks/dep-13de2fef.js');
require('../_chunks/dep-441ad6a5.js');
require('../_chunks/dep-2e455d91.js');
require('../_chunks/dep-894a5ece.js');
require('../_chunks/dep-bcd16957.js');
require('./hooks/useFormItemStyle.js');
require('../_util/parseTNode.js');
require('../_chunks/dep-b289b0be.js');
require('./utils/index.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-57a4aad8.js');
require('../_chunks/dep-ca918872.js');
require('../_chunks/dep-427fb3a1.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Form = _util_forwardRefWithStatics["default"](function (originalProps, ref) {
  var _form$getInternalHook, _form$getInternalHook2;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix,
    globalFormConfig = _useConfig.form;
  var props = hooks_useDefaultProps["default"](originalProps, form_defaultProps.formDefaultProps);
  var style = props.style,
    className = props.className,
    labelWidth = props.labelWidth,
    statusIcon = props.statusIcon,
    labelAlign = props.labelAlign,
    layout = props.layout,
    colon = props.colon,
    initialData = props.initialData,
    _props$requiredMark = props.requiredMark,
    requiredMark = _props$requiredMark === void 0 ? globalFormConfig.requiredMark : _props$requiredMark,
    _props$requiredMarkPo = props.requiredMarkPosition,
    requiredMarkPosition = _props$requiredMarkPo === void 0 ? globalFormConfig.requiredMarkPosition : _props$requiredMarkPo,
    scrollToFirstError = props.scrollToFirstError,
    showErrorMessage = props.showErrorMessage,
    resetType = props.resetType,
    rules = props.rules,
    _props$errorMessage = props.errorMessage,
    errorMessage = _props$errorMessage === void 0 ? globalFormConfig.errorMessage : _props$errorMessage,
    disabled = props.disabled,
    children = props.children,
    id = props.id,
    onReset = props.onReset,
    _props$onValuesChange = props.onValuesChange,
    onValuesChange = _props$onValuesChange === void 0 ? _util_noop["default"] : _props$onValuesChange;
  var formClass = classNames__default["default"]("".concat(classPrefix, "-form"), className, defineProperty._defineProperty({}, "".concat(classPrefix, "-form-inline"), layout === "inline"));
  var _useForm = form_hooks_useForm["default"](props.form),
    _useForm2 = slicedToArray._slicedToArray(_useForm, 1),
    form = _useForm2[0];
  var formRef = React.useRef(null);
  var formMapRef = React.useRef(/* @__PURE__ */new Map());
  var floatingFormDataRef = React.useRef({});
  var formInstance = form_hooks_useInstance["default"](props, formRef, formMapRef, floatingFormDataRef, form);
  React.useImperativeHandle(ref, function () {
    return formInstance;
  });
  Object.assign(form, _objectSpread({}, formInstance));
  form === null || form === void 0 || (_form$getInternalHook = form.getInternalHooks) === null || _form$getInternalHook === void 0 || (_form$getInternalHook = _form$getInternalHook.call(form, form_hooks_useForm.HOOK_MARK)) === null || _form$getInternalHook === void 0 || (_form$getInternalHook2 = _form$getInternalHook.setForm) === null || _form$getInternalHook2 === void 0 || _form$getInternalHook2.call(_form$getInternalHook, formInstance);
  React.useEffect(function () {
    var _form$getInternalHook3, _form$getInternalHook4;
    form === null || form === void 0 || (_form$getInternalHook3 = form.getInternalHooks) === null || _form$getInternalHook3 === void 0 || (_form$getInternalHook3 = _form$getInternalHook3.call(form, form_hooks_useForm.HOOK_MARK)) === null || _form$getInternalHook3 === void 0 || (_form$getInternalHook4 = _form$getInternalHook3.flashQueue) === null || _form$getInternalHook4 === void 0 || _form$getInternalHook4.call(_form$getInternalHook3);
  }, [form]);
  function onResetHandler(e) {
    var _form$getInternalHook5, _form$getInternalHook6;
    toConsumableArray._toConsumableArray(formMapRef.current.values()).forEach(function (formItemRef) {
      formItemRef === null || formItemRef === void 0 || formItemRef.current.resetField();
    });
    form === null || form === void 0 || (_form$getInternalHook5 = form.getInternalHooks) === null || _form$getInternalHook5 === void 0 || (_form$getInternalHook5 = _form$getInternalHook5.call(form, form_hooks_useForm.HOOK_MARK)) === null || _form$getInternalHook5 === void 0 || (_form$getInternalHook6 = _form$getInternalHook5.notifyWatch) === null || _form$getInternalHook6 === void 0 || _form$getInternalHook6.call(_form$getInternalHook5, []);
    form.store = {};
    floatingFormDataRef.current = {};
    onReset === null || onReset === void 0 || onReset({
      e: e
    });
  }
  function onFormItemValueChange(changedValue) {
    requestAnimationFrame(function () {
      var allFields = formInstance.getFieldsValue(true);
      onValuesChange(changedValue, allFields);
    });
  }
  return /* @__PURE__ */React__default["default"].createElement(form_FormContext["default"].Provider, {
    value: {
      form: form,
      labelWidth: labelWidth,
      statusIcon: statusIcon,
      labelAlign: labelAlign,
      layout: layout,
      colon: colon,
      initialData: initialData,
      requiredMark: requiredMark,
      requiredMarkPosition: requiredMarkPosition,
      errorMessage: errorMessage,
      showErrorMessage: showErrorMessage,
      scrollToFirstError: scrollToFirstError,
      resetType: resetType,
      rules: rules,
      disabled: disabled,
      readOnly: props.readOnly || props.readonly,
      formMapRef: formMapRef,
      floatingFormDataRef: floatingFormDataRef,
      onFormItemValueChange: onFormItemValueChange
    }
  }, /* @__PURE__ */React__default["default"].createElement("form", {
    ref: formRef,
    id: id,
    style: style,
    className: formClass,
    onSubmit: formInstance.submit,
    onReset: onResetHandler
  }, children));
}, {
  useForm: form_hooks_useForm["default"],
  useWatch: form_hooks_useWatch["default"],
  FormItem: form_FormItem["default"],
  FormList: form_FormList["default"]
});
Form.displayName = "Form";

exports["default"] = Form;
