// tdesign-react/cjs/form/FormList.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _typeof = require('../_chunks/dep-c4d1ce0c.js');
var toConsumableArray = require('../_chunks/dep-1d0dbfc5.js');
var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var log = require('../_chunks/dep-b289b0be.js');
var form_FormContext = require('./FormContext.js');
var form_hooks_useForm = require('./hooks/useForm.js');
var form_utils_index = require('./utils/index.js');
var get = require('../_chunks/dep-05df1ea3.js');
var cloneDeep = require('../_chunks/dep-4a562e04.js');
var set = require('../_chunks/dep-57a4aad8.js');
var merge = require('../_chunks/dep-043eee20.js');
var isEqual = require('../_chunks/dep-4e876c4b.js');
var unset = require('../_chunks/dep-441ad6a5.js');
var isArray = require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-427fb3a1.js');
require('../_chunks/dep-9b3a719d.js');
require('../_chunks/dep-13de2fef.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-ca918872.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-2e455d91.js');
require('../_chunks/dep-894a5ece.js');
require('../_chunks/dep-bcd16957.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

/**
 * Casts `value` as an array if it's not one.
 *
 * @static
 * @memberOf _
 * @since 4.4.0
 * @category Lang
 * @param {*} value The value to inspect.
 * @returns {Array} Returns the cast array.
 * @example
 *
 * _.castArray(1);
 * // => [1]
 *
 * _.castArray({ 'a': 1 });
 * // => [{ 'a': 1 }]
 *
 * _.castArray('abc');
 * // => ['abc']
 *
 * _.castArray(null);
 * // => [null]
 *
 * _.castArray(undefined);
 * // => [undefined]
 *
 * _.castArray();
 * // => []
 *
 * var array = [1, 2, 3];
 * console.log(_.castArray(array) === array);
 * // => true
 */
function castArray() {
  if (!arguments.length) {
    return [];
  }
  var value = arguments[0];
  return isArray.isArray(value) ? value : [value];
}

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var globalKey = 0;
var FormList = function FormList(props) {
  var name = props.name,
    rules = props.rules,
    children = props.children;
  var _useFormContext = form_FormContext.useFormContext(),
    formMapRef = _useFormContext.formMapRef,
    form = _useFormContext.form,
    onFormItemValueChange = _useFormContext.onFormItemValueChange,
    initialDataFromForm = _useFormContext.initialData,
    resetTypeFromContext = _useFormContext.resetType;
  var _useFormListContext = form_FormContext.useFormListContext(),
    parentFullPath = _useFormListContext.fullPath,
    parentInitialData = _useFormListContext.initialData;
  var fullPath = form_utils_index.concatName(parentFullPath, name);
  var initialData = React.useMemo(function () {
    var propsInitialData;
    if (props.initialData) {
      propsInitialData = props.initialData;
    } else if (parentFullPath && parentInitialData) {
      var relativePath = fullPath.slice(form_utils_index.convertNameToArray(parentFullPath).length);
      propsInitialData = get.get(parentInitialData, relativePath);
    } else {
      propsInitialData = get.get(initialDataFromForm, fullPath);
    }
    return cloneDeep.cloneDeep(propsInitialData || []);
  }, [props.initialData, fullPath, parentFullPath, parentInitialData, initialDataFromForm]);
  var _useState = React.useState(function () {
      return get.get(form === null || form === void 0 ? void 0 : form.store, fullPath) || initialData;
    }),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    formListValue = _useState2[0],
    setFormListValue = _useState2[1];
  var _useState3 = React.useState(function () {
      return formListValue.map(function (data, index) {
        return {
          data: _objectSpread({}, data),
          key: globalKey += 1,
          name: index,
          isListField: true
        };
      });
    }),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    fields = _useState4[0],
    setFields = _useState4[1];
  var formListRef = React.useRef(null);
  var formListMapRef = React.useRef(/* @__PURE__ */new Map());
  var snakeName = [].concat(name).filter(function (item) {
    return item !== void 0;
  }).toString();
  var updateFormList = function updateFormList(newFields, newFormListValue) {
    set.set(form === null || form === void 0 ? void 0 : form.store, fullPath, newFormListValue);
    setFields(newFields);
    setFormListValue(newFormListValue);
    var changeValue = form_utils_index.calcFieldValue(fullPath, newFormListValue);
    onFormItemValueChange === null || onFormItemValueChange === void 0 || onFormItemValueChange(changeValue);
  };
  var operation = {
    add: function add(defaultValue, insertIndex) {
      var newFields = toConsumableArray._toConsumableArray(fields);
      var index = insertIndex !== null && insertIndex !== void 0 ? insertIndex : newFields.length;
      newFields.splice(index, 0, {
        key: globalKey += 1,
        name: index,
        isListField: true
      });
      newFields.forEach(function (field, index2) {
        return Object.assign(field, {
          name: index2
        });
      });
      var newFormListValue = toConsumableArray._toConsumableArray(formListValue);
      newFormListValue.splice(index, 0, cloneDeep.cloneDeep(defaultValue));
      updateFormList(newFields, newFormListValue);
    },
    remove: function remove(index) {
      var indices = castArray(index);
      var newFields = toConsumableArray._toConsumableArray(fields).filter(function (f) {
        return !indices.includes(f.name);
      }).map(function (field, i) {
        return _objectSpread(_objectSpread({}, field), {}, {
          name: i
        });
      });
      var newFormListValue = toConsumableArray._toConsumableArray(formListValue).filter(function (_, i) {
        return !indices.includes(i);
      });
      unset.unset(form === null || form === void 0 ? void 0 : form.store, fullPath);
      updateFormList(newFields, newFormListValue);
    },
    move: function move(from, to) {
      var newFields = toConsumableArray._toConsumableArray(fields);
      var newFormListValue = toConsumableArray._toConsumableArray(formListValue);
      form_utils_index.swap(newFields, from, to);
      form_utils_index.swap(newFormListValue, from, to);
      newFields[from].name = from;
      newFields[to].name = to;
      updateFormList(newFields, newFormListValue);
    }
  };
  function setListFields(fieldData, callback) {
    if (isEqual.isEqual(formListValue, fieldData)) return;
    var newFieldData = toConsumableArray._toConsumableArray(fieldData);
    var newFields = newFieldData.map(function (_, index) {
      var currField = fields[index];
      var oldItem = formListValue[index];
      var newItem = newFieldData[index];
      var noChange = currField && isEqual.isEqual(oldItem, newItem);
      return {
        key: noChange ? currField.key : globalKey += 1,
        name: index,
        isListField: true
      };
    });
    Array.from(formListMapRef.current.values()).forEach(function (formItemRef) {
      if (!formItemRef.current) return;
      var childName = formItemRef.current.name;
      var data = get.get(newFieldData, childName);
      if (data !== void 0) callback(formItemRef, data);
    });
    updateFormList(newFields, newFieldData);
  }
  React.useEffect(function () {
    if (!name || !formMapRef) return;
    formMapRef.current.set(fullPath, formListRef);
    set.set(form === null || form === void 0 ? void 0 : form.store, fullPath, formListValue);
    return function () {
      formMapRef.current["delete"](fullPath);
    };
  }, [snakeName]);
  React.useEffect(function () {
    var _form$getInternalHook, _form$getInternalHook2;
    form === null || form === void 0 || (_form$getInternalHook = form.getInternalHooks) === null || _form$getInternalHook === void 0 || (_form$getInternalHook = _form$getInternalHook.call(form, form_hooks_useForm.HOOK_MARK)) === null || _form$getInternalHook === void 0 || (_form$getInternalHook2 = _form$getInternalHook.notifyWatch) === null || _form$getInternalHook2 === void 0 || _form$getInternalHook2.call(_form$getInternalHook, name);
  }, [form, snakeName, fields]);
  React.useImperativeHandle(formListRef, function () {
    return {
      name: name,
      fullPath: fullPath,
      value: formListValue,
      initialData: initialData,
      isFormList: true,
      formListMapRef: formListMapRef,
      getValue: function getValue() {
        return cloneDeep.cloneDeep(get.get(form === null || form === void 0 ? void 0 : form.store, fullPath));
      },
      validate: function validate() {
        var trigger = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "all";
        var resultList = [];
        var validates = toConsumableArray._toConsumableArray(formListMapRef.current.values()).map(function (formItemRef) {
          var _formItemRef$current, _formItemRef$current$;
          return formItemRef === null || formItemRef === void 0 || (_formItemRef$current = formItemRef.current) === null || _formItemRef$current === void 0 || (_formItemRef$current$ = _formItemRef$current.validate) === null || _formItemRef$current$ === void 0 ? void 0 : _formItemRef$current$.call(_formItemRef$current, trigger);
        });
        return new Promise(function (resolve) {
          Promise.all(validates).then(function (validateResult) {
            validateResult.forEach(function (result) {
              if (_typeof._typeof(result) !== "object") return;
              var errorValue = Object.values(result)[0];
              merge.merge(resultList, errorValue);
            });
            var errorItems = validateResult.filter(function (item) {
              if (_typeof._typeof(item) !== "object") return;
              return Object.values(item)[0] !== true;
            });
            if (errorItems.length) {
              resolve(defineProperty._defineProperty({}, snakeName, resultList));
            } else {
              resolve(defineProperty._defineProperty({}, snakeName, true));
            }
          });
        });
      },
      setValue: function setValue(fieldData) {
        setListFields(fieldData, function (formItemRef, data) {
          var _formItemRef$current2, _formItemRef$current3;
          formItemRef === null || formItemRef === void 0 || (_formItemRef$current2 = formItemRef.current) === null || _formItemRef$current2 === void 0 || (_formItemRef$current3 = _formItemRef$current2.setValue) === null || _formItemRef$current3 === void 0 || _formItemRef$current3.call(_formItemRef$current2, data);
        });
      },
      setField: function setField(fieldData) {
        var value = fieldData.value,
          status = fieldData.status;
        var currentValue = get.get(form === null || form === void 0 ? void 0 : form.store, fullPath) || [];
        if (isEqual.isEqual(currentValue, value)) return;
        setListFields(value, function (formItemRef, data) {
          var _formItemRef$current4, _formItemRef$current5;
          formItemRef === null || formItemRef === void 0 || (_formItemRef$current4 = formItemRef.current) === null || _formItemRef$current4 === void 0 || (_formItemRef$current5 = _formItemRef$current4.setField) === null || _formItemRef$current5 === void 0 || _formItemRef$current5.call(_formItemRef$current4, {
            value: data,
            status: status
          });
        });
      },
      resetField: function resetField(type) {
        var resetType = type || resetTypeFromContext;
        if (resetType === "initial") {
          var currentData = get.get(form === null || form === void 0 ? void 0 : form.store, fullPath);
          if (isEqual.isEqual(currentData, initialData)) return;
          set.set(form === null || form === void 0 ? void 0 : form.store, fullPath, initialData);
          setFormListValue(initialData);
          var newFields = initialData === null || initialData === void 0 ? void 0 : initialData.map(function (data, index) {
            return {
              data: _objectSpread({}, data),
              key: globalKey += 1,
              name: index,
              isListField: true
            };
          });
          setFields(newFields);
        } else {
          unset.unset(form === null || form === void 0 ? void 0 : form.store, fullPath);
          setFormListValue([]);
          setFields([]);
        }
      },
      setValidateMessage: function setValidateMessage(fieldData) {
        toConsumableArray._toConsumableArray(formListMapRef.current.values()).forEach(function (formItemRef) {
          var _formItemRef$current6, _formItemRef$current7;
          if (!formItemRef.current) return;
          var name2 = formItemRef.current.name;
          var data = get.get(fieldData, name2);
          formItemRef === null || formItemRef === void 0 || (_formItemRef$current6 = formItemRef.current) === null || _formItemRef$current6 === void 0 || (_formItemRef$current7 = _formItemRef$current6.setValidateMessage) === null || _formItemRef$current7 === void 0 || _formItemRef$current7.call(_formItemRef$current6, data);
        });
      },
      resetValidate: function resetValidate() {
        toConsumableArray._toConsumableArray(formListMapRef.current.values()).forEach(function (formItemRef) {
          var _formItemRef$current8, _formItemRef$current9;
          formItemRef === null || formItemRef === void 0 || (_formItemRef$current8 = formItemRef.current) === null || _formItemRef$current8 === void 0 || (_formItemRef$current9 = _formItemRef$current8.resetValidate) === null || _formItemRef$current9 === void 0 || _formItemRef$current9.call(_formItemRef$current8);
        });
      }
    };
  });
  if (typeof children !== "function") {
    log.log.error("Form", "FormList's children must be a function!");
    return null;
  }
  return /* @__PURE__ */React__default["default"].createElement(form_FormContext.FormListContext.Provider, {
    value: {
      name: name,
      fullPath: fullPath,
      rules: rules,
      formListMapRef: formListMapRef,
      initialData: initialData,
      form: form
    }
  }, children(fields, operation));
};
FormList.displayName = "FormList";

exports["default"] = FormList;
