// tdesign-react/es/select-input/SelectInput.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-5e72c8b9.js';
import React, { useRef, useImperativeHandle } from 'react';
import classNames from 'classnames';
import useConfig from '../hooks/useConfig.js';
import useDefaultProps from '../hooks/useDefaultProps.js';
import { Popup } from '../popup/index.js';
import { selectInputDefaultProps } from './defaultProps.js';
import useMultiple from './useMultiple.js';
import useOverlayInnerStyle from './useOverlayInnerStyle.js';
import useSingle from './useSingle.js';
import '../_chunks/dep-43f7da7b.js';
import '../config-provider/ConfigContext.js';
import 'lodash-es';
import '../locale/zh_CN.js';
import '../_chunks/dep-7d48ddac.js';
import 'dayjs';
import '../_chunks/dep-766c4603.js';
import '../popup/Popup.js';
import '../_chunks/dep-94a7af64.js';
import '../_chunks/dep-0054b574.js';
import '../_chunks/dep-8e8ac0d4.js';
import 'react-transition-group';
import '../_chunks/dep-a347b1b5.js';
import '../common/Portal.js';
import 'react-dom';
import '../hooks/useLayoutEffect.js';
import '../hooks/useAnimation.js';
import '../hooks/useAttach.js';
import '../hooks/useControlled.js';
import '../_util/noop.js';
import '../hooks/useMutationObserver.js';
import '../hooks/useLatest.js';
import '../hooks/usePopper.js';
import '../_chunks/dep-1e9f924f.js';
import '../_chunks/dep-57e67945.js';
import 'react-fast-compare';
import '@popperjs/core';
import '../hooks/useWindowSize.js';
import '../popup/defaultProps.js';
import '../popup/hooks/useTrigger.js';
import '../_util/listener.js';
import '../_util/ref.js';
import 'react-is';
import '../_util/isFragment.js';
import '../popup/utils/transition.js';
import '../popup/PopupPlugin.js';
import '../_chunks/dep-fe79dc5d.js';
import '../_util/react-render.js';
import '../common/PluginContainer.js';
import '../config-provider/index.js';
import '../config-provider/ConfigProvider.js';
import '../config-provider/type.js';
import '../popup/style/css.js';
import '../popup/type.js';
import '../tag-input/index.js';
import '../tag-input/TagInput.js';
import 'tdesign-icons-react';
import '../hooks/useDragSorter.js';
import '../hooks/useEventCallback.js';
import '../hooks/useGlobalIcon.js';
import '../input/index.js';
import '../input/Input.js';
import '../_chunks/dep-4461a684.js';
import '../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../_util/parseTNode.js';
import '../_chunks/dep-3561cd86.js';
import '../_chunks/dep-65d82dda.js';
import '../locale/LocalReceiver.js';
import '../input/defaultProps.js';
import '../input/InputGroup.js';
import '../input/useLengthLimit.js';
import '../_chunks/dep-9721fff3.js';
import '../input/style/css.js';
import '../input/type.js';
import '../tag-input/defaultProps.js';
import '../tag-input/useHover.js';
import '../tag-input/useTagList.js';
import '../tag/index.js';
import '../tag/Tag.js';
import '../_chunks/dep-dfdbe6bd.js';
import '../tag/defaultProps.js';
import '../tag/CheckTag.js';
import '../_chunks/dep-29a70a9d.js';
import '../tag/CheckTagGroup.js';
import '../tag/style/css.js';
import '../tag/type.js';
import '../tag-input/useTagScroll.js';
import '../tag-input/style/css.js';
import '../tag-input/type.js';
import '../hooks/useInnerPopupVisible.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../loading/defaultProps.js';
import '../loading/gradient.js';
import '../_chunks/dep-2481dfad.js';
import '../hooks/useDomRefCallback.js';
import '../loading/plugin.js';
import '../loading/style/css.js';
import '../loading/type.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var SelectInput = /*#__PURE__*/React.forwardRef(function (originalProps, ref) {
  var _useConfig = useConfig(),
    prefix = _useConfig.classPrefix;
  var props = useDefaultProps(originalProps, selectInputDefaultProps);
  var multiple = props.multiple,
    value = props.value,
    popupVisible = props.popupVisible,
    popupProps = props.popupProps,
    borderless = props.borderless,
    disabled = props.disabled;
  var selectInputRef = useRef(null);
  var selectInputWrapRef = useRef(null);
  var _useSingle = useSingle(props),
    commonInputProps = _useSingle.commonInputProps,
    inputRef = _useSingle.inputRef,
    singleInputValue = _useSingle.singleInputValue,
    onInnerClear = _useSingle.onInnerClear,
    renderSelectSingle = _useSingle.renderSelectSingle;
  var _useMultiple = useMultiple(props),
    tagInputRef = _useMultiple.tagInputRef,
    tags = _useMultiple.tags,
    multipleInputValue = _useMultiple.multipleInputValue,
    renderSelectMultiple = _useMultiple.renderSelectMultiple;
  var _useOverlayInnerStyle = useOverlayInnerStyle(props, {
      afterHidePopup: onInnerBlur
    }),
    tOverlayInnerStyle = _useOverlayInnerStyle.tOverlayInnerStyle,
    innerPopupVisible = _useOverlayInnerStyle.innerPopupVisible,
    onInnerPopupVisibleChange = _useOverlayInnerStyle.onInnerPopupVisibleChange,
    skipNextBlur = _useOverlayInnerStyle.skipNextBlur;
  var popupClasses = classNames([props.className, "".concat(prefix, "-select-input"), _defineProperty(_defineProperty(_defineProperty(_defineProperty({}, "".concat(prefix, "-select-input--borderless"), borderless), "".concat(prefix, "-select-input--multiple"), multiple), "".concat(prefix, "-select-input--popup-visible"), popupVisible !== null && popupVisible !== void 0 ? popupVisible : innerPopupVisible), "".concat(prefix, "-select-input--empty"), value instanceof Array ? !value.length : !value)]);
  useImperativeHandle(ref, function () {
    return _objectSpread(_objectSpread(_objectSpread({}, selectInputRef.current || {}), inputRef.current || {}), tagInputRef.current || {});
  });
  var visibleProps = {
    visible: popupVisible !== null && popupVisible !== void 0 ? popupVisible : innerPopupVisible
  };
  function onInnerBlur(ctx) {
    var _props$onBlur;
    if (skipNextBlur.current) return;
    var inputValue = props.multiple ? multipleInputValue : singleInputValue;
    var params = _objectSpread({
      e: ctx.e,
      inputValue: inputValue
    }, props.multiple && {
      tagInputValue: tags
    });
    (_props$onBlur = props.onBlur) === null || _props$onBlur === void 0 || _props$onBlur.call(props, props.value, params);
  }
  var mainContent = /* @__PURE__ */React.createElement("div", {
    className: popupClasses,
    style: props.style
  }, /* @__PURE__ */React.createElement(Popup, _objectSpread(_objectSpread(_objectSpread({
    ref: selectInputRef,
    trigger: (popupProps === null || popupProps === void 0 ? void 0 : popupProps.trigger) || "click",
    placement: "bottom-left",
    content: props.panel,
    hideEmptyPopup: true,
    onVisibleChange: onInnerPopupVisibleChange,
    updateScrollTop: props.updateScrollTop
  }, visibleProps), popupProps), {}, {
    disabled: disabled,
    overlayInnerStyle: tOverlayInnerStyle
  }), multiple ? renderSelectMultiple({
    commonInputProps: commonInputProps,
    popupVisible: visibleProps.visible,
    allowInput: props.allowInput,
    onInnerClear: onInnerClear,
    onInnerBlur: onInnerBlur
  }) : renderSelectSingle(visibleProps.visible, onInnerBlur)));
  if (!props.tips) {
    return mainContent;
  }
  return /* @__PURE__ */React.createElement("div", {
    ref: selectInputWrapRef,
    className: "".concat(prefix, "-select-input__wrap")
  }, mainContent, props.tips && /* @__PURE__ */React.createElement("div", {
    className: "".concat(prefix, "-input__tips ").concat(prefix, "-input__tips--").concat(props.status || "normal")
  }, props.tips));
});
SelectInput.displayName = "SelectInput";

export { SelectInput as default };
