// tdesign-react/cjs/typography/Text.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var typography_ellipsis_Ellipsis = require('./ellipsis/Ellipsis.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var typography_ellipsis_useEllipsis = require('./ellipsis/useEllipsis.js');
var button_Button = require('../button/Button.js');
var tooltip_index = require('../tooltip/index.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var typography_defaultProps = require('./defaultProps.js');
var _util_copyText = require('../_util/copyText.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('./ellipsis/Truncate.js');
require('../_chunks/dep-427fb3a1.js');
require('react-is');
require('../_chunks/dep-5c3139cb.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-2e455d91.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-894a5ece.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-63308b84.js');
require('../_chunks/dep-048e4314.js');
require('../_chunks/dep-7577d7c1.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../hooks/useDomRefCallback.js');
require('../hooks/useRipple.js');
require('../_chunks/dep-064e4dba.js');
require('../hooks/useAnimation.js');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-f0e884b9.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../button/defaultProps.js');
require('../_util/parseTNode.js');
require('../_chunks/dep-b289b0be.js');
require('../tooltip/Tooltip.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../hooks/useAttach.js');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/listener.js');
require('../_util/ref.js');
require('../_util/isFragment.js');
require('../popup/utils/transition.js');
require('../popup/PopupPlugin.js');
require('../tooltip/defaultProps.js');
require('../tooltip/TooltipLite.js');
require('../hooks/useSwitch.js');
require('../hooks/usePersistFn.js');
require('../_chunks/dep-2a85392e.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["theme", "disabled", "className", "copyable", "strong", "mark", "code", "keyboard", "underline", "delete", "italic", "children", "ellipsis"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Text = /*#__PURE__*/React.forwardRef(function (originalProps, ref) {
  var _copyable$tooltipProp;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var props = hooks_useDefaultProps["default"](originalProps, typography_defaultProps.textDefaultProps);
  var prefixCls = "".concat(classPrefix, "-typography");
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("typography"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var copiedText = t(local.copiedText);
  var theme = props.theme,
    disabled = props.disabled,
    className = props.className,
    copyable = props.copyable,
    strong = props.strong,
    mark = props.mark,
    code = props.code,
    keyboard = props.keyboard,
    underline = props.underline,
    deleteProp = props["delete"],
    italic = props.italic,
    children = props.children,
    ellipsis = props.ellipsis,
    rest = objectWithoutProperties._objectWithoutProperties(props, _excluded);
  var getComponent = function getComponent() {
    var _Object$entries$find;
    var componentMap = {
      strong: !!strong,
      mark: !!mark,
      code: !!code,
      kbd: !!keyboard,
      u: !!underline,
      del: !!deleteProp,
      i: !!italic
    };
    return (_Object$entries$find = Object.entries(componentMap).find(function (_ref) {
      var _ref2 = slicedToArray._slicedToArray(_ref, 2),
        condition = _ref2[1];
      return !!condition;
    })) === null || _Object$entries$find === void 0 ? void 0 : _Object$entries$find[0];
  };
  var currentRef = React.useRef(null);
  var _useEllipsis = typography_ellipsis_useEllipsis["default"](ellipsis),
    ellipsisProps = _useEllipsis.ellipsisProps;
  var Component = getComponent();
  var textEllipsisProps = _objectSpread({}, ellipsisProps);
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    isCopied = _useState2[0],
    setIsCopied = _useState2[1];
  var copyProps = typeof copyable === "boolean" ? {
    text: children === null || children === void 0 ? void 0 : children.toString(),
    onCopy: Function.prototype,
    tooltipProps: isCopied ? {
      content: copiedText
    } : null
  } : {
    text: (copyable === null || copyable === void 0 ? void 0 : copyable.text) || (children === null || children === void 0 ? void 0 : children.toString()),
    onCopy: copyable === null || copyable === void 0 ? void 0 : copyable.onCopy,
    tooltipProps: _objectSpread(_objectSpread({}, copyable === null || copyable === void 0 ? void 0 : copyable.tooltipProps), {}, {
      content: isCopied ? copiedText : copyable === null || copyable === void 0 || (_copyable$tooltipProp = copyable.tooltipProps) === null || _copyable$tooltipProp === void 0 ? void 0 : _copyable$tooltipProp.content
    }),
    suffix: copyable === null || copyable === void 0 ? void 0 : copyable.suffix
  };
  var handleCopy = function handleCopy() {
    _util_copyText["default"](copyProps === null || copyProps === void 0 ? void 0 : copyProps.text);
    setIsCopied(true);
    setTimeout(function () {
      setIsCopied(false);
    }, 1500);
    if (typeof copyProps.onCopy === "function") copyProps.onCopy();
  };
  var renderContent = function renderContent(withChildren) {
    var tooltipProps = copyProps.tooltipProps;
    var wrapWithTooltip = function wrapWithTooltip(wrapContent) {
      return tooltipProps ? /* @__PURE__ */React__default["default"].createElement(tooltip_index.Tooltip, _objectSpread({
        hideEmptyPopup: true
      }, tooltipProps), wrapContent) : wrapContent;
    };
    var getSuffix = function getSuffix() {
      if (typeof (copyProps === null || copyProps === void 0 ? void 0 : copyProps.suffix) === "function") {
        return copyProps.suffix({
          copied: isCopied
        });
      }
      return isCopied ? /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.CheckIcon, null) : /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.CopyIcon, null);
    };
    return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, withChildren ? children : null, copyable ? wrapWithTooltip(/* @__PURE__ */React__default["default"].createElement(button_Button["default"], {
      shape: "square",
      theme: "primary",
      variant: "text",
      icon: getSuffix(),
      onClick: handleCopy
    })) : null);
  };
  if (!ellipsis) {
    return Component ? /* @__PURE__ */React__default["default"].createElement("span", _objectSpread({
      className: classNames__default["default"](className, prefixCls),
      ref: ref || currentRef
    }, rest), /* @__PURE__ */React__default["default"].createElement(Component, null, renderContent(true))) : /* @__PURE__ */React__default["default"].createElement("span", _objectSpread({
      className: classNames__default["default"](className, defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(prefixCls, "--").concat(theme), theme), "".concat(prefixCls, "--disabled"), disabled))
    }, rest), renderContent(true));
  }
  return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, Component ? /* @__PURE__ */React__default["default"].createElement("span", _objectSpread({
    className: classNames__default["default"](className, prefixCls),
    ref: ref || currentRef
  }, rest), /* @__PURE__ */React__default["default"].createElement(typography_ellipsis_Ellipsis["default"], _objectSpread(_objectSpread({}, textEllipsisProps), {}, {
    component: Component
  }), children)) : /* @__PURE__ */React__default["default"].createElement(typography_ellipsis_Ellipsis["default"], _objectSpread(_objectSpread({}, textEllipsisProps), {}, {
    className: classNames__default["default"](className, defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(prefixCls, "--").concat(theme), theme), "".concat(prefixCls, "--disabled"), disabled))
  }, rest), children), renderContent(false));
});

exports["default"] = Text;
