// tdesign-react/cjs/select-input/useOverlayInnerStyle.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var _typeof = require('../_chunks/dep-c4d1ce0c.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var hooks_useControlled = require('../hooks/useControlled.js');
var hooks_useInnerPopupVisible = require('../hooks/useInnerPopupVisible.js');
var isFunction = require('../_chunks/dep-cba78412.js');
var isObject = require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-ecfe656e.js');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var MAX_POPUP_WIDTH = 1e3;
function useOverlayInnerStyle(props, extra) {
  var popupProps = props.popupProps,
    autoWidth = props.autoWidth,
    disabled = props.disabled,
    onPopupVisibleChange = props.onPopupVisibleChange,
    allowInput = props.allowInput;
  var readonly = props.readOnly || props.readonly;
  var _useControlled = hooks_useControlled["default"](props, "popupVisible", onPopupVisibleChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    innerPopupVisible = _useControlled2[0],
    setInnerPopupVisible = _useControlled2[1];
  var skipNextBlur = React.useRef(false);
  var matchWidthFunc = function matchWidthFunc(triggerElement, popupElement) {
    if (!triggerElement || !popupElement) return;
    var prevDisplay = popupElement.style.display;
    popupElement.style.display = "";
    var overlayScrollWidth = popupElement.offsetWidth - popupElement.scrollWidth;
    var width = popupElement.offsetWidth - overlayScrollWidth > triggerElement.offsetWidth ? popupElement.scrollWidth : triggerElement.offsetWidth - overlayScrollWidth;
    if (prevDisplay === "none") {
      popupElement.style.display = "none";
    }
    var otherOverlayInnerStyle = {};
    if (popupProps && _typeof._typeof(popupProps.overlayInnerStyle) === "object" && !popupProps.overlayInnerStyle.width) {
      otherOverlayInnerStyle = popupProps.overlayInnerStyle;
    }
    return _objectSpread({
      width: "".concat(Math.min(width, MAX_POPUP_WIDTH), "px")
    }, otherOverlayInnerStyle);
  };
  var onInnerPopupVisibleChange = hooks_useInnerPopupVisible["default"](function (visible, context) {
    skipNextBlur.current = false;
    if (disabled || readonly) return;
    var newVisible = context.trigger === "trigger-element-click" && allowInput ? true : visible;
    if (props.popupVisible !== newVisible) {
      setInnerPopupVisible(newVisible, context);
      if (!newVisible) {
        var _extra$afterHidePopup;
        extra === null || extra === void 0 || (_extra$afterHidePopup = extra.afterHidePopup) === null || _extra$afterHidePopup === void 0 || _extra$afterHidePopup.call(extra, context);
        skipNextBlur.current = true;
      }
    }
  });
  var tOverlayInnerStyle = React.useMemo(function () {
    var result = {};
    var overlayInnerStyle = (popupProps === null || popupProps === void 0 ? void 0 : popupProps.overlayInnerStyle) || {};
    if (isFunction.isFunction(overlayInnerStyle) || isObject.isObject(overlayInnerStyle) && overlayInnerStyle.width) {
      result = overlayInnerStyle;
    } else if (!autoWidth) {
      result = matchWidthFunc;
    }
    return result;
  }, [autoWidth, popupProps === null || popupProps === void 0 ? void 0 : popupProps.overlayInnerStyle]);
  return {
    tOverlayInnerStyle: tOverlayInnerStyle,
    innerPopupVisible: innerPopupVisible,
    skipNextBlur: skipNextBlur,
    onInnerPopupVisibleChange: onInnerPopupVisibleChange
  };
}

exports["default"] = useOverlayInnerStyle;
