// tdesign-react/cjs/list/hooks/useListVirtualScroll.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-2b5ba78d.js');
var React = require('react');
var log = require('../../_chunks/dep-b289b0be.js');
var hooks_useVirtualScroll = require('../../hooks/useVirtualScroll.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-1a3dddaf.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_chunks/dep-4e876c4b.js');
require('../../_chunks/dep-df8d2fc8.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-66f91e5d.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-ea361d31.js');
require('../../_chunks/dep-0bed591e.js');

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var useListVirtualScroll = function useListVirtualScroll(scroll, listRef, listItems) {
  var virtualScrollParams = React.useMemo(function () {
    return {
      data: listItems,
      scroll: scroll
    };
  }, [listItems, scroll]);
  var virtualConfig = hooks_useVirtualScroll["default"](listRef, virtualScrollParams);
  var isVirtualScroll = virtualConfig.isVirtualScroll;
  var lastScrollY = -1;
  var onInnerVirtualScroll = function onInnerVirtualScroll(e) {
    var target = e.target || e.srcElement;
    var top = target.scrollTop;
    if (lastScrollY !== top) {
      virtualConfig.isVirtualScroll && virtualConfig.handleScroll();
    } else {
      lastScrollY = -1;
    }
    lastScrollY = top;
  };
  var cursorStyle = React.useMemo(function () {
    return {
      position: "absolute",
      width: "1px",
      height: "1px",
      transition: "transform 0.2s",
      transform: "translate(0, ".concat(virtualConfig.scrollHeight, "px)"),
      msTransform: "translate(0, ".concat(virtualConfig.scrollHeight, "px)"),
      MozTransform: "translate(0, ".concat(virtualConfig.scrollHeight, "px)"),
      WebkitTransform: "translate(0, ".concat(virtualConfig.scrollHeight, "px)")
    };
  }, [virtualConfig.scrollHeight]);
  var listStyle = React.useMemo(function () {
    return {
      transform: "translate(0, ".concat(virtualConfig.translateY, "px)"),
      msTransform: "translate(0, ".concat(virtualConfig.translateY, "px)"),
      MozTransform: "translate(0, ".concat(virtualConfig.translateY, "px)"),
      WebkitTransform: "translate(0, ".concat(virtualConfig.translateY, "px)")
    };
  }, [virtualConfig.translateY]);
  var handleScrollTo = function handleScrollTo(params) {
    var index = params.index,
      key = params.key;
    var targetIndex = index === 0 ? index : index !== null && index !== void 0 ? index : Number(key);
    if (!targetIndex && targetIndex !== 0) {
      log.log.error("List", "scrollTo: `index` or `key` must exist.");
      return;
    }
    if (targetIndex < 0 || targetIndex >= listItems.length) {
      log.log.error("List", "".concat(targetIndex, " does not exist in data, check `index` or `key` please."));
      return;
    }
    virtualConfig.scrollToElement(_objectSpread(_objectSpread({}, params), {}, {
      index: targetIndex - 1
    }));
  };
  return {
    virtualConfig: virtualConfig,
    cursorStyle: cursorStyle,
    listStyle: listStyle,
    isVirtualScroll: isVirtualScroll,
    onInnerVirtualScroll: onInnerVirtualScroll,
    scrollToElement: handleScrollTo
  };
};

exports.useListVirtualScroll = useListVirtualScroll;
