// tdesign-react/lib/date-picker/base/Header.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _toConsumableArray } from '../../_chunks/dep-1e9f924f.js';
import { _ as _slicedToArray } from '../../_chunks/dep-94a7af64.js';
import React, { useRef, useCallback, useState, useMemo, useEffect } from 'react';
import useConfig from '../../hooks/useConfig.js';
import { useLocaleReceiver } from '../../locale/LocalReceiver.js';
import { PaginationMini } from '../../pagination/index.js';
import { Select } from '../../select/index.js';
import { useSelectRange } from '../hooks/useSelectRange.js';
import '../../_chunks/dep-8e8ac0d4.js';
import '../../_chunks/dep-57e67945.js';
import '../../_chunks/dep-0054b574.js';
import '../../config-provider/ConfigContext.js';
import '../../_chunks/dep-5e72c8b9.js';
import '../../_chunks/dep-43f7da7b.js';
import 'lodash-es';
import '../../_chunks/dep-923c86e8.js';
import '../../_chunks/dep-03cf63f2.js';
import 'dayjs';
import '../../config-provider/ConfigProvider.js';
import '../../pagination/Pagination.js';
import '../../_chunks/dep-4461a684.js';
import 'classnames';
import '../../_util/noop.js';
import '../../hooks/useControlled.js';
import '../../input-number/index.js';
import '../../input-number/InputNumber.js';
import 'tdesign-icons-react';
import '../../input/index.js';
import '../../input/Input.js';
import '../../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../../_util/parseTNode.js';
import '../../_chunks/dep-65d82dda.js';
import '../../hooks/useDefaultProps.js';
import '../../hooks/useGlobalIcon.js';
import '../../hooks/useLayoutEffect.js';
import '../../_chunks/dep-a347b1b5.js';
import '../../input/defaultProps.js';
import '../../input/InputGroup.js';
import '../../input/useLengthLimit.js';
import '../../_chunks/dep-ee4dfb46.js';
import '../../button/index.js';
import '../../button/Button.js';
import '../../hooks/useDomRefCallback.js';
import '../../hooks/useRipple.js';
import '../../_chunks/dep-2481dfad.js';
import '../../hooks/useAnimation.js';
import '../../loading/index.js';
import '../../loading/Loading.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../loading/defaultProps.js';
import '../../loading/gradient.js';
import '../../loading/plugin.js';
import '../../_util/react-render.js';
import '../../_chunks/dep-b9d1804c.js';
import '../../common/PluginContainer.js';
import '../../button/defaultProps.js';
import '../../input-number/useInputNumber.js';
import '../../_chunks/dep-35a55ece.js';
import '../../hooks/useCommonClassName.js';
import '../../input-number/defaultProps.js';
import '../../input-adornment/InputAdornment.js';
import '../../pagination/hooks/useBoundaryJumper.js';
import '../../pagination/hooks/usePrevNextJumper.js';
import '../../pagination/hooks/usePageNumber.js';
import '../../pagination/hooks/useTotal.js';
import '../../pagination/validators.js';
import '../../pagination/defaultProps.js';
import '../../select/base/Select.js';
import '../../_util/helper.js';
import '../../_util/ref.js';
import 'react-is';
import '../../_util/isFragment.js';
import '../../common/FakeArrow.js';
import '../../select-input/index.js';
import '../../select-input/SelectInput.js';
import '../../popup/index.js';
import '../../popup/Popup.js';
import 'react-transition-group';
import '../../hooks/useAttach.js';
import '../../hooks/useMutationObserver.js';
import '../../hooks/useLatest.js';
import '../../hooks/usePopper.js';
import 'react-fast-compare';
import '@popperjs/core';
import '../../hooks/useWindowSize.js';
import '../../popup/defaultProps.js';
import '../../popup/hooks/useTrigger.js';
import '../../_util/listener.js';
import '../../popup/utils/transition.js';
import '../../popup/PopupPlugin.js';
import '../../select-input/defaultProps.js';
import '../../select-input/useMultiple.js';
import '../../tag-input/index.js';
import '../../tag-input/TagInput.js';
import '../../hooks/useDragSorter.js';
import '../../hooks/useEventCallback.js';
import '../../tag-input/defaultProps.js';
import '../../tag-input/useHover.js';
import '../../tag-input/useTagList.js';
import '../../tag/index.js';
import '../../tag/Tag.js';
import '../../_chunks/dep-dfdbe6bd.js';
import '../../tag/defaultProps.js';
import '../../tag/CheckTag.js';
import '../../_chunks/dep-41651104.js';
import '../../tag/CheckTagGroup.js';
import '../../tag-input/useTagScroll.js';
import '../../select-input/useOverlayInnerStyle.js';
import '../../hooks/useInnerPopupVisible.js';
import '../../select-input/useSingle.js';
import '../../select/defaultProps.js';
import '../../select/hooks/useKeyboardControl.js';
import '../../_chunks/dep-a58e8ab4.js';
import '../../select/base/OptionGroup.js';
import '../../select/base/PopupContent.js';
import '../../select/hooks/usePanelVirtualScroll.js';
import '../../hooks/useVirtualScroll.js';
import '../../pagination/PaginationMini.js';
import '../../_chunks/dep-bd9ae7b8.js';
import '../../_chunks/dep-d0959166.js';
import '../../_chunks/dep-0b223aa8.js';
import '../../_chunks/dep-4dd692a9.js';

var useDatePickerLocalConfig = function useDatePickerLocalConfig() {
  var _useLocaleReceiver = useLocaleReceiver("datePicker"),
    _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  return {
    months: t(local.months),
    nextYear: t(local.nextYear),
    preYear: t(local.preYear),
    nextMonth: t(local.nextMonth),
    preMonth: t(local.preMonth),
    preDecade: t(local.preDecade),
    nextDecade: t(local.nextDecade),
    now: t(local.now)
  };
};
var DatePickerHeader = function DatePickerHeader(props) {
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix;
  var mode = props.mode,
    year = props.year,
    month = props.month,
    range = props.range,
    onMonthChange = props.onMonthChange,
    onYearChange = props.onYearChange,
    onJumperClick = props.onJumperClick,
    partial = props.partial,
    _props$internalYear = props.internalYear,
    internalYear = _props$internalYear === void 0 ? [] : _props$internalYear;
  var _useDatePickerLocalCo = useDatePickerLocalConfig(),
    now = _useDatePickerLocalCo.now,
    months = _useDatePickerLocalCo.months,
    preMonth = _useDatePickerLocalCo.preMonth,
    preYear = _useDatePickerLocalCo.preYear,
    nextMonth = _useDatePickerLocalCo.nextMonth,
    nextYear = _useDatePickerLocalCo.nextYear,
    preDecade = _useDatePickerLocalCo.preDecade,
    nextDecade = _useDatePickerLocalCo.nextDecade;
  var scrollAnchorRef = useRef("default");
  var _useSelectRange = useSelectRange({
      range: range,
      mode: mode,
      year: year,
      month: month
    }),
    paginationDisabled = _useSelectRange.paginationDisabled,
    monthHasAnyAllowed = _useSelectRange.monthHasAnyAllowed,
    yearHasAnyAllowed = _useSelectRange.yearHasAnyAllowed,
    decadeHasAnyAllowed = _useSelectRange.decadeHasAnyAllowed,
    canLoadMoreTop = _useSelectRange.canLoadMoreTop,
    canLoadMoreBottom = _useSelectRange.canLoadMoreBottom;
  var initOptions = useCallback(function (year2) {
    var options = [];
    if (mode === "year") {
      var extraYear = year2 % 10;
      var minYear = year2 - extraYear - 100;
      var maxYear = year2 - extraYear + 100;
      for (var i = minYear; i <= maxYear; i += 10) {
        var end = i + 9;
        if (decadeHasAnyAllowed(end)) {
          options.push({
            label: "".concat(i, " - ").concat(end),
            value: i + 9,
            disabled: false
          });
        }
      }
    } else {
      yearHasAnyAllowed(year2) && options.push({
        label: "".concat(year2),
        value: year2,
        disabled: false
      });
      for (var _i = 1; _i <= 10; _i++) {
        yearHasAnyAllowed(year2 + _i) && options.push({
          label: "".concat(year2 + _i),
          value: year2 + _i,
          disabled: false
        });
        yearHasAnyAllowed(year2 - _i) && options.unshift({
          label: "".concat(year2 - _i),
          value: year2 - _i,
          disabled: false
        });
      }
    }
    return options;
  }, [mode]);
  var _useState = useState(function () {
      return initOptions(year);
    }),
    _useState2 = _slicedToArray(_useState, 2),
    yearOptions = _useState2[0],
    setYearOptions = _useState2[1];
  var nearestYear = useMemo(function () {
    var _yearOptions$find;
    var extraYear = partial === "end" && mode === "year" && internalYear[1] - internalYear[0] <= 9 ? 9 : 0;
    return ((_yearOptions$find = yearOptions.find(function (option) {
      return option.value - (year + extraYear) <= 9 && option.value - (year + extraYear) >= 0;
    })) === null || _yearOptions$find === void 0 ? void 0 : _yearOptions$find.value) || year;
  }, [yearOptions, year, mode, partial, internalYear]);
  var monthOptions = months.map(function (item, index) {
    return {
      label: item,
      value: index,
      disabled: !monthHasAnyAllowed(year, index)
    };
  });
  var showPanelTop = useMemo(function () {
    var options = yearOptions;
    if (!options.length) return false;
    var first = options[0].value;
    return canLoadMoreTop(first);
  }, [canLoadMoreTop, yearOptions]);
  var showPanelBottom = useMemo(function () {
    var options = yearOptions;
    if (!options.length) return false;
    var last = options[options.length - 1].value;
    return canLoadMoreBottom(last);
  }, [canLoadMoreBottom, yearOptions]);
  function loadMoreYear(year2, type) {
    var options = [];
    if (mode === "year") {
      var extraYear = year2 % 10;
      if (type === "add") {
        for (var i = year2 - extraYear + 10; i <= year2 - extraYear + 50; i += 10) {
          var end = i + 9;
          decadeHasAnyAllowed(end) && options.push({
            label: "".concat(i, " - ").concat(end),
            value: i,
            disabled: false
          });
        }
      } else {
        for (var _i2 = year2 - extraYear - 1; _i2 > year2 - extraYear - 50; _i2 -= 10) {
          decadeHasAnyAllowed(_i2) && options.unshift({
            label: "".concat(_i2 - 9, " - ").concat(_i2),
            value: _i2,
            disabled: false
          });
        }
      }
    } else if (type === "add") {
      for (var _i3 = year2 + 1; _i3 <= year2 + 10; _i3++) {
        yearHasAnyAllowed(_i3) && options.push({
          label: "".concat(_i3),
          value: _i3,
          disabled: false
        });
      }
    } else {
      for (var _i4 = year2 - 1; _i4 > year2 - 10; _i4--) {
        yearHasAnyAllowed(_i4) && options.unshift({
          label: "".concat(_i4),
          value: _i4,
          disabled: false
        });
      }
    }
    return options;
  }
  var labelMap = {
    year: {
      prev: preDecade,
      current: now,
      next: nextDecade
    },
    month: {
      prev: preYear,
      current: now,
      next: nextYear
    },
    date: {
      prev: preMonth,
      current: now,
      next: nextMonth
    }
  };
  var headerClassName = "".concat(classPrefix, "-date-picker__header");
  var showMonthPicker = mode === "date" || mode === "week";
  function handlePanelTopClick(e) {
    var _e$stopPropagation, _e$nativeEvent, _e$nativeEvent$stopIm;
    e === null || e === void 0 || (_e$stopPropagation = e.stopPropagation) === null || _e$stopPropagation === void 0 || _e$stopPropagation.call(e);
    e === null || e === void 0 || (_e$nativeEvent = e.nativeEvent) === null || _e$nativeEvent === void 0 || (_e$nativeEvent$stopIm = _e$nativeEvent.stopImmediatePropagation) === null || _e$nativeEvent$stopIm === void 0 || _e$nativeEvent$stopIm.call(_e$nativeEvent);
    var firstYear = yearOptions[0].value;
    var options = loadMoreYear(firstYear, "reduce");
    setYearOptions([].concat(_toConsumableArray(options), _toConsumableArray(yearOptions)));
  }
  function handlePanelBottomClick(e) {
    var _e$stopPropagation2, _e$nativeEvent2;
    e === null || e === void 0 || (_e$stopPropagation2 = e.stopPropagation) === null || _e$stopPropagation2 === void 0 || _e$stopPropagation2.call(e);
    e === null || e === void 0 || (_e$nativeEvent2 = e.nativeEvent) === null || _e$nativeEvent2 === void 0 || _e$nativeEvent2.stopImmediatePropagation();
    var lastYear = yearOptions.slice(-1)[0].value;
    var options = loadMoreYear(lastYear, "add");
    setYearOptions([].concat(_toConsumableArray(yearOptions), _toConsumableArray(options)));
  }
  function handleScroll(_ref) {
    var e = _ref.e;
    if (e.target.scrollTop === 0) {
      showPanelTop && handlePanelTopClick();
      scrollAnchorRef.current = "top";
    } else if (e.target.scrollTop === e.target.scrollHeight - e.target.clientHeight) {
      showPanelBottom && handlePanelBottomClick();
      scrollAnchorRef.current = "bottom";
    }
  }
  function handleUpdateScrollTop(content) {
    if (scrollAnchorRef.current === "top") {
      content.scrollTop = 30 * 10;
    } else if (scrollAnchorRef.current === "bottom") {
      content.scrollTop = content.scrollHeight - 30 * 10;
    } else {
      var firstSelectedNode = content === null || content === void 0 ? void 0 : content.querySelector(".".concat(classPrefix, "-is-selected"));
      if (firstSelectedNode) {
        var _getComputedStyle = getComputedStyle(firstSelectedNode),
          paddingBottom = _getComputedStyle.paddingBottom;
        var _getComputedStyle2 = getComputedStyle(content),
          marginBottom = _getComputedStyle2.marginBottom;
        var elementBottomHeight = parseInt(paddingBottom, 10) + parseInt(marginBottom, 10);
        var updateValue = firstSelectedNode.offsetTop - content.offsetTop - (content.clientHeight - firstSelectedNode.clientHeight) + elementBottomHeight;
        content.scrollTop = updateValue;
      }
    }
  }
  useEffect(function () {
    var yearRange = initOptions(year);
    setYearOptions(yearRange);
  }, [initOptions, year]);
  return /* @__PURE__ */React.createElement("div", {
    className: headerClassName
  }, /* @__PURE__ */React.createElement("div", {
    className: "".concat(headerClassName, "-controller")
  }, showMonthPicker && /* @__PURE__ */React.createElement(Select, {
    className: "".concat(headerClassName, "-controller-month"),
    value: month,
    options: monthOptions,
    onChange: function onChange(val) {
      return onMonthChange(val);
    },
    popupProps: {
      attach: function attach(triggerElement) {
        return triggerElement.parentNode;
      },
      overlayClassName: "".concat(headerClassName, "-controller-month-popup")
    }
  }), /* @__PURE__ */React.createElement(Select, {
    className: "".concat(headerClassName, "-controller-year"),
    value: mode === "year" ? nearestYear : year,
    options: yearOptions,
    onChange: function onChange(val) {
      return onYearChange(val);
    },
    onPopupVisibleChange: function onPopupVisibleChange(visible) {
      if (!visible) scrollAnchorRef.current = "default";
    },
    popupProps: {
      onScroll: handleScroll,
      updateScrollTop: function updateScrollTop(el) {
        setTimeout(function () {
          handleUpdateScrollTop(el);
        }, 0);
      },
      attach: function attach(triggerElement) {
        return triggerElement.parentNode;
      },
      overlayClassName: "".concat(headerClassName, "-controller-year-popup")
    },
    panelTopContent: showPanelTop && /* @__PURE__ */React.createElement("div", {
      className: "".concat(classPrefix, "-select-option"),
      onClick: handlePanelTopClick
    }, "..."),
    panelBottomContent: showPanelBottom && /* @__PURE__ */React.createElement("div", {
      className: "".concat(classPrefix, "-select-option"),
      onClick: handlePanelBottomClick
    }, "...")
  })), /* @__PURE__ */React.createElement(PaginationMini, {
    tips: labelMap[mode],
    size: "small",
    disabled: paginationDisabled,
    onChange: onJumperClick
  }));
};
DatePickerHeader.displayName = "DatePickerHeader";
var DateHeader = /*#__PURE__*/React.memo(DatePickerHeader);

export { DateHeader as default };
