// tdesign-react/cjs/auto-complete/AutoComplete.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var hooks_useControlled = require('../hooks/useControlled.js');
var autoComplete_defaultProps = require('./defaultProps.js');
var hooks_useCommonClassName = require('../hooks/useCommonClassName.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var input_index = require('../input/index.js');
var popup_index = require('../popup/index.js');
var autoComplete_OptionList = require('./OptionList.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-ecfe656e.js');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-17e2e72f.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../input/Input.js');
require('../_chunks/dep-1c586f5c.js');
require('tdesign-icons-react');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../_util/parseTNode.js');
require('../_chunks/dep-b289b0be.js');
require('../hooks/useGlobalIcon.js');
require('../hooks/useLayoutEffect.js');
require('../_chunks/dep-68f5fe7b.js');
require('../input/defaultProps.js');
require('../input/InputGroup.js');
require('../input/useLengthLimit.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-f0e884b9.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useAnimation.js');
require('../hooks/useAttach.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/listener.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');
require('../popup/utils/transition.js');
require('../popup/PopupPlugin.js');
require('../_chunks/dep-e37cb830.js');
require('../_util/react-render.js');
require('../common/PluginContainer.js');
require('../_chunks/dep-26ad9068.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var AutoComplete = /*#__PURE__*/React.forwardRef(function (originalProps, ref) {
  var _props$placeholder;
  var props = hooks_useDefaultProps["default"](originalProps, autoComplete_defaultProps.autoCompleteDefaultProps);
  var readOnly = props.readOnly || props.readonly;
  var inputRef = React.useRef(null);
  var popupRef = React.useRef(null);
  var _useControlled = hooks_useControlled["default"](props, "value", props.onChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    tValue = _useControlled2[0],
    setTValue = _useControlled2[1];
  var _useCommonClassName = hooks_useCommonClassName["default"](),
    classPrefix = _useCommonClassName.classPrefix,
    sizeClassNames = _useCommonClassName.sizeClassNames;
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("input"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 1),
    global = _useLocaleReceiver2[0];
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    popupVisible = _useState2[0],
    setPopupVisible = _useState2[1];
  var optionListRef = React.useRef(null);
  React.useImperativeHandle(ref, function () {
    return {
      inputRef: inputRef.current,
      popupRef: popupRef.current
    };
  });
  var getOverlayStyle = function getOverlayStyle(trigger, popupElement) {
    var _props$popupProps;
    var triggerWidth = trigger.getBoundingClientRect().width || trigger.offsetWidth || trigger.clientWidth;
    var popupWidth = popupElement.getBoundingClientRect().width || popupElement.offsetWidth || popupElement.clientWidth;
    return _objectSpread({
      width: triggerWidth >= popupWidth ? "".concat(triggerWidth, "px") : "auto"
    }, (_props$popupProps = props.popupProps) === null || _props$popupProps === void 0 ? void 0 : _props$popupProps.overlayInnerStyle);
  };
  var classes = ["".concat(classPrefix, "-auto-complete")].concat(props.className);
  var popupClasses = function (_props$popupProps2) {
    var classes2 = ["".concat(classPrefix, "-select__dropdown")];
    if ((_props$popupProps2 = props.popupProps) !== null && _props$popupProps2 !== void 0 && _props$popupProps2.overlayClassName) {
      classes2 = classes2.concat(props.popupProps.overlayClassName);
    }
    return classNames__default["default"](classes2);
  }();
  var popupInnerClasses = function (_props$popupProps3) {
    var classes2 = ["".concat(classPrefix, "-select__dropdown-inner")];
    if ((_props$popupProps3 = props.popupProps) !== null && _props$popupProps3 !== void 0 && _props$popupProps3.overlayInnerClassName) {
      classes2 = classes2.concat(props.popupProps.overlayInnerClassName);
    }
    return classNames__default["default"](classes2);
  }();
  var onInputChange = function onInputChange(value, context) {
    if (!popupVisible) {
      setPopupVisible(true);
    }
    setTValue(value, context);
  };
  var innerInputProps = function () {
    var tProps = _objectSpread({
      value: tValue,
      size: props.size
    }, props.inputProps);
    return tProps;
  }();
  var onInnerFocus = function onInnerFocus(value, context) {
    var _props$onFocus;
    setPopupVisible(true);
    (_props$onFocus = props.onFocus) === null || _props$onFocus === void 0 || _props$onFocus.call(props, _objectSpread(_objectSpread({}, context), {}, {
      value: value
    }));
    var timer = setTimeout(function () {
      var _optionListRef$curren;
      (_optionListRef$curren = optionListRef.current) === null || _optionListRef$curren === void 0 || _optionListRef$curren.addKeyboardListener();
      clearTimeout(timer);
    }, 0);
  };
  var onInnerBlur = function onInnerBlur(value, context) {
    var _props$onBlur;
    (_props$onBlur = props.onBlur) === null || _props$onBlur === void 0 || _props$onBlur.call(props, _objectSpread(_objectSpread({}, context), {}, {
      value: value
    }));
  };
  var onInnerEnter = function onInnerEnter(value, context) {
    var _props$onEnter;
    (_props$onEnter = props.onEnter) === null || _props$onEnter === void 0 || _props$onEnter.call(props, _objectSpread(_objectSpread({}, context), {}, {
      value: value
    }));
  };
  var onInnerCompositionend = function onInnerCompositionend(value, context) {
    var _props$onCompositione;
    (_props$onCompositione = props.onCompositionend) === null || _props$onCompositione === void 0 || _props$onCompositione.call(props, _objectSpread(_objectSpread({}, context), {}, {
      value: value
    }));
  };
  var onInnerCompositionstart = function onInnerCompositionstart(value, context) {
    var _props$onCompositions;
    (_props$onCompositions = props.onCompositionstart) === null || _props$onCompositions === void 0 || _props$onCompositions.call(props, _objectSpread(_objectSpread({}, context), {}, {
      value: value
    }));
  };
  var onInnerSelect = function onInnerSelect(value, context) {
    var _props$onSelect;
    if (readOnly || props.disabled) return;
    setPopupVisible(false);
    setTValue(value, context);
    (_props$onSelect = props.onSelect) === null || _props$onSelect === void 0 || _props$onSelect.call(props, value, context);
  };
  var onPopupVisibleChange = function onPopupVisibleChange(visible, _ref) {
    var trigger = _ref.trigger;
    if (trigger !== "trigger-element-click") {
      setPopupVisible(visible);
    }
  };
  var triggerNode = props.triggerElement || props.children || /* @__PURE__ */React__default["default"].createElement(input_index.Input, _objectSpread({
    ref: inputRef,
    placeholder: (_props$placeholder = props.placeholder) !== null && _props$placeholder !== void 0 ? _props$placeholder : global.placeholder,
    tips: props.tips,
    status: props.status,
    readOnly: readOnly,
    disabled: props.disabled,
    clearable: props.clearable,
    autofocus: props.autofocus,
    borderless: props.borderless,
    onClear: props.onClear,
    onChange: onInputChange,
    onFocus: onInnerFocus,
    onBlur: onInnerBlur,
    onEnter: onInnerEnter,
    onCompositionend: onInnerCompositionend,
    onCompositionstart: onInnerCompositionstart
  }, innerInputProps));
  var listContent = Array.isArray(props.options) && /* @__PURE__ */React__default["default"].createElement(autoComplete_OptionList["default"], {
    ref: optionListRef,
    value: tValue,
    options: props.options,
    size: props.size,
    sizeClassNames: sizeClassNames,
    onSelect: onInnerSelect,
    popupVisible: popupVisible,
    highlightKeyword: props.highlightKeyword,
    filterable: props.filterable,
    filter: props.filter,
    empty: props.empty
  });
  var topContent = props.panelTopContent;
  var bottomContent = props.panelBottomContent;
  var panelContent = topContent || listContent || bottomContent ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-auto-complete__panel")
  }, topContent, listContent, bottomContent) : null;
  var popupProps = _objectSpread(_objectSpread({}, props.popupProps), {}, {
    overlayInnerStyle: getOverlayStyle,
    overlayInnerClassName: popupInnerClasses,
    overlayClassName: popupClasses
  });
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](classes),
    style: props.style
  }, /* @__PURE__ */React__default["default"].createElement(popup_index.Popup, _objectSpread({
    ref: popupRef,
    visible: popupVisible,
    onVisibleChange: onPopupVisibleChange,
    trigger: "focus",
    placement: "bottom-left",
    hideEmptyPopup: true,
    content: panelContent
  }, popupProps), triggerNode));
});
AutoComplete.displayName = "AutoComplete";

exports["default"] = AutoComplete;
