// tdesign-react/cjs/popconfirm/Popcontent.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _typeof = require('../_chunks/dep-c4d1ce0c.js');
var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var button_index = require('../button/index.js');
var _util_noop = require('../_util/noop.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var isString = require('../_chunks/dep-17e2e72f.js');
require('../button/Button.js');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-1c586f5c.js');
require('../hooks/useDomRefCallback.js');
require('../hooks/useRipple.js');
require('../_chunks/dep-064e4dba.js');
require('../hooks/useAnimation.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-68f5fe7b.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../hooks/useDefaultProps.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-f0e884b9.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../button/defaultProps.js');
require('../_util/parseTNode.js');
require('../_chunks/dep-b289b0be.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Popcontent = function Popcontent(props) {
  var content = props.content,
    cancelBtn = props.cancelBtn,
    confirmBtn = props.confirmBtn,
    icon = props.icon,
    theme = props.theme,
    _props$onCancel = props.onCancel,
    onCancel = _props$onCancel === void 0 ? _util_noop["default"] : _props$onCancel,
    _props$onConfirm = props.onConfirm,
    onConfirm = _props$onConfirm === void 0 ? _util_noop["default"] : _props$onConfirm,
    _props$onClose = props.onClose,
    onClose = _props$onClose === void 0 ? _util_noop["default"] : _props$onClose;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      InfoCircleFilledIcon: tdesignIconsReact.InfoCircleFilledIcon
    }),
    InfoCircleFilledIcon = _useGlobalIcon.InfoCircleFilledIcon;
  var hideCancel = cancelBtn === null || cancelBtn === void 0;
  var hideConfirm = confirmBtn === null || confirmBtn === void 0;
  function renderIcon() {
    var color = "#0052D9";
    var defaultIcon = /* @__PURE__ */React__default["default"].createElement(InfoCircleFilledIcon, null);
    switch (theme) {
      case "warning":
        color = "#FFAA00";
        break;
      case "danger":
        color = "#FF3E00";
        break;
    }
    var iconComponent = null;
    if (/*#__PURE__*/React__default["default"].isValidElement(icon)) {
      iconComponent = /*#__PURE__*/React__default["default"].cloneElement(icon, _objectSpread({
        style: {
          color: color
        }
      }, icon.props));
    } else if (defaultIcon) {
      iconComponent = /*#__PURE__*/React__default["default"].cloneElement(defaultIcon, {
        style: {
          color: color
        }
      });
    }
    return iconComponent;
  }
  function renderCancel() {
    if (/*#__PURE__*/React__default["default"].isValidElement(cancelBtn)) {
      return /*#__PURE__*/React__default["default"].cloneElement(cancelBtn, {
        onClick: function onClick(e) {
          var _cancelBtn$props, _cancelBtn$props$onCl;
          onClose({
            e: e,
            trigger: "cancel"
          });
          (_cancelBtn$props = cancelBtn.props) === null || _cancelBtn$props === void 0 || (_cancelBtn$props$onCl = _cancelBtn$props.onClick) === null || _cancelBtn$props$onCl === void 0 || _cancelBtn$props$onCl.call(_cancelBtn$props, e);
        }
      });
    }
    if (hideCancel) {
      return null;
    }
    return /* @__PURE__ */React__default["default"].createElement(button_index.Button, _objectSpread({
      size: "small",
      theme: "default",
      variant: "base",
      onClick: function onClick(e) {
        onClose({
          e: e,
          trigger: "cancel"
        });
        onCancel({
          e: e
        });
      }
    }, _typeof._typeof(cancelBtn) === "object" ? _objectSpread({}, cancelBtn) : {}), _typeof._typeof(cancelBtn) === "object" ? cancelBtn === null || cancelBtn === void 0 ? void 0 : cancelBtn.children : isString.isString(cancelBtn) && cancelBtn);
  }
  function renderConfirm() {
    if (/*#__PURE__*/React__default["default"].isValidElement(confirmBtn)) {
      return /*#__PURE__*/React__default["default"].cloneElement(confirmBtn, {
        onClick: function onClick(e) {
          var _confirmBtn$props, _confirmBtn$props$onC;
          onClose({
            e: e,
            trigger: "confirm"
          });
          (_confirmBtn$props = confirmBtn.props) === null || _confirmBtn$props === void 0 || (_confirmBtn$props$onC = _confirmBtn$props.onClick) === null || _confirmBtn$props$onC === void 0 || _confirmBtn$props$onC.call(_confirmBtn$props, e);
        }
      });
    }
    if (hideConfirm) {
      return null;
    }
    return /* @__PURE__ */React__default["default"].createElement(button_index.Button, _objectSpread({
      size: "small",
      theme: "primary",
      onClick: function onClick(e) {
        onClose({
          e: e,
          trigger: "confirm"
        });
        onConfirm({
          e: e
        });
      }
    }, _typeof._typeof(confirmBtn) === "object" ? _objectSpread({}, confirmBtn) : {}), _typeof._typeof(confirmBtn) === "object" ? confirmBtn === null || confirmBtn === void 0 ? void 0 : confirmBtn.children : isString.isString(confirmBtn) && confirmBtn);
  }
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-popconfirm__content"), props.className),
    style: props.style
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-popconfirm__body")
  }, renderIcon(), /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-popconfirm__inner")
  }, content)), /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-popconfirm__buttons")
  }, /* @__PURE__ */React__default["default"].createElement("span", {
    className: classNames__default["default"]("".concat(classPrefix, "-popconfirm__cancel"))
  }, renderCancel()), /* @__PURE__ */React__default["default"].createElement("span", {
    className: classNames__default["default"]("".concat(classPrefix, "-popconfirm__confirm"))
  }, renderConfirm())));
};
Popcontent.displayName = "Popcontent";

exports["default"] = Popcontent;
