// tdesign-react/esm/popconfirm/Popconfirm.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-5e72c8b9.js';
import { _ as _slicedToArray } from '../_chunks/dep-94a7af64.js';
import React, { forwardRef } from 'react';
import classNames from 'classnames';
import { Popup } from '../popup/index.js';
import useConfig from '../hooks/useConfig.js';
import useControlled from '../hooks/useControlled.js';
import { useLocaleReceiver } from '../locale/LocalReceiver.js';
import Popcontent from './Popcontent.js';
import { popconfirmDefaultProps } from './defaultProps.js';
import useDefaultProps from '../hooks/useDefaultProps.js';
import '../_chunks/dep-43f7da7b.js';
import '../_chunks/dep-0054b574.js';
import '../_chunks/dep-8e8ac0d4.js';
import '../popup/Popup.js';
import 'react-transition-group';
import 'lodash-es';
import '../_chunks/dep-a347b1b5.js';
import '../common/Portal.js';
import 'react-dom';
import '../hooks/useLayoutEffect.js';
import '../hooks/useAnimation.js';
import '../config-provider/ConfigContext.js';
import '../locale/zh_CN.js';
import '../_chunks/dep-7d48ddac.js';
import 'dayjs';
import '../_chunks/dep-766c4603.js';
import '../hooks/useAttach.js';
import '../hooks/useMutationObserver.js';
import '../hooks/useLatest.js';
import '../hooks/usePopper.js';
import '../_chunks/dep-1e9f924f.js';
import '../_chunks/dep-57e67945.js';
import 'react-fast-compare';
import '@popperjs/core';
import '../hooks/useWindowSize.js';
import '../popup/defaultProps.js';
import '../popup/hooks/useTrigger.js';
import '../_util/listener.js';
import '../_util/ref.js';
import 'react-is';
import '../_util/isFragment.js';
import '../popup/utils/transition.js';
import '../_util/noop.js';
import '../popup/PopupPlugin.js';
import '../_chunks/dep-fe79dc5d.js';
import '../_util/react-render.js';
import '../common/PluginContainer.js';
import '../config-provider/index.js';
import '../config-provider/ConfigProvider.js';
import '../config-provider/type.js';
import '../popup/style/index.js';
import '../popup/type.js';
import 'tdesign-icons-react';
import '../button/index.js';
import '../button/Button.js';
import '../_chunks/dep-4461a684.js';
import '../hooks/useDomRefCallback.js';
import '../hooks/useRipple.js';
import '../_chunks/dep-2481dfad.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../loading/defaultProps.js';
import '../loading/gradient.js';
import '../_chunks/dep-9721fff3.js';
import '../loading/plugin.js';
import '../loading/style/index.js';
import '../loading/type.js';
import '../button/defaultProps.js';
import '../_util/parseTNode.js';
import '../_chunks/dep-3561cd86.js';
import '../_chunks/dep-65d82dda.js';
import '../button/style/index.js';
import '../button/type.js';
import '../hooks/useGlobalIcon.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Popconfirm = /*#__PURE__*/forwardRef(function (originalProps, ref) {
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix;
  var _useLocaleReceiver = useLocaleReceiver("popconfirm"),
    _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var cancelContent = typeof local.cancel === "string" ? local.cancel : local.cancel.content;
  var confirmContent = typeof local.confirm === "string" ? local.confirm : local.confirm.content;
  var props = useDefaultProps(originalProps, popconfirmDefaultProps);
  var _props$cancelBtn = props.cancelBtn,
    cancelBtn = _props$cancelBtn === void 0 ? t(cancelContent) : _props$cancelBtn,
    _props$confirmBtn = props.confirmBtn,
    confirmBtn = _props$confirmBtn === void 0 ? t(confirmContent) : _props$confirmBtn;
  var _useControlled = useControlled(props, "visible", props.onVisibleChange),
    _useControlled2 = _slicedToArray(_useControlled, 2),
    visible = _useControlled2[0],
    setVisible = _useControlled2[1];
  return /* @__PURE__ */React.createElement(Popup, _objectSpread(_objectSpread({
    ref: ref
  }, props), {}, {
    visible: visible,
    trigger: "click",
    onVisibleChange: function onVisibleChange(visible2, context) {
      return setVisible(visible2, context);
    },
    overlayClassName: classNames("".concat(classPrefix, "-popconfirm")),
    content: /* @__PURE__ */React.createElement(Popcontent, _objectSpread(_objectSpread({
      cancelBtn: cancelBtn,
      confirmBtn: confirmBtn
    }, props), {}, {
      onClose: function onClose(context) {
        return setVisible(false, context);
      }
    }))
  }, props.popupProps));
});
Popconfirm.displayName = "Popconfirm";

export { Popconfirm as default };
