// tdesign-react/cjs/affix/Affix.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var _util_style = require('../_chunks/dep-68f5fe7b.js');
var _util_scroll = require('../_util/scroll.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var affix_defaultProps = require('./defaultProps.js');
var isFunction = require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-ecfe656e.js');
require('raf');
require('../_util/easing.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-2b5ba78d.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var Affix = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _useDefaultProps = hooks_useDefaultProps["default"](props, affix_defaultProps.affixDefaultProps),
    children = _useDefaultProps.children,
    content = _useDefaultProps.content,
    zIndex = _useDefaultProps.zIndex,
    container = _useDefaultProps.container,
    offsetBottom = _useDefaultProps.offsetBottom,
    offsetTop = _useDefaultProps.offsetTop,
    className = _useDefaultProps.className,
    style = _useDefaultProps.style,
    onFixedChange = _useDefaultProps.onFixedChange;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    containerReady = _useState2[0],
    setContainerReady = _useState2[1];
  var affixRef = React.useRef(null);
  var affixWrapRef = React.useRef(null);
  var placeholderEL = React.useRef(null);
  var scrollContainer = React.useRef(null);
  var ticking = React.useRef(false);
  var handleScroll = React.useCallback(function () {
    if (!ticking.current) {
      requestAnimationFrame(function () {
        var _affixWrapRef$current, _affixWrapRef$current2;
        var _ref = (_affixWrapRef$current = (_affixWrapRef$current2 = affixWrapRef.current) === null || _affixWrapRef$current2 === void 0 ? void 0 : _affixWrapRef$current2.getBoundingClientRect()) !== null && _affixWrapRef$current !== void 0 ? _affixWrapRef$current : {},
          _ref$top = _ref.top,
          wrapToTop = _ref$top === void 0 ? 0 : _ref$top,
          _ref$bottom = _ref.bottom,
          wrapToBottom = _ref$bottom === void 0 ? 0 : _ref$bottom,
          _ref$width = _ref.width,
          wrapWidth = _ref$width === void 0 ? 0 : _ref$width,
          _ref$height = _ref.height,
          wrapHeight = _ref$height === void 0 ? 0 : _ref$height;
        var containerToTop = 0;
        var containerToBottom = 0;
        if (_util_style.isWindow(scrollContainer.current)) {
          containerToBottom = scrollContainer.current.innerHeight;
        } else if (scrollContainer.current instanceof HTMLElement) {
          var rect = scrollContainer.current.getBoundingClientRect();
          containerToTop = rect.top;
          containerToBottom = rect.bottom;
        }
        var calcTop = wrapToTop - containerToTop;
        var fixedTop;
        if (props.offsetBottom !== void 0 && props.offsetTop === void 0) {
          var bottomThreshold = containerToBottom - (offsetBottom !== null && offsetBottom !== void 0 ? offsetBottom : 0);
          if (containerToBottom > 0 && wrapToBottom >= bottomThreshold) {
            fixedTop = bottomThreshold - wrapHeight;
          } else {
            fixedTop = false;
          }
        } else {
          var _scrollContainer$curr;
          var containerHeight = ((_scrollContainer$curr = scrollContainer.current) === null || _scrollContainer$curr === void 0 ? void 0 : _scrollContainer$curr[_util_style.isWindow(scrollContainer.current) ? "innerHeight" : "clientHeight"]) - wrapHeight;
          var calcBottom = containerToTop + containerHeight - (offsetBottom !== null && offsetBottom !== void 0 ? offsetBottom : 0);
          if (calcTop <= offsetTop) {
            fixedTop = containerToTop + offsetTop;
          } else if (wrapToTop >= calcBottom) {
            fixedTop = calcBottom;
          } else {
            fixedTop = false;
          }
        }
        if (affixRef.current) {
          var affixed = fixedTop !== false;
          var placeholderStatus = affixWrapRef.current.contains(placeholderEL.current);
          var prePlaceholderStatus = placeholderStatus;
          if (affixed) {
            affixRef.current.className = "".concat(classPrefix, "-affix");
            affixRef.current.style.top = "".concat(fixedTop, "px");
            affixRef.current.style.width = "".concat(wrapWidth, "px");
            affixRef.current.style.height = "".concat(wrapHeight, "px");
            if (zIndex) {
              affixRef.current.style.zIndex = "".concat(zIndex);
            }
            if (!placeholderStatus) {
              placeholderEL.current.style.width = "".concat(wrapWidth, "px");
              placeholderEL.current.style.height = "".concat(wrapHeight, "px");
              affixWrapRef.current.appendChild(placeholderEL.current);
              placeholderStatus = true;
            }
          } else {
            affixRef.current.removeAttribute("class");
            affixRef.current.removeAttribute("style");
            if (placeholderStatus) {
              placeholderEL.current.remove();
              placeholderStatus = false;
            }
          }
          if (prePlaceholderStatus !== placeholderStatus && isFunction.isFunction(onFixedChange)) {
            onFixedChange(affixed, {
              top: +fixedTop
            });
          }
        }
        ticking.current = false;
      });
    }
    ticking.current = true;
  }, [classPrefix, offsetBottom, offsetTop, zIndex, onFixedChange, props.offsetBottom, props.offsetTop]);
  React.useImperativeHandle(ref, function () {
    return {
      handleScroll: handleScroll
    };
  });
  React.useEffect(function () {
    placeholderEL.current = document.createElement("div");
  }, []);
  React.useEffect(function () {
    var checkContainerExist = function checkContainerExist() {
      var el = _util_scroll.getScrollContainer(container);
      var isReady = _util_style.isWindow(el) || el instanceof HTMLElement;
      setContainerReady(isReady);
      return isReady;
    };
    if (checkContainerExist()) return;
    var observer = new MutationObserver(function () {
      if (checkContainerExist()) {
        observer.disconnect();
      }
    });
    observer.observe(document.body, {
      childList: true,
      subtree: true
    });
    return function () {
      observer.disconnect();
    };
  }, [container]);
  React.useEffect(function () {
    if (!containerReady) return;
    var newContainer = _util_scroll.getScrollContainer(container);
    if (!newContainer) return;
    if (scrollContainer.current) {
      scrollContainer.current.removeEventListener("scroll", handleScroll);
    }
    scrollContainer.current = newContainer;
    handleScroll();
    scrollContainer.current.addEventListener("scroll", handleScroll);
    window.addEventListener("resize", handleScroll);
    var isContainerNotWindow = !_util_style.isWindow(scrollContainer.current);
    if (isContainerNotWindow) {
      window.addEventListener("scroll", handleScroll);
    }
    return function () {
      if (scrollContainer.current) {
        scrollContainer.current.removeEventListener("scroll", handleScroll);
      }
      window.removeEventListener("resize", handleScroll);
      if (isContainerNotWindow) {
        window.removeEventListener("scroll", handleScroll);
      }
    };
  }, [container, containerReady, handleScroll]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: affixWrapRef,
    className: className,
    style: style
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    ref: affixRef
  }, children || content));
});
Affix.displayName = "Affix";

exports["default"] = Affix;
