// tdesign-react/cjs/cascader/core/effect.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var treeNodeModel = require('../../_chunks/dep-91c4f617.js');
var cascader_core_helper = require('./helper.js');
var isNumber = require('../../_chunks/dep-f0e884b9.js');
var cloneDeep = require('../../_chunks/dep-4a562e04.js');
var isFunction = require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_chunks/dep-2b5ba78d.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-b289b0be.js');
require('../../_chunks/dep-f7bf3d17.js');
require('../../_chunks/dep-5e17a073.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-876d090e.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-b4442734.js');
require('../../_chunks/dep-ca918872.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-13de2fef.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-63308b84.js');
require('../../_chunks/dep-048e4314.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../_chunks/dep-089e8106.js');
require('../../_chunks/dep-5c3139cb.js');
require('../../_chunks/dep-a5b4810f.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-ea361d31.js');
require('../../_chunks/dep-2e455d91.js');
require('../../_chunks/dep-894a5ece.js');
require('../../_chunks/dep-bcd16957.js');
require('../../_chunks/dep-8f62fb96.js');
require('../../_chunks/dep-545dd164.js');

function expendClickEffect(propsTrigger, trigger, node, cascaderContext) {
  var checkStrictly = cascaderContext.checkStrictly,
    multiple = cascaderContext.multiple,
    treeStore = cascaderContext.treeStore,
    setVisible = cascaderContext.setVisible,
    setValue = cascaderContext.setValue,
    setTreeNodes = cascaderContext.setTreeNodes,
    setExpend = cascaderContext.setExpend,
    value = cascaderContext.value,
    max = cascaderContext.max,
    valueType = cascaderContext.valueType,
    isParentFilterable = cascaderContext.isParentFilterable;
  var isDisabled = node.disabled || multiple && value.length >= max && max !== 0;
  if (isDisabled) return;
  if (propsTrigger === trigger && node.children !== null) {
    var expanded = node.setExpanded(true);
    treeStore.refreshNodes();
    treeStore.replaceExpanded(expanded);
    var nodes = treeStore.getNodes().filter(function (node2) {
      return node2.visible;
    });
    setTreeNodes(nodes);
    if (multiple && !isParentFilterable) {
      setExpend(expanded);
    }
  }
  if (!multiple && (node.isLeaf() || checkStrictly) && trigger === "click") {
    var shouldClosePanel = !checkStrictly || propsTrigger === "hover";
    if (node.checked) {
      if (shouldClosePanel) {
        setVisible(false, {});
      }
      return;
    }
    treeStore.resetChecked();
    var checked = node.setChecked(true);
    var _checked = slicedToArray._slicedToArray(checked, 1),
      value2 = _checked[0];
    setValue(valueType === "single" ? value2 : node.getPath().map(function (item) {
      return item.value;
    }), "check", node.getModel());
    if (shouldClosePanel) {
      setVisible(false, {});
    }
  }
}
function valueChangeEffect(node, cascaderContext) {
  var disabled = cascaderContext.disabled,
    max = cascaderContext.max,
    inputVal = cascaderContext.inputVal,
    multiple = cascaderContext.multiple,
    setVisible = cascaderContext.setVisible,
    setValue = cascaderContext.setValue,
    treeNodes = cascaderContext.treeNodes,
    treeStore = cascaderContext.treeStore,
    valueType = cascaderContext.valueType;
  if (!node || disabled || node.disabled) {
    return;
  }
  var checked = node.setChecked(!node.isChecked());
  if (isNumber.isNumber(max) && max < 0) {
    console.warn("TDesign Warn:", "max should > 0");
  }
  if (checked.length > max && isNumber.isNumber(max) && max > 0) {
    return;
  }
  if (checked.length === 0) {
    var expanded = treeStore.getExpanded();
    setTimeout(function () {
      treeStore.replaceExpanded(expanded);
      treeStore.refreshNodes();
    }, 0);
  }
  if (!multiple) {
    setVisible(false, {});
  }
  var isSelectAll = treeNodes.every(function (item) {
    return checked.indexOf(item.value) > -1;
  });
  if (inputVal && isSelectAll) {
    setVisible(false, {});
  }
  var resValue = valueType === "single" ? checked : checked.map(function (val) {
    return treeStore.getNode(val).getPath().map(function (item) {
      return item.value;
    });
  });
  setValue(resValue, node.checked ? "uncheck" : "check", node.getModel());
}
function closeIconClickEffect(cascaderContext) {
  var setVisible = cascaderContext.setVisible,
    multiple = cascaderContext.multiple,
    setExpend = cascaderContext.setExpend,
    setValue = cascaderContext.setValue,
    treeStore = cascaderContext.treeStore;
  var expanded = treeStore.getExpanded();
  setTimeout(function () {
    treeStore.replaceExpanded(expanded);
    treeStore.refreshNodes();
  }, 0);
  setVisible(false, {});
  if (multiple) {
    setExpend([]);
  }
  setValue(multiple ? [] : "", "clear");
}
function handleRemoveTagEffect(cascaderContext, index, onRemove) {
  var disabled = cascaderContext.disabled,
    setValue = cascaderContext.setValue,
    value = cascaderContext.value,
    valueType = cascaderContext.valueType,
    treeStore = cascaderContext.treeStore;
  if (disabled) return;
  var newValue = cloneDeep.cloneDeep(value);
  var res = newValue.splice(index, 1);
  var node = treeStore.getNodes(res[0])[0];
  var checked = node.setChecked(!node.isChecked());
  if (valueType === "single") {
    setValue(newValue, "uncheck", node.getModel());
  } else {
    var resValue = checked.map(function (val) {
      return treeStore.getNode(val).getPath().map(function (item) {
        return item.value;
      });
    });
    setValue(resValue, "uncheck", node.getModel());
  }
  if (isFunction.isFunction(onRemove)) {
    onRemove({
      value: checked,
      node: node
    });
  }
}
var treeNodesEffect = function treeNodesEffect(inputVal, treeStore, setTreeNodes, filter, isParentFilterable) {
  if (!treeStore) return;
  var nodes = [];
  if (inputVal) {
    var filterMethods = function filterMethods(node) {
      if (!node.isLeaf() && !isParentFilterable) return;
      if (isFunction.isFunction(filter)) {
        return filter("".concat(inputVal), node);
      }
      var fullPathLabel = cascader_core_helper.getFullPathLabel(node, "");
      return fullPathLabel.indexOf("".concat(inputVal)) > -1;
    };
    nodes = treeStore.nodes.filter(filterMethods);
  } else {
    nodes = treeStore.getNodes().filter(function (node) {
      return node.visible;
    });
  }
  setTreeNodes(nodes);
};
var treeStoreExpendEffect = function treeStoreExpendEffect(treeStore, value, expend, valueType) {
  var _treeStore$config;
  var treeValue = cascader_core_helper.getTreeValue(value);
  if (!treeStore) return;
  var allowDuplicateValue = (_treeStore$config = treeStore.config) === null || _treeStore$config === void 0 ? void 0 : _treeStore$config.allowDuplicateValue;
  if (Array.isArray(treeValue) && expend.length === 0) {
    var expandedMap = /* @__PURE__ */new Map();
    var node = null;
    if (valueType === "full" && Array.isArray(value) && value.length > 0) {
      node = treeStore.getNode(value);
    } else {
      var _treeValue = slicedToArray._slicedToArray(treeValue, 1),
        val = _treeValue[0];
      if (val) {
        expandedMap.set(val, true);
        node = treeStore.getNode(val);
      }
    }
    if (node) {
      node.getParents().forEach(function (tn) {
        var expandedKey = allowDuplicateValue ? treeNodeModel.pathToKey(tn.getPath().map(function (n) {
          return n.value;
        })) : tn.value;
        expandedMap.set(expandedKey, true);
      });
      var expandedArr = Array.from(expandedMap.keys());
      treeStore.replaceExpanded(expandedArr);
    } else if (treeValue.length > 0) {
      treeStore.refreshNodes();
      return;
    } else {
      treeStore.resetExpanded();
    }
  }
  if (treeStore.getExpanded() && expend.length) {
    treeStore.replaceExpanded(expend);
  }
  treeStore.refreshNodes();
};

exports.closeIconClickEffect = closeIconClickEffect;
exports.expendClickEffect = expendClickEffect;
exports.handleRemoveTagEffect = handleRemoveTagEffect;
exports.treeNodesEffect = treeNodesEffect;
exports.treeStoreExpendEffect = treeStoreExpendEffect;
exports.valueChangeEffect = valueChangeEffect;
