// tdesign-react/cjs/cascader/core/helper.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _typeof = require('../../_chunks/dep-c4d1ce0c.js');
var treeNodeModel = require('../../_chunks/dep-91c4f617.js');
var isEmpty = require('../../_chunks/dep-545dd164.js');
require('../../_chunks/dep-2b5ba78d.js');
require('../../_chunks/dep-b289b0be.js');
require('../../_chunks/dep-f7bf3d17.js');
require('../../_chunks/dep-5e17a073.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-876d090e.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-b4442734.js');
require('../../_chunks/dep-ca918872.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-13de2fef.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-63308b84.js');
require('../../_chunks/dep-048e4314.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../_chunks/dep-089e8106.js');
require('../../_chunks/dep-5c3139cb.js');
require('../../_chunks/dep-a5b4810f.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-ea361d31.js');
require('../../_chunks/dep-2e455d91.js');
require('../../_chunks/dep-894a5ece.js');
require('../../_chunks/dep-bcd16957.js');
require('../../_chunks/dep-8f62fb96.js');

function getSingleContent(cascaderContext) {
  var value = cascaderContext.value,
    multiple = cascaderContext.multiple,
    treeStore = cascaderContext.treeStore,
    showAllLevels = cascaderContext.showAllLevels,
    valueType = cascaderContext.valueType;
  var isEmpty2 = !value && value !== 0 || Array.isArray(value) && value.length === 0;
  var isInvalidFullPathValue = valueType === "full" && !Array.isArray(value);
  if (multiple || isEmpty2 || isInvalidFullPathValue) return "";
  var formatContent = function formatContent(path2) {
    if (!(path2 !== null && path2 !== void 0 && path2.length)) return "";
    return showAllLevels ? path2.map(function (node) {
      return node.label;
    }).join(" ".concat(treeNodeModel.PATH_SEPARATOR, " ")) : path2[path2.length - 1].label;
  };
  var getDefaultDisplay = function getDefaultDisplay(val) {
    return Array.isArray(val) ? val.join(" ".concat(treeNodeModel.PATH_SEPARATOR, " ")) : String(val);
  };
  if (valueType === "full" && Array.isArray(value) && value.length > 0) {
    var node = treeStore === null || treeStore === void 0 ? void 0 : treeStore.getNode(value);
    var path2 = node === null || node === void 0 ? void 0 : node.getPath();
    return path2 !== null && path2 !== void 0 && path2.length ? formatContent(path2) : getDefaultDisplay(value);
  }
  var nodes = treeStore === null || treeStore === void 0 ? void 0 : treeStore.getNodes(value);
  if (!(nodes !== null && nodes !== void 0 && nodes.length)) {
    return getDefaultDisplay(value);
  }
  var path = nodes[0].getPath();
  return path !== null && path !== void 0 && path.length ? formatContent(path) : getDefaultDisplay(value);
}
function getMultipleContent(cascaderContext) {
  var value = cascaderContext.value,
    multiple = cascaderContext.multiple,
    treeStore = cascaderContext.treeStore,
    showAllLevels = cascaderContext.showAllLevels,
    valueType = cascaderContext.valueType;
  if (!multiple) return [];
  if (multiple && !Array.isArray(value)) return [];
  return value.map(function (item) {
    var _node;
    var node;
    if (valueType === "full" && Array.isArray(item)) {
      node = treeStore === null || treeStore === void 0 ? void 0 : treeStore.getNode(item);
    } else {
      var nodes = treeStore === null || treeStore === void 0 ? void 0 : treeStore.getNodes(item);
      node = nodes === null || nodes === void 0 ? void 0 : nodes[0];
    }
    if (!node) return void 0;
    return showAllLevels ? getFullPathLabel(node) : (_node = node) === null || _node === void 0 ? void 0 : _node.label;
  }).filter(function (item) {
    return !!item;
  });
}
function getPanels(treeNodes) {
  var panels = [];
  treeNodes.forEach(function (node) {
    if (panels[node.level]) {
      panels[node.level].push(node);
    } else {
      panels[node.level] = [node];
    }
  });
  return panels;
}
function getFullPathLabel(node) {
  var separator = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "/";
  return node === null || node === void 0 ? void 0 : node.getPath().map(function (node2) {
    return node2.label;
  }).join(separator);
}
var getTreeValue = function getTreeValue(value) {
  var treeValue = [];
  if (Array.isArray(value)) {
    if (value.length > 0 && _typeof._typeof(value[0]) === "object") {
      treeValue = value.map(function (val) {
        return val.value;
      });
    } else if (value.length) {
      treeValue = value;
    }
  } else if (value) {
    if (_typeof._typeof(value) === "object") {
      treeValue = [value.value];
    } else {
      treeValue = [value];
    }
  }
  return treeValue;
};
var getExpandValue = function getExpandValue(value, valueType) {
  if (valueType === "full" && Array.isArray(value) && value.length > 0) {
    return value[value.length - 1];
  }
  var treeValue = getTreeValue(value);
  return treeValue[0];
};
function isEmptyValues(value) {
  if (typeof value === "number" && !isNaN(value)) return false;
  return isEmpty.isEmpty(value);
}
function isValueInvalid(value, cascaderContext) {
  var multiple = cascaderContext.multiple,
    showAllLevels = cascaderContext.showAllLevels,
    valueType = cascaderContext.valueType;
  if (multiple) {
    return !Array.isArray(value);
  }
  if (valueType === "full") {
    return !Array.isArray(value);
  }
  if (Array.isArray(value)) {
    return !showAllLevels;
  }
  return false;
}

exports.getExpandValue = getExpandValue;
exports.getFullPathLabel = getFullPathLabel;
exports.getMultipleContent = getMultipleContent;
exports.getPanels = getPanels;
exports.getSingleContent = getSingleContent;
exports.getTreeValue = getTreeValue;
exports.isEmptyValues = isEmptyValues;
exports.isValueInvalid = isValueInvalid;
