// tdesign-react/cjs/date-picker/hooks/useDisableDate.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var dayjs = require('dayjs');
var utils = require('../../_chunks/dep-8227db70.js');
var isObject = require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-1a3dddaf.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_chunks/dep-2b5ba78d.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-06bf4625.js');
require('../../_chunks/dep-c990dbb4.js');
require('../../_chunks/dep-eab994e8.js');
require('../../_chunks/dep-876d090e.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-bf18bb11.js');
require('../../_chunks/dep-06ac19a1.js');
require('../../_chunks/dep-b289b0be.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-bcd16957.js');
require('../../_chunks/dep-2d5cf7d6.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-6abdd014.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

function useDisableDate(props) {
  var _disableDate = props.disableDate,
    format = props.format,
    mode = props.mode,
    start = props.start,
    end = props.end;
  return {
    disableDate: function disableDate(value) {
      return !utils.isEnabledDate({
        disableDate: _disableDate,
        format: format,
        mode: mode,
        value: value
      });
    },
    minDate: isObject.isObject(_disableDate) && "before" in _disableDate ? new Date(dayjs__default["default"](_disableDate.before).startOf("day").format()) : start,
    maxDate: isObject.isObject(_disableDate) && "after" in _disableDate ? new Date(dayjs__default["default"](_disableDate.after).endOf("day").format()) : end
  };
}

exports["default"] = useDisableDate;
