// tdesign-react/cjs/date-picker/hooks/useTableData.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var utils = require('../../_chunks/dep-8227db70.js');
var dayjs = require('dayjs');
var locale_LocalReceiver = require('../../locale/LocalReceiver.js');
var isArray = require('../../_chunks/dep-794b42ce.js');
var isFunction = require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_chunks/dep-2b5ba78d.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-06bf4625.js');
require('../../_chunks/dep-c990dbb4.js');
require('../../_chunks/dep-eab994e8.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-876d090e.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-bf18bb11.js');
require('../../_chunks/dep-06ac19a1.js');
require('../../_chunks/dep-b289b0be.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../_chunks/dep-bcd16957.js');
require('../../_chunks/dep-2d5cf7d6.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-6abdd014.js');
require('react');
require('../../config-provider/ConfigProvider.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-14013769.js');
require('../../_chunks/dep-043eee20.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../_chunks/dep-4a562e04.js');
require('../../_chunks/dep-a5b4810f.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-ea361d31.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

function useTableData(props) {
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("datePicker"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var monthLocal = t(local.months);
  var quarterLocal = t(local.quarters);
  var start = props.start,
    end = props.end,
    hoverStart = props.hoverStart,
    hoverEnd = props.hoverEnd,
    year = props.year,
    month = props.month,
    mode = props.mode,
    firstDayOfWeek = props.firstDayOfWeek,
    disableDate = props.disableDate,
    minDate = props.minDate,
    maxDate = props.maxDate,
    isRange = props.isRange,
    value = props.value,
    multiple = props.multiple;
  var options = {
    minDate: minDate,
    maxDate: maxDate,
    disableDate: disableDate,
    firstDayOfWeek: firstDayOfWeek,
    monthLocal: monthLocal,
    quarterLocal: quarterLocal,
    showWeekOfYear: mode === "week",
    dayjsLocale: local.dayjsLocale,
    cancelRangeSelectLimit: props.cancelRangeSelectLimit
  };
  var baseDisableDate = options.disableDate;
  var finalDisableDate = function finalDisableDate(date) {
    return baseDisableDate ? !!baseDisableDate(date) : false;
  };
  if (isArray.isArray(props.range)) {
    var _getRangeBounds = utils.getRangeBounds(props.range),
      rawMin = _getRangeBounds.min,
      rawMax = _getRangeBounds.max;
    var rangeMin = rawMin ? new Date(dayjs__default["default"](rawMin).startOf("day").format()) : null;
    var rangeMax = rawMax ? new Date(dayjs__default["default"](rawMax).endOf("day").format()) : null;
    var min = rangeMin;
    var max = rangeMax;
    if (["date", "week"].includes(props.mode)) {
      if (min) {
        options.minDate = options.minDate ? new Date(Math.max(options.minDate.getTime(), min.getTime())) : min;
      }
      if (max) {
        options.maxDate = options.maxDate ? new Date(Math.min(options.maxDate.getTime(), max.getTime())) : max;
      }
    }
    var intervalOverlap = function intervalOverlap(aStart, aEnd, bStart, bEnd) {
      var s1 = aStart.getTime();
      var e1 = aEnd.getTime();
      var s2 = bStart ? bStart.getTime() : -Infinity;
      var e2 = bEnd ? bEnd.getTime() : Infinity;
      return !(e1 < s2 || s1 > e2);
    };
    finalDisableDate = function finalDisableDate(date) {
      var base = baseDisableDate ? !!baseDisableDate(date) : false;
      if (props.mode === "date" || props.mode === "week") {
        var t2 = date.getTime();
        var aboveMin = min ? t2 >= min.getTime() : true;
        var belowMax = max ? t2 <= max.getTime() : true;
        var allowed = aboveMin && belowMax;
        return base || !allowed;
      }
      if (props.mode === "month") {
        var start2 = new Date(date.getFullYear(), date.getMonth(), 1);
        var end2 = new Date(date.getFullYear(), date.getMonth() + 1, 0);
        end2.setHours(23, 59, 59, 999);
        var _allowed = intervalOverlap(start2, end2, min, max);
        return base || !_allowed;
      }
      if (props.mode === "quarter") {
        var startMonth = Math.floor(date.getMonth() / 3) * 3;
        var _start = new Date(date.getFullYear(), startMonth, 1);
        var _end = new Date(date.getFullYear(), startMonth + 3, 0);
        _end.setHours(23, 59, 59, 999);
        var _allowed2 = intervalOverlap(_start, _end, min, max);
        return base || !_allowed2;
      }
      if (props.mode === "year") {
        var _start2 = new Date(date.getFullYear(), 0, 1);
        var _end2 = new Date(date.getFullYear(), 11, 31);
        _end2.setHours(23, 59, 59, 999);
        var _allowed3 = intervalOverlap(_start2, _end2, min, max);
        return base || !_allowed3;
      }
      return base;
    };
  } else if (isFunction.isFunction(props.range)) {
    var allow = props.range;
    finalDisableDate = function finalDisableDate(date) {
      var base = baseDisableDate ? !!baseDisableDate(date) : false;
      if (props.mode === "date" || props.mode === "week") {
        var disallowedByRange = !allow(date);
        return base || disallowedByRange;
      }
      if (props.mode === "month") {
        var allowed = utils.monthHasAnyAllowed(props.range, date.getFullYear(), date.getMonth());
        return base || !allowed;
      }
      if (props.mode === "quarter") {
        var startMonth = Math.floor(date.getMonth() / 3) * 3;
        var _allowed4 = false;
        for (var m = startMonth; m < startMonth + 3; m++) {
          if (utils.monthHasAnyAllowed(props.range, date.getFullYear(), m)) {
            _allowed4 = true;
            break;
          }
        }
        return base || !_allowed4;
      }
      if (props.mode === "year") {
        var _allowed5 = utils.yearHasAnyAllowed(props.range, date.getFullYear());
        return base || !_allowed5;
      }
      return base;
    };
  }
  options.disableDate = finalDisableDate;
  var data = [];
  if (mode === "date") {
    data = utils.getWeeks({
      year: year,
      month: month
    }, options);
  } else if (mode === "week") {
    data = utils.getWeeks({
      year: year,
      month: month
    }, options);
  } else if (mode === "quarter") {
    data = utils.getQuarters(year, options);
  } else if (mode === "month") {
    data = utils.getMonths(year, options);
  } else if (mode === "year") {
    data = utils.getYears(year, options);
  }
  return utils.flagActive(data, {
    start: start,
    end: end,
    hoverStart: hoverStart,
    hoverEnd: hoverEnd,
    type: mode,
    isRange: isRange,
    value: value,
    multiple: multiple
  });
}

exports["default"] = useTableData;
