// tdesign-react/cjs/anchor/AnchorItem.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var hooks_useConfig = require('../hooks/useConfig.js');
var anchor_AnchorContext = require('./AnchorContext.js');
var anchor_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_util/noop.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["href", "title", "target", "children", "className"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var AnchorItem = function AnchorItem(props) {
  var _useContext = React.useContext(anchor_AnchorContext.AnchorContext),
    onClick = _useContext.onClick,
    activeItem = _useContext.activeItem,
    registerItem = _useContext.registerItem,
    unregisterItem = _useContext.unregisterItem;
  var _useDefaultProps = hooks_useDefaultProps["default"](props, anchor_defaultProps.anchorItemDefaultProps),
    href = _useDefaultProps.href,
    title = _useDefaultProps.title,
    target = _useDefaultProps.target,
    children = _useDefaultProps.children,
    className = _useDefaultProps.className,
    rest = objectWithoutProperties._objectWithoutProperties(_useDefaultProps, _excluded);
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var titleAttr = typeof title === "string" ? title : null;
  var handleClick = function handleClick(e) {
    onClick({
      title: titleAttr,
      href: href
    }, e);
  };
  React.useEffect(function () {
    registerItem(href);
    return function () {
      return unregisterItem(href);
    };
  }, [href, registerItem, unregisterItem]);
  return /* @__PURE__ */React__default["default"].createElement("div", _objectSpread(_objectSpread({}, rest), {}, {
    className: classNames__default["default"]("".concat(classPrefix, "-anchor__item"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-active"), activeItem === href), className)
  }), /* @__PURE__ */React__default["default"].createElement("a", {
    href: href,
    className: classNames__default["default"]("".concat(classPrefix, "-anchor__item-link")),
    title: titleAttr,
    target: target,
    onClick: function onClick(e) {
      return handleClick(e);
    }
  }, title), children);
};
AnchorItem.displayName = "AnchorItem";

exports["default"] = AnchorItem;
