// tdesign-react/cjs/anchor/Anchor.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var hooks_useConfig = require('../hooks/useConfig.js');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var _util_style = require('../_chunks/dep-68f5fe7b.js');
var affix_index = require('../affix/index.js');
var anchor_AnchorContext = require('./AnchorContext.js');
var _util_scroll = require('../_util/scroll.js');
var anchor_AnchorItem = require('./AnchorItem.js');
var anchor_AnchorTarget = require('./AnchorTarget.js');
var anchor_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var isEmpty = require('../_chunks/dep-545dd164.js');
var isFunction = require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('hoist-non-react-statics');
require('../_chunks/dep-17e2e72f.js');
require('../affix/Affix.js');
require('../affix/defaultProps.js');
require('raf');
require('../_util/easing.js');
require('../_util/noop.js');
require('tdesign-icons-react');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../hooks/useAnimation.js');
require('../hooks/useAttach.js');
require('../hooks/useControlled.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../hooks/usePopper.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/listener.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');
require('../popup/utils/transition.js');
require('../popup/PopupPlugin.js');
require('../_chunks/dep-e37cb830.js');
require('../_util/react-render.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../message/index.js');
require('../message/Message.js');
require('../_chunks/dep-a7cf54af.js');
require('../hooks/usePersistFn.js');
require('../message/MessageClose.js');
require('../hooks/useGlobalIcon.js');
require('../message/useMessageClass.js');
require('../message/MessageIcon.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-f0e884b9.js');
require('../hooks/useDomRefCallback.js');
require('../loading/plugin.js');
require('../message/config.js');
require('../message/defaultProps.js');
require('../_chunks/dep-e44b03ec.js');
require('../_util/copyText.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["affixProps", "bounds", "targetOffset", "container", "size", "children", "cursor", "onClick", "onChange", "className", "getCurrentAnchor"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var ANCHOR_SHARP_REGEXP = /#(\S+)$/;
var Anchor = _util_forwardRefWithStatics["default"](function (props, ref) {
  var _useDefaultProps = hooks_useDefaultProps["default"](props, anchor_defaultProps.anchorDefaultProps),
    affixProps = _useDefaultProps.affixProps,
    bounds = _useDefaultProps.bounds,
    targetOffset = _useDefaultProps.targetOffset,
    container = _useDefaultProps.container,
    size = _useDefaultProps.size,
    children = _useDefaultProps.children,
    cursor = _useDefaultProps.cursor,
    onClick = _useDefaultProps.onClick,
    onChange = _useDefaultProps.onChange,
    className = _useDefaultProps.className,
    getCurrentAnchor = _useDefaultProps.getCurrentAnchor,
    rest = objectWithoutProperties._objectWithoutProperties(_useDefaultProps, _excluded);
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _useState = React.useState(""),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    activeItem = _useState2[0],
    setActiveItem = _useState2[1];
  var _useState3 = React.useState({
      top: "0px",
      height: "0px",
      opacity: 0
    }),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    cursorStyle = _useState4[0],
    setCursorStyle = _useState4[1];
  var anchorEl = React.useRef(null);
  var intervalRef = React.useRef({
    items: [],
    scrollContainer: _util_style.canUseDocument ? window : null,
    handleScrollLock: false
  });
  React.useImperativeHandle(ref, function () {
    return anchorEl.current;
  });
  var registerItem = function registerItem(href) {
    var items = intervalRef.current.items;
    if (ANCHOR_SHARP_REGEXP.test(href) && items.indexOf(href) < 0) items.push(href);
  };
  var unregisterItem = function unregisterItem(href) {
    var items = intervalRef.current.items;
    intervalRef.current.items = items.filter(function (item) {
      return href !== item;
    });
  };
  var getAnchorTarget = function getAnchorTarget(href) {
    return document.querySelector(href);
  };
  var handleScrollTo = function handleScrollTo(link) {
    var anchor = getAnchorTarget(link);
    if (!anchor) return;
    onChange === null || onChange === void 0 || onChange(link, activeItem);
    setActiveItem(link);
    intervalRef.current.handleScrollLock = true;
    var scrollContainer = intervalRef.current.scrollContainer;
    var scrollTop = _util_scroll.getScroll(scrollContainer);
    var offsetTop = _util_scroll.getOffsetTop(anchor, scrollContainer);
    var top = scrollTop + offsetTop - targetOffset;
    _util_scroll.scrollTo(top, {
      container: scrollContainer
    }).then(function () {
      intervalRef.current.handleScrollLock = false;
    });
  };
  var handleClick = function handleClick(item, e) {
    onClick === null || onClick === void 0 || onClick(_objectSpread({
      e: e
    }, item));
    handleScrollTo(item.href);
  };
  React.useEffect(function () {
    var _anchorEl$current$que, _anchorEl$current;
    var pointEl = (_anchorEl$current$que = (_anchorEl$current = anchorEl.current).querySelector) === null || _anchorEl$current$que === void 0 ? void 0 : _anchorEl$current$que.call(_anchorEl$current, ".".concat(classPrefix, "-is-active>a"));
    if (!pointEl) {
      setCursorStyle(null);
    } else {
      var top = pointEl.offsetTop,
        height = pointEl.offsetHeight;
      setCursorStyle({
        top: "".concat(top, "px"),
        height: "".concat(height, "px"),
        opacity: 1
      });
    }
  }, [activeItem, classPrefix]);
  var handleScroll = React.useCallback(function () {
    var _intervalRef$current = intervalRef.current,
      scrollContainer = _intervalRef$current.scrollContainer,
      handleScrollLock = _intervalRef$current.handleScrollLock;
    if (handleScrollLock) return;
    var items = intervalRef.current.items;
    var filters = [];
    var active = "";
    items.forEach(function (href) {
      var anchor = getAnchorTarget(href);
      if (!anchor) return;
      var top = _util_scroll.getOffsetTop(anchor, scrollContainer);
      if (top <= bounds + targetOffset) {
        filters.push({
          href: href,
          top: top
        });
      }
    });
    if (filters.length) {
      var latest = filters.reduce(function (prev, cur) {
        return prev.top > cur.top ? prev : cur;
      });
      active = latest.href;
    }
    if (active !== activeItem) {
      onChange === null || onChange === void 0 || onChange(active, activeItem);
      setActiveItem(active);
    }
  }, [activeItem, bounds, onChange, targetOffset]);
  React.useEffect(function () {
    intervalRef.current.scrollContainer = _util_scroll.getScrollContainer(container);
    var scrollContainer = intervalRef.current.scrollContainer;
    handleScroll();
    scrollContainer === null || scrollContainer === void 0 || scrollContainer.addEventListener("scroll", handleScroll);
    return function () {
      scrollContainer === null || scrollContainer === void 0 || scrollContainer.removeEventListener("scroll", handleScroll);
    };
  }, [container, handleScroll]);
  var anchorClass = classNames__default["default"]("".concat(classPrefix, "-anchor"), defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-size-s"), size === "small"), "".concat(classPrefix, "-size-m"), size === "medium"), "".concat(classPrefix, "-size-l"), size === "large"), className);
  var CursorCmp = function CursorCmp() {
    if (isFunction.isFunction(cursor)) return cursor();
    if (isEmpty.isEmpty(cursor)) return /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-anchor__line-cursor")
    });
    return cursor;
  };
  var Cmp = /* @__PURE__ */React__default["default"].createElement(anchor_AnchorContext.AnchorContext.Provider, {
    value: {
      onClick: handleClick,
      activeItem: (getCurrentAnchor === null || getCurrentAnchor === void 0 ? void 0 : getCurrentAnchor(activeItem)) || activeItem,
      registerItem: registerItem,
      unregisterItem: unregisterItem
    }
  }, /* @__PURE__ */React__default["default"].createElement("div", _objectSpread(_objectSpread({}, rest), {}, {
    className: anchorClass,
    ref: anchorEl
  }), /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-anchor__line")
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-anchor__line-cursor-wrapper"),
    style: cursorStyle
  }, CursorCmp())), children));
  return isEmpty.isEmpty(affixProps) ? Cmp : /* @__PURE__ */React__default["default"].createElement(affix_index.Affix, _objectSpread({}, affixProps), Cmp);
}, {
  AnchorItem: anchor_AnchorItem["default"],
  AnchorTarget: anchor_AnchorTarget["default"]
});
Anchor.displayName = "Anchor";

exports["default"] = Anchor;
