// tdesign-react/cjs/timeline/Timeline.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var classNames = require('classnames');
var React = require('react');
var hooks_useConfig = require('../hooks/useConfig.js');
var timeline_TimelineItem = require('./TimelineItem.js');
var timeline_TimelineContext = require('./TimelineContext.js');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var timeline_useAlign = require('./useAlign.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_util/parseTNode.js');
require('../_chunks/dep-b289b0be.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../hooks/useDefaultProps.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-1c586f5c.js');
require('../_chunks/dep-f0e884b9.js');
require('../hooks/useDomRefCallback.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-5c3139cb.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-2e455d91.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-894a5ece.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-63308b84.js');
require('../_chunks/dep-048e4314.js');
require('hoist-non-react-statics');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var Timeline = _util_forwardRefWithStatics["default"](function (props, ref) {
  var _props$theme = props.theme,
    theme = _props$theme === void 0 ? "default" : _props$theme,
    labelAlign = props.labelAlign,
    children = props.children,
    className = props.className,
    style = props.style,
    _props$reverse = props.reverse,
    reverse = _props$reverse === void 0 ? false : _props$reverse,
    _props$layout = props.layout,
    layout = _props$layout === void 0 ? "vertical" : _props$layout,
    _props$mode = props.mode,
    mode = _props$mode === void 0 ? "alternate" : _props$mode;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var renderAlign = timeline_useAlign.useAlign(labelAlign, layout);
  var timelineItems = React__default["default"].Children.toArray(children).filter(function (child) {
    var _child$type;
    return ((_child$type = child.type) === null || _child$type === void 0 ? void 0 : _child$type.displayName) === timeline_TimelineItem["default"].displayName;
  });
  var itemsStatus = React__default["default"].Children.map(timelineItems, function (child) {
    var _child$props;
    return ((_child$props = child.props) === null || _child$props === void 0 ? void 0 : _child$props.dotColor) || "primary";
  });
  var hasLabelItem = timelineItems.some(function (item) {
    var _item$props;
    return !!(item !== null && item !== void 0 && (_item$props = item.props) !== null && _item$props !== void 0 && _item$props.label);
  });
  if (reverse) {
    timelineItems.reverse();
  }
  var itemsCounts = React__default["default"].Children.count(timelineItems);
  var timelineClassName = classNames__default["default"]("".concat(classPrefix, "-timeline"), defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-timeline-").concat(renderAlign), true), "".concat(classPrefix, "-timeline-reverse"), reverse), "".concat(classPrefix, "-timeline-").concat(layout), true), "".concat(classPrefix, "-timeline-label"), hasLabelItem), "".concat(classPrefix, "-timeline-label--").concat(mode), true), className);
  return /* @__PURE__ */React__default["default"].createElement(timeline_TimelineContext["default"].Provider, {
    value: {
      theme: theme,
      reverse: reverse,
      itemsStatus: itemsStatus,
      layout: layout,
      globalAlign: labelAlign,
      mode: mode
    }
  }, /* @__PURE__ */React__default["default"].createElement("ul", {
    className: timelineClassName,
    style: style,
    ref: ref
  }, React__default["default"].Children.map(timelineItems, function (ele, index) {
    var _ele$props;
    return /*#__PURE__*/React__default["default"].cloneElement(ele, {
      index: index,
      className: classNames__default["default"]([ele === null || ele === void 0 || (_ele$props = ele.props) === null || _ele$props === void 0 ? void 0 : _ele$props.className], defineProperty._defineProperty({}, "".concat(classPrefix, "-timeline-item--last"), index === itemsCounts - 1))
    });
  })));
}, {
  Item: timeline_TimelineItem["default"]
});
Timeline.displayName = "Timeline";

exports["default"] = Timeline;
