// tdesign-react/cjs/timeline/TimelineItem.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var hooks_useConfig = require('../hooks/useConfig.js');
var timeline_TimelineContext = require('./TimelineContext.js');
var _util_parseTNode = require('../_util/parseTNode.js');
var timeline_useAlign = require('./useAlign.js');
var loading_index = require('../loading/index.js');
var omit = require('../_chunks/dep-5c3139cb.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-b289b0be.js');
require('../loading/Loading.js');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../hooks/useDefaultProps.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-1c586f5c.js');
require('../_chunks/dep-f0e884b9.js');
require('../hooks/useDomRefCallback.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-2e455d91.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-894a5ece.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-63308b84.js');
require('../_chunks/dep-048e4314.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var DefaultTheme = ["default", "primary", "success", "warning", "error"];
var TimelineItem = function TimelineItem(props) {
  var className = props.className,
    _props$style = props.style,
    style = _props$style === void 0 ? {} : _props$style,
    dot = props.dot,
    _props$dotColor = props.dotColor,
    dotColor = _props$dotColor === void 0 ? "primary" : _props$dotColor,
    labelAlign = props.labelAlign,
    children = props.children,
    index = props.index,
    content = props.content,
    label = props.label,
    _props$loading = props.loading,
    loading = _props$loading === void 0 ? false : _props$loading,
    onClick = props.onClick;
  var _useContext = React.useContext(timeline_TimelineContext["default"]),
    theme = _useContext.theme,
    reverse = _useContext.reverse,
    itemsStatus = _useContext.itemsStatus,
    layout = _useContext.layout,
    globalAlign = _useContext.globalAlign,
    mode = _useContext.mode;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var renderAlign = timeline_useAlign.useAlign(globalAlign, layout);
  var getPositionClassName = function getPositionClassName(index2) {
    var left = layout === "horizontal" ? "top" : "left";
    var right = layout === "horizontal" ? "bottom" : "right";
    if (renderAlign === "alternate") {
      return labelAlign || index2 % 2 === 0 ? "".concat(classPrefix, "-timeline-item-").concat(left) : "".concat(classPrefix, "-timeline-item-").concat(right);
    }
    if (renderAlign === "left" || renderAlign === "top") {
      return "".concat(classPrefix, "-timeline-item-").concat(left);
    }
    if (renderAlign === "right" || renderAlign === "bottom") {
      return "".concat(classPrefix, "-timeline-item-").concat(right);
    }
    return "";
  };
  var dotElement = React.useMemo(function () {
    var ele = _util_parseTNode["default"](dot);
    if (/*#__PURE__*/React__default["default"].isValidElement(ele)) {
      var _ele$props;
      return /*#__PURE__*/React__default["default"].cloneElement(ele, {
        className: classNames__default["default"](ele === null || ele === void 0 || (_ele$props = ele.props) === null || _ele$props === void 0 ? void 0 : _ele$props.className, "".concat(classPrefix, "-timeline-item__dot-content"))
      });
    }
    return ele;
  }, [dot, classPrefix]);
  var handleClick = function handleClick(e) {
    onClick === null || onClick === void 0 || onClick({
      e: e,
      item: omit.omit(props, ["children", "index", "onClick"])
    });
  };
  var itemClassName = classNames__default["default"](defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-timeline-item"), true), "".concat(getPositionClassName(index)), true), className);
  var tailClassName = classNames__default["default"](defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-timeline-item__tail"), true), "".concat(classPrefix, "-timeline-item__tail--theme-").concat(theme), true), "".concat(classPrefix, "-timeline-item__tail--status-").concat(itemsStatus[index]), reverse));
  var dotClassName = classNames__default["default"](defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-timeline-item__dot"), true), "".concat(classPrefix, "-timeline-item__dot--custom"), !!dotElement || !dotElement && loading), "".concat(classPrefix, "-timeline-item__dot--").concat(dotColor), DefaultTheme.includes(dotColor)));
  var labelClassName = classNames__default["default"]("".concat(classPrefix, "-timeline-item__label"), defineProperty._defineProperty({}, "".concat(classPrefix, "-timeline-item__label--").concat(mode), true));
  return /* @__PURE__ */React__default["default"].createElement("li", {
    className: itemClassName,
    style: style,
    onClick: handleClick
  }, mode === "alternate" && label && /* @__PURE__ */React__default["default"].createElement("div", {
    className: labelClassName
  }, label), /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-timeline-item__wrapper")
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: dotClassName,
    style: {
      borderColor: !DefaultTheme.includes(dotColor) && dotColor
    }
  }, !dotElement && loading && /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, {
    size: "12px",
    className: "".concat(classPrefix, "-timeline-item__dot-content")
  }), dotElement), /* @__PURE__ */React__default["default"].createElement("div", {
    className: tailClassName
  })), /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-timeline-item__content")
  }, content || children, mode === "same" && label && /* @__PURE__ */React__default["default"].createElement("div", {
    className: labelClassName
  }, label)));
};
TimelineItem.displayName = "TimelineItem";

exports["default"] = TimelineItem;
