// tdesign-react/cjs/collapse/Collapse.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var toConsumableArray = require('../_chunks/dep-1d0dbfc5.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var hooks_useConfig = require('../hooks/useConfig.js');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var hooks_useControlled = require('../hooks/useControlled.js');
var collapse_CollapsePanel = require('./CollapsePanel.js');
var collapse_CollapseContext = require('./CollapseContext.js');
var collapse_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-32593c54.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('hoist-non-react-statics');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('react-transition-group');
require('../common/FakeArrow.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-f0e884b9.js');
require('../_chunks/dep-17e2e72f.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["defaultExpandAll", "disabled", "expandIconPlacement", "expandOnRowClick", "expandIcon"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Collapse = _util_forwardRefWithStatics["default"](function (originalProps, ref) {
  var props = hooks_useDefaultProps["default"](originalProps, collapse_defaultProps.collapseDefaultProps);
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var componentName = "".concat(classPrefix, "-collapse");
  var borderlessClass = "".concat(classPrefix, "--border-less");
  var defaultExpandAll = props.defaultExpandAll,
    disabled = props.disabled,
    expandIconPlacement = props.expandIconPlacement,
    expandOnRowClick = props.expandOnRowClick,
    expandIcon = props.expandIcon,
    rest = objectWithoutProperties._objectWithoutProperties(props, _excluded);
  var children = rest.children,
    className = rest.className,
    style = rest.style,
    expandMutex = rest.expandMutex,
    borderless = rest.borderless,
    onChange = rest.onChange;
  var _useControlled = hooks_useControlled["default"](props, "value", onChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    collapseValue = _useControlled2[0],
    setCollapseValue = _useControlled2[1];
  var collapseValues = React.useRef(collapseValue);
  React.useEffect(function () {
    collapseValues.current = collapseValue;
  }, [collapseValue]);
  var updateCollapseValue = function updateCollapseValue(value, context) {
    var newValue = [].concat(collapseValues.current || []);
    var index = newValue.indexOf(value);
    if (index >= 0) {
      newValue.splice(index, 1);
    } else if (expandMutex) {
      newValue = [value];
    } else {
      newValue.push(value);
    }
    collapseValues.current = toConsumableArray._toConsumableArray(newValue);
    setCollapseValue(newValue, context);
  };
  var classes = [componentName, defineProperty._defineProperty({}, borderlessClass, !!borderless), className];
  var childrenList = React__default["default"].Children.toArray(children).filter(function (child) {
    return child.type.displayName === collapse_CollapsePanel["default"].displayName;
  });
  var collapsePanelList = function collapsePanelList() {
    return childrenList.map(function (child, index) {
      var key = child.key || String(index);
      var childProps = _objectSpread({
        key: key,
        index: index + 1
      }, child.props);
      return /*#__PURE__*/React__default["default"].cloneElement(child, childProps);
    });
  };
  return /* @__PURE__ */React__default["default"].createElement(collapse_CollapseContext["default"].Provider, {
    value: {
      defaultExpandAll: defaultExpandAll,
      disabled: disabled,
      expandIconPlacement: expandIconPlacement,
      expandOnRowClick: expandOnRowClick,
      expandIcon: expandIcon,
      updateCollapseValue: updateCollapseValue,
      collapseValue: collapseValue
    }
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](classes),
    style: style,
    ref: ref
  }, collapsePanelList()));
}, {
  Panel: collapse_CollapsePanel["default"]
});
Collapse.displayName = "Collapse";

exports["default"] = Collapse;
