// tdesign-react/cjs/collapse/CollapsePanel.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var reactTransitionGroup = require('react-transition-group');
var collapse_CollapseContext = require('./CollapseContext.js');
var common_FakeArrow = require('../common/FakeArrow.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var collapse_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-1c586f5c.js');
require('../_chunks/dep-f0e884b9.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-794b42ce.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var CollapsePanel = function CollapsePanel(props) {
  var _useDefaultProps = hooks_useDefaultProps["default"](props, collapse_defaultProps.collapsePanelDefaultProps),
    value = _useDefaultProps.value,
    disabled = _useDefaultProps.disabled,
    destroyOnCollapse = _useDefaultProps.destroyOnCollapse,
    expandIcon = _useDefaultProps.expandIcon,
    className = _useDefaultProps.className,
    style = _useDefaultProps.style,
    header = _useDefaultProps.header,
    headerRightContent = _useDefaultProps.headerRightContent,
    children = _useDefaultProps.children,
    index = _useDefaultProps.index;
  var _useCollapseContext = collapse_CollapseContext.useCollapseContext(),
    disableAll = _useCollapseContext.disabled,
    defaultExpandAll = _useCollapseContext.defaultExpandAll,
    expandIconPlacement = _useCollapseContext.expandIconPlacement,
    expandOnRowClick = _useCollapseContext.expandOnRowClick,
    expandIconAll = _useCollapseContext.expandIcon,
    updateCollapseValue = _useCollapseContext.updateCollapseValue,
    collapseValue = _useCollapseContext.collapseValue;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var componentName = "".concat(classPrefix, "-collapse-panel");
  var innerValue = value || index;
  var finalExpandIcon = (expandIcon === void 0 ? expandIconAll : expandIcon) || null;
  var headRef = React.useRef(null);
  var iconRef = React.useRef(null);
  var contentRef = React.useRef(null);
  var bodyRef = React.useRef(null);
  var isDisabled = disabled || !!disableAll;
  React.useEffect(function () {
    if (defaultExpandAll) {
      updateCollapseValue(innerValue);
    }
  }, []);
  var isActive = Array.isArray(collapseValue) ? collapseValue.includes(innerValue) : collapseValue === innerValue;
  var classes = classNames__default["default"](componentName, defineProperty._defineProperty({}, "".concat(classPrefix, "-is-disabled"), isDisabled), className);
  var handleClick = function handleClick(e) {
    var canExpand = expandOnRowClick && e.currentTarget === headRef.current || e.currentTarget === iconRef.current;
    if (canExpand && !isDisabled) {
      updateCollapseValue(innerValue, {
        e: e
      });
    }
    e.stopPropagation();
  };
  var renderIcon = function renderIcon() {
    var iconNode = null;
    if (/*#__PURE__*/React__default["default"].isValidElement(finalExpandIcon)) {
      iconNode = finalExpandIcon;
    } else if (finalExpandIcon) {
      iconNode = /* @__PURE__ */React__default["default"].createElement(common_FakeArrow["default"], {
        className: classNames__default["default"]("".concat(componentName, "__icon--default"))
      });
    }
    return iconNode && /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(componentName, "__icon ").concat(componentName, "__icon--").concat(expandIconPlacement, " ").concat(isActive ? "".concat(componentName, "__icon--active") : ""),
      ref: iconRef,
      onClick: handleClick
    }, iconNode);
  };
  var renderHeader = function renderHeader() {
    var cls = ["".concat(componentName, "__header"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-clickable"), expandOnRowClick && !isDisabled)];
    return /* @__PURE__ */React__default["default"].createElement("div", {
      ref: headRef,
      className: classNames__default["default"](cls),
      onClick: handleClick
    }, expandIconPlacement === "left" && renderIcon(), header, /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(componentName, "__header--blank")
    }), headerRightContent, expandIconPlacement === "right" && renderIcon());
  };
  var renderBody = function renderBody() {
    var transitionCallbacks = {
      onEnter: function onEnter() {
        bodyRef.current.style.height = "".concat(contentRef === null || contentRef === void 0 ? void 0 : contentRef.current.clientHeight, "px");
      },
      onEntered: function onEntered() {
        bodyRef.current.style.height = "auto";
      },
      onExit: function onExit() {
        bodyRef.current.style.height = "".concat(contentRef === null || contentRef === void 0 ? void 0 : contentRef.current.clientHeight, "px");
      },
      onExiting: function onExiting() {
        bodyRef.current.style.height = "0px";
      }
    };
    return /* @__PURE__ */React__default["default"].createElement(reactTransitionGroup.CSSTransition, _objectSpread({
      "in": isActive,
      appear: true,
      timeout: 300,
      nodeRef: bodyRef,
      unmountOnExit: destroyOnCollapse
    }, transitionCallbacks), /* @__PURE__ */React__default["default"].createElement("div", {
      style: {
        height: 0
      },
      className: classNames__default["default"]("".concat(componentName, "__body"), "".concat(classPrefix, "-slide-down-enter-active"), defineProperty._defineProperty({}, "".concat(componentName, "__body--collapsed"), !isActive)),
      ref: bodyRef
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(componentName, "__content"),
      ref: contentRef
    }, children)));
  };
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classes,
    style: _objectSpread({}, style)
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(componentName, "__wrapper")
  }, renderHeader(), renderBody()));
};
CollapsePanel.displayName = "CollapsePanel";

exports["default"] = CollapsePanel;
