// tdesign-react/cjs/dropdown/DropdownItem.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var helper = require('../_chunks/dep-b011b5d3.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useDomRefCallback = require('../hooks/useDomRefCallback.js');
var hooks_useRipple = require('../hooks/useRipple.js');
var dropdown_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-1c586f5c.js');
require('../_chunks/dep-f0e884b9.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-794b42ce.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-064e4dba.js');
require('../hooks/useAnimation.js');
require('../_chunks/dep-68f5fe7b.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var DropdownItem = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _useDefaultProps = hooks_useDefaultProps["default"](props, dropdown_defaultProps.dropdownItemDefaultProps),
    children = _useDefaultProps.children,
    className = _useDefaultProps.className,
    active = _useDefaultProps.active,
    disabled = _useDefaultProps.disabled,
    value = _useDefaultProps.value,
    theme = _useDefaultProps.theme,
    prefixIcon = _useDefaultProps.prefixIcon,
    maxColumnWidth = _useDefaultProps.maxColumnWidth,
    minColumnWidth = _useDefaultProps.minColumnWidth,
    onClick = _useDefaultProps.onClick,
    style = _useDefaultProps.style,
    isSubmenu = _useDefaultProps.isSubmenu;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _useDomRefCallback = hooks_useDomRefCallback["default"](),
    _useDomRefCallback2 = slicedToArray._slicedToArray(_useDomRefCallback, 2),
    dropdownItemDom = _useDomRefCallback2[0],
    setRefCurrent = _useDomRefCallback2[1];
  hooks_useRipple["default"](isSubmenu || disabled ? null : (ref === null || ref === void 0 ? void 0 : ref.current) || dropdownItemDom);
  var handleItemClick = function handleItemClick(e) {
    if (disabled) return;
    onClick === null || onClick === void 0 || onClick(value, {
      e: e
    });
  };
  return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement("li", {
    className: classNames__default["default"](className, "".concat(classPrefix, "-dropdown__item--theme-").concat(theme), defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-dropdown__item--active"), active), "".concat(classPrefix, "-dropdown__item--disabled"), disabled)),
    onClick: handleItemClick,
    style: _objectSpread({
      maxWidth: helper.pxCompat(maxColumnWidth),
      minWidth: helper.pxCompat(minColumnWidth)
    }, style),
    ref: ref || setRefCurrent
  }, prefixIcon ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-dropdown__item-icon")
  }, prefixIcon) : null, children));
});
DropdownItem.displayName = "DropdownItem";

exports["default"] = DropdownItem;
