// tdesign-react/cjs/input-number/InputNumber.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var classNames = require('classnames');
var input_index = require('../input/index.js');
var button_index = require('../button/index.js');
var inputNumber_useInputNumber = require('./useInputNumber.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var inputNumber_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../input/Input.js');
require('../_chunks/dep-1c586f5c.js');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../_util/parseTNode.js');
require('../_chunks/dep-b289b0be.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-aff91ab0.js');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../hooks/useLayoutEffect.js');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../locale/LocalReceiver.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../input/defaultProps.js');
require('../input/InputGroup.js');
require('../input/useLengthLimit.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-f0e884b9.js');
require('../button/Button.js');
require('../hooks/useDomRefCallback.js');
require('../hooks/useRipple.js');
require('../_chunks/dep-064e4dba.js');
require('../hooks/useAnimation.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../common/Portal.js');
require('react-dom');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../button/defaultProps.js');
require('../_chunks/dep-04deac0d.js');
require('../_chunks/dep-8f62fb96.js');
require('../hooks/useCommonClassName.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function TdInputNumber(originalProps, ref) {
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      ChevronDownIcon: tdesignIconsReact.ChevronDownIcon,
      RemoveIcon: tdesignIconsReact.RemoveIcon,
      ChevronUpIcon: tdesignIconsReact.ChevronUpIcon,
      AddIcon: tdesignIconsReact.AddIcon
    }),
    ChevronDownIcon = _useGlobalIcon.ChevronDownIcon,
    RemoveIcon = _useGlobalIcon.RemoveIcon,
    ChevronUpIcon = _useGlobalIcon.ChevronUpIcon,
    AddIcon = _useGlobalIcon.AddIcon;
  var props = hooks_useDefaultProps["default"](originalProps, inputNumber_defaultProps.inputNumberDefaultProps);
  var _useInputNumber = inputNumber_useInputNumber["default"](props),
    classPrefix = _useInputNumber.classPrefix,
    wrapClasses = _useInputNumber.wrapClasses,
    addClasses = _useInputNumber.addClasses,
    reduceClasses = _useInputNumber.reduceClasses,
    listeners = _useInputNumber.listeners,
    isError = _useInputNumber.isError,
    inputRef = _useInputNumber.inputRef,
    userInput = _useInputNumber.userInput,
    handleAdd = _useInputNumber.handleAdd,
    handleReduce = _useInputNumber.handleReduce,
    onInnerInputChange = _useInputNumber.onInnerInputChange;
  var wrapRef = React.useRef(null);
  var status = isError ? "error" : props.status;
  var addIcon = props.theme === "column" ? /* @__PURE__ */React__default["default"].createElement(ChevronUpIcon, {
    size: props.size
  }) : /* @__PURE__ */React__default["default"].createElement(AddIcon, {
    size: props.size
  });
  var reduceIcon = props.theme === "column" ? /* @__PURE__ */React__default["default"].createElement(ChevronDownIcon, {
    size: props.size
  }) : /* @__PURE__ */React__default["default"].createElement(RemoveIcon, {
    size: props.size
  });
  React.useImperativeHandle(ref, function () {
    return {
      currentElement: wrapRef.current,
      inputElement: inputRef.current
    };
  });
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](wrapClasses, props.className),
    style: props.style,
    ref: wrapRef
  }, props.theme !== "normal" && /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    className: reduceClasses,
    disabled: props.disabled,
    onClick: handleReduce,
    variant: "outline",
    shape: "square",
    icon: reduceIcon
  }), /* @__PURE__ */React__default["default"].createElement(input_index.Input, _objectSpread(_objectSpread({
    ref: inputRef,
    autocomplete: "off",
    disabled: props.disabled,
    readOnly: props.readOnly || props.readonly,
    placeholder: props.placeholder,
    autoWidth: props.autoWidth,
    align: props.align || (props.theme === "row" ? "center" : void 0),
    status: status,
    label: props.label,
    suffix: props.suffix,
    value: userInput,
    onChange: onInnerInputChange,
    size: props.size
  }, listeners), props.inputProps || {})), props.theme !== "normal" && /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    className: addClasses,
    disabled: props.disabled,
    onClick: handleAdd,
    variant: "outline",
    shape: "square",
    icon: addIcon
  }), props.tips && /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-input__tips"), "".concat(classPrefix, "-input__tips--").concat(status))
  }, props.tips));
}
var InputNumber = /*#__PURE__*/React.forwardRef(TdInputNumber);
InputNumber.displayName = "InputNumber";

exports["default"] = InputNumber;
