// tdesign-react/cjs/tabs/TabNavItem.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var hooks_useRipple = require('../hooks/useRipple.js');
var hooks_useDomRefCallback = require('../hooks/useDomRefCallback.js');
var _util_noop = require('../_util/noop.js');
var tabs_useTabClass = require('./useTabClass.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-064e4dba.js');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../hooks/useAnimation.js');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var TabNavItem = function TabNavItem(props) {
  var label = props.label,
    removable = props.removable,
    isActive = props.isActive,
    _props$onClick = props.onClick,
    onClick = _props$onClick === void 0 ? _util_noop["default"] : _props$onClick,
    theme = props.theme,
    placement = props.placement,
    _props$onRemove = props.onRemove,
    onRemove = _props$onRemove === void 0 ? _util_noop["default"] : _props$onRemove,
    value = props.value,
    _props$size = props.size,
    size = _props$size === void 0 ? "medium" : _props$size,
    _props$disabled = props.disabled,
    disabled = _props$disabled === void 0 ? false : _props$disabled,
    index = props.index,
    _props$onTabRemove = props.onTabRemove,
    onTabRemove = _props$onTabRemove === void 0 ? _util_noop["default"] : _props$onTabRemove,
    innerRef = props.innerRef,
    dragProps = props.dragProps;
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      CloseIcon: tdesignIconsReact.CloseIcon
    }),
    CloseIcon = _useGlobalIcon.CloseIcon;
  var isCard = theme === "card";
  var _useTabClass = tabs_useTabClass.useTabClass(),
    tdTabsClassGenerator = _useTabClass.tdTabsClassGenerator,
    tdClassGenerator = _useTabClass.tdClassGenerator,
    tdSizeClassGenerator = _useTabClass.tdSizeClassGenerator;
  var _useDomRefCallback = hooks_useDomRefCallback["default"](),
    _useDomRefCallback2 = slicedToArray._slicedToArray(_useDomRefCallback, 2),
    navItemDom = _useDomRefCallback2[0],
    setRefCurrent = _useDomRefCallback2[1];
  hooks_useRipple["default"](navItemDom);
  return /* @__PURE__ */React__default["default"].createElement("div", _objectSpread(_objectSpread({}, dragProps), {}, {
    ref: innerRef,
    onClick: disabled ? _util_noop["default"] : onClick,
    className: classNames__default["default"](tdTabsClassGenerator("nav-item"), isCard ? tdTabsClassGenerator("nav--card") : "", tdSizeClassGenerator(size), isActive ? tdClassGenerator("is-active") : "", tdClassGenerator("is-".concat(placement)), disabled ? tdClassGenerator("is-disabled") : "", props.className)
  }), isCard ? /* @__PURE__ */React__default["default"].createElement("span", {
    className: classNames__default["default"](tdTabsClassGenerator("nav-item-text-wrapper"))
  }, label) : /* @__PURE__ */React__default["default"].createElement("div", {
    ref: setRefCurrent,
    className: classNames__default["default"](tdTabsClassGenerator("nav-item-wrapper"))
  }, /* @__PURE__ */React__default["default"].createElement("span", {
    className: classNames__default["default"](tdTabsClassGenerator("nav-item-text-wrapper"))
  }, label)), removable && !disabled ? /* @__PURE__ */React__default["default"].createElement("span", {
    className: classNames__default["default"]("remove-btn"),
    onClick: function onClick(e) {
      e.stopPropagation();
      onRemove({
        value: value,
        e: e
      });
      onTabRemove({
        value: value,
        e: e,
        index: index
      });
    }
  }, /* @__PURE__ */React__default["default"].createElement(CloseIcon, null)) : null);
};

exports["default"] = TabNavItem;
