// tdesign-react/cjs/tabs/Tabs.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var tabs_TabNav = require('./TabNav.js');
var tabs_useTabClass = require('./useTabClass.js');
var tabs_TabPanel = require('./TabPanel.js');
var tabs_defaultProps = require('./defaultProps.js');
var hooks_useDragSorter = require('../hooks/useDragSorter.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var hooks_useControlled = require('../hooks/useControlled.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('hoist-non-react-statics');
require('tdesign-icons-react');
require('../_util/noop.js');
require('./TabNavItem.js');
require('../hooks/useRipple.js');
require('../_chunks/dep-064e4dba.js');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../hooks/useAnimation.js');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../hooks/useDomRefCallback.js');
require('../hooks/useGlobalIcon.js');
require('./TabBar.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../_chunks/dep-876d090e.js');
require('../hooks/useResizeObserver.js');
require('../hooks/useLayoutEffect.js');
require('../_util/parseTNode.js');
require('../_chunks/dep-b289b0be.js');
require('../_chunks/dep-5c3139cb.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-2e455d91.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-894a5ece.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-63308b84.js');
require('../_chunks/dep-048e4314.js');
require('../hooks/useEventCallback.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Tabs = _util_forwardRefWithStatics["default"](function (originalProps, ref) {
  var props = hooks_useDefaultProps["default"](originalProps, tabs_defaultProps.tabsDefaultProps);
  var children = props.children,
    list = props.list,
    placement = props.placement,
    dragSort = props.dragSort,
    className = props.className,
    style = props.style,
    onRemove = props.onRemove;
  var _useControlled = hooks_useControlled["default"](props, "value", props.onChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    value = _useControlled2[0],
    onChange = _useControlled2[1];
  var _useTabClass = tabs_useTabClass.useTabClass(),
    tdTabsClassPrefix = _useTabClass.tdTabsClassPrefix,
    tdTabsClassGenerator = _useTabClass.tdTabsClassGenerator,
    tdClassGenerator = _useTabClass.tdClassGenerator;
  var targetClassNameRegExpStr = "^".concat(tdTabsClassPrefix, "(__nav-item|__nav-item-wrapper|__nav-item-text-wrapper)");
  var _useDragSorter = hooks_useDragSorter["default"](_objectSpread(_objectSpread({}, props), {}, {
      sortOnDraggable: dragSort,
      onDragOverCheck: {
        x: true,
        targetClassNameRegExp: new RegExp(targetClassNameRegExpStr)
      }
    })),
    getDragProps = _useDragSorter.getDragProps;
  var memoChildren = React__default["default"].useMemo(function () {
    if (!list || list.length === 0) {
      return children;
    }
    return list.map(function (panelProps) {
      return /* @__PURE__ */React__default["default"].createElement(tabs_TabPanel["default"], _objectSpread({
        key: panelProps.value
      }, panelProps));
    });
  }, [children, list]);
  var itemList = React__default["default"].Children.map(memoChildren, function (child) {
    if (child && child.type === tabs_TabPanel["default"]) {
      return child.props;
    }
    return null;
  });
  var handleChange = React__default["default"].useCallback(function (v) {
    onChange === null || onChange === void 0 || onChange(v);
  }, [onChange]);
  var headerNode = React__default["default"].useMemo(function () {
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: classNames__default["default"](tdTabsClassGenerator("header"), tdClassGenerator("is-".concat(placement)))
    }, /* @__PURE__ */React__default["default"].createElement(tabs_TabNav["default"], _objectSpread(_objectSpread({}, props), {}, {
      getDragProps: getDragProps,
      activeValue: value,
      onRemove: onRemove,
      itemList: itemList,
      onChange: handleChange
    })));
  }, [props, getDragProps, value, onRemove, itemList, handleChange, placement, tdTabsClassGenerator, tdClassGenerator]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: ref,
    className: classNames__default["default"](tdTabsClassPrefix, className),
    style: style
  }, placement !== "bottom" ? headerNode : null, /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](tdTabsClassGenerator("content"), tdClassGenerator("is-".concat(placement)))
  }, React__default["default"].Children.map(memoChildren, function (child) {
    if (child && child.type === tabs_TabPanel["default"]) {
      return /* @__PURE__ */React__default["default"].createElement(tabs_TabPanel["default"], _objectSpread(_objectSpread({}, child.props), {}, {
        isActive: child.props.value === value
      }));
    }
    return null;
  })), placement === "bottom" ? headerNode : null);
}, {
  TabPanel: tabs_TabPanel["default"]
});
Tabs.displayName = "Tabs";

exports["default"] = Tabs;
