// tdesign-react/cjs/qrcode/QRCode.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var qrcode_QRCodeCanvas = require('./QRCodeCanvas.js');
var qrcode_QRCodeSVG = require('./QRCodeSVG.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var qrcode_defaultProps = require('./defaultProps.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var qrcode_QRCodeStatus = require('./QRCodeStatus.js');
var hooks_useVariables = require('../hooks/useVariables.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-2d6cca8e.js');
require('../_chunks/dep-427fb3a1.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-17e2e72f.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('tdesign-icons-react');
require('../loading/index.js');
require('../loading/Loading.js');
require('../_chunks/dep-68f5fe7b.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-f0e884b9.js');
require('../hooks/useDomRefCallback.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../_chunks/dep-dd0f5f6b.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../_chunks/dep-876d090e.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["className", "value", "borderless", "iconSize", "color", "bgColor", "style", "icon", "size", "type", "status", "onRefresh", "statusRender"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var QRCode = function QRCode(props) {
  var _iconSize$height, _iconSize$width, _style$width, _style$height;
  var _useDefaultProps = hooks_useDefaultProps["default"](props, qrcode_defaultProps.qRCodeDefaultProps),
    className = _useDefaultProps.className,
    value = _useDefaultProps.value,
    borderless = _useDefaultProps.borderless,
    iconSize = _useDefaultProps.iconSize,
    color = _useDefaultProps.color,
    bgColor = _useDefaultProps.bgColor,
    style = _useDefaultProps.style,
    icon = _useDefaultProps.icon,
    size = _useDefaultProps.size,
    type = _useDefaultProps.type,
    status = _useDefaultProps.status,
    onRefresh = _useDefaultProps.onRefresh,
    statusRender = _useDefaultProps.statusRender,
    rest = objectWithoutProperties._objectWithoutProperties(_useDefaultProps, _excluded);
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("qrcode"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 1),
    locale = _useLocaleReceiver2[0];
  var _useVariables = hooks_useVariables["default"]({
      themeFgColor: "--td-text-color-primary",
      themeBgColor: "--td-bg-color-specialcomponent"
    }),
    themeFgColor = _useVariables.themeFgColor,
    themeBgColor = _useVariables.themeBgColor;
  var finalBgColor = React.useMemo(function () {
    return bgColor || themeBgColor || "transparent";
  }, [bgColor, themeBgColor]);
  if (!value) {
    return null;
  }
  var imageSettings = {
    src: icon,
    x: void 0,
    y: void 0,
    height: typeof iconSize === "number" ? iconSize : (_iconSize$height = iconSize === null || iconSize === void 0 ? void 0 : iconSize.height) !== null && _iconSize$height !== void 0 ? _iconSize$height : 40,
    width: typeof iconSize === "number" ? iconSize : (_iconSize$width = iconSize === null || iconSize === void 0 ? void 0 : iconSize.width) !== null && _iconSize$width !== void 0 ? _iconSize$width : 40,
    excavate: true,
    crossOrigin: "anonymous"
  };
  var qrCodeProps = _objectSpread({
    value: value,
    size: size,
    bgColor: finalBgColor,
    fgColor: color || themeFgColor,
    imageSettings: icon ? imageSettings : void 0
  }, rest);
  var cls = classNames__default["default"]("".concat(classPrefix, "-qrcode"), defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-borderless"), borderless), "".concat(classPrefix, "-qrcode-svg"), type === "svg"), className);
  var mergedStyle = _objectSpread(_objectSpread({
    backgroundColor: finalBgColor
  }, style), {}, {
    width: (_style$width = style === null || style === void 0 ? void 0 : style.width) !== null && _style$width !== void 0 ? _style$width : size,
    height: (_style$height = style === null || style === void 0 ? void 0 : style.height) !== null && _style$height !== void 0 ? _style$height : size
  });
  return /* @__PURE__ */React__default["default"].createElement("div", _objectSpread(_objectSpread({}, rest), {}, {
    className: cls,
    style: mergedStyle
  }), status !== "active" && /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-mask")
  }, /* @__PURE__ */React__default["default"].createElement(qrcode_QRCodeStatus["default"], {
    classPrefix: classPrefix,
    locale: locale,
    status: status,
    onRefresh: onRefresh,
    statusRender: statusRender
  })), type === "canvas" ? /* @__PURE__ */React__default["default"].createElement(qrcode_QRCodeCanvas.QRCodeCanvas, _objectSpread(_objectSpread({}, qrCodeProps), {}, {
    size: size
  })) : /* @__PURE__ */React__default["default"].createElement(qrcode_QRCodeSVG.QRCodeSVG, _objectSpread(_objectSpread({}, qrCodeProps), {}, {
    size: size
  })));
};
QRCode.displayName = "QRCode";

exports["default"] = QRCode;
