// tdesign-react/cjs/qrcode/QRCodeStatus.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var loading_index = require('../loading/index.js');
require('../loading/Loading.js');
require('../_chunks/dep-2b5ba78d.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('classnames');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-ecfe656e.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../hooks/useLayoutEffect.js');
require('../hooks/useDefaultProps.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-1c586f5c.js');
require('../_chunks/dep-f0e884b9.js');
require('../hooks/useDomRefCallback.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var defaultSpin = /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, {
  size: "32px"
});
function QRcodeStatus(_ref) {
  var locale = _ref.locale,
    classPrefix = _ref.classPrefix,
    onRefresh = _ref.onRefresh,
    statusRender = _ref.statusRender,
    status = _ref.status;
  var defaultExpiredNode = React.useMemo(function () {
    return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement("p", {
      className: "".concat(classPrefix, "-expired__text")
    }, locale === null || locale === void 0 ? void 0 : locale.expiredText), onRefresh && /* @__PURE__ */React__default["default"].createElement("p", {
      className: "".concat(classPrefix, "-expired__button"),
      onClick: onRefresh
    }, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.RefreshIcon, {
      size: 16
    }), locale === null || locale === void 0 ? void 0 : locale.refreshText));
  }, [classPrefix, locale === null || locale === void 0 ? void 0 : locale.expiredText, locale === null || locale === void 0 ? void 0 : locale.refreshText, onRefresh]);
  var defaultScannedNode = React.useMemo(function () {
    return /* @__PURE__ */React__default["default"].createElement("p", {
      className: "".concat(classPrefix, "-scanned")
    }, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.CheckCircleFilledIcon, {
      size: 16,
      className: "".concat(classPrefix, "-scanned__icon")
    }), locale === null || locale === void 0 ? void 0 : locale.scannedText);
  }, [classPrefix, locale === null || locale === void 0 ? void 0 : locale.scannedText]);
  var defaultNodes = React.useMemo(function () {
    return {
      expired: defaultExpiredNode,
      loading: defaultSpin,
      scanned: defaultScannedNode
    };
  }, [defaultExpiredNode, defaultScannedNode]);
  var defaultStatusRender = React.useCallback(function (info) {
    return defaultNodes[info.status];
  }, [defaultNodes]);
  var mergedStatusRender = statusRender !== null && statusRender !== void 0 ? statusRender : defaultStatusRender;
  return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, mergedStatusRender({
    status: status,
    onRefresh: onRefresh
  }) || null);
}

exports["default"] = QRcodeStatus;
