// tdesign-react/cjs/breadcrumb/BreadcrumbItem.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var hooks_useCommonClassName = require('../hooks/useCommonClassName.js');
var breadcrumb_BreadcrumbContext = require('./BreadcrumbContext.js');
var _util_parseTNode = require('../_util/parseTNode.js');
var breadcrumb_defaultProps = require('./defaultProps.js');
var tooltip_index = require('../tooltip/index.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var _util_scroll = require('../_util/scroll.js');
var isFunction = require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-b289b0be.js');
require('../tooltip/Tooltip.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../hooks/useAnimation.js');
require('../hooks/useAttach.js');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../hooks/usePopper.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/listener.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');
require('../popup/utils/transition.js');
require('../popup/PopupPlugin.js');
require('../_chunks/dep-e37cb830.js');
require('../_util/react-render.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../tooltip/defaultProps.js');
require('../tooltip/TooltipLite.js');
require('../hooks/useSwitch.js');
require('../hooks/usePersistFn.js');
require('../_chunks/dep-2a85392e.js');
require('raf');
require('../_util/easing.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["children", "separator", "disabled", "maxItemWidth", "maxWidth", "icon", "href", "to", "target", "router", "replace", "className", "content", "onClick", "tooltipProps", "readOnly"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var BreadcrumbItem = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var commonClassNames = hooks_useCommonClassName["default"]();
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      ChevronRightIcon: tdesignIconsReact.ChevronRightIcon
    }),
    ChevronRightIcon = _useGlobalIcon.ChevronRightIcon;
  var _useDefaultProps = hooks_useDefaultProps["default"](props, breadcrumb_defaultProps.breadcrumbItemDefaultProps),
    children = _useDefaultProps.children,
    separator = _useDefaultProps.separator,
    disabled = _useDefaultProps.disabled,
    maxItemWidth = _useDefaultProps.maxItemWidth,
    maxWidth = _useDefaultProps.maxWidth,
    icon = _useDefaultProps.icon,
    href = _useDefaultProps.href,
    to = _useDefaultProps.to,
    target = _useDefaultProps.target,
    router = _useDefaultProps.router,
    replace = _useDefaultProps.replace,
    className = _useDefaultProps.className,
    content = _useDefaultProps.content,
    onClick = _useDefaultProps.onClick,
    tooltipProps = _useDefaultProps.tooltipProps,
    readonly = _useDefaultProps.readOnly,
    restProps = objectWithoutProperties._objectWithoutProperties(_useDefaultProps, _excluded);
  var _useContext = React.useContext(breadcrumb_BreadcrumbContext.BreadcrumbContext),
    maxItemWidthInContext = _useContext.maxItemWidthInContext,
    separatorInContext = _useContext.separator;
  var breadcrumbText = React.useRef(null);
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    isCutOff = _useState2[0],
    setIsCutOff = _useState2[1];
  var breadcrumbItemClassNames = classNames__default["default"]("".concat(classPrefix, "-breadcrumb__item"));
  var textWrapperClassName = "".concat(classPrefix, "-breadcrumb__inner");
  var textClassName = "".concat(classPrefix, "-breadcrumb__inner-text");
  var textClassNames = classNames__default["default"](defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-breadcrumb--text-overflow"), !readonly), commonClassNames.STATUS.disabled, disabled));
  var separatorClassName = "".concat(classPrefix, "-breadcrumb__separator");
  var linkClassName = "".concat(classPrefix, "-link");
  var maxWidthForItem = React.useMemo(function () {
    return {
      maxWidth: maxWidth || maxItemWidth || maxItemWidthInContext || "120px"
    };
  }, [maxItemWidth, maxWidth, maxItemWidthInContext]);
  React.useEffect(function () {
    if (!breadcrumbText.current) return;
    return setIsCutOff(_util_scroll.isNodeOverflow(breadcrumbText.current));
  }, [breadcrumbText]);
  var textContent = /* @__PURE__ */React__default["default"].createElement("span", {
    className: textWrapperClassName,
    style: maxWidthForItem
  }, isFunction.isFunction(icon) ? icon() : icon, /* @__PURE__ */React__default["default"].createElement("span", {
    ref: breadcrumbText,
    className: textClassName
  }, children || content));
  var itemContent = /* @__PURE__ */React__default["default"].createElement("span", {
    className: textClassNames
  }, textContent);
  if ((href || to) && !disabled) {
    var handleRouting = function handleRouting() {
      if (href || !router) return;
      replace ? router.replace(to) : router.push(to);
    };
    itemContent = /* @__PURE__ */React__default["default"].createElement("a", {
      className: classNames__default["default"](textClassNames, linkClassName),
      href: href,
      target: target,
      onClick: handleRouting
    }, textContent);
  }
  var separatorInProps = _util_parseTNode["default"](separator);
  var separatorContent = separatorInProps || separatorInContext || /* @__PURE__ */React__default["default"].createElement(ChevronRightIcon, {
    style: {
      opacity: ".5"
    }
  });
  var handleClick = function handleClick(e) {
    if (disabled) return;
    onClick === null || onClick === void 0 || onClick(e);
  };
  return /* @__PURE__ */React__default["default"].createElement("div", _objectSpread({
    className: classNames__default["default"](breadcrumbItemClassNames, className),
    ref: ref,
    onClick: handleClick
  }, restProps), isCutOff ? /* @__PURE__ */React__default["default"].createElement(tooltip_index.TooltipLite, _objectSpread({
    content: children || content
  }, tooltipProps), itemContent) : itemContent, /* @__PURE__ */React__default["default"].createElement("span", {
    className: separatorClassName
  }, separatorContent));
});
BreadcrumbItem.displayName = "BreadcrumbItem";

exports["default"] = BreadcrumbItem;
