// tdesign-react/cjs/breadcrumb/Breadcrumb.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var toConsumableArray = require('../_chunks/dep-1d0dbfc5.js');
var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var objectWithoutProperties = require('../_chunks/dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var log = require('../_chunks/dep-b289b0be.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var _util_forwardRefWithStatics = require('../_util/forwardRefWithStatics.js');
var breadcrumb_BreadcrumbItem = require('./BreadcrumbItem.js');
var breadcrumb_BreadcrumbContext = require('./BreadcrumbContext.js');
var breadcrumb_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var isFunction = require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('hoist-non-react-statics');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../hooks/useGlobalIcon.js');
require('../hooks/useCommonClassName.js');
require('../_util/parseTNode.js');
require('../tooltip/index.js');
require('../tooltip/Tooltip.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../hooks/useAnimation.js');
require('../hooks/useAttach.js');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-b4442734.js');
require('../_chunks/dep-876d090e.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-eab994e8.js');
require('../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/listener.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');
require('../popup/utils/transition.js');
require('../popup/PopupPlugin.js');
require('../_chunks/dep-e37cb830.js');
require('../_util/react-render.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../tooltip/defaultProps.js');
require('../tooltip/TooltipLite.js');
require('../hooks/useSwitch.js');
require('../hooks/usePersistFn.js');
require('../_chunks/dep-2a85392e.js');
require('../_util/scroll.js');
require('raf');
require('../_util/easing.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["children", "options", "separator", "maxItemWidth", "className", "maxItems", "itemsAfterCollapse", "itemsBeforeCollapse", "ellipsis"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Breadcrumb = _util_forwardRefWithStatics["default"](function (props, ref) {
  var _useDefaultProps = hooks_useDefaultProps["default"](props, breadcrumb_defaultProps.breadcrumbDefaultProps),
    children = _useDefaultProps.children,
    options = _useDefaultProps.options,
    separator = _useDefaultProps.separator,
    maxItemWidth = _useDefaultProps.maxItemWidth,
    className = _useDefaultProps.className,
    maxItems = _useDefaultProps.maxItems,
    itemsAfterCollapse = _useDefaultProps.itemsAfterCollapse,
    itemsBeforeCollapse = _useDefaultProps.itemsBeforeCollapse,
    ellipsis = _useDefaultProps.ellipsis,
    restProps = objectWithoutProperties._objectWithoutProperties(_useDefaultProps, _excluded);
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var renderOptionsToItems = React.useCallback(function (options2) {
    if (!options2 || !options2.length) {
      return void 0;
    }
    return options2.map(function (option, index) {
      var targetProps = {};
      if (option.target) {
        targetProps.target = option.target;
      }
      return /* @__PURE__ */React__default["default"].createElement(breadcrumb_BreadcrumbItem["default"], _objectSpread(_objectSpread({}, targetProps), {}, {
        key: index,
        maxWidth: option.maxWidth,
        disabled: option.disabled,
        href: option.href,
        to: option.to,
        router: option.router,
        replace: option.replace,
        separator: separator,
        maxItemWidth: maxItemWidth,
        icon: option.icon,
        content: option.content
      }), option.content || option.children);
    });
  }, [maxItemWidth, separator]);
  var renderEllipsis = React.useCallback(function (items) {
    if (!ellipsis) {
      return /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.EllipsisIcon, null);
    }
    if (isFunction.isFunction(ellipsis)) {
      return ellipsis({
        items: items,
        separator: separator
      });
    }
    return ellipsis;
  }, [ellipsis, separator]);
  var content = React.useMemo(function () {
    var _items;
    var items = renderOptionsToItems(options) || React.Children.toArray(children);
    var itemsCount = ((_items = items) === null || _items === void 0 ? void 0 : _items.length) || 0;
    if (!items || !maxItems) {
      return items;
    }
    if (maxItems >= itemsCount) {
      return items;
    }
    if (maxItems && (!itemsBeforeCollapse || !itemsAfterCollapse)) {
      log.log.error("Breadcrumb", "\u9700\u8981\u8BBE\u7F6E itemsBeforeCollapse \u548C itemsAfterCollapse \u5C5E\u6027\u6765\u63A7\u5236\u7701\u7565\u53F7\u524D\u540E\u7684\u663E\u793A\u9879\u6570\u3002");
      return items;
    }
    if (maxItems && maxItems < (itemsAfterCollapse || 0) + (itemsBeforeCollapse || 0)) {
      log.log.error("Breadcrumb", "maxItems={".concat(maxItems, "} \u5FC5\u987B\u5927\u4E8E\u6216\u7B49\u4E8E itemsBeforeCollapse={").concat(itemsBeforeCollapse, "} + itemsAfterCollapse={").concat(itemsAfterCollapse, "}"));
      return items;
    }
    var startCount = itemsBeforeCollapse;
    var endCount = itemsAfterCollapse;
    var collapsedItems = items.slice(startCount, itemsCount - endCount);
    var beforeItems = items.slice(0, startCount);
    var afterItems = items.slice(itemsCount - endCount);
    var ellipsis2 = renderEllipsis(collapsedItems.map(function (item) {
      return item.props;
    }));
    items = [].concat(toConsumableArray._toConsumableArray(beforeItems), [/* @__PURE__ */React__default["default"].createElement(breadcrumb_BreadcrumbItem["default"], {
      key: "ellipsis",
      readOnly: true
    }, ellipsis2)], toConsumableArray._toConsumableArray(afterItems));
    return items;
  }, [renderOptionsToItems, options, children, maxItems, itemsBeforeCollapse, itemsAfterCollapse, renderEllipsis]);
  return /* @__PURE__ */React__default["default"].createElement(breadcrumb_BreadcrumbContext.BreadcrumbContext.Provider, {
    value: {
      maxItemWidthInContext: maxItemWidth,
      separator: separator
    }
  }, /* @__PURE__ */React__default["default"].createElement("div", _objectSpread({
    ref: ref,
    className: classNames__default["default"]("".concat(classPrefix, "-breadcrumb"), className)
  }, restProps), content));
}, {
  BreadcrumbItem: breadcrumb_BreadcrumbItem["default"]
});
Breadcrumb.displayName = "Breadcrumb";

exports["default"] = Breadcrumb;
