// tdesign-react/lib/descriptions/Row.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-94a7af64.js';
import React, { useContext } from 'react';
import useConfig from '../hooks/useConfig.js';
import { DescriptionsContext } from './DescriptionsContext.js';
import { useLocaleReceiver } from '../locale/LocalReceiver.js';
import '../_chunks/dep-0054b574.js';
import '../_chunks/dep-8e8ac0d4.js';
import '../config-provider/ConfigContext.js';
import '../_chunks/dep-5e72c8b9.js';
import '../_chunks/dep-43f7da7b.js';
import 'lodash-es';
import '../_chunks/dep-923c86e8.js';
import '../_chunks/dep-03cf63f2.js';
import 'dayjs';
import '../config-provider/ConfigProvider.js';

var Row = function Row(props) {
  var row = props.row;
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix;
  var _useLocaleReceiver = useLocaleReceiver("descriptions"),
    _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
    locale = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var descriptionsContext = useContext(DescriptionsContext);
  var COMPONENT_NAME = "".concat(classPrefix, "-descriptions");
  var label = function label(node) {
    var layout = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "horizontal";
    var rowKey = arguments.length > 2 ? arguments[2] : undefined;
    var span = node.span;
    var labelSpan = layout === "horizontal" ? 1 : span;
    return /* @__PURE__ */React.createElement("td", {
      key: rowKey,
      colSpan: labelSpan,
      className: "".concat(COMPONENT_NAME, "__label"),
      style: descriptionsContext.labelStyle
    }, node.label, descriptionsContext.colon && t(locale.colonText));
  };
  var content = function content(node) {
    var layout = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "horizontal";
    var rowKey = arguments.length > 2 ? arguments[2] : undefined;
    var span = node.span;
    var contentSpan = span > 1 && layout === "horizontal" ? span * 2 - 1 : span;
    return /* @__PURE__ */React.createElement("td", {
      key: rowKey,
      colSpan: contentSpan,
      className: "".concat(COMPONENT_NAME, "__content"),
      style: descriptionsContext.contentStyle
    }, node.content);
  };
  var hh = function hh() {
    return /* @__PURE__ */React.createElement("tr", null, row.map(function (node, i) {
      return /* @__PURE__ */React.createElement(React.Fragment, {
        key: i
      }, label(node), content(node));
    }));
  };
  var hv = function hv() {
    return /* @__PURE__ */React.createElement(React.Fragment, null, /* @__PURE__ */React.createElement("tr", null, row.map(function (node, i) {
      return label(node, "vertical", "top_".concat(i));
    })), /* @__PURE__ */React.createElement("tr", null, row.map(function (node, i) {
      return content(node, "vertical", "bottom_".concat(i));
    })));
  };
  var vh = function vh() {
    return /* @__PURE__ */React.createElement(React.Fragment, null, row.map(function (node, i) {
      return /* @__PURE__ */React.createElement("tr", {
        key: i
      }, label(node), content(node));
    }));
  };
  var vv = function vv() {
    return /* @__PURE__ */React.createElement(React.Fragment, null, row.map(function (node, i) {
      return /* @__PURE__ */React.createElement(React.Fragment, {
        key: i
      }, /* @__PURE__ */React.createElement("tr", null, label(node)), /* @__PURE__ */React.createElement("tr", null, content(node)));
    }));
  };
  if (descriptionsContext.layout === "horizontal") {
    if (descriptionsContext.itemLayout === "horizontal") {
      return hh();
    }
    return hv();
  }
  if (descriptionsContext.itemLayout === "horizontal") {
    return vh();
  }
  return vv();
};
Row.displayName = "DescriptionsRow";

export { Row as default };
