// tdesign-react/cjs/date-picker/DateRangePicker.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var toConsumableArray = require('../_chunks/dep-1d0dbfc5.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var dayjs = require('dayjs');
var format = require('../_chunks/dep-06ac19a1.js');
var utils = require('../_chunks/dep-8227db70.js');
var log = require('../_chunks/dep-b289b0be.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var hooks_useLatest = require('../hooks/useLatest.js');
var hooks_useUpdateEffect = require('../hooks/useUpdateEffect.js');
var rangeInput_index = require('../range-input/index.js');
var datePicker_defaultProps = require('./defaultProps.js');
var datePicker_hooks_useRange = require('./hooks/useRange.js');
var datePicker_panel_RangePanel = require('./panel/RangePanel.js');
var datePicker_utils = require('./utils.js');
var isArray = require('../_chunks/dep-794b42ce.js');
var isFunction = require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-06bf4625.js');
require('../_chunks/dep-bf18bb11.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-c990dbb4.js');
require('../_chunks/dep-eab994e8.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-6abdd014.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../hooks/useIsFirstRender.js');
require('../range-input/RangeInput.js');
require('../_chunks/dep-1c586f5c.js');
require('tdesign-icons-react');
require('../input/index.js');
require('../input/Input.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../_util/parseTNode.js');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-b4442734.js');
require('../hooks/useGlobalIcon.js');
require('../hooks/useLayoutEffect.js');
require('../_chunks/dep-68f5fe7b.js');
require('../locale/LocalReceiver.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../input/defaultProps.js');
require('../input/InputGroup.js');
require('../input/useLengthLimit.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-f0e884b9.js');
require('../range-input/defaultProps.js');
require('../range-input/RangeInputPopup.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useAnimation.js');
require('../hooks/useAttach.js');
require('../hooks/useMutationObserver.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-d3af8eb7.js');
require('../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/listener.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');
require('../popup/utils/transition.js');
require('../popup/PopupPlugin.js');
require('../_chunks/dep-e37cb830.js');
require('../_util/react-render.js');
require('../common/PluginContainer.js');
require('../select-input/useOverlayInnerStyle.js');
require('../hooks/useInnerPopupVisible.js');
require('./hooks/useRangeValue.js');
require('./panel/PanelContent.js');
require('./base/Header.js');
require('../pagination/index.js');
require('../pagination/Pagination.js');
require('../select/index.js');
require('../select/base/Select.js');
require('../_util/helper.js');
require('../_chunks/dep-e44b03ec.js');
require('../common/FakeArrow.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../hooks/useDomRefCallback.js');
require('../loading/plugin.js');
require('../select-input/index.js');
require('../select-input/SelectInput.js');
require('../select-input/defaultProps.js');
require('../select-input/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/TagInput.js');
require('../hooks/useDragSorter.js');
require('../hooks/useEventCallback.js');
require('../tag-input/defaultProps.js');
require('../tag-input/useHover.js');
require('../tag-input/useTagList.js');
require('../tag/index.js');
require('../tag/Tag.js');
require('../_chunks/dep-d14bfcfa.js');
require('../tag/defaultProps.js');
require('../tag/CheckTag.js');
require('../_chunks/dep-dd0f5f6b.js');
require('../tag/CheckTagGroup.js');
require('../tag-input/useTagScroll.js');
require('../select-input/useSingle.js');
require('../_chunks/dep-f7bf3d17.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-ca918872.js');
require('../_chunks/dep-13de2fef.js');
require('../_chunks/dep-63308b84.js');
require('../_chunks/dep-048e4314.js');
require('../select/defaultProps.js');
require('../select/hooks/useKeyboardControl.js');
require('../_chunks/dep-bdc97672.js');
require('../hooks/useRipple.js');
require('../select/base/OptionGroup.js');
require('../_chunks/dep-05df1ea3.js');
require('../select/base/PopupContent.js');
require('../select/hooks/usePanelVirtualScroll.js');
require('../hooks/useVirtualScroll.js');
require('../input-number/index.js');
require('../input-number/InputNumber.js');
require('../button/index.js');
require('../button/Button.js');
require('../button/defaultProps.js');
require('../input-number/useInputNumber.js');
require('../_chunks/dep-04deac0d.js');
require('../_chunks/dep-8f62fb96.js');
require('../hooks/useCommonClassName.js');
require('../input-number/defaultProps.js');
require('../input-adornment/InputAdornment.js');
require('../pagination/hooks/useBoundaryJumper.js');
require('../pagination/hooks/usePrevNextJumper.js');
require('../pagination/hooks/usePageNumber.js');
require('../pagination/hooks/useTotal.js');
require('../pagination/validators.js');
require('../pagination/defaultProps.js');
require('../_chunks/dep-5c3139cb.js');
require('../_chunks/dep-2e455d91.js');
require('../_chunks/dep-894a5ece.js');
require('../pagination/PaginationMini.js');
require('./hooks/useSelectRange.js');
require('./base/Table.js');
require('./base/Cell.js');
require('../time-picker/index.js');
require('../time-picker/TimePicker.js');
require('../_chunks/dep-9220a4d6.js');
require('../time-picker/hooks/useTimePickerTextConfig.js');
require('../hooks/useDebounce.js');
require('../hooks/usePersistFn.js');
require('../hooks/usePropsRef.js');
require('../time-picker/TimeRangePicker.js');
require('../time-picker/panel/TimePickerPanel.js');
require('../time-picker/defaultProps.js');
require('./panel/ExtraContent.js');
require('./base/Footer.js');
require('./hooks/useTableData.js');
require('./hooks/useDisableDate.js');
require('../_chunks/dep-4d78a412.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var DateRangePicker = /*#__PURE__*/React.forwardRef(function (originalProps, ref) {
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix,
    globalDatePickerConfig = _useConfig.datePicker;
  var props = hooks_useDefaultProps["default"](originalProps, datePicker_defaultProps.dateRangePickerDefaultProps);
  var className = props.className,
    style = props.style,
    disabled = props.disabled,
    mode = props.mode,
    enableTimePicker = props.enableTimePicker,
    disableDate = props.disableDate,
    _props$firstDayOfWeek = props.firstDayOfWeek,
    firstDayOfWeek = _props$firstDayOfWeek === void 0 ? globalDatePickerConfig.firstDayOfWeek : _props$firstDayOfWeek,
    presets = props.presets,
    timePickerProps = props.timePickerProps,
    presetsPlacement = props.presetsPlacement,
    panelPreselection = props.panelPreselection,
    cancelRangeSelectLimit = props.cancelRangeSelectLimit,
    range = props.range,
    panelActiveDate = props.panelActiveDate,
    cell = props.cell,
    onPick = props.onPick,
    disableTime = props.disableTime,
    needConfirm = props.needConfirm;
  var _useRange = datePicker_hooks_useRange["default"](props),
    inputValue = _useRange.inputValue,
    popupVisible = _useRange.popupVisible,
    rangeInputProps = _useRange.rangeInputProps,
    popupProps = _useRange.popupProps,
    value = _useRange.value,
    year = _useRange.year,
    month = _useRange.month,
    time = _useRange.time,
    activeIndex = _useRange.activeIndex,
    isHoverCell = _useRange.isHoverCell,
    setActiveIndex = _useRange.setActiveIndex,
    onChange = _useRange.onChange,
    setIsHoverCell = _useRange.setIsHoverCell,
    setInputValue = _useRange.setInputValue,
    setPopupVisible = _useRange.setPopupVisible,
    setTime = _useRange.setTime,
    setYear = _useRange.setYear,
    setMonth = _useRange.setMonth,
    isFirstValueSelected = _useRange.isFirstValueSelected,
    setIsFirstValueSelected = _useRange.setIsFirstValueSelected,
    cacheValue = _useRange.cacheValue,
    setCacheValue = _useRange.setCacheValue,
    isSwitchTimeMode = _useRange.isSwitchTimeMode;
  var _getDefaultFormat = format.getDefaultFormat({
      mode: mode,
      enableTimePicker: enableTimePicker,
      format: props.format,
      valueType: props.valueType
    }),
    format$1 = _getDefaultFormat.format,
    timeFormat = _getDefaultFormat.timeFormat,
    valueType = _getDefaultFormat.valueType;
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    isSelected = _useState2[0],
    setIsSelected = _useState2[1];
  var handleSyncPanelValue = function handleSyncPanelValue(value2) {
    var nextMonth = value2.map(function (v) {
      return format.parseToDayjs(v, format$1).month();
    });
    var nextYear = value2.map(function (v) {
      return format.parseToDayjs(v, format$1).year();
    });
    if (nextYear[0] === nextYear[1] && nextMonth[0] === nextMonth[1]) {
      nextMonth[0] === 11 ? nextMonth[0] -= 1 : nextMonth[1] += 1;
    }
    if (["month", "quarter"].includes(props.mode) && nextYear[0] === nextYear[1]) {
      nextYear = [nextYear[0], nextYear[0] + 1];
    }
    setMonth(nextMonth);
    setYear(nextYear);
  };
  var handlePopupInvisible = function handlePopupInvisible() {
    var _props$popupProps, _props$popupProps$onV;
    setPopupVisible(false);
    (_props$popupProps = props.popupProps) === null || _props$popupProps === void 0 || (_props$popupProps$onV = _props$popupProps.onVisibleChange) === null || _props$popupProps$onV === void 0 || _props$popupProps$onV.call(_props$popupProps, false, {});
  };
  var onTriggerNeedConfirm = hooks_useLatest["default"](function () {
    if (needConfirm || !enableTimePicker || popupVisible) return;
    var nextValue = toConsumableArray._toConsumableArray(inputValue);
    var notValidIndex = nextValue.findIndex(function (v) {
      return !v || !format.isValidDate(v, format$1);
    });
    if (notValidIndex === -1 && nextValue.length === 2) {
      var currentValue = format.formatDate(value || [], {
        format: format$1
      });
      if (currentValue[0] !== nextValue[0] || currentValue[1] !== nextValue[1]) {
        var formattedValue = format.formatDate(nextValue, {
          format: format$1,
          targetFormat: valueType,
          autoSwap: true
        });
        onChange(formattedValue, {
          dayjsValue: nextValue.map(function (v) {
            return format.parseToDayjs(v, format$1);
          }),
          trigger: "confirm"
        });
      }
    } else {
      setInputValue(format.formatDate(value || [], {
        format: format$1
      }));
    }
  });
  React.useEffect(function () {
    if (value === cacheValue) return;
    if (popupVisible) {
      setIsSelected(false);
      setCacheValue(format.formatDate(value || [], {
        format: format$1
      }));
      setTime(format.formatTime(value || [dayjs__default["default"]().format(timeFormat), dayjs__default["default"]().format(timeFormat)], format$1, timeFormat, props.defaultTime));
      if (!value.length) {
        if (range && isArray.isArray(range) || panelActiveDate) {
          var _startRangeBounds$min, _startRangeBounds$min2, _startRangeBounds$max, _startRangeBounds$min3, _startRangeBounds$min4, _startRangeBounds$max2, _endRangeBounds$min$g, _endRangeBounds$min, _endRangeBounds$max, _endRangeBounds$min$g2, _endRangeBounds$min2, _endRangeBounds$max2, _startPanelActiveDate, _startPanelActiveDate2, _startPanelActiveDate3, _startPanelActiveDate4, _endPanelActiveDate$y, _endPanelActiveDate, _endPanelActiveDate2, _endPanelActiveDate3;
          var startRange = range;
          var endRange = range;
          if (isArray.isArray(range)) {
            var _range = slicedToArray._slicedToArray(range, 2),
              first = _range[0],
              second = _range[1];
            if (isArray.isArray(first) || isFunction.isFunction(first) || isArray.isArray(second) || isFunction.isFunction(second)) {
              startRange = first;
              endRange = second;
            }
          }
          var startRangeBounds = utils.getRangeBounds(startRange);
          var endRangeBounds = utils.getRangeBounds(endRange);
          var startYearFromRange = (_startRangeBounds$min = (_startRangeBounds$min2 = startRangeBounds.min) === null || _startRangeBounds$min2 === void 0 ? void 0 : _startRangeBounds$min2.getFullYear()) !== null && _startRangeBounds$min !== void 0 ? _startRangeBounds$min : (_startRangeBounds$max = startRangeBounds.max) === null || _startRangeBounds$max === void 0 ? void 0 : _startRangeBounds$max.getFullYear();
          var startMonthFromRange = (_startRangeBounds$min3 = (_startRangeBounds$min4 = startRangeBounds.min) === null || _startRangeBounds$min4 === void 0 ? void 0 : _startRangeBounds$min4.getMonth()) !== null && _startRangeBounds$min3 !== void 0 ? _startRangeBounds$min3 : (_startRangeBounds$max2 = startRangeBounds.max) === null || _startRangeBounds$max2 === void 0 ? void 0 : _startRangeBounds$max2.getMonth();
          var endYearFromRange = (_endRangeBounds$min$g = (_endRangeBounds$min = endRangeBounds.min) === null || _endRangeBounds$min === void 0 ? void 0 : _endRangeBounds$min.getFullYear()) !== null && _endRangeBounds$min$g !== void 0 ? _endRangeBounds$min$g : (_endRangeBounds$max = endRangeBounds.max) === null || _endRangeBounds$max === void 0 ? void 0 : _endRangeBounds$max.getFullYear();
          var endMonthFromRange = (_endRangeBounds$min$g2 = (_endRangeBounds$min2 = endRangeBounds.min) === null || _endRangeBounds$min2 === void 0 ? void 0 : _endRangeBounds$min2.getMonth()) !== null && _endRangeBounds$min$g2 !== void 0 ? _endRangeBounds$min$g2 : (_endRangeBounds$max2 = endRangeBounds.max) === null || _endRangeBounds$max2 === void 0 ? void 0 : _endRangeBounds$max2.getMonth();
          var startPanelActiveDate = panelActiveDate;
          var endPanelActiveDate = panelActiveDate;
          if (isArray.isArray(panelActiveDate)) {
            var _panelActiveDate = slicedToArray._slicedToArray(panelActiveDate, 2);
            startPanelActiveDate = _panelActiveDate[0];
            endPanelActiveDate = _panelActiveDate[1];
          }
          var leftYear = (_startPanelActiveDate = (_startPanelActiveDate2 = startPanelActiveDate) === null || _startPanelActiveDate2 === void 0 ? void 0 : _startPanelActiveDate2.year) !== null && _startPanelActiveDate !== void 0 ? _startPanelActiveDate : startYearFromRange;
          var leftMonth = (_startPanelActiveDate3 = startPanelActiveDate) !== null && _startPanelActiveDate3 !== void 0 && _startPanelActiveDate3.month ? Number((_startPanelActiveDate4 = startPanelActiveDate) === null || _startPanelActiveDate4 === void 0 ? void 0 : _startPanelActiveDate4.month) - 1 : startMonthFromRange;
          var rightYear = (_endPanelActiveDate$y = (_endPanelActiveDate = endPanelActiveDate) === null || _endPanelActiveDate === void 0 ? void 0 : _endPanelActiveDate.year) !== null && _endPanelActiveDate$y !== void 0 ? _endPanelActiveDate$y : endYearFromRange;
          var rightMonth = (_endPanelActiveDate2 = endPanelActiveDate) !== null && _endPanelActiveDate2 !== void 0 && _endPanelActiveDate2.month ? Number((_endPanelActiveDate3 = endPanelActiveDate) === null || _endPanelActiveDate3 === void 0 ? void 0 : _endPanelActiveDate3.month) - 1 : endMonthFromRange;
          var _initYearMonthTime = format.initYearMonthTime({
              value: value,
              mode: mode,
              format: format$1,
              enableTimePicker: enableTimePicker
            }),
            defaultYear = _initYearMonthTime.year,
            defaultMonth = _initYearMonthTime.month;
          var nextYear = [leftYear !== null && leftYear !== void 0 ? leftYear : defaultYear[0], rightYear !== null && rightYear !== void 0 ? rightYear : defaultYear[1]];
          var nextMonth = [leftMonth !== null && leftMonth !== void 0 ? leftMonth : defaultMonth[0], rightMonth !== null && rightMonth !== void 0 ? rightMonth : defaultMonth[1]];
          if (nextYear[0] === nextYear[1] && nextMonth[0] === nextMonth[1] && !enableTimePicker) {
            if (startRange === endRange) {
              if (mode === "year") nextYear[1] += 10;else if (mode === "month" || mode === "quarter") nextYear[1] += 1;else nextMonth[1] += 1;
            }
          }
          setYear(nextYear);
          setMonth(nextMonth);
        } else {
          var _initYearMonthTime2 = format.initYearMonthTime({
              value: value,
              mode: mode,
              format: format$1,
              enableTimePicker: enableTimePicker
            }),
            _defaultYear = _initYearMonthTime2.year,
            _defaultMonth = _initYearMonthTime2.month;
          setYear(_defaultYear);
          setMonth(_defaultMonth);
        }
      } else if (value.length === 2 && (!enableTimePicker || isSwitchTimeMode)) {
        handleSyncPanelValue(value);
      } else {
        var _nextYear = value.map(function (v) {
          return format.parseToDayjs(v, format$1).year();
        });
        if (_nextYear.length === 1) _nextYear = [_nextYear[0], _nextYear[0]];
        var _nextMonth = value.map(function (v) {
          return format.parseToDayjs(v, format$1).month();
        });
        if (_nextMonth.length === 1) _nextMonth = [_nextMonth[0], Math.min(_nextMonth[0] + 1, 11)];
        setYear(_nextYear);
        setMonth(_nextMonth);
      }
    } else {
      setActiveIndex(0);
      setIsHoverCell(false);
      setIsFirstValueSelected(false);
      setInputValue(format.formatDate(value || [], {
        format: format$1
      }));
    }
  }, [popupVisible]);
  hooks_useUpdateEffect["default"](function () {
    onTriggerNeedConfirm.current();
  }, [popupVisible]);
  function onCellMouseEnter(date) {
    setIsHoverCell(true);
    var nextValue = toConsumableArray._toConsumableArray(inputValue);
    nextValue[activeIndex] = format.formatDate(date, {
      format: format$1
    });
    setInputValue(nextValue);
  }
  function onCellMouseLeave() {
    setIsHoverCell(false);
    setInputValue(cacheValue);
  }
  function onCellClick(date, _ref) {
    var e = _ref.e;
    onPick === null || onPick === void 0 || onPick(date, {
      e: e,
      partial: activeIndex ? "end" : "start"
    });
    setIsHoverCell(false);
    setIsSelected(true);
    var nextValue = toConsumableArray._toConsumableArray(inputValue);
    nextValue[activeIndex] = format.formatDate(date, {
      format: format$1
    });
    setCacheValue(nextValue);
    setInputValue(nextValue);
    if (enableTimePicker && !isSwitchTimeMode) return;
    var notValidIndex = nextValue.findIndex(function (v) {
      return !v || !format.isValidDate(v, format$1);
    });
    if (notValidIndex === -1 && nextValue.length === 2) {
      if (!isFirstValueSelected && format.parseToDayjs(nextValue[0], format$1).isAfter(format.parseToDayjs(nextValue[1], format$1))) {
        nextValue[activeIndex ? 0 : 1] = "";
        setCacheValue(nextValue);
        setInputValue(nextValue);
      } else {
        onChange(format.formatDate(nextValue, {
          format: format$1,
          targetFormat: valueType,
          autoSwap: true
        }), {
          dayjsValue: nextValue.map(function (v) {
            return format.parseToDayjs(v, format$1);
          }),
          trigger: "pick"
        });
      }
    }
    if (!isFirstValueSelected || !activeIndex || isSwitchTimeMode) {
      var nextIndex = notValidIndex;
      if (isSwitchTimeMode && activeIndex === 1 && isFirstValueSelected) return;
      if (nextIndex === -1) nextIndex = activeIndex ? 0 : 1;
      setActiveIndex(nextIndex);
      setIsFirstValueSelected(!!nextValue[0]);
    } else {
      handlePopupInvisible();
    }
  }
  function onJumperClick(_ref2) {
    var trigger = _ref2.trigger,
      partial = _ref2.partial;
    var partialIndex = partial === "start" ? 0 : 1;
    var monthCountMap = {
      date: 1,
      week: 1,
      month: 12,
      quarter: 12,
      year: 120
    };
    var monthCount = monthCountMap[mode] || 0;
    var current = new Date(year[partialIndex], month[partialIndex]);
    var next = null;
    if (trigger === "prev") {
      next = utils.subtractMonth(current, monthCount);
    } else if (trigger === "current") {
      next = new Date();
    } else if (trigger === "next") {
      next = utils.addMonth(current, monthCount);
    }
    var nextYear = toConsumableArray._toConsumableArray(year);
    nextYear[partialIndex] = next.getFullYear();
    var nextMonth = toConsumableArray._toConsumableArray(month);
    nextMonth[partialIndex] = next.getMonth();
    var onlyYearSelect = ["year", "quarter", "month"].includes(mode);
    var correctedDate = datePicker_utils.dateCorrection(partialIndex, nextYear, nextMonth, onlyYearSelect);
    nextYear = correctedDate.nextYear;
    nextMonth = correctedDate.nextMonth;
    setYear(nextYear);
    setMonth(nextMonth);
    var yearChanged = year[partialIndex] !== nextYear[partialIndex];
    var monthChanged = month[partialIndex] !== nextMonth[partialIndex];
    if (yearChanged) {
      var _props$onYearChange;
      (_props$onYearChange = props.onYearChange) === null || _props$onYearChange === void 0 || _props$onYearChange.call(props, {
        partial: partial,
        year: nextYear[partialIndex],
        date: value.map(function (v) {
          return dayjs__default["default"](v).toDate();
        }),
        trigger: trigger === "current" ? "today" : "year-".concat(datePicker_utils.triggerMap[trigger])
      });
    }
    if (monthChanged) {
      var _props$onMonthChange;
      (_props$onMonthChange = props.onMonthChange) === null || _props$onMonthChange === void 0 || _props$onMonthChange.call(props, {
        partial: partial,
        month: nextMonth[partialIndex],
        date: value.map(function (v) {
          return dayjs__default["default"](v).toDate();
        }),
        trigger: trigger === "current" ? "today" : "month-".concat(datePicker_utils.triggerMap[trigger])
      });
    }
  }
  function onTimePickerChange(val, context) {
    var _context$activeIndex;
    var _extractTimeObj = utils.extractTimeObj(val),
      hours = _extractTimeObj.hours,
      minutes = _extractTimeObj.minutes,
      seconds = _extractTimeObj.seconds,
      milliseconds = _extractTimeObj.milliseconds,
      meridiem = _extractTimeObj.meridiem;
    var currentIndex = (_context$activeIndex = context.activeIndex) !== null && _context$activeIndex !== void 0 ? _context$activeIndex : activeIndex;
    var nextInputValue = toConsumableArray._toConsumableArray(inputValue);
    var changedInputValue = inputValue[currentIndex];
    var currentDate = !dayjs__default["default"](changedInputValue, format$1).isValid() ? dayjs__default["default"]().year(year[currentIndex]).month(month[currentIndex]) : dayjs__default["default"](changedInputValue, format$1);
    var nextHours = hours;
    if (/am/i.test(meridiem) && nextHours === 12) nextHours -= 12;
    if (/pm/i.test(meridiem) && nextHours < 12) nextHours += 12;
    var nextDate = currentDate.hour(nextHours).minute(minutes).second(seconds).millisecond(milliseconds).toDate();
    nextInputValue[currentIndex] = nextDate;
    var nextTime = toConsumableArray._toConsumableArray(time);
    nextTime[currentIndex] = val;
    setTime(nextTime);
    setIsSelected(true);
    setInputValue(format.formatDate(nextInputValue, {
      format: format$1
    }));
    setCacheValue(format.formatDate(nextInputValue, {
      format: format$1
    }));
  }
  function onConfirmClick(_ref3) {
    var _props$onConfirm;
    var e = _ref3.e;
    var nextValue = toConsumableArray._toConsumableArray(inputValue);
    props === null || props === void 0 || (_props$onConfirm = props.onConfirm) === null || _props$onConfirm === void 0 || _props$onConfirm.call(props, {
      e: e,
      date: nextValue,
      partial: activeIndex ? "end" : "start"
    });
    var notValidIndex = nextValue.findIndex(function (v) {
      return !v || !format.isValidDate(v, format$1);
    });
    if (notValidIndex === -1 && nextValue.length === 2) {
      if (!isFirstValueSelected && format.parseToDayjs(nextValue[0], format$1).isAfter(format.parseToDayjs(nextValue[1], format$1))) {
        nextValue[activeIndex ? 0 : 1] = "";
        setCacheValue(nextValue);
        setInputValue(nextValue);
      } else {
        onChange(format.formatDate(nextValue, {
          format: format$1,
          targetFormat: valueType,
          autoSwap: true
        }), {
          dayjsValue: nextValue.map(function (v) {
            return format.parseToDayjs(v, format$1);
          }),
          trigger: "confirm"
        });
      }
    }
    if (isSwitchTimeMode && nextValue.every(Boolean)) handlePopupInvisible();
    if (!isFirstValueSelected || !activeIndex) {
      var nextIndex = notValidIndex;
      if (nextIndex === -1) nextIndex = activeIndex ? 0 : 1;
      setActiveIndex(nextIndex);
      setIsFirstValueSelected(!!nextValue[0]);
    } else if (nextValue.length === 2) {
      handlePopupInvisible();
    }
  }
  function onPresetClick(preset, context) {
    var presetValue = preset;
    if (typeof preset === "function") {
      presetValue = preset();
    }
    if (!Array.isArray(presetValue)) {
      log.log.error("DateRangePicker", "preset: ".concat(preset, " must be Array!"));
    } else {
      var _props$onPresetClick;
      var formattedPreset = format.formatDate(presetValue, {
        format: format$1,
        targetFormat: valueType
      });
      setInputValue(formattedPreset);
      setCacheValue(formattedPreset);
      setTime(format.formatTime(formattedPreset, format$1, timeFormat, props.defaultTime));
      setIsSelected(true);
      setIsFirstValueSelected(true);
      handleSyncPanelValue(formattedPreset);
      handlePopupInvisible();
      onChange(formattedPreset, {
        dayjsValue: formattedPreset.map(function (p) {
          return format.parseToDayjs(p, format$1);
        }),
        trigger: "preset"
      });
      (_props$onPresetClick = props.onPresetClick) === null || _props$onPresetClick === void 0 || _props$onPresetClick.call(props, context);
    }
  }
  function onYearChange(nextVal, _ref4) {
    var _props$onYearChange2;
    var partial = _ref4.partial;
    var partialIndex = partial === "start" ? 0 : 1;
    if (enableTimePicker) partialIndex = activeIndex;
    var nextYear = toConsumableArray._toConsumableArray(year);
    var nextMonth = toConsumableArray._toConsumableArray(month);
    nextYear[partialIndex] = nextVal;
    var onlyYearSelect = ["year", "quarter", "month"].includes(mode);
    var correctedDate = datePicker_utils.dateCorrection(partialIndex, nextYear, nextMonth, onlyYearSelect);
    nextYear = correctedDate.nextYear;
    nextMonth = correctedDate.nextMonth;
    setYear(nextYear);
    !onlyYearSelect && setMonth(nextMonth);
    (_props$onYearChange2 = props.onYearChange) === null || _props$onYearChange2 === void 0 || _props$onYearChange2.call(props, {
      partial: partial,
      year: nextYear[partialIndex],
      date: value.map(function (v) {
        return dayjs__default["default"](v).toDate();
      }),
      trigger: "year-select"
    });
  }
  function onMonthChange(nextVal, _ref5) {
    var _props$onMonthChange2;
    var partial = _ref5.partial;
    var partialIndex = partial === "start" ? 0 : 1;
    var nextMonth = toConsumableArray._toConsumableArray(month);
    if (enableTimePicker) partialIndex = activeIndex;
    setMonth(function () {
      nextMonth[partialIndex] = nextVal;
      if (year[0] === year[1]) {
        if (partialIndex === 0) {
          if (nextMonth[1] <= nextMonth[0]) {
            nextMonth[1] = nextMonth[0] + 1;
            if (nextMonth[1] === 12) {
              nextMonth[1] = 0;
              setYear(function (currentYear) {
                return [currentYear[0], currentYear[1] + 1];
              });
            }
          }
        }
        if (partialIndex === 1) {
          nextMonth[0] = Math.min(nextMonth[0], nextMonth[1]);
          if (nextMonth[0] >= nextMonth[1]) {
            nextMonth[0] -= 1;
            if (nextMonth[0] === -1) {
              nextMonth[0] = 11;
              setYear(function (currentYear) {
                return [currentYear[0] - 1, currentYear[1]];
              });
            }
          }
        }
      }
      return nextMonth;
    });
    (_props$onMonthChange2 = props.onMonthChange) === null || _props$onMonthChange2 === void 0 || _props$onMonthChange2.call(props, {
      partial: partial,
      month: nextMonth[partialIndex],
      date: value.map(function (v) {
        return dayjs__default["default"](v).toDate();
      }),
      trigger: "month-select"
    });
  }
  var panelProps = {
    hoverValue: isHoverCell ? inputValue : [],
    value: isSelected ? cacheValue : value,
    isFirstValueSelected: isFirstValueSelected,
    panelPreselection: panelPreselection,
    year: year,
    month: month,
    range: range,
    mode: mode,
    format: format$1,
    presets: presets,
    time: time,
    disableDate: disableDate,
    firstDayOfWeek: firstDayOfWeek,
    timePickerProps: timePickerProps,
    enableTimePicker: enableTimePicker,
    presetsPlacement: presetsPlacement,
    activeIndex: activeIndex,
    popupVisible: popupVisible,
    cancelRangeSelectLimit: cancelRangeSelectLimit,
    needConfirm: needConfirm,
    cell: cell,
    onCellClick: onCellClick,
    onCellMouseEnter: onCellMouseEnter,
    onCellMouseLeave: onCellMouseLeave,
    onJumperClick: onJumperClick,
    onConfirmClick: onConfirmClick,
    onPresetClick: onPresetClick,
    onYearChange: onYearChange,
    onMonthChange: onMonthChange,
    onTimePickerChange: onTimePickerChange,
    disableTime: disableTime,
    isSwitchTimeMode: isSwitchTimeMode
  };
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-date-range-picker"), className),
    style: style,
    ref: ref
  }, /* @__PURE__ */React__default["default"].createElement(rangeInput_index.RangeInputPopup, {
    disabled: disabled,
    status: props.status,
    tips: props.tips,
    inputValue: inputValue,
    popupProps: popupProps,
    rangeInputProps: rangeInputProps,
    popupVisible: popupVisible,
    panel: /* @__PURE__ */React__default["default"].createElement(datePicker_panel_RangePanel["default"], _objectSpread({}, panelProps))
  }));
});
DateRangePicker.displayName = "DateRangePicker";

exports["default"] = DateRangePicker;
