// tdesign-react/cjs/date-picker/DateRangePickerPanel.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var toConsumableArray = require('../_chunks/dep-1d0dbfc5.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var dayjs = require('dayjs');
var format = require('../_chunks/dep-06ac19a1.js');
var utils = require('../_chunks/dep-8227db70.js');
var log = require('../_chunks/dep-b289b0be.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var datePicker_hooks_useRangeValue = require('./hooks/useRangeValue.js');
var datePicker_panel_RangePanel = require('./panel/RangePanel.js');
var datePicker_utils = require('./utils.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-06bf4625.js');
require('../_chunks/dep-bf18bb11.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-c990dbb4.js');
require('../_chunks/dep-eab994e8.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-6abdd014.js');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-b4442734.js');
require('classnames');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('./panel/PanelContent.js');
require('./base/Header.js');
require('../locale/LocalReceiver.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../pagination/index.js');
require('../pagination/Pagination.js');
require('../_chunks/dep-1c586f5c.js');
require('../select/index.js');
require('../select/base/Select.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../_util/helper.js');
require('../_chunks/dep-e44b03ec.js');
require('../_util/parseTNode.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');
require('../common/FakeArrow.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-f0e884b9.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../_chunks/dep-68f5fe7b.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../hooks/useDomRefCallback.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../select-input/index.js');
require('../select-input/SelectInput.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../hooks/useAnimation.js');
require('../hooks/useAttach.js');
require('../hooks/useMutationObserver.js');
require('../hooks/useLatest.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-d3af8eb7.js');
require('../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/listener.js');
require('../popup/utils/transition.js');
require('../popup/PopupPlugin.js');
require('../select-input/defaultProps.js');
require('../select-input/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/TagInput.js');
require('tdesign-icons-react');
require('../hooks/useDragSorter.js');
require('../hooks/useEventCallback.js');
require('../hooks/useGlobalIcon.js');
require('../input/index.js');
require('../input/Input.js');
require('../input/defaultProps.js');
require('../input/InputGroup.js');
require('../input/useLengthLimit.js');
require('../tag-input/defaultProps.js');
require('../tag-input/useHover.js');
require('../tag-input/useTagList.js');
require('../tag/index.js');
require('../tag/Tag.js');
require('../_chunks/dep-d14bfcfa.js');
require('../tag/defaultProps.js');
require('../tag/CheckTag.js');
require('../_chunks/dep-dd0f5f6b.js');
require('../tag/CheckTagGroup.js');
require('../tag-input/useTagScroll.js');
require('../select-input/useOverlayInnerStyle.js');
require('../hooks/useInnerPopupVisible.js');
require('../select-input/useSingle.js');
require('../_chunks/dep-f7bf3d17.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-ca918872.js');
require('../_chunks/dep-13de2fef.js');
require('../_chunks/dep-63308b84.js');
require('../_chunks/dep-048e4314.js');
require('../select/defaultProps.js');
require('../select/hooks/useKeyboardControl.js');
require('../_chunks/dep-bdc97672.js');
require('../hooks/useRipple.js');
require('../select/base/OptionGroup.js');
require('../_chunks/dep-05df1ea3.js');
require('../select/base/PopupContent.js');
require('../select/hooks/usePanelVirtualScroll.js');
require('../hooks/useVirtualScroll.js');
require('../input-number/index.js');
require('../input-number/InputNumber.js');
require('../button/index.js');
require('../button/Button.js');
require('../button/defaultProps.js');
require('../input-number/useInputNumber.js');
require('../_chunks/dep-04deac0d.js');
require('../_chunks/dep-8f62fb96.js');
require('../hooks/useCommonClassName.js');
require('../input-number/defaultProps.js');
require('../input-adornment/InputAdornment.js');
require('../pagination/hooks/useBoundaryJumper.js');
require('../pagination/hooks/usePrevNextJumper.js');
require('../pagination/hooks/usePageNumber.js');
require('../pagination/hooks/useTotal.js');
require('../pagination/validators.js');
require('../pagination/defaultProps.js');
require('../_chunks/dep-5c3139cb.js');
require('../_chunks/dep-2e455d91.js');
require('../_chunks/dep-894a5ece.js');
require('../pagination/PaginationMini.js');
require('./hooks/useSelectRange.js');
require('./base/Table.js');
require('./base/Cell.js');
require('../time-picker/index.js');
require('../time-picker/TimePicker.js');
require('../_chunks/dep-9220a4d6.js');
require('../time-picker/hooks/useTimePickerTextConfig.js');
require('../hooks/useDebounce.js');
require('../hooks/usePersistFn.js');
require('../hooks/usePropsRef.js');
require('../time-picker/TimeRangePicker.js');
require('../range-input/index.js');
require('../range-input/RangeInput.js');
require('../range-input/defaultProps.js');
require('../range-input/RangeInputPopup.js');
require('../time-picker/panel/TimePickerPanel.js');
require('../time-picker/defaultProps.js');
require('./panel/ExtraContent.js');
require('./base/Footer.js');
require('./hooks/useTableData.js');
require('./hooks/useDisableDate.js');
require('../_chunks/dep-4d78a412.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var DateRangePickerPanel = /*#__PURE__*/React.forwardRef(function (originalProps, ref) {
  var props = hooks_useDefaultProps["default"](originalProps, {
    mode: "date",
    defaultValue: [],
    panelPreselection: true
  });
  var className = props.className,
    style = props.style,
    mode = props.mode,
    enableTimePicker = props.enableTimePicker,
    disableDate = props.disableDate,
    firstDayOfWeek = props.firstDayOfWeek,
    presets = props.presets,
    timePickerProps = props.timePickerProps,
    presetsPlacement = props.presetsPlacement,
    panelPreselection = props.panelPreselection,
    onPanelClick = props.onPanelClick;
  var _useRangeValue = datePicker_hooks_useRangeValue["default"](props),
    value = _useRangeValue.value,
    year = _useRangeValue.year,
    month = _useRangeValue.month,
    _useRangeValue$time = _useRangeValue.time,
    time = _useRangeValue$time === void 0 ? [] : _useRangeValue$time,
    onChange = _useRangeValue.onChange,
    setTime = _useRangeValue.setTime,
    setYear = _useRangeValue.setYear,
    setMonth = _useRangeValue.setMonth,
    isFirstValueSelected = _useRangeValue.isFirstValueSelected,
    setIsFirstValueSelected = _useRangeValue.setIsFirstValueSelected,
    cacheValue = _useRangeValue.cacheValue,
    setCacheValue = _useRangeValue.setCacheValue,
    isSwitchTimeMode = _useRangeValue.isSwitchTimeMode;
  var _getDefaultFormat = format.getDefaultFormat({
      mode: mode,
      enableTimePicker: enableTimePicker,
      format: props.format
    }),
    format$1 = _getDefaultFormat.format,
    timeFormat = _getDefaultFormat.timeFormat;
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    isSelected = _useState2[0],
    setIsSelected = _useState2[1];
  var _useState3 = React.useState(false),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    isHoverCell = _useState4[0],
    setIsHoverCell = _useState4[1];
  var _useState5 = React.useState([]),
    _useState6 = slicedToArray._slicedToArray(_useState5, 2),
    hoverValue = _useState6[0],
    setHoverValue = _useState6[1];
  var activeIndex = React.useMemo(function () {
    return isFirstValueSelected ? 1 : 0;
  }, [isFirstValueSelected]);
  var handleSyncPanelValue = function handleSyncPanelValue(value2) {
    var nextMonth = value2.map(function (v) {
      return format.parseToDayjs(v, format$1).month();
    });
    var nextYear = value2.map(function (v) {
      return format.parseToDayjs(v, format$1).year();
    });
    if (nextYear[0] === nextYear[1] && nextMonth[0] === nextMonth[1]) {
      nextMonth[0] === 11 ? nextMonth[0] -= 1 : nextMonth[1] += 1;
    }
    setMonth(nextMonth);
    setYear(nextYear);
    setTime(format.formatTime(value2, format$1, timeFormat, props.defaultTime));
  };
  function onCellMouseEnter(date) {
    setIsHoverCell(true);
    var nextValue = toConsumableArray._toConsumableArray(hoverValue);
    nextValue[activeIndex] = format.formatDate(date, {
      format: format$1
    });
    setHoverValue(nextValue);
  }
  function onCellMouseLeave() {
    setIsHoverCell(false);
    setHoverValue(cacheValue);
  }
  function onCellClick(date, _ref) {
    var _props$onCellClick;
    var e = _ref.e,
      partial = _ref.partial;
    setIsSelected(true);
    var nextValue = toConsumableArray._toConsumableArray(cacheValue);
    nextValue[activeIndex] = format.formatDate(date, {
      format: format$1
    });
    setCacheValue(nextValue);
    (_props$onCellClick = props.onCellClick) === null || _props$onCellClick === void 0 || _props$onCellClick.call(props, {
      date: nextValue.map(function (v) {
        return dayjs__default["default"](v).toDate();
      }),
      e: e,
      partial: activeIndex ? "end" : "start"
    });
    if (mode === "date") {
      var partialIndex = partial === "start" ? 0 : 1;
      var isAdditional = dayjs__default["default"](date).month() !== month[partialIndex];
      if (isAdditional) {
        if (activeIndex === 0) setMonth([dayjs__default["default"](date).month(), Math.min(dayjs__default["default"](date).month() + 1, 11)]);
        if (activeIndex === 1) setMonth([Math.max(dayjs__default["default"](date).month() - 1, 0), dayjs__default["default"](date).month()]);
      }
    }
    if (enableTimePicker && !isSwitchTimeMode) return;
    if (nextValue.length === 2 && isFirstValueSelected) {
      onChange(format.formatDate(nextValue, {
        format: format$1,
        autoSwap: true
      }), {
        dayjsValue: nextValue.map(function (v) {
          return format.parseToDayjs(v, format$1);
        }),
        trigger: "pick"
      });
      setIsFirstValueSelected(false);
    } else {
      setIsFirstValueSelected(true);
    }
  }
  function onJumperClick(_ref2) {
    var trigger = _ref2.trigger,
      partial = _ref2.partial;
    var partialIndex = partial === "start" ? 0 : 1;
    var triggerMap = {
      "-1": "arrow-previous",
      1: "arrow-next"
    };
    var monthCountMap = {
      date: 1,
      week: 1,
      month: 12,
      quarter: 12,
      year: 120
    };
    var monthCount = monthCountMap[mode] || 0;
    var current = new Date(year[partialIndex], month[partialIndex]);
    var next = null;
    if (trigger === "prev") {
      next = utils.subtractMonth(current, monthCount);
    } else if (trigger === "current") {
      next = new Date();
    } else if (trigger === "next") {
      next = utils.addMonth(current, monthCount);
    }
    var nextYear = toConsumableArray._toConsumableArray(year);
    nextYear[partialIndex] = next.getFullYear();
    var nextMonth = toConsumableArray._toConsumableArray(month);
    nextMonth[partialIndex] = next.getMonth();
    var onlyYearSelect = ["year", "quarter", "month"].includes(mode);
    var correctedDate = datePicker_utils.dateCorrection(partialIndex, nextYear, nextMonth, onlyYearSelect);
    nextYear = correctedDate.nextYear;
    nextMonth = correctedDate.nextMonth;
    if (year.some(function (y) {
      return !nextYear.includes(y);
    })) {
      var _props$onYearChange;
      (_props$onYearChange = props.onYearChange) === null || _props$onYearChange === void 0 || _props$onYearChange.call(props, {
        partial: partial,
        year: nextYear[partialIndex],
        date: value.map(function (v) {
          return format.parseToDayjs(v, format$1).toDate();
        }),
        trigger: trigger === "current" ? "today" : "year-".concat(triggerMap[trigger])
      });
    }
    if (month.some(function (m) {
      return !nextMonth.includes(m);
    })) {
      var _props$onMonthChange;
      (_props$onMonthChange = props.onMonthChange) === null || _props$onMonthChange === void 0 || _props$onMonthChange.call(props, {
        partial: partial,
        month: nextMonth[partialIndex],
        date: value.map(function (v) {
          return format.parseToDayjs(v, format$1).toDate();
        }),
        trigger: trigger === "current" ? "today" : "month-".concat(triggerMap[trigger])
      });
    }
    setYear(nextYear);
    setMonth(nextMonth);
  }
  function onTimePickerChange(val, context) {
    var _context$activeIndex, _props$onTimeChange;
    var _extractTimeObj = utils.extractTimeObj(val),
      hours = _extractTimeObj.hours,
      minutes = _extractTimeObj.minutes,
      seconds = _extractTimeObj.seconds,
      milliseconds = _extractTimeObj.milliseconds,
      meridiem = _extractTimeObj.meridiem;
    var currentIndex = (_context$activeIndex = context.activeIndex) !== null && _context$activeIndex !== void 0 ? _context$activeIndex : activeIndex;
    var nextInputValue = toConsumableArray._toConsumableArray(cacheValue);
    var changedInputValue = cacheValue[currentIndex];
    var currentDate = !dayjs__default["default"](changedInputValue, format$1).isValid() ? dayjs__default["default"]().year(year[currentIndex]).month(month[currentIndex]) : dayjs__default["default"](changedInputValue, format$1);
    var nextHours = hours;
    if (/am/i.test(meridiem) && nextHours === 12) nextHours -= 12;
    if (/pm/i.test(meridiem) && nextHours < 12) nextHours += 12;
    var nextDate = currentDate.hour(nextHours).minute(minutes).second(seconds).millisecond(milliseconds).toDate();
    nextInputValue[currentIndex] = nextDate;
    var nextTime = toConsumableArray._toConsumableArray(time);
    nextTime[currentIndex] = val;
    setTime(nextTime);
    setIsSelected(true);
    setCacheValue(format.formatDate(nextInputValue, {
      format: format$1
    }));
    (_props$onTimeChange = props.onTimeChange) === null || _props$onTimeChange === void 0 || _props$onTimeChange.call(props, {
      time: val,
      partial: currentIndex ? "end" : "start",
      date: value.map(function (v) {
        return dayjs__default["default"](v).toDate();
      }),
      trigger: "time-hour"
    });
  }
  function onConfirmClick(_ref3) {
    var _props$onConfirm;
    var e = _ref3.e;
    var nextValue = toConsumableArray._toConsumableArray(cacheValue);
    if (nextValue.length === 2 && isFirstValueSelected) {
      onChange(format.formatDate(nextValue, {
        format: format$1,
        autoSwap: true
      }), {
        dayjsValue: nextValue.map(function (v) {
          return format.parseToDayjs(v, format$1);
        }),
        trigger: "confirm"
      });
      setYear(nextValue.map(function (v) {
        return dayjs__default["default"](v, format$1).year();
      }));
      setMonth(nextValue.map(function (v) {
        return dayjs__default["default"](v, format$1).month();
      }));
      setIsFirstValueSelected(false);
    } else {
      setIsFirstValueSelected(true);
    }
    (_props$onConfirm = props.onConfirm) === null || _props$onConfirm === void 0 || _props$onConfirm.call(props, {
      date: value.map(function (v) {
        return dayjs__default["default"](v).toDate();
      }),
      e: e
    });
  }
  function onPresetClick(presetValue, context) {
    var _props$onPresetClick;
    var presetVal = typeof presetValue === "function" ? presetValue() : presetValue;
    if (!Array.isArray(presetVal)) {
      log.log.error("DateRangePickerPanel", "preset: ".concat(presetValue, " must be Array!"));
    } else {
      var formattedPreset = format.formatDate(presetVal, {
        format: format$1
      });
      setCacheValue(formattedPreset);
      setIsFirstValueSelected(true);
      setIsSelected(true);
      handleSyncPanelValue(formattedPreset);
      onChange(formattedPreset, {
        dayjsValue: presetVal.map(function (p) {
          return format.parseToDayjs(p, format$1);
        }),
        trigger: "preset"
      });
    }
    (_props$onPresetClick = props.onPresetClick) === null || _props$onPresetClick === void 0 || _props$onPresetClick.call(props, context);
  }
  function onYearChange(nextVal, _ref4) {
    var _props$onYearChange2;
    var partial = _ref4.partial;
    var partialIndex = partial === "start" ? 0 : 1;
    if (enableTimePicker) partialIndex = activeIndex;
    var nextYear = toConsumableArray._toConsumableArray(year);
    nextYear[partialIndex] = nextVal;
    var nextMonth = toConsumableArray._toConsumableArray(month);
    nextYear[partialIndex] = nextVal;
    var onlyYearSelect = ["year", "quarter", "month"].includes(mode);
    var correctedDate = datePicker_utils.dateCorrection(partialIndex, nextYear, nextMonth, onlyYearSelect);
    nextYear = correctedDate.nextYear;
    nextMonth = correctedDate.nextMonth;
    setYear(nextYear);
    !onlyYearSelect && setMonth(nextMonth);
    (_props$onYearChange2 = props.onYearChange) === null || _props$onYearChange2 === void 0 || _props$onYearChange2.call(props, {
      partial: partial,
      year: nextYear[partialIndex],
      date: value.map(function (v) {
        return dayjs__default["default"](v).toDate();
      }),
      trigger: "year-select"
    });
  }
  function onMonthChange(nextVal, _ref5) {
    var _props$onMonthChange2;
    var partial = _ref5.partial;
    var partialIndex = partial === "start" ? 0 : 1;
    if (enableTimePicker) partialIndex = activeIndex;
    var nextMonth = toConsumableArray._toConsumableArray(month);
    nextMonth[partialIndex] = nextVal;
    if (year[0] === year[1]) {
      if (partialIndex === 0) {
        if (nextMonth[1] <= nextMonth[0]) {
          nextMonth[1] = nextMonth[0] + 1;
          if (nextMonth[1] === 12) {
            nextMonth[1] = 0;
            setYear(function (currentYear) {
              return [currentYear[0], currentYear[1] + 1];
            });
          }
        }
      }
      if (partialIndex === 1) {
        nextMonth[0] = Math.min(nextMonth[0], nextMonth[1]);
        if (nextMonth[0] >= nextMonth[1]) {
          nextMonth[0] -= 1;
          if (nextMonth[0] === -1) {
            nextMonth[0] = 11;
            setYear(function (currentYear) {
              return [currentYear[0] - 1, currentYear[1]];
            });
          }
        }
      }
    }
    setMonth(nextMonth);
    (_props$onMonthChange2 = props.onMonthChange) === null || _props$onMonthChange2 === void 0 || _props$onMonthChange2.call(props, {
      partial: partial,
      month: nextMonth[partialIndex],
      date: value.map(function (v) {
        return dayjs__default["default"](v).toDate();
      }),
      trigger: "month-select"
    });
  }
  var panelProps = {
    hoverValue: isHoverCell ? hoverValue : [],
    value: isSelected ? cacheValue : value,
    year: year,
    month: month,
    mode: mode,
    format: format$1,
    presets: presets,
    time: time,
    disableDate: disableDate,
    firstDayOfWeek: firstDayOfWeek,
    timePickerProps: timePickerProps,
    enableTimePicker: enableTimePicker,
    presetsPlacement: presetsPlacement,
    activeIndex: activeIndex,
    panelPreselection: panelPreselection,
    onCellMouseEnter: onCellMouseEnter,
    onCellMouseLeave: onCellMouseLeave,
    onCellClick: onCellClick,
    onJumperClick: onJumperClick,
    onConfirmClick: onConfirmClick,
    onPresetClick: onPresetClick,
    onYearChange: onYearChange,
    onMonthChange: onMonthChange,
    onTimePickerChange: onTimePickerChange,
    onPanelClick: onPanelClick,
    isSwitchTimeMode: isSwitchTimeMode
  };
  return /* @__PURE__ */React__default["default"].createElement(datePicker_panel_RangePanel["default"], _objectSpread({
    ref: ref,
    className: className,
    style: style
  }, panelProps));
});
DateRangePickerPanel.displayName = "DateRangePickerPanel";

exports["default"] = DateRangePickerPanel;
