// tdesign-react/cjs/date-picker/DatePicker.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var dayjs = require('dayjs');
var format = require('../_chunks/dep-06ac19a1.js');
var utils = require('../_chunks/dep-8227db70.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var hooks_useLatest = require('../hooks/useLatest.js');
var hooks_useUpdateEffect = require('../hooks/useUpdateEffect.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var selectInput_index = require('../select-input/index.js');
var datePicker_defaultProps = require('./defaultProps.js');
var datePicker_hooks_useSingle = require('./hooks/useSingle.js');
var datePicker_panel_SinglePanel = require('./panel/SinglePanel.js');
var datePicker_utils = require('./utils.js');
var _baseGetTag = require('../_chunks/dep-5f51e3c1.js');
var isObjectLike = require('../_chunks/dep-ecfe656e.js');
var _baseUnary = require('../_chunks/dep-746faf11.js');
var _overArg = require('../_chunks/dep-cd29b2b2.js');
var isArray = require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-06bf4625.js');
require('../_chunks/dep-bf18bb11.js');
require('../_chunks/dep-b289b0be.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-c990dbb4.js');
require('../_chunks/dep-eab994e8.js');
require('../_chunks/dep-876d090e.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-bcd16957.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-6abdd014.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../hooks/useIsFirstRender.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../select-input/SelectInput.js');
require('../popup/index.js');
require('../popup/Popup.js');
require('react-transition-group');
require('../_chunks/dep-68f5fe7b.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../hooks/useAnimation.js');
require('../hooks/useAttach.js');
require('../hooks/useControlled.js');
require('../_util/noop.js');
require('../_chunks/dep-eb1b216c.js');
require('../_chunks/dep-b529a0d7.js');
require('../_chunks/dep-b4442734.js');
require('../hooks/useMutationObserver.js');
require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-d3af8eb7.js');
require('../hooks/usePopper.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('react-fast-compare');
require('@popperjs/core');
require('../hooks/useWindowSize.js');
require('../popup/defaultProps.js');
require('../popup/hooks/useTrigger.js');
require('../_util/listener.js');
require('../_util/ref.js');
require('react-is');
require('../_util/isFragment.js');
require('../popup/utils/transition.js');
require('../popup/PopupPlugin.js');
require('../_chunks/dep-e37cb830.js');
require('../_util/react-render.js');
require('../common/PluginContainer.js');
require('../select-input/defaultProps.js');
require('../select-input/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/TagInput.js');
require('tdesign-icons-react');
require('../hooks/useDragSorter.js');
require('../hooks/useEventCallback.js');
require('../hooks/useGlobalIcon.js');
require('../input/index.js');
require('../input/Input.js');
require('../_chunks/dep-1c586f5c.js');
require('../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../_util/parseTNode.js');
require('../input/defaultProps.js');
require('../input/InputGroup.js');
require('../input/useLengthLimit.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-f0e884b9.js');
require('../tag-input/defaultProps.js');
require('../tag-input/useHover.js');
require('../tag-input/useTagList.js');
require('../tag/index.js');
require('../tag/Tag.js');
require('../_chunks/dep-d14bfcfa.js');
require('../tag/defaultProps.js');
require('../tag/CheckTag.js');
require('../_chunks/dep-dd0f5f6b.js');
require('../tag/CheckTagGroup.js');
require('../tag-input/useTagScroll.js');
require('../select-input/useOverlayInnerStyle.js');
require('../hooks/useInnerPopupVisible.js');
require('../select-input/useSingle.js');
require('../loading/index.js');
require('../loading/Loading.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../hooks/useDomRefCallback.js');
require('../loading/plugin.js');
require('../_chunks/dep-f7bf3d17.js');
require('../_chunks/dep-5e17a073.js');
require('../_chunks/dep-ca918872.js');
require('../_chunks/dep-13de2fef.js');
require('../_chunks/dep-63308b84.js');
require('../_chunks/dep-048e4314.js');
require('./hooks/useSingleValue.js');
require('../_chunks/dep-5c3139cb.js');
require('../_chunks/dep-2e455d91.js');
require('../_chunks/dep-894a5ece.js');
require('./panel/PanelContent.js');
require('./base/Header.js');
require('../pagination/index.js');
require('../pagination/Pagination.js');
require('../select/index.js');
require('../select/base/Select.js');
require('../_util/helper.js');
require('../_chunks/dep-e44b03ec.js');
require('../common/FakeArrow.js');
require('../select/defaultProps.js');
require('../select/hooks/useKeyboardControl.js');
require('../_chunks/dep-bdc97672.js');
require('../hooks/useRipple.js');
require('../select/base/OptionGroup.js');
require('../_chunks/dep-05df1ea3.js');
require('../select/base/PopupContent.js');
require('../select/hooks/usePanelVirtualScroll.js');
require('../hooks/useVirtualScroll.js');
require('../input-number/index.js');
require('../input-number/InputNumber.js');
require('../button/index.js');
require('../button/Button.js');
require('../button/defaultProps.js');
require('../input-number/useInputNumber.js');
require('../_chunks/dep-04deac0d.js');
require('../_chunks/dep-8f62fb96.js');
require('../hooks/useCommonClassName.js');
require('../input-number/defaultProps.js');
require('../input-adornment/InputAdornment.js');
require('../pagination/hooks/useBoundaryJumper.js');
require('../pagination/hooks/usePrevNextJumper.js');
require('../pagination/hooks/usePageNumber.js');
require('../pagination/hooks/useTotal.js');
require('../pagination/validators.js');
require('../pagination/defaultProps.js');
require('../pagination/PaginationMini.js');
require('./hooks/useSelectRange.js');
require('./base/Table.js');
require('./base/Cell.js');
require('../time-picker/index.js');
require('../time-picker/TimePicker.js');
require('../_chunks/dep-9220a4d6.js');
require('../time-picker/hooks/useTimePickerTextConfig.js');
require('../hooks/useDebounce.js');
require('../hooks/usePersistFn.js');
require('../hooks/usePropsRef.js');
require('../time-picker/TimeRangePicker.js');
require('../range-input/index.js');
require('../range-input/RangeInput.js');
require('../range-input/defaultProps.js');
require('../range-input/RangeInputPopup.js');
require('../time-picker/panel/TimePickerPanel.js');
require('../time-picker/defaultProps.js');
require('./panel/ExtraContent.js');
require('./base/Footer.js');
require('./hooks/useTableData.js');
require('./hooks/useDisableDate.js');
require('../_chunks/dep-4d78a412.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

/** `Object#toString` result references. */
var dateTag = '[object Date]';

/**
 * The base implementation of `_.isDate` without Node.js optimizations.
 *
 * @private
 * @param {*} value The value to check.
 * @returns {boolean} Returns `true` if `value` is a date object, else `false`.
 */
function baseIsDate(value) {
  return isObjectLike.isObjectLike(value) && _baseGetTag.baseGetTag(value) == dateTag;
}

/* Node.js helper references. */
var nodeIsDate = _overArg.nodeUtil && _overArg.nodeUtil.isDate;

/**
 * Checks if `value` is classified as a `Date` object.
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Lang
 * @param {*} value The value to check.
 * @returns {boolean} Returns `true` if `value` is a date object, else `false`.
 * @example
 *
 * _.isDate(new Date);
 * // => true
 *
 * _.isDate('Mon April 23 2012');
 * // => false
 */
var isDate = nodeIsDate ? _baseUnary.baseUnary(nodeIsDate) : baseIsDate;
var isDate$1 = isDate;

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var DatePicker = /*#__PURE__*/React.forwardRef(function (originalProps, ref) {
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var props = hooks_useDefaultProps["default"](originalProps, datePicker_defaultProps.datePickerDefaultProps);
  var className = props.className,
    style = props.style,
    disabled = props.disabled,
    mode = props.mode,
    enableTimePicker = props.enableTimePicker,
    disableDate = props.disableDate,
    firstDayOfWeek = props.firstDayOfWeek,
    presets = props.presets,
    defaultTime = props.defaultTime,
    timePickerProps = props.timePickerProps,
    presetsPlacement = props.presetsPlacement,
    needConfirm = props.needConfirm,
    multiple = props.multiple,
    label = props.label,
    range = props.range,
    cell = props.cell,
    disableTime = props.disableTime,
    onClear = props.onClear,
    onPick = props.onPick;
  var _useSingle = datePicker_hooks_useSingle["default"](props),
    inputValue = _useSingle.inputValue,
    popupVisible = _useSingle.popupVisible,
    inputProps = _useSingle.inputProps,
    popupProps = _useSingle.popupProps,
    tagInputProps = _useSingle.tagInputProps,
    value = _useSingle.value,
    year = _useSingle.year,
    month = _useSingle.month,
    time = _useSingle.time,
    onChange = _useSingle.onChange,
    setIsHoverCell = _useSingle.setIsHoverCell,
    setInputValue = _useSingle.setInputValue,
    setPopupVisible = _useSingle.setPopupVisible,
    setTime = _useSingle.setTime,
    setYear = _useSingle.setYear,
    setMonth = _useSingle.setMonth,
    cacheValue = _useSingle.cacheValue,
    setCacheValue = _useSingle.setCacheValue;
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("datePicker"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 1),
    local = _useLocaleReceiver2[0];
  var _getDefaultFormat = format.getDefaultFormat({
      mode: mode,
      format: props.format,
      valueType: props.valueType,
      enableTimePicker: multiple ? false : enableTimePicker
    }),
    format$1 = _getDefaultFormat.format,
    timeFormat = _getDefaultFormat.timeFormat,
    valueType = _getDefaultFormat.valueType;
  var onTriggerNeedConfirm = hooks_useLatest["default"](function () {
    if (needConfirm || !enableTimePicker || popupVisible) return;
    var nextValue = format.formatDate(inputValue, {
      format: format$1
    });
    if (nextValue) {
      var currentValue = format.formatDate(value, {
        format: format$1
      });
      if (currentValue === nextValue) return;
      onChange(format.formatDate(inputValue, {
        format: format$1,
        targetFormat: valueType
      }), {
        dayjsValue: format.parseToDayjs(inputValue, format$1),
        trigger: "confirm"
      });
    } else {
      setInputValue(format.formatDate(value, {
        format: format$1
      }));
    }
  });
  var handlePopupInvisible = function handlePopupInvisible() {
    var _props$popupProps, _props$popupProps$onV;
    setPopupVisible(false);
    (_props$popupProps = props.popupProps) === null || _props$popupProps === void 0 || (_props$popupProps$onV = _props$popupProps.onVisibleChange) === null || _props$popupProps$onV === void 0 || _props$popupProps$onV.call(_props$popupProps, false, {});
  };
  hooks_useUpdateEffect["default"](function () {
    onTriggerNeedConfirm.current();
  }, [popupVisible]);
  React.useEffect(function () {
    if (multiple) return;
    var dateValue = value && !isDate$1(value) && !["week", "quarter"].includes(props.mode) ? utils.covertToDate(value, valueType) : value;
    setCacheValue(format.formatDate(dateValue, {
      format: format$1
    }));
    setInputValue(format.formatDate(dateValue, {
      format: format$1
    }));
    if (popupVisible) {
      if ((props.range && isArray.isArray(props.range) || props.panelActiveDate) && !value) {
        var _rangeBounds$min$getF, _rangeBounds$min, _rangeBounds$max, _rangeBounds$min$getM, _rangeBounds$min2, _rangeBounds$max2, _props$panelActiveDat, _props$panelActiveDat2, _props$panelActiveDat3, _props$panelActiveDat4;
        var rangeBounds = utils.getRangeBounds(props.range);
        var yearFromRange = (_rangeBounds$min$getF = (_rangeBounds$min = rangeBounds.min) === null || _rangeBounds$min === void 0 ? void 0 : _rangeBounds$min.getFullYear()) !== null && _rangeBounds$min$getF !== void 0 ? _rangeBounds$min$getF : (_rangeBounds$max = rangeBounds.max) === null || _rangeBounds$max === void 0 ? void 0 : _rangeBounds$max.getFullYear();
        var monthFromRange = (_rangeBounds$min$getM = (_rangeBounds$min2 = rangeBounds.min) === null || _rangeBounds$min2 === void 0 ? void 0 : _rangeBounds$min2.getMonth()) !== null && _rangeBounds$min$getM !== void 0 ? _rangeBounds$min$getM : (_rangeBounds$max2 = rangeBounds.max) === null || _rangeBounds$max2 === void 0 ? void 0 : _rangeBounds$max2.getMonth();
        setYear((_props$panelActiveDat = (_props$panelActiveDat2 = props.panelActiveDate) === null || _props$panelActiveDat2 === void 0 ? void 0 : _props$panelActiveDat2.year) !== null && _props$panelActiveDat !== void 0 ? _props$panelActiveDat : yearFromRange);
        setMonth((_props$panelActiveDat3 = props.panelActiveDate) !== null && _props$panelActiveDat3 !== void 0 && _props$panelActiveDat3.month ? Number((_props$panelActiveDat4 = props.panelActiveDate) === null || _props$panelActiveDat4 === void 0 ? void 0 : _props$panelActiveDat4.month) - 1 : monthFromRange);
      } else {
        setYear(format.parseToDayjs(value, format$1).year());
        setMonth(format.parseToDayjs(value, format$1).month());
      }
      setTime(format.formatTime(value, format$1, timeFormat, defaultTime));
    } else {
      setIsHoverCell(false);
    }
  }, [popupVisible]);
  function onCellMouseEnter(date) {
    if (multiple) return;
    setIsHoverCell(true);
    setInputValue(format.formatDate(date, {
      format: format$1
    }));
  }
  function onCellMouseLeave() {
    if (multiple) return;
    setIsHoverCell(false);
    setInputValue(format.formatDate(cacheValue, {
      format: format$1
    }));
  }
  function onCellClick(date) {
    onPick === null || onPick === void 0 || onPick(date);
    setIsHoverCell(false);
    if (mode === "date") {
      setYear(date.getFullYear());
      setMonth(date.getMonth());
    }
    if (enableTimePicker) {
      setCacheValue(format.formatDate(date, {
        format: format$1
      }));
      if (props.needConfirm) return;
      onChange(format.formatDate(date, {
        format: format$1,
        targetFormat: valueType
      }), {
        dayjsValue: format.parseToDayjs(date, format$1),
        trigger: "pick"
      });
    } else {
      if (multiple) {
        var newDate = processDate(date);
        onChange(newDate, {
          dayjsValue: format.parseToDayjs(date, format$1),
          trigger: "pick"
        });
        return;
      }
      handlePopupInvisible();
      onChange(format.formatDate(date, {
        format: format$1,
        targetFormat: valueType
      }), {
        dayjsValue: format.parseToDayjs(date, format$1),
        trigger: "pick"
      });
    }
  }
  var onJumperClick = React__default["default"].useCallback(function (_ref) {
    var trigger = _ref.trigger;
    var monthCountMap = {
      date: 1,
      week: 1,
      month: 12,
      quarter: 12,
      year: 120
    };
    var monthCount = monthCountMap[mode] || 0;
    var current = new Date(year, month);
    var next = null;
    if (trigger === "prev") {
      next = utils.subtractMonth(current, monthCount);
    } else if (trigger === "current") {
      next = new Date();
    } else if (trigger === "next") {
      next = utils.addMonth(current, monthCount);
    }
    var nextYear = next.getFullYear();
    var nextMonth = next.getMonth();
    setYear(nextYear);
    setMonth(nextMonth);
    if (nextYear !== year) {
      var _props$onYearChange;
      (_props$onYearChange = props.onYearChange) === null || _props$onYearChange === void 0 || _props$onYearChange.call(props, {
        year: nextYear,
        date: format.parseToDayjs(value, format$1).toDate(),
        trigger: trigger === "current" ? "today" : "year-".concat(datePicker_utils.triggerMap[trigger])
      });
    }
    if (nextMonth !== month) {
      var _props$onMonthChange;
      (_props$onMonthChange = props.onMonthChange) === null || _props$onMonthChange === void 0 || _props$onMonthChange.call(props, {
        month: nextMonth,
        date: format.parseToDayjs(value, format$1).toDate(),
        trigger: trigger === "current" ? "today" : "month-".concat(datePicker_utils.triggerMap[trigger])
      });
    }
  }, [year, month, mode, setYear, setMonth, value]);
  function onTimePickerChange(val) {
    setTime(val);
    var _extractTimeObj = utils.extractTimeObj(val),
      hours = _extractTimeObj.hours,
      minutes = _extractTimeObj.minutes,
      seconds = _extractTimeObj.seconds,
      milliseconds = _extractTimeObj.milliseconds,
      meridiem = _extractTimeObj.meridiem;
    var nextHours = hours;
    if (/am/i.test(meridiem) && nextHours === 12) nextHours -= 12;
    if (/pm/i.test(meridiem) && nextHours < 12) nextHours += 12;
    var currentDate = !dayjs__default["default"](inputValue, format$1).isValid() ? dayjs__default["default"]() : dayjs__default["default"](inputValue, format$1);
    var nextDate = currentDate.hour(nextHours).minute(minutes).second(seconds).millisecond(milliseconds).toDate();
    setInputValue(format.formatDate(nextDate, {
      format: format$1
    }));
    setCacheValue(format.formatDate(nextDate, {
      format: format$1
    }));
    onPick === null || onPick === void 0 || onPick(nextDate);
  }
  function onConfirmClick(_ref2) {
    var _props$onConfirm;
    var e = _ref2.e;
    var nextValue = format.formatDate(inputValue, {
      format: format$1
    });
    props === null || props === void 0 || (_props$onConfirm = props.onConfirm) === null || _props$onConfirm === void 0 || _props$onConfirm.call(props, {
      e: e,
      date: nextValue
    });
    handlePopupInvisible();
    if (nextValue) {
      onChange(format.formatDate(inputValue, {
        format: format$1,
        targetFormat: valueType
      }), {
        dayjsValue: format.parseToDayjs(inputValue, format$1),
        trigger: "confirm"
      });
    } else {
      setInputValue(format.formatDate(value, {
        format: format$1
      }));
    }
  }
  function onPresetClick(preset, context) {
    var _props$onPresetClick;
    var presetValue = preset;
    if (typeof preset === "function") {
      presetValue = preset();
    }
    var formattedPreset = format.formatDate(presetValue, {
      format: format$1,
      targetFormat: valueType
    });
    var formattedInput = format.formatDate(presetValue, {
      format: format$1
    });
    setInputValue(formattedInput);
    setCacheValue(formattedInput);
    setTime(format.formatTime(presetValue, format$1, timeFormat, props.defaultTime));
    setYear(format.parseToDayjs(presetValue, format$1).year());
    setMonth(format.parseToDayjs(presetValue, format$1).month());
    handlePopupInvisible();
    onChange(formattedPreset, {
      dayjsValue: format.parseToDayjs(presetValue, format$1),
      trigger: "preset"
    });
    (_props$onPresetClick = props.onPresetClick) === null || _props$onPresetClick === void 0 || _props$onPresetClick.call(props, context);
  }
  var onYearChange = React.useCallback(function (year2) {
    var _props$onYearChange2;
    setYear(year2);
    (_props$onYearChange2 = props.onYearChange) === null || _props$onYearChange2 === void 0 || _props$onYearChange2.call(props, {
      year: year2,
      date: format.parseToDayjs(value, format$1).toDate(),
      trigger: "year-select"
    });
  }, [value]);
  var onMonthChange = React.useCallback(function (month2) {
    var _props$onMonthChange2;
    setMonth(month2);
    (_props$onMonthChange2 = props.onMonthChange) === null || _props$onMonthChange2 === void 0 || _props$onMonthChange2.call(props, {
      month: month2,
      date: format.parseToDayjs(value, format$1).toDate(),
      trigger: "month-select"
    });
  }, [value]);
  function processDate(date) {
    var _currentDate;
    var isSameDate;
    var currentValue = value || [];
    if (mode !== "week") isSameDate = currentValue.some(function (val) {
      return utils.isSame(format.parseToDayjs(val, format$1).toDate(), date, mode, local.dayjsLocale);
    });else {
      isSameDate = currentValue.some(function (val) {
        return val === dayjs__default["default"](date).locale(local.dayjsLocale).format(format$1);
      });
    }
    var currentDate;
    if (!isSameDate) {
      currentDate = currentValue.concat(format.formatDate(date, {
        format: format$1,
        targetFormat: valueType
      }));
    } else {
      currentDate = currentValue.filter(function (val) {
        return format.formatDate(val, {
          format: format$1,
          targetFormat: valueType
        }) !== format.formatDate(date, {
          format: format$1,
          targetFormat: valueType
        });
      });
    }
    return (_currentDate = currentDate) === null || _currentDate === void 0 ? void 0 : _currentDate.sort(function (a, b) {
      return dayjs__default["default"](a).valueOf() - dayjs__default["default"](b).valueOf();
    });
  }
  var onTagRemoveClick = function onTagRemoveClick(ctx) {
    if (["week", "quarter"].includes(props.mode)) {
      onChange === null || onChange === void 0 || onChange(ctx.value, {
        trigger: "tag-remove"
      });
      return;
    }
    var removeDate = dayjs__default["default"](ctx.item).toDate();
    var newDate = processDate(removeDate);
    onChange === null || onChange === void 0 || onChange(newDate, {
      dayjsValue: format.parseToDayjs(removeDate, format$1),
      trigger: "tag-remove"
    });
  };
  var onTagClearClick = function onTagClearClick(_ref3) {
    var e = _ref3.e;
    e.stopPropagation();
    handlePopupInvisible();
    onChange([], {
      dayjsValue: dayjs__default["default"](),
      trigger: "clear"
    });
    onClear === null || onClear === void 0 || onClear({
      e: e
    });
  };
  var panelProps = {
    value: cacheValue,
    year: year,
    month: month,
    mode: mode,
    format: format$1,
    presets: presets,
    time: multiple ? false : time,
    disableDate: disableDate,
    firstDayOfWeek: firstDayOfWeek,
    timePickerProps: timePickerProps,
    enableTimePicker: multiple ? false : enableTimePicker,
    presetsPlacement: presetsPlacement,
    popupVisible: popupVisible,
    needConfirm: needConfirm,
    multiple: multiple,
    range: range,
    cell: cell,
    onCellClick: onCellClick,
    onCellMouseEnter: onCellMouseEnter,
    onCellMouseLeave: onCellMouseLeave,
    onJumperClick: onJumperClick,
    onConfirmClick: onConfirmClick,
    onPresetClick: onPresetClick,
    onYearChange: onYearChange,
    onMonthChange: onMonthChange,
    onTimePickerChange: onTimePickerChange,
    disableTime: disableTime
  };
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"]("".concat(classPrefix, "-date-picker"), className),
    style: style,
    ref: ref
  }, /* @__PURE__ */React__default["default"].createElement(selectInput_index.SelectInput, {
    disabled: disabled,
    value: inputValue,
    status: props.status,
    tips: props.tips,
    borderless: props.borderless,
    label: label,
    popupProps: popupProps,
    inputProps: inputProps,
    popupVisible: popupVisible,
    panel: /* @__PURE__ */React__default["default"].createElement(datePicker_panel_SinglePanel["default"], _objectSpread({}, panelProps)),
    multiple: multiple,
    tagInputProps: _objectSpread({
      onRemove: onTagRemoveClick
    }, tagInputProps),
    onClear: onTagClearClick
  }));
});
DatePicker.displayName = "DatePicker";

exports["default"] = DatePicker;
