// tdesign-react/esm/date-picker/panel/SinglePanel.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-5e72c8b9.js';
import React, { forwardRef } from 'react';
import classNames from 'classnames';
import { isFunction } from 'lodash-es';
import { g as getDefaultFormat, p as parseToDayjs } from '../../_chunks/dep-07ae3139.js';
import useConfig from '../../hooks/useConfig.js';
import PanelContent from './PanelContent.js';
import ExtraContent from './ExtraContent.js';
import useTableData from '../hooks/useTableData.js';
import useDisableDate from '../hooks/useDisableDate.js';
import useDefaultProps from '../../hooks/useDefaultProps.js';
import { parseToDateTime } from '../utils.js';
import '../../_chunks/dep-43f7da7b.js';
import 'dayjs';
import '../../_chunks/dep-7d48ddac.js';
import '../../_chunks/dep-9b727dd4.js';
import '../../_chunks/dep-3561cd86.js';
import '../../_chunks/dep-65d82dda.js';
import '../../config-provider/ConfigContext.js';
import '../../locale/zh_CN.js';
import '../../_chunks/dep-766c4603.js';
import '../base/Header.js';
import '../../_chunks/dep-1e9f924f.js';
import '../../_chunks/dep-8e8ac0d4.js';
import '../../_chunks/dep-57e67945.js';
import '../../_chunks/dep-94a7af64.js';
import '../../_chunks/dep-0054b574.js';
import '../../locale/LocalReceiver.js';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';
import '../../config-provider/type.js';
import '../../pagination/index.js';
import '../../pagination/Pagination.js';
import '../../_chunks/dep-4461a684.js';
import '../../_util/noop.js';
import '../../hooks/useControlled.js';
import '../../select/index.js';
import '../../select/base/Select.js';
import '../../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../../_util/helper.js';
import '../../_util/parseTNode.js';
import '../../_util/ref.js';
import 'react-is';
import '../../_util/isFragment.js';
import '../../common/FakeArrow.js';
import '../../_chunks/dep-9721fff3.js';
import '../../loading/index.js';
import '../../loading/Loading.js';
import '../../_chunks/dep-a347b1b5.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../hooks/useLayoutEffect.js';
import '../../loading/defaultProps.js';
import '../../loading/gradient.js';
import '../../_chunks/dep-2481dfad.js';
import '../../hooks/useDomRefCallback.js';
import '../../loading/plugin.js';
import '../../_util/react-render.js';
import '../../_chunks/dep-fe79dc5d.js';
import '../../common/PluginContainer.js';
import '../../loading/style/index.js';
import '../../loading/type.js';
import '../../select-input/index.js';
import '../../select-input/SelectInput.js';
import '../../popup/index.js';
import '../../popup/Popup.js';
import 'react-transition-group';
import '../../hooks/useAnimation.js';
import '../../hooks/useAttach.js';
import '../../hooks/useMutationObserver.js';
import '../../hooks/useLatest.js';
import '../../hooks/usePopper.js';
import 'react-fast-compare';
import '@popperjs/core';
import '../../hooks/useWindowSize.js';
import '../../popup/defaultProps.js';
import '../../popup/hooks/useTrigger.js';
import '../../_util/listener.js';
import '../../popup/utils/transition.js';
import '../../popup/PopupPlugin.js';
import '../../popup/style/index.js';
import '../../popup/type.js';
import '../../select-input/defaultProps.js';
import '../../select-input/useMultiple.js';
import '../../tag-input/index.js';
import '../../tag-input/TagInput.js';
import 'tdesign-icons-react';
import '../../hooks/useDragSorter.js';
import '../../hooks/useEventCallback.js';
import '../../hooks/useGlobalIcon.js';
import '../../input/index.js';
import '../../input/Input.js';
import '../../input/defaultProps.js';
import '../../input/InputGroup.js';
import '../../input/useLengthLimit.js';
import '../../input/style/index.js';
import '../../input/type.js';
import '../../tag-input/defaultProps.js';
import '../../tag-input/useHover.js';
import '../../tag-input/useTagList.js';
import '../../tag/index.js';
import '../../tag/Tag.js';
import '../../_chunks/dep-dfdbe6bd.js';
import '../../tag/defaultProps.js';
import '../../tag/CheckTag.js';
import '../../_chunks/dep-29a70a9d.js';
import '../../tag/CheckTagGroup.js';
import '../../tag/style/index.js';
import '../../tag/type.js';
import '../../tag-input/useTagScroll.js';
import '../../tag-input/style/index.js';
import '../../tag-input/type.js';
import '../../select-input/useOverlayInnerStyle.js';
import '../../hooks/useInnerPopupVisible.js';
import '../../select-input/useSingle.js';
import '../../select-input/style/index.js';
import '../../select-input/type.js';
import '../../select/defaultProps.js';
import '../../select/hooks/useKeyboardControl.js';
import '../../_chunks/dep-29be6164.js';
import '../../hooks/useRipple.js';
import '../../select/base/OptionGroup.js';
import '../../select/base/PopupContent.js';
import '../../select/hooks/usePanelVirtualScroll.js';
import '../../hooks/useVirtualScroll.js';
import '../../select/style/index.js';
import '../../select/type.js';
import '../../input-number/index.js';
import '../../input-number/InputNumber.js';
import '../../button/index.js';
import '../../button/Button.js';
import '../../button/defaultProps.js';
import '../../button/style/index.js';
import '../../button/type.js';
import '../../input-number/useInputNumber.js';
import '../../_chunks/dep-35a55ece.js';
import '../../hooks/useCommonClassName.js';
import '../../input-number/defaultProps.js';
import '../../input-number/style/index.js';
import '../../input-number/type.js';
import '../../input-adornment/index.js';
import '../../input-adornment/InputAdornment.js';
import '../../input-adornment/style/index.js';
import '../../input-adornment/type.js';
import '../../pagination/hooks/useBoundaryJumper.js';
import '../../pagination/hooks/usePrevNextJumper.js';
import '../../pagination/hooks/usePageNumber.js';
import '../../pagination/hooks/useTotal.js';
import '../../pagination/validators.js';
import '../../pagination/defaultProps.js';
import '../../pagination/PaginationMini.js';
import '../../pagination/style/index.js';
import '../../pagination/type.js';
import '../hooks/useSelectRange.js';
import '../../_chunks/dep-7ff1070c.js';
import '../../_chunks/dep-35769bc6.js';
import '../base/Table.js';
import '../base/Cell.js';
import '../../time-picker/index.js';
import '../../time-picker/TimePicker.js';
import '../../_chunks/dep-35931e80.js';
import '../../time-picker/hooks/useTimePickerTextConfig.js';
import '../../hooks/useDebounce.js';
import '../../hooks/usePersistFn.js';
import '../../hooks/usePropsRef.js';
import '../../time-picker/TimeRangePicker.js';
import '../../range-input/index.js';
import '../../range-input/RangeInput.js';
import '../../range-input/defaultProps.js';
import '../../range-input/RangeInputPopup.js';
import '../../range-input/style/index.js';
import '../../range-input/type.js';
import '../../time-picker/panel/TimePickerPanel.js';
import '../../time-picker/defaultProps.js';
import '../../time-picker/style/index.js';
import '../../time-picker/type.js';
import '../base/Footer.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var SinglePanel = /*#__PURE__*/forwardRef(function (originalProps, ref) {
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix,
    globalDatePickerConfig = _useConfig.datePicker;
  var panelName = "".concat(classPrefix, "-date-picker__panel");
  var props = useDefaultProps(originalProps, {
    mode: "date",
    enableTimePicker: false,
    presetsPlacement: "bottom"
  });
  var value = props.value,
    mode = props.mode,
    presetsPlacement = props.presetsPlacement,
    _props$firstDayOfWeek = props.firstDayOfWeek,
    firstDayOfWeek = _props$firstDayOfWeek === void 0 ? globalDatePickerConfig.firstDayOfWeek : _props$firstDayOfWeek,
    needConfirm = props.needConfirm,
    style = props.style,
    className = props.className,
    year = props.year,
    month = props.month,
    range = props.range,
    cell = props.cell,
    onPanelClick = props.onPanelClick,
    disableTime = props.disableTime;
  var _getDefaultFormat = getDefaultFormat({
      mode: props.mode,
      format: props.format,
      enableTimePicker: props.enableTimePicker
    }),
    format = _getDefaultFormat.format;
  var disableDateOptions = useDisableDate({
    disableDate: props.disableDate,
    mode: props.mode,
    format: format
  });
  var disableTimeOptions = function disableTimeOptions(h, m, s, ms) {
    if (!isFunction(disableTime) || !value) {
      return {};
    }
    return disableTime(parseToDateTime(value, format, [h, m, s, ms]));
  };
  var tableData = useTableData(_objectSpread({
    value: value,
    year: year,
    month: month,
    range: range,
    mode: mode,
    start: value ? parseToDayjs(props.multiple ? value[0] : value, format).toDate() : void 0,
    firstDayOfWeek: firstDayOfWeek,
    multiple: props.multiple
  }, disableDateOptions));
  var panelContentProps = {
    mode: mode,
    value: value,
    year: year,
    month: month,
    range: range,
    format: format,
    cell: cell,
    firstDayOfWeek: firstDayOfWeek,
    tableData: tableData,
    popupVisible: props.popupVisible,
    multiple: props.multiple,
    time: props.time,
    timePickerProps: _objectSpread({
      disableTime: disableTimeOptions
    }, props.timePickerProps),
    enableTimePicker: props.enableTimePicker,
    onMonthChange: props.onMonthChange,
    onYearChange: props.onYearChange,
    onJumperClick: props.onJumperClick,
    onCellClick: props.onCellClick,
    onCellMouseEnter: props.onCellMouseEnter,
    onCellMouseLeave: props.onCellMouseLeave,
    onTimePickerChange: props.onTimePickerChange
  };
  var extraProps = {
    presets: props.presets,
    enableTimePicker: props.enableTimePicker,
    presetsPlacement: props.presetsPlacement,
    onPresetClick: props.onPresetClick,
    onConfirmClick: props.onConfirmClick,
    selectedValue: props.value,
    needConfirm: needConfirm
  };
  return /* @__PURE__ */React.createElement("div", {
    ref: ref,
    style: style,
    className: classNames(panelName, className, _defineProperty({}, "".concat(panelName, "--direction-row"), ["left", "right"].includes(presetsPlacement))),
    onClick: function onClick(e) {
      return onPanelClick === null || onPanelClick === void 0 ? void 0 : onPanelClick({
        e: e
      });
    }
  }, ["top", "left"].includes(presetsPlacement) ? /* @__PURE__ */React.createElement(ExtraContent, _objectSpread({}, extraProps)) : null, /* @__PURE__ */React.createElement(PanelContent, _objectSpread({}, panelContentProps)), ["bottom", "right"].includes(presetsPlacement) ? /* @__PURE__ */React.createElement(ExtraContent, _objectSpread({}, extraProps)) : null);
});
SinglePanel.displayName = "SinglePanel";

export { SinglePanel as default };
