// tdesign-react/cjs/empty/Empty.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
var empty_defaultProps = require('./defaultProps.js');
var image_index = require('../image/index.js');
var empty_assets_MaintenanceSvg = require('./assets/MaintenanceSvg.js');
var empty_assets_NetworkErrorSvg = require('./assets/NetworkErrorSvg.js');
var empty_assets_EmptySvg = require('./assets/EmptySvg.js');
var empty_assets_FailSvg = require('./assets/FailSvg.js');
var empty_assets_SuccessSvg = require('./assets/SuccessSvg.js');
var isObject = require('../_chunks/dep-aff91ab0.js');
var isString = require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../image/Image.js');
require('../_chunks/dep-1c586f5c.js');
require('tdesign-icons-react');
require('../_chunks/dep-e31ce0e8.js');
require('../image/defaultProps.js');
require('../space/index.js');
require('../space/Space.js');
require('react-is');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-f0e884b9.js');
require('../space/defaultProps.js');
require('../hooks/useGlobalIcon.js');
require('../hooks/useImagePreviewUrl.js');
require('../_chunks/dep-b289b0be.js');
require('../_chunks/dep-442716ca.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function getImageIns(data) {
  var result = data;
  if (/*#__PURE__*/React.isValidElement(data)) {
    result = data;
  } else if (isObject.isObject(data)) {
    result = /* @__PURE__ */React__default["default"].createElement(image_index.Image, _objectSpread({}, data));
  } else if (isString.isString(data)) {
    result = /* @__PURE__ */React__default["default"].createElement(image_index.Image, {
      src: data
    });
  }
  return data ? result : null;
}
var Empty = function Empty(props) {
  var _defaultMaps$type, _typeImageProps$image, _typeImageProps$title;
  var _useDefaultProps = hooks_useDefaultProps["default"](props, empty_defaultProps.emptyDefaultProps),
    propsImage = _useDefaultProps.image,
    imageStyle = _useDefaultProps.imageStyle,
    propsDescription = _useDefaultProps.description,
    propsTitle = _useDefaultProps.title,
    type = _useDefaultProps.type,
    action = _useDefaultProps.action,
    style = _useDefaultProps.style,
    className = _useDefaultProps.className,
    _useDefaultProps$size = _useDefaultProps.size,
    size = _useDefaultProps$size === void 0 ? "medium" : _useDefaultProps$size;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix,
    empty = _useConfig.empty;
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("empty"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var defaultMaps = {
    maintenance: {
      image: empty.image.maintenance || /* @__PURE__ */React__default["default"].createElement(empty_assets_MaintenanceSvg["default"], null),
      title: empty.titleText.maintenance || t(local.titleText.maintenance)
    },
    success: {
      image: empty.image.success || /* @__PURE__ */React__default["default"].createElement(empty_assets_SuccessSvg["default"], null),
      title: empty.titleText.success || t(local.titleText.success)
    },
    fail: {
      image: empty.image.fail || /* @__PURE__ */React__default["default"].createElement(empty_assets_FailSvg["default"], null),
      title: empty.titleText.fail || t(local.titleText.fail)
    },
    "network-error": {
      image: empty.image.networkError || /* @__PURE__ */React__default["default"].createElement(empty_assets_NetworkErrorSvg["default"], null),
      title: empty.titleText.networkError || t(local.titleText.networkError)
    },
    empty: {
      image: empty.image.empty || /* @__PURE__ */React__default["default"].createElement(empty_assets_EmptySvg["default"], null),
      title: empty.titleText.empty || t(local.titleText.empty)
    }
  };
  var defaultSize = {
    small: "".concat(classPrefix, "-size-s"),
    medium: "".concat(classPrefix, "-size"),
    large: "".concat(classPrefix, "-size-l")
  };
  var prefix = "".concat(classPrefix, "-empty");
  var emptyClasses = classNames__default["default"](prefix, className, defaultSize[size]);
  var titleClasses = classNames__default["default"]("".concat(prefix, "__title"));
  var imageClasses = classNames__default["default"]("".concat(prefix, "__image"));
  var descriptionClasses = classNames__default["default"]("".concat(prefix, "__description"));
  var actionCls = classNames__default["default"]("".concat(prefix, "__action"));
  var typeImageProps = (_defaultMaps$type = defaultMaps[type]) !== null && _defaultMaps$type !== void 0 ? _defaultMaps$type : null;
  var _image$title$descript = {
      image: propsImage ? propsImage : (_typeImageProps$image = typeImageProps === null || typeImageProps === void 0 ? void 0 : typeImageProps.image) !== null && _typeImageProps$image !== void 0 ? _typeImageProps$image : null,
      title: propsTitle ? propsTitle : (_typeImageProps$title = typeImageProps === null || typeImageProps === void 0 ? void 0 : typeImageProps.title) !== null && _typeImageProps$title !== void 0 ? _typeImageProps$title : null,
      description: propsDescription
    },
    image = _image$title$descript.image,
    description = _image$title$descript.description,
    title = _image$title$descript.title;
  function renderTitle() {
    if (!title) {
      return null;
    }
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: titleClasses
    }, title);
  }
  function renderDescription() {
    if (!description) {
      return null;
    }
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: descriptionClasses
    }, description);
  }
  var imageContent = getImageIns(image);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: emptyClasses,
    style: style
  }, imageContent ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: imageClasses,
    style: imageStyle
  }, imageContent) : null, renderTitle(), renderDescription(), action ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: actionCls
  }, action) : null);
};

exports["default"] = Empty;
