// tdesign-react/es/table/TBody.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-5e72c8b9.js';
import { _ as _slicedToArray } from '../_chunks/dep-94a7af64.js';
import React, { useMemo } from 'react';
import classNames from 'classnames';
import { camelCase, pick, get } from 'lodash-es';
import { useLocaleReceiver } from '../locale/LocalReceiver.js';
import useRowspanAndColspan from './hooks/useRowspanAndColspan.js';
import TR, { ROW_LISTENERS, TABLE_PROPS } from './TR.js';
import '../_chunks/dep-43f7da7b.js';
import '../_chunks/dep-0054b574.js';
import '../_chunks/dep-8e8ac0d4.js';
import '../config-provider/index.js';
import '../config-provider/ConfigProvider.js';
import '../config-provider/ConfigContext.js';
import '../locale/zh_CN.js';
import '../_chunks/dep-7d48ddac.js';
import 'dayjs';
import '../_chunks/dep-766c4603.js';
import '../config-provider/type.js';
import '../_chunks/dep-3561cd86.js';
import '../_chunks/dep-65d82dda.js';
import '../hooks/useLayoutEffect.js';
import '../_chunks/dep-a347b1b5.js';
import './Cell.js';
import './Ellipsis.js';
import '../_util/scroll.js';
import 'raf';
import '../_util/easing.js';
import '../tooltip/index.js';
import '../tooltip/Tooltip.js';
import '../_chunks/dep-4461a684.js';
import '../popup/index.js';
import '../popup/Popup.js';
import 'react-transition-group';
import '../common/Portal.js';
import 'react-dom';
import '../hooks/useConfig.js';
import '../hooks/useAnimation.js';
import '../hooks/useAttach.js';
import '../hooks/useControlled.js';
import '../_util/noop.js';
import '../hooks/useDefaultProps.js';
import '../hooks/useMutationObserver.js';
import '../hooks/useLatest.js';
import '../hooks/usePopper.js';
import '../_chunks/dep-1e9f924f.js';
import '../_chunks/dep-57e67945.js';
import 'react-fast-compare';
import '@popperjs/core';
import '../hooks/useWindowSize.js';
import '../popup/defaultProps.js';
import '../popup/hooks/useTrigger.js';
import '../_util/listener.js';
import '../_util/ref.js';
import 'react-is';
import '../_util/isFragment.js';
import '../popup/utils/transition.js';
import '../popup/PopupPlugin.js';
import '../_chunks/dep-fe79dc5d.js';
import '../_util/react-render.js';
import '../common/PluginContainer.js';
import '../popup/style/css.js';
import '../popup/type.js';
import '../tooltip/defaultProps.js';
import '../tooltip/TooltipLite.js';
import '../hooks/useSwitch.js';
import '../hooks/usePersistFn.js';
import '../tooltip/style/css.js';
import '../tooltip/type.js';
import '../hooks/useDebounce.js';
import './hooks/useFixed.js';
import '../_chunks/dep-eb4e35fe.js';
import '../_chunks/dep-9721fff3.js';
import '../hooks/useDeepEffect.js';
import '../hooks/usePrevious.js';
import './utils.js';
import './hooks/useClassName.js';
import './hooks/useLazyLoad.js';
import '../_chunks/dep-be76de2c.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var ROW_AND_TD_LISTENERS = ROW_LISTENERS.concat("cell-click");
var extendTableProps = ["rowKey", "rowClassName", "rowAttributes", "loading", "empty", "fixedRows", "firstFullRow", "lastFullRow", "rowspanAndColspan", "scroll", "cellEmptyContent", "onCellClick", "onPageChange", "onRowClick", "onRowDblclick", "onRowMouseover", "onRowMousedown", "onRowMouseenter", "onRowMouseleave", "onRowMouseup", "onScroll", "onScrollX", "onScrollY"];
var trProperties = ["classPrefix", "ellipsisOverlayClassName", "rowAndColFixedPosition", "scroll", "tableRef", "tableContentRef", "trs", "bufferSize", "isVirtual", "rowHeight", "scrollType"];
function TBody(props) {
  var data = props.data,
    columns = props.columns,
    rowKey = props.rowKey,
    firstFullRow = props.firstFullRow,
    lastFullRow = props.lastFullRow,
    virtualConfig = props.virtualConfig,
    allTableClasses = props.allTableClasses;
  var isVirtualScroll = virtualConfig.isVirtualScroll;
  var renderData = isVirtualScroll ? virtualConfig.visibleData : data;
  var _useLocaleReceiver = useLocaleReceiver("table"),
    _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
    global = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var _useRowspanAndColspan = useRowspanAndColspan(renderData, columns, rowKey, props.rowspanAndColspan),
    skipSpansMap = _useRowspanAndColspan.skipSpansMap;
  var isSkipSnapsMapNotFinish = Boolean(props.rowspanAndColspan && !skipSpansMap.size);
  var tableFullRowClasses = allTableClasses.tableFullRowClasses,
    tableBaseClass = allTableClasses.tableBaseClass;
  var tbodyClasses = useMemo(function () {
    return [tableBaseClass.body];
  }, [tableBaseClass.body]);
  var hasFullRowConfig = useMemo(function () {
    return firstFullRow || lastFullRow;
  }, [firstFullRow, lastFullRow]);
  var isEmpty = !(data !== null && data !== void 0 && data.length) && !props.loading && !hasFullRowConfig;
  var renderEmpty = function renderEmpty(columns2) {
    return /* @__PURE__ */React.createElement("tr", {
      className: classNames([tableBaseClass.emptyRow, _defineProperty({}, tableFullRowClasses.base, props.isWidthOverflow)])
    }, /* @__PURE__ */React.createElement("td", {
      colSpan: columns2.length
    }, /* @__PURE__ */React.createElement("div", {
      className: classNames([tableBaseClass.empty, _defineProperty({}, tableFullRowClasses.innerFullRow, props.isWidthOverflow)]),
      style: props.isWidthOverflow ? {
        width: "".concat(props.tableWidth, "px")
      } : {}
    }, props.empty || t(global.empty))));
  };
  var renderFullRow = function renderFullRow(type) {
    var fullRowNode = {
      "first-full-row": firstFullRow,
      "last-full-row": lastFullRow
    }[type];
    if (!fullRowNode) return null;
    var rowProps = {};
    if (isVirtualScroll) {
      var fullRowIndex = {
        "first-full-row": 0,
        "last-full-row": data.length - 1
      }[type];
      var rowVisible = renderData.some(function (item) {
        return item.__VIRTUAL_SCROLL_INDEX === fullRowIndex;
      });
      if (!rowVisible) return null;
      rowProps.ref = function (ref) {
        if (ref) {
          props === null || props === void 0 || props.handleRowMounted({
            ref: ref,
            data: data[fullRowIndex]
          });
        }
      };
    }
    var isFixedToLeft = props.isWidthOverflow && columns.find(function (col) {
      return col.fixed === "left";
    });
    var tType = camelCase(type);
    var classes = [tableFullRowClasses.base, tableFullRowClasses[tType]];
    return /* @__PURE__ */React.createElement("tr", _objectSpread({
      key: type,
      className: classNames(classes)
    }, rowProps), /* @__PURE__ */React.createElement("td", {
      colSpan: columns.length
    }, /* @__PURE__ */React.createElement("div", {
      className: classNames(_defineProperty({}, tableFullRowClasses.innerFullRow, isFixedToLeft)) || void 0,
      style: isFixedToLeft ? {
        width: "".concat(props.tableWidth, "px")
      } : {}
    }, /* @__PURE__ */React.createElement("div", {
      className: tableFullRowClasses.innerFullElement
    }, fullRowNode))));
  };
  var renderExpandedRows = function renderExpandedRows(row, rowIndex) {
    var _props$renderExpanded;
    var p = {
      row: row,
      index: rowIndex,
      columns: columns,
      tableWidth: props.tableWidth,
      isWidthOverflow: props.isWidthOverflow
    };
    return (_props$renderExpanded = props.renderExpandedRow) === null || _props$renderExpanded === void 0 ? void 0 : _props$renderExpanded.call(props, p);
  };
  var renderTRNodeList = function renderTRNodeList() {
    if (isSkipSnapsMapNotFinish) return null;
    var trNodeList = [];
    var firstFullRowNode = renderFullRow("first-full-row");
    firstFullRowNode && trNodeList.push(firstFullRowNode);
    renderData === null || renderData === void 0 || renderData.forEach(function (row, rowIndex) {
      if (row.__VIRTUAL_FAKE_DATA) return;
      var getRowIndex = function getRowIndex() {
        var virtualIndex = row.__VIRTUAL_SCROLL_INDEX;
        if (isVirtualScroll && firstFullRow && virtualIndex) {
          return virtualIndex - 1;
        }
        return virtualIndex !== null && virtualIndex !== void 0 ? virtualIndex : rowIndex;
      };
      var trProps = _objectSpread(_objectSpread(_objectSpread({}, pick(props, TABLE_PROPS)), {}, {
        rowKey: props.rowKey || "id",
        row: row,
        columns: columns,
        rowIndex: getRowIndex(),
        dataLength: data.length,
        skipSpansMap: skipSpansMap,
        virtualConfig: virtualConfig,
        classPrefix: props.classPrefix,
        ellipsisOverlayClassName: props.ellipsisOverlayClassName
      }, pick(props, trProperties)), {}, {
        pagination: props.pagination
      });
      if (props.onCellClick) {
        trProps.onCellClick = props.onCellClick;
      }
      var trNode = /* @__PURE__ */React.createElement(TR, _objectSpread(_objectSpread({
        key: get(row, props.rowKey || "id") || rowIndex
      }, trProps), {}, {
        onRowMounted: props.handleRowMounted
      }));
      trNodeList.push(trNode);
      var expandedRowNode = renderExpandedRows(row, rowIndex);
      expandedRowNode && trNodeList.push(expandedRowNode);
    });
    var lastFullRowNode = renderFullRow("last-full-row");
    lastFullRowNode && trNodeList.push(lastFullRowNode);
    return trNodeList;
  };
  var translate = "translateY(".concat(virtualConfig.translateY, "px)");
  var posStyle = isVirtualScroll ? {
    transform: translate,
    msTransform: translate,
    MozTransform: translate,
    WebkitTransform: translate
  } : void 0;
  return /* @__PURE__ */React.createElement("tbody", {
    className: classNames(tbodyClasses) || void 0,
    style: _objectSpread({}, posStyle)
  }, isEmpty ? renderEmpty(columns) : renderTRNodeList());
}
TBody.displayName = "TBody";

export { ROW_AND_TD_LISTENERS, TBody as default, extendTableProps };
