// tdesign-react/cjs/space/Space.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var require$$0 = require('react-is');
var helper = require('../_chunks/dep-b011b5d3.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var space_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-1c586f5c.js');
require('../_chunks/dep-f0e884b9.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-794b42ce.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var SizeMap = {
  small: "8px",
  medium: "16px",
  large: "24px"
};
var defaultNeedPolyfill = helper.getFlexGapPolyFill();
var _toArray = function toArray(children) {
  var ret = [];
  React__default["default"].Children.forEach(children, function (child) {
    if (child === void 0 || child === null) {
      return;
    }
    if (Array.isArray(child)) {
      ret = ret.concat(_toArray(child));
    } else if (require$$0.isFragment(child) && child.props) {
      ret = ret.concat(_toArray(child.props.children));
    } else {
      ret.push(child);
    }
  });
  return ret;
};
var EMPTY_NODE = ["", false, null, void 0];
var Space = /*#__PURE__*/React__default["default"].forwardRef(function (originalProps, ref) {
  var props = hooks_useDefaultProps["default"](originalProps, space_defaultProps.spaceDefaultProps);
  var className = props.className,
    style = props.style,
    align = props.align,
    direction = props.direction,
    size = props.size,
    breakLine = props.breakLine,
    separator = props.separator,
    forceFlexGapPolyfill = props.forceFlexGapPolyfill;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var needPolyfill = Boolean(forceFlexGapPolyfill || defaultNeedPolyfill);
  var renderStyle = React.useMemo(function () {
    var renderGap = "";
    if (Array.isArray(size)) {
      renderGap = size.map(function (s) {
        if (typeof s === "number") {
          return "".concat(s, "px");
        }
        if (typeof s === "string") {
          return SizeMap[s] || s;
        }
        return s;
      }).join(" ");
    } else if (typeof size === "string") {
      renderGap = SizeMap[size] || size;
    } else if (typeof size === "number") {
      renderGap = "".concat(size, "px");
    }
    var tStyle = _objectSpread({}, style);
    if (needPolyfill) {
      var _renderGap$split = renderGap.split(" "),
        _renderGap$split2 = slicedToArray._slicedToArray(_renderGap$split, 2),
        columnGap = _renderGap$split2[0],
        rowGap = _renderGap$split2[1];
      tStyle["--td-space-column-gap"] = columnGap;
      tStyle["--td-space-row-gap"] = rowGap || columnGap;
    } else {
      tStyle.gap = renderGap;
    }
    return tStyle;
  }, [style, size, needPolyfill]);
  var childrenNode = React__default["default"].useMemo(function () {
    var children = _toArray(props.children);
    var childCount = React__default["default"].Children.count(children);
    return React__default["default"].Children.map(children, function (child, index) {
      var showSeparator = index + 1 !== childCount && separator;
      return /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, EMPTY_NODE.includes(child) ? null : /* @__PURE__ */React__default["default"].createElement("div", {
        className: "".concat(classPrefix, "-space-item")
      }, child), showSeparator && /* @__PURE__ */React__default["default"].createElement("div", {
        className: "".concat(classPrefix, "-space-item-separator")
      }, separator));
    });
  }, [props.children, classPrefix, separator]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: ref,
    style: renderStyle,
    className: classNames__default["default"]("".concat(classPrefix, "-space"), className, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-space-align-").concat(align), align), "".concat(classPrefix, "-space-").concat(direction), direction), "".concat(classPrefix, "-space--break-line"), breakLine), "".concat(classPrefix, "-space--polyfill"), needPolyfill))
  }, childrenNode);
});
Space.displayName = "Space";

exports.SizeMap = SizeMap;
exports["default"] = Space;
