// tdesign-react/es/color-picker/components/panel/alpha.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../../_chunks/dep-5e72c8b9.js';
import { _ as _objectWithoutProperties } from '../../../_chunks/dep-4461a684.js';
import React from 'react';
import classNames from 'classnames';
import ColorSlider from './slider.js';
import '../../../_chunks/dep-43f7da7b.js';
import '../../../_chunks/dep-8df8dad3.js';
import '../../../hooks/useMouseEvent.js';
import '../../hooks/useStyles.js';
import '../../../_chunks/dep-94a7af64.js';
import '../../../_chunks/dep-0054b574.js';
import '../../../_chunks/dep-8e8ac0d4.js';

var _excluded = ["color", "baseClassName", "onChange"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Alpha = function Alpha(props) {
  var color = props.color,
    baseClassName = props.baseClassName,
    onChange = props.onChange,
    rest = _objectWithoutProperties(props, _excluded);
  var handleChange = function handleChange(v) {
    onChange(v / 100);
  };
  var railStyle = {
    background: "linear-gradient(to right, rgba(0, 0, 0, 0), ".concat(props.color.rgb, ")")
  };
  return /* @__PURE__ */React.createElement(ColorSlider, _objectSpread({
    baseClassName: baseClassName,
    className: classNames(["".concat(baseClassName, "__alpha"), "".concat(baseClassName, "--bg-alpha")]),
    color: color,
    value: color.alpha * 100,
    onChange: handleChange,
    railStyle: railStyle,
    type: "alpha",
    maxValue: 100
  }, rest));
};
var AlphaSlider = /*#__PURE__*/React.memo(Alpha);

export { AlphaSlider as default };
