// tdesign-react/cjs/typography/ellipsis/Ellipsis.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var objectWithoutProperties = require('../../_chunks/dep-1c586f5c.js');
var React = require('react');
var classNames = require('classnames');
var typography_ellipsis_Truncate = require('./Truncate.js');
var hooks_useConfig = require('../../hooks/useConfig.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_chunks/dep-427fb3a1.js');
require('react-is');
require('../../_chunks/dep-5c3139cb.js');
require('../../_chunks/dep-b4442734.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-876d090e.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-a5b4810f.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-ea361d31.js');
require('../../_chunks/dep-2e455d91.js');
require('../../_chunks/dep-5e17a073.js');
require('../../_chunks/dep-894a5ece.js');
require('../../_chunks/dep-bcd16957.js');
require('../../_chunks/dep-63308b84.js');
require('../../_chunks/dep-048e4314.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-14013769.js');
require('../../_chunks/dep-06bf4625.js');
require('dayjs');
require('../../_chunks/dep-043eee20.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-2d5cf7d6.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["className", "children", "lines", "ellipsisClassName", "ellipsisPrefix", "onToggleExpand", "width", "onTruncate", "component", "collapsible", "expandable", "more", "less"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Ellipsis = function Ellipsis(_ref) {
  var className = _ref.className,
    children = _ref.children,
    _ref$lines = _ref.lines,
    lines = _ref$lines === void 0 ? 1 : _ref$lines,
    ellipsisClassName = _ref.ellipsisClassName,
    _ref$ellipsisPrefix = _ref.ellipsisPrefix,
    ellipsisPrefix = _ref$ellipsisPrefix === void 0 ? "..." : _ref$ellipsisPrefix,
    onToggleExpand = _ref.onToggleExpand,
    _ref$width = _ref.width,
    width = _ref$width === void 0 ? 0 : _ref$width,
    onTruncate = _ref.onTruncate,
    _ref$component = _ref.component,
    Component = _ref$component === void 0 ? "div" : _ref$component,
    _ref$collapsible = _ref.collapsible,
    collapsible = _ref$collapsible === void 0 ? false : _ref$collapsible,
    _ref$expandable = _ref.expandable,
    expandable = _ref$expandable === void 0 ? false : _ref$expandable,
    more = _ref.more,
    less = _ref.less,
    rest = objectWithoutProperties._objectWithoutProperties(_ref, _excluded);
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var symbolClassName = ellipsisClassName || "".concat(classPrefix, "-typography-ellipsis-symbol");
  var isMountRef = React.useRef(false);
  React.useEffect(function () {
    isMountRef.current = true;
  }, []);
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    isExpanded = _useState2[0],
    setIsExpanded = _useState2[1];
  var handleToggleExpand = function handleToggleExpand(e) {
    if (!expandable) return;
    if (isMountRef.current) {
      setIsExpanded(!isExpanded);
      onToggleExpand === null || onToggleExpand === void 0 || onToggleExpand(!isExpanded, e);
    }
  };
  var truncateRef = React.useRef(null);
  var _useState3 = React.useState(false),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    isTruncated = _useState4[0],
    setTruncated = _useState4[1];
  var handleTruncate = function handleTruncate(truncated) {
    if (isMountRef.current && truncated !== isTruncated) {
      setTruncated(truncated);
      if (truncated && truncateRef.current) {
        var _truncateRef$current$, _truncateRef$current;
        (_truncateRef$current$ = (_truncateRef$current = truncateRef.current).onResize) === null || _truncateRef$current$ === void 0 || _truncateRef$current$.call(_truncateRef$current);
      }
      onTruncate === null || onTruncate === void 0 || onTruncate(truncated);
    }
  };
  var componentProps = _objectSpread({
    className: className
  }, rest);
  return /* @__PURE__ */React__default["default"].createElement(Component, _objectSpread({}, componentProps), /* @__PURE__ */React__default["default"].createElement(typography_ellipsis_Truncate["default"], {
    width: width,
    lines: !isExpanded && lines,
    className: "".concat(classPrefix, "-typography-ellipsis"),
    ellipsis: /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(classPrefix, "-typography-ellipsis-symbol-wrapper")
    }, ellipsisPrefix, /* @__PURE__ */React__default["default"].createElement("span", {
      className: symbolClassName,
      onClick: handleToggleExpand
    }, more)),
    onTruncate: handleTruncate,
    ref: truncateRef,
    lineClassName: "".concat(classPrefix, "-typography-ellipsis-line")
  }, children), !isTruncated && collapsible && isExpanded && /* @__PURE__ */React__default["default"].createElement("span", {
    className: classNames__default["default"](symbolClassName, "".concat(symbolClassName, "--expanded")),
    onClick: handleToggleExpand
  }, less));
};

exports["default"] = Ellipsis;
