// tdesign-react/cjs/typography/ellipsis/useEllipsis.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var React = require('react');
var tooltip_index = require('../../tooltip/index.js');
var locale_LocalReceiver = require('../../locale/LocalReceiver.js');
var isFunction = require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../tooltip/Tooltip.js');
require('../../_chunks/dep-1c586f5c.js');
require('classnames');
require('../../popup/index.js');
require('../../popup/Popup.js');
require('react-transition-group');
require('../../_chunks/dep-68f5fe7b.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../common/Portal.js');
require('react-dom');
require('../../hooks/useConfig.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-14013769.js');
require('../../_chunks/dep-06bf4625.js');
require('dayjs');
require('../../_chunks/dep-043eee20.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../_chunks/dep-2d5cf7d6.js');
require('../../hooks/useLayoutEffect.js');
require('../../hooks/useAnimation.js');
require('../../hooks/useAttach.js');
require('../../hooks/useControlled.js');
require('../../_util/noop.js');
require('../../_chunks/dep-eb1b216c.js');
require('../../_chunks/dep-b529a0d7.js');
require('../../_chunks/dep-bcd16957.js');
require('../../_chunks/dep-b4442734.js');
require('../../_chunks/dep-876d090e.js');
require('../../hooks/useDefaultProps.js');
require('../../hooks/useMutationObserver.js');
require('../../hooks/useLatest.js');
require('../../_chunks/dep-4e876c4b.js');
require('../../_chunks/dep-df8d2fc8.js');
require('../../_chunks/dep-66f91e5d.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-ea361d31.js');
require('../../_chunks/dep-d3af8eb7.js');
require('../../_chunks/dep-eab994e8.js');
require('../../hooks/usePopper.js');
require('../../_chunks/dep-1d0dbfc5.js');
require('../../_chunks/dep-3c2d5e8c.js');
require('react-fast-compare');
require('@popperjs/core');
require('../../hooks/useWindowSize.js');
require('../../popup/defaultProps.js');
require('../../popup/hooks/useTrigger.js');
require('../../_util/listener.js');
require('../../_util/ref.js');
require('react-is');
require('../../_util/isFragment.js');
require('../../popup/utils/transition.js');
require('../../popup/PopupPlugin.js');
require('../../_chunks/dep-e37cb830.js');
require('../../_util/react-render.js');
require('../../common/PluginContainer.js');
require('../../config-provider/ConfigProvider.js');
require('../../_chunks/dep-4a562e04.js');
require('../../_chunks/dep-a5b4810f.js');
require('../../tooltip/defaultProps.js');
require('../../tooltip/TooltipLite.js');
require('../../hooks/useSwitch.js');
require('../../hooks/usePersistFn.js');
require('../../_chunks/dep-2a85392e.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function useEllipsis(ellipsis) {
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("typography"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var expandText = t(local.expandText);
  var collapseText = t(local.collapseText);
  var formattedEllipsis = {};
  if (ellipsis) {
    var _ellipsis$expandable, _ellipsis$collapsible;
    formattedEllipsis = ellipsis === true ? {
      row: 1,
      expandable: false,
      tooltipProps: null,
      suffix: function suffix() {
        return "";
      },
      collapsible: true
    } : {
      row: ellipsis.row || 1,
      expandable: (_ellipsis$expandable = ellipsis.expandable) !== null && _ellipsis$expandable !== void 0 ? _ellipsis$expandable : false,
      tooltipProps: ellipsis.tooltipProps || null,
      suffix: function suffix(_ref) {
        var expanded = _ref.expanded;
        return typeof (ellipsis === null || ellipsis === void 0 ? void 0 : ellipsis.suffix) === "function" ? ellipsis === null || ellipsis === void 0 ? void 0 : ellipsis.suffix({
          expanded: expanded
        }) : expanded ? collapseText : ellipsis !== null && ellipsis !== void 0 && ellipsis.expandable ? "".concat(expandText) : "...";
      },
      collapsible: (_ellipsis$collapsible = ellipsis === null || ellipsis === void 0 ? void 0 : ellipsis.collapsible) !== null && _ellipsis$collapsible !== void 0 ? _ellipsis$collapsible : false
    };
  }
  var _useState = React.useState(true),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    isClamped = _useState2[0],
    setIsClamped = _useState2[1];
  var handleExpand = function handleExpand(expanded) {
    var _ellipsis$onExpand;
    if (typeof expanded !== "boolean") return;
    setIsClamped(!expanded);
    (_ellipsis$onExpand = ellipsis.onExpand) === null || _ellipsis$onExpand === void 0 || _ellipsis$onExpand.call(ellipsis, !expanded);
  };
  var getEllipsisSuffix = function getEllipsisSuffix() {
    var _formattedEllipsis$su, _formattedEllipsis, _formattedEllipsis2;
    var moreOrLess;
    if (isFunction.isFunction(formattedEllipsis.suffix)) moreOrLess = (_formattedEllipsis$su = (_formattedEllipsis = formattedEllipsis).suffix) === null || _formattedEllipsis$su === void 0 ? void 0 : _formattedEllipsis$su.call(_formattedEllipsis, {
      expanded: !isClamped
    });else moreOrLess = formattedEllipsis.suffix;
    if ((_formattedEllipsis2 = formattedEllipsis) !== null && _formattedEllipsis2 !== void 0 && _formattedEllipsis2.tooltipProps && !!moreOrLess) {
      var _formattedEllipsis3;
      return /* @__PURE__ */React__default["default"].createElement(tooltip_index.Tooltip, _objectSpread({}, (_formattedEllipsis3 = formattedEllipsis) === null || _formattedEllipsis3 === void 0 ? void 0 : _formattedEllipsis3.tooltipProps), moreOrLess);
    }
    return moreOrLess;
  };
  var getEllipsisPrefix = function getEllipsisPrefix() {
    var _formattedEllipsis$su2, _formattedEllipsis4, _formattedEllipsis5;
    var moreOrLess;
    if (isFunction.isFunction(formattedEllipsis.suffix)) moreOrLess = (_formattedEllipsis$su2 = (_formattedEllipsis4 = formattedEllipsis).suffix) === null || _formattedEllipsis$su2 === void 0 ? void 0 : _formattedEllipsis$su2.call(_formattedEllipsis4, {
      expanded: !isClamped
    });else moreOrLess = formattedEllipsis.suffix;
    if ((_formattedEllipsis5 = formattedEllipsis) !== null && _formattedEllipsis5 !== void 0 && _formattedEllipsis5.tooltipProps && !moreOrLess) {
      var _formattedEllipsis6;
      return /* @__PURE__ */React__default["default"].createElement(tooltip_index.Tooltip, _objectSpread({}, (_formattedEllipsis6 = formattedEllipsis) === null || _formattedEllipsis6 === void 0 ? void 0 : _formattedEllipsis6.tooltipProps), "...");
    }
    return "...";
  };
  var ellipsisProps = {
    lines: formattedEllipsis.row,
    ellipsisPrefix: getEllipsisPrefix(),
    more: getEllipsisSuffix(),
    less: getEllipsisSuffix(),
    onToggleExpand: handleExpand,
    expandable: formattedEllipsis.expandable,
    collapsible: formattedEllipsis.collapsible
  };
  return {
    ellipsisProps: ellipsisProps
  };
}

exports["default"] = useEllipsis;
