// tdesign-react/cjs/tree/hooks/useControllable.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var React = require('react');
var hooks_useControlled = require('../../hooks/useControlled.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_util/noop.js');
require('../../_chunks/dep-eb1b216c.js');
require('../../_chunks/dep-b529a0d7.js');
require('../../_chunks/dep-bcd16957.js');
require('../../_chunks/dep-b4442734.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-876d090e.js');
require('../../_chunks/dep-ecfe656e.js');

function useControllable(props) {
  var _useControlled = hooks_useControlled["default"](props, "value", props.onChange),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    value = _useControlled2[0],
    onChange = _useControlled2[1];
  var _useControlled3 = hooks_useControlled["default"](props, "expanded", props.onExpand),
    _useControlled4 = slicedToArray._slicedToArray(_useControlled3, 2),
    expanded = _useControlled4[0],
    onExpand = _useControlled4[1];
  var _useControlled5 = hooks_useControlled["default"](props, "actived", props.onActive),
    _useControlled6 = slicedToArray._slicedToArray(_useControlled5, 2),
    actived = _useControlled6[0],
    onActive = _useControlled6[1];
  var _useState = React.useState([]),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    indeterminate = _useState2[0],
    setTreeIndeterminate = _useState2[1];
  return {
    value: value,
    onChange: onChange,
    expanded: expanded,
    onExpand: onExpand,
    actived: actived,
    onActive: onActive,
    setTreeIndeterminate: setTreeIndeterminate,
    indeterminate: indeterminate
  };
}

exports["default"] = useControllable;
