// tdesign-react/cjs/tree/hooks/useStore.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var React = require('react');
var treeStore = require('../../_chunks/dep-5814cc47.js');
var hooks_useUpdateLayoutEffect = require('../../hooks/useUpdateLayoutEffect.js');
var hooks_usePrevious = require('../../hooks/usePrevious.js');
var hooks_usePersistFn = require('../../hooks/usePersistFn.js');
var cloneDeep = require('../../_chunks/dep-4a562e04.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_chunks/dep-2b5ba78d.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-427fb3a1.js');
require('mitt');
require('../../_chunks/dep-78c47269.js');
require('../../_chunks/dep-e37cb830.js');
require('../../_chunks/dep-1d0dbfc5.js');
require('../../_chunks/dep-3c2d5e8c.js');
require('../../_chunks/dep-b289b0be.js');
require('../../_chunks/dep-91c4f617.js');
require('../../_chunks/dep-f7bf3d17.js');
require('../../_chunks/dep-5e17a073.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-876d090e.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-b4442734.js');
require('../../_chunks/dep-ca918872.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-13de2fef.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-63308b84.js');
require('../../_chunks/dep-048e4314.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../_chunks/dep-089e8106.js');
require('../../_chunks/dep-5c3139cb.js');
require('../../_chunks/dep-a5b4810f.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-ea361d31.js');
require('../../_chunks/dep-2e455d91.js');
require('../../_chunks/dep-894a5ece.js');
require('../../_chunks/dep-bcd16957.js');
require('../../_chunks/dep-8f62fb96.js');
require('../../_chunks/dep-f0e884b9.js');
require('../../_chunks/dep-05df1ea3.js');
require('../../_chunks/dep-526abd62.js');
require('../../_chunks/dep-4d78a412.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../_chunks/dep-966765d8.js');
require('../../_chunks/dep-4ef3269f.js');
require('../../_chunks/dep-66f91e5d.js');
require('../../_chunks/dep-2a089253.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-e44b03ec.js');
require('../../_chunks/dep-eb1b216c.js');
require('../../_chunks/dep-b529a0d7.js');
require('../../hooks/useIsFirstRender.js');
require('../../hooks/useLayoutEffect.js');
require('../../_chunks/dep-68f5fe7b.js');

function useStore(props, refresh) {
  var storeRef = React.useRef(null);
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    filterChanged = _useState2[0],
    toggleFilterChanged = _useState2[1];
  var _useState3 = React.useState(null),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    prevExpanded = _useState4[0],
    changePrevExpanded = _useState4[1];
  var data = props.data,
    keys = props.keys,
    expandAll = props.expandAll,
    expandParent = props.expandParent,
    expanded = props.expanded,
    expandLevel = props.expandLevel,
    expandMutex = props.expandMutex,
    activable = props.activable,
    activeMultiple = props.activeMultiple,
    actived = props.actived,
    disabled = props.disabled,
    disableCheck = props.disableCheck,
    draggable = props.draggable,
    checkable = props.checkable,
    value = props.value,
    checkStrictly = props.checkStrictly,
    load = props.load,
    lazy = props.lazy,
    valueMode = props.valueMode,
    filter = props.filter,
    _onLoad = props.onLoad,
    indeterminate = props.indeterminate,
    setTreeIndeterminate = props.setTreeIndeterminate,
    _props$allowFoldNodeO = props.allowFoldNodeOnFilter,
    allowFoldNodeOnFilter = _props$allowFoldNodeO === void 0 ? false : _props$allowFoldNodeO;
  var preFilter = hooks_usePrevious["default"](filter);
  React.useEffect(function () {
    if (!allowFoldNodeOnFilter) return;
    toggleFilterChanged(JSON.stringify(preFilter) !== JSON.stringify(filter));
  }, [filter, allowFoldNodeOnFilter]);
  var expandFilterPath = function expandFilterPath() {
    if (!allowFoldNodeOnFilter || !filterChanged) return;
    toggleFilterChanged(false);
    var store2 = storeRef.current;
    if (props.filter) {
      if (!prevExpanded) changePrevExpanded(store2.getExpanded());
      var pathValues = [];
      var allNodes = store2.getNodes();
      allNodes.forEach(function (node) {
        if (node.vmIsLocked) {
          pathValues.push(node.value);
        }
      });
      store2.setExpanded(pathValues);
    } else if (prevExpanded) {
      store2.replaceExpanded(prevExpanded);
      changePrevExpanded(null);
    }
  };
  var handleUpdate = hooks_usePersistFn.usePersistFn(function () {
    expandFilterPath();
    refresh();
  });
  var getExpandedArr = function getExpandedArr(arr, store2) {
    var expandedMap = /* @__PURE__ */new Map();
    arr.forEach(function (val) {
      expandedMap.set(val, true);
      if (expandParent) {
        var node = store2.getNode(val);
        node === null || node === void 0 || node.getParents().forEach(function (tn) {
          expandedMap.set(tn.value, true);
        });
      }
    });
    return Array.from(expandedMap.keys());
  };
  var createStore = function createStore() {
    var store2 = new treeStore.TreeStore({
      keys: keys,
      activable: activable,
      activeMultiple: activeMultiple,
      checkable: checkable,
      checkStrictly: checkStrictly,
      expandAll: expandAll,
      expandLevel: expandLevel,
      expandMutex: expandMutex,
      expandParent: expandParent,
      disabled: disabled,
      disableCheck: disableCheck,
      draggable: draggable,
      load: load,
      lazy: lazy,
      valueMode: valueMode,
      filter: filter,
      onLoad: function onLoad(info) {
        var node = info.node;
        _onLoad === null || _onLoad === void 0 || _onLoad({
          node: node.getModel()
        });
      },
      onUpdate: handleUpdate,
      allowFoldNodeOnFilter: allowFoldNodeOnFilter
    });
    var list = cloneDeep.cloneDeep(data);
    if (!Array.isArray(list)) {
      list = [];
    }
    store2.append(list);
    store2.refreshNodes();
    if (Array.isArray(value)) {
      store2.setChecked(value);
    }
    if (Array.isArray(expanded)) {
      var expandedArr = getExpandedArr(expanded, store2);
      store2.setExpanded(expandedArr);
    }
    if (Array.isArray(actived)) {
      store2.setActived(actived);
    }
    store2.refreshNodes();
    return store2;
  };
  if (!storeRef.current) {
    storeRef.current = createStore();
  }
  var store = storeRef.current;
  hooks_useUpdateLayoutEffect["default"](function () {
    if (data && Array.isArray(data)) {
      var expanded2 = store.getExpanded();
      var checked = store.getChecked();
      var actived2 = store.getActived();
      store.removeAll();
      store.append(data);
      store.setChecked(checked);
      store.setActived(actived2);
      store.setExpanded(expanded2);
    }
  }, [data, store]);
  hooks_useUpdateLayoutEffect["default"](function () {
    store.setConfig({
      keys: keys,
      expandAll: expandAll,
      expandLevel: expandLevel,
      expandMutex: expandMutex,
      expandParent: expandParent,
      activable: activable,
      activeMultiple: activeMultiple,
      disabled: disabled,
      disableCheck: disableCheck,
      checkable: checkable,
      draggable: draggable,
      checkStrictly: checkStrictly,
      load: load,
      lazy: lazy,
      valueMode: valueMode
    });
    store.refreshState();
  }, [activable, activeMultiple, checkStrictly, draggable, checkable, disabled, disableCheck, expandAll, expandLevel, expandMutex, expandParent, keys, lazy, load, store, valueMode]);
  hooks_useUpdateLayoutEffect["default"](function () {
    if (expandAll) {
      var valueList = store.getNodes().filter(function (node) {
        return Array.isArray(node.children) && node.children.length;
      }).map(function (node) {
        return node.value;
      });
      store.setExpanded(valueList);
    } else {
      store.replaceExpanded(prevExpanded);
      changePrevExpanded(null);
    }
  }, [store, expandAll]);
  hooks_useUpdateLayoutEffect["default"](function () {
    if (Array.isArray(value)) {
      store.replaceChecked(value);
      var checkedValue = store.getCheckedNodes().map(function (v) {
        return v.data[(keys === null || keys === void 0 ? void 0 : keys.value) || "value"];
      });
      var indeterminateConflict = checkedValue.filter(function (v) {
        return indeterminate.includes(v);
      });
      if (indeterminateConflict.length) {
        setTreeIndeterminate(indeterminate.filter(function (v) {
          return !indeterminateConflict.includes(v);
        }));
      }
    }
  }, [store, value, data]);
  hooks_useUpdateLayoutEffect["default"](function () {
    if (Array.isArray(expanded)) {
      var expandedArr = getExpandedArr(expanded, store);
      store.replaceExpanded(expandedArr);
    }
  }, [expanded, store]);
  hooks_useUpdateLayoutEffect["default"](function () {
    if (Array.isArray(actived)) {
      store.replaceActived(actived);
    }
  }, [actived, store]);
  hooks_useUpdateLayoutEffect["default"](function () {
    if (Array.isArray(indeterminate)) {
      store.replaceIndeterminate(indeterminate);
    }
  }, [indeterminate, store, data]);
  hooks_useUpdateLayoutEffect["default"](function () {
    store.setConfig({
      filter: filter
    });
    store.updateAll();
  }, [filter, store]);
  return storeRef.current;
}

exports.useStore = useStore;
