// tdesign-react/cjs/card/Card.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var loading_index = require('../loading/index.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useCommonClassName = require('../hooks/useCommonClassName.js');
var card_defaultProps = require('./defaultProps.js');
var hooks_useDefaultProps = require('../hooks/useDefaultProps.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../loading/Loading.js');
require('../_chunks/dep-1a3dddaf.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-68f5fe7b.js');
require('../_chunks/dep-17e2e72f.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-ecfe656e.js');
require('../common/Portal.js');
require('react-dom');
require('../hooks/useLayoutEffect.js');
require('../loading/defaultProps.js');
require('../loading/gradient.js');
require('../_chunks/dep-064e4dba.js');
require('../_chunks/dep-b011b5d3.js');
require('../_chunks/dep-1d0dbfc5.js');
require('../_chunks/dep-3c2d5e8c.js');
require('../_chunks/dep-1c586f5c.js');
require('../_chunks/dep-f0e884b9.js');
require('../hooks/useDomRefCallback.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-14013769.js');
require('../_chunks/dep-06bf4625.js');
require('dayjs');
require('../_chunks/dep-043eee20.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-b21b548e.js');
require('../_chunks/dep-69c18aee.js');
require('../_chunks/dep-83fd21a7.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-5f500f41.js');
require('../_chunks/dep-7577d7c1.js');
require('../_chunks/dep-2d5cf7d6.js');
require('../loading/plugin.js');
require('../_util/react-render.js');
require('../_chunks/dep-e37cb830.js');
require('../common/PluginContainer.js');
require('../config-provider/ConfigProvider.js');
require('../_chunks/dep-4a562e04.js');
require('../_chunks/dep-a5b4810f.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Card = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _props$children;
  var _useDefaultProps = hooks_useDefaultProps["default"](props, card_defaultProps.cardDefaultProps),
    actions = _useDefaultProps.actions,
    avatar = _useDefaultProps.avatar,
    bordered = _useDefaultProps.bordered,
    className = _useDefaultProps.className,
    cover = _useDefaultProps.cover,
    description = _useDefaultProps.description,
    footer = _useDefaultProps.footer,
    header = _useDefaultProps.header,
    headerBordered = _useDefaultProps.headerBordered,
    hoverShadow = _useDefaultProps.hoverShadow,
    loading = _useDefaultProps.loading,
    shadow = _useDefaultProps.shadow,
    size = _useDefaultProps.size,
    style = _useDefaultProps.style,
    subtitle = _useDefaultProps.subtitle,
    title = _useDefaultProps.title,
    theme = _useDefaultProps.theme,
    status = _useDefaultProps.status,
    loadingProps = _useDefaultProps.loadingProps,
    bodyClassName = _useDefaultProps.bodyClassName,
    bodyStyle = _useDefaultProps.bodyStyle,
    footerClassName = _useDefaultProps.footerClassName,
    footerStyle = _useDefaultProps.footerStyle,
    headerClassName = _useDefaultProps.headerClassName,
    headerStyle = _useDefaultProps.headerStyle;
  var children = (_props$children = props.children) !== null && _props$children !== void 0 ? _props$children : props.content;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var commonClassNames = hooks_useCommonClassName["default"]();
  var isPoster2 = theme === "poster2";
  var cardClass = classNames__default["default"]("".concat(classPrefix, "-card"), className, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty({}, commonClassNames.SIZE.small, size === "small"), "".concat(classPrefix, "-card--bordered"), bordered), "".concat(classPrefix, "-card--shadow"), shadow), "".concat(classPrefix, "-card--shadow-hover"), hoverShadow));
  var showHeader = header || title || subtitle || description || avatar || actions && !isPoster2 || status && isPoster2;
  var headerClass = classNames__default["default"](headerClassName, defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-card__header"), showHeader), "".concat(classPrefix, "-card__title--bordered"), headerBordered));
  var titleClass = classNames__default["default"](defineProperty._defineProperty({}, "".concat(classPrefix, "-card__title"), title));
  var subtitleClass = classNames__default["default"](defineProperty._defineProperty({}, "".concat(classPrefix, "-card__subtitle"), subtitle));
  var actionClass = classNames__default["default"](defineProperty._defineProperty({}, "".concat(classPrefix, "-card__actions"), actions));
  var footerClass = classNames__default["default"](footerClassName, defineProperty._defineProperty({}, "".concat(classPrefix, "-card__footer"), footer));
  var coverClass = classNames__default["default"](defineProperty._defineProperty({}, "".concat(classPrefix, "-card__cover"), cover));
  var avatarClass = classNames__default["default"](defineProperty._defineProperty({}, "".concat(classPrefix, "-card__avatar"), avatar));
  var bodyClass = classNames__default["default"](bodyClassName, defineProperty._defineProperty({}, "".concat(classPrefix, "-card__body"), children));
  var descriptionClass = classNames__default["default"](defineProperty._defineProperty({}, "".concat(classPrefix, "-card__description"), description));
  var renderTitle = title ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: titleClass
  }, title) : null;
  var renderSubtitle = subtitle ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: subtitleClass
  }, subtitle) : null;
  var renderDescription = description ? /* @__PURE__ */React__default["default"].createElement("p", {
    className: descriptionClass
  }, description) : null;
  var renderAvatar = avatar && /* @__PURE__ */React__default["default"].createElement("div", {
    className: avatarClass
  }, avatar);
  var renderHeaderActions = actions && !isPoster2 && /* @__PURE__ */React__default["default"].createElement("div", {
    className: actionClass
  }, actions);
  var renderFooterActions = actions && isPoster2 && /* @__PURE__ */React__default["default"].createElement("div", {
    className: actionClass
  }, actions);
  var renderStatus = status && isPoster2 && /* @__PURE__ */React__default["default"].createElement("div", {
    className: actionClass
  }, status);
  var renderHeader = function renderHeader() {
    if (header) {
      return /* @__PURE__ */React__default["default"].createElement("div", {
        className: headerClass,
        style: headerStyle
      }, header);
    }
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: headerClass,
      style: headerStyle
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-card__header-wrapper")
    }, renderAvatar, /* @__PURE__ */React__default["default"].createElement("div", null, renderTitle, renderSubtitle, renderDescription)), renderHeaderActions, renderStatus);
  };
  var renderCover = cover ? /* @__PURE__ */React__default["default"].createElement("div", {
    className: coverClass
  }, typeof cover === "string" ? /* @__PURE__ */React__default["default"].createElement("img", {
    src: cover,
    alt: ""
  }) : cover) : null;
  var renderChildren = children && /* @__PURE__ */React__default["default"].createElement("div", {
    className: bodyClass,
    style: bodyStyle
  }, children);
  var renderFooter = footer && /* @__PURE__ */React__default["default"].createElement("div", {
    className: footerClass,
    style: footerStyle
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-card__footer-wrapper")
  }, footer), renderFooterActions);
  var card = /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, showHeader ? renderHeader() : null, renderCover, renderChildren, renderFooter);
  var childrenNode = null;
  if (!Reflect.has(props, "loading")) {
    childrenNode = card;
  } else if (/*#__PURE__*/React__default["default"].isValidElement(loading)) {
    childrenNode = /*#__PURE__*/React__default["default"].cloneElement(loading, null, card);
  } else {
    childrenNode = /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, _objectSpread(_objectSpread({}, loadingProps), {}, {
      loading: !!loading
    }), card);
  }
  return /* @__PURE__ */React__default["default"].createElement("div", {
    ref: ref,
    className: cardClass,
    style: style
  }, childrenNode);
});
Card.displayName = "Card";

exports["default"] = Card;
