// tdesign-react/cjs/upload/themes/ImageCard.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var toConsumableArray = require('../../_chunks/dep-1d0dbfc5.js');
var defineProperty = require('../../_chunks/dep-2b5ba78d.js');
var objectWithoutProperties = require('../../_chunks/dep-1c586f5c.js');
var React = require('react');
var tdesignIconsReact = require('tdesign-icons-react');
var classNames = require('classnames');
var utils = require('../../_chunks/dep-442716ca.js');
var loading_index = require('../../loading/index.js');
var hooks_useGlobalIcon = require('../../hooks/useGlobalIcon.js');
var imageViewer_index = require('../../image-viewer/index.js');
var _util_parseTNode = require('../../_util/parseTNode.js');
var link_index = require('../../link/index.js');
var image_index = require('../../image/index.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_chunks/dep-3c2d5e8c.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-b289b0be.js');
require('../../loading/Loading.js');
require('../../_chunks/dep-1a3dddaf.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-68f5fe7b.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../common/Portal.js');
require('react-dom');
require('../../hooks/useConfig.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-14013769.js');
require('../../_chunks/dep-06bf4625.js');
require('dayjs');
require('../../_chunks/dep-043eee20.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../_chunks/dep-2d5cf7d6.js');
require('../../hooks/useLayoutEffect.js');
require('../../hooks/useDefaultProps.js');
require('../../loading/defaultProps.js');
require('../../loading/gradient.js');
require('../../_chunks/dep-064e4dba.js');
require('../../_chunks/dep-b011b5d3.js');
require('../../_chunks/dep-f0e884b9.js');
require('../../hooks/useDomRefCallback.js');
require('../../loading/plugin.js');
require('../../_util/react-render.js');
require('../../_chunks/dep-e37cb830.js');
require('../../common/PluginContainer.js');
require('../../config-provider/ConfigProvider.js');
require('../../_chunks/dep-4a562e04.js');
require('../../_chunks/dep-a5b4810f.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-ea361d31.js');
require('../../image-viewer/ImageViewer.js');
require('../../_chunks/dep-995b39ff.js');
require('../../_chunks/dep-04deac0d.js');
require('../../hooks/useImagePreviewUrl.js');
require('../../locale/LocalReceiver.js');
require('../../tooltip/index.js');
require('../../tooltip/Tooltip.js');
require('../../popup/index.js');
require('../../popup/Popup.js');
require('react-transition-group');
require('../../hooks/useAnimation.js');
require('../../hooks/useAttach.js');
require('../../hooks/useControlled.js');
require('../../_util/noop.js');
require('../../_chunks/dep-eb1b216c.js');
require('../../_chunks/dep-b529a0d7.js');
require('../../_chunks/dep-bcd16957.js');
require('../../_chunks/dep-b4442734.js');
require('../../_chunks/dep-876d090e.js');
require('../../hooks/useMutationObserver.js');
require('../../hooks/useLatest.js');
require('../../_chunks/dep-4e876c4b.js');
require('../../_chunks/dep-df8d2fc8.js');
require('../../_chunks/dep-66f91e5d.js');
require('../../_chunks/dep-d3af8eb7.js');
require('../../_chunks/dep-eab994e8.js');
require('../../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../../hooks/useWindowSize.js');
require('../../popup/defaultProps.js');
require('../../popup/hooks/useTrigger.js');
require('../../_util/listener.js');
require('../../_util/ref.js');
require('react-is');
require('../../_util/isFragment.js');
require('../../popup/utils/transition.js');
require('../../popup/PopupPlugin.js');
require('../../tooltip/defaultProps.js');
require('../../tooltip/TooltipLite.js');
require('../../hooks/useSwitch.js');
require('../../hooks/usePersistFn.js');
require('../../_chunks/dep-2a85392e.js');
require('../../dialog/index.js');
require('../../dialog/Dialog.js');
require('../../hooks/useDeepEffect.js');
require('../../hooks/useSetState.js');
require('../../dialog/defaultProps.js');
require('../../dialog/DialogCard.js');
require('../../button/index.js');
require('../../button/Button.js');
require('../../hooks/useRipple.js');
require('../../button/defaultProps.js');
require('../../dialog/hooks/useDialogDrag.js');
require('../../hooks/useMouseEvent.js');
require('../../dialog/hooks/useDialogEsc.js');
require('../../dialog/hooks/useDialogPosition.js');
require('../../dialog/hooks/useLockStyle.js');
require('../../_chunks/dep-fb673f7a.js');
require('../../_chunks/dep-8f62fb96.js');
require('../../dialog/plugin.js');
require('../../image-viewer/hooks/useIconMap.js');
require('../../image-viewer/hooks/useIndex.js');
require('../../image-viewer/hooks/useMirror.js');
require('../../image-viewer/hooks/usePosition.js');
require('../../image-viewer/hooks/useRotate.js');
require('../../image-viewer/hooks/useScale.js');
require('../../image-viewer/defaultProps.js');
require('../../image-viewer/hooks/useImageScale.js');
require('../../image-viewer/hooks/useViewerScale.js');
require('../../image-viewer/defaultTrigger.js');
require('../../image/Image.js');
require('../../_chunks/dep-e31ce0e8.js');
require('../../image/defaultProps.js');
require('../../space/index.js');
require('../../space/Space.js');
require('../../space/defaultProps.js');
require('../../link/Link.js');
require('../../link/defaultProps.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var _excluded = ["className"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var ImageCard = function ImageCard(props) {
  var _locale$triggerUpload;
  var displayFiles = props.displayFiles,
    locale = props.locale,
    classPrefix = props.classPrefix,
    multiple = props.multiple,
    _props$max = props.max,
    max = _props$max === void 0 ? 0 : _props$max,
    onRemove = props.onRemove,
    disabled = props.disabled,
    fileListDisplay = props.fileListDisplay,
    _props$imageProps = props.imageProps,
    imageProps = _props$imageProps === void 0 ? {} : _props$imageProps;
  var _useGlobalIcon = hooks_useGlobalIcon["default"]({
      AddIcon: tdesignIconsReact.AddIcon,
      BrowseIcon: tdesignIconsReact.BrowseIcon,
      DeleteIcon: tdesignIconsReact.DeleteIcon,
      ErrorCircleFilledIcon: tdesignIconsReact.ErrorCircleFilledIcon
    }),
    BrowseIcon = _useGlobalIcon.BrowseIcon,
    DeleteIcon = _useGlobalIcon.DeleteIcon,
    AddIcon = _useGlobalIcon.AddIcon,
    ErrorCircleFilledIcon = _useGlobalIcon.ErrorCircleFilledIcon;
  var imageClassName = imageProps.className,
    restImageProps = objectWithoutProperties._objectWithoutProperties(imageProps, _excluded);
  var showTrigger = React__default["default"].useMemo(function () {
    if (multiple) {
      return !max || displayFiles.length < max;
    }
    return !(displayFiles !== null && displayFiles !== void 0 && displayFiles[0]);
  }, [displayFiles, max, multiple]);
  var renderMainContent = function renderMainContent(file, index) {
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-upload__card-content ").concat(classPrefix, "-upload__card-box")
    }, /* @__PURE__ */React__default["default"].createElement(image_index.Image, _objectSpread(_objectSpread({
      fit: "contain",
      className: classNames__default["default"]("".concat(classPrefix, "-upload__card-image"), imageClassName)
    }, restImageProps), {}, {
      src: file.url || file.raw
    })), /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-upload__card-mask")
    }, /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(classPrefix, "-upload__card-mask-item"),
      onClick: function onClick(e) {
        return e.stopPropagation();
      }
    }, /* @__PURE__ */React__default["default"].createElement(imageViewer_index.ImageViewer, _objectSpread({
      trigger: function trigger(_ref) {
        var open = _ref.open;
        return /* @__PURE__ */React__default["default"].createElement(BrowseIcon, {
          onClick: function onClick(e) {
            var _props$onPreview;
            (_props$onPreview = props.onPreview) === null || _props$onPreview === void 0 || _props$onPreview.call(props, {
              file: file,
              index: index,
              e: e
            });
            open();
          }
        });
      },
      images: displayFiles.map(function (t) {
        return t.url || t.raw;
      }),
      defaultIndex: index
    }, props.imageViewerProps))), !disabled && /* @__PURE__ */React__default["default"].createElement(React__default["default"].Fragment, null, /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(classPrefix, "-upload__card-mask-item-divider")
    }), /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(classPrefix, "-upload__card-mask-item"),
      onClick: function onClick(e) {
        return e.stopPropagation();
      }
    }, /* @__PURE__ */React__default["default"].createElement(DeleteIcon, {
      onClick: function onClick(e) {
        return onRemove === null || onRemove === void 0 ? void 0 : onRemove({
          e: e,
          file: file,
          index: index
        });
      }
    })))));
  };
  var renderProgressFile = function renderProgressFile(file, loadCard) {
    var _locale$progress;
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: classNames__default["default"]([loadCard, "".concat(classPrefix, "-upload__").concat(props.theme, "-").concat(file.status)])
    }, /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, {
      loading: true,
      size: "medium"
    }), /* @__PURE__ */React__default["default"].createElement("p", null, locale === null || locale === void 0 || (_locale$progress = locale.progress) === null || _locale$progress === void 0 ? void 0 : _locale$progress.uploadingText, props.showUploadProgress ? " ".concat(file.percent, "%") : ""));
  };
  var renderFailFile = function renderFailFile(file, index, loadCard) {
    var _file$response, _locale$progress2;
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: loadCard
    }, /* @__PURE__ */React__default["default"].createElement(ErrorCircleFilledIcon, null), /* @__PURE__ */React__default["default"].createElement("p", null, ((_file$response = file.response) === null || _file$response === void 0 ? void 0 : _file$response.error) || (locale === null || locale === void 0 || (_locale$progress2 = locale.progress) === null || _locale$progress2 === void 0 ? void 0 : _locale$progress2.failText)), /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-upload__card-mask")
    }, /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(classPrefix, "-upload__card-mask-item"),
      onClick: function onClick(e) {
        return e.stopPropagation();
      }
    }, /* @__PURE__ */React__default["default"].createElement(DeleteIcon, {
      onClick: function onClick(e) {
        var _props$onRemove;
        return (_props$onRemove = props.onRemove) === null || _props$onRemove === void 0 ? void 0 : _props$onRemove.call(props, {
          e: e,
          file: file,
          index: index
        });
      }
    }))));
  };
  var cardItemClasses = "".concat(classPrefix, "-upload__card-item ").concat(classPrefix, "-is-background");
  if (fileListDisplay) {
    return /* @__PURE__ */React__default["default"].createElement("div", null, _util_parseTNode["default"](fileListDisplay, {
      triggerUpload: props.triggerUpload,
      uploadFiles: props.uploadFiles,
      cancelUpload: props.cancelUpload,
      onPreview: props.onPreview,
      onRemove: props.onRemove,
      toUploadFiles: props.toUploadFiles,
      sizeOverLimitMessage: props.sizeOverLimitMessage,
      locale: props.locale,
      files: displayFiles
    }));
  }
  return /* @__PURE__ */React__default["default"].createElement("div", null, /* @__PURE__ */React__default["default"].createElement("ul", {
    className: "".concat(classPrefix, "-upload__card")
  }, displayFiles === null || displayFiles === void 0 ? void 0 : displayFiles.map(function (file, index) {
    var loadCard = "".concat(classPrefix, "-upload__card-container ").concat(classPrefix, "-upload__card-box");
    var fileName = props.abridgeName ? utils.abridgeName.apply(void 0, [file.name].concat(toConsumableArray._toConsumableArray(props.abridgeName))) : file.name;
    var fileNameClassName = "".concat(classPrefix, "-upload__card-name");
    return /* @__PURE__ */React__default["default"].createElement("li", {
      className: cardItemClasses,
      key: index
    }, file.status === "progress" && renderProgressFile(file, loadCard), file.status === "fail" && renderFailFile(file, index, loadCard), !["progress", "fail"].includes(file.status) && renderMainContent(file, index), fileName && props.showImageFileName && (file.url ? /* @__PURE__ */React__default["default"].createElement(link_index.Link, {
      href: file.url,
      className: fileNameClassName,
      target: "_blank",
      hover: "color",
      size: "small"
    }, fileName) : /* @__PURE__ */React__default["default"].createElement("span", {
      className: fileNameClassName
    }, fileName)));
  }), showTrigger && /* @__PURE__ */React__default["default"].createElement("li", {
    className: cardItemClasses,
    onClick: props.triggerUpload
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: classNames__default["default"](["".concat(classPrefix, "-upload__image-add"), "".concat(classPrefix, "-upload__card-container"), "".concat(classPrefix, "-upload__card-box"), defineProperty._defineProperty({}, "".concat(classPrefix, "-is-disabled"), props.disabled)])
  }, /* @__PURE__ */React__default["default"].createElement(AddIcon, null), /* @__PURE__ */React__default["default"].createElement("p", {
    className: classNames__default["default"](["".concat(classPrefix, "-size-s"), "".concat(classPrefix, "-upload__add-text")])
  }, locale === null || locale === void 0 || (_locale$triggerUpload = locale.triggerUploadText) === null || _locale$triggerUpload === void 0 ? void 0 : _locale$triggerUpload.image)))));
};
ImageCard.displayName = "ImageCard";

exports["default"] = ImageCard;
