// tdesign-react/lib/affix/Affix.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-94a7af64.js';
import React, { forwardRef, useState, useRef, useCallback, useImperativeHandle, useEffect } from 'react';
import { isFunction } from 'lodash-es';
import { i as isWindow } from '../_chunks/dep-a347b1b5.js';
import { getScrollContainer } from '../_util/scroll.js';
import useConfig from '../hooks/useConfig.js';
import useDefaultProps from '../hooks/useDefaultProps.js';
import { affixDefaultProps } from './defaultProps.js';
import '../_chunks/dep-0054b574.js';
import '../_chunks/dep-8e8ac0d4.js';
import 'raf';
import '../_util/easing.js';
import '../config-provider/ConfigContext.js';
import '../_chunks/dep-5e72c8b9.js';
import '../_chunks/dep-43f7da7b.js';
import '../_chunks/dep-923c86e8.js';
import '../_chunks/dep-03cf63f2.js';
import 'dayjs';

var Affix = /*#__PURE__*/forwardRef(function (props, ref) {
  var _useDefaultProps = useDefaultProps(props, affixDefaultProps),
    children = _useDefaultProps.children,
    content = _useDefaultProps.content,
    zIndex = _useDefaultProps.zIndex,
    container = _useDefaultProps.container,
    offsetBottom = _useDefaultProps.offsetBottom,
    offsetTop = _useDefaultProps.offsetTop,
    className = _useDefaultProps.className,
    style = _useDefaultProps.style,
    onFixedChange = _useDefaultProps.onFixedChange;
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix;
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    containerReady = _useState2[0],
    setContainerReady = _useState2[1];
  var affixRef = useRef(null);
  var affixWrapRef = useRef(null);
  var placeholderEL = useRef(null);
  var scrollContainer = useRef(null);
  var ticking = useRef(false);
  var handleScroll = useCallback(function () {
    if (!ticking.current) {
      requestAnimationFrame(function () {
        var _affixWrapRef$current, _affixWrapRef$current2;
        var _ref = (_affixWrapRef$current = (_affixWrapRef$current2 = affixWrapRef.current) === null || _affixWrapRef$current2 === void 0 ? void 0 : _affixWrapRef$current2.getBoundingClientRect()) !== null && _affixWrapRef$current !== void 0 ? _affixWrapRef$current : {},
          _ref$top = _ref.top,
          wrapToTop = _ref$top === void 0 ? 0 : _ref$top,
          _ref$bottom = _ref.bottom,
          wrapToBottom = _ref$bottom === void 0 ? 0 : _ref$bottom,
          _ref$width = _ref.width,
          wrapWidth = _ref$width === void 0 ? 0 : _ref$width,
          _ref$height = _ref.height,
          wrapHeight = _ref$height === void 0 ? 0 : _ref$height;
        var containerToTop = 0;
        var containerToBottom = 0;
        if (isWindow(scrollContainer.current)) {
          containerToBottom = scrollContainer.current.innerHeight;
        } else if (scrollContainer.current instanceof HTMLElement) {
          var rect = scrollContainer.current.getBoundingClientRect();
          containerToTop = rect.top;
          containerToBottom = rect.bottom;
        }
        var calcTop = wrapToTop - containerToTop;
        var fixedTop;
        if (props.offsetBottom !== void 0 && props.offsetTop === void 0) {
          var bottomThreshold = containerToBottom - (offsetBottom !== null && offsetBottom !== void 0 ? offsetBottom : 0);
          if (containerToBottom > 0 && wrapToBottom >= bottomThreshold) {
            fixedTop = bottomThreshold - wrapHeight;
          } else {
            fixedTop = false;
          }
        } else {
          var _scrollContainer$curr;
          var containerHeight = ((_scrollContainer$curr = scrollContainer.current) === null || _scrollContainer$curr === void 0 ? void 0 : _scrollContainer$curr[isWindow(scrollContainer.current) ? "innerHeight" : "clientHeight"]) - wrapHeight;
          var calcBottom = containerToTop + containerHeight - (offsetBottom !== null && offsetBottom !== void 0 ? offsetBottom : 0);
          if (calcTop <= offsetTop) {
            fixedTop = containerToTop + offsetTop;
          } else if (wrapToTop >= calcBottom) {
            fixedTop = calcBottom;
          } else {
            fixedTop = false;
          }
        }
        if (affixRef.current) {
          var affixed = fixedTop !== false;
          var placeholderStatus = affixWrapRef.current.contains(placeholderEL.current);
          var prePlaceholderStatus = placeholderStatus;
          if (affixed) {
            affixRef.current.className = "".concat(classPrefix, "-affix");
            affixRef.current.style.top = "".concat(fixedTop, "px");
            affixRef.current.style.width = "".concat(wrapWidth, "px");
            affixRef.current.style.height = "".concat(wrapHeight, "px");
            if (zIndex) {
              affixRef.current.style.zIndex = "".concat(zIndex);
            }
            if (!placeholderStatus) {
              placeholderEL.current.style.width = "".concat(wrapWidth, "px");
              placeholderEL.current.style.height = "".concat(wrapHeight, "px");
              affixWrapRef.current.appendChild(placeholderEL.current);
              placeholderStatus = true;
            }
          } else {
            affixRef.current.removeAttribute("class");
            affixRef.current.removeAttribute("style");
            if (placeholderStatus) {
              placeholderEL.current.remove();
              placeholderStatus = false;
            }
          }
          if (prePlaceholderStatus !== placeholderStatus && isFunction(onFixedChange)) {
            onFixedChange(affixed, {
              top: +fixedTop
            });
          }
        }
        ticking.current = false;
      });
    }
    ticking.current = true;
  }, [classPrefix, offsetBottom, offsetTop, zIndex, onFixedChange, props.offsetBottom, props.offsetTop]);
  useImperativeHandle(ref, function () {
    return {
      handleScroll: handleScroll
    };
  });
  useEffect(function () {
    placeholderEL.current = document.createElement("div");
  }, []);
  useEffect(function () {
    var checkContainerExist = function checkContainerExist() {
      var el = getScrollContainer(container);
      var isReady = isWindow(el) || el instanceof HTMLElement;
      setContainerReady(isReady);
      return isReady;
    };
    if (checkContainerExist()) return;
    var observer = new MutationObserver(function () {
      if (checkContainerExist()) {
        observer.disconnect();
      }
    });
    observer.observe(document.body, {
      childList: true,
      subtree: true
    });
    return function () {
      observer.disconnect();
    };
  }, [container]);
  useEffect(function () {
    if (!containerReady) return;
    var newContainer = getScrollContainer(container);
    if (!newContainer) return;
    if (scrollContainer.current) {
      scrollContainer.current.removeEventListener("scroll", handleScroll);
    }
    scrollContainer.current = newContainer;
    handleScroll();
    scrollContainer.current.addEventListener("scroll", handleScroll);
    window.addEventListener("resize", handleScroll);
    var isContainerNotWindow = !isWindow(scrollContainer.current);
    if (isContainerNotWindow) {
      window.addEventListener("scroll", handleScroll);
    }
    return function () {
      if (scrollContainer.current) {
        scrollContainer.current.removeEventListener("scroll", handleScroll);
      }
      window.removeEventListener("resize", handleScroll);
      if (isContainerNotWindow) {
        window.removeEventListener("scroll", handleScroll);
      }
    };
  }, [container, containerReady, handleScroll]);
  return /* @__PURE__ */React.createElement("div", {
    ref: affixWrapRef,
    className: className,
    style: style
  }, /* @__PURE__ */React.createElement("div", {
    ref: affixRef
  }, children || content));
});
Affix.displayName = "Affix";

export { Affix as default };
