// tdesign-react/cjs/date-picker/base/Cell.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-2b5ba78d.js');
var React = require('react');
var classNames = require('classnames');
var utils = require('../../_chunks/dep-8227db70.js');
var hooks_useConfig = require('../../hooks/useConfig.js');
var _util_parseTNode = require('../../_util/parseTNode.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-1a3dddaf.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('dayjs');
require('../../_chunks/dep-06bf4625.js');
require('../../_chunks/dep-c990dbb4.js');
require('../../_chunks/dep-eab994e8.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-876d090e.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-bf18bb11.js');
require('../../_chunks/dep-06ac19a1.js');
require('../../_chunks/dep-b289b0be.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-bcd16957.js');
require('../../_chunks/dep-2d5cf7d6.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-6abdd014.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-14013769.js');
require('../../_chunks/dep-043eee20.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-7577d7c1.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var DatePickerCell = function DatePickerCell(props) {
  var _classNames, _parseTNode;
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var cellClass = classNames__default["default"]("".concat(classPrefix, "-date-picker__cell"), (_classNames = {}, defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-date-picker__cell--now"), props.now), "".concat(classPrefix, "-date-picker__cell--active"), props.active), "".concat(classPrefix, "-date-picker__cell--disabled"), props.disabled), "".concat(classPrefix, "-date-picker__cell--highlight"), props.highlight), "".concat(classPrefix, "-date-picker__cell--hover-highlight"), props.hoverHighlight), "".concat(classPrefix, "-date-picker__cell--active-start"), props.startOfRange), "".concat(classPrefix, "-date-picker__cell--active-end"), props.endOfRange), "".concat(classPrefix, "-date-picker__cell--hover-start"), props.hoverStartOfRange), "".concat(classPrefix, "-date-picker__cell--hover-end"), props.hoverEndOfRange), "".concat(classPrefix, "-date-picker__cell--additional"), props.additional), defineProperty._defineProperty(defineProperty._defineProperty(defineProperty._defineProperty(_classNames, "".concat(classPrefix, "-date-picker__cell--first-day-of-month"), props.firstDayOfMonth), "".concat(classPrefix, "-date-picker__cell--last-day-of-month"), props.lastDayOfMonth), "".concat(classPrefix, "-date-picker__cell--week-of-year"), props.weekOfYear)));
  function handleClick(e) {
    var _props$onClick;
    if (props.disabled) return;
    if (props.time) {
      var _extractTimeObj = utils.extractTimeObj(props.time),
        hours = _extractTimeObj.hours,
        minutes = _extractTimeObj.minutes,
        seconds = _extractTimeObj.seconds,
        milliseconds = _extractTimeObj.milliseconds,
        meridiem = _extractTimeObj.meridiem;
      var nextHours = hours;
      if (/am/i.test(meridiem) && nextHours === 12) nextHours -= 12;
      if (/pm/i.test(meridiem) && nextHours < 12) nextHours += 12;
      props.value.setHours(nextHours);
      props.value.setMinutes(minutes);
      props.value.setSeconds(seconds);
      props.value.setMilliseconds(milliseconds);
    }
    (_props$onClick = props.onClick) === null || _props$onClick === void 0 || _props$onClick.call(props, props.value, {
      e: e
    });
  }
  function handleMouseEnter() {
    var _props$onMouseEnter;
    if (props.disabled) return;
    if (props.time) {
      var _extractTimeObj2 = utils.extractTimeObj(props.time),
        hours = _extractTimeObj2.hours,
        minutes = _extractTimeObj2.minutes,
        seconds = _extractTimeObj2.seconds,
        milliseconds = _extractTimeObj2.milliseconds,
        meridiem = _extractTimeObj2.meridiem;
      var nextHours = hours;
      if (/am/i.test(meridiem) && nextHours === 12) nextHours -= 12;
      if (/pm/i.test(meridiem) && nextHours < 12) nextHours += 12;
      props.value.setHours(nextHours);
      props.value.setMinutes(minutes);
      props.value.setSeconds(seconds);
      props.value.setMilliseconds(milliseconds);
    }
    (_props$onMouseEnter = props.onMouseEnter) === null || _props$onMouseEnter === void 0 || _props$onMouseEnter.call(props, props.value);
  }
  return /* @__PURE__ */React__default["default"].createElement("td", {
    className: cellClass,
    onClick: handleClick,
    onMouseEnter: handleMouseEnter
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-date-picker__cell-inner")
  }, (_parseTNode = _util_parseTNode["default"](props.cell, {
    value: props.value
  })) !== null && _parseTNode !== void 0 ? _parseTNode : props.text));
};
DatePickerCell.displayName = "DatePickerCell";

exports["default"] = DatePickerCell;
