// tdesign-react/cjs/date-picker/base/Footer.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var locale_LocalReceiver = require('../../locale/LocalReceiver.js');
var button_index = require('../../button/index.js');
var hooks_useConfig = require('../../hooks/useConfig.js');
var isFunction = require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../config-provider/ConfigProvider.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-14013769.js');
require('../../_chunks/dep-06bf4625.js');
require('dayjs');
require('../../_chunks/dep-043eee20.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../_chunks/dep-2d5cf7d6.js');
require('../../_chunks/dep-4a562e04.js');
require('../../_chunks/dep-a5b4810f.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-ea361d31.js');
require('../../button/Button.js');
require('../../_chunks/dep-1c586f5c.js');
require('../../hooks/useDomRefCallback.js');
require('../../hooks/useRipple.js');
require('../../_chunks/dep-064e4dba.js');
require('../../hooks/useAnimation.js');
require('../../_chunks/dep-68f5fe7b.js');
require('../../loading/index.js');
require('../../loading/Loading.js');
require('../../common/Portal.js');
require('react-dom');
require('../../hooks/useLayoutEffect.js');
require('../../hooks/useDefaultProps.js');
require('../../loading/defaultProps.js');
require('../../loading/gradient.js');
require('../../_chunks/dep-b011b5d3.js');
require('../../_chunks/dep-1d0dbfc5.js');
require('../../_chunks/dep-3c2d5e8c.js');
require('../../_chunks/dep-f0e884b9.js');
require('../../loading/plugin.js');
require('../../_util/react-render.js');
require('../../_chunks/dep-e37cb830.js');
require('../../common/PluginContainer.js');
require('../../button/defaultProps.js');
require('../../_util/parseTNode.js');
require('../../_chunks/dep-b289b0be.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var DatePickerFooter = function DatePickerFooter(props) {
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("datePicker"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var confirmText = t(local.confirm);
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var enableTimePicker = props.enableTimePicker,
    onConfirmClick = props.onConfirmClick,
    _props$presetsPlaceme = props.presetsPlacement,
    presetsPlacement = _props$presetsPlaceme === void 0 ? "bottom" : _props$presetsPlaceme,
    presets = props.presets,
    onPresetClick = props.onPresetClick,
    selectedValue = props.selectedValue,
    needConfirm = props.needConfirm,
    onTimePanelChange = props.onTimePanelChange,
    isSwitchTimeMode = props.isSwitchTimeMode;
  var footerClass = classNames__default["default"]("".concat(classPrefix, "-date-picker__footer"), "".concat(classPrefix, "-date-picker__footer--").concat(presetsPlacement));
  var renderPresets = function renderPresets() {
    if (presets) {
      if (/*#__PURE__*/React__default["default"].isValidElement(presets)) return presets;
      if (isFunction.isFunction(presets)) return presets();
      return Object.keys(presets).map(function (key) {
        return /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
          key: key,
          size: "small",
          variant: "text",
          onClick: function onClick(e) {
            return onPresetClick(presets[key], {
              e: e,
              preset: defineProperty._defineProperty({}, key, presets[key])
            });
          }
        }, key);
      });
    }
    return null;
  };
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: footerClass
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-date-picker__presets")
  }, renderPresets()), /* @__PURE__ */React__default["default"].createElement("div", null, isSwitchTimeMode && /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    style: {
      marginRight: 16
    },
    size: "small",
    theme: "primary",
    variant: "text",
    onClick: onTimePanelChange
  }, props.isDateRangeContent ? t(local.selectTime) : t(local.selectDate)), enableTimePicker && needConfirm && /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    disabled: !selectedValue,
    size: "small",
    theme: "primary",
    onClick: function onClick(e) {
      return onConfirmClick({
        e: e
      });
    }
  }, confirmText)));
};
DatePickerFooter.displayName = "DatePickerFooter";

exports["default"] = DatePickerFooter;
