// tdesign-react/cjs/date-picker/base/Header.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var toConsumableArray = require('../../_chunks/dep-1d0dbfc5.js');
var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var React = require('react');
var hooks_useConfig = require('../../hooks/useConfig.js');
var locale_LocalReceiver = require('../../locale/LocalReceiver.js');
var pagination_index = require('../../pagination/index.js');
var select_index = require('../../select/index.js');
var datePicker_hooks_useSelectRange = require('../hooks/useSelectRange.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_chunks/dep-3c2d5e8c.js');
require('../../_chunks/dep-32593c54.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-2b5ba78d.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-14013769.js');
require('../../_chunks/dep-06bf4625.js');
require('dayjs');
require('../../_chunks/dep-043eee20.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../_chunks/dep-2d5cf7d6.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../config-provider/ConfigProvider.js');
require('../../_chunks/dep-4a562e04.js');
require('../../_chunks/dep-a5b4810f.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-ea361d31.js');
require('../../pagination/Pagination.js');
require('../../_chunks/dep-1c586f5c.js');
require('classnames');
require('../../_util/noop.js');
require('../../hooks/useControlled.js');
require('../../_chunks/dep-eb1b216c.js');
require('../../_chunks/dep-b529a0d7.js');
require('../../_chunks/dep-bcd16957.js');
require('../../_chunks/dep-b4442734.js');
require('../../_chunks/dep-876d090e.js');
require('../../input-number/index.js');
require('../../input-number/InputNumber.js');
require('tdesign-icons-react');
require('../../input/index.js');
require('../../input/Input.js');
require('../../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../../_util/parseTNode.js');
require('../../_chunks/dep-b289b0be.js');
require('../../hooks/useDefaultProps.js');
require('../../hooks/useGlobalIcon.js');
require('../../hooks/useLayoutEffect.js');
require('../../_chunks/dep-68f5fe7b.js');
require('../../input/defaultProps.js');
require('../../input/InputGroup.js');
require('../../input/useLengthLimit.js');
require('../../_chunks/dep-b011b5d3.js');
require('../../_chunks/dep-f0e884b9.js');
require('../../button/index.js');
require('../../button/Button.js');
require('../../hooks/useDomRefCallback.js');
require('../../hooks/useRipple.js');
require('../../_chunks/dep-064e4dba.js');
require('../../hooks/useAnimation.js');
require('../../loading/index.js');
require('../../loading/Loading.js');
require('../../common/Portal.js');
require('react-dom');
require('../../loading/defaultProps.js');
require('../../loading/gradient.js');
require('../../loading/plugin.js');
require('../../_util/react-render.js');
require('../../_chunks/dep-e37cb830.js');
require('../../common/PluginContainer.js');
require('../../button/defaultProps.js');
require('../../input-number/useInputNumber.js');
require('../../_chunks/dep-04deac0d.js');
require('../../_chunks/dep-8f62fb96.js');
require('../../hooks/useCommonClassName.js');
require('../../input-number/defaultProps.js');
require('../../input-adornment/InputAdornment.js');
require('../../pagination/hooks/useBoundaryJumper.js');
require('../../pagination/hooks/usePrevNextJumper.js');
require('../../pagination/hooks/usePageNumber.js');
require('../../pagination/hooks/useTotal.js');
require('../../pagination/validators.js');
require('../../pagination/defaultProps.js');
require('../../_chunks/dep-5c3139cb.js');
require('../../_chunks/dep-2e455d91.js');
require('../../_chunks/dep-5e17a073.js');
require('../../_chunks/dep-894a5ece.js');
require('../../_chunks/dep-63308b84.js');
require('../../_chunks/dep-048e4314.js');
require('../../select/base/Select.js');
require('../../_util/helper.js');
require('../../_chunks/dep-e44b03ec.js');
require('../../_util/ref.js');
require('react-is');
require('../../_util/isFragment.js');
require('../../common/FakeArrow.js');
require('../../select-input/index.js');
require('../../select-input/SelectInput.js');
require('../../popup/index.js');
require('../../popup/Popup.js');
require('react-transition-group');
require('../../hooks/useAttach.js');
require('../../hooks/useMutationObserver.js');
require('../../hooks/useLatest.js');
require('../../_chunks/dep-4e876c4b.js');
require('../../_chunks/dep-df8d2fc8.js');
require('../../_chunks/dep-66f91e5d.js');
require('../../_chunks/dep-d3af8eb7.js');
require('../../_chunks/dep-eab994e8.js');
require('../../hooks/usePopper.js');
require('react-fast-compare');
require('@popperjs/core');
require('../../hooks/useWindowSize.js');
require('../../popup/defaultProps.js');
require('../../popup/hooks/useTrigger.js');
require('../../_util/listener.js');
require('../../popup/utils/transition.js');
require('../../popup/PopupPlugin.js');
require('../../select-input/defaultProps.js');
require('../../select-input/useMultiple.js');
require('../../tag-input/index.js');
require('../../tag-input/TagInput.js');
require('../../hooks/useDragSorter.js');
require('../../hooks/useEventCallback.js');
require('../../tag-input/defaultProps.js');
require('../../tag-input/useHover.js');
require('../../tag-input/useTagList.js');
require('../../tag/index.js');
require('../../tag/Tag.js');
require('../../_chunks/dep-d14bfcfa.js');
require('../../tag/defaultProps.js');
require('../../tag/CheckTag.js');
require('../../_chunks/dep-dd0f5f6b.js');
require('../../tag/CheckTagGroup.js');
require('../../tag-input/useTagScroll.js');
require('../../select-input/useOverlayInnerStyle.js');
require('../../hooks/useInnerPopupVisible.js');
require('../../select-input/useSingle.js');
require('../../_chunks/dep-f7bf3d17.js');
require('../../_chunks/dep-ca918872.js');
require('../../_chunks/dep-13de2fef.js');
require('../../select/defaultProps.js');
require('../../select/hooks/useKeyboardControl.js');
require('../../_chunks/dep-bdc97672.js');
require('../../select/base/OptionGroup.js');
require('../../_chunks/dep-05df1ea3.js');
require('../../select/base/PopupContent.js');
require('../../select/hooks/usePanelVirtualScroll.js');
require('../../hooks/useVirtualScroll.js');
require('../../pagination/PaginationMini.js');
require('../../_chunks/dep-8227db70.js');
require('../../_chunks/dep-c990dbb4.js');
require('../../_chunks/dep-bf18bb11.js');
require('../../_chunks/dep-06ac19a1.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var useDatePickerLocalConfig = function useDatePickerLocalConfig() {
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("datePicker"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  return {
    months: t(local.months),
    nextYear: t(local.nextYear),
    preYear: t(local.preYear),
    nextMonth: t(local.nextMonth),
    preMonth: t(local.preMonth),
    preDecade: t(local.preDecade),
    nextDecade: t(local.nextDecade),
    now: t(local.now)
  };
};
var DatePickerHeader = function DatePickerHeader(props) {
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var mode = props.mode,
    year = props.year,
    month = props.month,
    range = props.range,
    onMonthChange = props.onMonthChange,
    onYearChange = props.onYearChange,
    onJumperClick = props.onJumperClick,
    partial = props.partial,
    _props$internalYear = props.internalYear,
    internalYear = _props$internalYear === void 0 ? [] : _props$internalYear;
  var _useDatePickerLocalCo = useDatePickerLocalConfig(),
    now = _useDatePickerLocalCo.now,
    months = _useDatePickerLocalCo.months,
    preMonth = _useDatePickerLocalCo.preMonth,
    preYear = _useDatePickerLocalCo.preYear,
    nextMonth = _useDatePickerLocalCo.nextMonth,
    nextYear = _useDatePickerLocalCo.nextYear,
    preDecade = _useDatePickerLocalCo.preDecade,
    nextDecade = _useDatePickerLocalCo.nextDecade;
  var scrollAnchorRef = React.useRef("default");
  var _useSelectRange = datePicker_hooks_useSelectRange.useSelectRange({
      range: range,
      mode: mode,
      year: year,
      month: month
    }),
    paginationDisabled = _useSelectRange.paginationDisabled,
    monthHasAnyAllowed = _useSelectRange.monthHasAnyAllowed,
    yearHasAnyAllowed = _useSelectRange.yearHasAnyAllowed,
    decadeHasAnyAllowed = _useSelectRange.decadeHasAnyAllowed,
    canLoadMoreTop = _useSelectRange.canLoadMoreTop,
    canLoadMoreBottom = _useSelectRange.canLoadMoreBottom;
  var initOptions = React.useCallback(function (year2) {
    var options = [];
    if (mode === "year") {
      var extraYear = year2 % 10;
      var minYear = year2 - extraYear - 100;
      var maxYear = year2 - extraYear + 100;
      for (var i = minYear; i <= maxYear; i += 10) {
        var end = i + 9;
        if (decadeHasAnyAllowed(end)) {
          options.push({
            label: "".concat(i, " - ").concat(end),
            value: i + 9,
            disabled: false
          });
        }
      }
    } else {
      yearHasAnyAllowed(year2) && options.push({
        label: "".concat(year2),
        value: year2,
        disabled: false
      });
      for (var _i = 1; _i <= 10; _i++) {
        yearHasAnyAllowed(year2 + _i) && options.push({
          label: "".concat(year2 + _i),
          value: year2 + _i,
          disabled: false
        });
        yearHasAnyAllowed(year2 - _i) && options.unshift({
          label: "".concat(year2 - _i),
          value: year2 - _i,
          disabled: false
        });
      }
    }
    return options;
  }, [mode]);
  var _useState = React.useState(function () {
      return initOptions(year);
    }),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    yearOptions = _useState2[0],
    setYearOptions = _useState2[1];
  var nearestYear = React.useMemo(function () {
    var _yearOptions$find;
    var extraYear = partial === "end" && mode === "year" && internalYear[1] - internalYear[0] <= 9 ? 9 : 0;
    return ((_yearOptions$find = yearOptions.find(function (option) {
      return option.value - (year + extraYear) <= 9 && option.value - (year + extraYear) >= 0;
    })) === null || _yearOptions$find === void 0 ? void 0 : _yearOptions$find.value) || year;
  }, [yearOptions, year, mode, partial, internalYear]);
  var monthOptions = months.map(function (item, index) {
    return {
      label: item,
      value: index,
      disabled: !monthHasAnyAllowed(year, index)
    };
  });
  var showPanelTop = React.useMemo(function () {
    var options = yearOptions;
    if (!options.length) return false;
    var first = options[0].value;
    return canLoadMoreTop(first);
  }, [canLoadMoreTop, yearOptions]);
  var showPanelBottom = React.useMemo(function () {
    var options = yearOptions;
    if (!options.length) return false;
    var last = options[options.length - 1].value;
    return canLoadMoreBottom(last);
  }, [canLoadMoreBottom, yearOptions]);
  function loadMoreYear(year2, type) {
    var options = [];
    if (mode === "year") {
      var extraYear = year2 % 10;
      if (type === "add") {
        for (var i = year2 - extraYear + 10; i <= year2 - extraYear + 50; i += 10) {
          var end = i + 9;
          decadeHasAnyAllowed(end) && options.push({
            label: "".concat(i, " - ").concat(end),
            value: i,
            disabled: false
          });
        }
      } else {
        for (var _i2 = year2 - extraYear - 1; _i2 > year2 - extraYear - 50; _i2 -= 10) {
          decadeHasAnyAllowed(_i2) && options.unshift({
            label: "".concat(_i2 - 9, " - ").concat(_i2),
            value: _i2,
            disabled: false
          });
        }
      }
    } else if (type === "add") {
      for (var _i3 = year2 + 1; _i3 <= year2 + 10; _i3++) {
        yearHasAnyAllowed(_i3) && options.push({
          label: "".concat(_i3),
          value: _i3,
          disabled: false
        });
      }
    } else {
      for (var _i4 = year2 - 1; _i4 > year2 - 10; _i4--) {
        yearHasAnyAllowed(_i4) && options.unshift({
          label: "".concat(_i4),
          value: _i4,
          disabled: false
        });
      }
    }
    return options;
  }
  var labelMap = {
    year: {
      prev: preDecade,
      current: now,
      next: nextDecade
    },
    month: {
      prev: preYear,
      current: now,
      next: nextYear
    },
    date: {
      prev: preMonth,
      current: now,
      next: nextMonth
    }
  };
  var headerClassName = "".concat(classPrefix, "-date-picker__header");
  var showMonthPicker = mode === "date" || mode === "week";
  function handlePanelTopClick(e) {
    var _e$stopPropagation, _e$nativeEvent, _e$nativeEvent$stopIm;
    e === null || e === void 0 || (_e$stopPropagation = e.stopPropagation) === null || _e$stopPropagation === void 0 || _e$stopPropagation.call(e);
    e === null || e === void 0 || (_e$nativeEvent = e.nativeEvent) === null || _e$nativeEvent === void 0 || (_e$nativeEvent$stopIm = _e$nativeEvent.stopImmediatePropagation) === null || _e$nativeEvent$stopIm === void 0 || _e$nativeEvent$stopIm.call(_e$nativeEvent);
    var firstYear = yearOptions[0].value;
    var options = loadMoreYear(firstYear, "reduce");
    setYearOptions([].concat(toConsumableArray._toConsumableArray(options), toConsumableArray._toConsumableArray(yearOptions)));
  }
  function handlePanelBottomClick(e) {
    var _e$stopPropagation2, _e$nativeEvent2;
    e === null || e === void 0 || (_e$stopPropagation2 = e.stopPropagation) === null || _e$stopPropagation2 === void 0 || _e$stopPropagation2.call(e);
    e === null || e === void 0 || (_e$nativeEvent2 = e.nativeEvent) === null || _e$nativeEvent2 === void 0 || _e$nativeEvent2.stopImmediatePropagation();
    var lastYear = yearOptions.slice(-1)[0].value;
    var options = loadMoreYear(lastYear, "add");
    setYearOptions([].concat(toConsumableArray._toConsumableArray(yearOptions), toConsumableArray._toConsumableArray(options)));
  }
  function handleScroll(_ref) {
    var e = _ref.e;
    if (e.target.scrollTop === 0) {
      showPanelTop && handlePanelTopClick();
      scrollAnchorRef.current = "top";
    } else if (e.target.scrollTop === e.target.scrollHeight - e.target.clientHeight) {
      showPanelBottom && handlePanelBottomClick();
      scrollAnchorRef.current = "bottom";
    }
  }
  function handleUpdateScrollTop(content) {
    if (scrollAnchorRef.current === "top") {
      content.scrollTop = 30 * 10;
    } else if (scrollAnchorRef.current === "bottom") {
      content.scrollTop = content.scrollHeight - 30 * 10;
    } else {
      var firstSelectedNode = content === null || content === void 0 ? void 0 : content.querySelector(".".concat(classPrefix, "-is-selected"));
      if (firstSelectedNode) {
        var _getComputedStyle = getComputedStyle(firstSelectedNode),
          paddingBottom = _getComputedStyle.paddingBottom;
        var _getComputedStyle2 = getComputedStyle(content),
          marginBottom = _getComputedStyle2.marginBottom;
        var elementBottomHeight = parseInt(paddingBottom, 10) + parseInt(marginBottom, 10);
        var updateValue = firstSelectedNode.offsetTop - content.offsetTop - (content.clientHeight - firstSelectedNode.clientHeight) + elementBottomHeight;
        content.scrollTop = updateValue;
      }
    }
  }
  React.useEffect(function () {
    var yearRange = initOptions(year);
    setYearOptions(yearRange);
  }, [initOptions, year]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: headerClassName
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(headerClassName, "-controller")
  }, showMonthPicker && /* @__PURE__ */React__default["default"].createElement(select_index.Select, {
    className: "".concat(headerClassName, "-controller-month"),
    value: month,
    options: monthOptions,
    onChange: function onChange(val) {
      return onMonthChange(val);
    },
    popupProps: {
      attach: function attach(triggerElement) {
        return triggerElement.parentNode;
      },
      overlayClassName: "".concat(headerClassName, "-controller-month-popup")
    }
  }), /* @__PURE__ */React__default["default"].createElement(select_index.Select, {
    className: "".concat(headerClassName, "-controller-year"),
    value: mode === "year" ? nearestYear : year,
    options: yearOptions,
    onChange: function onChange(val) {
      return onYearChange(val);
    },
    onPopupVisibleChange: function onPopupVisibleChange(visible) {
      if (!visible) scrollAnchorRef.current = "default";
    },
    popupProps: {
      onScroll: handleScroll,
      updateScrollTop: function updateScrollTop(el) {
        setTimeout(function () {
          handleUpdateScrollTop(el);
        }, 0);
      },
      attach: function attach(triggerElement) {
        return triggerElement.parentNode;
      },
      overlayClassName: "".concat(headerClassName, "-controller-year-popup")
    },
    panelTopContent: showPanelTop && /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-select-option"),
      onClick: handlePanelTopClick
    }, "..."),
    panelBottomContent: showPanelBottom && /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-select-option"),
      onClick: handlePanelBottomClick
    }, "...")
  })), /* @__PURE__ */React__default["default"].createElement(pagination_index.PaginationMini, {
    tips: labelMap[mode],
    size: "small",
    disabled: paginationDisabled,
    onChange: onJumperClick
  }));
};
DatePickerHeader.displayName = "DatePickerHeader";
var DateHeader = /*#__PURE__*/React__default["default"].memo(DatePickerHeader);

exports["default"] = DateHeader;
