// tdesign-react/cjs/date-picker/base/Table.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-2b5ba78d.js');
var slicedToArray = require('../../_chunks/dep-1a3dddaf.js');
var React = require('react');
var classNames = require('classnames');
var dayjs = require('dayjs');
var _commonjsHelpers = require('../../_chunks/dep-06bf4625.js');
var format = require('../../_chunks/dep-06ac19a1.js');
var locale_LocalReceiver = require('../../locale/LocalReceiver.js');
var hooks_useConfig = require('../../hooks/useConfig.js');
var datePicker_base_Cell = require('./Cell.js');
require('../../_chunks/dep-c4d1ce0c.js');
require('../../_chunks/dep-32593c54.js');
require('../../_chunks/dep-51fc57b3.js');
require('../../_chunks/dep-bf18bb11.js');
require('../../_chunks/dep-b289b0be.js');
require('../../_chunks/dep-17e2e72f.js');
require('../../_chunks/dep-5f51e3c1.js');
require('../../_chunks/dep-794b42ce.js');
require('../../_chunks/dep-ecfe656e.js');
require('../../_chunks/dep-aff91ab0.js');
require('../../config-provider/ConfigProvider.js');
require('../../config-provider/ConfigContext.js');
require('../../_chunks/dep-14013769.js');
require('../../_chunks/dep-043eee20.js');
require('../../_chunks/dep-e51f8066.js');
require('../../_chunks/dep-67a947f0.js');
require('../../_chunks/dep-cd29b2b2.js');
require('../../_chunks/dep-ed46627b.js');
require('../../_chunks/dep-746faf11.js');
require('../../_chunks/dep-6abdd014.js');
require('../../_chunks/dep-6d96dce8.js');
require('../../_chunks/dep-977a0080.js');
require('../../_chunks/dep-cba78412.js');
require('../../_chunks/dep-41c55905.js');
require('../../_chunks/dep-b21b548e.js');
require('../../_chunks/dep-69c18aee.js');
require('../../_chunks/dep-83fd21a7.js');
require('../../_chunks/dep-0bed591e.js');
require('../../_chunks/dep-5f500f41.js');
require('../../_chunks/dep-7577d7c1.js');
require('../../_chunks/dep-2d5cf7d6.js');
require('../../_chunks/dep-4a562e04.js');
require('../../_chunks/dep-a5b4810f.js');
require('../../_chunks/dep-de8b5dd9.js');
require('../../_chunks/dep-fc111b3d.js');
require('../../_chunks/dep-ea361d31.js');
require('../../_chunks/dep-8227db70.js');
require('../../_chunks/dep-c990dbb4.js');
require('../../_chunks/dep-eab994e8.js');
require('../../_chunks/dep-876d090e.js');
require('../../_chunks/dep-bcd16957.js');
require('../../_util/parseTNode.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

var isoWeek$1 = {exports: {}};

(function (module, exports) {
  !function (e, t) {
    module.exports = t() ;
  }(_commonjsHelpers.commonjsGlobal, function () {

    var e = "day";
    return function (t, i, s) {
      var a = function a(t) {
          return t.add(4 - t.isoWeekday(), e);
        },
        d = i.prototype;
      d.isoWeekYear = function () {
        return a(this).year();
      }, d.isoWeek = function (t) {
        if (!this.$utils().u(t)) return this.add(7 * (t - this.isoWeek()), e);
        var i,
          d,
          n,
          o,
          r = a(this),
          u = (i = this.isoWeekYear(), d = this.$u, n = (d ? s.utc : s)().year(i).startOf("year"), o = 4 - n.isoWeekday(), n.isoWeekday() > 4 && (o += 7), n.add(o, e));
        return r.diff(u, "week") + 1;
      }, d.isoWeekday = function (e) {
        return this.$utils().u(e) ? this.day() || 7 : this.day(this.day() % 7 ? e : e - 7);
      };
      var n = d.startOf;
      d.startOf = function (e, t) {
        var i = this.$utils(),
          s = !!i.u(t) || t;
        return "isoweek" === i.p(e) ? s ? this.date(this.date() - (this.isoWeekday() - 1)).startOf("day") : this.date(this.date() - 1 - (this.isoWeekday() - 1) + 7).endOf("day") : n.bind(this)(e, t);
      };
    };
  });
})(isoWeek$1);
var isoWeek = isoWeek$1.exports;

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
dayjs__default["default"].extend(isoWeek);
var DatePickerTable = function DatePickerTable(props) {
  var _useConfig = hooks_useConfig["default"](),
    classPrefix = _useConfig.classPrefix;
  var value = props.value,
    format$1 = props.format,
    mode = props.mode,
    data = props.data,
    time = props.time,
    cell = props.cell,
    onCellClick = props.onCellClick,
    onCellMouseEnter = props.onCellMouseEnter,
    onCellMouseLeave = props.onCellMouseLeave,
    firstDayOfWeek = props.firstDayOfWeek;
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("datePicker"),
    _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
    local = _useLocaleReceiver2[0],
    t = _useLocaleReceiver2[1];
  var weekdays = t(local.weekdays);
  var weekAbbreviation = t(local.weekAbbreviation);
  var weekArr = [];
  var wi = firstDayOfWeek - 1;
  var len = weekdays.length;
  while (weekArr.length < len) {
    weekArr.push(weekdays[wi]);
    wi = (wi + len + 1) % len;
  }
  if (mode === "week") weekArr.unshift(weekAbbreviation);
  var showThead = mode === "date" || mode === "week";
  var valueYearWeek = React.useMemo(function () {
    if (mode !== "week" || !value) return {};
    if (Array.isArray(value)) {
      var _startObj$year, _startObj$locale, _startObj$locale$week, _endObj$year, _endObj$locale, _endObj$locale$week;
      if (!value.length) return {};
      var _value$map = value.map(function (v) {
          return v && format.parseToDayjs(v, format$1);
        }),
        _value$map2 = slicedToArray._slicedToArray(_value$map, 2),
        startObj = _value$map2[0],
        endObj = _value$map2[1];
      var startYear = startObj === null || startObj === void 0 || (_startObj$year = startObj.year) === null || _startObj$year === void 0 ? void 0 : _startObj$year.call(startObj);
      var startWeek = startObj === null || startObj === void 0 || (_startObj$locale = startObj.locale) === null || _startObj$locale === void 0 || (_startObj$locale = _startObj$locale.call(startObj, local.dayjsLocale)) === null || _startObj$locale === void 0 || (_startObj$locale$week = _startObj$locale.week) === null || _startObj$locale$week === void 0 ? void 0 : _startObj$locale$week.call(_startObj$locale);
      var endYear = endObj === null || endObj === void 0 || (_endObj$year = endObj.year) === null || _endObj$year === void 0 ? void 0 : _endObj$year.call(endObj);
      var endWeek = endObj === null || endObj === void 0 || (_endObj$locale = endObj.locale) === null || _endObj$locale === void 0 || (_endObj$locale = _endObj$locale.call(endObj, local.dayjsLocale)) === null || _endObj$locale === void 0 || (_endObj$locale$week = _endObj$locale.week) === null || _endObj$locale$week === void 0 ? void 0 : _endObj$locale$week.call(_endObj$locale);
      return {
        startYear: startYear,
        startWeek: startWeek,
        endYear: endYear,
        endWeek: endWeek
      };
    }
    var valueObj = format.parseToDayjs(value, format$1).locale(local.dayjsLocale);
    return {
      year: valueObj.year(),
      week: valueObj.week()
    };
  }, [mode, value, format$1]);
  var weekRowClass = function weekRowClass(value2, targetDayjs) {
    if (mode !== "week" || !value2) return {};
    if (Array.isArray(value2)) {
      if (!value2.length) return {};
      var targetYear = targetDayjs.isoWeekYear();
      var targetWeek = targetDayjs.isoWeek();
      var isActive = targetYear === valueYearWeek.startYear && targetWeek === valueYearWeek.startWeek || targetYear === valueYearWeek.endYear && targetWeek === valueYearWeek.endWeek;
      var isRange = (targetYear > valueYearWeek.startYear || targetYear === valueYearWeek.startYear && targetWeek > valueYearWeek.startWeek) && (targetYear < valueYearWeek.endYear || targetYear === valueYearWeek.endYear && targetWeek < valueYearWeek.endWeek);
      return defineProperty._defineProperty(defineProperty._defineProperty({}, "".concat(classPrefix, "-date-picker__table-").concat(mode, "-row--active"), isActive), "".concat(classPrefix, "-date-picker__table-").concat(mode, "-row--range"), isRange);
    }
    return defineProperty._defineProperty({}, "".concat(classPrefix, "-date-picker__table-").concat(mode, "-row--active"), valueYearWeek.year === targetDayjs.year() && valueYearWeek.week === targetDayjs.week());
  };
  var multipleWeekRowClass = function multipleWeekRowClass(value2, targetDayjs) {
    if (mode !== "week" || Array.isArray(value2) && !value2.length) return {};
    var isSomeYearWeek = value2.map(function (v) {
      return format.parseToDayjs(v, format$1);
    }).some(function (item) {
      return item.isoWeek() === targetDayjs.isoWeek() && item.isoWeekYear() === targetDayjs.isoWeekYear();
    });
    return defineProperty._defineProperty({}, "".concat(classPrefix, "-date-picker__table-").concat(mode, "-row--active"), isSomeYearWeek);
  };
  var activeRowCss = props.multiple ? multipleWeekRowClass : weekRowClass;
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(classPrefix, "-date-picker__table"),
    onMouseLeave: function onMouseLeave(e) {
      return onCellMouseLeave === null || onCellMouseLeave === void 0 ? void 0 : onCellMouseLeave({
        e: e
      });
    }
  }, /* @__PURE__ */React__default["default"].createElement("table", null, showThead && /* @__PURE__ */React__default["default"].createElement("thead", null, /* @__PURE__ */React__default["default"].createElement("tr", null, weekArr.map(function (value2, i) {
    return /* @__PURE__ */React__default["default"].createElement("th", {
      key: i
    }, value2);
  }))), /* @__PURE__ */React__default["default"].createElement("tbody", null, data.map(function (row, i) {
    return /* @__PURE__ */React__default["default"].createElement("tr", {
      key: i,
      className: classNames__default["default"]("".concat(classPrefix, "-date-picker__table-").concat(mode, "-row"), _objectSpread({}, activeRowCss(value, row[0].dayjsObj)))
    }, row.map(function (col, j) {
      return /* @__PURE__ */React__default["default"].createElement(datePicker_base_Cell["default"], _objectSpread(_objectSpread({}, col), {}, {
        key: j,
        time: time,
        cell: cell,
        onClick: onCellClick,
        onMouseEnter: onCellMouseEnter
      }));
    }));
  }))));
};
DatePickerTable.displayName = "DatePickerTable";

exports["default"] = DatePickerTable;
