// tdesign-react/cjs/hooks/useMutationObserver.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-2b5ba78d.js');
var React = require('react');
var hooks_useLatest = require('./useLatest.js');
var isEqual = require('../_chunks/dep-4e876c4b.js');
var debounce = require('../_chunks/dep-d3af8eb7.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-0bed591e.js');
require('../_chunks/dep-eab994e8.js');
require('../_chunks/dep-876d090e.js');

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var DEFAULT_OPTIONS = {
  debounceTime: 0,
  config: {
    attributes: true,
    childList: true,
    characterData: true,
    subtree: true
  }
};
function useMutationObservable(targetEl, cb, options) {
  var _options$debounceTime;
  var optionsRef = React.useRef(null);
  var signalRef = React.useRef(0);
  var callbackRef = hooks_useLatest["default"](cb);
  var mergedOptions = {
    debounceTime: (_options$debounceTime = options === null || options === void 0 ? void 0 : options.debounceTime) !== null && _options$debounceTime !== void 0 ? _options$debounceTime : DEFAULT_OPTIONS.debounceTime,
    config: _objectSpread(_objectSpread({}, DEFAULT_OPTIONS.config), options === null || options === void 0 ? void 0 : options.config)
  };
  if (!isEqual.isEqual(mergedOptions, optionsRef.current)) {
    signalRef.current += 1;
  }
  optionsRef.current = mergedOptions;
  React.useEffect(function () {
    if (!targetEl || !(targetEl !== null && targetEl !== void 0 && targetEl.nodeType)) return;
    var observer = null;
    try {
      var _optionsRef$current = optionsRef.current,
        debounceTime = _optionsRef$current.debounceTime,
        config = _optionsRef$current.config;
      var mutationCallback = function mutationCallback() {
        callbackRef.current.apply(callbackRef, arguments);
      };
      observer = new MutationObserver(debounceTime > 0 ? debounce.debounce(mutationCallback, debounceTime) : mutationCallback);
      observer.observe(targetEl, config);
    } catch (e) {
      console.error(e);
    }
    return function () {
      if (observer) {
        observer.disconnect();
        observer = null;
      }
    };
  }, [targetEl, signalRef.current]);
}

exports["default"] = useMutationObservable;
