// tdesign-react/cjs/hooks/useVirtualScroll.js
/**
 * tdesign v1.16.9
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../_chunks/dep-1a3dddaf.js');
var React = require('react');
var isEqual = require('../_chunks/dep-4e876c4b.js');
require('../_chunks/dep-32593c54.js');
require('../_chunks/dep-51fc57b3.js');
require('../_chunks/dep-df8d2fc8.js');
require('../_chunks/dep-e51f8066.js');
require('../_chunks/dep-67a947f0.js');
require('../_chunks/dep-5f51e3c1.js');
require('../_chunks/dep-c4d1ce0c.js');
require('../_chunks/dep-ecfe656e.js');
require('../_chunks/dep-794b42ce.js');
require('../_chunks/dep-cd29b2b2.js');
require('../_chunks/dep-ed46627b.js');
require('../_chunks/dep-746faf11.js');
require('../_chunks/dep-6abdd014.js');
require('../_chunks/dep-6d96dce8.js');
require('../_chunks/dep-977a0080.js');
require('../_chunks/dep-cba78412.js');
require('../_chunks/dep-aff91ab0.js');
require('../_chunks/dep-41c55905.js');
require('../_chunks/dep-66f91e5d.js');
require('../_chunks/dep-de8b5dd9.js');
require('../_chunks/dep-fc111b3d.js');
require('../_chunks/dep-ea361d31.js');
require('../_chunks/dep-0bed591e.js');

var requestAnimationFrame = (typeof window === "undefined" ? false : window.requestAnimationFrame) || function (cb) {
  return setTimeout(cb, 16.6);
};
var useVirtualScroll = function useVirtualScroll(container, params) {
  var _params$enable = params.enable,
    enable = _params$enable === void 0 ? true : _params$enable,
    data = params.data,
    scroll = params.scroll;
  var dataRef = React.useRef(data);
  var containerHeight = React.useRef(0);
  var _useState = React.useState([]),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    visibleData = _useState2[0],
    setVisibleData = _useState2[1];
  var _useState3 = React.useState(0),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    translateY = _useState4[0],
    setTranslateY = _useState4[1];
  var _useState5 = React.useState(0),
    _useState6 = slicedToArray._slicedToArray(_useState5, 2),
    scrollHeight = _useState6[0],
    setScrollHeight = _useState6[1];
  var trScrollTopHeightList = React.useRef([]);
  var _useState7 = React.useState([]),
    _useState8 = slicedToArray._slicedToArray(_useState7, 2),
    trHeightList = _useState8[0],
    setTrHeightList = _useState8[1];
  var tScroll = React.useMemo(function () {
    var _scroll$isFixedRowHei, _scroll$fixedRows;
    if (!scroll) return {};
    return {
      bufferSize: scroll.bufferSize || 10,
      isFixedRowHeight: (_scroll$isFixedRowHei = scroll.isFixedRowHeight) !== null && _scroll$isFixedRowHei !== void 0 ? _scroll$isFixedRowHei : false,
      rowHeight: scroll.rowHeight || 47,
      threshold: scroll.threshold || 100,
      type: scroll.type,
      fixedRows: (_scroll$fixedRows = scroll.fixedRows) !== null && _scroll$fixedRows !== void 0 ? _scroll$fixedRows : [0, 0]
    };
  }, [scroll]);
  var enableVirtualScroll = React.useMemo(function () {
    return enable && tScroll.type === "virtual";
  }, [enable, tScroll]);
  var isVirtualScroll = React.useMemo(function () {
    return enableVirtualScroll && tScroll.threshold < data.length;
  }, [enableVirtualScroll, tScroll.threshold, data.length]);
  var visibleCount = Math.min(tScroll.bufferSize * 3, data.length);
  var _useState9 = React.useState(function () {
      return [0, visibleCount];
    }),
    _useState0 = slicedToArray._slicedToArray(_useState9, 2),
    startAndEndIndex = _useState0[0],
    setStartAndEndIndex = _useState0[1];
  var getTrScrollTopHeightList = function getTrScrollTopHeightList(trHeightList2) {
    var list = [];
    for (var i = 0, len = data.length; i < len; i++) {
      list[i] = (list[i - 1] || 0) + (trHeightList2[i] || tScroll.rowHeight);
    }
    return list;
  };
  var updateVisibleData = function updateVisibleData(trScrollTopHeightList2, scrollTop) {
    var _data$, _data;
    var currentIndex = -1;
    for (var i = 0, len = trScrollTopHeightList2.length; i < len; i++) {
      if (trScrollTopHeightList2[i] >= scrollTop) {
        currentIndex = i;
        break;
      }
    }
    var lastIndex = trScrollTopHeightList2.length;
    var containerCurrentHeight = containerHeight.current || container.current.getBoundingClientRect().height;
    var scrollBottom = scrollTop + containerCurrentHeight;
    for (var _i = currentIndex, _len = trScrollTopHeightList2.length; _i < _len; _i++) {
      if (trScrollTopHeightList2[_i] >= scrollBottom) {
        lastIndex = _i;
        break;
      }
    }
    if (currentIndex < 0) return;
    var startIndex = Math.max(currentIndex - tScroll.bufferSize, 0);
    var endIndex = Math.min(lastIndex + tScroll.bufferSize, trScrollTopHeightList2.length);
    var fixedRows = tScroll.fixedRows;
    var _fixedRows = slicedToArray._slicedToArray(fixedRows, 2),
      fixedStart = _fixedRows[0],
      fixedEnd = _fixedRows[1];
    var firstIsFake = (_data$ = data[0]) === null || _data$ === void 0 ? void 0 : _data$.__VIRTUAL_FAKE_DATA;
    var isScrollStart = lastIndex - visibleCount <= 0;
    var startOffset = firstIsFake && !isScrollStart ? 1 : 0;
    var fixedStartData = fixedStart ? data.slice(startOffset, fixedStart + startOffset) : [];
    if (fixedStart && startIndex < fixedStart) {
      fixedStartData = fixedStartData.slice(0, startIndex);
    }
    var lastIsFake = (_data = data[data.length - 1]) === null || _data === void 0 ? void 0 : _data.__VIRTUAL_FAKE_DATA;
    var isScrollEnd = lastIndex === trScrollTopHeightList2.length;
    var endOffset = lastIsFake && !isScrollEnd ? 1 : 0;
    var fixedEndData = fixedEnd ? data.slice(data.length - fixedEnd - endOffset) : [];
    var bottomStartIndex = endIndex - data.length + 1 + (fixedEnd !== null && fixedEnd !== void 0 ? fixedEnd : 0);
    if (fixedEnd && bottomStartIndex > 0) {
      fixedEndData = fixedEndData.slice(bottomStartIndex);
    }
    if (startAndEndIndex.join() !== [startIndex, endIndex].join() && startIndex >= 0) {
      var tmpVisibleData = fixedStartData.concat(data.slice(startIndex, endIndex)).concat(fixedEndData);
      setVisibleData(tmpVisibleData);
      var lastScrollTop = trScrollTopHeightList2[startIndex - 1];
      var top = lastScrollTop > 0 ? lastScrollTop : 0;
      var stickyHeight = trScrollTopHeightList2[Math.min(startIndex, fixedStart) - 1] || 0;
      setTranslateY(top - stickyHeight);
      setStartAndEndIndex([startIndex, endIndex]);
    }
  };
  var handleRowMounted = function handleRowMounted(rowData) {
    if (!isVirtualScroll || !rowData || tScroll.isFixedRowHeight || !(container !== null && container !== void 0 && container.current)) return;
    var trHeight = rowData.ref.offsetHeight;
    var rowIndex = rowData.data.__VIRTUAL_SCROLL_INDEX;
    var newTrHeightList = trHeightList;
    if (newTrHeightList[rowIndex] !== trHeight) {
      newTrHeightList[rowIndex] = trHeight;
      setTrHeightList(newTrHeightList);
      var scrollTopHeightList = getTrScrollTopHeightList(newTrHeightList);
      trScrollTopHeightList.current = scrollTopHeightList;
      var lastIndex = scrollTopHeightList.length - 1;
      setScrollHeight(scrollTopHeightList[lastIndex] - containerHeight.current);
      updateVisibleData(scrollTopHeightList, container.current.scrollTop);
    }
  };
  var handleScroll = function handleScroll() {
    if (!isVirtualScroll) return;
    updateVisibleData(trScrollTopHeightList.current, container.current.scrollTop);
  };
  var addIndexToData = function addIndexToData(data2) {
    data2.forEach(function (item, index) {
      Reflect.set(item, "__VIRTUAL_SCROLL_INDEX", index);
    });
  };
  var updateScrollTop = function updateScrollTop(_ref) {
    var _container$current;
    var index = _ref.index,
      _ref$top = _ref.top,
      top = _ref$top === void 0 ? 0 : _ref$top,
      behavior = _ref.behavior;
    var scrollTop = trScrollTopHeightList.current[index] - top;
    (_container$current = container.current) === null || _container$current === void 0 || _container$current.scrollTo({
      top: scrollTop,
      behavior: behavior || "auto"
    });
  };
  var scrollToElement = function scrollToElement(p) {
    updateScrollTop(p);
    if (!tScroll.isFixedRowHeight) {
      requestAnimationFrame(function () {
        var _p$time;
        var duration = (_p$time = p.time) !== null && _p$time !== void 0 ? _p$time : 60;
        var timer = setTimeout(function () {
          updateScrollTop(p);
          clearTimeout(timer);
        }, duration);
      });
    }
  };
  React.useEffect(function () {
    if (!enableVirtualScroll) return;
    if (!isVirtualScroll) {
      if (!data.length) return;
      setVisibleData(data);
      dataRef.current = data;
      trScrollTopHeightList.current = getTrScrollTopHeightList(trHeightList);
      return;
    }
    addIndexToData(data);
    var dataChanged = !isEqual.isEqual(dataRef.current, data);
    var scrollTopHeightList = trScrollTopHeightList.current;
    if ((scrollTopHeightList === null || scrollTopHeightList === void 0 ? void 0 : scrollTopHeightList.length) === (data === null || data === void 0 ? void 0 : data.length) && !dataChanged) {
      var lastIndex = scrollTopHeightList.length - 1;
      setScrollHeight(scrollTopHeightList[lastIndex]);
      updateVisibleData(scrollTopHeightList, container.current.scrollTop);
    } else {
      var _container$current2;
      dataRef.current = data;
      setScrollHeight(data.length * tScroll.rowHeight);
      var currentScrollTop = ((_container$current2 = container.current) === null || _container$current2 === void 0 ? void 0 : _container$current2.scrollTop) || 0;
      var currentIndex = Math.floor(currentScrollTop / tScroll.rowHeight);
      for (var i = 0; i < (scrollTopHeightList === null || scrollTopHeightList === void 0 ? void 0 : scrollTopHeightList.length); i++) {
        if (scrollTopHeightList[i] >= currentScrollTop) {
          currentIndex = i;
          break;
        }
      }
      var startIndex = Math.max(currentIndex - tScroll.bufferSize, 0);
      var endIndex = Math.min(startIndex + visibleCount, data.length);
      var tmpData = data.slice(startIndex, endIndex);
      var translateY2 = startIndex * tScroll.rowHeight;
      if ((scrollTopHeightList === null || scrollTopHeightList === void 0 ? void 0 : scrollTopHeightList.length) > 0 && startIndex > 0) {
        var prevHeight = scrollTopHeightList[Math.min(startIndex - 1, scrollTopHeightList.length - 1)] || 0;
        translateY2 = Math.max(0, prevHeight);
      }
      setVisibleData(tmpData);
      setTranslateY(translateY2);
    }
    var timer = setTimeout(function () {
      if (container.current) {
        var tmpContainerHeight = container.current.getBoundingClientRect().height;
        containerHeight.current = tmpContainerHeight;
        var scrollTopHeightList2 = getTrScrollTopHeightList(trHeightList);
        trScrollTopHeightList.current = scrollTopHeightList2;
        clearTimeout(timer);
      }
    }, 1);
  }, [container, data, tScroll, isVirtualScroll, startAndEndIndex, trHeightList]);
  return {
    visibleData: visibleData,
    translateY: translateY,
    scrollHeight: scrollHeight,
    isVirtualScroll: isVirtualScroll,
    handleScroll: handleScroll,
    handleRowMounted: handleRowMounted,
    scrollToElement: scrollToElement
  };
};

exports["default"] = useVirtualScroll;
